{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-13 20:47:46.557220\n",
      "Analysis started on 2024-08-13 at  2024-08-13 20:47:46.557220\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/13 20:47:53 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-13 20:47:53.623034\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading sources\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:KeyboardInterrupt while sending command.                (0 + 8) / 98]\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\", line 1038, in send_command\n",
      "    response = connection.send_command(command)\n",
      "  File \"/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/clientserver.py\", line 511, in send_command\n",
      "    answer = smart_decode(self.stream.readline()[:-1])\n",
      "  File \"/opt/conda/miniconda3/lib/python3.10/socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_34878/3260239266.py\u001b[0m in \u001b[0;36m<cell line: 703>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmetric_list\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    704\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"max_L2GScore\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 705\u001b[0;31m         datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n\u001b[0m\u001b[1;32m    706\u001b[0m             \u001b[0mdiscrepancifier\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0motGenetics\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    707\u001b[0m         )\n",
      "\u001b[0;32m/tmp/ipykernel_34878/3260239266.py\u001b[0m in \u001b[0;36mbenchmarkOT\u001b[0;34m(discrepancifier, otGenetics, metric)\u001b[0m\n\u001b[1;32m    525\u001b[0m                 \u001b[0;34mf\"{value}\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    526\u001b[0m             )  ##### modifications here to include the groups of ranking/distances to TSS\n\u001b[0;32m--> 527\u001b[0;31m             \u001b[0;34m.\u001b[0m\u001b[0mpivot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"homogenized\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    528\u001b[0m             \u001b[0;34m.\u001b[0m\u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"targetId\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    529\u001b[0m         )\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/group.py\u001b[0m in \u001b[0;36mpivot\u001b[0;34m(self, pivot_col, values)\u001b[0m\n\u001b[1;32m    275\u001b[0m         \"\"\"\n\u001b[1;32m    276\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mvalues\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 277\u001b[0;31m             \u001b[0mjgd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jgd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpivot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpivot_col\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    278\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m             \u001b[0mjgd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jgd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpivot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpivot_col\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1318\u001b[0m             \u001b[0mproto\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEND_COMMAND_PART\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1319\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1320\u001b[0;31m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1321\u001b[0m         return_value = get_return_value(\n\u001b[1;32m   1322\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36msend_command\u001b[0;34m(self, command, retry, binary)\u001b[0m\n\u001b[1;32m   1036\u001b[0m         \u001b[0mconnection\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_connection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1037\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1038\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconnection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1039\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbinary\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1040\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_create_connection_guard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconnection\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/clientserver.py\u001b[0m in \u001b[0;36msend_command\u001b[0;34m(self, command)\u001b[0m\n\u001b[1;32m    509\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    510\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 511\u001b[0;31m                 \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msmart_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    512\u001b[0m                 \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Answer received: {0}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0manswer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    513\u001b[0m                 \u001b[0;31m# Happens when a the other end is dead. There might be an empty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/miniconda3/lib/python3.10/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    704\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 705\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    706\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    707\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#####\n",
    "#######\n",
    "## ANALYSIS FOR L2G Scores, genetic evidence and Direction of Effect\n",
    "## Original, propagated and Other Vs Oncology\n",
    "#######\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "### Now , filter by rank, and join with the info from Ot genetics and run the DoE.\n",
    "ranking = Window.partitionBy(\"studyId\", \"variantId\")\n",
    "### union with the other datasources\n",
    "platform_v = \"24.06\"\n",
    "\n",
    "target_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/targets/\"\n",
    ")\n",
    "target = spark.read.parquet(target_path)\n",
    "\n",
    "disease_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/diseases/\"\n",
    ")\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "mecact_path = f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/mechanismOfAction/\"\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "evidences = spark.read.parquet(\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/evidence\"\n",
    ").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "# 1# Make a list of variant of interest (Sequence ontology terms) to subset data of interest.\n",
    "### Bear in mind that SO works with ontology structure as: SO:XXXXXX, but databases has the SO as: SO_XXXXXX\n",
    "var_filter_lof = [\n",
    "    ### High impact variants https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html\n",
    "    \"SO_0001589\",  ## frameshit_variant\n",
    "    \"SO_0001587\",  ## stop_gained\n",
    "    \"SO_0001574\",  ## splice_acceptor_variant\n",
    "    \"SO_0001575\",  ## splice_donor_variant\n",
    "    \"SO_0002012\",  ## start_lost\n",
    "    \"SO_0001578\",  ## stop_lost\n",
    "    \"SO_0001893\",  ## transcript_ablation\n",
    "    # \"SO:0001889\", ## transcript_amplification ## the Only HIGH impact that increase protein.\n",
    "]\n",
    "\n",
    "gof = [\"SO_0002053\"]\n",
    "lof = [\"SO_0002054\"]\n",
    "\n",
    "print(\"loading sources\")\n",
    "\n",
    "## Building Sequence Ontology\n",
    "so_path = \"gs://ot-team/jroldan/sequenceOntology_20221118.csv\"\n",
    "so_ontology = spark.read.csv(so_path, header=True)\n",
    "building = (\n",
    "    so_ontology.select(F.col(\"Accession\"), F.col(\"Parents\"))\n",
    "    .withColumn(\"Parentalind\", F.split(F.col(\"Parents\"), \",\"))\n",
    "    .withColumn(\"Parentalind\", F.explode_outer(\"Parentalind\"))\n",
    "    .groupBy(\"Parentalind\")\n",
    "    .agg(F.collect_list(F.col(\"Accession\")).alias(\"childrens\"))\n",
    "    .join(so_ontology, F.col(\"Parentalind\") == so_ontology.Accession, \"right\")\n",
    ")\n",
    "## annotate TSG/oncogene/bivalent using 'hallmarks.attributes'\n",
    "oncotsg_list = [\n",
    "    \"TSG\",\n",
    "    \"oncogene\",\n",
    "    \"Oncogene\",\n",
    "    \"oncogene\",\n",
    "    \"oncogene,TSG\",\n",
    "    \"TSG,oncogene\",\n",
    "    \"fusion,oncogene\",\n",
    "    \"oncogene,fusion\",\n",
    "]\n",
    "\n",
    "#### rlike('('+Keywords+')(\\s|$)'\n",
    "### on 03.07.2023 we add the categories:\n",
    "# DISRUPTING AGENT - inhibitor\n",
    "# STABILISER - activator\n",
    "\n",
    "### Hacer el join del actionType con el chembl para sacar los mecanismos de accion.\n",
    "inhibitors = [\n",
    "    \"RNAI INHIBITOR\",\n",
    "    \"NEGATIVE MODULATOR\",\n",
    "    \"NEGATIVE ALLOSTERIC MODULATOR\",\n",
    "    \"ANTAGONIST\",\n",
    "    \"ANTISENSE INHIBITOR\",\n",
    "    \"BLOCKER\",\n",
    "    \"INHIBITOR\",\n",
    "    \"DEGRADER\",\n",
    "    \"INVERSE AGONIST\",\n",
    "    \"ALLOSTERIC ANTAGONIST\",\n",
    "    \"DISRUPTING AGENT\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "activators = [\n",
    "    \"PARTIAL AGONIST\",\n",
    "    \"ACTIVATOR\",\n",
    "    \"POSITIVE ALLOSTERIC MODULATOR\",\n",
    "    \"POSITIVE MODULATOR\",\n",
    "    \"AGONIST\",\n",
    "    \"SEQUESTERING AGENT\",\n",
    "    \"STABILISER\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "columnas = [\"activator\", \"inhibitor\"]\n",
    "both = activators + inhibitors\n",
    "\n",
    "actionType = (\n",
    "    mecact.select(\n",
    "        F.explode_outer(\"chemblIds\").alias(\"drugId2\"),\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "        \"targets\",\n",
    "    )\n",
    "    .select(\n",
    "        F.explode_outer(\"targets\").alias(\"targetId2\"),\n",
    "        \"drugId2\",\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "    )\n",
    "    .groupBy(\"targetId2\", \"drugId2\")\n",
    "    .agg(\n",
    "        F.collect_set(\"actionType\").alias(\"actionType\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "oncolabel = (\n",
    "    target.select(\n",
    "        \"id\", \"approvedSymbol\", F.explode_outer(F.col(\"hallmarks.attributes\"))\n",
    "    )\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.description\")\n",
    "    .filter(F.col(\"description\").isin(oncotsg_list))\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"description\").alias(\"description\"))\n",
    "    .withColumn(\"description_splited\", F.concat_ws(\",\", F.col(\"description\")))\n",
    "    .withColumn(\n",
    "        \"TSorOncogene\",\n",
    "        F.when(\n",
    "            (\n",
    "                F.col(\"description_splited\").rlike(\"ncogene\")\n",
    "                & F.col(\"description_splited\").rlike(\"TSG\")\n",
    "            ),\n",
    "            F.lit(\"bivalent\"),\n",
    "        )\n",
    "        .when(F.col(\"description_splited\").rlike(\"ncogene(\\s|$)\"), F.lit(\"oncogene\"))\n",
    "        .when(F.col(\"description_splited\").rlike(\"TSG(\\s|$)\"), F.lit(\"TSG\"))\n",
    "        .otherwise(F.lit(\"noEvaluable\")),  ####\n",
    "    )\n",
    "    .withColumnRenamed(\"id\", \"target_id\")\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "\n",
    "windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "sincgc = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "germline = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"chembl\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "#### version all gene burden\n",
    "prueba_assessment = (\n",
    "    directionOfEffect(evidences, platform_v)\n",
    "    .withColumn(\n",
    "        \"rank\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"average\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "genEvidDataset = (\n",
    "    prueba_assessment.filter(F.col(\"datasourceId\") != \"chembl\")  #### checked 31.05.2023\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.count(\"targetId\").alias(\"Nr_evidences\"))\n",
    "    .select(\"targetId\", \"diseaseId\", \"Nr_evidences\")\n",
    "    .withColumn(\"geneticEvidence\", F.lit(\"hasGeneticEvidence\"))\n",
    ")\n",
    "\n",
    "coherency_toAssess_others_datasource = (  #### checked 31.05.2023\n",
    "    prueba_assessment.filter(\n",
    "        (F.col(\"homogenized\").isin(columns)) & (F.col(\"datasourceId\") != \"chembl\")\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"datasourceId\").alias(\"datasourceIds\"))\n",
    ")\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "v2g = spark.read.parquet(\"gs://genetics-portal-dev-data/22.09.1/outputs/v2g\")\n",
    "varDistToGene = v2g.select(\n",
    "    F.concat_ws(\"_\", \"chr_id\", \"position\", \"ref_allele\", \"alt_allele\").alias(\n",
    "        \"variantId\"\n",
    "    ),\n",
    "    F.col(\"gene_id\").alias(\"targetId\"),\n",
    "    \"source_id\",\n",
    "    \"d\",\n",
    "    \"distance_score\",\n",
    ").filter(F.col(\"source_id\") == \"canonical_tss\")\n",
    "\n",
    "ranking = Window.partitionBy(\"studyId\", \"variantId\")\n",
    "\n",
    "\n",
    "#######\n",
    "# Build Ot genetics dataset as supporting evidence\n",
    "#######\n",
    "otGenetics = (\n",
    "    prueba_assessment.filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    # .filter((F.col(\"homogenized\") != \"noEvaluable\"))\n",
    "    .join(varDistToGene, on=[\"variantId\", \"targetId\"], how=\"left\")\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.collect_set(\"datasourceId\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"L2G_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageL2G\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageCanonicalTSSDistance\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"d\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.when(\n",
    "            F.col(\"rank\").isNull(),\n",
    "            F.array_remove(F.col(\"datasources\"), \"ot_genetics_portal\"),\n",
    "        ).otherwise(F.col(\"datasources\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"distance_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"d\").asc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"ChemblL2gRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"L2G_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"chemblDistanceRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"distance_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"frontierValue\",\n",
    "        ## ot genetics portal\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"beta\") <= 0.1) & (F.col(\"beta\") >= -0.1),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"OddsRatio\") <= 1.1) & (F.col(\"OddsRatio\") >= 0.9),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.lit(\"noValue\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "#####\n",
    "# function for interpreting DoE and coherencies/discrepancies\n",
    "#####\n",
    "\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "analysis_chembl = discrepancifier(\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\") == \"chembl\"))\n",
    "    .withColumn(\n",
    "        \"maxClinPhase\",\n",
    "        F.max(F.col(\"clinicalPhase\")).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "    .pivot(\"homogenized\")\n",
    "    .agg(F.count(\"targetId\"))\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### propag OtGenetics:\n",
    "otGenetics_propag = (\n",
    "    otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "    .join(\n",
    "        diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "        on=\"diseaseId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "    .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    ").persist()\n",
    "\n",
    "\n",
    "#### include dictionary for calling dataframes:\n",
    "# max_L2GScore\n",
    "# min_distance_ranking\n",
    "\n",
    "\n",
    "def benchmarkOT(discrepancifier, otGenetics, metric):\n",
    "    dict_comb = {}\n",
    "    dict_comb = {\n",
    "        \"hasDirectionOfEffect\": f\"{metric}\",\n",
    "        \"diagonalYes\": f\"{metric}\",\n",
    "        \"oneCellYes\": f\"{metric}\",\n",
    "    }\n",
    "    list_l2g = [\n",
    "        0.1,\n",
    "        0.15,\n",
    "        0.2,\n",
    "        0.25,\n",
    "        0.3,\n",
    "        0.35,\n",
    "        0.4,\n",
    "        0.45,\n",
    "        0.5,\n",
    "        0.55,\n",
    "        0.6,\n",
    "        0.65,\n",
    "        0.7,\n",
    "        0.75,\n",
    "        0.8,\n",
    "        0.85,\n",
    "        0.9,\n",
    "        0.95,\n",
    "    ]\n",
    "    list_dist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
    "    return (\n",
    "        discrepancifier(\n",
    "            otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "            .withColumn(\n",
    "                \"min_distance_ranking\",\n",
    "                F.min(\"distance_ranking\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(  ### take maximum L2G score per T-D\n",
    "                \"max_L2GScore\",\n",
    "                F.max(\"resourceScore\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .groupBy(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                f\"{value}\",\n",
    "            )  ##### modifications here to include the groups of ranking/distances to TSS\n",
    "            .pivot(\"homogenized\")\n",
    "            .agg(F.count(\"targetId\"))\n",
    "        )\n",
    "        .selectExpr(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "            f\"{metric}\",\n",
    "            \"coherencyDiagonal as coherencyDiagonal\",\n",
    "            \"coherencyOneCell as coherencyOneCell\",\n",
    "            \"LoF_protect\",\n",
    "            \"GoF_protect\",\n",
    "            \"LoF_risk\",\n",
    "            \"GoF_risk\",\n",
    "        )\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(f\"{metric}\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(  #### new column to modify it\n",
    "            \"hasdirectionOfEffect\",\n",
    "            F.when(F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"yes\")).otherwise(\n",
    "                F.lit(\"no\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .select(\n",
    "            [\"*\"]\n",
    "            + (\n",
    "                [  ### single columns\n",
    "                    F.when(F.col(f\"{metric}\") >= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}>={str(n).replace('.', '_')}\")\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"  # Adjust this condition as needed\n",
    "                else [\n",
    "                    F.when(F.col(f\"{metric}\") <= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}<={n}\")\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "            + (\n",
    "                [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"\n",
    "                else [\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) <= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}<={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "metric_list = [\"max_L2GScore\", \"min_distance_ranking\"]\n",
    "datasetDict = {}\n",
    "for value in metric_list:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_l2g_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )\n",
    "    elif value == \"min_distance_ranking\":\n",
    "        datasetDict[f\"df_distance_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_distance_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df(test_propag) -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    toAnalysis = test_propag.drop(\n",
    "        \"Phase4\",\n",
    "        \"Phase>=3\",\n",
    "        \"Phase>=2\",\n",
    "        \"Phase>=1\",\n",
    "        \"Phase0\",\n",
    "        \"clinicalStatus\",\n",
    "        \"prediction\",\n",
    "        \"count\",\n",
    "        \"PhaseT\",\n",
    "        \"taLabelSimple\",\n",
    "    ).columns[17:]\n",
    "    dataType = [\"byDatatype\"] * len(toAnalysis)\n",
    "    l_studies = []\n",
    "    l_studies.extend([list(a) for a in zip(toAnalysis, dataType)])\n",
    "\n",
    "    schema = StructType(\n",
    "        [\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "            StructField(\"comparisonType\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    comparisons = spark.createDataFrame(l_studies, schema=schema)\n",
    "    ### include all the columns as predictor\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "    wComparison = Window.partitionBy(F.col(comparisonColumn))\n",
    "    wPrediction = Window.partitionBy(F.col(predictionColumn))\n",
    "    wPredictionComparison = Window.partitionBy(\n",
    "        F.col(comparisonColumn), F.col(predictionColumn)\n",
    "    )\n",
    "    \"\"\"\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "    \n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    \"\"\"\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "\n",
    "\n",
    "c = datetime.now()\n",
    "\n",
    "print(\"start doing aggregations and writing\")\n",
    "today_date = str(date.today())\n",
    "listado = []\n",
    "#### Run analysis using loop\n",
    "for key, df in datasetDict.items():\n",
    "    print(key)\n",
    "    df = df.persist()\n",
    "    aggSetups_original = comparisons_df(df)\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "##### read files and make spreadsheet\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio\n",
    "\n",
    "\n",
    "def convertTuple(tup):\n",
    "    st = \",\".join(map(str, tup))\n",
    "    return st\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "#### update the dictionary dfs with other columns included in the analysis\n",
    "key_list = [\n",
    "    \"hasGeneticEvidence\",\n",
    "    \"oneCell\",\n",
    "    \"diagonal\",\n",
    "    \"hasDirectionOfEffect\",\n",
    "    \"max_L2GScore\",\n",
    "    \"min_distance_ranking\",\n",
    "]\n",
    "value_list = [\n",
    "    \"geneticEvidence\",\n",
    "    \"oneCellDoE\",\n",
    "    \"diagonalDoE\",\n",
    "    \"hasDirectionOfEffect\",\n",
    "    \"L2GScore\",\n",
    "    \"TSSDistance\",\n",
    "]\n",
    "\n",
    "dfs = {}\n",
    "\n",
    "\n",
    "def create_dict_column(dfs, key_list, value_list):\n",
    "    if len(key_list) != len(value_list):\n",
    "        raise ValueError(\"lists of different length\")\n",
    "    dfs.update(zip(key_list, value_list))\n",
    "    return dfs\n",
    "\n",
    "\n",
    "dfs = create_dict_column(dfs, key_list, value_list)\n",
    "\n",
    "# Define the lists of possible substrings\n",
    "phase_opt = [\n",
    "    \"Phase4\",\n",
    "    \"Phase>=3\",\n",
    "    \"Phase>=2\",\n",
    "    \"Phase>=1\",\n",
    "    \"PhaseT\",\n",
    "]\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "results = []\n",
    "\n",
    "# Iterate over the sample strings and extract the desired substrings\n",
    "for path in listado:\n",
    "    array1 = np.delete(\n",
    "        (\n",
    "            spark.read.parquet(path)\n",
    "            .join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "            .groupBy(\"comparison\")\n",
    "            .pivot(\"prediction\")\n",
    "            .agg(F.first(\"a\"))\n",
    "            .sort(F.col(\"comparison\").desc())\n",
    "            .select(\"comparison\", \"yes\", \"no\")\n",
    "            .fillna(0)\n",
    "            .toPandas()\n",
    "            .to_numpy()\n",
    "        ),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    # Check for comparison options\n",
    "    for key, value in dfs.items():\n",
    "        if key in path and \"combined\" in path:\n",
    "            if \"oneCell\" in path:\n",
    "                comparison = dfs[key]\n",
    "                group = \"combinedOneCell\"\n",
    "            elif \"diagonal\" in path:\n",
    "                comparison = dfs[key]\n",
    "                group = \"combinedDiagonal\"\n",
    "            elif \"Direction\" in path:\n",
    "                comparison = dfs[key]\n",
    "                group = \"combinedHasDoE\"\n",
    "        elif key in path:\n",
    "            comparison = dfs[key]\n",
    "            group = key\n",
    "\n",
    "    # Check for phase options\n",
    "    for substr in phase_opt:\n",
    "        if substr in path:\n",
    "            phase = substr\n",
    "\n",
    "    if \"original\" in path:\n",
    "        dimension = \"original\"\n",
    "    elif \"propag\" in path:\n",
    "        dimension = \"propagated\"\n",
    "    if \"l2g\" in path:\n",
    "        dataset = \"l2gScore\"\n",
    "    elif \"distance\" in path:\n",
    "        dataset = \"TSSdistance\"\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            group,\n",
    "            dataset,\n",
    "            comparison,\n",
    "            dimension,\n",
    "            phase,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            res_npPhaseX,\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            path,\n",
    "        ]\n",
    "    )\n",
    "    print(path)\n",
    "# Convert the results to a pandas DataFrame\n",
    "df = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"group\",\n",
    "        \"dataset\",\n",
    "        \"comparison\",\n",
    "        \"dimension\",\n",
    "        \"phase\",\n",
    "        \"oddsRatio\",\n",
    "        \"pValue\",\n",
    "        \"lowerInterval\",\n",
    "        \"upperInterval\",\n",
    "        \"total\",\n",
    "        \"values\",\n",
    "        \"relSuccess\",\n",
    "        \"rsLower\",\n",
    "        \"rsUpper\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")\n",
    "df = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"group\",\n",
    "        \"dataset\",\n",
    "        \"comparison\",\n",
    "        \"dimension\",\n",
    "        \"phase\",\n",
    "        \"oddsRatio\",\n",
    "        \"pValue\",\n",
    "        \"lowerInterval\",\n",
    "        \"upperInterval\",\n",
    "        \"total\",\n",
    "        \"values\",\n",
    "        \"relSuccess\",\n",
    "        \"rsLower\",\n",
    "        \"rsUpper\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-13 20:52:22.664204\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "#####\n",
    "#######\n",
    "## ANALYSIS FOR L2G Scores, genetic evidence and Direction of Effect\n",
    "## Original, propagated and Other Vs Oncology\n",
    "#######\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### combined list:\n",
    "\n",
    "list_l2g = [\n",
    "    0.1,\n",
    "    0.15,\n",
    "    0.2,\n",
    "    0.25,\n",
    "    0.3,\n",
    "    0.35,\n",
    "    0.4,\n",
    "    0.45,\n",
    "    0.5,\n",
    "    0.55,\n",
    "    0.6,\n",
    "    0.65,\n",
    "    0.7,\n",
    "    0.75,\n",
    "    0.8,\n",
    "    0.85,\n",
    "    0.9,\n",
    "    0.95,\n",
    "]\n",
    "list_dist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.1\n",
      "0.15\n",
      "0.2\n",
      "0.25\n",
      "0.3\n",
      "0.35\n",
      "0.4\n",
      "0.45\n",
      "0.5\n",
      "0.55\n",
      "0.6\n",
      "0.65\n",
      "0.7\n",
      "0.75\n",
      "0.8\n",
      "0.85\n",
      "0.9\n",
      "0.95\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "for l in list_l2g:\n",
    "    print(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start doing aggregations and writing\n",
      "df_l2g_original\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/09 12:33:24 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-09 12:33:24.590569\n",
      "2024-08-09 12:33:24.600035\n",
      "2024-08-09 12:33:24.608987\n",
      "2024-08-09 12:33:24.618684\n",
      "2024-08-09 12:33:24.628586\n",
      "2024-08-09 12:33:24.636323\n",
      "2024-08-09 12:33:24.644112\n",
      "2024-08-09 12:33:24.652456\n",
      "2024-08-09 12:33:24.661620\n",
      "2024-08-09 12:33:24.669845\n",
      "2024-08-09 12:33:24.678966\n",
      "2024-08-09 12:33:24.687293\n",
      "2024-08-09 12:33:24.694914\n",
      "2024-08-09 12:33:24.703302\n",
      "2024-08-09 12:33:24.711203\n",
      "2024-08-09 12:33:24.718911\n",
      "2024-08-09 12:33:24.726355\n",
      "2024-08-09 12:33:24.733893\n",
      "2024-08-09 12:33:24.742051\n",
      "2024-08-09 12:33:24.754343\n",
      "2024-08-09 12:33:24.762562\n",
      "2024-08-09 12:33:24.770371\n",
      "2024-08-09 12:33:24.779529\n",
      "2024-08-09 12:33:24.789073\n",
      "2024-08-09 12:33:24.798847\n",
      "2024-08-09 12:33:24.807340\n",
      "2024-08-09 12:33:24.814584\n",
      "2024-08-09 12:33:24.821711\n",
      "2024-08-09 12:33:24.828721\n",
      "2024-08-09 12:33:24.835681\n",
      "2024-08-09 12:33:24.842667\n",
      "2024-08-09 12:33:24.850163\n",
      "2024-08-09 12:33:24.857628\n",
      "2024-08-09 12:33:24.865090\n",
      "2024-08-09 12:33:24.872801\n",
      "2024-08-09 12:33:24.880072\n",
      "2024-08-09 12:33:24.888156\n",
      "2024-08-09 12:33:24.896080\n",
      "2024-08-09 12:33:24.903778\n",
      "2024-08-09 12:33:24.912140\n",
      "2024-08-09 12:33:24.920424\n",
      "2024-08-09 12:33:24.928979\n",
      "2024-08-09 12:33:24.936941\n",
      "2024-08-09 12:33:24.943995\n",
      "2024-08-09 12:33:24.951359\n",
      "2024-08-09 12:33:24.959828\n",
      "2024-08-09 12:33:24.969787\n",
      "2024-08-09 12:33:24.978618\n",
      "2024-08-09 12:33:24.986189\n",
      "2024-08-09 12:33:24.995469\n",
      "2024-08-09 12:33:25.003774\n",
      "2024-08-09 12:33:25.011008\n",
      "2024-08-09 12:33:25.019082\n",
      "2024-08-09 12:33:25.027758\n",
      "2024-08-09 12:33:25.038049\n",
      "2024-08-09 12:33:25.049715\n",
      "2024-08-09 12:33:25.060430\n",
      "2024-08-09 12:33:25.071785\n",
      "2024-08-09 12:33:25.082164\n",
      "2024-08-09 12:33:25.091564\n",
      "2024-08-09 12:33:25.100175\n",
      "2024-08-09 12:33:25.109357\n",
      "2024-08-09 12:33:25.118826\n",
      "2024-08-09 12:33:25.127233\n",
      "2024-08-09 12:33:25.135129\n",
      "2024-08-09 12:33:25.145004\n",
      "2024-08-09 12:33:25.153676\n",
      "2024-08-09 12:33:25.163897\n",
      "2024-08-09 12:33:25.171935\n",
      "2024-08-09 12:33:25.179915\n",
      "2024-08-09 12:33:25.188899\n",
      "2024-08-09 12:33:25.198032\n",
      "2024-08-09 12:33:25.206944\n",
      "2024-08-09 12:33:25.214201\n",
      "2024-08-09 12:33:25.221305\n",
      "2024-08-09 12:33:25.232403\n",
      "2024-08-09 12:33:25.241494\n",
      "2024-08-09 12:33:25.249727\n",
      "2024-08-09 12:33:25.257454\n",
      "2024-08-09 12:33:25.266031\n",
      "2024-08-09 12:33:25.275295\n",
      "2024-08-09 12:33:25.291133\n",
      "2024-08-09 12:33:25.301213\n",
      "2024-08-09 12:33:25.311122\n",
      "2024-08-09 12:33:25.319958\n",
      "2024-08-09 12:33:25.328959\n",
      "2024-08-09 12:33:25.337051\n",
      "2024-08-09 12:33:25.344838\n",
      "2024-08-09 12:33:25.352323\n",
      "2024-08-09 12:33:25.360442\n",
      "2024-08-09 12:33:25.368510\n",
      "2024-08-09 12:33:25.376300\n",
      "2024-08-09 12:33:25.384736\n",
      "2024-08-09 12:33:25.392304\n",
      "2024-08-09 12:33:25.401180\n",
      "2024-08-09 12:33:25.408849\n",
      "2024-08-09 12:33:25.417045\n",
      "2024-08-09 12:33:25.424736\n",
      "2024-08-09 12:33:25.432596\n",
      "2024-08-09 12:33:25.439910\n",
      "2024-08-09 12:33:25.447677\n",
      "2024-08-09 12:33:25.454917\n",
      "2024-08-09 12:33:25.465438\n",
      "2024-08-09 12:33:25.476873\n",
      "2024-08-09 12:33:25.485185\n",
      "2024-08-09 12:33:25.492314\n",
      "2024-08-09 12:33:25.499555\n",
      "2024-08-09 12:33:25.506755\n",
      "2024-08-09 12:33:25.513895\n",
      "2024-08-09 12:33:25.521398\n",
      "2024-08-09 12:33:25.528196\n",
      "2024-08-09 12:33:25.536010\n",
      "2024-08-09 12:33:25.544503\n",
      "2024-08-09 12:33:25.552231\n",
      "2024-08-09 12:33:25.560205\n",
      "2024-08-09 12:33:25.568406\n",
      "2024-08-09 12:33:25.576195\n",
      "2024-08-09 12:33:25.584570\n",
      "2024-08-09 12:33:25.591966\n",
      "2024-08-09 12:33:25.600561\n",
      "2024-08-09 12:33:25.610720\n",
      "2024-08-09 12:33:25.620049\n",
      "2024-08-09 12:33:25.628646\n",
      "2024-08-09 12:33:25.636571\n",
      "2024-08-09 12:33:25.645314\n",
      "2024-08-09 12:33:25.652919\n",
      "2024-08-09 12:33:25.660638\n",
      "2024-08-09 12:33:25.668015\n",
      "2024-08-09 12:33:25.675955\n",
      "2024-08-09 12:33:25.684018\n",
      "2024-08-09 12:33:25.693075\n",
      "2024-08-09 12:33:25.701450\n",
      "2024-08-09 12:33:25.711233\n",
      "2024-08-09 12:33:25.719263\n",
      "2024-08-09 12:33:25.726873\n",
      "2024-08-09 12:33:25.735764\n",
      "2024-08-09 12:33:25.743948\n",
      "2024-08-09 12:33:25.752337\n",
      "2024-08-09 12:33:25.761376\n",
      "2024-08-09 12:33:25.771939\n",
      "2024-08-09 12:33:25.779735\n",
      "2024-08-09 12:33:25.788288\n",
      "2024-08-09 12:33:25.796291\n",
      "2024-08-09 12:33:25.803770\n",
      "2024-08-09 12:33:25.812452\n",
      "2024-08-09 12:33:25.824024\n",
      "2024-08-09 12:33:25.832005\n",
      "2024-08-09 12:33:25.839579\n",
      "2024-08-09 12:33:25.847690\n",
      "2024-08-09 12:33:25.855946\n",
      "2024-08-09 12:33:25.862866\n",
      "2024-08-09 12:33:25.870227\n",
      "2024-08-09 12:33:25.877561\n",
      "2024-08-09 12:33:25.884573\n",
      "2024-08-09 12:33:25.891776\n",
      "2024-08-09 12:33:25.899394\n",
      "2024-08-09 12:33:25.906457\n",
      "2024-08-09 12:33:25.914067\n",
      "2024-08-09 12:33:25.921384\n",
      "2024-08-09 12:33:25.930761\n",
      "2024-08-09 12:33:25.939690\n",
      "2024-08-09 12:33:25.947046\n",
      "2024-08-09 12:33:25.955196\n",
      "2024-08-09 12:33:25.962392\n",
      "2024-08-09 12:33:25.970086\n",
      "2024-08-09 12:33:25.977422\n",
      "2024-08-09 12:33:25.985118\n",
      "2024-08-09 12:33:25.993326\n",
      "2024-08-09 12:33:26.000129\n",
      "2024-08-09 12:33:26.007032\n",
      "2024-08-09 12:33:26.014411\n",
      "2024-08-09 12:33:26.022395\n",
      "2024-08-09 12:33:26.030018\n",
      "2024-08-09 12:33:26.038344\n",
      "2024-08-09 12:33:26.046928\n",
      "2024-08-09 12:33:26.057212\n",
      "2024-08-09 12:33:26.064969\n",
      "2024-08-09 12:33:26.072923\n",
      "2024-08-09 12:33:26.080205\n",
      "2024-08-09 12:33:26.089131\n",
      "2024-08-09 12:33:26.097779\n",
      "2024-08-09 12:33:26.106927\n",
      "2024-08-09 12:33:26.116649\n",
      "2024-08-09 12:33:26.125819\n",
      "2024-08-09 12:33:26.133669\n",
      "2024-08-09 12:33:26.141920\n",
      "2024-08-09 12:33:26.149923\n",
      "2024-08-09 12:33:26.159609\n",
      "2024-08-09 12:33:26.169205\n",
      "2024-08-09 12:33:26.178345\n",
      "2024-08-09 12:33:26.187950\n",
      "2024-08-09 12:33:26.200172\n",
      "2024-08-09 12:33:26.210732\n",
      "2024-08-09 12:33:26.220500\n",
      "2024-08-09 12:33:26.230551\n",
      "2024-08-09 12:33:26.238925\n",
      "2024-08-09 12:33:26.248809\n",
      "2024-08-09 12:33:26.257272\n",
      "2024-08-09 12:33:26.265138\n",
      "2024-08-09 12:33:26.277394\n",
      "2024-08-09 12:33:26.287764\n",
      "2024-08-09 12:33:26.299099\n",
      "2024-08-09 12:33:26.310222\n",
      "2024-08-09 12:33:26.319735\n",
      "2024-08-09 12:33:26.328720\n",
      "2024-08-09 12:33:26.337142\n",
      "2024-08-09 12:33:26.344564\n",
      "2024-08-09 12:33:26.351660\n",
      "2024-08-09 12:33:26.359544\n",
      "2024-08-09 12:33:26.368944\n",
      "2024-08-09 12:33:26.377096\n",
      "2024-08-09 12:33:26.385472\n",
      "2024-08-09 12:33:26.397726\n",
      "2024-08-09 12:33:26.405493\n",
      "2024-08-09 12:33:26.414923\n",
      "2024-08-09 12:33:26.423282\n",
      "2024-08-09 12:33:26.430820\n",
      "2024-08-09 12:33:26.438713\n",
      "2024-08-09 12:33:26.445750\n",
      "2024-08-09 12:33:26.455320\n",
      "2024-08-09 12:33:26.464284\n",
      "2024-08-09 12:33:26.473284\n",
      "2024-08-09 12:33:26.480593\n",
      "2024-08-09 12:33:26.488415\n",
      "2024-08-09 12:33:26.497437\n",
      "2024-08-09 12:33:26.507883\n",
      "2024-08-09 12:33:26.517492\n",
      "2024-08-09 12:33:26.527862\n",
      "2024-08-09 12:33:26.537166\n",
      "2024-08-09 12:33:26.546527\n",
      "2024-08-09 12:33:26.554602\n",
      "2024-08-09 12:33:26.562693\n",
      "2024-08-09 12:33:26.571091\n",
      "2024-08-09 12:33:26.579382\n",
      "2024-08-09 12:33:26.587684\n",
      "2024-08-09 12:33:26.596362\n",
      "2024-08-09 12:33:26.604995\n",
      "2024-08-09 12:33:26.613962\n",
      "2024-08-09 12:33:26.622255\n",
      "2024-08-09 12:33:26.630433\n",
      "2024-08-09 12:33:26.638108\n",
      "2024-08-09 12:33:26.647186\n",
      "2024-08-09 12:33:26.654709\n",
      "2024-08-09 12:33:26.663562\n",
      "2024-08-09 12:33:26.671794\n",
      "2024-08-09 12:33:26.679167\n",
      "2024-08-09 12:33:26.687021\n",
      "2024-08-09 12:33:26.695643\n",
      "2024-08-09 12:33:26.703409\n",
      "2024-08-09 12:33:26.710617\n",
      "2024-08-09 12:33:26.718022\n",
      "2024-08-09 12:33:26.725865\n",
      "2024-08-09 12:33:26.734920\n",
      "2024-08-09 12:33:26.742289\n",
      "2024-08-09 12:33:26.750249\n",
      "2024-08-09 12:33:26.758970\n",
      "2024-08-09 12:33:26.766608\n",
      "2024-08-09 12:33:26.773482\n",
      "2024-08-09 12:33:26.781280\n",
      "2024-08-09 12:33:26.788586\n",
      "2024-08-09 12:33:26.796887\n",
      "2024-08-09 12:33:26.805285\n",
      "2024-08-09 12:33:26.813764\n",
      "2024-08-09 12:33:26.821565\n",
      "2024-08-09 12:33:26.829158\n",
      "2024-08-09 12:33:26.838043\n",
      "2024-08-09 12:33:26.846119\n",
      "2024-08-09 12:33:26.854527\n",
      "2024-08-09 12:33:26.861881\n",
      "2024-08-09 12:33:26.871262\n",
      "2024-08-09 12:33:26.878694\n",
      "2024-08-09 12:33:26.886584\n",
      "2024-08-09 12:33:26.894348\n",
      "2024-08-09 12:33:26.901765\n",
      "2024-08-09 12:33:26.909437\n",
      "2024-08-09 12:33:26.916245\n",
      "2024-08-09 12:33:26.923175\n",
      "2024-08-09 12:33:26.930365\n",
      "2024-08-09 12:33:26.938134\n",
      "2024-08-09 12:33:26.946658\n",
      "2024-08-09 12:33:26.959423\n",
      "2024-08-09 12:33:26.970362\n",
      "2024-08-09 12:33:26.978892\n",
      "2024-08-09 12:33:26.987421\n",
      "2024-08-09 12:33:26.995136\n",
      "2024-08-09 12:33:27.001970\n",
      "2024-08-09 12:33:27.008750\n",
      "2024-08-09 12:33:27.015800\n",
      "2024-08-09 12:33:27.023499\n",
      "2024-08-09 12:33:27.032864\n",
      "2024-08-09 12:33:27.041549\n",
      "2024-08-09 12:33:27.049233\n",
      "2024-08-09 12:33:27.056341\n",
      "2024-08-09 12:33:27.063769\n",
      "2024-08-09 12:33:27.070975\n",
      "2024-08-09 12:33:27.079566\n",
      "2024-08-09 12:33:27.090311\n",
      "2024-08-09 12:33:27.101478\n",
      "2024-08-09 12:33:27.109415\n",
      "2024-08-09 12:33:27.117233\n",
      "2024-08-09 12:33:27.124657\n",
      "2024-08-09 12:33:27.132005\n",
      "2024-08-09 12:33:27.139241\n",
      "2024-08-09 12:33:27.147314\n",
      "2024-08-09 12:33:27.155687\n",
      "2024-08-09 12:33:27.163244\n",
      "2024-08-09 12:33:27.170406\n",
      "2024-08-09 12:33:27.177390\n",
      "2024-08-09 12:33:27.185442\n",
      "2024-08-09 12:33:27.194992\n",
      "2024-08-09 12:33:27.204520\n",
      "2024-08-09 12:33:27.215184\n",
      "2024-08-09 12:33:27.223900\n",
      "2024-08-09 12:33:27.232022\n",
      "2024-08-09 12:33:27.240520\n",
      "2024-08-09 12:33:27.248588\n",
      "2024-08-09 12:33:27.255925\n",
      "2024-08-09 12:33:27.263136\n",
      "2024-08-09 12:33:27.273110\n",
      "2024-08-09 12:33:27.282902\n",
      "2024-08-09 12:33:27.292510\n",
      "2024-08-09 12:33:27.301395\n",
      "2024-08-09 12:33:27.308503\n",
      "2024-08-09 12:33:27.315889\n",
      "2024-08-09 12:33:27.324077\n",
      "2024-08-09 12:33:27.332446\n",
      "2024-08-09 12:33:27.340520\n",
      "2024-08-09 12:33:27.348846\n",
      "2024-08-09 12:33:27.356697\n",
      "2024-08-09 12:33:27.364515\n",
      "2024-08-09 12:33:27.372380\n",
      "2024-08-09 12:33:27.380249\n",
      "2024-08-09 12:33:27.388172\n",
      "2024-08-09 12:33:27.396044\n",
      "2024-08-09 12:33:27.405099\n",
      "2024-08-09 12:33:27.412915\n",
      "2024-08-09 12:33:27.420931\n",
      "2024-08-09 12:33:27.429035\n",
      "2024-08-09 12:33:27.437039\n",
      "2024-08-09 12:33:27.444836\n",
      "2024-08-09 12:33:27.453939\n",
      "2024-08-09 12:33:27.461617\n",
      "2024-08-09 12:33:27.469091\n",
      "2024-08-09 12:33:27.476248\n",
      "2024-08-09 12:33:27.483494\n",
      "2024-08-09 12:33:27.491217\n",
      "2024-08-09 12:33:27.499833\n",
      "2024-08-09 12:33:27.509343\n",
      "2024-08-09 12:33:27.516463\n",
      "2024-08-09 12:33:27.523917\n",
      "2024-08-09 12:33:27.535160\n",
      "2024-08-09 12:33:27.542691\n",
      "2024-08-09 12:33:27.550174\n",
      "2024-08-09 12:33:27.557889\n",
      "2024-08-09 12:33:27.565437\n",
      "2024-08-09 12:33:27.574513\n",
      "2024-08-09 12:33:27.584113\n",
      "2024-08-09 12:33:27.592008\n",
      "2024-08-09 12:33:27.600240\n",
      "2024-08-09 12:33:27.608671\n",
      "2024-08-09 12:33:27.615994\n",
      "2024-08-09 12:33:27.625189\n",
      "2024-08-09 12:33:27.634362\n",
      "2024-08-09 12:33:27.643676\n",
      "2024-08-09 12:33:27.653506\n",
      "2024-08-09 12:33:27.661165\n",
      "2024-08-09 12:33:27.668708\n",
      "2024-08-09 12:33:27.677160\n",
      "2024-08-09 12:33:27.684275\n",
      "2024-08-09 12:33:27.692818\n",
      "2024-08-09 12:33:27.700950\n",
      "2024-08-09 12:33:27.709620\n",
      "2024-08-09 12:33:27.717473\n",
      "2024-08-09 12:33:27.726218\n",
      "2024-08-09 12:33:27.734337\n",
      "2024-08-09 12:33:27.742839\n",
      "2024-08-09 12:33:27.751172\n",
      "2024-08-09 12:33:27.759457\n",
      "2024-08-09 12:33:27.767323\n",
      "2024-08-09 12:33:27.775636\n",
      "df_l2g_original df unpersisted\n",
      "df_l2g_propagated\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/09 12:33:27 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-09 12:33:28.273044\n",
      "2024-08-09 12:33:28.282922\n",
      "2024-08-09 12:33:28.291608\n",
      "2024-08-09 12:33:28.299506\n",
      "2024-08-09 12:33:28.307860\n",
      "2024-08-09 12:33:28.316238\n",
      "2024-08-09 12:33:28.324657\n",
      "2024-08-09 12:33:28.332700\n",
      "2024-08-09 12:33:28.341796\n",
      "2024-08-09 12:33:28.349845\n",
      "2024-08-09 12:33:28.358681\n",
      "2024-08-09 12:33:28.366449\n",
      "2024-08-09 12:33:28.374013\n",
      "2024-08-09 12:33:28.381620\n",
      "2024-08-09 12:33:28.389001\n",
      "2024-08-09 12:33:28.395976\n",
      "2024-08-09 12:33:28.403716\n",
      "2024-08-09 12:33:28.411079\n",
      "2024-08-09 12:33:28.417874\n",
      "2024-08-09 12:33:28.426193\n",
      "2024-08-09 12:33:28.434824\n",
      "2024-08-09 12:33:28.444227\n",
      "2024-08-09 12:33:28.452889\n",
      "2024-08-09 12:33:28.462659\n",
      "2024-08-09 12:33:28.473108\n",
      "2024-08-09 12:33:28.481591\n",
      "2024-08-09 12:33:28.489437\n",
      "2024-08-09 12:33:28.497218\n",
      "2024-08-09 12:33:28.504568\n",
      "2024-08-09 12:33:28.512139\n",
      "2024-08-09 12:33:28.520219\n",
      "2024-08-09 12:33:28.527461\n",
      "2024-08-09 12:33:28.535625\n",
      "2024-08-09 12:33:28.542778\n",
      "2024-08-09 12:33:28.550495\n",
      "2024-08-09 12:33:28.558552\n",
      "2024-08-09 12:33:28.567835\n",
      "2024-08-09 12:33:28.576784\n",
      "2024-08-09 12:33:28.585276\n",
      "2024-08-09 12:33:28.593140\n",
      "2024-08-09 12:33:28.601476\n",
      "2024-08-09 12:33:28.610173\n",
      "2024-08-09 12:33:28.619598\n",
      "2024-08-09 12:33:28.627804\n",
      "2024-08-09 12:33:28.636580\n",
      "2024-08-09 12:33:28.644657\n",
      "2024-08-09 12:33:28.651801\n",
      "2024-08-09 12:33:28.658744\n",
      "2024-08-09 12:33:28.665933\n",
      "2024-08-09 12:33:28.674322\n",
      "2024-08-09 12:33:28.681674\n",
      "2024-08-09 12:33:28.692228\n",
      "2024-08-09 12:33:28.702268\n",
      "2024-08-09 12:33:28.711127\n",
      "2024-08-09 12:33:28.719577\n",
      "2024-08-09 12:33:28.727974\n",
      "2024-08-09 12:33:28.736436\n",
      "2024-08-09 12:33:28.746614\n",
      "2024-08-09 12:33:28.755674\n",
      "2024-08-09 12:33:28.764713\n",
      "2024-08-09 12:33:28.772482\n",
      "2024-08-09 12:33:28.783172\n",
      "2024-08-09 12:33:28.792022\n",
      "2024-08-09 12:33:28.799946\n",
      "2024-08-09 12:33:28.807226\n",
      "2024-08-09 12:33:28.814242\n",
      "2024-08-09 12:33:28.821457\n",
      "2024-08-09 12:33:28.833071\n",
      "2024-08-09 12:33:28.841778\n",
      "2024-08-09 12:33:28.849719\n",
      "2024-08-09 12:33:28.857994\n",
      "2024-08-09 12:33:28.865608\n",
      "2024-08-09 12:33:28.873822\n",
      "2024-08-09 12:33:28.881970\n",
      "2024-08-09 12:33:28.890712\n",
      "2024-08-09 12:33:28.900022\n",
      "2024-08-09 12:33:28.909045\n",
      "2024-08-09 12:33:28.918020\n",
      "2024-08-09 12:33:28.927719\n",
      "2024-08-09 12:33:28.937214\n",
      "2024-08-09 12:33:28.947135\n",
      "2024-08-09 12:33:28.955073\n",
      "2024-08-09 12:33:28.962462\n",
      "2024-08-09 12:33:28.975945\n",
      "2024-08-09 12:33:28.984060\n",
      "2024-08-09 12:33:28.992566\n",
      "2024-08-09 12:33:29.000370\n",
      "2024-08-09 12:33:29.009252\n",
      "2024-08-09 12:33:29.017827\n",
      "2024-08-09 12:33:29.027811\n",
      "2024-08-09 12:33:29.036337\n",
      "2024-08-09 12:33:29.044834\n",
      "2024-08-09 12:33:29.052147\n",
      "2024-08-09 12:33:29.058990\n",
      "2024-08-09 12:33:29.066053\n",
      "2024-08-09 12:33:29.074222\n",
      "2024-08-09 12:33:29.081553\n",
      "2024-08-09 12:33:29.089581\n",
      "2024-08-09 12:33:29.097967\n",
      "2024-08-09 12:33:29.106676\n",
      "2024-08-09 12:33:29.114506\n",
      "2024-08-09 12:33:29.122132\n",
      "2024-08-09 12:33:29.129585\n",
      "2024-08-09 12:33:29.137816\n",
      "2024-08-09 12:33:29.146703\n",
      "2024-08-09 12:33:29.158748\n",
      "2024-08-09 12:33:29.166962\n",
      "2024-08-09 12:33:29.175306\n",
      "2024-08-09 12:33:29.183366\n",
      "2024-08-09 12:33:29.191905\n",
      "2024-08-09 12:33:29.200943\n",
      "2024-08-09 12:33:29.209293\n",
      "2024-08-09 12:33:29.217434\n",
      "2024-08-09 12:33:29.225407\n",
      "2024-08-09 12:33:29.233399\n",
      "2024-08-09 12:33:29.241538\n",
      "2024-08-09 12:33:29.250417\n",
      "2024-08-09 12:33:29.257604\n",
      "2024-08-09 12:33:29.265181\n",
      "2024-08-09 12:33:29.273609\n",
      "2024-08-09 12:33:29.281132\n",
      "2024-08-09 12:33:29.288531\n",
      "2024-08-09 12:33:29.295519\n",
      "2024-08-09 12:33:29.305267\n",
      "2024-08-09 12:33:29.314555\n",
      "2024-08-09 12:33:29.323944\n",
      "2024-08-09 12:33:29.331637\n",
      "2024-08-09 12:33:29.338765\n",
      "2024-08-09 12:33:29.347306\n",
      "2024-08-09 12:33:29.354996\n",
      "2024-08-09 12:33:29.362545\n",
      "2024-08-09 12:33:29.370363\n",
      "2024-08-09 12:33:29.377693\n",
      "2024-08-09 12:33:29.386519\n",
      "2024-08-09 12:33:29.394334\n",
      "2024-08-09 12:33:29.402133\n",
      "2024-08-09 12:33:29.410237\n",
      "2024-08-09 12:33:29.417309\n",
      "2024-08-09 12:33:29.425073\n",
      "2024-08-09 12:33:29.433087\n",
      "2024-08-09 12:33:29.441022\n",
      "2024-08-09 12:33:29.447867\n",
      "2024-08-09 12:33:29.455154\n",
      "2024-08-09 12:33:29.462556\n",
      "2024-08-09 12:33:29.470012\n",
      "2024-08-09 12:33:29.477824\n",
      "2024-08-09 12:33:29.485034\n",
      "2024-08-09 12:33:29.493428\n",
      "2024-08-09 12:33:29.500745\n",
      "2024-08-09 12:33:29.509129\n",
      "2024-08-09 12:33:29.519427\n",
      "2024-08-09 12:33:29.527825\n",
      "2024-08-09 12:33:29.536138\n",
      "2024-08-09 12:33:29.544170\n",
      "2024-08-09 12:33:29.552658\n",
      "2024-08-09 12:33:29.561063\n",
      "2024-08-09 12:33:29.568512\n",
      "2024-08-09 12:33:29.577339\n",
      "2024-08-09 12:33:29.588448\n",
      "2024-08-09 12:33:29.596321\n",
      "2024-08-09 12:33:29.605061\n",
      "2024-08-09 12:33:29.614372\n",
      "2024-08-09 12:33:29.623569\n",
      "2024-08-09 12:33:29.633077\n",
      "2024-08-09 12:33:29.641750\n",
      "2024-08-09 12:33:29.649737\n",
      "2024-08-09 12:33:29.657924\n",
      "2024-08-09 12:33:29.665494\n",
      "2024-08-09 12:33:29.674363\n",
      "2024-08-09 12:33:29.682415\n",
      "2024-08-09 12:33:29.692168\n",
      "2024-08-09 12:33:29.701561\n",
      "2024-08-09 12:33:29.710246\n",
      "2024-08-09 12:33:29.718086\n",
      "2024-08-09 12:33:29.727701\n",
      "2024-08-09 12:33:29.736790\n",
      "2024-08-09 12:33:29.744470\n",
      "2024-08-09 12:33:29.751992\n",
      "2024-08-09 12:33:29.759067\n",
      "2024-08-09 12:33:29.767661\n",
      "2024-08-09 12:33:29.777170\n",
      "2024-08-09 12:33:29.785152\n",
      "2024-08-09 12:33:29.793162\n",
      "2024-08-09 12:33:29.802120\n",
      "2024-08-09 12:33:29.810235\n",
      "2024-08-09 12:33:29.819070\n",
      "2024-08-09 12:33:29.827616\n",
      "2024-08-09 12:33:29.836486\n",
      "2024-08-09 12:33:29.845557\n",
      "2024-08-09 12:33:29.853180\n",
      "2024-08-09 12:33:29.860558\n",
      "2024-08-09 12:33:29.868344\n",
      "2024-08-09 12:33:29.875648\n",
      "2024-08-09 12:33:29.882420\n",
      "2024-08-09 12:33:29.889780\n",
      "2024-08-09 12:33:29.897038\n",
      "2024-08-09 12:33:29.905628\n",
      "2024-08-09 12:33:29.912806\n",
      "2024-08-09 12:33:29.919923\n",
      "2024-08-09 12:33:29.926863\n",
      "2024-08-09 12:33:29.934396\n",
      "2024-08-09 12:33:29.941854\n",
      "2024-08-09 12:33:29.950731\n",
      "2024-08-09 12:33:29.960312\n",
      "2024-08-09 12:33:29.968338\n",
      "2024-08-09 12:33:29.977653\n",
      "2024-08-09 12:33:29.985653\n",
      "2024-08-09 12:33:29.993160\n",
      "2024-08-09 12:33:30.000954\n",
      "2024-08-09 12:33:30.008024\n",
      "2024-08-09 12:33:30.015531\n",
      "2024-08-09 12:33:30.024209\n",
      "2024-08-09 12:33:30.032427\n",
      "2024-08-09 12:33:30.040447\n",
      "2024-08-09 12:33:30.047426\n",
      "2024-08-09 12:33:30.055288\n",
      "2024-08-09 12:33:30.062961\n",
      "2024-08-09 12:33:30.071791\n",
      "2024-08-09 12:33:30.079392\n",
      "2024-08-09 12:33:30.086313\n",
      "2024-08-09 12:33:30.093270\n",
      "2024-08-09 12:33:30.100621\n",
      "2024-08-09 12:33:30.107528\n",
      "2024-08-09 12:33:30.114840\n",
      "2024-08-09 12:33:30.121915\n",
      "2024-08-09 12:33:30.132363\n",
      "2024-08-09 12:33:30.141958\n",
      "2024-08-09 12:33:30.151400\n",
      "2024-08-09 12:33:30.159745\n",
      "2024-08-09 12:33:30.167585\n",
      "2024-08-09 12:33:30.175412\n",
      "2024-08-09 12:33:30.183339\n",
      "2024-08-09 12:33:30.191076\n",
      "2024-08-09 12:33:30.198427\n",
      "2024-08-09 12:33:30.206166\n",
      "2024-08-09 12:33:30.216938\n",
      "2024-08-09 12:33:30.224662\n",
      "2024-08-09 12:33:30.233133\n",
      "2024-08-09 12:33:30.241498\n",
      "2024-08-09 12:33:30.249325\n",
      "2024-08-09 12:33:30.257680\n",
      "2024-08-09 12:33:30.265668\n",
      "2024-08-09 12:33:30.274106\n",
      "2024-08-09 12:33:30.281442\n",
      "2024-08-09 12:33:30.289204\n",
      "2024-08-09 12:33:30.296595\n",
      "2024-08-09 12:33:30.303916\n",
      "2024-08-09 12:33:30.311424\n",
      "2024-08-09 12:33:30.319369\n",
      "2024-08-09 12:33:30.327304\n",
      "2024-08-09 12:33:30.336118\n",
      "2024-08-09 12:33:30.345371\n",
      "2024-08-09 12:33:30.353901\n",
      "2024-08-09 12:33:30.361506\n",
      "2024-08-09 12:33:30.369052\n",
      "2024-08-09 12:33:30.376627\n",
      "2024-08-09 12:33:30.383913\n",
      "2024-08-09 12:33:30.392627\n",
      "2024-08-09 12:33:30.399985\n",
      "2024-08-09 12:33:30.407909\n",
      "2024-08-09 12:33:30.416187\n",
      "2024-08-09 12:33:30.424017\n",
      "2024-08-09 12:33:30.431735\n",
      "2024-08-09 12:33:30.439275\n",
      "2024-08-09 12:33:30.446894\n",
      "2024-08-09 12:33:30.454291\n",
      "2024-08-09 12:33:30.461783\n",
      "2024-08-09 12:33:30.468981\n",
      "2024-08-09 12:33:30.478124\n",
      "2024-08-09 12:33:30.486329\n",
      "2024-08-09 12:33:30.495484\n",
      "2024-08-09 12:33:30.504259\n",
      "2024-08-09 12:33:30.512591\n",
      "2024-08-09 12:33:30.520827\n",
      "2024-08-09 12:33:30.528965\n",
      "2024-08-09 12:33:30.536589\n",
      "2024-08-09 12:33:30.543428\n",
      "2024-08-09 12:33:30.550831\n",
      "2024-08-09 12:33:30.558406\n",
      "2024-08-09 12:33:30.566141\n",
      "2024-08-09 12:33:30.573375\n",
      "2024-08-09 12:33:30.581010\n",
      "2024-08-09 12:33:30.588322\n",
      "2024-08-09 12:33:30.595969\n",
      "2024-08-09 12:33:30.603254\n",
      "2024-08-09 12:33:30.610384\n",
      "2024-08-09 12:33:30.617717\n",
      "2024-08-09 12:33:30.627286\n",
      "2024-08-09 12:33:30.636248\n",
      "2024-08-09 12:33:30.646185\n",
      "2024-08-09 12:33:30.654051\n",
      "2024-08-09 12:33:30.661837\n",
      "2024-08-09 12:33:30.669800\n",
      "2024-08-09 12:33:30.677352\n",
      "2024-08-09 12:33:30.685032\n",
      "2024-08-09 12:33:30.695315\n",
      "2024-08-09 12:33:30.706985\n",
      "2024-08-09 12:33:30.716372\n",
      "2024-08-09 12:33:30.724428\n",
      "2024-08-09 12:33:30.733646\n",
      "2024-08-09 12:33:30.743439\n",
      "2024-08-09 12:33:30.752994\n",
      "2024-08-09 12:33:30.762122\n",
      "2024-08-09 12:33:30.770702\n",
      "2024-08-09 12:33:30.778770\n",
      "2024-08-09 12:33:30.787655\n",
      "2024-08-09 12:33:30.796396\n",
      "2024-08-09 12:33:30.804516\n",
      "2024-08-09 12:33:30.813404\n",
      "2024-08-09 12:33:30.821700\n",
      "2024-08-09 12:33:30.829407\n",
      "2024-08-09 12:33:30.842392\n",
      "2024-08-09 12:33:30.851602\n",
      "2024-08-09 12:33:30.860747\n",
      "2024-08-09 12:33:30.869791\n",
      "2024-08-09 12:33:30.877679\n",
      "2024-08-09 12:33:30.885309\n",
      "2024-08-09 12:33:30.893510\n",
      "2024-08-09 12:33:30.902383\n",
      "2024-08-09 12:33:30.909600\n",
      "2024-08-09 12:33:30.917349\n",
      "2024-08-09 12:33:30.924386\n",
      "2024-08-09 12:33:30.931945\n",
      "2024-08-09 12:33:30.939438\n",
      "2024-08-09 12:33:30.946709\n",
      "2024-08-09 12:33:30.954703\n",
      "2024-08-09 12:33:30.963080\n",
      "2024-08-09 12:33:30.970736\n",
      "2024-08-09 12:33:30.978463\n",
      "2024-08-09 12:33:30.989361\n",
      "2024-08-09 12:33:30.997715\n",
      "2024-08-09 12:33:31.007643\n",
      "2024-08-09 12:33:31.017294\n",
      "2024-08-09 12:33:31.027184\n",
      "2024-08-09 12:33:31.036007\n",
      "2024-08-09 12:33:31.046043\n",
      "2024-08-09 12:33:31.056340\n",
      "2024-08-09 12:33:31.065616\n",
      "2024-08-09 12:33:31.073829\n",
      "2024-08-09 12:33:31.081612\n",
      "2024-08-09 12:33:31.089320\n",
      "2024-08-09 12:33:31.097659\n",
      "2024-08-09 12:33:31.106253\n",
      "2024-08-09 12:33:31.118616\n",
      "2024-08-09 12:33:31.127273\n",
      "2024-08-09 12:33:31.134834\n",
      "2024-08-09 12:33:31.143825\n",
      "2024-08-09 12:33:31.153167\n",
      "2024-08-09 12:33:31.164853\n",
      "2024-08-09 12:33:31.173175\n",
      "2024-08-09 12:33:31.180726\n",
      "2024-08-09 12:33:31.188934\n",
      "2024-08-09 12:33:31.197568\n",
      "2024-08-09 12:33:31.208104\n",
      "2024-08-09 12:33:31.215802\n",
      "2024-08-09 12:33:31.223192\n",
      "2024-08-09 12:33:31.231626\n",
      "2024-08-09 12:33:31.238910\n",
      "2024-08-09 12:33:31.248948\n",
      "2024-08-09 12:33:31.261389\n",
      "2024-08-09 12:33:31.273319\n",
      "2024-08-09 12:33:31.282644\n",
      "2024-08-09 12:33:31.290855\n",
      "2024-08-09 12:33:31.298059\n",
      "2024-08-09 12:33:31.306094\n",
      "2024-08-09 12:33:31.313842\n",
      "2024-08-09 12:33:31.321303\n",
      "2024-08-09 12:33:31.329761\n",
      "2024-08-09 12:33:31.337464\n",
      "2024-08-09 12:33:31.344603\n",
      "2024-08-09 12:33:31.352019\n",
      "2024-08-09 12:33:31.358889\n",
      "2024-08-09 12:33:31.366516\n",
      "2024-08-09 12:33:31.374234\n",
      "2024-08-09 12:33:31.381672\n",
      "2024-08-09 12:33:31.389070\n",
      "2024-08-09 12:33:31.396513\n",
      "2024-08-09 12:33:31.405588\n",
      "2024-08-09 12:33:31.413680\n",
      "2024-08-09 12:33:31.423137\n",
      "df_l2g_propagated df unpersisted\n",
      "df_distance_original\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/09 12:33:31 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-09 12:33:31.825454\n",
      "2024-08-09 12:33:31.834162\n",
      "2024-08-09 12:33:31.842037\n",
      "2024-08-09 12:33:31.849578\n",
      "2024-08-09 12:33:31.860280\n",
      "2024-08-09 12:33:31.871016\n",
      "2024-08-09 12:33:31.880868\n",
      "2024-08-09 12:33:31.888208\n",
      "2024-08-09 12:33:31.895682\n",
      "2024-08-09 12:33:31.904884\n",
      "2024-08-09 12:33:31.912843\n",
      "2024-08-09 12:33:31.921091\n",
      "2024-08-09 12:33:31.930211\n",
      "2024-08-09 12:33:31.937657\n",
      "2024-08-09 12:33:31.944840\n",
      "2024-08-09 12:33:31.952132\n",
      "2024-08-09 12:33:31.958923\n",
      "2024-08-09 12:33:31.966736\n",
      "2024-08-09 12:33:31.974401\n",
      "2024-08-09 12:33:31.982122\n",
      "2024-08-09 12:33:31.990485\n",
      "2024-08-09 12:33:32.000044\n",
      "2024-08-09 12:33:32.008071\n",
      "2024-08-09 12:33:32.015681\n",
      "2024-08-09 12:33:32.023646\n",
      "2024-08-09 12:33:32.032921\n",
      "2024-08-09 12:33:32.041494\n",
      "2024-08-09 12:33:32.050181\n",
      "2024-08-09 12:33:32.058692\n",
      "2024-08-09 12:33:32.069904\n",
      "2024-08-09 12:33:32.078995\n",
      "2024-08-09 12:33:32.087122\n",
      "2024-08-09 12:33:32.094096\n",
      "2024-08-09 12:33:32.103027\n",
      "2024-08-09 12:33:32.110447\n",
      "2024-08-09 12:33:32.118839\n",
      "2024-08-09 12:33:32.127041\n",
      "2024-08-09 12:33:32.134576\n",
      "2024-08-09 12:33:32.142532\n",
      "2024-08-09 12:33:32.149645\n",
      "2024-08-09 12:33:32.157312\n",
      "2024-08-09 12:33:32.164920\n",
      "2024-08-09 12:33:32.172016\n",
      "2024-08-09 12:33:32.179484\n",
      "2024-08-09 12:33:32.186743\n",
      "2024-08-09 12:33:32.193827\n",
      "2024-08-09 12:33:32.201415\n",
      "2024-08-09 12:33:32.209176\n",
      "2024-08-09 12:33:32.216476\n",
      "2024-08-09 12:33:32.223543\n",
      "2024-08-09 12:33:32.230706\n",
      "2024-08-09 12:33:32.238730\n",
      "2024-08-09 12:33:32.247045\n",
      "2024-08-09 12:33:32.254925\n",
      "2024-08-09 12:33:32.265396\n",
      "2024-08-09 12:33:32.273642\n",
      "2024-08-09 12:33:32.283242\n",
      "2024-08-09 12:33:32.292151\n",
      "2024-08-09 12:33:32.299535\n",
      "2024-08-09 12:33:32.306214\n",
      "2024-08-09 12:33:32.313924\n",
      "2024-08-09 12:33:32.322505\n",
      "2024-08-09 12:33:32.329772\n",
      "2024-08-09 12:33:32.336510\n",
      "2024-08-09 12:33:32.343692\n",
      "2024-08-09 12:33:32.351343\n",
      "2024-08-09 12:33:32.358290\n",
      "2024-08-09 12:33:32.364961\n",
      "2024-08-09 12:33:32.371715\n",
      "2024-08-09 12:33:32.378358\n",
      "2024-08-09 12:33:32.385505\n",
      "2024-08-09 12:33:32.392660\n",
      "2024-08-09 12:33:32.408887\n",
      "2024-08-09 12:33:32.418311\n",
      "2024-08-09 12:33:32.426319\n",
      "2024-08-09 12:33:32.434257\n",
      "2024-08-09 12:33:32.443256\n",
      "2024-08-09 12:33:32.450478\n",
      "2024-08-09 12:33:32.457762\n",
      "2024-08-09 12:33:32.465809\n",
      "2024-08-09 12:33:32.474533\n",
      "2024-08-09 12:33:32.482254\n",
      "2024-08-09 12:33:32.491620\n",
      "2024-08-09 12:33:32.499294\n",
      "2024-08-09 12:33:32.507007\n",
      "2024-08-09 12:33:32.516373\n",
      "2024-08-09 12:33:32.524921\n",
      "2024-08-09 12:33:32.532914\n",
      "2024-08-09 12:33:32.540929\n",
      "2024-08-09 12:33:32.550047\n",
      "2024-08-09 12:33:32.566169\n",
      "2024-08-09 12:33:32.576107\n",
      "2024-08-09 12:33:32.584246\n",
      "2024-08-09 12:33:32.592064\n",
      "2024-08-09 12:33:32.600689\n",
      "2024-08-09 12:33:32.608816\n",
      "2024-08-09 12:33:32.616047\n",
      "2024-08-09 12:33:32.623038\n",
      "2024-08-09 12:33:32.630156\n",
      "2024-08-09 12:33:32.637708\n",
      "2024-08-09 12:33:32.647275\n",
      "2024-08-09 12:33:32.655262\n",
      "2024-08-09 12:33:32.663810\n",
      "2024-08-09 12:33:32.673236\n",
      "2024-08-09 12:33:32.682044\n",
      "2024-08-09 12:33:32.694536\n",
      "2024-08-09 12:33:32.702877\n",
      "2024-08-09 12:33:32.711203\n",
      "2024-08-09 12:33:32.719078\n",
      "2024-08-09 12:33:32.726411\n",
      "2024-08-09 12:33:32.734093\n",
      "2024-08-09 12:33:32.741188\n",
      "2024-08-09 12:33:32.750937\n",
      "2024-08-09 12:33:32.759354\n",
      "2024-08-09 12:33:32.768468\n",
      "2024-08-09 12:33:32.777303\n",
      "2024-08-09 12:33:32.785635\n",
      "2024-08-09 12:33:32.793422\n",
      "2024-08-09 12:33:32.800513\n",
      "2024-08-09 12:33:32.808295\n",
      "2024-08-09 12:33:32.817333\n",
      "2024-08-09 12:33:32.825477\n",
      "2024-08-09 12:33:32.834406\n",
      "2024-08-09 12:33:32.842849\n",
      "2024-08-09 12:33:32.851730\n",
      "2024-08-09 12:33:32.861804\n",
      "2024-08-09 12:33:32.870221\n",
      "2024-08-09 12:33:32.877414\n",
      "2024-08-09 12:33:32.886886\n",
      "2024-08-09 12:33:32.895004\n",
      "2024-08-09 12:33:32.903489\n",
      "2024-08-09 12:33:32.912637\n",
      "2024-08-09 12:33:32.921325\n",
      "2024-08-09 12:33:32.928825\n",
      "2024-08-09 12:33:32.936830\n",
      "2024-08-09 12:33:32.946133\n",
      "2024-08-09 12:33:32.955528\n",
      "2024-08-09 12:33:32.964044\n",
      "2024-08-09 12:33:32.975334\n",
      "2024-08-09 12:33:32.984855\n",
      "2024-08-09 12:33:32.992265\n",
      "2024-08-09 12:33:32.999553\n",
      "2024-08-09 12:33:33.006823\n",
      "2024-08-09 12:33:33.014162\n",
      "2024-08-09 12:33:33.020806\n",
      "2024-08-09 12:33:33.027827\n",
      "2024-08-09 12:33:33.035549\n",
      "2024-08-09 12:33:33.043245\n",
      "2024-08-09 12:33:33.050922\n",
      "2024-08-09 12:33:33.059535\n",
      "2024-08-09 12:33:33.068836\n",
      "2024-08-09 12:33:33.076762\n",
      "2024-08-09 12:33:33.085239\n",
      "2024-08-09 12:33:33.092629\n",
      "2024-08-09 12:33:33.101428\n",
      "2024-08-09 12:33:33.110290\n",
      "2024-08-09 12:33:33.118957\n",
      "2024-08-09 12:33:33.127719\n",
      "2024-08-09 12:33:33.137903\n",
      "2024-08-09 12:33:33.147606\n",
      "2024-08-09 12:33:33.159293\n",
      "2024-08-09 12:33:33.171873\n",
      "2024-08-09 12:33:33.182123\n",
      "2024-08-09 12:33:33.192373\n",
      "2024-08-09 12:33:33.200785\n",
      "2024-08-09 12:33:33.208915\n",
      "2024-08-09 12:33:33.216490\n",
      "2024-08-09 12:33:33.228103\n",
      "2024-08-09 12:33:33.236871\n",
      "2024-08-09 12:33:33.246552\n",
      "2024-08-09 12:33:33.256494\n",
      "2024-08-09 12:33:33.266647\n",
      "2024-08-09 12:33:33.276115\n",
      "2024-08-09 12:33:33.284786\n",
      "2024-08-09 12:33:33.294157\n",
      "2024-08-09 12:33:33.302610\n",
      "2024-08-09 12:33:33.311436\n",
      "2024-08-09 12:33:33.320021\n",
      "2024-08-09 12:33:33.327604\n",
      "2024-08-09 12:33:33.335412\n",
      "2024-08-09 12:33:33.342956\n",
      "2024-08-09 12:33:33.350017\n",
      "2024-08-09 12:33:33.357303\n",
      "2024-08-09 12:33:33.364797\n",
      "2024-08-09 12:33:33.374083\n",
      "2024-08-09 12:33:33.383110\n",
      "2024-08-09 12:33:33.391254\n",
      "2024-08-09 12:33:33.398895\n",
      "2024-08-09 12:33:33.407604\n",
      "2024-08-09 12:33:33.417111\n",
      "2024-08-09 12:33:33.425423\n",
      "2024-08-09 12:33:33.432597\n",
      "2024-08-09 12:33:33.441118\n",
      "2024-08-09 12:33:33.448229\n",
      "2024-08-09 12:33:33.455370\n",
      "2024-08-09 12:33:33.462934\n",
      "2024-08-09 12:33:33.470545\n",
      "2024-08-09 12:33:33.479923\n",
      "2024-08-09 12:33:33.487646\n",
      "2024-08-09 12:33:33.495225\n",
      "2024-08-09 12:33:33.502500\n",
      "2024-08-09 12:33:33.510753\n",
      "2024-08-09 12:33:33.518043\n",
      "2024-08-09 12:33:33.525672\n",
      "2024-08-09 12:33:33.533687\n",
      "2024-08-09 12:33:33.540925\n",
      "2024-08-09 12:33:33.549134\n",
      "2024-08-09 12:33:33.558645\n",
      "2024-08-09 12:33:33.567497\n",
      "2024-08-09 12:33:33.576300\n",
      "2024-08-09 12:33:33.585092\n",
      "2024-08-09 12:33:33.593318\n",
      "2024-08-09 12:33:33.601281\n",
      "2024-08-09 12:33:33.610920\n",
      "2024-08-09 12:33:33.619055\n",
      "2024-08-09 12:33:33.626671\n",
      "2024-08-09 12:33:33.633564\n",
      "2024-08-09 12:33:33.640420\n",
      "2024-08-09 12:33:33.647529\n",
      "2024-08-09 12:33:33.654322\n",
      "df_distance_original df unpersisted\n",
      "df_distance_propagated\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/09 12:33:33 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-09 12:33:34.045640\n",
      "2024-08-09 12:33:34.058181\n",
      "2024-08-09 12:33:34.066922\n",
      "2024-08-09 12:33:34.077316\n",
      "2024-08-09 12:33:34.086147\n",
      "2024-08-09 12:33:34.095247\n",
      "2024-08-09 12:33:34.103705\n",
      "2024-08-09 12:33:34.116731\n",
      "2024-08-09 12:33:34.129532\n",
      "2024-08-09 12:33:34.141066\n",
      "2024-08-09 12:33:34.151217\n",
      "2024-08-09 12:33:34.158535\n",
      "2024-08-09 12:33:34.165414\n",
      "2024-08-09 12:33:34.172313\n",
      "2024-08-09 12:33:34.179431\n",
      "2024-08-09 12:33:34.186486\n",
      "2024-08-09 12:33:34.194727\n",
      "2024-08-09 12:33:34.202559\n",
      "2024-08-09 12:33:34.210778\n",
      "2024-08-09 12:33:34.218617\n",
      "2024-08-09 12:33:34.226690\n",
      "2024-08-09 12:33:34.236517\n",
      "2024-08-09 12:33:34.245726\n",
      "2024-08-09 12:33:34.254151\n",
      "2024-08-09 12:33:34.261525\n",
      "2024-08-09 12:33:34.269795\n",
      "2024-08-09 12:33:34.279101\n",
      "2024-08-09 12:33:34.286995\n",
      "2024-08-09 12:33:34.295307\n",
      "2024-08-09 12:33:34.305636\n",
      "2024-08-09 12:33:34.315286\n",
      "2024-08-09 12:33:34.323886\n",
      "2024-08-09 12:33:34.333009\n",
      "2024-08-09 12:33:34.341362\n",
      "2024-08-09 12:33:34.350162\n",
      "2024-08-09 12:33:34.357920\n",
      "2024-08-09 12:33:34.366009\n",
      "2024-08-09 12:33:34.373590\n",
      "2024-08-09 12:33:34.381219\n",
      "2024-08-09 12:33:34.388379\n",
      "2024-08-09 12:33:34.396133\n",
      "2024-08-09 12:33:34.406470\n",
      "2024-08-09 12:33:34.416048\n",
      "2024-08-09 12:33:34.424183\n",
      "2024-08-09 12:33:34.432234\n",
      "2024-08-09 12:33:34.439968\n",
      "2024-08-09 12:33:34.448116\n",
      "2024-08-09 12:33:34.456338\n",
      "2024-08-09 12:33:34.466220\n",
      "2024-08-09 12:33:34.479449\n",
      "2024-08-09 12:33:34.490805\n",
      "2024-08-09 12:33:34.502999\n",
      "2024-08-09 12:33:34.510685\n",
      "2024-08-09 12:33:34.519430\n",
      "2024-08-09 12:33:34.526862\n",
      "2024-08-09 12:33:34.534989\n",
      "2024-08-09 12:33:34.542091\n",
      "2024-08-09 12:33:34.549404\n",
      "2024-08-09 12:33:34.556757\n",
      "2024-08-09 12:33:34.564498\n",
      "2024-08-09 12:33:34.572166\n",
      "2024-08-09 12:33:34.580165\n",
      "2024-08-09 12:33:34.587522\n",
      "2024-08-09 12:33:34.594701\n",
      "2024-08-09 12:33:34.602745\n",
      "2024-08-09 12:33:34.610063\n",
      "2024-08-09 12:33:34.617094\n",
      "2024-08-09 12:33:34.624370\n",
      "2024-08-09 12:33:34.632518\n",
      "2024-08-09 12:33:34.641370\n",
      "2024-08-09 12:33:34.650484\n",
      "2024-08-09 12:33:34.658298\n",
      "2024-08-09 12:33:34.670387\n",
      "2024-08-09 12:33:34.679396\n",
      "2024-08-09 12:33:34.687410\n",
      "2024-08-09 12:33:34.694582\n",
      "2024-08-09 12:33:34.702585\n",
      "2024-08-09 12:33:34.709768\n",
      "2024-08-09 12:33:34.717886\n",
      "2024-08-09 12:33:34.726266\n",
      "2024-08-09 12:33:34.734729\n",
      "2024-08-09 12:33:34.742398\n",
      "2024-08-09 12:33:34.749938\n",
      "2024-08-09 12:33:34.757870\n",
      "2024-08-09 12:33:34.765345\n",
      "2024-08-09 12:33:34.773170\n",
      "2024-08-09 12:33:34.780091\n",
      "2024-08-09 12:33:34.787373\n",
      "2024-08-09 12:33:34.795291\n",
      "2024-08-09 12:33:34.803805\n",
      "2024-08-09 12:33:34.816257\n",
      "2024-08-09 12:33:34.826294\n",
      "2024-08-09 12:33:34.833534\n",
      "2024-08-09 12:33:34.840579\n",
      "2024-08-09 12:33:34.848977\n",
      "2024-08-09 12:33:34.858012\n",
      "2024-08-09 12:33:34.867777\n",
      "2024-08-09 12:33:34.878763\n",
      "2024-08-09 12:33:34.889188\n",
      "2024-08-09 12:33:34.897705\n",
      "2024-08-09 12:33:34.906583\n",
      "2024-08-09 12:33:34.915612\n",
      "2024-08-09 12:33:34.924483\n",
      "2024-08-09 12:33:34.932674\n",
      "2024-08-09 12:33:34.941662\n",
      "2024-08-09 12:33:34.949792\n",
      "2024-08-09 12:33:34.957902\n",
      "2024-08-09 12:33:34.967521\n",
      "2024-08-09 12:33:34.977184\n",
      "2024-08-09 12:33:34.988445\n",
      "2024-08-09 12:33:34.997081\n",
      "2024-08-09 12:33:35.004499\n",
      "2024-08-09 12:33:35.012128\n",
      "2024-08-09 12:33:35.019356\n",
      "2024-08-09 12:33:35.027666\n",
      "2024-08-09 12:33:35.035563\n",
      "2024-08-09 12:33:35.045675\n",
      "2024-08-09 12:33:35.055737\n",
      "2024-08-09 12:33:35.065091\n",
      "2024-08-09 12:33:35.074384\n",
      "2024-08-09 12:33:35.083324\n",
      "2024-08-09 12:33:35.091356\n",
      "2024-08-09 12:33:35.099730\n",
      "2024-08-09 12:33:35.107859\n",
      "2024-08-09 12:33:35.115800\n",
      "2024-08-09 12:33:35.123301\n",
      "2024-08-09 12:33:35.131185\n",
      "2024-08-09 12:33:35.139219\n",
      "2024-08-09 12:33:35.147319\n",
      "2024-08-09 12:33:35.155530\n",
      "2024-08-09 12:33:35.163292\n",
      "2024-08-09 12:33:35.170510\n",
      "2024-08-09 12:33:35.178208\n",
      "2024-08-09 12:33:35.185821\n",
      "2024-08-09 12:33:35.194609\n",
      "2024-08-09 12:33:35.203466\n",
      "2024-08-09 12:33:35.211923\n",
      "2024-08-09 12:33:35.220344\n",
      "2024-08-09 12:33:35.228511\n",
      "2024-08-09 12:33:35.237199\n",
      "2024-08-09 12:33:35.244509\n",
      "2024-08-09 12:33:35.252069\n",
      "2024-08-09 12:33:35.260299\n",
      "2024-08-09 12:33:35.270067\n",
      "2024-08-09 12:33:35.277605\n",
      "2024-08-09 12:33:35.285209\n",
      "2024-08-09 12:33:35.293177\n",
      "2024-08-09 12:33:35.302033\n",
      "2024-08-09 12:33:35.310223\n",
      "2024-08-09 12:33:35.318281\n",
      "2024-08-09 12:33:35.327689\n",
      "2024-08-09 12:33:35.335564\n",
      "2024-08-09 12:33:35.343427\n",
      "2024-08-09 12:33:35.351285\n",
      "2024-08-09 12:33:35.359657\n",
      "2024-08-09 12:33:35.367872\n",
      "2024-08-09 12:33:35.376209\n",
      "2024-08-09 12:33:35.383346\n",
      "2024-08-09 12:33:35.390554\n",
      "2024-08-09 12:33:35.398250\n",
      "2024-08-09 12:33:35.406094\n",
      "2024-08-09 12:33:35.413480\n",
      "2024-08-09 12:33:35.421536\n",
      "2024-08-09 12:33:35.430259\n",
      "2024-08-09 12:33:35.439755\n",
      "2024-08-09 12:33:35.449314\n",
      "2024-08-09 12:33:35.459236\n",
      "2024-08-09 12:33:35.471006\n",
      "2024-08-09 12:33:35.479538\n",
      "2024-08-09 12:33:35.488023\n",
      "2024-08-09 12:33:35.496938\n",
      "2024-08-09 12:33:35.505493\n",
      "2024-08-09 12:33:35.513329\n",
      "2024-08-09 12:33:35.521928\n",
      "2024-08-09 12:33:35.529395\n",
      "2024-08-09 12:33:35.538707\n",
      "2024-08-09 12:33:35.548539\n",
      "2024-08-09 12:33:35.558482\n",
      "2024-08-09 12:33:35.567342\n",
      "2024-08-09 12:33:35.576657\n",
      "2024-08-09 12:33:35.585390\n",
      "2024-08-09 12:33:35.593094\n",
      "2024-08-09 12:33:35.601480\n",
      "2024-08-09 12:33:35.613817\n",
      "2024-08-09 12:33:35.625423\n",
      "2024-08-09 12:33:35.635034\n",
      "2024-08-09 12:33:35.645025\n",
      "2024-08-09 12:33:35.653180\n",
      "2024-08-09 12:33:35.662686\n",
      "2024-08-09 12:33:35.672907\n",
      "2024-08-09 12:33:35.681329\n",
      "2024-08-09 12:33:35.689301\n",
      "2024-08-09 12:33:35.697958\n",
      "2024-08-09 12:33:35.707381\n",
      "2024-08-09 12:33:35.715713\n",
      "2024-08-09 12:33:35.724246\n",
      "2024-08-09 12:33:35.734430\n",
      "2024-08-09 12:33:35.743447\n",
      "2024-08-09 12:33:35.751980\n",
      "2024-08-09 12:33:35.759435\n",
      "2024-08-09 12:33:35.767901\n",
      "2024-08-09 12:33:35.775525\n",
      "2024-08-09 12:33:35.783563\n",
      "2024-08-09 12:33:35.792778\n",
      "2024-08-09 12:33:35.801351\n",
      "2024-08-09 12:33:35.810056\n",
      "2024-08-09 12:33:35.819407\n",
      "2024-08-09 12:33:35.827323\n",
      "2024-08-09 12:33:35.835452\n",
      "2024-08-09 12:33:35.843107\n",
      "2024-08-09 12:33:35.851381\n",
      "2024-08-09 12:33:35.860451\n",
      "2024-08-09 12:33:35.869095\n",
      "2024-08-09 12:33:35.876357\n",
      "2024-08-09 12:33:35.884424\n",
      "2024-08-09 12:33:35.892045\n",
      "2024-08-09 12:33:35.900578\n",
      "2024-08-09 12:33:35.908826\n",
      "2024-08-09 12:33:35.916322\n",
      "2024-08-09 12:33:35.923785\n",
      "df_distance_propagated df unpersisted\n"
     ]
    },
    {
     "ename": "AnalysisException",
     "evalue": "Path does not exist: gs://ot-team/jroldan/2024-08-09_analysis/df_l2g_original/hasGeneticEvidence_Phase4.parquet",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAnalysisException\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_525536/2386158578.py\u001b[0m in \u001b[0;36m<cell line: 243>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    244\u001b[0m     array1 = np.delete(\n\u001b[1;32m    245\u001b[0m         (\n\u001b[0;32m--> 246\u001b[0;31m             \u001b[0mspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparquet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    247\u001b[0m             \u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfull_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"prediction\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"comparison\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"outer\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    248\u001b[0m             \u001b[0;34m.\u001b[0m\u001b[0mgroupBy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"comparison\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/readwriter.py\u001b[0m in \u001b[0;36mparquet\u001b[0;34m(self, *paths, **options)\u001b[0m\n\u001b[1;32m    362\u001b[0m         )\n\u001b[1;32m    363\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 364\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jreader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparquet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_to_seq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_spark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpaths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    365\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    366\u001b[0m     def text(\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1319\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1321\u001b[0;31m         return_value = get_return_value(\n\u001b[0m\u001b[1;32m   1322\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[1;32m   1323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/utils.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    194\u001b[0m                 \u001b[0;31m# Hide where the exception came from that shows a non-Pythonic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m                 \u001b[0;31m# JVM exception message.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mconverted\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAnalysisException\u001b[0m: Path does not exist: gs://ot-team/jroldan/2024-08-09_analysis/df_l2g_original/hasGeneticEvidence_Phase4.parquet"
     ]
    }
   ],
   "source": [
    "def comparisons_df(test_propag) -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    toAnalysis = test_propag.drop(\n",
    "        \"Phase4\",\n",
    "        \"Phase>=3\",\n",
    "        \"Phase>=2\",\n",
    "        \"Phase>=1\",\n",
    "        \"Phase0\",\n",
    "        \"clinicalStatus\",\n",
    "        \"prediction\",\n",
    "        \"count\",\n",
    "        \"PhaseT\",\n",
    "        \"taLabelSimple\",\n",
    "    ).columns[17:]\n",
    "    dataType = [\"byDatatype\"] * len(toAnalysis)\n",
    "    l_studies = []\n",
    "    l_studies.extend([list(a) for a in zip(toAnalysis, dataType)])\n",
    "\n",
    "    schema = StructType(\n",
    "        [\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "            StructField(\"comparisonType\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    comparisons = spark.createDataFrame(l_studies, schema=schema)\n",
    "    ### include all the columns as predictor\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "    wComparison = Window.partitionBy(F.col(comparisonColumn))\n",
    "    wPrediction = Window.partitionBy(F.col(predictionColumn))\n",
    "    wPredictionComparison = Window.partitionBy(\n",
    "        F.col(comparisonColumn), F.col(predictionColumn)\n",
    "    )\n",
    "    \"\"\"\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "    \n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    \"\"\"\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "\n",
    "\n",
    "c = datetime.now()\n",
    "\n",
    "print(\"start doing aggregations and writing\")\n",
    "today_date = str(date.today())\n",
    "listado = []\n",
    "#### Run analysis using loop\n",
    "for key, df in datasetDict.items():\n",
    "    print(key)\n",
    "    df = df.persist()\n",
    "    aggSetups_original = comparisons_df(df)\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "##### read files and make spreadsheet\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio\n",
    "\n",
    "\n",
    "def convertTuple(tup):\n",
    "    st = \",\".join(map(str, tup))\n",
    "    return st\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "#### update the dictionary dfs with other columns included in the analysis\n",
    "key_list = [\n",
    "    \"hasGeneticEvidence\",\n",
    "    \"oneCell\",\n",
    "    \"diagonal\",\n",
    "    \"hasDirectionOfEffect\",\n",
    "    \"max_L2GScore\",\n",
    "    \"min_distance_ranking\",\n",
    "]\n",
    "value_list = [\n",
    "    \"geneticEvidence\",\n",
    "    \"oneCellDoE\",\n",
    "    \"diagonalDoE\",\n",
    "    \"hasDirectionOfEffect\",\n",
    "    \"L2GScore\",\n",
    "    \"TSSDistance\",\n",
    "]\n",
    "\n",
    "dfs = {}\n",
    "\n",
    "\n",
    "def create_dict_column(dfs, key_list, value_list):\n",
    "    if len(key_list) != len(value_list):\n",
    "        raise ValueError(\"lists of different length\")\n",
    "    dfs.update(zip(key_list, value_list))\n",
    "    return dfs\n",
    "\n",
    "\n",
    "dfs = create_dict_column(dfs, key_list, value_list)\n",
    "\n",
    "# Define the lists of possible substrings\n",
    "phase_opt = [\n",
    "    \"Phase4\",\n",
    "    \"Phase>=3\",\n",
    "    \"Phase>=2\",\n",
    "    \"Phase>=1\",\n",
    "    \"PhaseT\",\n",
    "]\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "results = []\n",
    "\n",
    "# Iterate over the sample strings and extract the desired substrings\n",
    "for path in listado:\n",
    "    array1 = np.delete(\n",
    "        (\n",
    "            spark.read.parquet(path)\n",
    "            .join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "            .groupBy(\"comparison\")\n",
    "            .pivot(\"prediction\")\n",
    "            .agg(F.first(\"a\"))\n",
    "            .sort(F.col(\"comparison\").desc())\n",
    "            .select(\"comparison\", \"yes\", \"no\")\n",
    "            .fillna(0)\n",
    "            .toPandas()\n",
    "            .to_numpy()\n",
    "        ),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    # Check for comparison options\n",
    "    for key, value in dfs.items():\n",
    "        if key in path and \"combined\" in path:\n",
    "            if \"oneCell\" in path:\n",
    "                comparison = dfs[key]\n",
    "                group = \"combinedOneCell\"\n",
    "            elif \"diagonal\" in path:\n",
    "                comparison = dfs[key]\n",
    "                group = \"combinedDiagonal\"\n",
    "            elif \"Direction\" in path:\n",
    "                comparison = dfs[key]\n",
    "                group = \"combinedHasDoE\"\n",
    "        elif key in path:\n",
    "            comparison = dfs[key]\n",
    "            group = key\n",
    "\n",
    "    # Check for phase options\n",
    "    for substr in phase_opt:\n",
    "        if substr in path:\n",
    "            phase = substr\n",
    "\n",
    "    if \"original\" in path:\n",
    "        dimension = \"original\"\n",
    "    elif \"propag\" in path:\n",
    "        dimension = \"propagated\"\n",
    "    if \"l2g\" in path:\n",
    "        dataset = \"l2gScore\"\n",
    "    elif \"distance\" in path:\n",
    "        dataset = \"TSSdistance\"\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            group,\n",
    "            dataset,\n",
    "            comparison,\n",
    "            dimension,\n",
    "            phase,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            res_npPhaseX,\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            path,\n",
    "        ]\n",
    "    )\n",
    "    print(path)\n",
    "# Convert the results to a pandas DataFrame\n",
    "df = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"group\",\n",
    "        \"dataset\",\n",
    "        \"comparison\",\n",
    "        \"dimension\",\n",
    "        \"phase\",\n",
    "        \"oddsRatio\",\n",
    "        \"pValue\",\n",
    "        \"lowerInterval\",\n",
    "        \"upperInterval\",\n",
    "        \"total\",\n",
    "        \"values\",\n",
    "        \"relSuccess\",\n",
    "        \"rsLower\",\n",
    "        \"rsUpper\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")\n",
    "df = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"group\",\n",
    "        \"dataset\",\n",
    "        \"comparison\",\n",
    "        \"dimension\",\n",
    "        \"phase\",\n",
    "        \"oddsRatio\",\n",
    "        \"pValue\",\n",
    "        \"lowerInterval\",\n",
    "        \"upperInterval\",\n",
    "        \"total\",\n",
    "        \"values\",\n",
    "        \"relSuccess\",\n",
    "        \"rsLower\",\n",
    "        \"rsUpper\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+-------------------------------------------------+--------------------------------------+------------------------------------------------+-------------------------------------+-----------------------------------------------+-----+\n",
      "|max_L2GScore>=0_5|max_L2GScore>=0_5&diagonalAgreeWithDrugs_combined|max_L2GScore>=0_5&diagonalYes_combined|max_L2GScore>=0_5&oneCellAgreeWithDrugs_combined|max_L2GScore>=0_5&oneCellYes_combined|max_L2GScore>=0_5&hasDirectionOfEffect_combined|count|\n",
      "+-----------------+-------------------------------------------------+--------------------------------------+------------------------------------------------+-------------------------------------+-----------------------------------------------+-----+\n",
      "|               no|                                               no|                                    no|                                              no|                                   no|                                             no|80961|\n",
      "|              yes|                                               no|                                    no|                                              no|                                   no|                                            yes|  302|\n",
      "|              yes|                                              yes|                                   yes|                                              no|                                   no|                                            yes|   43|\n",
      "|              yes|                                              yes|                                   yes|                                             yes|                                  yes|                                            yes|   25|\n",
      "+-----------------+-------------------------------------------------+--------------------------------------+------------------------------------------------+-------------------------------------+-----------------------------------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "datasetDict[f\"df_l2g_original\"].groupBy(\n",
    "    \"max_L2GScore>=0_5\",\n",
    "    \"max_L2GScore>=0_5&diagonalAgreeWithDrugs_combined\",\n",
    "    \"max_L2GScore>=0_5&diagonalYes_combined\",\n",
    "    \"max_L2GScore>=0_5&oneCellAgreeWithDrugs_combined\",\n",
    "    \"max_L2GScore>=0_5&oneCellYes_combined\",\n",
    "    \"max_L2GScore>=0_5&hasDirectionOfEffect_combined\",\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hasGeneticEvidence',\n",
       " 'hasdirectionOfEffect',\n",
       " 'diagonalYes',\n",
       " 'oneCellYes',\n",
       " 'max_L2GScore>=0_1',\n",
       " 'max_L2GScore>=0_15',\n",
       " 'max_L2GScore>=0_2',\n",
       " 'max_L2GScore>=0_25',\n",
       " 'max_L2GScore>=0_3',\n",
       " 'max_L2GScore>=0_35',\n",
       " 'max_L2GScore>=0_4',\n",
       " 'max_L2GScore>=0_45',\n",
       " 'max_L2GScore>=0_5',\n",
       " 'max_L2GScore>=0_55',\n",
       " 'max_L2GScore>=0_6',\n",
       " 'max_L2GScore>=0_65',\n",
       " 'max_L2GScore>=0_7',\n",
       " 'max_L2GScore>=0_75',\n",
       " 'max_L2GScore>=0_8',\n",
       " 'max_L2GScore>=0_85',\n",
       " 'max_L2GScore>=0_9',\n",
       " 'max_L2GScore>=0_95',\n",
       " 'max_L2GScore>=0_1&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_15&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_2&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_25&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_3&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_35&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_4&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_45&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_5&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_55&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_6&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_65&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_7&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_75&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_8&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_85&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_9&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_95&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_1&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_15&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_2&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_25&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_3&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_35&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_4&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_45&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_5&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_55&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_6&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_65&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_7&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_75&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_8&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_85&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_9&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_95&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_1&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_15&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_2&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_25&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_3&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_35&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_4&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_45&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_5&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_55&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_6&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_65&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_7&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_75&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_8&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_85&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_9&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_95&oneCellYes_combined']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasetDict[f\"df_l2g_original\"].drop(\n",
    "    \"Phase4\",\n",
    "    \"Phase>=3\",\n",
    "    \"Phase>=2\",\n",
    "    \"Phase>=1\",\n",
    "    \"Phase0\",\n",
    "    \"clinicalStatus\",\n",
    "    \"prediction\",\n",
    "    \"count\",\n",
    "    \"PhaseT\",\n",
    "    \"taLabelSimple\",\n",
    ").columns[17:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['diseaseId',\n",
       " 'targetId',\n",
       " 'max_L2GScore',\n",
       " 'coherencyDiagonal',\n",
       " 'coherencyOneCell',\n",
       " 'LoF_protect',\n",
       " 'GoF_protect',\n",
       " 'LoF_risk',\n",
       " 'GoF_risk',\n",
       " 'maxClinPhase',\n",
       " 'coherencyDiagonal_ch',\n",
       " 'coherencyOneCell_ch',\n",
       " 'LoF_protect_ch',\n",
       " 'GoF_protect_ch',\n",
       " 'geneticEvidence',\n",
       " 'diagonalAgreeWithDrugs',\n",
       " 'oneCellAgreeWithDrugs',\n",
       " 'Phase4',\n",
       " 'Phase>=3',\n",
       " 'Phase>=2',\n",
       " 'Phase>=1',\n",
       " 'Phase0',\n",
       " 'clinicalStatus',\n",
       " 'prediction',\n",
       " 'PhaseT',\n",
       " 'taLabelSimple',\n",
       " 'hasGeneticEvidence',\n",
       " 'hasdirectionOfEffect',\n",
       " 'diagonalYes',\n",
       " 'oneCellYes',\n",
       " 'max_L2GScore>=0_1',\n",
       " 'max_L2GScore>=0_15',\n",
       " 'max_L2GScore>=0_2',\n",
       " 'max_L2GScore>=0_25',\n",
       " 'max_L2GScore>=0_3',\n",
       " 'max_L2GScore>=0_35',\n",
       " 'max_L2GScore>=0_4',\n",
       " 'max_L2GScore>=0_45',\n",
       " 'max_L2GScore>=0_5',\n",
       " 'max_L2GScore>=0_55',\n",
       " 'max_L2GScore>=0_6',\n",
       " 'max_L2GScore>=0_65',\n",
       " 'max_L2GScore>=0_7',\n",
       " 'max_L2GScore>=0_75',\n",
       " 'max_L2GScore>=0_8',\n",
       " 'max_L2GScore>=0_85',\n",
       " 'max_L2GScore>=0_9',\n",
       " 'max_L2GScore>=0_95',\n",
       " 'max_L2GScore>=0_1&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_15&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_2&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_25&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_3&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_35&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_4&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_45&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_5&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_55&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_6&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_65&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_7&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_75&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_8&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_85&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_9&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_95&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_1&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_15&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_2&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_25&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_3&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_35&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_4&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_45&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_5&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_55&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_6&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_65&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_7&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_75&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_8&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_85&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_9&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_95&diagonalAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_1&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_15&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_2&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_25&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_3&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_35&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_4&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_45&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_5&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_55&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_6&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_65&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_7&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_75&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_8&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_85&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_9&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_95&oneCellAgreeWithDrugs_combined',\n",
       " 'max_L2GScore>=0_1&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_15&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_2&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_25&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_3&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_35&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_4&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_45&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_5&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_55&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_6&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_65&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_7&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_75&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_8&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_85&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_9&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_95&hasDirectionOfEffect_combined',\n",
       " 'max_L2GScore>=0_1&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_15&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_2&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_25&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_3&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_35&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_4&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_45&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_5&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_55&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_6&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_65&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_7&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_75&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_8&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_85&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_9&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_95&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_1&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_15&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_2&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_25&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_3&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_35&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_4&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_45&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_5&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_55&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_6&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_65&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_7&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_75&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_8&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_85&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_9&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_95&oneCellYes_combined']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasetDict[f\"df_l2g_original\"].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'dict_comb' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_525536/1489734980.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdict_comb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'dict_comb' is not defined"
     ]
    }
   ],
   "source": [
    "for a, x in dict_comb.items():\n",
    "    print(a, x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+--------------------+-----------+----------+-----+\n",
      "|hasGeneticEvidence|hasdirectionOfEffect|diagonalYes|oneCellYes|count|\n",
      "+------------------+--------------------+-----------+----------+-----+\n",
      "|                no|                  no|         no|        no|80653|\n",
      "|               yes|                 yes|         no|        no|  575|\n",
      "|               yes|                 yes|        yes|        no|   63|\n",
      "|               yes|                 yes|        yes|       yes|   40|\n",
      "+------------------+--------------------+-----------+----------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### compare whether the column of direction of effect has the right information\n",
    "\n",
    "datasetDict[f\"df_l2g_original\"].groupBy(\n",
    "    \"hasGeneticEvidence\", \"hasdirectionOfEffect\", \"diagonalYes\", \"oneCellYes\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-13 21:36:40.200693\n",
      "loading sources\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/13 21:36:44 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/13 21:36:44 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/13 21:36:45 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/13 21:36:46 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/13 21:36:46 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/13 21:36:46 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/13 21:36:46 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "#####\n",
    "#######\n",
    "## ANALYSIS FOR L2G Scores, genetic evidence and Direction of Effect\n",
    "## Original, propagated and Other Vs Oncology\n",
    "#######\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "### Now , filter by rank, and join with the info from Ot genetics and run the DoE.\n",
    "ranking = Window.partitionBy(\"studyId\", \"variantId\")\n",
    "### union with the other datasources\n",
    "platform_v = \"24.06\"\n",
    "\n",
    "target_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/targets/\"\n",
    ")\n",
    "target = spark.read.parquet(target_path)\n",
    "\n",
    "disease_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/diseases/\"\n",
    ")\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "mecact_path = f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/mechanismOfAction/\"\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "evidences = spark.read.parquet(\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/evidence\"\n",
    ").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "# 1# Make a list of variant of interest (Sequence ontology terms) to subset data of interest.\n",
    "### Bear in mind that SO works with ontology structure as: SO:XXXXXX, but databases has the SO as: SO_XXXXXX\n",
    "var_filter_lof = [\n",
    "    ### High impact variants https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html\n",
    "    \"SO_0001589\",  ## frameshit_variant\n",
    "    \"SO_0001587\",  ## stop_gained\n",
    "    \"SO_0001574\",  ## splice_acceptor_variant\n",
    "    \"SO_0001575\",  ## splice_donor_variant\n",
    "    \"SO_0002012\",  ## start_lost\n",
    "    \"SO_0001578\",  ## stop_lost\n",
    "    \"SO_0001893\",  ## transcript_ablation\n",
    "    # \"SO:0001889\", ## transcript_amplification ## the Only HIGH impact that increase protein.\n",
    "]\n",
    "\n",
    "gof = [\"SO_0002053\"]\n",
    "lof = [\"SO_0002054\"]\n",
    "\n",
    "print(\"loading sources\")\n",
    "\n",
    "## Building Sequence Ontology\n",
    "so_path = \"gs://ot-team/jroldan/sequenceOntology_20221118.csv\"\n",
    "so_ontology = spark.read.csv(so_path, header=True)\n",
    "building = (\n",
    "    so_ontology.select(F.col(\"Accession\"), F.col(\"Parents\"))\n",
    "    .withColumn(\"Parentalind\", F.split(F.col(\"Parents\"), \",\"))\n",
    "    .withColumn(\"Parentalind\", F.explode_outer(\"Parentalind\"))\n",
    "    .groupBy(\"Parentalind\")\n",
    "    .agg(F.collect_list(F.col(\"Accession\")).alias(\"childrens\"))\n",
    "    .join(so_ontology, F.col(\"Parentalind\") == so_ontology.Accession, \"right\")\n",
    ")\n",
    "## annotate TSG/oncogene/bivalent using 'hallmarks.attributes'\n",
    "oncotsg_list = [\n",
    "    \"TSG\",\n",
    "    \"oncogene\",\n",
    "    \"Oncogene\",\n",
    "    \"oncogene\",\n",
    "    \"oncogene,TSG\",\n",
    "    \"TSG,oncogene\",\n",
    "    \"fusion,oncogene\",\n",
    "    \"oncogene,fusion\",\n",
    "]\n",
    "\n",
    "#### rlike('('+Keywords+')(\\s|$)'\n",
    "### on 03.07.2023 we add the categories:\n",
    "# DISRUPTING AGENT - inhibitor\n",
    "# STABILISER - activator\n",
    "\n",
    "### Hacer el join del actionType con el chembl para sacar los mecanismos de accion.\n",
    "inhibitors = [\n",
    "    \"RNAI INHIBITOR\",\n",
    "    \"NEGATIVE MODULATOR\",\n",
    "    \"NEGATIVE ALLOSTERIC MODULATOR\",\n",
    "    \"ANTAGONIST\",\n",
    "    \"ANTISENSE INHIBITOR\",\n",
    "    \"BLOCKER\",\n",
    "    \"INHIBITOR\",\n",
    "    \"DEGRADER\",\n",
    "    \"INVERSE AGONIST\",\n",
    "    \"ALLOSTERIC ANTAGONIST\",\n",
    "    \"DISRUPTING AGENT\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "activators = [\n",
    "    \"PARTIAL AGONIST\",\n",
    "    \"ACTIVATOR\",\n",
    "    \"POSITIVE ALLOSTERIC MODULATOR\",\n",
    "    \"POSITIVE MODULATOR\",\n",
    "    \"AGONIST\",\n",
    "    \"SEQUESTERING AGENT\",\n",
    "    \"STABILISER\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "columnas = [\"activator\", \"inhibitor\"]\n",
    "both = activators + inhibitors\n",
    "\n",
    "actionType = (\n",
    "    mecact.select(\n",
    "        F.explode_outer(\"chemblIds\").alias(\"drugId2\"),\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "        \"targets\",\n",
    "    )\n",
    "    .select(\n",
    "        F.explode_outer(\"targets\").alias(\"targetId2\"),\n",
    "        \"drugId2\",\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "    )\n",
    "    .groupBy(\"targetId2\", \"drugId2\")\n",
    "    .agg(\n",
    "        F.collect_set(\"actionType\").alias(\"actionType\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "oncolabel = (\n",
    "    target.select(\n",
    "        \"id\", \"approvedSymbol\", F.explode_outer(F.col(\"hallmarks.attributes\"))\n",
    "    )\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.description\")\n",
    "    .filter(F.col(\"description\").isin(oncotsg_list))\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"description\").alias(\"description\"))\n",
    "    .withColumn(\"description_splited\", F.concat_ws(\",\", F.col(\"description\")))\n",
    "    .withColumn(\n",
    "        \"TSorOncogene\",\n",
    "        F.when(\n",
    "            (\n",
    "                F.col(\"description_splited\").rlike(\"ncogene\")\n",
    "                & F.col(\"description_splited\").rlike(\"TSG\")\n",
    "            ),\n",
    "            F.lit(\"bivalent\"),\n",
    "        )\n",
    "        .when(F.col(\"description_splited\").rlike(\"ncogene(\\s|$)\"), F.lit(\"oncogene\"))\n",
    "        .when(F.col(\"description_splited\").rlike(\"TSG(\\s|$)\"), F.lit(\"TSG\"))\n",
    "        .otherwise(F.lit(\"noEvaluable\")),  ####\n",
    "    )\n",
    "    .withColumnRenamed(\"id\", \"target_id\")\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "\n",
    "windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "sincgc = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "germline = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"chembl\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "#### version all gene burden\n",
    "prueba_assessment = (\n",
    "    directionOfEffect(evidences, platform_v)\n",
    "    .withColumn(\n",
    "        \"rank\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"average\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "genEvidDataset = (\n",
    "    prueba_assessment.filter(F.col(\"datasourceId\") != \"chembl\")  #### checked 31.05.2023\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.count(\"targetId\").alias(\"Nr_evidences\"))\n",
    "    .select(\"targetId\", \"diseaseId\", \"Nr_evidences\")\n",
    "    .withColumn(\"geneticEvidence\", F.lit(\"hasGeneticEvidence\"))\n",
    "    .drop(\"Nr_evidences\")\n",
    ")\n",
    "\n",
    "coherency_toAssess_others_datasource = (  #### checked 31.05.2023\n",
    "    prueba_assessment.filter(\n",
    "        (F.col(\"homogenized\").isin(columns)) & (F.col(\"datasourceId\") != \"chembl\")\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"datasourceId\").alias(\"datasourceIds\"))\n",
    ")\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "v2g = spark.read.parquet(\"gs://genetics-portal-dev-data/22.09.1/outputs/v2g\")\n",
    "varDistToGene = v2g.select(\n",
    "    F.concat_ws(\"_\", \"chr_id\", \"position\", \"ref_allele\", \"alt_allele\").alias(\n",
    "        \"variantId\"\n",
    "    ),\n",
    "    F.col(\"gene_id\").alias(\"targetId\"),\n",
    "    \"source_id\",\n",
    "    \"d\",\n",
    "    \"distance_score\",\n",
    ").filter(F.col(\"source_id\") == \"canonical_tss\")\n",
    "\n",
    "ranking = Window.partitionBy(\"studyId\", \"variantId\")\n",
    "\n",
    "\n",
    "#######\n",
    "# Build Ot genetics dataset as supporting evidence\n",
    "#######\n",
    "otGenetics = (\n",
    "    prueba_assessment.filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    # .filter((F.col(\"homogenized\") != \"noEvaluable\"))\n",
    "    .join(varDistToGene, on=[\"variantId\", \"targetId\"], how=\"left\")\n",
    "    .join(genEvidDataset, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.collect_set(\"datasourceId\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"L2G_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageL2G\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageCanonicalTSSDistance\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"d\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.when(\n",
    "            F.col(\"rank\").isNull(),\n",
    "            F.array_remove(F.col(\"datasources\"), \"ot_genetics_portal\"),\n",
    "        ).otherwise(F.col(\"datasources\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"distance_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"d\").asc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"ChemblL2gRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"L2G_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"chemblDistanceRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"distance_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"frontierValue\",\n",
    "        ## ot genetics portal\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"beta\") <= 0.1) & (F.col(\"beta\") >= -0.1),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"OddsRatio\") <= 1.1) & (F.col(\"OddsRatio\") >= 0.9),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.lit(\"noValue\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "#####\n",
    "# function for interpreting DoE and coherencies/discrepancies\n",
    "#####\n",
    "\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "analysis_chembl = discrepancifier(\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\") == \"chembl\"))\n",
    "    .withColumn(\n",
    "        \"maxClinPhase\",\n",
    "        F.max(F.col(\"clinicalPhase\")).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "    .pivot(\"homogenized\")\n",
    "    .agg(F.count(\"targetId\"))\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### propag OtGenetics:\n",
    "otGenetics_propag = (\n",
    "    otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "    .join(\n",
    "        diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "        on=\"diseaseId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "    .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    ").persist()\n",
    "\n",
    "\n",
    "#### include dictionary for calling dataframes:\n",
    "# max_L2GScore\n",
    "# min_distance_ranking\n",
    "\n",
    "\n",
    "def benchmarkOT(discrepancifier, otGenetics, metric):\n",
    "    dict_comb = {}\n",
    "    dict_comb = {\n",
    "        \"hasDirectionOfEffect\": f\"{metric}\",\n",
    "        \"diagonalYes\": f\"{metric}\",\n",
    "        \"oneCellYes\": f\"{metric}\",\n",
    "    }\n",
    "    list_l2g = [\n",
    "        0.1,\n",
    "        0.15,\n",
    "        0.2,\n",
    "        0.25,\n",
    "        0.3,\n",
    "        0.35,\n",
    "        0.4,\n",
    "        0.45,\n",
    "        0.5,\n",
    "        0.55,\n",
    "        0.6,\n",
    "        0.65,\n",
    "        0.7,\n",
    "        0.75,\n",
    "        0.8,\n",
    "        0.85,\n",
    "        0.9,\n",
    "        0.95,\n",
    "    ]\n",
    "    list_dist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
    "    return (\n",
    "        discrepancifier(\n",
    "            otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "            .withColumn(\n",
    "                \"min_distance_ranking\",\n",
    "                F.min(\"distance_ranking\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(  ### take maximum L2G score per T-D\n",
    "                \"max_L2GScore\",\n",
    "                F.max(\"resourceScore\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .groupBy(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"geneticEvidence\",\n",
    "                f\"{value}\",\n",
    "            )  ##### modifications here to include the groups of ranking/distances to TSS\n",
    "            .pivot(\"homogenized\")\n",
    "            .agg(F.count(\"targetId\"))\n",
    "        )\n",
    "        .selectExpr(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "            \"geneticEvidence\",\n",
    "            f\"{metric}\",\n",
    "            \"coherencyDiagonal as coherencyDiagonal\",\n",
    "            \"coherencyOneCell as coherencyOneCell\",\n",
    "            \"LoF_protect\",\n",
    "            \"GoF_protect\",\n",
    "            \"LoF_risk\",\n",
    "            \"GoF_risk\",\n",
    "        )\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        # .withColumn(\n",
    "        #    \"geneticEvidence\",\n",
    "        #    F.when(\n",
    "        #        F.col(f\"{metric}\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "        #    ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        # )\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(  #### new column to modify it\n",
    "            \"hasdirectionOfEffect\",\n",
    "            F.when(F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"yes\")).otherwise(\n",
    "                F.lit(\"no\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .select(\n",
    "            [\"*\"]\n",
    "            + (\n",
    "                [  ### single columns\n",
    "                    F.when(F.col(f\"{metric}\") >= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}>={str(n).replace('.', '_')}\")\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"  # Adjust this condition as needed\n",
    "                else [\n",
    "                    F.when(F.col(f\"{metric}\") <= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}<={n}\")\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "            + (\n",
    "                [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"\n",
    "                else [\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) <= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}<={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .persist()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "metric_list = [\"max_L2GScore\", \"min_distance_ranking\"]\n",
    "\n",
    "datasetDict = {}\n",
    "for value in metric_list:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_l2g_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )\n",
    "    elif value == \"min_distance_ranking\":\n",
    "        datasetDict[f\"df_distance_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_distance_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-14 11:36:51.363460\n",
      "loading sources\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/14 11:36:55 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/14 11:36:55 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/14 11:36:56 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/14 11:36:57 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/14 11:36:57 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/14 11:36:57 WARN CacheManager: Asked to cache already cached data.\n",
      "24/08/14 11:36:57 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "#########\n",
    "#####\n",
    "#######\n",
    "## ANALYSIS FOR L2G Scores, genetic evidence and Direction of Effect\n",
    "## Original, propagated and Other Vs Oncology\n",
    "#######\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "### Now , filter by rank, and join with the info from Ot genetics and run the DoE.\n",
    "ranking = Window.partitionBy(\"studyId\", \"variantId\")\n",
    "### union with the other datasources\n",
    "platform_v = \"24.06\"\n",
    "\n",
    "target_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/targets/\"\n",
    ")\n",
    "target = spark.read.parquet(target_path)\n",
    "\n",
    "disease_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/diseases/\"\n",
    ")\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "mecact_path = f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/mechanismOfAction/\"\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "evidences = spark.read.parquet(\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/evidence\"\n",
    ").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "# 1# Make a list of variant of interest (Sequence ontology terms) to subset data of interest.\n",
    "### Bear in mind that SO works with ontology structure as: SO:XXXXXX, but databases has the SO as: SO_XXXXXX\n",
    "var_filter_lof = [\n",
    "    ### High impact variants https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html\n",
    "    \"SO_0001589\",  ## frameshit_variant\n",
    "    \"SO_0001587\",  ## stop_gained\n",
    "    \"SO_0001574\",  ## splice_acceptor_variant\n",
    "    \"SO_0001575\",  ## splice_donor_variant\n",
    "    \"SO_0002012\",  ## start_lost\n",
    "    \"SO_0001578\",  ## stop_lost\n",
    "    \"SO_0001893\",  ## transcript_ablation\n",
    "    # \"SO:0001889\", ## transcript_amplification ## the Only HIGH impact that increase protein.\n",
    "]\n",
    "\n",
    "gof = [\"SO_0002053\"]\n",
    "lof = [\"SO_0002054\"]\n",
    "\n",
    "print(\"loading sources\")\n",
    "\n",
    "## Building Sequence Ontology\n",
    "so_path = \"gs://ot-team/jroldan/sequenceOntology_20221118.csv\"\n",
    "so_ontology = spark.read.csv(so_path, header=True)\n",
    "building = (\n",
    "    so_ontology.select(F.col(\"Accession\"), F.col(\"Parents\"))\n",
    "    .withColumn(\"Parentalind\", F.split(F.col(\"Parents\"), \",\"))\n",
    "    .withColumn(\"Parentalind\", F.explode_outer(\"Parentalind\"))\n",
    "    .groupBy(\"Parentalind\")\n",
    "    .agg(F.collect_list(F.col(\"Accession\")).alias(\"childrens\"))\n",
    "    .join(so_ontology, F.col(\"Parentalind\") == so_ontology.Accession, \"right\")\n",
    ")\n",
    "## annotate TSG/oncogene/bivalent using 'hallmarks.attributes'\n",
    "oncotsg_list = [\n",
    "    \"TSG\",\n",
    "    \"oncogene\",\n",
    "    \"Oncogene\",\n",
    "    \"oncogene\",\n",
    "    \"oncogene,TSG\",\n",
    "    \"TSG,oncogene\",\n",
    "    \"fusion,oncogene\",\n",
    "    \"oncogene,fusion\",\n",
    "]\n",
    "\n",
    "#### rlike('('+Keywords+')(\\s|$)'\n",
    "### on 03.07.2023 we add the categories:\n",
    "# DISRUPTING AGENT - inhibitor\n",
    "# STABILISER - activator\n",
    "\n",
    "### Hacer el join del actionType con el chembl para sacar los mecanismos de accion.\n",
    "inhibitors = [\n",
    "    \"RNAI INHIBITOR\",\n",
    "    \"NEGATIVE MODULATOR\",\n",
    "    \"NEGATIVE ALLOSTERIC MODULATOR\",\n",
    "    \"ANTAGONIST\",\n",
    "    \"ANTISENSE INHIBITOR\",\n",
    "    \"BLOCKER\",\n",
    "    \"INHIBITOR\",\n",
    "    \"DEGRADER\",\n",
    "    \"INVERSE AGONIST\",\n",
    "    \"ALLOSTERIC ANTAGONIST\",\n",
    "    \"DISRUPTING AGENT\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "activators = [\n",
    "    \"PARTIAL AGONIST\",\n",
    "    \"ACTIVATOR\",\n",
    "    \"POSITIVE ALLOSTERIC MODULATOR\",\n",
    "    \"POSITIVE MODULATOR\",\n",
    "    \"AGONIST\",\n",
    "    \"SEQUESTERING AGENT\",\n",
    "    \"STABILISER\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "columnas = [\"activator\", \"inhibitor\"]\n",
    "both = activators + inhibitors\n",
    "\n",
    "actionType = (\n",
    "    mecact.select(\n",
    "        F.explode_outer(\"chemblIds\").alias(\"drugId2\"),\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "        \"targets\",\n",
    "    )\n",
    "    .select(\n",
    "        F.explode_outer(\"targets\").alias(\"targetId2\"),\n",
    "        \"drugId2\",\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "    )\n",
    "    .groupBy(\"targetId2\", \"drugId2\")\n",
    "    .agg(\n",
    "        F.collect_set(\"actionType\").alias(\"actionType\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "oncolabel = (\n",
    "    target.select(\n",
    "        \"id\", \"approvedSymbol\", F.explode_outer(F.col(\"hallmarks.attributes\"))\n",
    "    )\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.description\")\n",
    "    .filter(F.col(\"description\").isin(oncotsg_list))\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"description\").alias(\"description\"))\n",
    "    .withColumn(\"description_splited\", F.concat_ws(\",\", F.col(\"description\")))\n",
    "    .withColumn(\n",
    "        \"TSorOncogene\",\n",
    "        F.when(\n",
    "            (\n",
    "                F.col(\"description_splited\").rlike(\"ncogene\")\n",
    "                & F.col(\"description_splited\").rlike(\"TSG\")\n",
    "            ),\n",
    "            F.lit(\"bivalent\"),\n",
    "        )\n",
    "        .when(F.col(\"description_splited\").rlike(\"ncogene(\\s|$)\"), F.lit(\"oncogene\"))\n",
    "        .when(F.col(\"description_splited\").rlike(\"TSG(\\s|$)\"), F.lit(\"TSG\"))\n",
    "        .otherwise(F.lit(\"noEvaluable\")),  ####\n",
    "    )\n",
    "    .withColumnRenamed(\"id\", \"target_id\")\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "\n",
    "windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "sincgc = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "germline = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"chembl\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "#### version all gene burden\n",
    "prueba_assessment = (\n",
    "    directionOfEffect(evidences, platform_v)\n",
    "    .withColumn(\n",
    "        \"rank\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"average\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "genEvidDataset = (\n",
    "    prueba_assessment.filter(F.col(\"datasourceId\") != \"chembl\")  #### checked 31.05.2023\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.count(\"targetId\").alias(\"Nr_evidences\"))\n",
    "    .select(\"targetId\", \"diseaseId\", \"Nr_evidences\")\n",
    "    .withColumn(\"geneticEvidence\", F.lit(\"hasGeneticEvidence\"))\n",
    "    .drop(\"Nr_evidences\")\n",
    ")\n",
    "\n",
    "coherency_toAssess_others_datasource = (  #### checked 31.05.2023\n",
    "    prueba_assessment.filter(\n",
    "        (F.col(\"homogenized\").isin(columns)) & (F.col(\"datasourceId\") != \"chembl\")\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"datasourceId\").alias(\"datasourceIds\"))\n",
    ")\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "v2g = spark.read.parquet(\"gs://genetics-portal-dev-data/22.09.1/outputs/v2g\")\n",
    "varDistToGene = v2g.select(\n",
    "    F.concat_ws(\"_\", \"chr_id\", \"position\", \"ref_allele\", \"alt_allele\").alias(\n",
    "        \"variantId\"\n",
    "    ),\n",
    "    F.col(\"gene_id\").alias(\"targetId\"),\n",
    "    \"source_id\",\n",
    "    \"d\",\n",
    "    \"distance_score\",\n",
    ").filter(F.col(\"source_id\") == \"canonical_tss\")\n",
    "\n",
    "ranking = Window.partitionBy(\"studyId\", \"variantId\")\n",
    "\n",
    "\n",
    "#######\n",
    "# Build Ot genetics dataset as supporting evidence\n",
    "#######\n",
    "otGenetics = (\n",
    "    prueba_assessment.filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    # .filter((F.col(\"homogenized\") != \"noEvaluable\"))\n",
    "    .join(varDistToGene, on=[\"variantId\", \"targetId\"], how=\"left\")\n",
    "    .join(genEvidDataset, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.collect_set(\"datasourceId\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"L2G_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageL2G\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageCanonicalTSSDistance\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"d\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.when(\n",
    "            F.col(\"rank\").isNull(),\n",
    "            F.array_remove(F.col(\"datasources\"), \"ot_genetics_portal\"),\n",
    "        ).otherwise(F.col(\"datasources\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"distance_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"d\").asc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"ChemblL2gRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"L2G_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"chemblDistanceRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"distance_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"frontierValue\",\n",
    "        ## ot genetics portal\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"beta\") <= 0.1) & (F.col(\"beta\") >= -0.1),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"OddsRatio\") <= 1.1) & (F.col(\"OddsRatio\") >= 0.9),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.lit(\"noValue\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "#####\n",
    "# function for interpreting DoE and coherencies/discrepancies\n",
    "#####\n",
    "\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "analysis_chembl = discrepancifier(\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\") == \"chembl\"))\n",
    "    .withColumn(\n",
    "        \"maxClinPhase\",\n",
    "        F.max(F.col(\"clinicalPhase\")).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "    .pivot(\"homogenized\")\n",
    "    .agg(F.count(\"targetId\"))\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### propag OtGenetics:\n",
    "otGenetics_propag = (\n",
    "    otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "    .join(\n",
    "        diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "        on=\"diseaseId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "    .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    ").persist()\n",
    "\n",
    "\n",
    "#### include dictionary for calling dataframes:\n",
    "# max_L2GScore\n",
    "# min_distance_ranking\n",
    "\n",
    "\n",
    "def benchmarkOT(discrepancifier, otGenetics, metric):\n",
    "    dict_comb = {}\n",
    "    dict_comb = {\n",
    "        \"hasDirectionOfEffect\": f\"{metric}\",\n",
    "        \"diagonalYes\": f\"{metric}\",\n",
    "        \"oneCellYes\": f\"{metric}\",\n",
    "        \"L2GbutNotDoE\": f\"{metric}\",\n",
    "    }\n",
    "    list_l2g = [\n",
    "        0.1,\n",
    "        0.15,\n",
    "        0.2,\n",
    "        0.25,\n",
    "        0.3,\n",
    "        0.35,\n",
    "        0.4,\n",
    "        0.45,\n",
    "        0.5,\n",
    "        0.55,\n",
    "        0.6,\n",
    "        0.65,\n",
    "        0.7,\n",
    "        0.75,\n",
    "        0.8,\n",
    "        0.85,\n",
    "        0.9,\n",
    "        0.95,\n",
    "    ]\n",
    "    list_dist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
    "    return (\n",
    "        discrepancifier(\n",
    "            otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "            .withColumn(\n",
    "                \"min_distance_ranking\",\n",
    "                F.min(\"distance_ranking\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(  ### take maximum L2G score per T-D\n",
    "                \"max_L2GScore\",\n",
    "                F.max(\"resourceScore\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .groupBy(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"geneticEvidence\",\n",
    "                f\"{value}\",\n",
    "            )  ##### modifications here to include the groups of ranking/distances to TSS\n",
    "            .pivot(\"homogenized\")\n",
    "            .agg(F.count(\"targetId\"))\n",
    "        )\n",
    "        .selectExpr(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "            \"geneticEvidence\",\n",
    "            f\"{metric}\",\n",
    "            \"coherencyDiagonal as coherencyDiagonal\",\n",
    "            \"coherencyOneCell as coherencyOneCell\",\n",
    "            \"LoF_protect\",\n",
    "            \"GoF_protect\",\n",
    "            \"LoF_risk\",\n",
    "            \"GoF_risk\",\n",
    "        )\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        # .withColumn(\n",
    "        #    \"geneticEvidence\",\n",
    "        #    F.when(\n",
    "        #        F.col(f\"{metric}\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "        #    ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        # )\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(  #### new column to modify it\n",
    "            \"hasdirectionOfEffect\",\n",
    "            F.when(F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"yes\")).otherwise(\n",
    "                F.lit(\"no\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"L2GAndColoc\",\n",
    "            F.when(\n",
    "                (F.col(\"geneticEvidence\") == \"hasGeneticEvidence\")\n",
    "                & (F.col(\"coherencyDiagonal\").isin([\"coherent\", \"dispar\"])),\n",
    "                F.lit(\"yes\"),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .select(\n",
    "            [\"*\"]\n",
    "            + (\n",
    "                [  ### single columns\n",
    "                    F.when(F.col(f\"{metric}\") >= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}>={str(n).replace('.', '_')}\")\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"  # Adjust this condition as needed\n",
    "                else [\n",
    "                    F.when(F.col(f\"{metric}\") <= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}<={n}\")\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "            + (\n",
    "                [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"\n",
    "                else [\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) <= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}<={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "metric_list = [\"max_L2GScore\", \"min_distance_ranking\"]\n",
    "datasetDict = {}\n",
    "for value in metric_list:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_l2g_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )\n",
    "    elif value == \"min_distance_ranking\":\n",
    "        datasetDict[f\"df_distance_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_distance_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hasGeneticEvidence',\n",
       " 'hasdirectionOfEffect',\n",
       " 'diagonalYes',\n",
       " 'oneCellYes',\n",
       " 'L2GbutNotDoE',\n",
       " 'min_distance_ranking<=1',\n",
       " 'min_distance_ranking<=2',\n",
       " 'min_distance_ranking<=3',\n",
       " 'min_distance_ranking<=4',\n",
       " 'min_distance_ranking<=5',\n",
       " 'min_distance_ranking<=6',\n",
       " 'min_distance_ranking<=7',\n",
       " 'min_distance_ranking<=8',\n",
       " 'min_distance_ranking<=9',\n",
       " 'min_distance_ranking<=10',\n",
       " 'min_distance_ranking<=1&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=2&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=3&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=4&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=5&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=6&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=7&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=8&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=9&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=10&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=1&diagonalYes_combined',\n",
       " 'min_distance_ranking<=2&diagonalYes_combined',\n",
       " 'min_distance_ranking<=3&diagonalYes_combined',\n",
       " 'min_distance_ranking<=4&diagonalYes_combined',\n",
       " 'min_distance_ranking<=5&diagonalYes_combined',\n",
       " 'min_distance_ranking<=6&diagonalYes_combined',\n",
       " 'min_distance_ranking<=7&diagonalYes_combined',\n",
       " 'min_distance_ranking<=8&diagonalYes_combined',\n",
       " 'min_distance_ranking<=9&diagonalYes_combined',\n",
       " 'min_distance_ranking<=10&diagonalYes_combined',\n",
       " 'min_distance_ranking<=1&oneCellYes_combined',\n",
       " 'min_distance_ranking<=2&oneCellYes_combined',\n",
       " 'min_distance_ranking<=3&oneCellYes_combined',\n",
       " 'min_distance_ranking<=4&oneCellYes_combined',\n",
       " 'min_distance_ranking<=5&oneCellYes_combined',\n",
       " 'min_distance_ranking<=6&oneCellYes_combined',\n",
       " 'min_distance_ranking<=7&oneCellYes_combined',\n",
       " 'min_distance_ranking<=8&oneCellYes_combined',\n",
       " 'min_distance_ranking<=9&oneCellYes_combined',\n",
       " 'min_distance_ranking<=10&oneCellYes_combined',\n",
       " 'min_distance_ranking<=1&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=2&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=3&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=4&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=5&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=6&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=7&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=8&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=9&L2GbutNotDoE_combined',\n",
       " 'min_distance_ranking<=10&L2GbutNotDoE_combined']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasetDict[f\"df_distance_propagated\"].drop(\n",
    "    \"Phase4\",\n",
    "    \"Phase>=3\",\n",
    "    \"Phase>=2\",\n",
    "    \"Phase>=1\",\n",
    "    \"Phase0\",\n",
    "    \"clinicalStatus\",\n",
    "    \"prediction\",\n",
    "    \"count\",\n",
    "    \"PhaseT\",\n",
    "    \"taLabelSimple\",\n",
    ").columns[17:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 233:======================================>             (148 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------+-----------------+-----+\n",
      "|L2GData|coherencyDiagonal|count|\n",
      "+-------+-----------------+-----+\n",
      "|  noL2G|             null|80304|\n",
      "| hasL2G|       EvidNotDoE|  766|\n",
      "| hasL2G|         coherent|  248|\n",
      "| hasL2G|           dispar|   13|\n",
      "+-------+-----------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "datasetDict[f\"df_distance_propagated\"].withColumn(\n",
    "    \"L2GData\",\n",
    "    F.when(F.col(\"min_distance_ranking\").isNotNull(), F.lit(\"hasL2G\")).otherwise(\n",
    "        F.lit(\"noL2G\")\n",
    "    ),\n",
    ").groupBy(\"L2GData\", \"coherencyDiagonal\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'dict_comb' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_8354/4115502485.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdict_comb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'dict_comb' is not defined"
     ]
    }
   ],
   "source": [
    "for a, x in dict_comb.items():\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------+------------------+------------+-----------------+-------------------------------------------+-----+\n",
      "|L2GData|   geneticEvidence|L2GbutNotDoE|coherencyDiagonal|min_distance_ranking<=6&oneCellYes_combined|count|\n",
      "+-------+------------------+------------+-----------------+-------------------------------------------+-----+\n",
      "| hasL2G|hasGeneticEvidence|          no|       EvidNotDoE|                                         no|  766|\n",
      "|  noL2G|              null|          no|             null|                                         no|80304|\n",
      "| hasL2G|hasGeneticEvidence|         yes|         coherent|                                         no|  193|\n",
      "| hasL2G|hasGeneticEvidence|         yes|         coherent|                                        yes|   55|\n",
      "| hasL2G|hasGeneticEvidence|         yes|           dispar|                                         no|   13|\n",
      "+-------+------------------+------------+-----------------+-------------------------------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "datasetDict[f\"df_distance_propagated\"].withColumn(\n",
    "    \"L2GData\",\n",
    "    F.when(F.col(\"min_distance_ranking\").isNotNull(), F.lit(\"hasL2G\")).otherwise(\n",
    "        F.lit(\"noL2G\")\n",
    "    ),\n",
    ").withColumn(\n",
    "    \"L2GbutNotDoE\",\n",
    "    F.when(\n",
    "        (F.col(\"geneticEvidence\") == \"hasGeneticEvidence\")\n",
    "        & (F.col(\"coherencyDiagonal\").isin([\"coherent\", \"dispar\"])),\n",
    "        F.lit(\"yes\"),\n",
    "    ).otherwise(F.lit(\"no\")),\n",
    ").groupBy(\n",
    "    \"L2GData\",\n",
    "    \"geneticEvidence\",\n",
    "    \"L2GbutNotDoE\",\n",
    "    \"coherencyDiagonal\",\n",
    "    \"min_distance_ranking<=6&oneCellYes_combined\",\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+------+\n",
      "|hasScore| count|\n",
      "+--------+------+\n",
      "|     yes|781213|\n",
      "+--------+------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\").withColumn(\n",
    "    \"hasScore\", F.when(F.col(\"score\").isNotNull(), F.lit(\"yes\")).otherwise(F.lit(\"now\"))\n",
    ").groupBy(\"hasScore\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['diseaseId',\n",
       " 'targetId',\n",
       " 'geneticEvidence',\n",
       " 'min_distance_ranking',\n",
       " 'coherencyDiagonal',\n",
       " 'coherencyOneCell',\n",
       " 'LoF_protect',\n",
       " 'GoF_protect',\n",
       " 'LoF_risk',\n",
       " 'GoF_risk',\n",
       " 'maxClinPhase',\n",
       " 'coherencyDiagonal_ch',\n",
       " 'coherencyOneCell_ch',\n",
       " 'LoF_protect_ch',\n",
       " 'GoF_protect_ch',\n",
       " 'diagonalAgreeWithDrugs',\n",
       " 'oneCellAgreeWithDrugs',\n",
       " 'Phase4',\n",
       " 'Phase>=3',\n",
       " 'Phase>=2',\n",
       " 'Phase>=1',\n",
       " 'Phase0',\n",
       " 'clinicalStatus',\n",
       " 'prediction',\n",
       " 'PhaseT',\n",
       " 'taLabelSimple',\n",
       " 'hasGeneticEvidence',\n",
       " 'hasdirectionOfEffect',\n",
       " 'diagonalYes',\n",
       " 'oneCellYes',\n",
       " 'min_distance_ranking<=1',\n",
       " 'min_distance_ranking<=2',\n",
       " 'min_distance_ranking<=3',\n",
       " 'min_distance_ranking<=4',\n",
       " 'min_distance_ranking<=5',\n",
       " 'min_distance_ranking<=6',\n",
       " 'min_distance_ranking<=7',\n",
       " 'min_distance_ranking<=8',\n",
       " 'min_distance_ranking<=9',\n",
       " 'min_distance_ranking<=10',\n",
       " 'min_distance_ranking<=1&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=2&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=3&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=4&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=5&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=6&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=7&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=8&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=9&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=10&hasDirectionOfEffect_combined',\n",
       " 'min_distance_ranking<=1&diagonalYes_combined',\n",
       " 'min_distance_ranking<=2&diagonalYes_combined',\n",
       " 'min_distance_ranking<=3&diagonalYes_combined',\n",
       " 'min_distance_ranking<=4&diagonalYes_combined',\n",
       " 'min_distance_ranking<=5&diagonalYes_combined',\n",
       " 'min_distance_ranking<=6&diagonalYes_combined',\n",
       " 'min_distance_ranking<=7&diagonalYes_combined',\n",
       " 'min_distance_ranking<=8&diagonalYes_combined',\n",
       " 'min_distance_ranking<=9&diagonalYes_combined',\n",
       " 'min_distance_ranking<=10&diagonalYes_combined',\n",
       " 'min_distance_ranking<=1&oneCellYes_combined',\n",
       " 'min_distance_ranking<=2&oneCellYes_combined',\n",
       " 'min_distance_ranking<=3&oneCellYes_combined',\n",
       " 'min_distance_ranking<=4&oneCellYes_combined',\n",
       " 'min_distance_ranking<=5&oneCellYes_combined',\n",
       " 'min_distance_ranking<=6&oneCellYes_combined',\n",
       " 'min_distance_ranking<=7&oneCellYes_combined',\n",
       " 'min_distance_ranking<=8&oneCellYes_combined',\n",
       " 'min_distance_ranking<=9&oneCellYes_combined',\n",
       " 'min_distance_ranking<=10&oneCellYes_combined']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasetDict[f\"df_distance_propagated\"].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### How many T-D have L2G score?\n",
    "### How many T-D overlap with L2G and Coloc/DoE?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### modify the spreadsheet of coloc from 02.08.2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-15 09:00:58.613710\n",
      "Analysis started on 2024-08-15 at  2024-08-15 09:00:58.613710\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/15 09:01:06 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    }
   ],
   "source": [
    "#### now with propagation\n",
    "\"\"\" this scripts run the analysis for comparing QTL studies, tissues together with therapy areas matched\"\"\"\n",
    "from functions import relative_success\n",
    "from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/15 09:07:28 WARN CacheManager: Asked to cache already cached data.        \n",
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "created full_data and lists\n"
     ]
    }
   ],
   "source": [
    "#### now with propagation\n",
    "\"\"\" this scripts run the analysis for comparing QTL studies, tissues together with therapy areas matched\"\"\"\n",
    "from functions import relative_success\n",
    "from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "platform_v = \"24.06\"\n",
    "\n",
    "target_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/targets/\"\n",
    ")\n",
    "target = spark.read.parquet(target_path)\n",
    "\n",
    "disease_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/diseases/\"\n",
    ")\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "\n",
    "evidences = spark.read.parquet(\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/evidence\"\n",
    ")\n",
    "\n",
    "coloc = spark.read.parquet(\n",
    "    \"gs://genetics-portal-dev-data/22.09.1/outputs/v2d_coloc\"\n",
    ").filter(F.col(\"right_type\") != \"gwas\")\n",
    "\n",
    "\n",
    "### laod sample sizes - look at the document\n",
    "sampleSize = spark.read.csv(\"gs://ot-team/jroldan/colocSampleSize.csv\", header=True)\n",
    "\n",
    "terminated = terminated_td(\n",
    "    spark, \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "### load QTL tissues mapped to therapy areas\n",
    "onto_samples = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/20240112_mappedSQLtissuesGTP3_5.csv\", header=True\n",
    ")\n",
    "\n",
    "#### GSEA annotation for hallmark inflamation targets\n",
    "immflam_annot = (\n",
    "    spark.read.json(\n",
    "        \"gs://ot-team/jroldan/analysis/HALLMARK_INFLAMMATORY_RESPONSE.v2023.2.Hs_edited.json\"\n",
    "    )\n",
    "    .select(F.explode_outer(\"geneSymbols\").alias(\"approvedSymbol\"))\n",
    "    .withColumn(\"isInflam\", F.lit(\"yes\"))\n",
    ")\n",
    "\n",
    "#### Build the uniqueBetas_analyse\n",
    "targetType = (\n",
    "    target.select(\"id\", \"approvedSymbol\", F.explode_outer(\"targetClass\"))\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.label\")\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"label\").alias(\"label\"))\n",
    "    .filter(F.col(\"label\").isNotNull())\n",
    "    .selectExpr(\n",
    "        \"id as targetIdtargetType\",\n",
    "        \"approvedSymbol as approvedSymbol\",\n",
    "        \"label as targetType\",\n",
    "    )\n",
    "    .join(immflam_annot, on=\"approvedSymbol\", how=\"left\")\n",
    ")\n",
    "\n",
    "### take ontology of samples\n",
    "samplesOnto = (\n",
    "    onto_samples.withColumn(\n",
    "        \"right_bio_feature\", F.split(F.col(\"original\"), \" - \").getItem(0)\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"therapyArea\", F.split(F.col(\"20231207_curated_simplified\"), \" - \").getItem(2)\n",
    "    )\n",
    "    .withColumn(\"EFO\", F.split(F.col(\"20231207_curated_simplified\"), \" - \").getItem(1))\n",
    "    .withColumn(\n",
    "        \"right_bio_feature2\",\n",
    "        F.split(F.col(\"20231207_curated_simplified\"), \" - \").getItem(0),\n",
    "    )\n",
    "    .drop(\n",
    "        \"curated_simplified\",\n",
    "        \"20231207_curated_simplified\",\n",
    "        \"original\",\n",
    "        \"curated\",\n",
    "        \"_c3\",\n",
    "        \"_c4\",\n",
    "    )\n",
    ")\n",
    "\n",
    "coloc2 = (\n",
    "    coloc.select(\n",
    "        F.concat_ws(\"_\", \"left_chrom\", \"left_pos\", \"left_ref\", \"left_alt\").alias(\n",
    "            \"left_locus_id\"\n",
    "        ),\n",
    "        F.concat_ws(\"_\", \"right_chrom\", \"right_pos\", \"right_ref\", \"right_alt\").alias(\n",
    "            \"right_locus_id\"\n",
    "        ),\n",
    "        F.col(\"left_study\").alias(\"left_study_id\"),\n",
    "        F.col(\"right_study\").alias(\"right_study_id\"),\n",
    "        \"right_gene_id\",\n",
    "        \"coloc_h4\",\n",
    "        \"left_var_right_study_beta\",\n",
    "        \"right_phenotype\",\n",
    "        F.col(\"left_type\"),\n",
    "        F.col(\"right_type\"),\n",
    "        F.col(\"right_bio_feature\"),\n",
    "        F.col(\"is_flipped\"),\n",
    "        \"left_var_right_study_pval\",\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"beta_assessed\",  ### diferent from sQTL and oQTL\n",
    "        F.when(\n",
    "            (F.col(\"left_var_right_study_beta\") > 0)\n",
    "            & (F.col(\"right_study_id\") != \"GTEx-sQTL\"),\n",
    "            F.lit(\"gof\"),\n",
    "        ).when(\n",
    "            (F.col(\"left_var_right_study_beta\") < 0)\n",
    "            & (F.col(\"right_study_id\") != \"GTEx-sQTL\"),\n",
    "            F.lit(\"lof\"),\n",
    "        )\n",
    "        #### for sQTL is the opposite\n",
    "        .when(\n",
    "            (F.col(\"left_var_right_study_beta\") > 0)\n",
    "            & (F.col(\"right_study_id\") == \"GTEx-sQTL\"),\n",
    "            F.lit(\"lof\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"left_var_right_study_beta\") < 0)\n",
    "            & (F.col(\"right_study_id\") == \"GTEx-sQTL\"),\n",
    "            F.lit(\"gof\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"neutral\")),\n",
    "    )\n",
    "    .join(samplesOnto, on=[\"right_bio_feature\"], how=\"left\")\n",
    "    .drop(\"right_bio_feature\")\n",
    ")\n",
    "### check for disparities using count of different assessment for beta for target\n",
    "\n",
    "disparities = coloc2.groupBy(\"left_locus_id\", \"left_study_id\", \"right_gene_id\").agg(\n",
    "    F.size(F.collect_set(\"beta_assessed\")).alias(\"count\"),\n",
    ")\n",
    "### add the label of which left_locus_id,left_study_id and right_gene_id are having contradictions\n",
    "\n",
    "coloc3 = (\n",
    "    coloc2.withColumnRenamed(\"right_bio_feature2\", \"right_bio_feature\")\n",
    "    .join(\n",
    "        disparities, on=[\"left_locus_id\", \"left_study_id\", \"right_gene_id\"], how=\"left\"\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "#### Run directionOfEffect\n",
    "prueba_assessment = directionOfEffect(evidences, platform_v)\n",
    "\n",
    "## add therapeuticArea name to  diseases\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"OTAR_0000020\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "#### prepare for disease propagation, but explode at the end\n",
    "diseases2 = (\n",
    "    diseases.withColumn(\"taId\", F.explode_outer(\"therapeuticAreas\"))\n",
    "    .join(taDf.drop(\"taLabelSimple\"), on=\"taId\", how=\"left\")\n",
    "    .groupBy(\"id\", \"parents\", \"name\")\n",
    "    .agg(\n",
    "        F.collect_set(\"taId\").alias(\"taId\"),\n",
    "        F.collect_set(\"taLabel\").alias(\"diseaseTherapyAreas\"),\n",
    "    )\n",
    ")\n",
    "#### load data from ot genetics with the assessments\n",
    "delimiter = \",\"\n",
    "evidence_ot_genetics = (\n",
    "    (\n",
    "        prueba_assessment.filter((F.col(\"datasourceId\").isin([\"ot_genetics_portal\"])))\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"variantId\", \"studyId\")\n",
    "        .pivot(\"directionOnTrait\")\n",
    "        .count()\n",
    "        .drop(\"noEvaluable\", \"conflict/noEvaluable\")\n",
    "        .persist()\n",
    "    )\n",
    "    .join(diseases2, F.col(\"diseaseId\") == diseases2.id, \"left\")\n",
    "    .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "    .withColumn(\n",
    "        \"diseaseId\",\n",
    "        F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    "    )\n",
    "    .drop(\"id\")\n",
    ")\n",
    "\n",
    "coloc_otgene = (\n",
    "    coloc3.withColumnRenamed(\"left_study_id\", \"studyId\")\n",
    "    .withColumnRenamed(\"left_locus_id\", \"locusId\")\n",
    "    .withColumnRenamed(\"right_gene_id\", \"targetId\")\n",
    "    .join(\n",
    "        evidence_ot_genetics.withColumnRenamed(\"variantId\", \"locusId\"),\n",
    "        on=[\"studyId\", \"locusId\", \"targetId\"],\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .withColumnRenamed(\"GoF\", \"GoF_OT\")  # remove\n",
    "    .withColumnRenamed(\"LoF\", \"LoF_OT\")  # remove\n",
    ").persist()\n",
    "\n",
    "chembl_trials = (\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "chembl = (\n",
    "    (\n",
    "        prueba_assessment.filter(\n",
    "            (F.col(\"datasourceId\").isin([\"chembl\"]))\n",
    "            & (F.col(\"homogenized\").isin([\"noEvaluable\", \"dispar\"]) == False)\n",
    "        )\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .pivot(\"variantEffect\")\n",
    "    .count()\n",
    "    .withColumnRenamed(\"LoF\", \"LoF_Ch\")\n",
    "    .withColumnRenamed(\"GoF\", \"GoF_Ch\")\n",
    "    .join(chembl_trials, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "coloc_bnch = coloc_otgene.join(chembl, on=[\"targetId\", \"diseaseId\"], how=\"inner\")\n",
    "coloc_bnch2 = coloc_bnch.join(\n",
    "    sampleSize.select(\"right_study_id\", \"sampleSize\"), on=\"right_study_id\", how=\"left\"\n",
    ").persist()\n",
    "\n",
    "withEvidence = (\n",
    "    coloc_bnch2.withColumn(\n",
    "        \"ChEMBL\",\n",
    "        F.when(\n",
    "            (F.col(\"GoF_Ch\").isNotNull()) & (F.col(\"LoF_Ch\").isNotNull()),\n",
    "            F.lit(\"gof&lof\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"LoF_Ch\").isNotNull()) & (F.col(\"GoF_Ch\").isNull()),\n",
    "            F.lit(F.lit(\"lof\")),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"GoF_Ch\").isNotNull()) & (F.col(\"LoF_Ch\").isNull()),\n",
    "            F.lit(F.lit(\"gof\")),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Coherency_chembl\",\n",
    "        F.when(  ### there are cases of drug with gof&lof\n",
    "            (F.col(\"protect\").isNotNull()),\n",
    "            F.when(\n",
    "                (F.col(\"beta_assessed\") == \"gof\"),\n",
    "                F.when(\n",
    "                    (F.col(\"GoF_Ch\").isNotNull()) & (F.col(\"LoF_Ch\").isNull()),\n",
    "                    F.lit(\"coherent\"),\n",
    "                ).when(\n",
    "                    (F.col(\"LoF_Ch\").isNotNull()) & (F.col(\"GoF_Ch\").isNull()),\n",
    "                    F.lit(\"dispar\"),\n",
    "                ),\n",
    "            ).when(\n",
    "                (F.col(\"beta_assessed\") == \"lof\"),\n",
    "                F.when(\n",
    "                    (F.col(\"GoF_Ch\").isNotNull()) & (F.col(\"LoF_Ch\").isNull()),\n",
    "                    F.lit(\"dispar\"),\n",
    "                ).when(\n",
    "                    (F.col(\"LoF_Ch\").isNotNull()) & (F.col(\"GoF_Ch\").isNull()),\n",
    "                    F.lit(\"coherent\"),\n",
    "                ),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .join(target.select(\"id\", \"approvedSymbol\"), target.id == F.col(\"targetId\"), \"left\")\n",
    "    .drop(\"id\")\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "uniqueBetas = withEvidence.persist()\n",
    "\n",
    "custom_schema = StructType(\n",
    "    [\n",
    "        StructField(\"_c0\", StringType(), True),\n",
    "        StructField(\"_c1\", StringType(), True),\n",
    "        StructField(\"_c2\", StringType(), True),\n",
    "        StructField(\"_c3\", DecimalType(38, 37), True),\n",
    "        # Add more fields as needed\n",
    "    ]\n",
    ")\n",
    "\n",
    "### windows function for adjusting pvalue using benjamini-hochberg correction\n",
    "window = Window.partitionBy(\"studyId\").orderBy(\"pVal\")\n",
    "window2 = Window.partitionBy(\"studyId\")\n",
    "\n",
    "### read the tissue enrichent's file\n",
    "tissueEnrichment = (\n",
    "    spark.read.csv(\n",
    "        \"gs://ot-team/jroldan/Tissue_enrichment_results/\",\n",
    "        sep=\"\\t\",\n",
    "        schema=custom_schema,\n",
    "    )\n",
    "    .selectExpr(\n",
    "        \"_c0 as studyId\",\n",
    "        \"_c1 as tissueEnriched\",\n",
    "        \"_c3 as pVal\",\n",
    "        #### instead of removing, keeping the ones with pvalues=0 to not perturb the rank\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"pVal\",\n",
    "        F.when(F.col(\"pVal\") == 0e-37, F.lit(1e-37)).otherwise(F.col(\"pVal\")),\n",
    "        ### ranking column\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"index\",\n",
    "        F.row_number().over(window),\n",
    "        ### total number of rows per study column\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"length\",\n",
    "        F.last(\"index\").over(window2),\n",
    "        ### BH adjusted p values calculation\n",
    "    )\n",
    "    .withColumn(\"adjPVal\", (F.col(\"pVal\") * F.col(\"length\")) / F.col(\"index\"))\n",
    ")\n",
    "\n",
    "### read the file with matching tissue enriched with therapy areas\n",
    "tissueEnrichTherAreas = (\n",
    "    spark.read.csv(\n",
    "        \"gs://ot-team/jroldan/analysis/20231207_gwasTissueEnrrichedToTherapyAreas.csv/\",\n",
    "        sep=\",\",\n",
    "        header=True,\n",
    "    )\n",
    "    .withColumnRenamed(\"tissue\", \"tissueEnriched\")\n",
    "    .withColumn(\n",
    "        \"tissueEnrichedTherapyAreas\",\n",
    "        F.array(F.col(\"TherapyArea\"), F.col(\"Alternative\")),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "### prepare the file of tissue enriched with their therapy areas\n",
    "studyEnrichTherArea = (\n",
    "    tissueEnrichment.join(\n",
    "        tissueEnrichTherAreas.select(\"tissueEnriched\", \"tissueEnrichedTherapyAreas\"),\n",
    "        on=\"tissueEnriched\",\n",
    "        how=\"left\",\n",
    "        ## filter by p value <0.05\n",
    "    )\n",
    "    ### filter by adjusted p values below 0.05\n",
    "    .filter(F.col(\"adjPVal\") < 0.05)\n",
    "    .groupBy(\"studyId\")\n",
    "    .agg(\n",
    "        F.array_except(\n",
    "            F.flatten(F.collect_set(\"tissueEnrichedTherapyAreas\")), F.array(F.lit(None))\n",
    "        ).alias(\"studyTherapyArea\"),\n",
    "        F.collect_set(\"tissueEnriched\").alias(\"tissuesEnriched\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "#### mapping right_bio_feature - disease\n",
    "schema_rbf = StructType(\n",
    "    [\n",
    "        StructField(\"remove\", StringType(), True),\n",
    "        StructField(\"right_bio_feature\", StringType(), True),\n",
    "        StructField(\"name\", StringType(), True),\n",
    "        StructField(\"matchRBFToDisease\", StringType(), True),\n",
    "        # Add more fields as needed\n",
    "    ]\n",
    ")\n",
    "\n",
    "### read the tissue enrichent's file\n",
    "tissueToDisease = (\n",
    "    spark.read.csv(\n",
    "        \"gs://ot-team/jroldan/analysis/20240801_newcurationRBFToDisease.csv/\",\n",
    "        header=True,\n",
    "    )\n",
    "    .select(\"name\", \"right_bio_feature\", \"matchRBFToDisease\")\n",
    "    .distinct()\n",
    ")\n",
    "\n",
    "\n",
    "uniqueBetas_location = (\n",
    "    (\n",
    "        target_membrane(spark, target, uniqueBetas).drop(\n",
    "            \"mb\",\n",
    "            \"counted\",\n",
    "            \"HPA_membrane\",\n",
    "            \"HPA_secreted\",\n",
    "            \"uniprot_membrane\",\n",
    "            \"uniprot_secreted\",\n",
    "            \"loc\",\n",
    "            \"location_id\",\n",
    "            # \"result\",\n",
    "            \"loc_id\",\n",
    "        )\n",
    "        # .join(targetType, on=[\"targetId\"], how=\"left\")\n",
    "    )\n",
    "    .join(tissueToDisease, on=[\"name\", \"right_bio_feature\"], how=\"left\")\n",
    "    .withColumn(\"qtlTherapyArea\", F.split(F.col(\"therapyArea\"), \",\"))\n",
    "    .join(studyEnrichTherArea, on=\"studyId\", how=\"left\")\n",
    "    .withColumn(\n",
    "        \"taIntersectQtlDisease\",\n",
    "        F.array_intersect(F.col(\"qtlTherapyArea\"), F.col(\"diseaseTherapyAreas\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"taIntersectQtlTissueEnriched\",\n",
    "        F.array_intersect(F.col(\"qtlTherapyArea\"), F.col(\"studyTherapyArea\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"taNIntersectQtlDisease\",\n",
    "        F.when(\n",
    "            F.size(\n",
    "                F.array_intersect(F.col(\"qtlTherapyArea\"), F.col(\"diseaseTherapyAreas\"))\n",
    "            )\n",
    "            >= 1,\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"taNIntersectQtlTissueEnriched\",\n",
    "        F.when(\n",
    "            F.size(\n",
    "                F.array_intersect(F.col(\"qtlTherapyArea\"), F.col(\"studyTherapyArea\"))\n",
    "            )\n",
    "            >= 1,\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .join(targetType, on=\"approvedSymbol\", how=\"left\")\n",
    ")\n",
    "\n",
    "\n",
    "\"\"\"single column comparisons\"\"\"\n",
    "dfs = {}\n",
    "claves = []\n",
    "columns = []\n",
    "list_columns = [\n",
    "    \"right_bio_feature\",\n",
    "    \"right_type\",\n",
    "    \"beta_assessed\",\n",
    "    \"right_study_id\",\n",
    "    \"therapyArea\",\n",
    "]\n",
    "\n",
    "for x in list_columns:\n",
    "    a = [data[0] for data in uniqueBetas_location.select(x).distinct().collect()]\n",
    "    for y in a:\n",
    "        claves.append(y)\n",
    "        columns.append(x)\n",
    "#### remove the \".0\" that is problematic\n",
    "claves2 = [str(s).replace(\".0\", \"\") for s in claves]\n",
    "\n",
    "for key, value in zip(claves2, columns):\n",
    "    dfs[key] = value\n",
    "\n",
    "\"\"\"combined column comparisons\"\"\"\n",
    "list1 = []\n",
    "list2 = []\n",
    "cols = uniqueBetas_location.drop(\"Coherency_geneBurden\").columns\n",
    "\n",
    "### make list per column to be combined\n",
    "study_column = [\n",
    "    str(row.right_study_id)\n",
    "    for row in uniqueBetas_location.filter(F.col(\"right_study_id\").isNotNull())\n",
    "    .select(\"right_study_id\")\n",
    "    .distinct()\n",
    "    .collect()\n",
    "]\n",
    "\n",
    "qtl_type = [\n",
    "    str(row.right_type)\n",
    "    for row in uniqueBetas_location.filter(F.col(\"right_type\").isNotNull())\n",
    "    .select(\"right_type\")\n",
    "    .distinct()\n",
    "    .collect()  ## there are \"Nulls\" in this column\n",
    "]\n",
    "\n",
    "biofeature_column = [\n",
    "    str(row.right_bio_feature)\n",
    "    for row in uniqueBetas_location.filter(F.col(\"right_bio_feature\").isNotNull())\n",
    "    .select(\"right_bio_feature\")\n",
    "    .distinct()\n",
    "    .collect()\n",
    "]\n",
    "\n",
    "### make a dictionary with column combinations study & qtl\n",
    "for x in study_column:\n",
    "    for y in qtl_type:\n",
    "        list1.append(x)\n",
    "        list2.append(y)\n",
    "\n",
    "### make a dictionary with column combinations biofeature & qtl\n",
    "for x in biofeature_column:\n",
    "    for y in qtl_type:\n",
    "        list1.append(x)\n",
    "        list2.append(y)\n",
    "\n",
    "combined_lists = [list(pair) for pair in zip(list1, list2)]\n",
    "\n",
    "\n",
    "## collect all columns in an array\n",
    "### drop Coherency_geneBurden because it contains \"coherent\" and \"dispar\" words\n",
    "def dict_comb_comp(combined_lists, study_column, qtl_type):\n",
    "    new2 = {}\n",
    "    for x, n in combined_lists:\n",
    "        if x in study_column:\n",
    "            if n in qtl_type:\n",
    "                new2.update(\n",
    "                    {f\"{x}&{n}\": {\"right_type\": {f\"{n}\": {\"right_study_id\": f\"{x}\"}}}}\n",
    "                )\n",
    "    return new2\n",
    "\n",
    "\n",
    "new2 = dict_comb_comp(combined_lists, study_column, qtl_type)\n",
    "\n",
    "df_string = (\n",
    "    uniqueBetas_location.drop(\"Coherency_geneBurden\")\n",
    "    .select([F.col(col_name).cast(\"string\").alias(col_name) for col_name in cols])\n",
    "    .withColumn(\"array\", F.array(cols))\n",
    ")\n",
    "\n",
    "\"\"\"make the dataframe with column combinations\"\"\"\n",
    "df_string2 = df_string.select(\n",
    "    [\"*\"]\n",
    "    + [\n",
    "        F.when(\n",
    "            (F.col(i) == b)\n",
    "            & (F.col(c) == t)\n",
    "            & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "            F.lit(\"yes\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"no\"))\n",
    "        .alias(a)\n",
    "        for a, x in new2.items()\n",
    "        for i, z in x.items()\n",
    "        for b, y in z.items()\n",
    "        for c, t in y.items()\n",
    "        # print(a, i, b, c, t)\n",
    "    ]\n",
    ").persist()\n",
    "\n",
    "delimiter = \",\"  ### to convert string of propagated traits to array\n",
    "\n",
    "uniqueBetas_analyse = (\n",
    "    df_string2.select(\n",
    "        [\"*\"]\n",
    "        + [  #### only doe\n",
    "            F.when(\n",
    "                ((f\"{x}\") == F.col(c)) & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"no\"))\n",
    "            .alias(f\"{x}_{c}\")\n",
    "            for x, c in dfs.items()\n",
    "        ]\n",
    "    )\n",
    "    .join(terminated, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "    .withColumn(\n",
    "        \"Phase4\",\n",
    "        F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=3\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=2\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=1\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase0\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"allPhases\",\n",
    "        F.when(F.col(\"maxClinPhase\").isNotNull(), F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"PhaseT\",\n",
    "        F.when(F.col(\"prediction\").isNotNull(), F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"coloc>_60\",\n",
    "        F.when(\n",
    "            (F.col(\"coloc_h4\") >= 0.60) & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "            F.lit(\"yes\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") >= 0.60) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") < 0.60) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\").isNull()) | (F.col(\"Coherency_chembl\").isNull()),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"coloc>_80\",\n",
    "        F.when(\n",
    "            (F.col(\"coloc_h4\") >= 0.80) & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "            F.lit(\"yes\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") >= 0.80) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") < 0.80) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\").isNull()) | (F.col(\"Coherency_chembl\").isNull()),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"coloc>_85\",\n",
    "        F.when(\n",
    "            (F.col(\"coloc_h4\") >= 0.85) & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "            F.lit(\"yes\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") >= 0.85) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") < 0.85) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\").isNull()) | (F.col(\"Coherency_chembl\").isNull()),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"coloc>_90\",\n",
    "        F.when(\n",
    "            (F.col(\"coloc_h4\") >= 0.90) & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "            F.lit(\"yes\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") >= 0.90) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") < 0.90) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\").isNull()) | (F.col(\"Coherency_chembl\").isNull()),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"coloc>_95\",\n",
    "        F.when(\n",
    "            (F.col(\"coloc_h4\") >= 0.95) & (F.col(\"Coherency_chembl\") == \"coherent\"),\n",
    "            F.lit(\"yes\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") >= 0.95) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\") < 0.95) & (F.col(\"Coherency_chembl\") == \"dispar\"),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"coloc_h4\").isNull()) | (F.col(\"Coherency_chembl\").isNull()),\n",
    "            F.lit(\"no\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"secreted\",\n",
    "        F.when(F.col(\"Nr_secreted\") == 1, F.lit(\"yes\"))\n",
    "        .when(F.col(\"Nr_secreted\") == 0, F.lit(\"no\"))\n",
    "        .otherwise(None),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"matchQtlTeTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"taNIntersectQtlTissueEnriched\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"matchQtlDiseaseTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"taNIntersectQtlDisease\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"matchQtlRBFrelevantToDisease\",\n",
    "        F.when(\n",
    "            (F.col(\"matchRBFToDisease\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doeCoherent\",\n",
    "        F.when(F.col(\"Coherency_chembl\") == \"coherent\", F.lit(\"yes\")).otherwise(\n",
    "            F.lit(\"no\")\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doeNotCoherent\",\n",
    "        F.when(F.col(\"Coherency_chembl\") == \"coherent\", F.lit(\"yes\")).otherwise(\n",
    "            F.lit(\"no\")\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doe&matchQtlTeTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"Coherency_chembl\") == \"coherent\")\n",
    "            & (F.col(\"taNIntersectQtlTissueEnriched\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doe&matchQtlDiseaseTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"Coherency_chembl\") == \"coherent\")\n",
    "            & (F.col(\"taNIntersectQtlDisease\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doe&matchQtlRBFrelevantToDisease\",\n",
    "        F.when(\n",
    "            (F.col(\"Coherency_chembl\") == \"coherent\")\n",
    "            & (F.col(\"matchRBFToDisease\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doeInflam\",\n",
    "        F.when(\n",
    "            (F.col(\"coherency_chembl\") == \"coherent\") & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"isInflam\",\n",
    "        F.when(F.col(\"isInflam\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    ### columns combinations for tissue enrichment and therapy areas\n",
    "    .withColumn(\n",
    "        \"doeInflamMatchQtlTeTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"Coherency_chembl\") == \"coherent\")\n",
    "            & (F.col(\"taNIntersectQtlTissueEnriched\") == \"yes\")\n",
    "            & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doeInflamMatchQtlDiseaseTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"Coherency_chembl\") == \"coherent\")\n",
    "            & (F.col(\"taNIntersectQtlDisease\") == \"yes\")\n",
    "            & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"doeInflamMatchQtlRBFrelevantToDisease\",\n",
    "        F.when(\n",
    "            (F.col(\"Coherency_chembl\") == \"coherent\")\n",
    "            & (F.col(\"matchRBFToDisease\") == \"yes\")\n",
    "            & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    #### columns considering only inflam and tissue/therapy areas\n",
    "    .withColumn(\n",
    "        \"inflamMatchQtlTeTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"taNIntersectQtlTissueEnriched\") == \"yes\")\n",
    "            & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"inflamMatchQtlDiseaseTherArea\",\n",
    "        F.when(\n",
    "            (F.col(\"taNIntersectQtlDisease\") == \"yes\") & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"inflamMatchQtlRBFrelevantToDisease\",\n",
    "        F.when(\n",
    "            (F.col(\"matchRBFToDisease\") == \"yes\") & (F.col(\"isInflam\") == \"yes\"),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .drop(\"targetType\")  ### remove nonsense columns for this\n",
    "    .filter(\n",
    "        (F.col(\"coherency_chembl\").isNotNull())\n",
    "        & (F.col(\"name\") != \"COVID-19\")\n",
    "        # | (F.col(\"name\").isNull())\n",
    "    )\n",
    "    .drop(\n",
    "        \"array\",\n",
    "        \"remove\",\n",
    "        \"Phase0\",\n",
    "        \"allPhases\",\n",
    "        \"Terminated\",\n",
    "        \"propagatedTraits\",\n",
    "        \"count\",\n",
    "        \"clinicalStatus\",\n",
    "        \"prediction\",\n",
    "        \"allPhases\",\n",
    "        \"propagatedTraits\",\n",
    "        \"targetType\",\n",
    "    )\n",
    "    .repartition(100)\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "\n",
    "def comparisons_df(toAnalysis_studies) -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "\n",
    "    dataType = [\"byDatatype\"] * len(toAnalysis_studies)\n",
    "    l_studies = []\n",
    "    l_studies.extend([list(a) for a in zip(toAnalysis_studies, dataType)])\n",
    "\n",
    "    schema = StructType(\n",
    "        [\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "            StructField(\"comparisonType\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    comparisons = spark.createDataFrame(l_studies, schema=schema)\n",
    "    ### include all the columns as predictor\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "print(\"created full_data and lists\")\n",
    "\n",
    "toAnalysis_studies = uniqueBetas_analyse.drop(\n",
    "    \"Phase4\",\n",
    "    \"Phase>=3\",\n",
    "    \"Phase>=2\",\n",
    "    \"Phase>=1\",\n",
    "    \"PhaseT\",\n",
    ").columns[\n",
    "    157:\n",
    "]  ### removed column combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-08-19 08:22:07.236163\n",
      "Analysis started on 2024-08-19 at  2024-08-19 08:22:07.236163\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/08/19 08:22:12 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    }
   ],
   "source": [
    "#### now with propagation\n",
    "\"\"\" this scripts run the analysis for comparing QTL studies, tissues together with therapy areas matched\"\"\"\n",
    "from functions import relative_success\n",
    "from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.functions import format_number, udf\n",
    "\n",
    "test = (\n",
    "    spark.read.option(\"header\", \"true\")\n",
    "    .option(\"quote\", '\"')\n",
    "    .option(\"escape\", '\"')\n",
    "    .option(\"multiLine\", \"true\")\n",
    "    .csv(\"gs://ot-team/jroldan/analysis/2024-08-02_colocDoEanalysis.csv\", header=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_pvalue(val):\n",
    "    if val is not None:\n",
    "        if val < 0.005:\n",
    "            return \"{:.2e}\".format(val)  # Scientific notation\n",
    "        else:\n",
    "            return \"{:.4f}\".format(val)  # Fixed decimal notation\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "# Register the UDF\n",
    "format_pvalue_udf = udf(format_pvalue, StringType())\n",
    "\n",
    "regex = r\"\\[\\[\\s*(\\d+)\\s+(\\d+)\\]\\s*\\[\\s*(\\d+)\\s+(\\d+)\\]\\]\"\n",
    "\n",
    "test2 = (\n",
    "    test.withColumn(\n",
    "        \"significant\",\n",
    "        F.when(F.col(\"pValue\") < 0.0001, F.lit(\"****\"))\n",
    "        .when((F.col(\"pValue\") >= 0.0001) & (F.col(\"pValue\") < 0.001), F.lit(\"***\"))\n",
    "        .when((F.col(\"pValue\") >= 0.001) & (F.col(\"pValue\") < 0.01), F.lit(\"**\"))\n",
    "        .when((F.col(\"pValue\") >= 0.01) & (F.col(\"pValue\") < 0.05), F.lit(\"*\"))\n",
    "        .when(F.col(\"pValue\") >= 0.05, F.lit(\"ns\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"writeFigure\",\n",
    "        F.concat(\n",
    "            F.round(F.col(\"oddsRatio\"), 2),\n",
    "            F.lit(\" \"),\n",
    "            F.lit(\"(\"),\n",
    "            F.round(F.col(\"lowerInterval\"), 2),\n",
    "            F.lit(\"-\"),\n",
    "            F.round(F.col(\"upperInterval\"), 2),\n",
    "            F.lit(\")\"),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\"val1\", F.regexp_extract(\"values\", regex, 1))\n",
    "    .withColumn(\"val2\", F.regexp_extract(\"values\", regex, 2))\n",
    "    .withColumn(\"val3\", F.regexp_extract(\"values\", regex, 3))\n",
    "    .withColumn(\"val4\", F.regexp_extract(\"values\", regex, 4))\n",
    "    .withColumn(\"numerator\", (F.col(\"val1\") + F.col(\"val2\")).cast(\"int\"))\n",
    "    .withColumn(\"denominator\", (F.col(\"val3\") + F.col(\"val4\")).cast(\"int\"))\n",
    "    .withColumn(\"pValue\", F.col(\"pValue\").cast(DoubleType()))\n",
    "    .withColumn(\n",
    "        \"pValue_formatted\",\n",
    "        F.when(\n",
    "            F.col(\"pValue\") < 0.0001, F.col(\"pValue\").cast(\"string\")\n",
    "        )  # Check if value matches scientific notation range\n",
    "        # .when(F.col(\"pValue\") < 0.05, F.format_number(F.col(\"pValue\"), 2))\n",
    "        .otherwise(F.format_number(F.col(\"pValue\"), 4)),\n",
    "    )\n",
    ").withColumn(\"numDen\", F.concat_ws(\"/\", F.col(\"numerator\"), F.col(\"denominator\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_mappings = {\n",
    "    \"biofeature\": [\n",
    "        \"Stomach\",\n",
    "        \"BRAIN_PUTAMEN\",\n",
    "        \"Brain_Hypothalamus\",\n",
    "        \"Adipose_Subcutaneous\",\n",
    "        \"BLOOD\",\n",
    "        \"IPSC\",\n",
    "        \"Artery_Aorta\",\n",
    "        \"NEUTROPHIL\",\n",
    "        \"MONOCYTE_IAV\",\n",
    "        \"Vagina\",\n",
    "        \"BRAIN_NAIVE\",\n",
    "        \"Muscle_Skeletal\",\n",
    "        \"Lung\",\n",
    "        \"DLPFC_NAIVE\",\n",
    "        \"Testis\",\n",
    "        \"CD8_T-CELL_NAIVE\",\n",
    "        \"CD4_T-CELL_NAIVE\",\n",
    "        \"TREG_MEMORY\",\n",
    "        \"B-CELL_NAIVE\",\n",
    "        \"NK-CELL_NAIVE\",\n",
    "        \"BRAIN_SPINAL_CORD\",\n",
    "        \"BRAIN_FRONTAL_CORTEX\",\n",
    "        \"LCL_STATIN\",\n",
    "        \"MUSCLE\",\n",
    "        \"PANCREATIC_ISLET\",\n",
    "        \"Cells_EBV-transformed_lymphocytes\",\n",
    "        \"CD8_T-CELL_ANTI-CD3-CD28\",\n",
    "        \"ADIPOSE_NAIVE\",\n",
    "        \"MONOCYTE_NAIVE\",\n",
    "        \"Adrenal_Gland\",\n",
    "        \"FAT\",\n",
    "        \"Minor_Salivary_Gland\",\n",
    "        \"Pituitary\",\n",
    "        \"MONOCYTE_CD14\",\n",
    "        \"PLACENTA_NAIVE\",\n",
    "        \"ILEUM\",\n",
    "        \"Colon_Transverse\",\n",
    "        \"Pancreas\",\n",
    "        \"Esophagus_Muscularis\",\n",
    "        \"SMALL_INTESTINE\",\n",
    "        \"MONOCYTE_LPS2\",\n",
    "        \"Nerve_Tibial\",\n",
    "        \"Artery_Tibial\",\n",
    "        \"MACROPHAGE_IFNG\",\n",
    "        \"BLOOD PLASMA\",\n",
    "        \"Brain_Nucleus_accumbens_basal_ganglia\",\n",
    "        \"Brain_Substantia_nigra\",\n",
    "        \"MACROPHAGE_IFNG+SALMONELLA\",\n",
    "        \"LCL\",\n",
    "        \"Brain_Spinal_cord_cervical_c-1\",\n",
    "        \"Cells_Cultured_fibroblasts\",\n",
    "        \"Prostate\",\n",
    "        \"FIBROBLAST\",\n",
    "        \"SKIN_SUN_EXPOSED\",\n",
    "        \"MONOCYTE_LPS\",\n",
    "        \"Brain_Cerebellum\",\n",
    "        \"Brain_Cortex\",\n",
    "        \"Brain_Cerebellar_Hemisphere\",\n",
    "        \"ADIPOSE_VISCERAL\",\n",
    "        \"Thyroid\",\n",
    "        \"Skin_Not_Sun_Exposed_Suprapubic\",\n",
    "        \"Brain_Amygdala\",\n",
    "        \"Esophagus_Gastroesophageal_Junction\",\n",
    "        \"Heart_Left_Ventricle\",\n",
    "        \"SKIN_NOT_SUN_EXPOSED\",\n",
    "        \"MONOCYTE_IFN24\",\n",
    "        \"Spleen\",\n",
    "        \"Whole_Blood\",\n",
    "        \"TH2_MEMORY\",\n",
    "        \"Skin_Sun_Exposed_Lower_leg\",\n",
    "        \"T-CELL_CD4\",\n",
    "        \"MONOCYTE_LPS24\",\n",
    "        \"T-CELL\",\n",
    "        \"B-CELL_CD19\",\n",
    "        \"HLC\",\n",
    "        \"Breast_Mammary_Tissue\",\n",
    "        \"Adipose_Visceral_Omentum\",\n",
    "        \"TH17_MEMORY\",\n",
    "        \"SKIN\",\n",
    "        \"MONOCYTE\",\n",
    "        \"PLATELET\",\n",
    "        \"Small_Intestine_Terminal_Ileum\",\n",
    "        \"Kidney_Cortex\",\n",
    "        \"Brain_Anterior_cingulate_cortex_BA24\",\n",
    "        \"MONOCYTE_CD16_NAIVE\",\n",
    "        \"MACROPHAGE_NAIVE\",\n",
    "        \"NEUTROPHIL_CD16\",\n",
    "        \"BREAST\",\n",
    "        \"Liver\",\n",
    "        \"MACROPHAGE_SALMONELLA\",\n",
    "        \"MONOCYTE_R848\",\n",
    "        \"Colon_Sigmoid\",\n",
    "        \"Brain_Hippocampus\",\n",
    "        \"Brain_Caudate_basal_ganglia\",\n",
    "        \"Heart_Atrial_Appendage\",\n",
    "        \"Artery_Coronary\",\n",
    "        \"TH1_MEMORY\",\n",
    "        \"Esophagus_Mucosa\",\n",
    "        \"BRAIN_NUCLEUS_ACCUMBENS\",\n",
    "        \"MACROPHAGE_LISTERIA\",\n",
    "        \"MUSCLE_NAIVE\",\n",
    "        \"BRAIN\",\n",
    "        \"Ovary\",\n",
    "        \"MONOCYTE_PAM3CSK4\",\n",
    "        \"BRAIN_CAUDATE\",\n",
    "        \"SUBSTANTIA_NIGRA\",\n",
    "        \"Uterus\",\n",
    "        \"LOW_GRADE_CARTILAGE_NAIVE\",\n",
    "        \"T-CELL_CD8\",\n",
    "        \"TFH_MEMORY\",\n",
    "        \"TREG_NAIVE\",\n",
    "        \"CD4_T-CELL_ANTI-CD3-CD28\",\n",
    "        \"Brain_Frontal_Cortex_BA9\",\n",
    "        \"Brain_Putamen_basal_ganglia\",\n",
    "        \"RECTUM\",\n",
    "        \"NEUTROPHIL_CD15\",\n",
    "        \"TH1-17_MEMORY\",\n",
    "        \"HIGH_GRADE_CARTILAGE_NAIVE\",\n",
    "        \"MICROGLIA_NAIVE\",\n",
    "        \"LCL_NAIVE\",\n",
    "    ],\n",
    "    \"qtl_type\": [\"sqtl\", \"eqtl\", \"pqtl\"],\n",
    "    \"right_study_id\": [\n",
    "        \"Fairfax_2014\",\n",
    "        \"CEDAR\",\n",
    "        \"ROSMAP\",\n",
    "        \"GTEx-eQTL\",\n",
    "        \"BLUEPRINT\",\n",
    "        \"TwinsUK\",\n",
    "        \"Schmiedel_2018\",\n",
    "        \"Lepik_2017\",\n",
    "        \"GTEx-sQTL\",\n",
    "        \"Quach_2016\",\n",
    "        \"eQTLGen\",\n",
    "        \"Nedelec_2016\",\n",
    "        \"FUSION\",\n",
    "        \"Steinberg_2020\",\n",
    "        \"PhLiPS\",\n",
    "        \"GENCORD\",\n",
    "        \"CommonMind\",\n",
    "        \"OLLI_2016\",\n",
    "        \"FOLKERSEN_2020\",\n",
    "        \"GEUVADIS\",\n",
    "        \"CAP\",\n",
    "        \"SUHRE_2017\",\n",
    "        \"PIETZNER_2020\",\n",
    "        \"HipSci\",\n",
    "        \"Alasoo_2018\",\n",
    "        \"Fairfax_2012\",\n",
    "        \"Kasela_2017\",\n",
    "        \"Peng_2018\",\n",
    "        \"SUN2018\",\n",
    "        \"Braineac2\",\n",
    "        \"HILLARY_2019\",\n",
    "        \"BrainSeq\",\n",
    "        \"van_de_Bunt_2015\",\n",
    "        \"FOLKERSEN_2017\",\n",
    "        \"Young_2019\",\n",
    "        \"iPSCORE\",\n",
    "        \"Naranbhai_2015\",\n",
    "    ],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.functions import coalesce\n",
    "\n",
    "conditions = []\n",
    "conditions2 = []\n",
    "\n",
    "for list_name, substrings in list_mappings.items():\n",
    "    for substr in substrings:\n",
    "        # Create a condition column for each substring\n",
    "        condition_column = F.when(F.col(\"path\").contains(substr), F.lit(f\"{substr}\"))\n",
    "        condition_column2 = F.when(\n",
    "            F.col(\"path\").contains(substr), F.lit(f\"{list_name}\")\n",
    "        )\n",
    "        conditions.append(condition_column)\n",
    "        conditions2.append(condition_column2)\n",
    "\n",
    "# Combine all conditions using coalesce to get the first non-null result\n",
    "final_condition_column = coalesce(*conditions)\n",
    "final_condition_column2 = coalesce(*conditions2)\n",
    "\n",
    "df = test2.withColumn(\"name\", final_condition_column).withColumn(\n",
    "    \"data_type\", final_condition_column2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------------------------+\n",
      "|name_rbf                         |\n",
      "+---------------------------------+\n",
      "|                                 |\n",
      "|                                 |\n",
      "|                                 |\n",
      "|                                 |\n",
      "|                                 |\n",
      "|Stomach                          |\n",
      "|Stomach                          |\n",
      "|Stomach                          |\n",
      "|Stomach                          |\n",
      "|Stomach                          |\n",
      "|Cells_EBV-transformed_lymphocytes|\n",
      "|Cells_EBV-transformed_lymphocytes|\n",
      "|Cells_EBV-transformed_lymphocytes|\n",
      "|Cells_EBV-transformed_lymphocytes|\n",
      "|Cells_EBV-transformed_lymphocytes|\n",
      "|Nerve_Tibial                     |\n",
      "|Nerve_Tibial                     |\n",
      "|Nerve_Tibial                     |\n",
      "|Nerve_Tibial                     |\n",
      "|Nerve_Tibial                     |\n",
      "+---------------------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "### extract the right name from the right_bio_feature one\n",
    "\n",
    "# Extract the part of the string between '/propagated/' and '_right_bio_feature'\n",
    "df = df.withColumn(\n",
    "    \"name_rbf\",\n",
    "    F.regexp_extract(F.col(\"path\"), r\"/propagated/(.+)_right_bio_feature\", 1),\n",
    ")\n",
    "\n",
    "# Show result\n",
    "df.select(\"name_rbf\").show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "df.toPandas().to_csv(\n",
    "    \"gs://ot-team/jroldan/analysis/2024-08-02_colocDoEanalysis_processed_rbfCorrected.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### dataframe of L2G coherency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g = (\n",
    "    spark.read.option(\"header\", \"true\")\n",
    "    .option(\"quote\", '\"')\n",
    "    .option(\"escape\", '\"')\n",
    "    .option(\"multiLine\", \"true\")\n",
    "    .csv(\n",
    "        \"gs://ot-team/jroldan/analysis/2024-08-14_analysis_coherencyL2G.csv\",\n",
    "        header=True,\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.functions import format_number, udf\n",
    "\n",
    "\n",
    "def format_pvalue(val):\n",
    "    if val is not None:\n",
    "        if val < 0.005:\n",
    "            return \"{:.2e}\".format(val)  # Scientific notation\n",
    "        else:\n",
    "            return \"{:.4f}\".format(val)  # Fixed decimal notation\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "# Register the UDF\n",
    "format_pvalue_udf = udf(format_pvalue, StringType())\n",
    "\n",
    "regex = r\"\\[\\[\\s*(\\d+)\\s+(\\d+)\\]\\s*\\[\\s*(\\d+)\\s+(\\d+)\\]\\]\"\n",
    "\n",
    "l2g2 = (\n",
    "    l2g.withColumn(\n",
    "        \"significant\",\n",
    "        F.when(F.col(\"pValue\") < 0.0001, F.lit(\"****\"))\n",
    "        .when((F.col(\"pValue\") >= 0.0001) & (F.col(\"pValue\") < 0.001), F.lit(\"***\"))\n",
    "        .when((F.col(\"pValue\") >= 0.001) & (F.col(\"pValue\") < 0.01), F.lit(\"**\"))\n",
    "        .when((F.col(\"pValue\") >= 0.01) & (F.col(\"pValue\") < 0.05), F.lit(\"*\"))\n",
    "        .when(F.col(\"pValue\") >= 0.05, F.lit(\"ns\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"writeFigure\",\n",
    "        F.concat(\n",
    "            F.round(F.col(\"oddsRatio\"), 2),\n",
    "            F.lit(\" \"),\n",
    "            F.lit(\"(\"),\n",
    "            F.round(F.col(\"lowerInterval\"), 2),\n",
    "            F.lit(\"-\"),\n",
    "            F.round(F.col(\"upperInterval\"), 2),\n",
    "            F.lit(\")\"),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\"val1\", F.regexp_extract(\"values\", regex, 1))\n",
    "    .withColumn(\"val2\", F.regexp_extract(\"values\", regex, 2))\n",
    "    .withColumn(\"val3\", F.regexp_extract(\"values\", regex, 3))\n",
    "    .withColumn(\"val4\", F.regexp_extract(\"values\", regex, 4))\n",
    "    .withColumn(\"numerator\", (F.col(\"val1\") + F.col(\"val2\")).cast(\"int\"))\n",
    "    .withColumn(\"denominator\", (F.col(\"val3\") + F.col(\"val4\")).cast(\"int\"))\n",
    "    .withColumn(\"pValue\", F.col(\"pValue\").cast(DoubleType()))\n",
    "    .withColumn(\n",
    "        \"pValue_formatted\",\n",
    "        F.when(\n",
    "            F.col(\"pValue\") < 0.0001, F.col(\"pValue\").cast(\"string\")\n",
    "        )  # Check if value matches scientific notation range\n",
    "        # .when(F.col(\"pValue\") < 0.05, F.format_number(F.col(\"pValue\"), 2))\n",
    "        .otherwise(F.format_number(F.col(\"pValue\"), 4)),\n",
    "    )\n",
    ").withColumn(\"numDen\", F.concat_ws(\"/\", F.col(\"numerator\"), F.col(\"denominator\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.functions import regexp_extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapl2g = {\n",
    "    \"l2gScore\": [\n",
    "        \"0_1\",\n",
    "        \"0_15\",\n",
    "        \"0_2\",\n",
    "        \"0_25\",\n",
    "        \"0_3\",\n",
    "        \"0_35\",\n",
    "        \"0_4\",\n",
    "        \"0_45\",\n",
    "        \"0_5\",\n",
    "        \"0_55\",\n",
    "        \"0_6\",\n",
    "        \"0_65\",\n",
    "        \"0_7\",\n",
    "        \"0_75\",\n",
    "        \"0_8\",\n",
    "        \"0_85\",\n",
    "        \"0_9\",\n",
    "        \"0_95\",\n",
    "    ],\n",
    "    \"distanceRank\": [\n",
    "        \"<=1\",\n",
    "        \"<=2\",\n",
    "        \"<=3\",\n",
    "        \"<=4\",\n",
    "        \"<=5\",\n",
    "        \"<=6\",\n",
    "        \"<=7\",\n",
    "        \"<=8\",\n",
    "        \"<=9\",\n",
    "        \"<=10\",\n",
    "    ],\n",
    "}\n",
    "from pyspark.sql.functions import coalesce\n",
    "\n",
    "conditions = []\n",
    "conditions2 = []\n",
    "\n",
    "for list_name, substrings in mapl2g.items():\n",
    "    for substr in substrings:\n",
    "        # Create a condition column for each substring\n",
    "        condition_column = F.when(F.col(\"path\").contains(substr), F.lit(f\"{substr}\"))\n",
    "        condition_column2 = F.when(\n",
    "            F.col(\"path\").contains(substr), F.lit(f\"{list_name}\")\n",
    "        )\n",
    "        conditions.append(condition_column)\n",
    "        conditions2.append(condition_column2)\n",
    "\n",
    "# Combine all conditions using coalesce to get the first non-null result\n",
    "final_condition_column = coalesce(*conditions)\n",
    "final_condition_column2 = coalesce(*conditions2)\n",
    "\n",
    "l2g3 = (\n",
    "    l2g2.withColumn(\"name2\", final_condition_column).withColumn(\n",
    "        \"data_type\", final_condition_column2\n",
    "    )\n",
    "    #    .filter(F.col(\"rank\") == 0.1)\n",
    "    # .show(truncate=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "### modify the 0_1 numbers.\n",
    "# Start with a default value for the match column\n",
    "match_column = F.lit(None)\n",
    "\n",
    "# Iterate over all substrings in the dictionary\n",
    "for key, substrings in mapl2g.items():\n",
    "    for substring in substrings:\n",
    "        pattern = f\"{substring}[_&]\"\n",
    "        match_column = F.when(F.col(\"path\").rlike(pattern), substring).otherwise(\n",
    "            match_column\n",
    "        )\n",
    "\n",
    "# Add the new column to the DataFrame\n",
    "l2g4 = l2g3.withColumn(\"rankTrue\", F.regexp_replace(match_column, \"_\", \".\"))\n",
    "\n",
    "# df_with_match.withColumn(\n",
    "#    \"matched_substring2\", F.regexp_replace(F.col(\"matched_substring\"), \"_\", \".\")\n",
    "# ).show(1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Start with a default value for the match column\n",
    "match_column = F.lit(None)\n",
    "\n",
    "# Iterate over all substrings in the dictionary\n",
    "for key, substrings in mapl2g.items():\n",
    "    for substring in substrings:\n",
    "        pattern = f\"{substring}[_&]\"\n",
    "        # Build the match column with the correct substring if found\n",
    "        match_column = F.when(F.col(\"path\").rlike(pattern), substring).otherwise(\n",
    "            match_column\n",
    "        )\n",
    "\n",
    "# Extract the number from patterns like \"<=n\"\n",
    "number_column = F.regexp_extract(match_column, r\"<=([\\d]+)\", 2)\n",
    "\n",
    "# Replace underscores with periods for patterns like \"0_1\"\n",
    "processed_match_column = F.when(\n",
    "    F.col(\"path\").rlike(r\"[_&]\"), F.regexp_replace(match_column, \"_\", \".\")\n",
    ").otherwise(number_column)\n",
    "\n",
    "# Add the new column to the DataFrame\n",
    "l2g4 = l2g3.withColumn(\"rankTrue\", processed_match_column).withColumn(\n",
    "    \"fixed\",  ### process the strings <=n, to take only n in the distance rank cases\n",
    "    F.when(\n",
    "        F.col(\"rankTrue\").contains(\"<=\"),\n",
    "        F.expr(\"substring(rankTrue, 3, length(rankTrue) - 2)\"),\n",
    "    ).otherwise(F.col(\"rankTrue\")),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g4.toPandas().to_csv(\n",
    "    \"gs://ot-team/jroldan/analysis/2024-08-14_processed_analysis_coherencyL2G.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fix the group for the L2G score and Coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of substrings of interest\n",
    "group_mapping = [\"hasGeneticEvidence\", \"L2GAndColoc\", \"diagonalYes\", \"oneCellYes\"]\n",
    "\n",
    "# Initialize a default column with None\n",
    "match_column = F.lit(None)\n",
    "\n",
    "# Iterate over the substrings and update the match_column when a substring is found\n",
    "for substring in group_mapping:\n",
    "    match_column = F.when(F.col(\"path\").rlike(substring), substring).otherwise(\n",
    "        match_column\n",
    "    )\n",
    "\n",
    "# Add the new column with the matched substring\n",
    "l2g5 = l2g4.withColumn(\"group_fixed\", match_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2:>                                                          (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+-----+\n",
      "|       group_fixed|count|\n",
      "+------------------+-----+\n",
      "|              null|  280|\n",
      "|       diagonalYes|  300|\n",
      "|       L2GAndColoc|  300|\n",
      "|hasGeneticEvidence|  300|\n",
      "|        oneCellYes|  300|\n",
      "+------------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "l2g5.groupBy(\"group_fixed\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------------------------------------------------------------+\n",
      "|path                                                                                        |\n",
      "+--------------------------------------------------------------------------------------------+\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_1_Phase4.parquet   |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_1_Phase>=3.parquet |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_1_Phase>=2.parquet |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_1_Phase>=1.parquet |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_1_PhaseT.parquet   |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_15_Phase4.parquet  |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_15_Phase>=3.parquet|\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_15_Phase>=2.parquet|\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_15_Phase>=1.parquet|\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_15_PhaseT.parquet  |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_2_Phase4.parquet   |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_2_Phase>=3.parquet |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_2_Phase>=2.parquet |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_2_Phase>=1.parquet |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_2_PhaseT.parquet   |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_25_Phase4.parquet  |\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_25_Phase>=3.parquet|\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_25_Phase>=2.parquet|\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_25_Phase>=1.parquet|\n",
      "|gs://ot-team/jroldan/2024-08-14_analysis/df_l2g_original/max_L2GScore>=0_25_PhaseT.parquet  |\n",
      "+--------------------------------------------------------------------------------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "l2g5.filter(F.col(\"group_fixed\").isNull()).select(\"path\").show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g5.toPandas().to_csv(\n",
    "    \"gs://ot-team/jroldan/analysis/2024-08-14_processedLatest_analysis_coherencyL2G_v19Aug.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trying to see whether the analysis can be done as the code it is in whe notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-01-23 12:10:51.668473\n",
      "Analysis started on 2025-01-23 at  2025-01-23 12:10:51.668473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/23 12:10:58 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-01-23 12:10:58.122712\n"
     ]
    },
    {
     "ename": "AnalysisException",
     "evalue": "Column 'leftVariantId' does not exist. Did you mean one of the following? [variantId, targetId, variantRsId, reactionId, targetId2, cohortId, contrast, datatypeId, literature, target_id, variantHgvsId, ancestryId, beta, datasourceId, description, diseaseId, drugId, projectId, releaseDate, sourceId, studyId, OddsRatio, actionType, alleleOrigins, ancestry, biomarkers, clinicalStatus, drugId2, homogenized, id, intogenAnnot, pathways, reactionName, releaseVersion, studyCases, variantEffect, biomarkerName, cellType, clinicalPhase, confidence, drugResponse, pValueMantissa, pmcIds, publicationYear, resourceScore, score, sex, statisticalMethod, studyStartDate, targetInModel, urls, TSorOncogene, activators_list, biologicalModelId, geneticBackground, mutatedSamples, pValueExponent, studyOverview, approvedSymbol, cohortDescription, cohortPhenotypes, cohortShortName, description_splited, directionOnTrait, drugFromSource, inhibitors_list, intogen_function, studySampleSize, studyStopReason, targetModulation, textMiningSentences, allelicRequirements, diseaseCellLines, diseaseFromSource, diseaseFromSourceId, statisticalTestTail, targetFromSource, targetFromSourceId, targetInModelMgiId, clinicalSignificances, log2FoldChangeValue, crisprScreenLibrary, biosamplesFromSource, publicationFirstAuthor, significantDriverMethods, targetInModelEnsemblId, diseaseFromSourceMappedId, statisticalMethodOverview, studyStopReasonCategories, betaConfidenceIntervalLower, betaConfidenceIntervalUpper, log2FoldChangePercentileRank, studyCasesWithQualifyingVariants, variantAminoacidDescriptions, variantFunctionalConsequenceId, biologicalModelGeneticBackground, oddsRatioConfidenceIntervalLower, oddsRatioConfidenceIntervalUpper, biologicalModelAllelicComposition, variantFunctionalConsequenceFromQtlId, diseaseModelAssociatedHumanPhenotypes, diseaseModelAssociatedModelPhenotypes];\n'Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 79 more fields]\n+- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 78 more fields]\n   +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 77 more fields]\n      +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 77 more fields]\n         +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 77 more fields]\n            +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 78 more fields]\n               +- Window [collect_set(intogen_function#1938, 0, 0) windowspecdefinition(targetId#782, diseaseId#864, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#2043], [targetId#782, diseaseId#864]\n                  +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 76 more fields]\n                     +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 76 more fields]\n                        +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 75 more fields]\n                           +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 74 more fields]\n                              +- Join LeftOuter, ((drugId2#1031 = drugId#815) AND (targetId2#1038 = targetId#782))\n                                 :- Join LeftOuter, (target_id#1081 = targetId#782)\n                                 :  :- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, concat_ws(,, clinicalSignificances#798) AS clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 65 more fields]\n                                 :  :  +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#798, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 65 more fields]\n                                 :  :     +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, cast(beta#787 as double) AS beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#798, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 65 more fields]\n                                 :  :        +- Filter datasourceId#781 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                 :  :           +- Filter datasourceId#781 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                 :  :              +- Relation [datasourceId#781,targetId#782,alleleOrigins#783,allelicRequirements#784,ancestry#785,ancestryId#786,beta#787,betaConfidenceIntervalLower#788,betaConfidenceIntervalUpper#789,biologicalModelAllelicComposition#790,biologicalModelGeneticBackground#791,biologicalModelId#792,biomarkerName#793,biomarkers#794,biosamplesFromSource#795,cellType#796,clinicalPhase#797,clinicalSignificances#798,clinicalStatus#799,cohortDescription#800,cohortId#801,cohortPhenotypes#802,cohortShortName#803,confidence#804,... 65 more fields] parquet\n                                 :  +- Project [id#960 AS target_id#1081, approvedSymbol#961, description#1066, description_splited#1070, TSorOncogene#1075]\n                                 :     +- Project [id#960, approvedSymbol#961, description#1066, description_splited#1070, CASE WHEN (RLIKE(description_splited#1070, ncogene) AND RLIKE(description_splited#1070, TSG)) THEN bivalent WHEN RLIKE(description_splited#1070, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#1070, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#1075]\n                                 :        +- Project [id#960, approvedSymbol#961, description#1066, concat_ws(,, description#1066) AS description_splited#1070]\n                                 :           +- Aggregate [id#960, approvedSymbol#961], [id#960, approvedSymbol#961, collect_set(description#1058, 0, 0) AS description#1066]\n                                 :              +- Filter description#1058 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                 :                 +- Project [id#960, approvedSymbol#961, col#1053.description AS description#1058]\n                                 :                    +- Project [id#960, approvedSymbol#961, col#1053]\n                                 :                       +- Generate explode(hallmarks#970.attributes), true, [col#1053]\n                                 :                          +- Relation [id#960,approvedSymbol#961,biotype#962,transcriptIds#963,canonicalTranscript#964,canonicalExons#965,genomicLocation#966,alternativeGenes#967,approvedName#968,go#969,hallmarks#970,synonyms#971,symbolSynonyms#972,nameSynonyms#973,functionDescriptions#974,subcellularLocations#975,targetClass#976,obsoleteSymbols#977,obsoleteNames#978,constraint#979,tep#980,proteinIds#981,dbXrefs#982,chemicalProbes#983,... 4 more fields] parquet\n                                 +- Aggregate [targetId2#1038, drugId2#1031], [targetId2#1038, drugId2#1031, collect_set(actionType#1016, 0, 0) AS actionType#1048]\n                                    +- Project [targetId2#1038, drugId2#1031, actionType#1016, mechanismOfAction#1017]\n                                       +- Generate explode(targets#1021), true, [targetId2#1038]\n                                          +- Project [drugId2#1031, actionType#1016, mechanismOfAction#1017, targets#1021]\n                                             +- Generate explode(chemblIds#1018), true, [drugId2#1031]\n                                                +- Relation [actionType#1016,mechanismOfAction#1017,chemblIds#1018,targetName#1019,targetType#1020,targets#1021,references#1022] parquet\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAnalysisException\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_8478/1531421275.py\u001b[0m in \u001b[0;36m<cell line: 132>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    131\u001b[0m prueba_assessment = (\n\u001b[1;32m    132\u001b[0m     \u001b[0mdirectionOfEffect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mevidences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplatform_v\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 133\u001b[0;31m     .withColumn(\n\u001b[0m\u001b[1;32m    134\u001b[0m         \u001b[0;34m\"rank\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m         F.when(\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/dataframe.py\u001b[0m in \u001b[0;36mwithColumn\u001b[0;34m(self, colName, col)\u001b[0m\n\u001b[1;32m   3034\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mColumn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3035\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"col should be Column\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3036\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwithColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolName\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparkSession\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3037\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3038\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwithColumnRenamed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexisting\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m\"DataFrame\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1319\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1321\u001b[0;31m         return_value = get_return_value(\n\u001b[0m\u001b[1;32m   1322\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[1;32m   1323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/utils.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    194\u001b[0m                 \u001b[0;31m# Hide where the exception came from that shows a non-Pythonic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m                 \u001b[0;31m# JVM exception message.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mconverted\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAnalysisException\u001b[0m: Column 'leftVariantId' does not exist. Did you mean one of the following? [variantId, targetId, variantRsId, reactionId, targetId2, cohortId, contrast, datatypeId, literature, target_id, variantHgvsId, ancestryId, beta, datasourceId, description, diseaseId, drugId, projectId, releaseDate, sourceId, studyId, OddsRatio, actionType, alleleOrigins, ancestry, biomarkers, clinicalStatus, drugId2, homogenized, id, intogenAnnot, pathways, reactionName, releaseVersion, studyCases, variantEffect, biomarkerName, cellType, clinicalPhase, confidence, drugResponse, pValueMantissa, pmcIds, publicationYear, resourceScore, score, sex, statisticalMethod, studyStartDate, targetInModel, urls, TSorOncogene, activators_list, biologicalModelId, geneticBackground, mutatedSamples, pValueExponent, studyOverview, approvedSymbol, cohortDescription, cohortPhenotypes, cohortShortName, description_splited, directionOnTrait, drugFromSource, inhibitors_list, intogen_function, studySampleSize, studyStopReason, targetModulation, textMiningSentences, allelicRequirements, diseaseCellLines, diseaseFromSource, diseaseFromSourceId, statisticalTestTail, targetFromSource, targetFromSourceId, targetInModelMgiId, clinicalSignificances, log2FoldChangeValue, crisprScreenLibrary, biosamplesFromSource, publicationFirstAuthor, significantDriverMethods, targetInModelEnsemblId, diseaseFromSourceMappedId, statisticalMethodOverview, studyStopReasonCategories, betaConfidenceIntervalLower, betaConfidenceIntervalUpper, log2FoldChangePercentileRank, studyCasesWithQualifyingVariants, variantAminoacidDescriptions, variantFunctionalConsequenceId, biologicalModelGeneticBackground, oddsRatioConfidenceIntervalLower, oddsRatioConfidenceIntervalUpper, biologicalModelAllelicComposition, variantFunctionalConsequenceFromQtlId, diseaseModelAssociatedHumanPhenotypes, diseaseModelAssociatedModelPhenotypes];\n'Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 79 more fields]\n+- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 78 more fields]\n   +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 77 more fields]\n      +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 77 more fields]\n         +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 77 more fields]\n            +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 78 more fields]\n               +- Window [collect_set(intogen_function#1938, 0, 0) windowspecdefinition(targetId#782, diseaseId#864, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#2043], [targetId#782, diseaseId#864]\n                  +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 76 more fields]\n                     +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 76 more fields]\n                        +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 75 more fields]\n                           +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 74 more fields]\n                              +- Join LeftOuter, ((drugId2#1031 = drugId#815) AND (targetId2#1038 = targetId#782))\n                                 :- Join LeftOuter, (target_id#1081 = targetId#782)\n                                 :  :- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, concat_ws(,, clinicalSignificances#798) AS clinicalSignificances#1267, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 65 more fields]\n                                 :  :  +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#798, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 65 more fields]\n                                 :  :     +- Project [datasourceId#781, targetId#782, alleleOrigins#783, allelicRequirements#784, ancestry#785, ancestryId#786, cast(beta#787 as double) AS beta#1087, betaConfidenceIntervalLower#788, betaConfidenceIntervalUpper#789, biologicalModelAllelicComposition#790, biologicalModelGeneticBackground#791, biologicalModelId#792, biomarkerName#793, biomarkers#794, biosamplesFromSource#795, cellType#796, clinicalPhase#797, clinicalSignificances#798, clinicalStatus#799, cohortDescription#800, cohortId#801, cohortPhenotypes#802, cohortShortName#803, confidence#804, ... 65 more fields]\n                                 :  :        +- Filter datasourceId#781 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                 :  :           +- Filter datasourceId#781 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                 :  :              +- Relation [datasourceId#781,targetId#782,alleleOrigins#783,allelicRequirements#784,ancestry#785,ancestryId#786,beta#787,betaConfidenceIntervalLower#788,betaConfidenceIntervalUpper#789,biologicalModelAllelicComposition#790,biologicalModelGeneticBackground#791,biologicalModelId#792,biomarkerName#793,biomarkers#794,biosamplesFromSource#795,cellType#796,clinicalPhase#797,clinicalSignificances#798,clinicalStatus#799,cohortDescription#800,cohortId#801,cohortPhenotypes#802,cohortShortName#803,confidence#804,... 65 more fields] parquet\n                                 :  +- Project [id#960 AS target_id#1081, approvedSymbol#961, description#1066, description_splited#1070, TSorOncogene#1075]\n                                 :     +- Project [id#960, approvedSymbol#961, description#1066, description_splited#1070, CASE WHEN (RLIKE(description_splited#1070, ncogene) AND RLIKE(description_splited#1070, TSG)) THEN bivalent WHEN RLIKE(description_splited#1070, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#1070, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#1075]\n                                 :        +- Project [id#960, approvedSymbol#961, description#1066, concat_ws(,, description#1066) AS description_splited#1070]\n                                 :           +- Aggregate [id#960, approvedSymbol#961], [id#960, approvedSymbol#961, collect_set(description#1058, 0, 0) AS description#1066]\n                                 :              +- Filter description#1058 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                 :                 +- Project [id#960, approvedSymbol#961, col#1053.description AS description#1058]\n                                 :                    +- Project [id#960, approvedSymbol#961, col#1053]\n                                 :                       +- Generate explode(hallmarks#970.attributes), true, [col#1053]\n                                 :                          +- Relation [id#960,approvedSymbol#961,biotype#962,transcriptIds#963,canonicalTranscript#964,canonicalExons#965,genomicLocation#966,alternativeGenes#967,approvedName#968,go#969,hallmarks#970,synonyms#971,symbolSynonyms#972,nameSynonyms#973,functionDescriptions#974,subcellularLocations#975,targetClass#976,obsoleteSymbols#977,obsoleteNames#978,constraint#979,tep#980,proteinIds#981,dbXrefs#982,chemicalProbes#983,... 4 more fields] parquet\n                                 +- Aggregate [targetId2#1038, drugId2#1031], [targetId2#1038, drugId2#1031, collect_set(actionType#1016, 0, 0) AS actionType#1048]\n                                    +- Project [targetId2#1038, drugId2#1031, actionType#1016, mechanismOfAction#1017]\n                                       +- Generate explode(targets#1021), true, [targetId2#1038]\n                                          +- Project [drugId2#1031, actionType#1016, mechanismOfAction#1017, targets#1021]\n                                             +- Generate explode(chemblIds#1018), true, [drugId2#1031]\n                                                +- Relation [actionType#1016,mechanismOfAction#1017,chemblIds#1018,targetName#1019,targetType#1020,targets#1021,references#1022] parquet\n"
     ]
    }
   ],
   "source": [
    "#########\n",
    "#####\n",
    "#######\n",
    "## ANALYSIS FOR L2G Scores, genetic evidence and Direction of Effect\n",
    "## Original, propagated and Other Vs Oncology\n",
    "#######\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "### New coloc evidences == leftVariantId\n",
    "ranking = Window.partitionBy(\"studyId\", \"leftVariantId\")\n",
    "\n",
    "### union with the other datasources\n",
    "platform_v = \"24.09\"\n",
    "\n",
    "target_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/targets/\"\n",
    ")\n",
    "target = spark.read.parquet(target_path)\n",
    "\n",
    "disease_path = (\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/diseases/\"\n",
    ")\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "mecact_path = f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/mechanismOfAction/\"\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "evidences = spark.read.parquet(\n",
    "    f\"gs://open-targets-data-releases/{platform_v}/output/etl/parquet/evidence\"\n",
    ").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "\n",
    "windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "sincgc = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "germline = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"chembl\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "#### Build the evidences from new coloc to be able to do the ranking.\n",
    "\n",
    "\n",
    "#### version all gene burden\n",
    "prueba_assessment = (\n",
    "    directionOfEffect(evidences, platform_v)\n",
    "    .withColumn(\n",
    "        \"rank\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"average\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-01-24 11:28:26.526966\n",
      "Analysis started on 2025-01-24 at  2025-01-24 11:28:26.526966\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/24 11:28:28 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n",
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "#### BUILDING THE NEW GWAS GENETIC EVIDENCE FROM COLOC\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from datetime import datetime\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "credibleEvidence = spark.read.parquet(f\"{path}evidence\").filter(\n",
    "    F.col(\"datasourceId\").isin([\"gwas_credible_sets\"])\n",
    ")\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "# remove columns without content (only null values on them)\n",
    "df = credibleEvidence.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns).persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "### bring directionality from QTL\n",
    "\n",
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "### modify to include more information\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )\n",
    "    .select(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"homogenized\",\n",
    "        \"leftStudyLocusId\",\n",
    "        \"h4\",\n",
    "        \"datasourceId\",\n",
    "        \"resourceScore\",\n",
    "        \"leftVariantId\",\n",
    "        \"credibleLeftStudyType\",\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")  ### there will be duplicates TargetId-DiseaseId because we are taking the most significant DoE\n",
    "\n",
    "### Now I have the DoE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "#### LOAD STUDYLOCUSID AND VARIANT DISTANCES\n",
    "l2gPred = spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/locusToGenePredictions\"\n",
    ")\n",
    "l2gTable = (\n",
    "    l2gPred.select(\"studyLocusId\", \"geneId\", F.explode_outer(\"locusToGeneFeatures\"))\n",
    "    .filter(F.col(\"key\").isin([\"distanceFootprintMean\", \"distanceTssMean\"]))\n",
    "    .groupBy(\"studyLocusId\", \"geneId\")\n",
    "    .pivot(\"key\")\n",
    "    .agg(F.first(\"value\"))\n",
    ")\n",
    "\n",
    "gwasCredibleAssocDistances = gwasCredibleAssoc.join(\n",
    "    l2gTable.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\").withColumnRenamed(\n",
    "        \"geneId\", \"targetId\"\n",
    "    ),\n",
    "    on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "    how=\"left\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 49:>                                                         (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------------+---------------+\n",
      "|    leftStudyLocusId|       targetId|  diseaseId|homogenized|                h4|      datasourceId|      resourceScore|  leftVariantId|credibleLeftStudyType|distanceFootprintMean|distanceTssMean|\n",
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------------+---------------+\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|            0.9999886|      0.9997445|\n",
      "|b9e47164bf645be64...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|           0.99987966|      0.9996356|\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004611|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004611|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|            0.9999886|      0.9997445|\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|            0.9999886|      0.9997445|\n",
      "|b9e47164bf645be64...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|           0.99987966|      0.9996356|\n",
      "|a84c1695b952c7f7a...|ENSG00000000419|EFO_0000313|noEvaluable|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|             0.993824|      0.9904567|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|            0.9773113|      0.9773113|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|            0.9543429|      0.9543429|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|             0.939027|       0.939027|\n",
      "|57969a20046c11221...|ENSG00000000457|EFO_0004286|noEvaluable|              null|gwas_credible_sets|0.06451956210793872|           null|                 null|           0.95156074|     0.95156074|\n",
      "|1eb9f83b09dd97b97...|ENSG00000000460|EFO_0003911|noEvaluable|              null|gwas_credible_sets|0.06116443109603223|           null|                 null|            0.9369747|      0.9143192|\n",
      "|1eb9f83b09dd97b97...|ENSG00000000460|EFO_0004269|noEvaluable|              null|gwas_credible_sets|0.06116443109603223|           null|                 null|            0.9369747|      0.9143192|\n",
      "|253165fd0bbff490a...|ENSG00000000460|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.35074980670304023|           null|                 null|            0.9489437|      0.9489436|\n",
      "|0447c925d979dac73...|ENSG00000000938|EFO_0004742|noEvaluable|              null|gwas_credible_sets|0.11826405203479222|           null|                 null|           0.96459347|     0.95905524|\n",
      "|cf0c2add70777ac20...|ENSG00000000971|EFO_0001365|noEvaluable|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|            0.9988011|      0.9848842|\n",
      "|cf0c2add70777ac20...|ENSG00000000971|EFO_0001365|noEvaluable|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|            0.9988011|      0.9848842|\n",
      "|cf0c2add70777ac20...|ENSG00000000971|EFO_0001365|noEvaluable|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|            0.9988011|      0.9848842|\n",
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------------+---------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssocDistances.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['leftStudyLocusId',\n",
       " 'targetId',\n",
       " 'diseaseId',\n",
       " 'homogenized',\n",
       " 'h4',\n",
       " 'datasourceId',\n",
       " 'resourceScore',\n",
       " 'leftVariantId',\n",
       " 'credibleLeftStudyType',\n",
       " 'geneId',\n",
       " 'distanceFootprintMean',\n",
       " 'distanceTssMean']"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasCredibleAssocDistances.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 195:>                                                      (0 + 16) / 17]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+------+\n",
      "|       targetId|  diseaseId|counts|\n",
      "+---------------+-----------+------+\n",
      "|ENSG00000166035|EFO_0004529|  2634|\n",
      "|ENSG00000175445|EFO_0004529|  2074|\n",
      "|ENSG00000134824|EFO_0004529|  1914|\n",
      "|ENSG00000149485|EFO_0004529|  1891|\n",
      "|ENSG00000087237|EFO_0004529|  1699|\n",
      "|ENSG00000084674|EFO_0004529|  1641|\n",
      "|ENSG00000218819|EFO_0004529|  1598|\n",
      "|ENSG00000166035|EFO_0004732|  1577|\n",
      "|ENSG00000038427|EFO_0004346|  1562|\n",
      "|ENSG00000038427|EFO_0001444|  1555|\n",
      "|ENSG00000198670|EFO_0004529|  1463|\n",
      "|ENSG00000109917|EFO_0004529|  1441|\n",
      "|ENSG00000122194|EFO_0004529|  1437|\n",
      "|ENSG00000134825|EFO_0004529|  1418|\n",
      "|ENSG00000130164|EFO_0004529|  1403|\n",
      "|ENSG00000000971|EFO_0004747|  1382|\n",
      "|ENSG00000110243|EFO_0004529|  1362|\n",
      "|ENSG00000116785|EFO_0004747|  1295|\n",
      "|ENSG00000130203|EFO_0004529|  1278|\n",
      "|ENSG00000069399|EFO_0004529|  1260|\n",
      "+---------------+-----------+------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "### For a given T-D there could be multiples studylocusId\n",
    "gwasCredibleAssocDistances.groupBy(\"targetId\", \"diseaseId\").agg(\n",
    "    F.size(F.collect_set(\"leftStudyLocusId\")).alias(\"counts\")\n",
    ").sort(F.col(\"counts\").desc()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------+---------------+-----------+-----------+--------------------+------------------+-------------------+--------------+---------------------+---------------------+---------------+\n",
      "|leftStudyLocusId                |targetId       |diseaseId  |homogenized|h4                  |datasourceId      |resourceScore      |leftVariantId |credibleLeftStudyType|distanceFootprintMean|distanceTssMean|\n",
      "+--------------------------------+---------------+-----------+-----------+--------------------+------------------+-------------------+--------------+---------------------+---------------------+---------------+\n",
      "|74da780a06c89e32672ccfe39eafdb41|ENSG00000001617|EFO_0005856|GoF_protect|0.8813741513582225  |gwas_credible_sets|0.09022955545837465|3_50132282_C_T|gwas                 |0.9783658            |0.9783658      |\n",
      "|74da780a06c89e32672ccfe39eafdb41|ENSG00000001617|EFO_0005856|GoF_protect|0.027484611508886925|gwas_credible_sets|0.09022955545837465|3_50132282_C_T|gwas                 |0.9783658            |0.9783658      |\n",
      "|74da780a06c89e32672ccfe39eafdb41|ENSG00000001617|EFO_0005856|GoF_protect|0.6743326348063136  |gwas_credible_sets|0.09022955545837465|3_50132282_C_T|gwas                 |0.9783658            |0.9783658      |\n",
      "+--------------------------------+---------------+-----------+-----------+--------------------+------------------+-------------------+--------------+---------------------+---------------------+---------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssocDistances.filter(\n",
    "    (F.col(\"targetId\") == \"ENSG00000001617\") & (F.col(\"diseaseId\") == \"EFO_0005856\")\n",
    ").show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "analysis_chembl = (\n",
    "    discrepancifier(\n",
    "        directionOfEffect(\n",
    "            evidences.filter((F.col(\"datasourceId\") == \"chembl\")), \"24.09\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "    .filter(  ### ensure drug has annotated MoA and is coherent per Target-Disease\n",
    "        ((F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull())\n",
    "        & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    )\n",
    "    .selectExpr(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"maxClinPhase\",\n",
    "        \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "        \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "        \"LoF_protect as LoF_protect_ch\",\n",
    "        \"GoF_protect as GoF_protect_ch\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "### pivot colocdoE grouping by T-D-studyLocusId-distances\n",
    "\n",
    "values = [\"max_L2GScore\", \"min_footPrintDistance_rank\", \"min_tssDistance_rank\"]\n",
    "value = \"max_L2GScore\"\n",
    "dict_comb = {\n",
    "    \"hasGeneticEvidence\": f\"{value}\",\n",
    "    \"diagonalYes\": f\"{value}\",\n",
    "    \"oneCellYes\": f\"{value}\",\n",
    "    \"L2GAndColoc\": f\"{value}\",\n",
    "}\n",
    "list_l2g = [\n",
    "    0.10,\n",
    "    0.15,\n",
    "    0.20,\n",
    "    0.25,\n",
    "    0.30,\n",
    "    0.35,\n",
    "    0.40,\n",
    "    0.45,\n",
    "    0.50,\n",
    "    0.55,\n",
    "    0.60,\n",
    "    0.65,\n",
    "    0.70,\n",
    "    0.75,\n",
    "    0.80,\n",
    "    0.85,\n",
    "    0.90,\n",
    "    0.95,\n",
    "]\n",
    "\n",
    "dataset = (\n",
    "    discrepancifier(\n",
    "        gwasCredibleAssocDistances\n",
    "        # .filter(F.col(\"h4\").isNotNull()) #### not filter by this because we want to include the L2G AND Coloc question\n",
    "        .withColumn(  ### take maximum L2G score per T-D\n",
    "            \"max_L2GScore\",\n",
    "            F.max(\"resourceScore\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"min_footPrintDistance_rank\",\n",
    "            F.min(\"distanceFootprintMean\").over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"min_tssDistance_rank\",\n",
    "            F.min(\"distanceTssMean\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "            f\"{value}\",\n",
    "            # \"leftStudyLocusId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .count()\n",
    "    )\n",
    "    .join(analysis_chembl, on=[\"targetId\", \"diseaseId\"], how=\"right\")\n",
    "    .withColumn(\n",
    "        \"diagonalAgreeWithDrugs\",\n",
    "        F.when(\n",
    "            (\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\").isNotNull())\n",
    "            )\n",
    "            # & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "            ,\n",
    "            F.when(\n",
    "                (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                & (F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                & (F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"dispar\")),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"oneCellAgreeWithDrugs\",\n",
    "        F.when(\n",
    "            (\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\").isNotNull())\n",
    "            )\n",
    "            # & (F.col(\"coherencyOneCell\") == \"coherent\")\n",
    "            ,\n",
    "            F.when(\n",
    "                (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                & (\n",
    "                    (F.col(\"LoF_protect\").isNotNull())\n",
    "                    & (F.col(\"LoF_risk\").isNull())\n",
    "                    & (F.col(\"GoF_protect\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                & (\n",
    "                    (F.col(\"GoF_protect\").isNotNull())\n",
    "                    & (F.col(\"LoF_risk\").isNull())\n",
    "                    & (F.col(\"LoF_protect\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"dispar\")),\n",
    "        ),\n",
    "        # ).filter(\n",
    "        #    F.col(\"diagonalAgreeWithDrugs\").isNotNull()\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase4\",\n",
    "        F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=3\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=2\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=1\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase0\",\n",
    "        F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"diagonalYes\",\n",
    "        F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")).otherwise(\n",
    "            F.lit(\"no\")\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"oneCellYes\",\n",
    "        F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")).otherwise(\n",
    "            F.lit(\"no\")\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"hasGeneticEvidence\",\n",
    "        F.when(F.col(\"max_L2GScore\").isNotNull(), F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"L2GAndColoc\",\n",
    "        F.when(\n",
    "            (F.col(f\"{value}\").isNotNull())\n",
    "            & (F.col(\"coherencyDiagonal\").isin([\"coherent\", \"dispar\"])),\n",
    "            F.lit(\"yes\"),\n",
    "        ).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .select(\n",
    "        [\"*\"]\n",
    "        + (\n",
    "            [  ### single columns\n",
    "                F.when(F.col(f\"{value}\") >= n, F.lit(\"yes\"))\n",
    "                .otherwise(F.lit(\"no\"))\n",
    "                .alias(f\"{value}>={str(n).replace('.', '_')}\")\n",
    "                for n in list_l2g\n",
    "            ]\n",
    "        )\n",
    "        + (\n",
    "            [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                .otherwise(F.lit(\"no\"))\n",
    "                .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                for a, x in dict_comb.items()\n",
    "                for n in list_l2g\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "### HERE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Make the loop and see if that make sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Make all the datasets\n",
    "values = [\"max_L2GScore\", \"min_footPrintDistance_rank\", \"min_tssDistance_rank\"]\n",
    "\n",
    "datasetDict = {}\n",
    "for value in values:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_l2g_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )\n",
    "    elif value == \"min_distance_ranking\":\n",
    "        datasetDict[f\"df_distance_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_distance_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def comparisons_df(test_propag) -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    toAnalysis = dataset.columns[22:]\n",
    "    dataType = [\"byDatatype\"] * len(toAnalysis)\n",
    "    l_studies = []\n",
    "    l_studies.extend([list(a) for a in zip(toAnalysis, dataType)])\n",
    "\n",
    "    schema = StructType(\n",
    "        [\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "            StructField(\"comparisonType\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    comparisons = spark.createDataFrame(l_studies, schema=schema)\n",
    "    ### include all the columns as predictor\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            # (\"Phase>=3\", \"clinical\"),\n",
    "            # (\"Phase>=2\", \"clinical\"),\n",
    "            # (\"Phase>=1\", \"clinical\"),\n",
    "            # (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "\n",
    "    wComparison = Window.partitionBy(F.col(comparisonColumn))\n",
    "    wPrediction = Window.partitionBy(F.col(predictionColumn))\n",
    "    wPredictionComparison = Window.partitionBy(\n",
    "        F.col(comparisonColumn), F.col(predictionColumn)\n",
    "    )\n",
    "\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "    \"\"\"\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    filePath = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            comparisonType,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            filePath,\n",
    "        ]\n",
    "    )\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/diagonalYes_Phase4.parquet\n",
      "2025-01-24 13:48:02.567385\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread \"serve-DataFrame\" java.net.SocketTimeoutException: Accept timed out\n",
      "\tat java.base/java.net.PlainSocketImpl.socketAccept(Native Method)\n",
      "\tat java.base/java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:474)\n",
      "\tat java.base/java.net.ServerSocket.implAccept(ServerSocket.java:565)\n",
      "\tat java.base/java.net.ServerSocket.accept(ServerSocket.java:533)\n",
      "\tat org.apache.spark.security.SocketAuthServer$$anon$1.run(SocketAuthServer.scala:64)\n",
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/oneCellYes_Phase4.parquet\n",
      "2025-01-24 13:48:29.285365\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/hasGeneticEvidence_Phase4.parquet\n",
      "2025-01-24 13:48:55.759779\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/L2GAndColoc_Phase4.parquet\n",
      "2025-01-24 13:49:19.984382\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1_Phase4.parquet\n",
      "2025-01-24 13:49:46.456749\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15_Phase4.parquet\n",
      "2025-01-24 13:50:13.758986\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2_Phase4.parquet\n",
      "2025-01-24 13:50:38.469266\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25_Phase4.parquet\n",
      "2025-01-24 13:51:02.176871\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3_Phase4.parquet\n",
      "2025-01-24 13:51:25.474986\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35_Phase4.parquet\n",
      "2025-01-24 13:51:50.037965\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4_Phase4.parquet\n",
      "2025-01-24 13:52:14.255274\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45_Phase4.parquet\n",
      "2025-01-24 13:52:37.702930\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5_Phase4.parquet\n",
      "2025-01-24 13:53:02.433922\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55_Phase4.parquet\n",
      "2025-01-24 13:53:25.855291\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6_Phase4.parquet\n",
      "2025-01-24 13:53:49.078757\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65_Phase4.parquet\n",
      "2025-01-24 13:54:13.704120\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7_Phase4.parquet\n",
      "2025-01-24 13:54:36.906718\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75_Phase4.parquet\n",
      "2025-01-24 13:54:59.822479\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8_Phase4.parquet\n",
      "2025-01-24 13:55:26.063421\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85_Phase4.parquet\n",
      "2025-01-24 13:55:49.409704\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9_Phase4.parquet\n",
      "2025-01-24 13:56:13.114374\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95_Phase4.parquet\n",
      "2025-01-24 13:56:38.148650\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:57:03.434358\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:57:26.207100\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:57:50.347361\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:58:14.525416\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:58:37.009774\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:59:00.995739\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:59:24.655329\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 13:59:48.331202\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:00:13.933699\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:00:38.613016\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:01:01.386085\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:01:24.964907\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:01:53.121509\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:02:16.785295\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:02:39.551917\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:03:03.986713\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:03:26.651291\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&hasGeneticEvidence_combined_Phase4.parquet\n",
      "2025-01-24 14:03:50.173742\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:04:14.760426\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:04:41.902378\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:05:06.772834\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:05:32.394684\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:05:58.777523\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:06:24.689999\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:06:49.924568\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:07:15.333694\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:07:41.036188\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:08:07.142632\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:08:32.785918\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:08:58.365627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:09:25.263246\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:09:51.059274\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:10:18.460608\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:10:44.923492\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:11:10.748078\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&diagonalYes_combined_Phase4.parquet\n",
      "2025-01-24 14:11:36.316350\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:12:01.436673\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:12:28.952148\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:12:55.746970\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:13:20.667611\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:13:45.942858\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:14:11.690364\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:14:38.478899\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:15:05.488663\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:15:30.864056\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                6]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:15:56.497687\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:16:21.694651\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:16:48.401367\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:17:14.524949\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                6]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:17:39.106764\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:18:04.167550\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:18:29.769816\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:18:55.840624\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&oneCellYes_combined_Phase4.parquet\n",
      "2025-01-24 14:19:20.762628\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:19:46.075609\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:20:11.583557\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:20:37.562144\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:21:03.195342\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:21:27.993809\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:21:53.874462\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:22:20.350230\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:22:47.253558\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:23:12.846449\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]6]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:23:38.345145\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:24:04.600772\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:24:31.039574\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:24:57.199812\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                6]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:25:21.516819\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:25:47.292305\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:26:13.126445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:26:39.312637\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&L2GAndColoc_combined_Phase4.parquet\n",
      "2025-01-24 14:27:04.090736\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "result_st = []\n",
    "result_ci = []\n",
    "results = []\n",
    "\n",
    "\n",
    "def convertTuple(tup):\n",
    "    st = \",\".join(map(str, tup))\n",
    "    return st\n",
    "\n",
    "\n",
    "aggSetups_original = comparisons_df(df)\n",
    "listado = []\n",
    "today_date = str(date.today())\n",
    "\n",
    "for row in aggSetups_original:\n",
    "    aggregations_original(dataset, value, listado, *row, today_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"type\",\n",
    "        \"criteria\",\n",
    "        \"phase\",\n",
    "        \"OR\",\n",
    "        \"pValue\",\n",
    "        \"LowCI\",\n",
    "        \"HighCI\",\n",
    "        \"total\",\n",
    "        \"array\",\n",
    "        \"rs\",\n",
    "        \"lowRs\",\n",
    "        \"HighRs\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>criteria</th>\n",
       "      <th>phase</th>\n",
       "      <th>OR</th>\n",
       "      <th>pValue</th>\n",
       "      <th>LowCI</th>\n",
       "      <th>HighCI</th>\n",
       "      <th>total</th>\n",
       "      <th>array</th>\n",
       "      <th>rs</th>\n",
       "      <th>lowRs</th>\n",
       "      <th>HighRs</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>diagonalYes</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>2.90</td>\n",
       "      <td>2.282597e-13</td>\n",
       "      <td>2.17</td>\n",
       "      <td>3.88</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[100, 98], [19534, 55546]]</td>\n",
       "      <td>1.94</td>\n",
       "      <td>1.69</td>\n",
       "      <td>2.23</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>oneCellYes</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>3.96</td>\n",
       "      <td>1.302329e-10</td>\n",
       "      <td>2.56</td>\n",
       "      <td>6.18</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[53, 38], [19581, 55606]]</td>\n",
       "      <td>2.24</td>\n",
       "      <td>1.88</td>\n",
       "      <td>2.66</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>hasGeneticEvidence</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>1.98</td>\n",
       "      <td>5.966624e-27</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2.23</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[453, 657], [19181, 54987]]</td>\n",
       "      <td>1.58</td>\n",
       "      <td>1.47</td>\n",
       "      <td>1.70</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>L2GAndColoc</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>2.60</td>\n",
       "      <td>6.797493e-17</td>\n",
       "      <td>2.07</td>\n",
       "      <td>3.25</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[155, 170], [19479, 55474]]</td>\n",
       "      <td>1.84</td>\n",
       "      <td>1.64</td>\n",
       "      <td>2.06</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>max_L2GScore&gt;=0_1</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>2.01</td>\n",
       "      <td>4.654861e-24</td>\n",
       "      <td>1.76</td>\n",
       "      <td>2.29</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[385, 549], [19249, 55095]]</td>\n",
       "      <td>1.59</td>\n",
       "      <td>1.47</td>\n",
       "      <td>1.72</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>max_L2GScore&gt;=0_75&amp;L2GAndColoc_combined</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>2.78</td>\n",
       "      <td>2.032015e-11</td>\n",
       "      <td>2.05</td>\n",
       "      <td>3.77</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[89, 91], [19545, 55553]]</td>\n",
       "      <td>1.90</td>\n",
       "      <td>1.64</td>\n",
       "      <td>2.20</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>max_L2GScore&gt;=0_8&amp;L2GAndColoc_combined</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>2.84</td>\n",
       "      <td>3.538646e-10</td>\n",
       "      <td>2.04</td>\n",
       "      <td>3.96</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[76, 76], [19558, 55568]]</td>\n",
       "      <td>1.92</td>\n",
       "      <td>1.64</td>\n",
       "      <td>2.25</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>max_L2GScore&gt;=0_85&amp;L2GAndColoc_combined</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>3.24</td>\n",
       "      <td>2.767114e-10</td>\n",
       "      <td>2.23</td>\n",
       "      <td>4.71</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[65, 57], [19569, 55587]]</td>\n",
       "      <td>2.05</td>\n",
       "      <td>1.73</td>\n",
       "      <td>2.42</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>max_L2GScore&gt;=0_9&amp;L2GAndColoc_combined</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>3.85</td>\n",
       "      <td>6.480108e-08</td>\n",
       "      <td>2.30</td>\n",
       "      <td>6.52</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[38, 28], [19596, 55616]]</td>\n",
       "      <td>2.21</td>\n",
       "      <td>1.80</td>\n",
       "      <td>2.72</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>byDatatype</td>\n",
       "      <td>max_L2GScore&gt;=0_95&amp;L2GAndColoc_combined</td>\n",
       "      <td>Phase4</td>\n",
       "      <td>6.62</td>\n",
       "      <td>6.038798e-07</td>\n",
       "      <td>2.91</td>\n",
       "      <td>16.42</td>\n",
       "      <td>75278</td>\n",
       "      <td>[[21, 9], [19613, 55635]]</td>\n",
       "      <td>2.69</td>\n",
       "      <td>2.12</td>\n",
       "      <td>3.40</td>\n",
       "      <td>gs://ot-team/jroldan/2025-01-24_analysis/max_L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          type                                 criteria   phase    OR  \\\n",
       "0   byDatatype                              diagonalYes  Phase4  2.90   \n",
       "1   byDatatype                               oneCellYes  Phase4  3.96   \n",
       "2   byDatatype                       hasGeneticEvidence  Phase4  1.98   \n",
       "3   byDatatype                              L2GAndColoc  Phase4  2.60   \n",
       "4   byDatatype                        max_L2GScore>=0_1  Phase4  2.01   \n",
       "..         ...                                      ...     ...   ...   \n",
       "89  byDatatype  max_L2GScore>=0_75&L2GAndColoc_combined  Phase4  2.78   \n",
       "90  byDatatype   max_L2GScore>=0_8&L2GAndColoc_combined  Phase4  2.84   \n",
       "91  byDatatype  max_L2GScore>=0_85&L2GAndColoc_combined  Phase4  3.24   \n",
       "92  byDatatype   max_L2GScore>=0_9&L2GAndColoc_combined  Phase4  3.85   \n",
       "93  byDatatype  max_L2GScore>=0_95&L2GAndColoc_combined  Phase4  6.62   \n",
       "\n",
       "          pValue  LowCI  HighCI  total                         array    rs  \\\n",
       "0   2.282597e-13   2.17    3.88  75278   [[100, 98], [19534, 55546]]  1.94   \n",
       "1   1.302329e-10   2.56    6.18  75278    [[53, 38], [19581, 55606]]  2.24   \n",
       "2   5.966624e-27   1.75    2.23  75278  [[453, 657], [19181, 54987]]  1.58   \n",
       "3   6.797493e-17   2.07    3.25  75278  [[155, 170], [19479, 55474]]  1.84   \n",
       "4   4.654861e-24   1.76    2.29  75278  [[385, 549], [19249, 55095]]  1.59   \n",
       "..           ...    ...     ...    ...                           ...   ...   \n",
       "89  2.032015e-11   2.05    3.77  75278    [[89, 91], [19545, 55553]]  1.90   \n",
       "90  3.538646e-10   2.04    3.96  75278    [[76, 76], [19558, 55568]]  1.92   \n",
       "91  2.767114e-10   2.23    4.71  75278    [[65, 57], [19569, 55587]]  2.05   \n",
       "92  6.480108e-08   2.30    6.52  75278    [[38, 28], [19596, 55616]]  2.21   \n",
       "93  6.038798e-07   2.91   16.42  75278     [[21, 9], [19613, 55635]]  2.69   \n",
       "\n",
       "    lowRs  HighRs                                               path  \n",
       "0    1.69    2.23  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "1    1.88    2.66  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "2    1.47    1.70  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "3    1.64    2.06  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "4    1.47    1.72  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "..    ...     ...                                                ...  \n",
       "89   1.64    2.20  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "90   1.64    2.25  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "91   1.73    2.42  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "92   1.80    2.72  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "93   2.12    3.40  gs://ot-team/jroldan/2025-01-24_analysis/max_L...  \n",
       "\n",
       "[94 rows x 13 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['byDatatype',\n",
       "  'diagonalYes',\n",
       "  'Phase4',\n",
       "  2.9,\n",
       "  2.2825971081982413e-13,\n",
       "  2.17,\n",
       "  3.88,\n",
       "  '75278',\n",
       "  [[100, 98], [19534, 55546]],\n",
       "  1.94,\n",
       "  1.69,\n",
       "  2.23,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/diagonalYes_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'oneCellYes',\n",
       "  'Phase4',\n",
       "  3.96,\n",
       "  1.3023288254998088e-10,\n",
       "  2.56,\n",
       "  6.18,\n",
       "  '75278',\n",
       "  [[53, 38], [19581, 55606]],\n",
       "  2.24,\n",
       "  1.88,\n",
       "  2.66,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/oneCellYes_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'hasGeneticEvidence',\n",
       "  'Phase4',\n",
       "  1.98,\n",
       "  5.966623858885305e-27,\n",
       "  1.75,\n",
       "  2.23,\n",
       "  '75278',\n",
       "  [[453, 657], [19181, 54987]],\n",
       "  1.58,\n",
       "  1.47,\n",
       "  1.7,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/hasGeneticEvidence_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'L2GAndColoc',\n",
       "  'Phase4',\n",
       "  2.6,\n",
       "  6.797492523446592e-17,\n",
       "  2.07,\n",
       "  3.25,\n",
       "  '75278',\n",
       "  [[155, 170], [19479, 55474]],\n",
       "  1.84,\n",
       "  1.64,\n",
       "  2.06,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/L2GAndColoc_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_1',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  4.6548607258271906e-24,\n",
       "  1.76,\n",
       "  2.29,\n",
       "  '75278',\n",
       "  [[385, 549], [19249, 55095]],\n",
       "  1.59,\n",
       "  1.47,\n",
       "  1.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_15',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  7.629298987536408e-22,\n",
       "  1.74,\n",
       "  2.31,\n",
       "  '75278',\n",
       "  [[348, 496], [19286, 55148]],\n",
       "  1.59,\n",
       "  1.47,\n",
       "  1.73,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_2',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  5.959374507776837e-21,\n",
       "  1.74,\n",
       "  2.32,\n",
       "  '75278',\n",
       "  [[332, 472], [19302, 55172]],\n",
       "  1.59,\n",
       "  1.47,\n",
       "  1.73,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_25',\n",
       "  'Phase4',\n",
       "  2.03,\n",
       "  7.703926580815119e-21,\n",
       "  1.76,\n",
       "  2.36,\n",
       "  '75278',\n",
       "  [[319, 448], [19315, 55196]],\n",
       "  1.6,\n",
       "  1.47,\n",
       "  1.75,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_3',\n",
       "  'Phase4',\n",
       "  2.05,\n",
       "  8.580671168975091e-21,\n",
       "  1.77,\n",
       "  2.38,\n",
       "  '75278',\n",
       "  [[313, 436], [19321, 55208]],\n",
       "  1.61,\n",
       "  1.48,\n",
       "  1.76,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_35',\n",
       "  'Phase4',\n",
       "  2.03,\n",
       "  4.357979880765691e-19,\n",
       "  1.74,\n",
       "  2.36,\n",
       "  '75278',\n",
       "  [[293, 413], [19341, 55231]],\n",
       "  1.6,\n",
       "  1.46,\n",
       "  1.75,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_4',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  2.8314721827452475e-18,\n",
       "  1.72,\n",
       "  2.35,\n",
       "  '75278',\n",
       "  [[285, 405], [19349, 55239]],\n",
       "  1.59,\n",
       "  1.46,\n",
       "  1.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_45',\n",
       "  'Phase4',\n",
       "  2.0,\n",
       "  8.485954806265006e-18,\n",
       "  1.71,\n",
       "  2.34,\n",
       "  '75278',\n",
       "  [[278, 396], [19356, 55248]],\n",
       "  1.59,\n",
       "  1.45,\n",
       "  1.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_5',\n",
       "  'Phase4',\n",
       "  2.0,\n",
       "  3.901753133431875e-17,\n",
       "  1.7,\n",
       "  2.34,\n",
       "  '75278',\n",
       "  [[269, 384], [19365, 55260]],\n",
       "  1.59,\n",
       "  1.45,\n",
       "  1.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_55',\n",
       "  'Phase4',\n",
       "  1.95,\n",
       "  1.6857436200812383e-15,\n",
       "  1.66,\n",
       "  2.3,\n",
       "  '75278',\n",
       "  [[256, 374], [19378, 55270]],\n",
       "  1.57,\n",
       "  1.42,\n",
       "  1.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_6',\n",
       "  'Phase4',\n",
       "  1.94,\n",
       "  1.8889457400065935e-14,\n",
       "  1.64,\n",
       "  2.29,\n",
       "  '75278',\n",
       "  [[241, 355], [19393, 55289]],\n",
       "  1.56,\n",
       "  1.41,\n",
       "  1.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_65',\n",
       "  'Phase4',\n",
       "  1.98,\n",
       "  1.0117386066034168e-14,\n",
       "  1.67,\n",
       "  2.36,\n",
       "  '75278',\n",
       "  [[231, 332], [19403, 55312]],\n",
       "  1.58,\n",
       "  1.43,\n",
       "  1.75,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_7',\n",
       "  'Phase4',\n",
       "  2.02,\n",
       "  1.8629168663379455e-14,\n",
       "  1.69,\n",
       "  2.41,\n",
       "  '75278',\n",
       "  [[218, 308], [19416, 55336]],\n",
       "  1.6,\n",
       "  1.44,\n",
       "  1.77,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_75',\n",
       "  'Phase4',\n",
       "  2.1,\n",
       "  1.4180192968301012e-13,\n",
       "  1.73,\n",
       "  2.56,\n",
       "  '75278',\n",
       "  [[184, 249], [19450, 55395]],\n",
       "  1.64,\n",
       "  1.46,\n",
       "  1.83,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_8',\n",
       "  'Phase4',\n",
       "  2.27,\n",
       "  2.99505174983308e-13,\n",
       "  1.82,\n",
       "  2.83,\n",
       "  '75278',\n",
       "  [[151, 189], [19483, 55455]],\n",
       "  1.71,\n",
       "  1.52,\n",
       "  1.93,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_85',\n",
       "  'Phase4',\n",
       "  2.5,\n",
       "  3.204930082139683e-12,\n",
       "  1.93,\n",
       "  3.23,\n",
       "  '75278',\n",
       "  [[115, 131], [19519, 55513]],\n",
       "  1.8,\n",
       "  1.57,\n",
       "  2.05,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_9',\n",
       "  'Phase4',\n",
       "  3.32,\n",
       "  3.071297568724775e-10,\n",
       "  2.26,\n",
       "  4.89,\n",
       "  '75278',\n",
       "  [[62, 53], [19572, 55591]],\n",
       "  2.07,\n",
       "  1.75,\n",
       "  2.45,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_95',\n",
       "  'Phase4',\n",
       "  4.67,\n",
       "  3.657794719519248e-07,\n",
       "  2.47,\n",
       "  9.1,\n",
       "  '75278',\n",
       "  [[28, 17], [19606, 55627]],\n",
       "  2.39,\n",
       "  1.9,\n",
       "  3.0,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_1&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  4.6548607258271906e-24,\n",
       "  1.76,\n",
       "  2.29,\n",
       "  '75278',\n",
       "  [[385, 549], [19249, 55095]],\n",
       "  1.59,\n",
       "  1.47,\n",
       "  1.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_15&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  7.629298987536408e-22,\n",
       "  1.74,\n",
       "  2.31,\n",
       "  '75278',\n",
       "  [[348, 496], [19286, 55148]],\n",
       "  1.59,\n",
       "  1.47,\n",
       "  1.73,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_2&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  5.959374507776837e-21,\n",
       "  1.74,\n",
       "  2.32,\n",
       "  '75278',\n",
       "  [[332, 472], [19302, 55172]],\n",
       "  1.59,\n",
       "  1.47,\n",
       "  1.73,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_25&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.03,\n",
       "  7.703926580815119e-21,\n",
       "  1.76,\n",
       "  2.36,\n",
       "  '75278',\n",
       "  [[319, 448], [19315, 55196]],\n",
       "  1.6,\n",
       "  1.47,\n",
       "  1.75,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_3&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.05,\n",
       "  8.580671168975091e-21,\n",
       "  1.77,\n",
       "  2.38,\n",
       "  '75278',\n",
       "  [[313, 436], [19321, 55208]],\n",
       "  1.61,\n",
       "  1.48,\n",
       "  1.76,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_35&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.03,\n",
       "  4.357979880765691e-19,\n",
       "  1.74,\n",
       "  2.36,\n",
       "  '75278',\n",
       "  [[293, 413], [19341, 55231]],\n",
       "  1.6,\n",
       "  1.46,\n",
       "  1.75,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_4&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.01,\n",
       "  2.8314721827452475e-18,\n",
       "  1.72,\n",
       "  2.35,\n",
       "  '75278',\n",
       "  [[285, 405], [19349, 55239]],\n",
       "  1.59,\n",
       "  1.46,\n",
       "  1.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_45&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.0,\n",
       "  8.485954806265006e-18,\n",
       "  1.71,\n",
       "  2.34,\n",
       "  '75278',\n",
       "  [[278, 396], [19356, 55248]],\n",
       "  1.59,\n",
       "  1.45,\n",
       "  1.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_5&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.0,\n",
       "  3.901753133431875e-17,\n",
       "  1.7,\n",
       "  2.34,\n",
       "  '75278',\n",
       "  [[269, 384], [19365, 55260]],\n",
       "  1.59,\n",
       "  1.45,\n",
       "  1.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_55&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  1.95,\n",
       "  1.6857436200812383e-15,\n",
       "  1.66,\n",
       "  2.3,\n",
       "  '75278',\n",
       "  [[256, 374], [19378, 55270]],\n",
       "  1.57,\n",
       "  1.42,\n",
       "  1.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_6&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  1.94,\n",
       "  1.8889457400065935e-14,\n",
       "  1.64,\n",
       "  2.29,\n",
       "  '75278',\n",
       "  [[241, 355], [19393, 55289]],\n",
       "  1.56,\n",
       "  1.41,\n",
       "  1.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_65&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  1.98,\n",
       "  1.0117386066034168e-14,\n",
       "  1.67,\n",
       "  2.36,\n",
       "  '75278',\n",
       "  [[231, 332], [19403, 55312]],\n",
       "  1.58,\n",
       "  1.43,\n",
       "  1.75,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_7&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.02,\n",
       "  1.8629168663379455e-14,\n",
       "  1.69,\n",
       "  2.41,\n",
       "  '75278',\n",
       "  [[218, 308], [19416, 55336]],\n",
       "  1.6,\n",
       "  1.44,\n",
       "  1.77,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_75&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.1,\n",
       "  1.4180192968301012e-13,\n",
       "  1.73,\n",
       "  2.56,\n",
       "  '75278',\n",
       "  [[184, 249], [19450, 55395]],\n",
       "  1.64,\n",
       "  1.46,\n",
       "  1.83,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_8&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.27,\n",
       "  2.99505174983308e-13,\n",
       "  1.82,\n",
       "  2.83,\n",
       "  '75278',\n",
       "  [[151, 189], [19483, 55455]],\n",
       "  1.71,\n",
       "  1.52,\n",
       "  1.93,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_85&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  2.5,\n",
       "  3.204930082139683e-12,\n",
       "  1.93,\n",
       "  3.23,\n",
       "  '75278',\n",
       "  [[115, 131], [19519, 55513]],\n",
       "  1.8,\n",
       "  1.57,\n",
       "  2.05,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_9&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  3.32,\n",
       "  3.071297568724775e-10,\n",
       "  2.26,\n",
       "  4.89,\n",
       "  '75278',\n",
       "  [[62, 53], [19572, 55591]],\n",
       "  2.07,\n",
       "  1.75,\n",
       "  2.45,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_95&hasGeneticEvidence_combined',\n",
       "  'Phase4',\n",
       "  4.67,\n",
       "  3.657794719519248e-07,\n",
       "  2.47,\n",
       "  9.1,\n",
       "  '75278',\n",
       "  [[28, 17], [19606, 55627]],\n",
       "  2.39,\n",
       "  1.9,\n",
       "  3.0,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&hasGeneticEvidence_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_1&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.88,\n",
       "  8.909955747340488e-12,\n",
       "  2.11,\n",
       "  3.91,\n",
       "  '75278',\n",
       "  [[88, 87], [19546, 55557]],\n",
       "  1.93,\n",
       "  1.67,\n",
       "  2.24,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_15&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.81,\n",
       "  7.173331526879397e-11,\n",
       "  2.05,\n",
       "  3.85,\n",
       "  '75278',\n",
       "  [[83, 84], [19551, 55560]],\n",
       "  1.91,\n",
       "  1.64,\n",
       "  2.23,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_2&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.77,\n",
       "  1.28208460492264e-10,\n",
       "  2.02,\n",
       "  3.81,\n",
       "  '75278',\n",
       "  [[82, 84], [19552, 55560]],\n",
       "  1.9,\n",
       "  1.63,\n",
       "  2.21,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_25&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.74,\n",
       "  5.671426211228725e-10,\n",
       "  1.98,\n",
       "  3.78,\n",
       "  '75278',\n",
       "  [[78, 81], [19556, 55563]],\n",
       "  1.88,\n",
       "  1.61,\n",
       "  2.21,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_3&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.81,\n",
       "  3.81222940253273e-10,\n",
       "  2.02,\n",
       "  3.89,\n",
       "  '75278',\n",
       "  [[78, 79], [19556, 55565]],\n",
       "  1.91,\n",
       "  1.63,\n",
       "  2.23,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_35&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.84,\n",
       "  4.956804433928456e-10,\n",
       "  2.03,\n",
       "  3.97,\n",
       "  '75278',\n",
       "  [[75, 75], [19559, 55569]],\n",
       "  1.92,\n",
       "  1.64,\n",
       "  2.25,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_4&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.8,\n",
       "  8.465350609949025e-10,\n",
       "  2.0,\n",
       "  3.92,\n",
       "  '75278',\n",
       "  [[74, 75], [19560, 55569]],\n",
       "  1.91,\n",
       "  1.62,\n",
       "  2.24,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_45&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.8,\n",
       "  8.465350609949025e-10,\n",
       "  2.0,\n",
       "  3.92,\n",
       "  '75278',\n",
       "  [[74, 75], [19560, 55569]],\n",
       "  1.91,\n",
       "  1.62,\n",
       "  2.24,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_5&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.8,\n",
       "  8.465350609949025e-10,\n",
       "  2.0,\n",
       "  3.92,\n",
       "  '75278',\n",
       "  [[74, 75], [19560, 55569]],\n",
       "  1.91,\n",
       "  1.62,\n",
       "  2.24,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_55&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.76,\n",
       "  2.028500218281168e-09,\n",
       "  1.97,\n",
       "  3.88,\n",
       "  '75278',\n",
       "  [[72, 74], [19562, 55570]],\n",
       "  1.89,\n",
       "  1.61,\n",
       "  2.23,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_6&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.61,\n",
       "  3.519703685080545e-08,\n",
       "  1.84,\n",
       "  3.68,\n",
       "  '75278',\n",
       "  [[67, 73], [19567, 55571]],\n",
       "  1.84,\n",
       "  1.55,\n",
       "  2.19,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_65&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.64,\n",
       "  4.182110949027236e-08,\n",
       "  1.86,\n",
       "  3.75,\n",
       "  '75278',\n",
       "  [[66, 71], [19568, 55573]],\n",
       "  1.85,\n",
       "  1.55,\n",
       "  2.2,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_7&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.71,\n",
       "  3.539047844757356e-08,\n",
       "  1.89,\n",
       "  3.88,\n",
       "  '75278',\n",
       "  [[64, 67], [19570, 55577]],\n",
       "  1.88,\n",
       "  1.57,\n",
       "  2.24,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_75&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.79,\n",
       "  6.371555170984652e-08,\n",
       "  1.91,\n",
       "  4.08,\n",
       "  '75278',\n",
       "  [[58, 59], [19576, 55585]],\n",
       "  1.9,\n",
       "  1.58,\n",
       "  2.29,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_8&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  2.68,\n",
       "  9.604788189767918e-07,\n",
       "  1.79,\n",
       "  4.01,\n",
       "  '75278',\n",
       "  [[51, 54], [19583, 55590]],\n",
       "  1.86,\n",
       "  1.53,\n",
       "  2.27,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_85&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  3.19,\n",
       "  1.5721164289004196e-07,\n",
       "  2.04,\n",
       "  5.02,\n",
       "  '75278',\n",
       "  [[45, 40], [19589, 55604]],\n",
       "  2.03,\n",
       "  1.66,\n",
       "  2.48,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_9&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  4.97,\n",
       "  1.8417412700463693e-07,\n",
       "  2.59,\n",
       "  9.83,\n",
       "  '75278',\n",
       "  [[28, 16], [19606, 55628]],\n",
       "  2.44,\n",
       "  1.95,\n",
       "  3.05,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_95&diagonalYes_combined',\n",
       "  'Phase4',\n",
       "  8.51,\n",
       "  7.650448386598998e-07,\n",
       "  3.24,\n",
       "  26.19,\n",
       "  '75278',\n",
       "  [[18, 6], [19616, 55638]],\n",
       "  2.88,\n",
       "  2.28,\n",
       "  3.63,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&diagonalYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_1&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.84,\n",
       "  3.726912813901655e-09,\n",
       "  2.41,\n",
       "  6.17,\n",
       "  '75278',\n",
       "  [[46, 34], [19588, 55610]],\n",
       "  2.21,\n",
       "  1.83,\n",
       "  2.67,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_15&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.81,\n",
       "  1.4219607585175595e-08,\n",
       "  2.36,\n",
       "  6.23,\n",
       "  '75278',\n",
       "  [[43, 32], [19591, 55612]],\n",
       "  2.2,\n",
       "  1.81,\n",
       "  2.68,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_2&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.81,\n",
       "  1.4219607585175595e-08,\n",
       "  2.36,\n",
       "  6.23,\n",
       "  '75278',\n",
       "  [[43, 32], [19591, 55612]],\n",
       "  2.2,\n",
       "  1.81,\n",
       "  2.68,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_25&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.11,\n",
       "  4.67383658510029e-09,\n",
       "  2.5,\n",
       "  6.85,\n",
       "  '75278',\n",
       "  [[42, 29], [19592, 55615]],\n",
       "  2.27,\n",
       "  1.87,\n",
       "  2.76,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_3&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.11,\n",
       "  4.67383658510029e-09,\n",
       "  2.5,\n",
       "  6.85,\n",
       "  '75278',\n",
       "  [[42, 29], [19592, 55615]],\n",
       "  2.27,\n",
       "  1.87,\n",
       "  2.76,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_35&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.05,\n",
       "  1.3403971627186301e-08,\n",
       "  2.44,\n",
       "  6.83,\n",
       "  '75278',\n",
       "  [[40, 28], [19594, 55616]],\n",
       "  2.26,\n",
       "  1.85,\n",
       "  2.76,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_4&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.95,\n",
       "  2.9409332826299174e-08,\n",
       "  2.37,\n",
       "  6.67,\n",
       "  '75278',\n",
       "  [[39, 28], [19595, 55616]],\n",
       "  2.23,\n",
       "  1.82,\n",
       "  2.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_45&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.95,\n",
       "  2.9409332826299174e-08,\n",
       "  2.37,\n",
       "  6.67,\n",
       "  '75278',\n",
       "  [[39, 28], [19595, 55616]],\n",
       "  2.23,\n",
       "  1.82,\n",
       "  2.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_5&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.95,\n",
       "  2.9409332826299174e-08,\n",
       "  2.37,\n",
       "  6.67,\n",
       "  '75278',\n",
       "  [[39, 28], [19595, 55616]],\n",
       "  2.23,\n",
       "  1.82,\n",
       "  2.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_55&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.85,\n",
       "  6.480107510562942e-08,\n",
       "  2.3,\n",
       "  6.52,\n",
       "  '75278',\n",
       "  [[38, 28], [19596, 55616]],\n",
       "  2.21,\n",
       "  1.8,\n",
       "  2.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_6&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  3.89,\n",
       "  8.430784357214069e-08,\n",
       "  2.3,\n",
       "  6.64,\n",
       "  '75278',\n",
       "  [[37, 27], [19597, 55617]],\n",
       "  2.22,\n",
       "  1.8,\n",
       "  2.74,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_65&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.04,\n",
       "  5.0647104071245976e-08,\n",
       "  2.38,\n",
       "  6.95,\n",
       "  '75278',\n",
       "  [[37, 26], [19597, 55618]],\n",
       "  2.25,\n",
       "  1.83,\n",
       "  2.77,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_7&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.14,\n",
       "  9.122423496793306e-08,\n",
       "  2.39,\n",
       "  7.27,\n",
       "  '75278',\n",
       "  [[35, 24], [19599, 55620]],\n",
       "  2.28,\n",
       "  1.84,\n",
       "  2.81,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_75&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.07,\n",
       "  2.5906807890473836e-07,\n",
       "  2.32,\n",
       "  7.26,\n",
       "  '75278',\n",
       "  [[33, 23], [19601, 55621]],\n",
       "  2.26,\n",
       "  1.82,\n",
       "  2.82,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_8&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.54,\n",
       "  7.860556789963332e-08,\n",
       "  2.52,\n",
       "  8.38,\n",
       "  '75278',\n",
       "  [[32, 20], [19602, 55624]],\n",
       "  2.36,\n",
       "  1.9,\n",
       "  2.93,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_85&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  4.89,\n",
       "  4.965438473809932e-08,\n",
       "  2.65,\n",
       "  9.28,\n",
       "  '75278',\n",
       "  [[31, 18], [19603, 55626]],\n",
       "  2.43,\n",
       "  1.96,\n",
       "  3.01,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_9&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  16.31,\n",
       "  2.0764948692818374e-10,\n",
       "  5.57,\n",
       "  64.9,\n",
       "  '75278',\n",
       "  [[23, 4], [19611, 55640]],\n",
       "  3.27,\n",
       "  2.79,\n",
       "  3.83,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_95&oneCellYes_combined',\n",
       "  'Phase4',\n",
       "  22.69,\n",
       "  3.9797184370540546e-08,\n",
       "  5.33,\n",
       "  203.43,\n",
       "  '75278',\n",
       "  [[16, 2], [19618, 55642]],\n",
       "  3.41,\n",
       "  2.89,\n",
       "  4.02,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&oneCellYes_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_1&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.62,\n",
       "  1.8103354814778916e-15,\n",
       "  2.06,\n",
       "  3.32,\n",
       "  '75278',\n",
       "  [[139, 151], [19495, 55493]],\n",
       "  1.84,\n",
       "  1.63,\n",
       "  2.08,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_1&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_15&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.57,\n",
       "  1.7468389417731163e-14,\n",
       "  2.02,\n",
       "  3.28,\n",
       "  '75278',\n",
       "  [[133, 147], [19501, 55497]],\n",
       "  1.83,\n",
       "  1.61,\n",
       "  2.07,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_15&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_2&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.59,\n",
       "  2.464102921456772e-14,\n",
       "  2.02,\n",
       "  3.31,\n",
       "  '75278',\n",
       "  [[130, 143], [19504, 55501]],\n",
       "  1.83,\n",
       "  1.62,\n",
       "  2.08,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_2&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_25&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.56,\n",
       "  1.4536824809500435e-13,\n",
       "  1.99,\n",
       "  3.28,\n",
       "  '75278',\n",
       "  [[125, 139], [19509, 55505]],\n",
       "  1.82,\n",
       "  1.6,\n",
       "  2.07,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_25&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_3&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.61,\n",
       "  6.936009257018499e-14,\n",
       "  2.03,\n",
       "  3.36,\n",
       "  '75278',\n",
       "  [[123, 134], [19511, 55510]],\n",
       "  1.84,\n",
       "  1.62,\n",
       "  2.09,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_3&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_35&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.69,\n",
       "  3.578082306588671e-14,\n",
       "  2.08,\n",
       "  3.48,\n",
       "  '75278',\n",
       "  [[120, 127], [19514, 55517]],\n",
       "  1.87,\n",
       "  1.64,\n",
       "  2.12,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_35&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_4&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.69,\n",
       "  4.862728469939691e-14,\n",
       "  2.07,\n",
       "  3.48,\n",
       "  '75278',\n",
       "  [[119, 126], [19515, 55518]],\n",
       "  1.87,\n",
       "  1.64,\n",
       "  2.13,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_4&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_45&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.64,\n",
       "  2.682111335930093e-13,\n",
       "  2.03,\n",
       "  3.43,\n",
       "  '75278',\n",
       "  [[115, 124], [19519, 55520]],\n",
       "  1.85,\n",
       "  1.62,\n",
       "  2.11,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_45&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_5&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.59,\n",
       "  1.0779650797981123e-12,\n",
       "  1.99,\n",
       "  3.37,\n",
       "  '75278',\n",
       "  [[113, 124], [19521, 55520]],\n",
       "  1.83,\n",
       "  1.6,\n",
       "  2.1,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_5&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_55&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.59,\n",
       "  2.0608224138431207e-12,\n",
       "  1.98,\n",
       "  3.38,\n",
       "  '75278',\n",
       "  [[111, 122], [19523, 55522]],\n",
       "  1.83,\n",
       "  1.6,\n",
       "  2.1,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_55&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_6&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.51,\n",
       "  2.328102897616538e-11,\n",
       "  1.91,\n",
       "  3.29,\n",
       "  '75278',\n",
       "  [[105, 119], [19529, 55525]],\n",
       "  1.8,\n",
       "  1.57,\n",
       "  2.07,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_6&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_65&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.59,\n",
       "  1.294308478350323e-11,\n",
       "  1.96,\n",
       "  3.42,\n",
       "  '75278',\n",
       "  [[102, 112], [19532, 55532]],\n",
       "  1.83,\n",
       "  1.59,\n",
       "  2.11,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_65&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_7&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.71,\n",
       "  5.487736775642916e-12,\n",
       "  2.03,\n",
       "  3.6,\n",
       "  '75278',\n",
       "  [[99, 104], [19535, 55540]],\n",
       "  1.87,\n",
       "  1.63,\n",
       "  2.16,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_7&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_75&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.78,\n",
       "  2.0320147644252384e-11,\n",
       "  2.05,\n",
       "  3.77,\n",
       "  '75278',\n",
       "  [[89, 91], [19545, 55553]],\n",
       "  1.9,\n",
       "  1.64,\n",
       "  2.2,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_75&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_8&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  2.84,\n",
       "  3.538645525059619e-10,\n",
       "  2.04,\n",
       "  3.96,\n",
       "  '75278',\n",
       "  [[76, 76], [19558, 55568]],\n",
       "  1.92,\n",
       "  1.64,\n",
       "  2.25,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_8&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_85&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  3.24,\n",
       "  2.767113521279882e-10,\n",
       "  2.23,\n",
       "  4.71,\n",
       "  '75278',\n",
       "  [[65, 57], [19569, 55587]],\n",
       "  2.05,\n",
       "  1.73,\n",
       "  2.42,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_85&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_9&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  3.85,\n",
       "  6.480107510562942e-08,\n",
       "  2.3,\n",
       "  6.52,\n",
       "  '75278',\n",
       "  [[38, 28], [19596, 55616]],\n",
       "  2.21,\n",
       "  1.8,\n",
       "  2.72,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_9&L2GAndColoc_combined_Phase4.parquet'],\n",
       " ['byDatatype',\n",
       "  'max_L2GScore>=0_95&L2GAndColoc_combined',\n",
       "  'Phase4',\n",
       "  6.62,\n",
       "  6.038797564487963e-07,\n",
       "  2.91,\n",
       "  16.42,\n",
       "  '75278',\n",
       "  [[21, 9], [19613, 55635]],\n",
       "  2.69,\n",
       "  2.12,\n",
       "  3.4,\n",
       "  'gs://ot-team/jroldan/2025-01-24_analysis/max_L2GScore/max_L2GScore>=0_95&L2GAndColoc_combined_Phase4.parquet']]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['diagonalYes',\n",
       " 'oneCellYes',\n",
       " 'hasGeneticEvidence',\n",
       " 'L2GAndColoc',\n",
       " 'max_L2GScore>=0_1',\n",
       " 'max_L2GScore>=0_15',\n",
       " 'max_L2GScore>=0_2',\n",
       " 'max_L2GScore>=0_25',\n",
       " 'max_L2GScore>=0_3',\n",
       " 'max_L2GScore>=0_35',\n",
       " 'max_L2GScore>=0_4',\n",
       " 'max_L2GScore>=0_45',\n",
       " 'max_L2GScore>=0_5',\n",
       " 'max_L2GScore>=0_55',\n",
       " 'max_L2GScore>=0_6',\n",
       " 'max_L2GScore>=0_65',\n",
       " 'max_L2GScore>=0_7',\n",
       " 'max_L2GScore>=0_75',\n",
       " 'max_L2GScore>=0_8',\n",
       " 'max_L2GScore>=0_85',\n",
       " 'max_L2GScore>=0_9',\n",
       " 'max_L2GScore>=0_95',\n",
       " 'max_L2GScore>=0_1&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_15&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_2&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_25&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_3&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_35&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_4&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_45&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_5&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_55&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_6&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_65&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_7&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_75&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_8&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_85&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_9&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_95&hasGeneticEvidence_combined',\n",
       " 'max_L2GScore>=0_1&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_15&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_2&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_25&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_3&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_35&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_4&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_45&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_5&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_55&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_6&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_65&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_7&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_75&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_8&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_85&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_9&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_95&diagonalYes_combined',\n",
       " 'max_L2GScore>=0_1&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_15&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_2&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_25&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_3&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_35&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_4&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_45&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_5&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_55&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_6&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_65&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_7&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_75&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_8&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_85&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_9&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_95&oneCellYes_combined',\n",
       " 'max_L2GScore>=0_1&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_15&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_2&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_25&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_3&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_35&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_4&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_45&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_5&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_55&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_6&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_65&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_7&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_75&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_8&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_85&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_9&L2GAndColoc_combined',\n",
       " 'max_L2GScore>=0_95&L2GAndColoc_combined']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.columns[22:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1968:==================================================>   (16 + 1) / 17]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+-------+\n",
      "|coherencyDiagonal|  count|\n",
      "+-----------------+-------+\n",
      "|       EvidNotDoE|1359051|\n",
      "|         coherent|1036658|\n",
      "+-----------------+-------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(\n",
    "    gwasCredibleAssocDistances.groupBy(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        # \"leftStudyLocusId\",\n",
    "        \"distanceFootprintMean\",\n",
    "        \"distanceTssMean\",\n",
    "    )\n",
    "    .pivot(\"homogenized\")\n",
    "    .count()\n",
    ").groupBy(\"coherencyDiagonal\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "from DoEAssessment import directionOfEffect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+----------------+-----------+----+------------+-------------+-------------+---------------------+------+---------------------+---------------+------------+--------------------+-------------------+--------------+--------------+\n",
      "|       targetId|    diseaseId|leftStudyLocusId|homogenized|  h4|datasourceId|resourceScore|leftVariantId|credibleLeftStudyType|geneId|distanceFootprintMean|distanceTssMean|maxClinPhase|coherencyDiagonal_ch|coherencyOneCell_ch|LoF_protect_ch|GoF_protect_ch|\n",
      "+---------------+-------------+----------------+-----------+----+------------+-------------+-------------+---------------------+------+---------------------+---------------+------------+--------------------+-------------------+--------------+--------------+\n",
      "|ENSG00000007314|  EFO_0000555|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|             3|          null|\n",
      "|ENSG00000007314|  EFO_0003102|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         0.5|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000007314|  EFO_0003894|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         4.0|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000007314|  EFO_0004699|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|            coherent|           coherent|             3|          null|\n",
      "|ENSG00000007314|  EFO_0801084|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000007314|  EFO_1000249|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000008018|  EFO_1000453|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000010310|  EFO_0003884|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|          null|             1|\n",
      "|ENSG00000012504|  EFO_0004210|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         4.0|            coherent|           coherent|          null|             1|\n",
      "|ENSG00000012504|MONDO_0019052|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         4.0|            coherent|           coherent|          null|             1|\n",
      "|ENSG00000012779|MONDO_0004235|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|            coherent|           coherent|             4|          null|\n",
      "|ENSG00000014138|  EFO_1001945|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|            coherent|           coherent|             4|          null|\n",
      "|ENSG00000023228|  EFO_1000657|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|             3|          null|\n",
      "|ENSG00000023445|  EFO_0000616|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         1.0|            coherent|           coherent|             2|          null|\n",
      "|ENSG00000025708|  EFO_0000702|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|             2|          null|\n",
      "|ENSG00000037280|  EFO_0003897|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|            11|          null|\n",
      "|ENSG00000058091|  EFO_0000574|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         1.0|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000062822|  EFO_0004991|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|            coherent|           coherent|             1|          null|\n",
      "|ENSG00000062822|  EFO_0005537|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|            coherent|           coherent|             7|          null|\n",
      "|ENSG00000062822|  EFO_0005952|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|            coherent|           coherent|            37|          null|\n",
      "+---------------+-------------+----------------+-----------+----+------------+-------------+-------------+---------------------+------+---------------------+---------------+------------+--------------------+-------------------+--------------+--------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssocDistances.groupBy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/23 17:24:59 WARN CacheManager: Asked to cache already cached data.\n",
      "25/01/23 17:25:00 WARN CacheManager: Asked to cache already cached data.\n",
      "25/01/23 17:25:00 WARN CacheManager: Asked to cache already cached data.\n",
      "25/01/23 17:25:00 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "ename": "AnalysisException",
     "evalue": "Column 'GoF_risk' does not exist. Did you mean one of the following? [geneId, h4, targetId, GoF_protect_ch, diseaseId, homogenized, LoF_protect_ch, maxClinPhase, datasourceId, leftVariantId, resourceScore, distanceTssMean, geneticEvidence, leftStudyLocusId, coherencyDiagonal_ch, coherencyOneCell_ch, distanceFootprintMean, credibleLeftStudyType];\n'Project [targetId#47413, diseaseId#47513, leftStudyLocusId#48830, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301, maxClinPhase#113689, coherencyDiagonal_ch#117011, coherencyOneCell_ch#117012, LoF_protect_ch#117013L, GoF_protect_ch#117014L, geneticEvidence#117039, CASE WHEN (coherencyDiagonal_ch#117011 = coherent) THEN CASE WHEN (isnotnull(LoF_protect_ch#117013L) AND (isnotnull('GoF_risk) OR isnotnull('LoF_protect))) THEN coherent WHEN (isnotnull(GoF_protect_ch#117014L) AND (isnotnull('LoF_risk) OR isnotnull('GoF_protect))) THEN coherent ELSE dispar END END AS diagonalAgreeWithDrugs#117058]\n+- Project [targetId#47413, diseaseId#47513, leftStudyLocusId#48830, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301, maxClinPhase#113689, coherencyDiagonal_ch#117011, coherencyOneCell_ch#117012, LoF_protect_ch#117013L, GoF_protect_ch#117014L, CASE WHEN isnotnull(resourceScore#47697) THEN yes END AS geneticEvidence#117039]\n   +- Project [targetId#47413, diseaseId#47513, leftStudyLocusId#48830, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301, maxClinPhase#113689, coherencyDiagonal_ch#117011, coherencyOneCell_ch#117012, LoF_protect_ch#117013L, GoF_protect_ch#117014L]\n      +- Join RightOuter, ((targetId#47628 = targetId#47413) AND (diseaseId#48955 = diseaseId#47513))\n         :- Project [leftStudyLocusId#48830, targetId#47628, diseaseId#48955, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301]\n         :  +- Join LeftOuter, (leftStudyLocusId#48830 = leftStudyLocusId#59310)\n         :     :- Project [targetId#47628, diseaseId#48955, CASE WHEN isnull(homogenized#49074) THEN noEvaluable ELSE homogenized#49074 END AS homogenized#49124, leftStudyLocusId#48830, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025]\n         :     :  +- Project [targetId#47628, diseaseId#48955, homogenized#49074, leftStudyLocusId#48830, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025]\n         :     :     +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 16 more fields]\n         :     :        +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 17 more fields]\n         :     :           +- Window [first(colocDoE#49033, true) windowspecdefinition(targetId#47628, diseaseId#48955, pValueExponent#48855 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS homogenized#49074], [targetId#47628, diseaseId#48955], [pValueExponent#48855 ASC NULLS FIRST]\n         :     :              +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                 +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                    +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 14 more fields]\n         :     :                       +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                          +- Generate explode(concat(array(diseaseId#47728), parents#47389)), true, [diseaseId#48955]\n         :     :                             +- Project [diseaseId#47728, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                                +- Join LeftOuter, (diseaseId#47728 = diseaseId#48909)\n         :     :                                   :- Project [leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, datasourceId#47627, ... 12 more fields]\n         :     :                                   :  +- Join RightOuter, ((leftStudyLocusId#47954 = leftStudyLocusId#48830) AND (targetId#48806 = targetId#47628))\n         :     :                                   :     :- Project [rightStudyId#48048, rightStudyLocusId#47955, leftStudyLocusId#47954, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, geneId#47895 AS targetId#48806, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923]\n         :     :                                   :     :  +- Filter NOT (rightStudyType#47957 = gwas)\n         :     :                                   :     :     +- Project [rightStudyId#48048, rightStudyLocusId#47955, leftStudyLocusId#47954, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, geneId#47895, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923]\n         :     :                                   :     :        +- Join LeftOuter, (rightStudyId#48048 = rightStudyId#48100)\n         :     :                                   :     :           :- Project [rightStudyLocusId#47955, leftStudyLocusId#47954, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightStudyId#48048, rightVariantId#48049, credibleRightStudyType#48050]\n         :     :                                   :     :           :  +- Join LeftOuter, (rightStudyLocusId#47955 = rightStudyLocusId#48047)\n         :     :                                   :     :           :     :- Project [leftStudyLocusId#47954, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025]\n         :     :                                   :     :           :     :  +- Join LeftOuter, (leftStudyLocusId#47954 = leftStudyLocusId#48022)\n         :     :                                   :     :           :     :     :- Relation [leftStudyLocusId#47954,rightStudyLocusId#47955,chromosome#47956,rightStudyType#47957,numberColocalisingVariants#47958L,h0#47959,h1#47960,h2#47961,h3#47962,h4#47963,colocalisationMethod#47964,betaRatioSignAverage#47965] parquet\n         :     :                                   :     :           :     :     +- Project [studyLocusId#47842 AS leftStudyLocusId#48022, StudyId#47843 AS leftStudyId#48023, variantId#47844 AS leftVariantId#48024, studyType#47867 AS credibleLeftStudyType#48025]\n         :     :                                   :     :           :     :        +- Relation [studyLocusId#47842,studyId#47843,variantId#47844,chromosome#47845,position#47846,region#47847,beta#47848,zScore#47849,pValueMantissa#47850,pValueExponent#47851,effectAlleleFrequencyFromSource#47852,standardError#47853,subStudyDescription#47854,qualityControls#47855,finemappingMethod#47856,credibleSetIndex#47857,credibleSetlog10BF#47858,purityMeanR2#47859,purityMinR2#47860,locusStart#47861,locusEnd#47862,sampleSize#47863,ldSet#47864,locus#47865,... 2 more fields] parquet\n         :     :                                   :     :           :     +- Project [studyLocusId#48055 AS rightStudyLocusId#48047, studyId#48056 AS rightStudyId#48048, variantId#48057 AS rightVariantId#48049, studyType#48080 AS credibleRightStudyType#48050]\n         :     :                                   :     :           :        +- Relation [studyLocusId#48055,studyId#48056,variantId#48057,chromosome#48058,position#48059,region#48060,beta#48061,zScore#48062,pValueMantissa#48063,pValueExponent#48064,effectAlleleFrequencyFromSource#48065,standardError#48066,subStudyDescription#48067,qualityControls#48068,finemappingMethod#48069,credibleSetIndex#48070,credibleSetlog10BF#48071,purityMeanR2#48072,purityMinR2#48073,locusStart#48074,locusEnd#48075,sampleSize#48076,ldSet#48077,locus#48078,... 2 more fields] parquet\n         :     :                                   :     :           +- Project [studyId#47894 AS rightStudyId#48100, geneId#47895, projectId#47896, studyType#47897 AS indexStudyType#48101, condition#47919, biosampleId#47923]\n         :     :                                   :     :              +- Relation [studyId#47894,geneId#47895,projectId#47896,studyType#47897,traitFromSource#47898,traitFromSourceMappedIds#47899,biosampleFromSourceId#47900,pubmedId#47901,publicationTitle#47902,publicationFirstAuthor#47903,publicationDate#47904,publicationJournal#47905,backgroundTraitFromSourceMappedIds#47906,initialSampleSize#47907,nCases#47908,nControls#47909,nSamples#47910,cohorts#47911,ldPopulationStructure#47912,discoverySamples#47913,replicationSamples#47914,qualityControls#47915,analysisFlags#47916,summarystatsLocation#47917,... 6 more fields] parquet\n         :     :                                   :     +- Project [studyLocusId#47706 AS leftStudyLocusId#48830, datasourceId#47627, targetId#47628, datatypeId#47657, diseaseFromSourceMappedId#47661, resourceScore#47697, targetFromSourceId#47713, diseaseId#47728, id#47729, score#47730, sourceId#47733, studyId#48847, variantId#48848, betaGwas#48785, pValueExponent#48855]\n         :     :                                   :        +- Project [studyLocusId#47706, datasourceId#47627, targetId#47628, datatypeId#47657, diseaseFromSourceMappedId#47661, resourceScore#47697, targetFromSourceId#47713, diseaseId#47728, id#47729, score#47730, sourceId#47733, studyId#48847, variantId#48848, betaGwas#48785, pValueExponent#48855]\n         :     :                                   :           +- Join LeftOuter, (studyLocusId#47706 = studyLocusId#48846)\n         :     :                                   :              :- Project [datasourceId#47627, targetId#47628, datatypeId#47657, diseaseFromSourceMappedId#47661, resourceScore#47697, studyLocusId#47706, targetFromSourceId#47713, diseaseId#47728, id#47729, score#47730, sourceId#47733]\n         :     :                                   :              :  +- Filter (datasourceId#47627 = gwas_credible_sets)\n         :     :                                   :              :     +- Filter datasourceId#47627 IN (gwas_credible_sets)\n         :     :                                   :              :        +- Relation [datasourceId#47627,targetId#47628,alleleOrigins#47629,allelicRequirements#47630,ancestry#47631,ancestryId#47632,assays#47633,assessments#47634,beta#47635,betaConfidenceIntervalLower#47636,betaConfidenceIntervalUpper#47637,biologicalModelAllelicComposition#47638,biologicalModelGeneticBackground#47639,biologicalModelId#47640,biomarkerList#47641,biomarkerName#47642,biomarkers#47643,biosamplesFromSource#47644,cellLineBackground#47645,cellType#47646,clinicalPhase#47647,clinicalSignificances#47648,clinicalStatus#47649,cohortDescription#47650,... 83 more fields] parquet\n         :     :                                   :              +- Project [studyLocusId#48846, studyId#48847, variantId#48848, beta#48852 AS betaGwas#48785, pValueExponent#48855]\n         :     :                                   :                 +- Relation [studyLocusId#48846,studyId#48847,variantId#48848,chromosome#48849,position#48850,region#48851,beta#48852,zScore#48853,pValueMantissa#48854,pValueExponent#48855,effectAlleleFrequencyFromSource#48856,standardError#48857,subStudyDescription#48858,qualityControls#48859,finemappingMethod#48860,credibleSetIndex#48861,credibleSetlog10BF#48862,purityMeanR2#48863,purityMinR2#48864,locusStart#48865,locusEnd#48866,sampleSize#48867,ldSet#48868,locus#48869,... 2 more fields] parquet\n         :     :                                   +- Project [id#47382 AS diseaseId#48909, name#47386, parents#47389, therapeuticAreas#47394]\n         :     :                                      +- Relation [id#47382,code#47383,dbXRefs#47384,description#47385,name#47386,directLocationIds#47387,obsoleteTerms#47388,parents#47389,synonyms#47390,ancestors#47391,descendants#47392,children#47393,therapeuticAreas#47394,indirectLocationIds#47395,ontology#47396] parquet\n         :     +- Project [studyLocusId#59269 AS leftStudyLocusId#59310, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301]\n         :        +- Project [studyLocusId#59269, geneId#59270, __pivot_first(value) AS `first(value)`#59299[0] AS distanceFootprintMean#59300, __pivot_first(value) AS `first(value)`#59299[1] AS distanceTssMean#59301]\n         :           +- Aggregate [studyLocusId#59269, geneId#59270], [studyLocusId#59269, geneId#59270, pivotfirst(key#59277, first(value)#59293, distanceFootprintMean, distanceTssMean, 0, 0) AS __pivot_first(value) AS `first(value)`#59299]\n         :              +- Aggregate [studyLocusId#59269, geneId#59270, key#59277], [studyLocusId#59269, geneId#59270, key#59277, first(value#59278, false) AS first(value)#59293]\n         :                 +- Filter key#59277 IN (distanceFootprintMean,distanceTssMean)\n         :                    +- Project [studyLocusId#59269, geneId#59270, key#59277, value#59278]\n         :                       +- Generate explode(locusToGeneFeatures#59272), true, [key#59277, value#59278]\n         :                          +- Relation [studyLocusId#59269,geneId#59270,score#59271,locusToGeneFeatures#59272] parquet\n         +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, coherencyDiagonal#116990 AS coherencyDiagonal_ch#117011, coherencyOneCell#117000 AS coherencyOneCell_ch#117012, LoF_protect#116947L AS LoF_protect_ch#117013L, GoF_protect#116946L AS GoF_protect_ch#117014L]\n            +- Filter ((isnotnull(GoF_protect#116946L) OR isnotnull(LoF_protect#116947L)) AND (coherencyDiagonal#116990 = coherent))\n               +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, GoF_risk#116973, LoF_risk#116981, coherencyDiagonal#116990, CASE WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnull(noEvaluable#116948L)) THEN noEvid WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnotnull(noEvaluable#116948L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#116981) OR isnotnull(LoF_protect#116947L)) OR isnotnull(GoF_risk#116973)) OR isnotnull(GoF_protect#116946L)) THEN CASE WHEN (isnotnull(LoF_risk#116981) AND ((isnull(LoF_protect#116947L) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L))) THEN coherent WHEN (isnotnull(GoF_risk#116973) AND ((isnull(LoF_protect#116947L) AND isnull(LoF_risk#116981)) AND isnull(GoF_protect#116946L))) THEN coherent WHEN (isnotnull(LoF_protect#116947L) AND ((isnull(LoF_risk#116981) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L))) THEN coherent WHEN (isnotnull(GoF_protect#116946L) AND ((isnull(LoF_protect#116947L) AND isnull(GoF_risk#116973)) AND isnull(LoF_risk#116981))) THEN coherent ELSE dispar END END AS coherencyOneCell#117000]\n                  +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, GoF_risk#116973, LoF_risk#116981, CASE WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnull(noEvaluable#116948L)) THEN noEvid WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnotnull(noEvaluable#116948L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#116981) OR isnotnull(LoF_protect#116947L)) OR isnotnull(GoF_risk#116973)) OR isnotnull(GoF_protect#116946L)) THEN CASE WHEN (isnotnull(GoF_risk#116973) AND isnotnull(LoF_risk#116981)) THEN dispar WHEN (isnotnull(LoF_protect#116947L) AND isnotnull(LoF_risk#116981)) THEN dispar WHEN (isnotnull(GoF_protect#116946L) AND isnotnull(GoF_risk#116973)) THEN dispar WHEN (isnotnull(GoF_protect#116946L) AND isnotnull(LoF_protect#116947L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#116990]\n                     +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, GoF_risk#116973, null AS LoF_risk#116981]\n                        +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, null AS GoF_risk#116973]\n                           +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, __pivot_count(targetId) AS `count(targetId)`#116945[0] AS GoF_protect#116946L, __pivot_count(targetId) AS `count(targetId)`#116945[1] AS LoF_protect#116947L, __pivot_count(targetId) AS `count(targetId)`#116945[2] AS noEvaluable#116948L]\n                              +- Aggregate [targetId#47413, diseaseId#47513, maxClinPhase#113689], [targetId#47413, diseaseId#47513, maxClinPhase#113689, pivotfirst(homogenized#113567, count(targetId)#116937L, GoF_protect, LoF_protect, noEvaluable, 0, 0) AS __pivot_count(targetId) AS `count(targetId)`#116945]\n                                 +- Aggregate [targetId#47413, diseaseId#47513, maxClinPhase#113689, homogenized#113567], [targetId#47413, diseaseId#47513, maxClinPhase#113689, homogenized#113567, count(targetId#47413) AS count(targetId)#116937L]\n                                    +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 97 more fields]\n                                       +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 98 more fields]\n                                          +- Window [max(clinicalPhase#47432) windowspecdefinition(targetId#47413, diseaseId#47513, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS maxClinPhase#113689], [targetId#47413, diseaseId#47513]\n                                             +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 96 more fields]\n                                                +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 96 more fields]\n                                                   +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 95 more fields]\n                                                      +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 95 more fields]\n                                                         +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 95 more fields]\n                                                            +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 96 more fields]\n                                                               +- Window [collect_set(intogen_function#113082, 0, 0) windowspecdefinition(targetId#47413, diseaseId#47513, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#113205], [targetId#47413, diseaseId#47513]\n                                                                  +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 94 more fields]\n                                                                     +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 94 more fields]\n                                                                        +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 93 more fields]\n                                                                           +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 92 more fields]\n                                                                              +- Join LeftOuter, ((drugId2#112013 = drugId#47450) AND (targetId2#112020 = targetId#47413))\n                                                                                 :- Join LeftOuter, (target_id#112063 = targetId#47413)\n                                                                                 :  :- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, concat_ws(,, clinicalSignificances#47433) AS clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 83 more fields]\n                                                                                 :  :  +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#47433, clinicalStatus#47434, cohortDescription#47435, ... 83 more fields]\n                                                                                 :  :     +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, cast(beta#47420 as double) AS beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#47433, clinicalStatus#47434, cohortDescription#47435, ... 83 more fields]\n                                                                                 :  :        +- Filter datasourceId#47412 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                 :  :           +- Filter (datasourceId#47412 = chembl)\n                                                                                 :  :              +- Filter datasourceId#47412 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                 :  :                 +- Relation [datasourceId#47412,targetId#47413,alleleOrigins#47414,allelicRequirements#47415,ancestry#47416,ancestryId#47417,assays#47418,assessments#47419,beta#47420,betaConfidenceIntervalLower#47421,betaConfidenceIntervalUpper#47422,biologicalModelAllelicComposition#47423,biologicalModelGeneticBackground#47424,biologicalModelId#47425,biomarkerList#47426,biomarkerName#47427,biomarkers#47428,biosamplesFromSource#47429,cellLineBackground#47430,cellType#47431,clinicalPhase#47432,clinicalSignificances#47433,clinicalStatus#47434,cohortDescription#47435,... 83 more fields] parquet\n                                                                                 :  +- Project [id#111942 AS target_id#112063, approvedSymbol#111943, description#112048, description_splited#112052, TSorOncogene#112057]\n                                                                                 :     +- Project [id#111942, approvedSymbol#111943, description#112048, description_splited#112052, CASE WHEN (RLIKE(description_splited#112052, ncogene) AND RLIKE(description_splited#112052, TSG)) THEN bivalent WHEN RLIKE(description_splited#112052, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#112052, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#112057]\n                                                                                 :        +- Project [id#111942, approvedSymbol#111943, description#112048, concat_ws(,, description#112048) AS description_splited#112052]\n                                                                                 :           +- Aggregate [id#111942, approvedSymbol#111943], [id#111942, approvedSymbol#111943, collect_set(description#112040, 0, 0) AS description#112048]\n                                                                                 :              +- Filter description#112040 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                                                                 :                 +- Project [id#111942, approvedSymbol#111943, col#112035.description AS description#112040]\n                                                                                 :                    +- Project [id#111942, approvedSymbol#111943, col#112035]\n                                                                                 :                       +- Generate explode(hallmarks#111952.attributes), true, [col#112035]\n                                                                                 :                          +- Relation [id#111942,approvedSymbol#111943,biotype#111944,transcriptIds#111945,canonicalTranscript#111946,canonicalExons#111947,genomicLocation#111948,alternativeGenes#111949,approvedName#111950,go#111951,hallmarks#111952,synonyms#111953,symbolSynonyms#111954,nameSynonyms#111955,functionDescriptions#111956,subcellularLocations#111957,targetClass#111958,obsoleteSymbols#111959,obsoleteNames#111960,constraint#111961,tep#111962,proteinIds#111963,dbXrefs#111964,chemicalProbes#111965,... 4 more fields] parquet\n                                                                                 +- Aggregate [targetId2#112020, drugId2#112013], [targetId2#112020, drugId2#112013, collect_set(actionType#111998, 0, 0) AS actionType#112030]\n                                                                                    +- Project [targetId2#112020, drugId2#112013, actionType#111998, mechanismOfAction#111999]\n                                                                                       +- Generate explode(targets#112003), true, [targetId2#112020]\n                                                                                          +- Project [drugId2#112013, actionType#111998, mechanismOfAction#111999, targets#112003]\n                                                                                             +- Generate explode(chemblIds#112000), true, [drugId2#112013]\n                                                                                                +- Relation [actionType#111998,mechanismOfAction#111999,chemblIds#112000,targetName#112001,targetType#112002,targets#112003,references#112004] parquet\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAnalysisException\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_8478/4243798011.py\u001b[0m in \u001b[0;36m<cell line: 32>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     36\u001b[0m         \u001b[0;34m\"geneticEvidence\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwhen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"resourceScore\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misNotNull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"yes\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m     )\n\u001b[0;32m---> 38\u001b[0;31m     .withColumn(\n\u001b[0m\u001b[1;32m     39\u001b[0m         \u001b[0;34m\"diagonalAgreeWithDrugs\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m         F.when(\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/dataframe.py\u001b[0m in \u001b[0;36mwithColumn\u001b[0;34m(self, colName, col)\u001b[0m\n\u001b[1;32m   3034\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mColumn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3035\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"col should be Column\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3036\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwithColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolName\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparkSession\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3037\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3038\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwithColumnRenamed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexisting\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m\"DataFrame\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1319\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1321\u001b[0;31m         return_value = get_return_value(\n\u001b[0m\u001b[1;32m   1322\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[1;32m   1323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/utils.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    194\u001b[0m                 \u001b[0;31m# Hide where the exception came from that shows a non-Pythonic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m                 \u001b[0;31m# JVM exception message.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mconverted\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAnalysisException\u001b[0m: Column 'GoF_risk' does not exist. Did you mean one of the following? [geneId, h4, targetId, GoF_protect_ch, diseaseId, homogenized, LoF_protect_ch, maxClinPhase, datasourceId, leftVariantId, resourceScore, distanceTssMean, geneticEvidence, leftStudyLocusId, coherencyDiagonal_ch, coherencyOneCell_ch, distanceFootprintMean, credibleLeftStudyType];\n'Project [targetId#47413, diseaseId#47513, leftStudyLocusId#48830, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301, maxClinPhase#113689, coherencyDiagonal_ch#117011, coherencyOneCell_ch#117012, LoF_protect_ch#117013L, GoF_protect_ch#117014L, geneticEvidence#117039, CASE WHEN (coherencyDiagonal_ch#117011 = coherent) THEN CASE WHEN (isnotnull(LoF_protect_ch#117013L) AND (isnotnull('GoF_risk) OR isnotnull('LoF_protect))) THEN coherent WHEN (isnotnull(GoF_protect_ch#117014L) AND (isnotnull('LoF_risk) OR isnotnull('GoF_protect))) THEN coherent ELSE dispar END END AS diagonalAgreeWithDrugs#117058]\n+- Project [targetId#47413, diseaseId#47513, leftStudyLocusId#48830, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301, maxClinPhase#113689, coherencyDiagonal_ch#117011, coherencyOneCell_ch#117012, LoF_protect_ch#117013L, GoF_protect_ch#117014L, CASE WHEN isnotnull(resourceScore#47697) THEN yes END AS geneticEvidence#117039]\n   +- Project [targetId#47413, diseaseId#47513, leftStudyLocusId#48830, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301, maxClinPhase#113689, coherencyDiagonal_ch#117011, coherencyOneCell_ch#117012, LoF_protect_ch#117013L, GoF_protect_ch#117014L]\n      +- Join RightOuter, ((targetId#47628 = targetId#47413) AND (diseaseId#48955 = diseaseId#47513))\n         :- Project [leftStudyLocusId#48830, targetId#47628, diseaseId#48955, homogenized#49124, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301]\n         :  +- Join LeftOuter, (leftStudyLocusId#48830 = leftStudyLocusId#59310)\n         :     :- Project [targetId#47628, diseaseId#48955, CASE WHEN isnull(homogenized#49074) THEN noEvaluable ELSE homogenized#49074 END AS homogenized#49124, leftStudyLocusId#48830, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025]\n         :     :  +- Project [targetId#47628, diseaseId#48955, homogenized#49074, leftStudyLocusId#48830, h4#47963, datasourceId#47627, resourceScore#47697, leftVariantId#48024, credibleLeftStudyType#48025]\n         :     :     +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 16 more fields]\n         :     :        +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 17 more fields]\n         :     :           +- Window [first(colocDoE#49033, true) windowspecdefinition(targetId#47628, diseaseId#48955, pValueExponent#48855 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS homogenized#49074], [targetId#47628, diseaseId#48955], [pValueExponent#48855 ASC NULLS FIRST]\n         :     :              +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                 +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                    +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 14 more fields]\n         :     :                       +- Project [diseaseId#48955, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                          +- Generate explode(concat(array(diseaseId#47728), parents#47389)), true, [diseaseId#48955]\n         :     :                             +- Project [diseaseId#47728, leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, ... 15 more fields]\n         :     :                                +- Join LeftOuter, (diseaseId#47728 = diseaseId#48909)\n         :     :                                   :- Project [leftStudyLocusId#48830, targetId#47628, rightStudyId#48048, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923, datasourceId#47627, ... 12 more fields]\n         :     :                                   :  +- Join RightOuter, ((leftStudyLocusId#47954 = leftStudyLocusId#48830) AND (targetId#48806 = targetId#47628))\n         :     :                                   :     :- Project [rightStudyId#48048, rightStudyLocusId#47955, leftStudyLocusId#47954, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, geneId#47895 AS targetId#48806, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923]\n         :     :                                   :     :  +- Filter NOT (rightStudyType#47957 = gwas)\n         :     :                                   :     :     +- Project [rightStudyId#48048, rightStudyLocusId#47955, leftStudyLocusId#47954, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightVariantId#48049, credibleRightStudyType#48050, geneId#47895, projectId#47896, indexStudyType#48101, condition#47919, biosampleId#47923]\n         :     :                                   :     :        +- Join LeftOuter, (rightStudyId#48048 = rightStudyId#48100)\n         :     :                                   :     :           :- Project [rightStudyLocusId#47955, leftStudyLocusId#47954, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025, rightStudyId#48048, rightVariantId#48049, credibleRightStudyType#48050]\n         :     :                                   :     :           :  +- Join LeftOuter, (rightStudyLocusId#47955 = rightStudyLocusId#48047)\n         :     :                                   :     :           :     :- Project [leftStudyLocusId#47954, rightStudyLocusId#47955, chromosome#47956, rightStudyType#47957, numberColocalisingVariants#47958L, h0#47959, h1#47960, h2#47961, h3#47962, h4#47963, colocalisationMethod#47964, betaRatioSignAverage#47965, leftStudyId#48023, leftVariantId#48024, credibleLeftStudyType#48025]\n         :     :                                   :     :           :     :  +- Join LeftOuter, (leftStudyLocusId#47954 = leftStudyLocusId#48022)\n         :     :                                   :     :           :     :     :- Relation [leftStudyLocusId#47954,rightStudyLocusId#47955,chromosome#47956,rightStudyType#47957,numberColocalisingVariants#47958L,h0#47959,h1#47960,h2#47961,h3#47962,h4#47963,colocalisationMethod#47964,betaRatioSignAverage#47965] parquet\n         :     :                                   :     :           :     :     +- Project [studyLocusId#47842 AS leftStudyLocusId#48022, StudyId#47843 AS leftStudyId#48023, variantId#47844 AS leftVariantId#48024, studyType#47867 AS credibleLeftStudyType#48025]\n         :     :                                   :     :           :     :        +- Relation [studyLocusId#47842,studyId#47843,variantId#47844,chromosome#47845,position#47846,region#47847,beta#47848,zScore#47849,pValueMantissa#47850,pValueExponent#47851,effectAlleleFrequencyFromSource#47852,standardError#47853,subStudyDescription#47854,qualityControls#47855,finemappingMethod#47856,credibleSetIndex#47857,credibleSetlog10BF#47858,purityMeanR2#47859,purityMinR2#47860,locusStart#47861,locusEnd#47862,sampleSize#47863,ldSet#47864,locus#47865,... 2 more fields] parquet\n         :     :                                   :     :           :     +- Project [studyLocusId#48055 AS rightStudyLocusId#48047, studyId#48056 AS rightStudyId#48048, variantId#48057 AS rightVariantId#48049, studyType#48080 AS credibleRightStudyType#48050]\n         :     :                                   :     :           :        +- Relation [studyLocusId#48055,studyId#48056,variantId#48057,chromosome#48058,position#48059,region#48060,beta#48061,zScore#48062,pValueMantissa#48063,pValueExponent#48064,effectAlleleFrequencyFromSource#48065,standardError#48066,subStudyDescription#48067,qualityControls#48068,finemappingMethod#48069,credibleSetIndex#48070,credibleSetlog10BF#48071,purityMeanR2#48072,purityMinR2#48073,locusStart#48074,locusEnd#48075,sampleSize#48076,ldSet#48077,locus#48078,... 2 more fields] parquet\n         :     :                                   :     :           +- Project [studyId#47894 AS rightStudyId#48100, geneId#47895, projectId#47896, studyType#47897 AS indexStudyType#48101, condition#47919, biosampleId#47923]\n         :     :                                   :     :              +- Relation [studyId#47894,geneId#47895,projectId#47896,studyType#47897,traitFromSource#47898,traitFromSourceMappedIds#47899,biosampleFromSourceId#47900,pubmedId#47901,publicationTitle#47902,publicationFirstAuthor#47903,publicationDate#47904,publicationJournal#47905,backgroundTraitFromSourceMappedIds#47906,initialSampleSize#47907,nCases#47908,nControls#47909,nSamples#47910,cohorts#47911,ldPopulationStructure#47912,discoverySamples#47913,replicationSamples#47914,qualityControls#47915,analysisFlags#47916,summarystatsLocation#47917,... 6 more fields] parquet\n         :     :                                   :     +- Project [studyLocusId#47706 AS leftStudyLocusId#48830, datasourceId#47627, targetId#47628, datatypeId#47657, diseaseFromSourceMappedId#47661, resourceScore#47697, targetFromSourceId#47713, diseaseId#47728, id#47729, score#47730, sourceId#47733, studyId#48847, variantId#48848, betaGwas#48785, pValueExponent#48855]\n         :     :                                   :        +- Project [studyLocusId#47706, datasourceId#47627, targetId#47628, datatypeId#47657, diseaseFromSourceMappedId#47661, resourceScore#47697, targetFromSourceId#47713, diseaseId#47728, id#47729, score#47730, sourceId#47733, studyId#48847, variantId#48848, betaGwas#48785, pValueExponent#48855]\n         :     :                                   :           +- Join LeftOuter, (studyLocusId#47706 = studyLocusId#48846)\n         :     :                                   :              :- Project [datasourceId#47627, targetId#47628, datatypeId#47657, diseaseFromSourceMappedId#47661, resourceScore#47697, studyLocusId#47706, targetFromSourceId#47713, diseaseId#47728, id#47729, score#47730, sourceId#47733]\n         :     :                                   :              :  +- Filter (datasourceId#47627 = gwas_credible_sets)\n         :     :                                   :              :     +- Filter datasourceId#47627 IN (gwas_credible_sets)\n         :     :                                   :              :        +- Relation [datasourceId#47627,targetId#47628,alleleOrigins#47629,allelicRequirements#47630,ancestry#47631,ancestryId#47632,assays#47633,assessments#47634,beta#47635,betaConfidenceIntervalLower#47636,betaConfidenceIntervalUpper#47637,biologicalModelAllelicComposition#47638,biologicalModelGeneticBackground#47639,biologicalModelId#47640,biomarkerList#47641,biomarkerName#47642,biomarkers#47643,biosamplesFromSource#47644,cellLineBackground#47645,cellType#47646,clinicalPhase#47647,clinicalSignificances#47648,clinicalStatus#47649,cohortDescription#47650,... 83 more fields] parquet\n         :     :                                   :              +- Project [studyLocusId#48846, studyId#48847, variantId#48848, beta#48852 AS betaGwas#48785, pValueExponent#48855]\n         :     :                                   :                 +- Relation [studyLocusId#48846,studyId#48847,variantId#48848,chromosome#48849,position#48850,region#48851,beta#48852,zScore#48853,pValueMantissa#48854,pValueExponent#48855,effectAlleleFrequencyFromSource#48856,standardError#48857,subStudyDescription#48858,qualityControls#48859,finemappingMethod#48860,credibleSetIndex#48861,credibleSetlog10BF#48862,purityMeanR2#48863,purityMinR2#48864,locusStart#48865,locusEnd#48866,sampleSize#48867,ldSet#48868,locus#48869,... 2 more fields] parquet\n         :     :                                   +- Project [id#47382 AS diseaseId#48909, name#47386, parents#47389, therapeuticAreas#47394]\n         :     :                                      +- Relation [id#47382,code#47383,dbXRefs#47384,description#47385,name#47386,directLocationIds#47387,obsoleteTerms#47388,parents#47389,synonyms#47390,ancestors#47391,descendants#47392,children#47393,therapeuticAreas#47394,indirectLocationIds#47395,ontology#47396] parquet\n         :     +- Project [studyLocusId#59269 AS leftStudyLocusId#59310, geneId#59270, distanceFootprintMean#59300, distanceTssMean#59301]\n         :        +- Project [studyLocusId#59269, geneId#59270, __pivot_first(value) AS `first(value)`#59299[0] AS distanceFootprintMean#59300, __pivot_first(value) AS `first(value)`#59299[1] AS distanceTssMean#59301]\n         :           +- Aggregate [studyLocusId#59269, geneId#59270], [studyLocusId#59269, geneId#59270, pivotfirst(key#59277, first(value)#59293, distanceFootprintMean, distanceTssMean, 0, 0) AS __pivot_first(value) AS `first(value)`#59299]\n         :              +- Aggregate [studyLocusId#59269, geneId#59270, key#59277], [studyLocusId#59269, geneId#59270, key#59277, first(value#59278, false) AS first(value)#59293]\n         :                 +- Filter key#59277 IN (distanceFootprintMean,distanceTssMean)\n         :                    +- Project [studyLocusId#59269, geneId#59270, key#59277, value#59278]\n         :                       +- Generate explode(locusToGeneFeatures#59272), true, [key#59277, value#59278]\n         :                          +- Relation [studyLocusId#59269,geneId#59270,score#59271,locusToGeneFeatures#59272] parquet\n         +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, coherencyDiagonal#116990 AS coherencyDiagonal_ch#117011, coherencyOneCell#117000 AS coherencyOneCell_ch#117012, LoF_protect#116947L AS LoF_protect_ch#117013L, GoF_protect#116946L AS GoF_protect_ch#117014L]\n            +- Filter ((isnotnull(GoF_protect#116946L) OR isnotnull(LoF_protect#116947L)) AND (coherencyDiagonal#116990 = coherent))\n               +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, GoF_risk#116973, LoF_risk#116981, coherencyDiagonal#116990, CASE WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnull(noEvaluable#116948L)) THEN noEvid WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnotnull(noEvaluable#116948L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#116981) OR isnotnull(LoF_protect#116947L)) OR isnotnull(GoF_risk#116973)) OR isnotnull(GoF_protect#116946L)) THEN CASE WHEN (isnotnull(LoF_risk#116981) AND ((isnull(LoF_protect#116947L) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L))) THEN coherent WHEN (isnotnull(GoF_risk#116973) AND ((isnull(LoF_protect#116947L) AND isnull(LoF_risk#116981)) AND isnull(GoF_protect#116946L))) THEN coherent WHEN (isnotnull(LoF_protect#116947L) AND ((isnull(LoF_risk#116981) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L))) THEN coherent WHEN (isnotnull(GoF_protect#116946L) AND ((isnull(LoF_protect#116947L) AND isnull(GoF_risk#116973)) AND isnull(LoF_risk#116981))) THEN coherent ELSE dispar END END AS coherencyOneCell#117000]\n                  +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, GoF_risk#116973, LoF_risk#116981, CASE WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnull(noEvaluable#116948L)) THEN noEvid WHEN ((((isnull(LoF_risk#116981) AND isnull(LoF_protect#116947L)) AND isnull(GoF_risk#116973)) AND isnull(GoF_protect#116946L)) AND isnotnull(noEvaluable#116948L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#116981) OR isnotnull(LoF_protect#116947L)) OR isnotnull(GoF_risk#116973)) OR isnotnull(GoF_protect#116946L)) THEN CASE WHEN (isnotnull(GoF_risk#116973) AND isnotnull(LoF_risk#116981)) THEN dispar WHEN (isnotnull(LoF_protect#116947L) AND isnotnull(LoF_risk#116981)) THEN dispar WHEN (isnotnull(GoF_protect#116946L) AND isnotnull(GoF_risk#116973)) THEN dispar WHEN (isnotnull(GoF_protect#116946L) AND isnotnull(LoF_protect#116947L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#116990]\n                     +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, GoF_risk#116973, null AS LoF_risk#116981]\n                        +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, GoF_protect#116946L, LoF_protect#116947L, noEvaluable#116948L, null AS GoF_risk#116973]\n                           +- Project [targetId#47413, diseaseId#47513, maxClinPhase#113689, __pivot_count(targetId) AS `count(targetId)`#116945[0] AS GoF_protect#116946L, __pivot_count(targetId) AS `count(targetId)`#116945[1] AS LoF_protect#116947L, __pivot_count(targetId) AS `count(targetId)`#116945[2] AS noEvaluable#116948L]\n                              +- Aggregate [targetId#47413, diseaseId#47513, maxClinPhase#113689], [targetId#47413, diseaseId#47513, maxClinPhase#113689, pivotfirst(homogenized#113567, count(targetId)#116937L, GoF_protect, LoF_protect, noEvaluable, 0, 0) AS __pivot_count(targetId) AS `count(targetId)`#116945]\n                                 +- Aggregate [targetId#47413, diseaseId#47513, maxClinPhase#113689, homogenized#113567], [targetId#47413, diseaseId#47513, maxClinPhase#113689, homogenized#113567, count(targetId#47413) AS count(targetId)#116937L]\n                                    +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 97 more fields]\n                                       +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 98 more fields]\n                                          +- Window [max(clinicalPhase#47432) windowspecdefinition(targetId#47413, diseaseId#47513, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS maxClinPhase#113689], [targetId#47413, diseaseId#47513]\n                                             +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 96 more fields]\n                                                +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 96 more fields]\n                                                   +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 95 more fields]\n                                                      +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 95 more fields]\n                                                         +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 95 more fields]\n                                                            +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 96 more fields]\n                                                               +- Window [collect_set(intogen_function#113082, 0, 0) windowspecdefinition(targetId#47413, diseaseId#47513, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#113205], [targetId#47413, diseaseId#47513]\n                                                                  +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 94 more fields]\n                                                                     +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 94 more fields]\n                                                                        +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 93 more fields]\n                                                                           +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 92 more fields]\n                                                                              +- Join LeftOuter, ((drugId2#112013 = drugId#47450) AND (targetId2#112020 = targetId#47413))\n                                                                                 :- Join LeftOuter, (target_id#112063 = targetId#47413)\n                                                                                 :  :- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, concat_ws(,, clinicalSignificances#47433) AS clinicalSignificances#112285, clinicalStatus#47434, cohortDescription#47435, ... 83 more fields]\n                                                                                 :  :  +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#47433, clinicalStatus#47434, cohortDescription#47435, ... 83 more fields]\n                                                                                 :  :     +- Project [datasourceId#47412, targetId#47413, alleleOrigins#47414, allelicRequirements#47415, ancestry#47416, ancestryId#47417, assays#47418, assessments#47419, cast(beta#47420 as double) AS beta#112069, betaConfidenceIntervalLower#47421, betaConfidenceIntervalUpper#47422, biologicalModelAllelicComposition#47423, biologicalModelGeneticBackground#47424, biologicalModelId#47425, biomarkerList#47426, biomarkerName#47427, biomarkers#47428, biosamplesFromSource#47429, cellLineBackground#47430, cellType#47431, clinicalPhase#47432, clinicalSignificances#47433, clinicalStatus#47434, cohortDescription#47435, ... 83 more fields]\n                                                                                 :  :        +- Filter datasourceId#47412 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                 :  :           +- Filter (datasourceId#47412 = chembl)\n                                                                                 :  :              +- Filter datasourceId#47412 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                 :  :                 +- Relation [datasourceId#47412,targetId#47413,alleleOrigins#47414,allelicRequirements#47415,ancestry#47416,ancestryId#47417,assays#47418,assessments#47419,beta#47420,betaConfidenceIntervalLower#47421,betaConfidenceIntervalUpper#47422,biologicalModelAllelicComposition#47423,biologicalModelGeneticBackground#47424,biologicalModelId#47425,biomarkerList#47426,biomarkerName#47427,biomarkers#47428,biosamplesFromSource#47429,cellLineBackground#47430,cellType#47431,clinicalPhase#47432,clinicalSignificances#47433,clinicalStatus#47434,cohortDescription#47435,... 83 more fields] parquet\n                                                                                 :  +- Project [id#111942 AS target_id#112063, approvedSymbol#111943, description#112048, description_splited#112052, TSorOncogene#112057]\n                                                                                 :     +- Project [id#111942, approvedSymbol#111943, description#112048, description_splited#112052, CASE WHEN (RLIKE(description_splited#112052, ncogene) AND RLIKE(description_splited#112052, TSG)) THEN bivalent WHEN RLIKE(description_splited#112052, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#112052, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#112057]\n                                                                                 :        +- Project [id#111942, approvedSymbol#111943, description#112048, concat_ws(,, description#112048) AS description_splited#112052]\n                                                                                 :           +- Aggregate [id#111942, approvedSymbol#111943], [id#111942, approvedSymbol#111943, collect_set(description#112040, 0, 0) AS description#112048]\n                                                                                 :              +- Filter description#112040 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                                                                 :                 +- Project [id#111942, approvedSymbol#111943, col#112035.description AS description#112040]\n                                                                                 :                    +- Project [id#111942, approvedSymbol#111943, col#112035]\n                                                                                 :                       +- Generate explode(hallmarks#111952.attributes), true, [col#112035]\n                                                                                 :                          +- Relation [id#111942,approvedSymbol#111943,biotype#111944,transcriptIds#111945,canonicalTranscript#111946,canonicalExons#111947,genomicLocation#111948,alternativeGenes#111949,approvedName#111950,go#111951,hallmarks#111952,synonyms#111953,symbolSynonyms#111954,nameSynonyms#111955,functionDescriptions#111956,subcellularLocations#111957,targetClass#111958,obsoleteSymbols#111959,obsoleteNames#111960,constraint#111961,tep#111962,proteinIds#111963,dbXrefs#111964,chemicalProbes#111965,... 4 more fields] parquet\n                                                                                 +- Aggregate [targetId2#112020, drugId2#112013], [targetId2#112020, drugId2#112013, collect_set(actionType#111998, 0, 0) AS actionType#112030]\n                                                                                    +- Project [targetId2#112020, drugId2#112013, actionType#111998, mechanismOfAction#111999]\n                                                                                       +- Generate explode(targets#112003), true, [targetId2#112020]\n                                                                                          +- Project [drugId2#112013, actionType#111998, mechanismOfAction#111999, targets#112003]\n                                                                                             +- Generate explode(chemblIds#112000), true, [drugId2#112013]\n                                                                                                +- Relation [actionType#111998,mechanismOfAction#111999,chemblIds#112000,targetName#112001,targetType#112002,targets#112003,references#112004] parquet\n"
     ]
    }
   ],
   "source": [
    "analysis_chembl = (\n",
    "    discrepancifier(\n",
    "        directionOfEffect(\n",
    "            evidences.filter((F.col(\"datasourceId\") == \"chembl\")), \"24.09\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "    .filter(  ### ensure drug has annotated MoA and is coherent per Target-Disease\n",
    "        ((F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull())\n",
    "        & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    )\n",
    "    .selectExpr(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"maxClinPhase\",\n",
    "        \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "        \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "        \"LoF_protect as LoF_protect_ch\",\n",
    "        \"GoF_protect as GoF_protect_ch\",\n",
    "    )\n",
    ")\n",
    "drugsCredibleDistance = (\n",
    "    gwasCredibleAssocDistances.join(\n",
    "        analysis_chembl, on=[\"targetId\", \"diseaseId\"], how=\"right\"\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"geneticEvidence\", F.when(F.col(\"resourceScore\").isNotNull(), F.lit(\"yes\"))\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"diagonalAgreeWithDrugs\",\n",
    "        F.when(\n",
    "            (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "            # & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "            ,\n",
    "            F.when(\n",
    "                (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                & (F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                & (F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"dispar\")),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"oneCellAgreeWithDrugs\",\n",
    "        F.when(\n",
    "            (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "            # & (F.col(\"coherencyOneCell\") == \"coherent\")\n",
    "            ,\n",
    "            F.when(\n",
    "                (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                & (\n",
    "                    (F.col(\"LoF_protect\").isNotNull())\n",
    "                    & (F.col(\"LoF_risk\").isNull())\n",
    "                    & (F.col(\"GoF_protect\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                & (\n",
    "                    (F.col(\"GoF_protect\").isNotNull())\n",
    "                    & (F.col(\"LoF_risk\").isNull())\n",
    "                    & (F.col(\"LoF_protect\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"dispar\")),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase4\",\n",
    "        F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=3\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=2\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"Phase>=1\",\n",
    "        F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Join the gwasCredibleAssocDistances with the drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+----------------+-----------+----+------------+-------------+-------------+---------------------+------+---------------------+---------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|leftStudyLocusId|homogenized|  h4|datasourceId|resourceScore|leftVariantId|credibleLeftStudyType|geneId|distanceFootprintMean|distanceTssMean|maxClinPhase|GoF_protect|LoF_protect|noEvaluable|GoF_risk|LoF_risk|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+----------------+-----------+----+------------+-------------+-------------+---------------------+------+---------------------+---------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+\n",
      "|ENSG00000007314|  EFO_0000555|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|          3|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0003102|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         0.5|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0003894|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         4.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0004699|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|       null|          3|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0801084|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_1000249|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000008018|  EFO_1000453|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000010310|  EFO_0003884|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|          1|       null|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000012504|  EFO_0004210|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         4.0|          1|       null|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000012504|MONDO_0019052|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         4.0|          1|       null|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000012779|MONDO_0004235|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|       null|          4|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000014138|  EFO_1001945|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|       null|          4|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000023228|  EFO_1000657|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|          3|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000023445|  EFO_0000616|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         1.0|       null|          2|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000025708|  EFO_0000702|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|          2|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000037280|  EFO_0003897|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|         11|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000058091|  EFO_0000574|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000062822|  EFO_0004991|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         2.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000062822|  EFO_0005537|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|       null|          7|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000062822|  EFO_0005952|            null|       null|null|        null|         null|         null|                 null|  null|                 null|           null|         3.0|       null|         37|       null|    null|    null|         coherent|        coherent|\n",
      "+---------------+-------------+----------------+-----------+----+------------+-------------+-------------+---------------------+------+---------------------+---------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "drugsCredibleDistance.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------------+-----------+------------------+------------------+-------------------+--------------+---------------------+---------------+---------------------+---------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+-----------------------+-----------------------------+------------------+\n",
      "|       targetId|  diseaseId|    leftStudyLocusId|homogenized|                h4|      datasourceId|      resourceScore| leftVariantId|credibleLeftStudyType|         geneId|distanceFootprintMean|distanceTssMean|maxClinPhase|GoF_protect|LoF_protect|noEvaluable|GoF_risk|LoF_risk|coherencyDiagonal|coherencyOneCell|min_tssDistance_ranking|min_footprintDistance_ranking|      max_L2GScore|\n",
      "+---------------+-----------+--------------------+-----------+------------------+------------------+-------------------+--------------+---------------------+---------------+---------------------+---------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+-----------------------+-----------------------------+------------------+\n",
      "|ENSG00000141510|EFO_0004251|c3ea50415a067f75b...|noEvaluable|              null|gwas_credible_sets|0.26301918037582356|          null|                 null|ENSG00000161955|           0.99763584|     0.99749476|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|c3ea50415a067f75b...|noEvaluable|              null|gwas_credible_sets|0.26301918037582356|          null|                 null|ENSG00000141510|           0.97872424|      0.9748645|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999595818679|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999595818679|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936680837|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936680837|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9999999936689647|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9980984347524174|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000141510|             0.999998|      0.9970369|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "|ENSG00000141510|EFO_0004251|917f8f710eb850396...|   GoF_risk|0.9980984347524174|gwas_credible_sets|0.25636546962760653|17_7668434_T_G|                 gwas|ENSG00000129244|            0.9983551|     0.99728173|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|              0.9748645|                   0.97872424|0.2652722291957128|\n",
      "+---------------+-----------+--------------------+-----------+------------------+------------------+-------------------+--------------+---------------------+---------------+---------------------+---------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+-----------------------+-----------------------------+------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "drugsCredibleDistance.withColumn(\n",
    "    \"min_tssDistance_ranking\",\n",
    "    F.min(\n",
    "        \"distanceTssMean\",\n",
    "    ).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    ").withColumn(\n",
    "    \"min_footprintDistance_ranking\",\n",
    "    F.min(\n",
    "        \"distanceFootprintMean\",\n",
    "    ).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    ").withColumn(  ### take maximum L2G score per T-D\n",
    "    \"max_L2GScore\",\n",
    "    F.max(\n",
    "        \"resourceScore\",\n",
    "    ).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    ").filter(\n",
    "    F.col(\"homogenized\").isNotNull()\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_chembl.filter(\n",
    "    (F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull()\n",
    ").filter(F.col())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_chembl.filter(\n",
    "    (F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull()\n",
    ").groupBy(\"coherencyDiagonal\").count().grupByshow()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|maxClinPhase|GoF_protect|LoF_protect|noEvaluable|GoF_risk|LoF_risk|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+\n",
      "|ENSG00000007314|  EFO_0000555|         2.0|       null|          3|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0003102|         0.5|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0003894|         4.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0004699|         3.0|       null|          3|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_0801084|         2.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000007314|  EFO_1000249|         3.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000008018|  EFO_1000453|         2.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000010310|  EFO_0003884|         2.0|          1|       null|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000012504|  EFO_0004210|         4.0|          1|       null|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000012504|MONDO_0019052|         4.0|          1|       null|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000012779|MONDO_0004235|         3.0|       null|          4|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000014138|  EFO_1001945|         3.0|       null|          4|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000023228|  EFO_1000657|         2.0|       null|          3|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000023445|  EFO_0000616|         1.0|       null|          2|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000025708|  EFO_0000702|         2.0|       null|          2|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000037280|  EFO_0003897|         2.0|       null|         11|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000058091|  EFO_0000574|         1.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000062822|  EFO_0004991|         2.0|       null|          1|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000062822|  EFO_0005537|         3.0|       null|          7|       null|    null|    null|         coherent|        coherent|\n",
      "|ENSG00000062822|  EFO_0005952|         3.0|       null|         37|       null|    null|    null|         coherent|        coherent|\n",
      "+---------------+-------------+------------+-----------+-----------+-----------+--------+--------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "analysis_chembl.filter(\n",
    "    (F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull()\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1026:>                                                       (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------+---------------------+---------------+\n",
      "|    leftStudyLocusId|       targetId|  diseaseId|homogenized|                h4|      datasourceId|      resourceScore|  leftVariantId|credibleLeftStudyType|         geneId|distanceFootprintMean|distanceTssMean|\n",
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------+---------------------+---------------+\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|            0.9999886|      0.9997445|\n",
      "|b9e47164bf645be64...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99987966|      0.9996356|\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004611|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004611|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|            0.9999886|      0.9997445|\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|            0.9999886|      0.9997445|\n",
      "|b9e47164bf645be64...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99987966|      0.9996356|\n",
      "|a84c1695b952c7f7a...|ENSG00000000419|EFO_0000313|noEvaluable|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|ENSG00000101126|           0.99596417|      0.9936219|\n",
      "|a84c1695b952c7f7a...|ENSG00000000419|EFO_0000313|noEvaluable|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|ENSG00000000419|             0.993824|      0.9904567|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000026559|             0.989532|       0.989532|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000000419|            0.9773113|      0.9773113|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000124217|            0.9784112|        0.97737|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000101126|            0.9715362|      0.9715362|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|ENSG00000000419|            0.9543429|      0.9543429|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|ENSG00000124217|           0.95582587|      0.9544222|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|ENSG00000026559|           0.97043514|     0.97043514|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|ENSG00000075945|           0.98231506|     0.98231506|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|ENSG00000000457|             0.939027|       0.939027|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|ENSG00000203740|            0.9954075|      0.9931712|\n",
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------+---------------------+---------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssocDistances.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 994:>                                                        (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------+---------------------+---------------+\n",
      "|    leftStudyLocusId|       targetId|  diseaseId|homogenized|                h4|      datasourceId|      resourceScore|  leftVariantId|credibleLeftStudyType|         geneId|distanceFootprintMean|distanceTssMean|\n",
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------+---------------------+---------------+\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|            0.9999886|      0.9997445|\n",
      "|b9e47164bf645be64...|ENSG00000000003|EFO_0004529|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99987966|      0.9996356|\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004611|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004611|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|            0.9999886|      0.9997445|\n",
      "|9b68d48251bfb71e9...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99999464|     0.99975055|\n",
      "|cb3db2374fb1fbeda...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|            0.9999886|      0.9997445|\n",
      "|b9e47164bf645be64...|ENSG00000000003|EFO_0004732|noEvaluable|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|ENSG00000000003|           0.99987966|      0.9996356|\n",
      "|a84c1695b952c7f7a...|ENSG00000000419|EFO_0000313|noEvaluable|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|ENSG00000000419|             0.993824|      0.9904567|\n",
      "|a84c1695b952c7f7a...|ENSG00000000419|EFO_0000313|noEvaluable|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|ENSG00000101126|           0.99596417|      0.9936219|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000026559|             0.989532|       0.989532|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000000419|            0.9773113|      0.9773113|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000124217|            0.9784112|        0.97737|\n",
      "|c5c09f45e16bc37f9...|ENSG00000000419|EFO_0004339|noEvaluable|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|ENSG00000101126|            0.9715362|      0.9715362|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|ENSG00000000419|            0.9543429|      0.9543429|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|ENSG00000124217|           0.95582587|      0.9544222|\n",
      "|f2140bdfde8f977e4...|ENSG00000000419|EFO_0004503|noEvaluable|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|ENSG00000026559|           0.97043514|     0.97043514|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|ENSG00000075945|           0.98231506|     0.98231506|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|ENSG00000000457|             0.939027|       0.939027|\n",
      "|2abe5b5bb1ca56812...|ENSG00000000457|EFO_0003872|GoF_protect|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|ENSG00000203740|            0.9954075|      0.9931712|\n",
      "+--------------------+---------------+-----------+-----------+------------------+------------------+-------------------+---------------+---------------------+---------------+---------------------+---------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssoc.join(\n",
    "    l2gTable.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "    on=\"leftStudyLocusId\",\n",
    "    how=\"left\",\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18333585"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasResolvedColoc.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "17244441"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasCredibleAssoc.distinct().count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16378337"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasCredibleAssoc.filter(F.col(\"colocDoE\").isNotNull()).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 881:>                                                        (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------------------+------------------+------------------+-------------------+---------------+---------------------+\n",
      "|       targetId|  diseaseId|homogenized|    leftStudyLocusId|                h4|      datasourceId|      resourceScore|  leftVariantId|credibleLeftStudyType|\n",
      "+---------------+-----------+-----------+--------------------+------------------+------------------+-------------------+---------------+---------------------+\n",
      "|ENSG00000000003|EFO_0004529|noEvaluable|9b68d48251bfb71e9...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004529|noEvaluable|cb3db2374fb1fbeda...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004529|noEvaluable|b9e47164bf645be64...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004611|noEvaluable|9b68d48251bfb71e9...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004611|noEvaluable|cb3db2374fb1fbeda...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004732|noEvaluable|9b68d48251bfb71e9...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004732|noEvaluable|cb3db2374fb1fbeda...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004732|noEvaluable|b9e47164bf645be64...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000419|EFO_0000313|noEvaluable|a84c1695b952c7f7a...|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|\n",
      "|ENSG00000000419|EFO_0004339|noEvaluable|c5c09f45e16bc37f9...|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|\n",
      "|ENSG00000000419|EFO_0004503|noEvaluable|f2140bdfde8f977e4...|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|\n",
      "|ENSG00000000457|EFO_0003872|GoF_protect|2abe5b5bb1ca56812...|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|\n",
      "|ENSG00000000457|EFO_0004286|noEvaluable|57969a20046c11221...|              null|gwas_credible_sets|0.06451956210793872|           null|                 null|\n",
      "|ENSG00000000460|EFO_0003911|noEvaluable|1eb9f83b09dd97b97...|              null|gwas_credible_sets|0.06116443109603223|           null|                 null|\n",
      "|ENSG00000000460|EFO_0004269|noEvaluable|1eb9f83b09dd97b97...|              null|gwas_credible_sets|0.06116443109603223|           null|                 null|\n",
      "|ENSG00000000460|EFO_0004339|noEvaluable|253165fd0bbff490a...|              null|gwas_credible_sets|0.35074980670304023|           null|                 null|\n",
      "|ENSG00000000938|EFO_0004742|noEvaluable|0447c925d979dac73...|              null|gwas_credible_sets|0.11826405203479222|           null|                 null|\n",
      "|ENSG00000000971|EFO_0001365|noEvaluable|cf0c2add70777ac20...|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|\n",
      "|ENSG00000000971|EFO_0001365|noEvaluable|cf0c2add70777ac20...|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|\n",
      "|ENSG00000000971|EFO_0001365|noEvaluable|cf0c2add70777ac20...|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|\n",
      "+---------------+-----------+-----------+--------------------+------------------+------------------+-------------------+---------------+---------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssoc.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "205167"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasCredibleAssoc.filter(F.col(\"homogenized\") != \"noEvaluable\").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"homogenized\"\n",
    ").count().count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "716633"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasCredibleAssoc.filter(F.col(\"homogenized\").isNotNull()).groupBy(\n",
    "    \"targetId\", \"diseaseId\",\"\"homogenized\"\"\n",
    ").count().count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18333585"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasCredibleAssoc.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+------------------+-------------------+--------------------+--------------------+------------+---------------+---------------------+-------------------+----------------------+---------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+------------+---------------+--------------------+--------------+--------------------+----------------+-----------+\n",
      "|  diseaseId|    leftStudyLocusId|       targetId|        rightStudyId|   rightStudyLocusId|chromosome|rightStudyType|numberColocalisingVariants|                  h0|                  h1|                  h2|                h3|                 h4|colocalisationMethod|betaRatioSignAverage| leftStudyId|  leftVariantId|credibleLeftStudyType|     rightVariantId|credibleRightStudyType|projectId|indexStudyType|condition|   biosampleId|      datasourceId|         datatypeId|diseaseFromSourceMappedId|      resourceScore|targetFromSourceId|                  id|              score|          sourceId|     studyId|      variantId|            betaGwas|pValueExponent|                name|therapeuticAreas|   colocDoE|\n",
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+------------------+-------------------+--------------------+--------------------+------------+---------------+---------------------+-------------------+----------------------+---------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+------------+---------------+--------------------+--------------+--------------------+----------------+-----------+\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|Braineac2_ge_subs...|60418c4b2bd9a44c1...|         2|          eqtl|                        75|2.158417791830621...|9.448151303386143...|6.654449402922124E-7|0.2905791572126367|  0.709420177247939|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232692936_C_T|                  eqtl|Braineac2|          eqtl|    naive|UBERON_0002038|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|Braineac2_ge_subs...|60418c4b2bd9a44c1...|         2|          eqtl|                        75|2.158417791830621...|9.448151303386143...|6.654449402922124E-7|0.2905791572126367|  0.709420177247939|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232692936_C_T|                  eqtl|Braineac2|          eqtl|    naive|UBERON_0002038|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|ROSMAP_leafcutter...|8e4541d2fe5f9814f...|         2|          sqtl|                         7|6.758486212870377...|2.958426317055846...|2.180262654145907...| 0.954331721227678|0.04566609850966728|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|2_232888546_T_TACAC|                  sqtl|   ROSMAP|          sqtl|    naive|UBERON_0009834|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|ROSMAP_leafcutter...|8e4541d2fe5f9814f...|         2|          sqtl|                         7|6.758486212870377...|2.958426317055846...|2.180262654145907...| 0.954331721227678|0.04566609850966728|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|2_232888546_T_TACAC|                  sqtl|   ROSMAP|          sqtl|    naive|UBERON_0009834|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|GTEx_tx_brain_spi...|bcef0afaf13e83196...|         2|          eqtl|                        52|1.633453531902611...|7.150198681595005E-4|1.585537081586007...|0.6937396144679886|0.30554377849331604|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232886667_A_G|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0006469|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|LoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|GTEx_tx_brain_spi...|bcef0afaf13e83196...|         2|          eqtl|                        52|1.633453531902611...|7.150198681595005E-4|1.585537081586007...|0.6937396144679886|0.30554377849331604|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232886667_A_G|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0006469|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|LoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|GTEx_txrev_brain_...|4765753cd7aa0937b...|         2|         tuqtl|                        69|1.016043804208129...|4.447579892748256E-6|7.947437823870326E-7|0.3472344520164696| 0.6527603056496945|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232866534_G_C|                 tuqtl|     GTEx|         tuqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|GTEx_txrev_brain_...|4765753cd7aa0937b...|         2|         tuqtl|                        69|1.016043804208129...|4.447579892748256E-6|7.947437823870326E-7|0.3472344520164696| 0.6527603056496945|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232866534_G_C|                 tuqtl|     GTEx|         tuqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|GTEx_ge_brain_cer...|58901258a26bad3d2...|         2|          eqtl|                        21|6.737528607159511...|2.949252441115216...|2.084478491736212E-6|0.9123616665521692|0.08763624893984978|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232872422_C_A|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|GTEx_ge_brain_cer...|58901258a26bad3d2...|         2|          eqtl|                        21|6.737528607159511...|2.949252441115216...|2.084478491736212E-6|0.9123616665521692|0.08763624893984978|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232872422_C_A|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004611|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0004732|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0005278|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0004529|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0020945|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0020945|0.07495761685624613|   ENSG00000154229|89c484048dd512d9b...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|free cholesterol:...|   [EFO_0001444]|       null|\n",
      "|EFO_0004529|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0020945|0.07495761685624613|   ENSG00000154229|89c484048dd512d9b...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|free cholesterol:...|   [EFO_0001444]|       null|\n",
      "|EFO_0800060|00117f0551bb5139e...|ENSG00000077616|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0800060| 0.8128893925911914|   ENSG00000077616|deb5a78336256c162...| 0.8128893925911914|gwas_credible_sets|GCST90199911|11_90163402_C_T| 0.14911349596103268|           -18|beta-citrylglutam...|   [EFO_0001444]|       null|\n",
      "|EFO_0004725|00117f0551bb5139e...|ENSG00000077616|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0800060| 0.8128893925911914|   ENSG00000077616|deb5a78336256c162...| 0.8128893925911914|gwas_credible_sets|GCST90199911|11_90163402_C_T| 0.14911349596103268|           -18|beta-citrylglutam...|   [EFO_0001444]|       null|\n",
      "|EFO_0010228|001893cc5566155b3...|ENSG00000166035|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0010228|  0.964496680049245|   ENSG00000166035|0b79549d99fc5e005...|  0.964496680049245|gwas_credible_sets|GCST90024356|15_58563549_C_T| 0.08450956600909004|            -9|phosphatidylethan...|   [EFO_0001444]|       null|\n",
      "|EFO_0010968|001893cc5566155b3...|ENSG00000166035|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0010228|  0.964496680049245|   ENSG00000166035|0b79549d99fc5e005...|  0.964496680049245|gwas_credible_sets|GCST90024356|15_58563549_C_T| 0.08450956600909004|            -9|phosphatidylethan...|   [EFO_0001444]|       null|\n",
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+------------------+-------------------+--------------------+--------------------+------------+---------------+---------------------+-------------------+----------------------+---------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+------------+---------------+--------------------+--------------+--------------------+----------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gwasResolvedColoc.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+--------------------+-------------------+--------------------+--------------------+-------------------+---------------+---------------------+----------------+----------------------+-----------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+-------------------+---------------+------------------+--------------+--------------------+--------------------+--------+\n",
      "|  diseaseId|    leftStudyLocusId|       targetId|        rightStudyId|   rightStudyLocusId|chromosome|rightStudyType|numberColocalisingVariants|                  h0|                  h1|                  h2|                  h3|                 h4|colocalisationMethod|betaRatioSignAverage|        leftStudyId|  leftVariantId|credibleLeftStudyType|  rightVariantId|credibleRightStudyType|  projectId|indexStudyType|condition|   biosampleId|      datasourceId|         datatypeId|diseaseFromSourceMappedId|      resourceScore|targetFromSourceId|                  id|              score|          sourceId|            studyId|      variantId|          betaGwas|pValueExponent|                name|    therapeuticAreas|colocDoE|\n",
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+--------------------+-------------------+--------------------+--------------------+-------------------+---------------+---------------------+----------------+----------------------+-----------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+-------------------+---------------+------------------+--------------+--------------------+--------------------+--------+\n",
      "|EFO_0001365|530c242206d0f48e3...|ENSG00000000971|                null|                null|      null|          null|                      null|                null|                null|                null|                null|               null|                null|                null|               null|           null|                 null|            null|                  null|       null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0001365| 0.4804678866007057|   ENSG00000000971|9a2672820aadd5c5a...| 0.4804678866007057|gwas_credible_sets|       GCST90250835|1_196665380_A_G|             -0.02|           -11|age-related macul...|[MONDO_0002025, O...|    null|\n",
      "|EFO_0001365|201f2c0647e9739dc...|ENSG00000000971|                null|                null|      null|          null|                      null|                null|                null|                null|                null|               null|                null|                null|               null|           null|                 null|            null|                  null|       null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0001365| 0.4841512340146833|   ENSG00000000971|fbd97826d8c045064...| 0.4841512340146833|gwas_credible_sets|       GCST90086110|1_196335068_T_C|           0.54431|           -12|age-related macul...|[MONDO_0002025, O...|    null|\n",
      "|EFO_0001365|b6b18799734a0e98a...|ENSG00000000971|UKB_PPP_EUR_CFH_P...|8b962c17fec250d39...|         1|          pqtl|                         1|7.290598617650363...|1.668381507631917...|0.021341075666305106|  0.4878786645816905|0.49078025975200457|               COLOC|                -1.0| FINNGEN_R11_H7_AMD|1_196414256_T_G|                 gwas| 1_196518011_A_C|                  pqtl|UKB_PPP_EUR|          pqtl|     null|UBERON_0001969|gwas_credible_sets|genetic_association|              EFO_0001365| 0.7886887071190862|   ENSG00000000971|e198fa82d38f147ad...| 0.7886887071190862|gwas_credible_sets| FINNGEN_R11_H7_AMD|1_196414256_T_G|0.0358279712273804|           -17|age-related macul...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|b6b18799734a0e98a...|ENSG00000000971|UKB_PPP_EUR_CFH_P...|d3f2cf893f5f6598e...|         1|          pqtl|                         5|2.620275367217217E-6|5.996267659504628E-5|7.459021281324574E-4|0.016086202990679963| 0.9831053119292256|               COLOC|                -1.0| FINNGEN_R11_H7_AMD|1_196414256_T_G|                 gwas| 1_196737512_C_T|                  pqtl|UKB_PPP_EUR|          pqtl|     null|UBERON_0001969|gwas_credible_sets|genetic_association|              EFO_0001365| 0.7886887071190862|   ENSG00000000971|e198fa82d38f147ad...| 0.7886887071190862|gwas_credible_sets| FINNGEN_R11_H7_AMD|1_196414256_T_G|0.0358279712273804|           -17|age-related macul...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|6584a07d78f85a9c1...|ENSG00000000971|UKB_PPP_EUR_CFH_P...|8b962c17fec250d39...|         1|          pqtl|                         1|7.052024460376012...|3.461826200806243...|2.064272023495675...| 0.00914263774113796| 0.9908367195386293|               COLOC|                -1.0|FINNGEN_R11_DRY_AMD|1_196619333_T_G|                 gwas| 1_196518011_A_C|                  pqtl|UKB_PPP_EUR|          pqtl|     null|UBERON_0001969|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.9623436973587878|   ENSG00000000971|eda142ef772235679...| 0.9623436973587878|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196619333_T_G|0.0628317994207065|           -21|dry age related m...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|706c8ec796a869ad1...|ENSG00000000971|                null|                null|      null|          null|                      null|                null|                null|                null|                null|               null|                null|                null|               null|           null|                 null|            null|                  null|       null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0001365| 0.6178769225770331|   ENSG00000000971|d19afd8a6ff804299...| 0.6178769225770331|gwas_credible_sets|       GCST90250837|1_196665380_A_G|             -0.01|           -22|age-related macul...|[MONDO_0002025, O...|    null|\n",
      "|EFO_0001365|50f30281f847da629...|ENSG00000000971|                null|                null|      null|          null|                      null|                null|                null|                null|                null|               null|                null|                null|               null|           null|                 null|            null|                  null|       null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0001365|0.31032362161006505|   ENSG00000000971|5b2f118dcd821b2cf...|0.31032362161006505|gwas_credible_sets|         GCST001987|1_196704863_A_C|0.9669838461896731|           -24|age-related macul...|[MONDO_0002025, O...|    null|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_tx_artery_ao...|20bf26b5823b6c86f...|         1|          eqtl|                         4|9.74270214780558E-11|1.760705458736370...|2.539966779467673E-5|  0.4584824339401917| 0.5414904055891289|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196706226_G_A|                  eqtl|       GTEx|          eqtl|    naive|UBERON_0001496|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_leafcutter_c...|29562aff1ba463496...|         1|          sqtl|                        79|1.215071567962966...|2.195882612062425...|6.640309114783812E-6| 0.11912330411286505| 0.8806504551660996|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196855148_A_G|                  sqtl|       GTEx|          sqtl|    naive|UBERON_0001157|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_tx_small_int...|d1f815338e36976df...|         1|          eqtl|                         3|3.384186888064291...|6.115917572765925E-6|3.009499871738586E-5|  0.5434216111158117| 0.4565421776294782|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196710552_C_T|                  eqtl|       GTEx|          eqtl|    naive|UBERON_0001211|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|UKB_PPP_EUR_CFH_P...|743bd10e3bbb422b6...|         1|          pqtl|                        57|2.914805597094355...|5.267649706718783...|2.737322211595228...|   0.494184305573193| 0.5057883212047093|               COLOC|                -1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_197058476_A_G|                  pqtl|UKB_PPP_EUR|          pqtl|     null|UBERON_0001969|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_exon_adipose...|d081122a62cd0943d...|         1|          eqtl|                        11|1.345310303649142...|2.431250956868485...|1.258126831881903...| 0.22659611047476574| 0.7731481697081263|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196705584_T_A|                  eqtl|       GTEx|          eqtl|    naive|UBERON_0001013|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|TwinsUK_tx_fat_EN...|f766ef3fcb64a8db8...|         1|          eqtl|                        66|1.360499389436279E-8|2.458700932903683...|7.852947275399569E-6| 0.14106014855714322| 0.8586861147972965|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196890414_G_A|                  eqtl|    TwinsUK|          eqtl|    naive|UBERON_0001013|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_txrev_adipos...|824ea07c6e0f2e53a...|         1|         tuqtl|                        28|5.589808467786663...|1.010192763673557...|1.319756008994187...|   0.237744673345885| 0.7622411188453648|               COLOC|                -1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196705584_T_A|                 tuqtl|       GTEx|         tuqtl|    naive|UBERON_0001013|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|FUSION_exon_muscl...|0b9b560b904afea05...|         1|          eqtl|                         4|1.743002270416398...|3.149961739813277E-5|1.815034808370219E-5| 0.32734131842357445| 0.6726090298679421|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196683704_C_T|                  eqtl|     FUSION|          eqtl|    naive|UBERON_0001134|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|TwinsUK_txrev_fat...|9335b972199d20b53...|         1|         tuqtl|                        57|8.880750744562661...|1.604933156197625...|6.565411105880568...| 0.11776818727245003|  0.882209197096808|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196890414_G_A|                 tuqtl|    TwinsUK|         tuqtl|    naive|UBERON_0001013|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_txrev_lung_E...|3c37e8fc87dd61baf...|         1|         tuqtl|                         2|1.077555184934397...|1.947362537022238E-9|2.654854572573687E-5|  0.4792657949910025|  0.520707654515801|               COLOC|                -1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196710552_C_T|                 tuqtl|       GTEx|         tuqtl|    naive|UBERON_0008952|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|LoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|ROSMAP_txrev_brai...|7b5a7076407636226...|         1|         tuqtl|                        20|3.033476796474597...|0.005482113526499141|4.737920187875709...|  0.8561007941669286| 0.1383694097570129|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas|1_196849157_G_GT|                 tuqtl|     ROSMAP|         tuqtl|    naive|UBERON_0009834|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|4bd19c6aa1b2b4ef7...|ENSG00000000971|GTEx_tx_lung_ENST...|6981f1fbed494d528...|         1|          eqtl|                        14|4.148704659559032...|7.497557899388134E-8|1.228665242430588...| 0.22126621911088576| 0.7787214192569629|               COLOC|                 1.0|FINNGEN_R11_DRY_AMD|1_196993852_A_G|                 gwas| 1_196710552_C_T|                  eqtl|       GTEx|          eqtl|    naive|UBERON_0008952|gwas_credible_sets|genetic_association|            MONDO_0100114| 0.4536236865987458|   ENSG00000000971|ac149dbaaf443aa08...| 0.4536236865987458|gwas_credible_sets|FINNGEN_R11_DRY_AMD|1_196993852_A_G|0.0124160006584544|           -24|dry age related m...|[MONDO_0002025, O...|GoF_risk|\n",
      "|EFO_0001365|bcb23efe5a1f0a53c...|ENSG00000000971|                null|                null|      null|          null|                      null|                null|                null|                null|                null|               null|                null|                null|               null|           null|                 null|            null|                  null|       null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0001365|0.42773188122842004|   ENSG00000000971|cbf61fc7985719b80...|0.42773188122842004|gwas_credible_sets|       GCST90134545|1_196691131_A_G|        -0.2086145|           -25|age-related macul...|[MONDO_0002025, O...|    null|\n",
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+--------------------+-------------------+--------------------+--------------------+-------------------+---------------+---------------------+----------------+----------------------+-----------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+-------------------+---------------+------------------+--------------+--------------------+--------------------+--------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gwasResolvedColoc.filter(\n",
    "    (F.col(\"targetId\") == \"ENSG00000000971\") & (F.col(\"diseaseId\") == \"EFO_0001365\")\n",
    ").sort(F.col(\"pValueExponent\").desc()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 414:>                                                        (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------------+-----+\n",
      "|       targetId|  diseaseId|   colocDoE|homogenizedRaw|count|\n",
      "+---------------+-----------+-----------+--------------+-----+\n",
      "|ENSG00000000003|EFO_0004529|       null|          null|    3|\n",
      "|ENSG00000000003|EFO_0004611|       null|          null|    2|\n",
      "|ENSG00000000003|EFO_0004732|       null|          null|    3|\n",
      "|ENSG00000000419|EFO_0000313|       null|          null|    1|\n",
      "|ENSG00000000419|EFO_0004339|       null|          null|    1|\n",
      "|ENSG00000000419|EFO_0004503|       null|          null|    1|\n",
      "|ENSG00000000457|EFO_0003872|GoF_protect|   GoF_protect|    1|\n",
      "|ENSG00000000457|EFO_0004286|       null|          null|    1|\n",
      "|ENSG00000000460|EFO_0003911|       null|          null|    1|\n",
      "|ENSG00000000460|EFO_0004269|       null|          null|    1|\n",
      "|ENSG00000000460|EFO_0004339|       null|          null|    1|\n",
      "|ENSG00000000938|EFO_0004742|       null|          null|    1|\n",
      "|ENSG00000000971|EFO_0001365|       null|          null|   33|\n",
      "|ENSG00000000971|EFO_0001365|GoF_protect|          null|  127|\n",
      "|ENSG00000000971|EFO_0001365|LoF_protect|          null|  118|\n",
      "|ENSG00000000971|EFO_0001365|   GoF_risk|          null|    8|\n",
      "|ENSG00000000971|EFO_0001365|   LoF_risk|          null|    7|\n",
      "|ENSG00000000971|EFO_0004128|       null|          null|    3|\n",
      "|ENSG00000000971|EFO_0004555|       null|          null|    8|\n",
      "|ENSG00000000971|EFO_0004555|   LoF_risk|          null|   23|\n",
      "+---------------+-----------+-----------+--------------+-----+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasResolvedColoc.withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(), F.first(\"colocDoE\").over(window_spec)\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "375518"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasResolvedColoc.filter(F.col(\"colocDoE\").isNotNull()).withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(), F.first(\"colocDoE\").over(window_spec)\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().filter(\n",
    "    F.col(\"homogenizedRaw\").isNotNull()\n",
    ").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwasResolvedColoc.filter(F.col(\"colocDoE\").isNotNull()).withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(), F.first(\"colocDoE\").over(window_spec)\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().filter(\n",
    "    F.col(\"homogenizedRaw\").isNotNull()\n",
    ").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 737:===================================================>   (16 + 1) / 17]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+------+\n",
      "|colocDoE| count|\n",
      "+--------+------+\n",
      "|    null|622321|\n",
      "+--------+------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "# ignorenulls=True\n",
    "gwasResolvedColoc.withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(),\n",
    "        F.first(\"colocDoE\", ignorenulls=True).over(window_spec),\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().filter(\n",
    "    F.col(\"homogenizedRaw\").isNull()\n",
    ").groupBy(\n",
    "    \"colocDoE\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "997839"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasResolvedColoc.withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(), F.first(\"colocDoE\").over(window_spec)\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "742682"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwasResolvedColoc.withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(), F.first(\"colocDoE\").over(window_spec)\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().filter(\n",
    "    F.col(\"homogenizedRaw\").isNull()\n",
    ").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 373:>                                                        (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------------+-----+\n",
      "|       targetId|  diseaseId|   colocDoE|homogenizedRaw|count|\n",
      "+---------------+-----------+-----------+--------------+-----+\n",
      "|ENSG00000000003|EFO_0004529|       null|          null|    3|\n",
      "|ENSG00000000003|EFO_0004611|       null|          null|    2|\n",
      "|ENSG00000000003|EFO_0004732|       null|          null|    3|\n",
      "|ENSG00000000419|EFO_0000313|       null|          null|    1|\n",
      "|ENSG00000000419|EFO_0004339|       null|          null|    1|\n",
      "|ENSG00000000419|EFO_0004503|       null|          null|    1|\n",
      "|ENSG00000000457|EFO_0003872|GoF_protect|   GoF_protect|    1|\n",
      "|ENSG00000000457|EFO_0004286|       null|          null|    1|\n",
      "|ENSG00000000460|EFO_0003911|       null|          null|    1|\n",
      "|ENSG00000000460|EFO_0004269|       null|          null|    1|\n",
      "|ENSG00000000460|EFO_0004339|       null|          null|    1|\n",
      "|ENSG00000000938|EFO_0004742|       null|          null|    1|\n",
      "|ENSG00000000971|EFO_0001365|       null|          null|   33|\n",
      "|ENSG00000000971|EFO_0001365|GoF_protect|          null|  127|\n",
      "|ENSG00000000971|EFO_0001365|LoF_protect|          null|  118|\n",
      "|ENSG00000000971|EFO_0001365|   GoF_risk|          null|    8|\n",
      "|ENSG00000000971|EFO_0001365|   LoF_risk|          null|    7|\n",
      "|ENSG00000000971|EFO_0004128|       null|          null|    3|\n",
      "|ENSG00000000971|EFO_0004555|       null|          null|    8|\n",
      "|ENSG00000000971|EFO_0004555|   LoF_risk|          null|   23|\n",
      "+---------------+-----------+-----------+--------------+-----+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasResolvedColoc.withColumn(\n",
    "    \"homogenizedRaw\",\n",
    "    F.when(\n",
    "        F.col(\"colocDoE\").isNotNull(), F.first(\"colocDoE\").over(window_spec)\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\"targetId\", \"diseaseId\", \"colocDoE\", \"homogenizedRaw\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+---------+--------+--------------+--------------------+-----+-----------+------+\n",
      "|targetId|diseaseId|colocDoE|pValueExponent|betaRatioSignAverage|count|homogenized|allDoE|\n",
      "+--------+---------+--------+--------------+--------------------+-----+-----------+------+\n",
      "+--------+---------+--------+--------------+--------------------+-----+-----------+------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gwasResolvedColoc.groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"colocDoE\", \"pValueExponent\", \"betaRatioSignAverage\"\n",
    ").count().withColumn(\"homogenized\", F.first(\"colocDoE\").over(window_spec)).withColumn(\n",
    "    \"allDoE\", F.size(F.collect_set(\"homogenized\").over(window_spec))\n",
    ").filter(\n",
    "    F.col(\"allDoE\") > 1\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 164:>                                                        (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------------------+------------------+------------------+-------------------+---------------+---------------------+\n",
      "|       targetId|  diseaseId|homogenized|    leftStudyLocusId|                h4|      datasourceId|      resourceScore|  leftVariantId|credibleLeftStudyType|\n",
      "+---------------+-----------+-----------+--------------------+------------------+------------------+-------------------+---------------+---------------------+\n",
      "|ENSG00000000003|EFO_0004529|noEvaluable|9b68d48251bfb71e9...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004529|noEvaluable|cb3db2374fb1fbeda...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004529|noEvaluable|b9e47164bf645be64...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004611|noEvaluable|9b68d48251bfb71e9...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004611|noEvaluable|cb3db2374fb1fbeda...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004732|noEvaluable|9b68d48251bfb71e9...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004732|noEvaluable|cb3db2374fb1fbeda...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000003|EFO_0004732|noEvaluable|b9e47164bf645be64...|              null|gwas_credible_sets| 0.9525927165569602|           null|                 null|\n",
      "|ENSG00000000419|EFO_0000313|noEvaluable|a84c1695b952c7f7a...|              null|gwas_credible_sets|0.24191390254194645|           null|                 null|\n",
      "|ENSG00000000419|EFO_0004339|noEvaluable|c5c09f45e16bc37f9...|              null|gwas_credible_sets|0.07392676917421312|           null|                 null|\n",
      "|ENSG00000000419|EFO_0004503|noEvaluable|f2140bdfde8f977e4...|              null|gwas_credible_sets|0.19619841264085358|           null|                 null|\n",
      "|ENSG00000000457|EFO_0003872|GoF_protect|2abe5b5bb1ca56812...|0.9757945803024912|gwas_credible_sets|0.07566412965156898|1_170170367_T_C|                 gwas|\n",
      "|ENSG00000000457|EFO_0004286|noEvaluable|57969a20046c11221...|              null|gwas_credible_sets|0.06451956210793872|           null|                 null|\n",
      "|ENSG00000000460|EFO_0003911|noEvaluable|1eb9f83b09dd97b97...|              null|gwas_credible_sets|0.06116443109603223|           null|                 null|\n",
      "|ENSG00000000460|EFO_0004269|noEvaluable|1eb9f83b09dd97b97...|              null|gwas_credible_sets|0.06116443109603223|           null|                 null|\n",
      "|ENSG00000000460|EFO_0004339|noEvaluable|253165fd0bbff490a...|              null|gwas_credible_sets|0.35074980670304023|           null|                 null|\n",
      "|ENSG00000000938|EFO_0004742|noEvaluable|0447c925d979dac73...|              null|gwas_credible_sets|0.11826405203479222|           null|                 null|\n",
      "|ENSG00000000971|EFO_0001365|noEvaluable|cf0c2add70777ac20...|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|\n",
      "|ENSG00000000971|EFO_0001365|noEvaluable|cf0c2add70777ac20...|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|\n",
      "|ENSG00000000971|EFO_0001365|noEvaluable|cf0c2add70777ac20...|              null|gwas_credible_sets| 0.8061740627656493|           null|                 null|\n",
      "+---------------+-----------+-----------+--------------------+------------------+------------------+-------------------+---------------+---------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "gwasCredibleAssoc.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+------------------+-------------------+--------------------+--------------------+------------+---------------+---------------------+-------------------+----------------------+---------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+------------+---------------+--------------------+--------------+--------------------+----------------+-----------+\n",
      "|  diseaseId|    leftStudyLocusId|       targetId|        rightStudyId|   rightStudyLocusId|chromosome|rightStudyType|numberColocalisingVariants|                  h0|                  h1|                  h2|                h3|                 h4|colocalisationMethod|betaRatioSignAverage| leftStudyId|  leftVariantId|credibleLeftStudyType|     rightVariantId|credibleRightStudyType|projectId|indexStudyType|condition|   biosampleId|      datasourceId|         datatypeId|diseaseFromSourceMappedId|      resourceScore|targetFromSourceId|                  id|              score|          sourceId|     studyId|      variantId|            betaGwas|pValueExponent|                name|therapeuticAreas|   colocDoE|\n",
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+------------------+-------------------+--------------------+--------------------+------------+---------------+---------------------+-------------------+----------------------+---------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+------------+---------------+--------------------+--------------+--------------------+----------------+-----------+\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|Braineac2_ge_subs...|60418c4b2bd9a44c1...|         2|          eqtl|                        75|2.158417791830621...|9.448151303386143...|6.654449402922124E-7|0.2905791572126367|  0.709420177247939|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232692936_C_T|                  eqtl|Braineac2|          eqtl|    naive|UBERON_0002038|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|Braineac2_ge_subs...|60418c4b2bd9a44c1...|         2|          eqtl|                        75|2.158417791830621...|9.448151303386143...|6.654449402922124E-7|0.2905791572126367|  0.709420177247939|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232692936_C_T|                  eqtl|Braineac2|          eqtl|    naive|UBERON_0002038|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|ROSMAP_leafcutter...|8e4541d2fe5f9814f...|         2|          sqtl|                         7|6.758486212870377...|2.958426317055846...|2.180262654145907...| 0.954331721227678|0.04566609850966728|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|2_232888546_T_TACAC|                  sqtl|   ROSMAP|          sqtl|    naive|UBERON_0009834|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|ROSMAP_leafcutter...|8e4541d2fe5f9814f...|         2|          sqtl|                         7|6.758486212870377...|2.958426317055846...|2.180262654145907...| 0.954331721227678|0.04566609850966728|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|2_232888546_T_TACAC|                  sqtl|   ROSMAP|          sqtl|    naive|UBERON_0009834|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|GTEx_tx_brain_spi...|bcef0afaf13e83196...|         2|          eqtl|                        52|1.633453531902611...|7.150198681595005E-4|1.585537081586007...|0.6937396144679886|0.30554377849331604|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232886667_A_G|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0006469|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|LoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|GTEx_tx_brain_spi...|bcef0afaf13e83196...|         2|          eqtl|                        52|1.633453531902611...|7.150198681595005E-4|1.585537081586007...|0.6937396144679886|0.30554377849331604|               COLOC|                 1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232886667_A_G|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0006469|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|LoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|GTEx_txrev_brain_...|4765753cd7aa0937b...|         2|         tuqtl|                        69|1.016043804208129...|4.447579892748256E-6|7.947437823870326E-7|0.3472344520164696| 0.6527603056496945|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232866534_G_C|                 tuqtl|     GTEx|         tuqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|GTEx_txrev_brain_...|4765753cd7aa0937b...|         2|         tuqtl|                        69|1.016043804208129...|4.447579892748256E-6|7.947437823870326E-7|0.3472344520164696| 0.6527603056496945|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232866534_G_C|                 tuqtl|     GTEx|         tuqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004838|000b0186be8b2498b...|ENSG00000182600|GTEx_ge_brain_cer...|58901258a26bad3d2...|         2|          eqtl|                        21|6.737528607159511...|2.949252441115216...|2.084478491736212E-6|0.9123616665521692|0.08763624893984978|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232872422_C_A|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0001444|000b0186be8b2498b...|ENSG00000182600|GTEx_ge_brain_cer...|58901258a26bad3d2...|         2|          eqtl|                        21|6.737528607159511...|2.949252441115216...|2.084478491736212E-6|0.9123616665521692|0.08763624893984978|               COLOC|                -1.0|GCST90018951|2_232818535_G_C|                 gwas|    2_232872422_C_A|                  eqtl|     GTEx|          eqtl|    naive|UBERON_0002037|gwas_credible_sets|genetic_association|              EFO_0004838|0.08914426799481445|   ENSG00000182600|4c13bab779458c4a7...|0.08914426799481445|gwas_credible_sets|GCST90018951|2_232818535_G_C|-0.01019498607227...|           -11| calcium measurement|   [EFO_0001444]|GoF_protect|\n",
      "|EFO_0004611|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0004732|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0005278|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0004529|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0004611|0.07495761685624613|   ENSG00000154229|d7cbe495598171e75...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|low density lipop...|   [EFO_0001444]|       null|\n",
      "|EFO_0020945|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0020945|0.07495761685624613|   ENSG00000154229|89c484048dd512d9b...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|free cholesterol:...|   [EFO_0001444]|       null|\n",
      "|EFO_0004529|00111a60f430e7ba0...|ENSG00000154229|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0020945|0.07495761685624613|   ENSG00000154229|89c484048dd512d9b...|0.07495761685624613|gwas_credible_sets|GCST90302048|17_66214462_A_C|-0.01636083112161...|           -10|free cholesterol:...|   [EFO_0001444]|       null|\n",
      "|EFO_0800060|00117f0551bb5139e...|ENSG00000077616|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0800060| 0.8128893925911914|   ENSG00000077616|deb5a78336256c162...| 0.8128893925911914|gwas_credible_sets|GCST90199911|11_90163402_C_T| 0.14911349596103268|           -18|beta-citrylglutam...|   [EFO_0001444]|       null|\n",
      "|EFO_0004725|00117f0551bb5139e...|ENSG00000077616|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0800060| 0.8128893925911914|   ENSG00000077616|deb5a78336256c162...| 0.8128893925911914|gwas_credible_sets|GCST90199911|11_90163402_C_T| 0.14911349596103268|           -18|beta-citrylglutam...|   [EFO_0001444]|       null|\n",
      "|EFO_0010228|001893cc5566155b3...|ENSG00000166035|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0010228|  0.964496680049245|   ENSG00000166035|0b79549d99fc5e005...|  0.964496680049245|gwas_credible_sets|GCST90024356|15_58563549_C_T| 0.08450956600909004|            -9|phosphatidylethan...|   [EFO_0001444]|       null|\n",
      "|EFO_0010968|001893cc5566155b3...|ENSG00000166035|                null|                null|      null|          null|                      null|                null|                null|                null|              null|               null|                null|                null|        null|           null|                 null|               null|                  null|     null|          null|     null|          null|gwas_credible_sets|genetic_association|              EFO_0010228|  0.964496680049245|   ENSG00000166035|0b79549d99fc5e005...|  0.964496680049245|gwas_credible_sets|GCST90024356|15_58563549_C_T| 0.08450956600909004|            -9|phosphatidylethan...|   [EFO_0001444]|       null|\n",
      "+-----------+--------------------+---------------+--------------------+--------------------+----------+--------------+--------------------------+--------------------+--------------------+--------------------+------------------+-------------------+--------------------+--------------------+------------+---------------+---------------------+-------------------+----------------------+---------+--------------+---------+--------------+------------------+-------------------+-------------------------+-------------------+------------------+--------------------+-------------------+------------------+------------+---------------+--------------------+--------------+--------------------+----------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gwasResolvedColoc.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### part 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genEvidDataset = (\n",
    "    prueba_assessment.filter(F.col(\"datasourceId\") != \"chembl\")  #### checked 31.05.2023\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.count(\"targetId\").alias(\"Nr_evidences\"))\n",
    "    .select(\"targetId\", \"diseaseId\", \"Nr_evidences\")\n",
    "    .withColumn(\"geneticEvidence\", F.lit(\"hasGeneticEvidence\"))\n",
    "    .drop(\"Nr_evidences\")\n",
    ")\n",
    "\n",
    "coherency_toAssess_others_datasource = (  #### checked 31.05.2023\n",
    "    prueba_assessment.filter(\n",
    "        (F.col(\"homogenized\").isin(columns)) & (F.col(\"datasourceId\") != \"chembl\")\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"datasourceId\").alias(\"datasourceIds\"))\n",
    ")\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "l2gPred = spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/locusToGenePredictions\"\n",
    ")\n",
    "l2gTable = (\n",
    "    l2gPred.select(\"studyLocusId\", \"geneId\", F.explode_outer(\"locusToGeneFeatures\"))\n",
    "    .filter(F.col(\"key\").isin([\"distanceFootprintMean\", \"distanceTssMean\"]))\n",
    "    .groupBy(\"studyLocusId\", \"geneId\")\n",
    "    .pivot(\"key\")\n",
    "    .agg(F.first(\"value\"))\n",
    ")\n",
    "\n",
    "\"\"\"\n",
    "v2g = spark.read.parquet(\"gs://genetics-portal-dev-data/22.09.1/outputs/v2g\")\n",
    "varDistToGene = v2g.select(\n",
    "    F.concat_ws(\"_\", \"chr_id\", \"position\", \"ref_allele\", \"alt_allele\").alias(\n",
    "        \"variantId\"\n",
    "    ),\n",
    "    F.col(\"gene_id\").alias(\"targetId\"),\n",
    "    \"source_id\",\n",
    "    \"d\",\n",
    "    \"distance_score\",\n",
    ").filter(F.col(\"source_id\") == \"canonical_tss\")\n",
    "\"\"\"\n",
    "\n",
    "#######\n",
    "# Build Ot genetics dataset as supporting evidence\n",
    "#######\n",
    "otGenetics = (\n",
    "    prueba_assessment.filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    # .filter((F.col(\"homogenized\") != \"noEvaluable\"))\n",
    "    .join(\n",
    "        varDistToGene, on=[\"variantId\", \"targetId\"], how=\"left\"\n",
    "    )  ### substitute by l2gTable\n",
    "    .join(genEvidDataset, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.collect_set(\"datasourceId\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"L2G_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageL2G\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"resourceScore\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"averageCanonicalTSSDistance\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.avg(\"d\").over(ranking.orderBy(F.col(\"resourceScore\").desc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"datasources\",\n",
    "        F.when(\n",
    "            F.col(\"rank\").isNull(),\n",
    "            F.array_remove(F.col(\"datasources\"), \"ot_genetics_portal\"),\n",
    "        ).otherwise(F.col(\"datasources\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"distance_ranking\",\n",
    "        F.when(\n",
    "            (F.col(\"datasourceId\") == \"ot_genetics_portal\"),\n",
    "            F.row_number().over(ranking.orderBy(F.col(\"d\").asc())),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"ChemblL2gRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"L2G_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"chemblDistanceRanking\",\n",
    "        F.when(\n",
    "            (F.array_contains(F.col(\"datasources\"), \"chembl\"))\n",
    "            & (F.array_contains(F.col(\"datasources\"), \"ot_genetics_portal\")),\n",
    "            F.lit(F.col(\"distance_ranking\")),\n",
    "        ).otherwise(F.lit(None)),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"frontierValue\",\n",
    "        ## ot genetics portal\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"beta\") <= 0.1) & (F.col(\"beta\") >= -0.1),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.when(\n",
    "                    (F.col(\"OddsRatio\") <= 1.1) & (F.col(\"OddsRatio\") >= 0.9),\n",
    "                    F.lit(\"limitValue\"),\n",
    "                ).otherwise(F.lit(\"noLimitValue\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.lit(\"noValue\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "#####\n",
    "# function for interpreting DoE and coherencies/discrepancies\n",
    "#####\n",
    "\n",
    "analysis_chembl = discrepancifier(\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\") == \"chembl\"))\n",
    "    .withColumn(\n",
    "        \"maxClinPhase\",\n",
    "        F.max(F.col(\"clinicalPhase\")).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "    .pivot(\"homogenized\")\n",
    "    .agg(F.count(\"targetId\"))\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### propag OtGenetics:\n",
    "otGenetics_propag = (\n",
    "    otGenetics.join(  ### propagated using parent terms\n",
    "        diseases.selectExpr(\n",
    "            \"id as diseaseId\",\n",
    "            \"parents\",\n",
    "        ),\n",
    "        on=\"diseaseId\",\n",
    "        how=\"left\",\n",
    "    ).withColumn(  ### propagating ###\n",
    "        \"diseaseId\",\n",
    "        F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "\n",
    "#### include dictionary for calling dataframes:\n",
    "# max_L2GScore\n",
    "# min_distance_ranking\n",
    "\n",
    "\n",
    "def benchmarkOT(discrepancifier, otGenetics, metric):\n",
    "    dict_comb = {}\n",
    "    dict_comb = {\n",
    "        \"hasGeneticEvidence\": f\"{metric}\",\n",
    "        \"diagonalYes\": f\"{metric}\",\n",
    "        \"oneCellYes\": f\"{metric}\",\n",
    "        \"L2GAndColoc\": f\"{metric}\",\n",
    "    }\n",
    "    list_l2g = [\n",
    "        0.10,\n",
    "        0.15,\n",
    "        0.20,\n",
    "        0.25,\n",
    "        0.30,\n",
    "        0.35,\n",
    "        0.40,\n",
    "        0.45,\n",
    "        0.50,\n",
    "        0.55,\n",
    "        0.60,\n",
    "        0.65,\n",
    "        0.70,\n",
    "        0.75,\n",
    "        0.80,\n",
    "        0.85,\n",
    "        0.90,\n",
    "        0.95,\n",
    "    ]\n",
    "    list_dist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
    "    return (\n",
    "        discrepancifier(\n",
    "            otGenetics.filter((F.col(\"datasourceId\") == \"ot_genetics_portal\"))\n",
    "            .withColumn(\n",
    "                \"min_distance_ranking\",\n",
    "                F.min(\"distance_ranking\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(  ### take maximum L2G score per T-D\n",
    "                \"max_L2GScore\",\n",
    "                F.max(\"resourceScore\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .groupBy(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"geneticEvidence\",\n",
    "                f\"{value}\",\n",
    "            )  ##### modifications here to include the groups of ranking/distances to TSS\n",
    "            .pivot(\"homogenized\")\n",
    "            .agg(F.count(\"targetId\"))\n",
    "        )\n",
    "        .selectExpr(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "            \"geneticEvidence\",\n",
    "            f\"{metric}\",\n",
    "            \"coherencyDiagonal as coherencyDiagonal\",\n",
    "            \"coherencyOneCell as coherencyOneCell\",\n",
    "            \"LoF_protect\",\n",
    "            \"GoF_protect\",\n",
    "            \"LoF_risk\",\n",
    "            \"GoF_risk\",\n",
    "        )\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(f\"{metric}\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"L2GAndColoc\",\n",
    "            F.when(\n",
    "                (F.col(\"geneticEvidence\") == \"hasGeneticEvidence\")\n",
    "                & (F.col(\"coherencyDiagonal\").isin([\"coherent\", \"dispar\"])),\n",
    "                F.lit(\"yes\"),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .select(\n",
    "            [\"*\"]\n",
    "            + (\n",
    "                [  ### single columns\n",
    "                    F.when(F.col(f\"{metric}\") >= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}>={str(n).replace('.', '_')}\")\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"  # Adjust this condition as needed\n",
    "                else [\n",
    "                    F.when(F.col(f\"{metric}\") <= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{metric}<={n}\")\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "            + (\n",
    "                [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "                if metric == \"max_L2GScore\"\n",
    "                else [\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) <= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}<={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_dist\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "metric_list = [\"max_L2GScore\", \"min_distance_ranking\"]\n",
    "datasetDict = {}\n",
    "for value in metric_list:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_l2g_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )\n",
    "    elif value == \"min_distance_ranking\":\n",
    "        datasetDict[f\"df_distance_original\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics, value\n",
    "        )\n",
    "        datasetDict[f\"df_distance_propagated\"] = benchmarkOT(\n",
    "            discrepancifier, otGenetics_propag, value\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-01-27 15:17:49.358444\n",
      "Analysis started on 2025-01-27 at  2025-01-27 15:17:49.358444\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/27 15:17:54 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read spark files\n",
      "fixing scXQTL and XQTL studies\n",
      "fixed scXQTL and XQTL studies\n",
      "creating new coloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating new gwasResolvedColoc\n",
      "creating new gwasCredibleAssoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating gwasCredibleAssocDistances\n",
      "creating analysis_chembl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating benchmarkOT function\n",
      "creating dataframes in loop\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'function' object has no attribute 'withColumn'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_15181/658469586.py\u001b[0m in \u001b[0;36m<cell line: 578>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    578\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    579\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"max_L2GScore\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 580\u001b[0;31m         datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n\u001b[0m\u001b[1;32m    581\u001b[0m             \u001b[0mdiscrepancifier\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    582\u001b[0m             \u001b[0mgwasCredibleAssocDistances\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_15181/658469586.py\u001b[0m in \u001b[0;36mbenchmarkOT\u001b[0;34m(gwasCredibleAssocDistances, analysis_chembl, list_l2g, dict_comb)\u001b[0m\n\u001b[1;32m    410\u001b[0m             \u001b[0mgwasCredibleAssocDistances\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    411\u001b[0m             \u001b[0;31m# .filter(F.col(\"h4\").isNotNull()) #### not filter by this because we want to include the L2G AND Coloc question\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 412\u001b[0;31m             .withColumn(  ### take maximum L2G score per T-D\n\u001b[0m\u001b[1;32m    413\u001b[0m                 \u001b[0;34m\"max_L2GScore\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    414\u001b[0m                 F.max(\"resourceScore\").over(\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'function' object has no attribute 'withColumn'"
     ]
    }
   ],
   "source": [
    "#### BUILDING THE NEW GWAS GENETIC EVIDENCE FROM COLOC\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date, datetime\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    "    IntegerType,\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "credibleEvidence = spark.read.parquet(f\"{path}evidence\").filter(\n",
    "    F.col(\"datasourceId\").isin([\"gwas_credible_sets\"])\n",
    ")\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "print(\"read spark files\")\n",
    "\n",
    "print(\"fixing scXQTL and XQTL studies\")\n",
    "#### Fixing scXQTL as XQTLs:\n",
    "## code provided by @ireneisdoomed\n",
    "pd.DataFrame.iteritems = pd.DataFrame.items\n",
    "\n",
    "raw_studies_metadata_schema: StructType = StructType(\n",
    "    [\n",
    "        StructField(\"study_id\", StringType(), True),\n",
    "        StructField(\"dataset_id\", StringType(), True),\n",
    "        StructField(\"study_label\", StringType(), True),\n",
    "        StructField(\"sample_group\", StringType(), True),\n",
    "        StructField(\"tissue_id\", StringType(), True),\n",
    "        StructField(\"tissue_label\", StringType(), True),\n",
    "        StructField(\"condition_label\", StringType(), True),\n",
    "        StructField(\"sample_size\", IntegerType(), True),\n",
    "        StructField(\"quant_method\", StringType(), True),\n",
    "        StructField(\"pmid\", StringType(), True),\n",
    "        StructField(\"study_type\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "    pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "    schema=raw_studies_metadata_schema,\n",
    ")\n",
    "\n",
    "# index = spark.read.parquet(\"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/genetics/parquet/study_index\")\n",
    "\n",
    "study_index_w_correct_type = (\n",
    "    study_table.select(\n",
    "        F.concat_ws(\n",
    "            \"_\",\n",
    "            F.col(\"study_label\"),\n",
    "            F.col(\"quant_method\"),\n",
    "            F.col(\"sample_group\"),\n",
    "        ).alias(\"extracted_column\"),\n",
    "        \"study_type\",\n",
    "    )\n",
    "    .join(\n",
    "        index\n",
    "        # Get eQTL Catalogue studies\n",
    "        .filter(F.col(\"studyType\") != \"gwas\").filter(\n",
    "            ~F.col(\"studyId\").startswith(\"UKB_PPP\")\n",
    "        )\n",
    "        # Remove measured trait\n",
    "        .withColumn(\n",
    "            \"extracted_column\",\n",
    "            F.regexp_replace(F.col(\"studyId\"), r\"(_ENS.*|_ILMN.*|_X.*|_[0-9]+:.*)\", \"\"),\n",
    "        ).withColumn(\n",
    "            \"extracted_column\",\n",
    "            # After the previous cleanup, there are multiple traits from the same publication starting with the gene symbol that need to be removed (e.g. `Sun_2018_aptamer_plasma_ANXA2.4961.17.1..1`)\n",
    "            F.when(\n",
    "                F.col(\"extracted_column\").startswith(\"Sun_2018_aptamer_plasma\"),\n",
    "                F.lit(\"Sun_2018_aptamer_plasma\"),\n",
    "            ).otherwise(F.col(\"extracted_column\")),\n",
    "        ),\n",
    "        on=\"extracted_column\",\n",
    "        how=\"right\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "fixed = (\n",
    "    study_index_w_correct_type.withColumn(\n",
    "        \"toFix\",\n",
    "        F.when(\n",
    "            (F.col(\"study_type\") != \"single-cell\")\n",
    "            & (F.col(\"studyType\").startswith(\"sc\")),\n",
    "            F.lit(True),\n",
    "        ).otherwise(F.lit(False)),\n",
    "    )\n",
    "    # Remove the substring \"sc\" from the studyType column\n",
    "    .withColumn(\n",
    "        \"newStudyType\",\n",
    "        F.when(\n",
    "            F.col(\"toFix\"), F.regexp_replace(F.col(\"studyType\"), r\"sc\", \"\")\n",
    "        ).otherwise(F.col(\"studyType\")),\n",
    "    ).drop(\"toFix\", \"extracted_column\", \"study_type\")\n",
    ").persist()\n",
    "all_studies = index.join(\n",
    "    fixed.selectExpr(\"studyId\", \"newStudyType\"), on=\"studyId\", how=\"left\"\n",
    ").persist()\n",
    "fixedIndex = all_studies.withColumn(\n",
    "    \"studyType\",\n",
    "    F.when(F.col(\"newStudyType\").isNotNull(), F.col(\"newStudyType\")).otherwise(\n",
    "        F.col(\"studyType\")\n",
    "    ),\n",
    ").drop(\"newStudyType\")\n",
    "\n",
    "print(\"fixed scXQTL and XQTL studies\")\n",
    "\n",
    "print(\"creating new coloc\")\n",
    "\n",
    "#### fixed\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        fixedIndex.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "# remove columns without content (only null values on them)\n",
    "df = credibleEvidence.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns).persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")\n",
    "print(\"creating new gwasResolvedColoc\")\n",
    "\n",
    "### bring directionality from QTL\n",
    "\n",
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "\n",
    "print(\"creating new gwasCredibleAssoc\")\n",
    "\n",
    "### modify to include more information\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )\n",
    "    .select(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"homogenized\",\n",
    "        \"leftStudyLocusId\",\n",
    "        \"h4\",\n",
    "        \"datasourceId\",\n",
    "        \"resourceScore\",\n",
    "        \"leftVariantId\",\n",
    "        \"credibleLeftStudyType\",\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")  ### there will be duplicates TargetId-DiseaseId because we are taking the most significant DoE\n",
    "\n",
    "#### LOAD STUDYLOCUSID AND VARIANT DISTANCES\n",
    "l2gPred = spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/locusToGenePredictions\"\n",
    ")\n",
    "l2gTable = (\n",
    "    l2gPred.select(\"studyLocusId\", \"geneId\", F.explode_outer(\"locusToGeneFeatures\"))\n",
    "    .filter(F.col(\"key\").isin([\"distanceFootprintMean\", \"distanceTssMean\"]))\n",
    "    .groupBy(\"studyLocusId\", \"geneId\")\n",
    "    .pivot(\"key\")\n",
    "    .agg(F.first(\"value\"))\n",
    ")\n",
    "print(\"creating gwasCredibleAssocDistances\")\n",
    "gwasCredibleAssocDistances = gwasCredibleAssoc.join(\n",
    "    l2gTable.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\").withColumnRenamed(\n",
    "        \"geneId\", \"targetId\"\n",
    "    ),\n",
    "    on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "print(\"creating analysis_chembl\")\n",
    "analysis_chembl = (\n",
    "    discrepancifier(\n",
    "        directionOfEffect(\n",
    "            evidences.filter((F.col(\"datasourceId\") == \"chembl\")), \"24.09\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "    .filter(  ### ensure drug has annotated MoA and is coherent per Target-Disease\n",
    "        ((F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull())\n",
    "        & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    )\n",
    "    .selectExpr(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"maxClinPhase\",\n",
    "        \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "        \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "        \"LoF_protect as LoF_protect_ch\",\n",
    "        \"GoF_protect as GoF_protect_ch\",\n",
    "    )\n",
    ")\n",
    "\n",
    "### pivot colocdoE grouping by T-D-studyLocusId-distances\n",
    "\n",
    "values = [\"max_L2GScore\", \"min_footPrintDistance_rank\", \"min_tssDistance_rank\"]\n",
    "value = \"max_L2GScore\"\n",
    "dict_comb = {\n",
    "    \"hasGeneticEvidence\": f\"{value}\",\n",
    "    \"diagonalYes\": f\"{value}\",\n",
    "    \"oneCellYes\": f\"{value}\",\n",
    "    \"L2GAndColoc\": f\"{value}\",\n",
    "}\n",
    "list_l2g = [\n",
    "    0.10,\n",
    "    0.15,\n",
    "    0.20,\n",
    "    0.25,\n",
    "    0.30,\n",
    "    0.35,\n",
    "    0.40,\n",
    "    0.45,\n",
    "    0.50,\n",
    "    0.55,\n",
    "    0.60,\n",
    "    0.65,\n",
    "    0.70,\n",
    "    0.75,\n",
    "    0.80,\n",
    "    0.85,\n",
    "    0.90,\n",
    "    0.95,\n",
    "]\n",
    "print(\"creating benchmarkOT function\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def benchmarkOT(\n",
    "    gwasCredibleAssocDistances, value, analysis_chembl, list_l2g, dict_comb\n",
    "):\n",
    "    return (\n",
    "        discrepancifier(\n",
    "            gwasCredibleAssocDistances\n",
    "            # .filter(F.col(\"h4\").isNotNull()) #### not filter by this because we want to include the L2G AND Coloc question\n",
    "            .withColumn(  ### take maximum L2G score per T-D\n",
    "                \"max_L2GScore\",\n",
    "                F.max(\"resourceScore\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(\n",
    "                \"min_footPrintDistance_rank\",\n",
    "                F.min(\"distanceFootprintMean\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(\n",
    "                \"min_tssDistance_rank\",\n",
    "                F.min(\"distanceTssMean\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .groupBy(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                f\"{value}\",\n",
    "                # \"leftStudyLocusId\",\n",
    "            )\n",
    "            .pivot(\"homogenized\")\n",
    "            .count()\n",
    "        )\n",
    "        .join(analysis_chembl, on=[\"targetId\", \"diseaseId\"], how=\"right\")\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (\n",
    "                    (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                    & (F.col(\"coherencyDiagonal\").isNotNull())\n",
    "                )\n",
    "                # & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "                ,\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (\n",
    "                    (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                    & (F.col(\"coherencyDiagonal\").isNotNull())\n",
    "                )\n",
    "                # & (F.col(\"coherencyOneCell\") == \"coherent\")\n",
    "                ,\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "            # ).filter(\n",
    "            #    F.col(\"diagonalAgreeWithDrugs\").isNotNull()\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(F.col(\"max_L2GScore\").isNotNull(), F.lit(\"yes\")).otherwise(\n",
    "                F.lit(\"no\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"L2GAndColoc\",\n",
    "            F.when(\n",
    "                (F.col(f\"{value}\").isNotNull())\n",
    "                & (F.col(\"coherencyDiagonal\").isin([\"coherent\", \"dispar\"])),\n",
    "                F.lit(\"yes\"),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .select(\n",
    "            [\"*\"]\n",
    "            + (\n",
    "                [  ### single columns\n",
    "                    F.when(F.col(f\"{value}\") >= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{value}>={str(n).replace('.', '_')}\")\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "            )\n",
    "            + (\n",
    "                [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "### HERE\n",
    "\n",
    "### Make all the datasets\n",
    "values = [\"max_L2GScore\", \"min_footPrintDistance_rank\", \"min_tssDistance_rank\"]\n",
    "\n",
    "print(\"creating dataframes in loop\")\n",
    "\n",
    "datasetDict = {}\n",
    "for value in values:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            discrepancifier,\n",
    "            gwasCredibleAssocDistances,\n",
    "            value,\n",
    "            analysis_chembl,\n",
    "        )\n",
    "    else:\n",
    "        datasetDict[f\"{value}\"] = benchmarkOT(\n",
    "            discrepancifier, gwasCredibleAssocDistances, value, analysis_chembl\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-01-27 22:16:36.625418\n",
      "Analysis started on 2025-01-27 at  2025-01-27 22:16:36.625418\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/27 22:16:41 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n",
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read spark files\n",
      "fixing scXQTL and XQTL studies\n",
      "fixed scXQTL and XQTL studies\n",
      "creating new coloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating new gwasResolvedColoc\n",
      "creating new gwasCredibleAssoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating gwasCredibleAssocDistances\n",
      "creating analysis_chembl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating benchmarkOT function\n",
      "creating dataframes in loop\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "launched function to run analysis\n",
      "corresponding dataframe key:  df_l2g_original\n",
      "df_l2g_original min_tssDistance_rank\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/27 22:21:03 WARN CacheManager: Asked to cache already cached data.\n",
      "                                                                                \r"
     ]
    },
    {
     "ename": "AnalysisException",
     "evalue": "Column 'clinicalStatus' does not exist. Did you mean one of the following? [diagonalYes, noEvaluable, oneCellYes, diseaseId, predictionType, targetId, total, GoF_protect, L2GAndColoc, LoF_protect, Phase0, Phase4, Phase>=1, Phase>=2, Phase>=3, comparisonType, maxClinPhase, max_L2GScore, GoF_protect_ch, GoF_risk, LoF_protect_ch, LoF_risk, coherencyOneCell, coherencyDiagonal, hasGeneticEvidence, coherencyDiagonal_ch, max_L2GScore>=0_1, max_L2GScore>=0_2, max_L2GScore>=0_3, max_L2GScore>=0_4, max_L2GScore>=0_5, max_L2GScore>=0_6, max_L2GScore>=0_7, max_L2GScore>=0_8, max_L2GScore>=0_9, coherencyOneCell_ch, diagonalAgreeWithDrugs, max_L2GScore>=0_15, max_L2GScore>=0_25, max_L2GScore>=0_35, max_L2GScore>=0_45, max_L2GScore>=0_55, max_L2GScore>=0_65, max_L2GScore>=0_75, max_L2GScore>=0_85, max_L2GScore>=0_95, oneCellAgreeWithDrugs, max_L2GScore>=0_1&diagonalYes_combined, max_L2GScore>=0_1&oneCellYes_combined, max_L2GScore>=0_2&diagonalYes_combined, max_L2GScore>=0_2&oneCellYes_combined, max_L2GScore>=0_3&diagonalYes_combined, max_L2GScore>=0_3&oneCellYes_combined, max_L2GScore>=0_4&diagonalYes_combined, max_L2GScore>=0_4&oneCellYes_combined, max_L2GScore>=0_5&diagonalYes_combined, max_L2GScore>=0_5&oneCellYes_combined, max_L2GScore>=0_6&diagonalYes_combined, max_L2GScore>=0_6&oneCellYes_combined, max_L2GScore>=0_7&diagonalYes_combined, max_L2GScore>=0_7&oneCellYes_combined, max_L2GScore>=0_8&diagonalYes_combined, max_L2GScore>=0_8&oneCellYes_combined, max_L2GScore>=0_9&diagonalYes_combined, max_L2GScore>=0_9&oneCellYes_combined, max_L2GScore>=0_1&L2GAndColoc_combined, max_L2GScore>=0_15&diagonalYes_combined, max_L2GScore>=0_15&oneCellYes_combined, max_L2GScore>=0_2&L2GAndColoc_combined, max_L2GScore>=0_25&diagonalYes_combined, max_L2GScore>=0_25&oneCellYes_combined, max_L2GScore>=0_3&L2GAndColoc_combined, max_L2GScore>=0_35&diagonalYes_combined, max_L2GScore>=0_35&oneCellYes_combined, max_L2GScore>=0_4&L2GAndColoc_combined, max_L2GScore>=0_45&diagonalYes_combined, max_L2GScore>=0_45&oneCellYes_combined, max_L2GScore>=0_5&L2GAndColoc_combined, max_L2GScore>=0_55&diagonalYes_combined, max_L2GScore>=0_55&oneCellYes_combined, max_L2GScore>=0_6&L2GAndColoc_combined, max_L2GScore>=0_65&diagonalYes_combined, max_L2GScore>=0_65&oneCellYes_combined, max_L2GScore>=0_7&L2GAndColoc_combined, max_L2GScore>=0_75&diagonalYes_combined, max_L2GScore>=0_75&oneCellYes_combined, max_L2GScore>=0_8&L2GAndColoc_combined, max_L2GScore>=0_85&diagonalYes_combined, max_L2GScore>=0_85&oneCellYes_combined, max_L2GScore>=0_9&L2GAndColoc_combined, max_L2GScore>=0_95&diagonalYes_combined, max_L2GScore>=0_95&oneCellYes_combined, max_L2GScore>=0_15&L2GAndColoc_combined, max_L2GScore>=0_25&L2GAndColoc_combined, max_L2GScore>=0_35&L2GAndColoc_combined, max_L2GScore>=0_45&L2GAndColoc_combined, max_L2GScore>=0_55&L2GAndColoc_combined, max_L2GScore>=0_65&L2GAndColoc_combined, max_L2GScore>=0_75&L2GAndColoc_combined, max_L2GScore>=0_85&L2GAndColoc_combined, max_L2GScore>=0_95&L2GAndColoc_combined, max_L2GScore>=0_1&hasGeneticEvidence_combined, max_L2GScore>=0_2&hasGeneticEvidence_combined, max_L2GScore>=0_3&hasGeneticEvidence_combined, max_L2GScore>=0_4&hasGeneticEvidence_combined, max_L2GScore>=0_5&hasGeneticEvidence_combined, max_L2GScore>=0_6&hasGeneticEvidence_combined, max_L2GScore>=0_7&hasGeneticEvidence_combined, max_L2GScore>=0_8&hasGeneticEvidence_combined, max_L2GScore>=0_9&hasGeneticEvidence_combined, max_L2GScore>=0_15&hasGeneticEvidence_combined, max_L2GScore>=0_25&hasGeneticEvidence_combined, max_L2GScore>=0_35&hasGeneticEvidence_combined, max_L2GScore>=0_45&hasGeneticEvidence_combined, max_L2GScore>=0_55&hasGeneticEvidence_combined, max_L2GScore>=0_65&hasGeneticEvidence_combined, max_L2GScore>=0_75&hasGeneticEvidence_combined, max_L2GScore>=0_85&hasGeneticEvidence_combined, max_L2GScore>=0_95&hasGeneticEvidence_combined];\n'Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 96 more fields]\n+- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 95 more fields]\n   +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 94 more fields]\n      +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 93 more fields]\n         +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 92 more fields]\n            +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 2 more fields]\n               +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, CASE WHEN isnotnull(max_L2GScore#13786) THEN yes ELSE no END AS hasGeneticEvidence#15733]\n                  +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, CASE WHEN (oneCellAgreeWithDrugs#15561 = coherent) THEN yes ELSE no END AS oneCellYes#15708]\n                     +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, CASE WHEN (diagonalAgreeWithDrugs#15544 = coherent) THEN yes ELSE no END AS diagonalYes#15684]\n                        +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, CASE WHEN (maxClinPhase#8314 = cast(0 as double)) THEN yes ELSE no END AS Phase0#15661]\n                           +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, CASE WHEN (maxClinPhase#8314 >= cast(1 as double)) THEN yes ELSE no END AS Phase>=1#15639]\n                              +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, CASE WHEN (maxClinPhase#8314 >= cast(2 as double)) THEN yes ELSE no END AS Phase>=2#15618]\n                                 +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, CASE WHEN (maxClinPhase#8314 >= cast(3 as double)) THEN yes ELSE no END AS Phase>=3#15598]\n                                    +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, CASE WHEN (maxClinPhase#8314 = cast(4 as double)) THEN yes ELSE no END AS Phase4#15579]\n                                       +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, CASE WHEN ((coherencyOneCell_ch#13775 = coherent) AND isnotnull(coherencyDiagonal#15508)) THEN CASE WHEN (isnotnull(LoF_protect_ch#13776L) AND (((isnotnull(LoF_protect#15487L) AND isnull(LoF_risk#15488L)) AND isnull(GoF_protect#15485L)) AND isnull(GoF_risk#15486L))) THEN coherent WHEN (isnotnull(GoF_protect_ch#13777L) AND (((isnotnull(GoF_protect#15485L) AND isnull(LoF_risk#15488L)) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L))) THEN coherent ELSE dispar END END AS oneCellAgreeWithDrugs#15561]\n                                          +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, CASE WHEN ((coherencyDiagonal_ch#13774 = coherent) AND isnotnull(coherencyDiagonal#15508)) THEN CASE WHEN (isnotnull(LoF_protect_ch#13776L) AND (isnotnull(GoF_risk#15486L) OR isnotnull(LoF_protect#15487L))) THEN coherent WHEN (isnotnull(GoF_protect_ch#13777L) AND (isnotnull(LoF_risk#15488L) OR isnotnull(GoF_protect#15485L))) THEN coherent ELSE dispar END END AS diagonalAgreeWithDrugs#15544]\n                                             +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L]\n                                                +- Join RightOuter, ((targetId#933 = targetId#718) AND (diseaseId#4812 = diseaseId#818))\n                                                   :- Project [targetId#933, diseaseId#4812, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, CASE WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnull(noEvaluable#15489L)) THEN noEvid WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnotnull(noEvaluable#15489L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#15488L) OR isnotnull(LoF_protect#15487L)) OR isnotnull(GoF_risk#15486L)) OR isnotnull(GoF_protect#15485L)) THEN CASE WHEN (isnotnull(LoF_risk#15488L) AND ((isnull(LoF_protect#15487L) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L))) THEN coherent WHEN (isnotnull(GoF_risk#15486L) AND ((isnull(LoF_protect#15487L) AND isnull(LoF_risk#15488L)) AND isnull(GoF_protect#15485L))) THEN coherent WHEN (isnotnull(LoF_protect#15487L) AND ((isnull(LoF_risk#15488L) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L))) THEN coherent WHEN (isnotnull(GoF_protect#15485L) AND ((isnull(LoF_protect#15487L) AND isnull(GoF_risk#15486L)) AND isnull(LoF_risk#15488L))) THEN coherent ELSE dispar END END AS coherencyOneCell#15518]\n                                                   :  +- Project [targetId#933, diseaseId#4812, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, CASE WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnull(noEvaluable#15489L)) THEN noEvid WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnotnull(noEvaluable#15489L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#15488L) OR isnotnull(LoF_protect#15487L)) OR isnotnull(GoF_risk#15486L)) OR isnotnull(GoF_protect#15485L)) THEN CASE WHEN (isnotnull(GoF_risk#15486L) AND isnotnull(LoF_risk#15488L)) THEN dispar WHEN (isnotnull(LoF_protect#15487L) AND isnotnull(LoF_risk#15488L)) THEN dispar WHEN (isnotnull(GoF_protect#15485L) AND isnotnull(GoF_risk#15486L)) THEN dispar WHEN (isnotnull(GoF_protect#15485L) AND isnotnull(LoF_protect#15487L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#15508]\n                                                   :     +- Project [targetId#933, diseaseId#4812, max_L2GScore#13786, __pivot_count(1) AS count AS `count(1) AS count`#15484[0] AS GoF_protect#15485L, __pivot_count(1) AS count AS `count(1) AS count`#15484[1] AS GoF_risk#15486L, __pivot_count(1) AS count AS `count(1) AS count`#15484[2] AS LoF_protect#15487L, __pivot_count(1) AS count AS `count(1) AS count`#15484[3] AS LoF_risk#15488L, __pivot_count(1) AS count AS `count(1) AS count`#15484[4] AS noEvaluable#15489L]\n                                                   :        +- Aggregate [targetId#933, diseaseId#4812, max_L2GScore#13786], [targetId#933, diseaseId#4812, max_L2GScore#13786, pivotfirst(homogenized#5856, count(1) AS count#15472L, GoF_protect, GoF_risk, LoF_protect, LoF_risk, noEvaluable, 0, 0) AS __pivot_count(1) AS count AS `count(1) AS count`#15484]\n                                                   :           +- Aggregate [targetId#933, diseaseId#4812, max_L2GScore#13786, homogenized#5856], [targetId#933, diseaseId#4812, max_L2GScore#13786, homogenized#5856, count(1) AS count(1) AS count#15472L]\n                                                   :              +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800, min_tssDistance_rank#13815]\n                                                   :                 +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800, min_tssDistance_rank#13815, min_tssDistance_rank#13815]\n                                                   :                    +- Window [min(distanceTssMean#5898) windowspecdefinition(targetId#933, diseaseId#4812, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS min_tssDistance_rank#13815], [targetId#933, diseaseId#4812]\n                                                   :                       +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800]\n                                                   :                          +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800]\n                                                   :                             +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800, min_footPrintDistance_rank#13800]\n                                                   :                                +- Window [min(distanceFootprintMean#5897) windowspecdefinition(targetId#933, diseaseId#4812, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS min_footPrintDistance_rank#13800], [targetId#933, diseaseId#4812]\n                                                   :                                   +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786]\n                                                   :                                      +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786]\n                                                   :                                         +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, max_L2GScore#13786]\n                                                   :                                            +- Window [max(resourceScore#1002) windowspecdefinition(targetId#933, diseaseId#4812, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS max_L2GScore#13786], [targetId#933, diseaseId#4812]\n                                                   :                                               +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                  +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                     +- Join LeftOuter, ((leftStudyLocusId#4687 = leftStudyLocusId#5907) AND (targetId#933 = targetId#5912))\n                                                   :                                                        :- Project [targetId#933, diseaseId#4812, CASE WHEN isnull(homogenized#5806) THEN noEvaluable ELSE homogenized#5806 END AS homogenized#5856, leftStudyLocusId#4687, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248]\n                                                   :                                                        :  +- Project [targetId#933, diseaseId#4812, homogenized#5806, leftStudyLocusId#4687, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248]\n                                                   :                                                        :     +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 16 more fields]\n                                                   :                                                        :        +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 17 more fields]\n                                                   :                                                        :           +- Window [first(colocDoE#4890, true) windowspecdefinition(targetId#933, diseaseId#4812, pValueExponent#4712 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS homogenized#5806], [targetId#933, diseaseId#4812], [pValueExponent#4712 ASC NULLS FIRST]\n                                                   :                                                        :              +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                 +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                    +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 14 more fields]\n                                                   :                                                        :                       +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                          +- Generate explode(concat(array(diseaseId#1033), parents#694)), true, [diseaseId#4812]\n                                                   :                                                        :                             +- Project [diseaseId#1033, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                                +- Join LeftOuter, (diseaseId#1033 = diseaseId#4766)\n                                                   :                                                        :                                   :- Project [leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, datasourceId#932, ... 12 more fields]\n                                                   :                                                        :                                   :  +- Join RightOuter, ((leftStudyLocusId#1259 = leftStudyLocusId#4687) AND (targetId#4663 = targetId#933))\n                                                   :                                                        :                                   :     :- Project [rightStudyId#3271, rightStudyLocusId#1260, leftStudyLocusId#1259, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, geneId#1200 AS targetId#4663, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228]\n                                                   :                                                        :                                   :     :  +- Filter NOT (rightStudyType#1262 = gwas)\n                                                   :                                                        :                                   :     :     +- Project [rightStudyId#3271, rightStudyLocusId#1260, leftStudyLocusId#1259, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, geneId#1200, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228]\n                                                   :                                                        :                                   :     :        +- Join LeftOuter, (rightStudyId#3271 = rightStudyId#3323)\n                                                   :                                                        :                                   :     :           :- Project [rightStudyLocusId#1260, leftStudyLocusId#1259, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightStudyId#3271, rightVariantId#3272, credibleRightStudyType#3273]\n                                                   :                                                        :                                   :     :           :  +- Join LeftOuter, (rightStudyLocusId#1260 = rightStudyLocusId#3270)\n                                                   :                                                        :                                   :     :           :     :- Project [leftStudyLocusId#1259, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248]\n                                                   :                                                        :                                   :     :           :     :  +- Join LeftOuter, (leftStudyLocusId#1259 = leftStudyLocusId#3245)\n                                                   :                                                        :                                   :     :           :     :     :- Relation [leftStudyLocusId#1259,rightStudyLocusId#1260,chromosome#1261,rightStudyType#1262,numberColocalisingVariants#1263L,h0#1264,h1#1265,h2#1266,h3#1267,h4#1268,colocalisationMethod#1269,betaRatioSignAverage#1270] parquet\n                                                   :                                                        :                                   :     :           :     :     +- Project [studyLocusId#1147 AS leftStudyLocusId#3245, StudyId#1148 AS leftStudyId#3246, variantId#1149 AS leftVariantId#3247, studyType#1172 AS credibleLeftStudyType#3248]\n                                                   :                                                        :                                   :     :           :     :        +- Relation [studyLocusId#1147,studyId#1148,variantId#1149,chromosome#1150,position#1151,region#1152,beta#1153,zScore#1154,pValueMantissa#1155,pValueExponent#1156,effectAlleleFrequencyFromSource#1157,standardError#1158,subStudyDescription#1159,qualityControls#1160,finemappingMethod#1161,credibleSetIndex#1162,credibleSetlog10BF#1163,purityMeanR2#1164,purityMinR2#1165,locusStart#1166,locusEnd#1167,sampleSize#1168,ldSet#1169,locus#1170,... 2 more fields] parquet\n                                                   :                                                        :                                   :     :           :     +- Project [studyLocusId#3278 AS rightStudyLocusId#3270, studyId#3279 AS rightStudyId#3271, variantId#3280 AS rightVariantId#3272, studyType#3303 AS credibleRightStudyType#3273]\n                                                   :                                                        :                                   :     :           :        +- Relation [studyLocusId#3278,studyId#3279,variantId#3280,chromosome#3281,position#3282,region#3283,beta#3284,zScore#3285,pValueMantissa#3286,pValueExponent#3287,effectAlleleFrequencyFromSource#3288,standardError#3289,subStudyDescription#3290,qualityControls#3291,finemappingMethod#3292,credibleSetIndex#3293,credibleSetlog10BF#3294,purityMeanR2#3295,purityMinR2#3296,locusStart#3297,locusEnd#3298,sampleSize#3299,ldSet#3300,locus#3301,... 2 more fields] parquet\n                                                   :                                                        :                                   :     :           +- Project [studyId#1199 AS rightStudyId#3323, geneId#1200, projectId#1201, studyType#3183 AS indexStudyType#3324, condition#1224, biosampleId#1228]\n                                                   :                                                        :                                   :     :              +- Project [studyId#1199, geneId#1200, projectId#1201, studyType#3183, traitFromSource#1203, traitFromSourceMappedIds#1204, biosampleFromSourceId#1205, pubmedId#1206, publicationTitle#1207, publicationFirstAuthor#1208, publicationDate#1209, publicationJournal#1210, backgroundTraitFromSourceMappedIds#1211, initialSampleSize#1212, nCases#1213, nControls#1214, nSamples#1215, cohorts#1216, ldPopulationStructure#1217, discoverySamples#1218, replicationSamples#1219, qualityControls#1220, analysisFlags#1221, summarystatsLocation#1222, ... 6 more fields]\n                                                   :                                                        :                                   :     :                 +- Project [studyId#1199, geneId#1200, projectId#1201, CASE WHEN isnotnull(newStudyType#1643) THEN newStudyType#1643 ELSE studyType#1202 END AS studyType#3183, traitFromSource#1203, traitFromSourceMappedIds#1204, biosampleFromSourceId#1205, pubmedId#1206, publicationTitle#1207, publicationFirstAuthor#1208, publicationDate#1209, publicationJournal#1210, backgroundTraitFromSourceMappedIds#1211, initialSampleSize#1212, nCases#1213, nControls#1214, nSamples#1215, cohorts#1216, ldPopulationStructure#1217, discoverySamples#1218, replicationSamples#1219, qualityControls#1220, analysisFlags#1221, summarystatsLocation#1222, ... 7 more fields]\n                                                   :                                                        :                                   :     :                    +- Project [studyId#1199, geneId#1200, projectId#1201, studyType#1202, traitFromSource#1203, traitFromSourceMappedIds#1204, biosampleFromSourceId#1205, pubmedId#1206, publicationTitle#1207, publicationFirstAuthor#1208, publicationDate#1209, publicationJournal#1210, backgroundTraitFromSourceMappedIds#1211, initialSampleSize#1212, nCases#1213, nControls#1214, nSamples#1215, cohorts#1216, ldPopulationStructure#1217, discoverySamples#1218, replicationSamples#1219, qualityControls#1220, analysisFlags#1221, summarystatsLocation#1222, ... 7 more fields]\n                                                   :                                                        :                                   :     :                       +- Join LeftOuter, (studyId#1199 = studyId#2346)\n                                                   :                                                        :                                   :     :                          :- Relation [studyId#1199,geneId#1200,projectId#1201,studyType#1202,traitFromSource#1203,traitFromSourceMappedIds#1204,biosampleFromSourceId#1205,pubmedId#1206,publicationTitle#1207,publicationFirstAuthor#1208,publicationDate#1209,publicationJournal#1210,backgroundTraitFromSourceMappedIds#1211,initialSampleSize#1212,nCases#1213,nControls#1214,nSamples#1215,cohorts#1216,ldPopulationStructure#1217,discoverySamples#1218,replicationSamples#1219,qualityControls#1220,analysisFlags#1221,summarystatsLocation#1222,... 6 more fields] parquet\n                                                   :                                                        :                                   :     :                          +- Project [studyId#2346, newStudyType#1643]\n                                                   :                                                        :                                   :     :                             +- Project [studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, analysisFlags#2368, summarystatsLocation#2369, ... 7 more fields]\n                                                   :                                                        :                                   :     :                                +- Project [extracted_column#1385, study_type#1337, studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, ... 10 more fields]\n                                                   :                                                        :                                   :     :                                   +- Project [extracted_column#1385, study_type#1337, studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, ... 9 more fields]\n                                                   :                                                        :                                   :     :                                      +- Project [extracted_column#1385, study_type#1337, studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, ... 8 more fields]\n                                                   :                                                        :                                   :     :                                         +- Join RightOuter, (extracted_column#1349 = extracted_column#1385)\n                                                   :                                                        :                                   :     :                                            :- Project [concat_ws(_, study_label#1329, quant_method#1335, sample_group#1330) AS extracted_column#1349, study_type#1337]\n                                                   :                                                        :                                   :     :                                            :  +- LogicalRDD [study_id#1327, dataset_id#1328, study_label#1329, sample_group#1330, tissue_id#1331, tissue_label#1332, condition_label#1333, sample_size#1334, quant_method#1335, pmid#1336, study_type#1337], false\n                                                   :                                                        :                                   :     :                                            +- Project [studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, analysisFlags#2368, summarystatsLocation#2369, ... 7 more fields]\n                                                   :                                                        :                                   :     :                                               +- Project [studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, analysisFlags#2368, summarystatsLocation#2369, ... 7 more fields]\n                                                   :                                                        :                                   :     :                                                  +- Filter NOT StartsWith(studyId#2346, UKB_PPP)\n                                                   :                                                        :                                   :     :                                                     +- Filter NOT (studyType#2349 = gwas)\n                                                   :                                                        :                                   :     :                                                        +- Relation [studyId#2346,geneId#2347,projectId#2348,studyType#2349,traitFromSource#2350,traitFromSourceMappedIds#2351,biosampleFromSourceId#2352,pubmedId#2353,publicationTitle#2354,publicationFirstAuthor#2355,publicationDate#2356,publicationJournal#2357,backgroundTraitFromSourceMappedIds#2358,initialSampleSize#2359,nCases#2360,nControls#2361,nSamples#2362,cohorts#2363,ldPopulationStructure#2364,discoverySamples#2365,replicationSamples#2366,qualityControls#2367,analysisFlags#2368,summarystatsLocation#2369,... 6 more fields] parquet\n                                                   :                                                        :                                   :     +- Project [studyLocusId#1011 AS leftStudyLocusId#4687, datasourceId#932, targetId#933, datatypeId#962, diseaseFromSourceMappedId#966, resourceScore#1002, targetFromSourceId#1018, diseaseId#1033, id#1034, score#1035, sourceId#1038, studyId#4704, variantId#4705, betaGwas#4642, pValueExponent#4712]\n                                                   :                                                        :                                   :        +- Project [studyLocusId#1011, datasourceId#932, targetId#933, datatypeId#962, diseaseFromSourceMappedId#966, resourceScore#1002, targetFromSourceId#1018, diseaseId#1033, id#1034, score#1035, sourceId#1038, studyId#4704, variantId#4705, betaGwas#4642, pValueExponent#4712]\n                                                   :                                                        :                                   :           +- Join LeftOuter, (studyLocusId#1011 = studyLocusId#4703)\n                                                   :                                                        :                                   :              :- Project [datasourceId#932, targetId#933, datatypeId#962, diseaseFromSourceMappedId#966, resourceScore#1002, studyLocusId#1011, targetFromSourceId#1018, diseaseId#1033, id#1034, score#1035, sourceId#1038]\n                                                   :                                                        :                                   :              :  +- Filter (datasourceId#932 = gwas_credible_sets)\n                                                   :                                                        :                                   :              :     +- Filter datasourceId#932 IN (gwas_credible_sets)\n                                                   :                                                        :                                   :              :        +- Relation [datasourceId#932,targetId#933,alleleOrigins#934,allelicRequirements#935,ancestry#936,ancestryId#937,assays#938,assessments#939,beta#940,betaConfidenceIntervalLower#941,betaConfidenceIntervalUpper#942,biologicalModelAllelicComposition#943,biologicalModelGeneticBackground#944,biologicalModelId#945,biomarkerList#946,biomarkerName#947,biomarkers#948,biosamplesFromSource#949,cellLineBackground#950,cellType#951,clinicalPhase#952,clinicalSignificances#953,clinicalStatus#954,cohortDescription#955,... 83 more fields] parquet\n                                                   :                                                        :                                   :              +- Project [studyLocusId#4703, studyId#4704, variantId#4705, beta#4709 AS betaGwas#4642, pValueExponent#4712]\n                                                   :                                                        :                                   :                 +- Relation [studyLocusId#4703,studyId#4704,variantId#4705,chromosome#4706,position#4707,region#4708,beta#4709,zScore#4710,pValueMantissa#4711,pValueExponent#4712,effectAlleleFrequencyFromSource#4713,standardError#4714,subStudyDescription#4715,qualityControls#4716,finemappingMethod#4717,credibleSetIndex#4718,credibleSetlog10BF#4719,purityMeanR2#4720,purityMinR2#4721,locusStart#4722,locusEnd#4723,sampleSize#4724,ldSet#4725,locus#4726,... 2 more fields] parquet\n                                                   :                                                        :                                   +- Project [id#687 AS diseaseId#4766, name#691, parents#694, therapeuticAreas#699]\n                                                   :                                                        :                                      +- Relation [id#687,code#688,dbXRefs#689,description#690,name#691,directLocationIds#692,obsoleteTerms#693,parents#694,synonyms#695,ancestors#696,descendants#697,children#698,therapeuticAreas#699,indirectLocationIds#700,ontology#701] parquet\n                                                   :                                                        +- Project [leftStudyLocusId#5907, geneId#5867 AS targetId#5912, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                           +- Project [studyLocusId#5866 AS leftStudyLocusId#5907, geneId#5867, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                              +- Project [studyLocusId#5866, geneId#5867, __pivot_first(value) AS `first(value)`#5896[0] AS distanceFootprintMean#5897, __pivot_first(value) AS `first(value)`#5896[1] AS distanceTssMean#5898]\n                                                   :                                                                 +- Aggregate [studyLocusId#5866, geneId#5867], [studyLocusId#5866, geneId#5867, pivotfirst(key#5874, first(value)#5890, distanceFootprintMean, distanceTssMean, 0, 0) AS __pivot_first(value) AS `first(value)`#5896]\n                                                   :                                                                    +- Aggregate [studyLocusId#5866, geneId#5867, key#5874], [studyLocusId#5866, geneId#5867, key#5874, first(value#5875, false) AS first(value)#5890]\n                                                   :                                                                       +- Filter key#5874 IN (distanceFootprintMean,distanceTssMean)\n                                                   :                                                                          +- Project [studyLocusId#5866, geneId#5867, key#5874, value#5875]\n                                                   :                                                                             +- Generate explode(locusToGeneFeatures#5869), true, [key#5874, value#5875]\n                                                   :                                                                                +- Relation [studyLocusId#5866,geneId#5867,score#5868,locusToGeneFeatures#5869] parquet\n                                                   +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, coherencyDiagonal#13753 AS coherencyDiagonal_ch#13774, coherencyOneCell#13763 AS coherencyOneCell_ch#13775, LoF_protect#11572L AS LoF_protect_ch#13776L, GoF_protect#11571L AS GoF_protect_ch#13777L]\n                                                      +- Filter ((isnotnull(GoF_protect#11571L) OR isnotnull(LoF_protect#11572L)) AND (coherencyDiagonal#13753 = coherent))\n                                                         +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, GoF_risk#13442, LoF_risk#13587, coherencyDiagonal#13753, CASE WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnull(noEvaluable#11573L)) THEN noEvid WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnotnull(noEvaluable#11573L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#13587) OR isnotnull(LoF_protect#11572L)) OR isnotnull(GoF_risk#13442)) OR isnotnull(GoF_protect#11571L)) THEN CASE WHEN (isnotnull(LoF_risk#13587) AND ((isnull(LoF_protect#11572L) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L))) THEN coherent WHEN (isnotnull(GoF_risk#13442) AND ((isnull(LoF_protect#11572L) AND isnull(LoF_risk#13587)) AND isnull(GoF_protect#11571L))) THEN coherent WHEN (isnotnull(LoF_protect#11572L) AND ((isnull(LoF_risk#13587) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L))) THEN coherent WHEN (isnotnull(GoF_protect#11571L) AND ((isnull(LoF_protect#11572L) AND isnull(GoF_risk#13442)) AND isnull(LoF_risk#13587))) THEN coherent ELSE dispar END END AS coherencyOneCell#13763]\n                                                            +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, GoF_risk#13442, LoF_risk#13587, CASE WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnull(noEvaluable#11573L)) THEN noEvid WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnotnull(noEvaluable#11573L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#13587) OR isnotnull(LoF_protect#11572L)) OR isnotnull(GoF_risk#13442)) OR isnotnull(GoF_protect#11571L)) THEN CASE WHEN (isnotnull(GoF_risk#13442) AND isnotnull(LoF_risk#13587)) THEN dispar WHEN (isnotnull(LoF_protect#11572L) AND isnotnull(LoF_risk#13587)) THEN dispar WHEN (isnotnull(GoF_protect#11571L) AND isnotnull(GoF_risk#13442)) THEN dispar WHEN (isnotnull(GoF_protect#11571L) AND isnotnull(LoF_protect#11572L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#13753]\n                                                               +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, GoF_risk#13442, null AS LoF_risk#13587]\n                                                                  +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, null AS GoF_risk#13442]\n                                                                     +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, __pivot_count(targetId) AS `count(targetId)`#11570[0] AS GoF_protect#11571L, __pivot_count(targetId) AS `count(targetId)`#11570[1] AS LoF_protect#11572L, __pivot_count(targetId) AS `count(targetId)`#11570[2] AS noEvaluable#11573L]\n                                                                        +- Aggregate [targetId#718, diseaseId#818, maxClinPhase#8314], [targetId#718, diseaseId#818, maxClinPhase#8314, pivotfirst(homogenized#7553, count(targetId)#11562L, GoF_protect, LoF_protect, noEvaluable, 0, 0) AS __pivot_count(targetId) AS `count(targetId)`#11570]\n                                                                           +- Aggregate [targetId#718, diseaseId#818, maxClinPhase#8314, homogenized#7553], [targetId#718, diseaseId#818, maxClinPhase#8314, homogenized#7553, count(targetId#718) AS count(targetId)#11562L]\n                                                                              +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 97 more fields]\n                                                                                 +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 98 more fields]\n                                                                                    +- Window [max(clinicalPhase#737) windowspecdefinition(targetId#718, diseaseId#818, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS maxClinPhase#8314], [targetId#718, diseaseId#818]\n                                                                                       +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 96 more fields]\n                                                                                          +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 96 more fields]\n                                                                                             +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 95 more fields]\n                                                                                                +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 95 more fields]\n                                                                                                   +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 95 more fields]\n                                                                                                      +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 96 more fields]\n                                                                                                         +- Window [collect_set(intogen_function#7068, 0, 0) windowspecdefinition(targetId#718, diseaseId#818, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#7191], [targetId#718, diseaseId#818]\n                                                                                                            +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 94 more fields]\n                                                                                                               +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 94 more fields]\n                                                                                                                  +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 93 more fields]\n                                                                                                                     +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 92 more fields]\n                                                                                                                        +- Join LeftOuter, ((drugId2#5999 = drugId#755) AND (targetId2#6006 = targetId#718))\n                                                                                                                           :- Join LeftOuter, (target_id#6049 = targetId#718)\n                                                                                                                           :  :- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, concat_ws(,, clinicalSignificances#738) AS clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 83 more fields]\n                                                                                                                           :  :  +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#738, clinicalStatus#739, cohortDescription#740, ... 83 more fields]\n                                                                                                                           :  :     +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, cast(beta#725 as double) AS beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#738, clinicalStatus#739, cohortDescription#740, ... 83 more fields]\n                                                                                                                           :  :        +- Filter datasourceId#717 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                                                           :  :           +- Filter (datasourceId#717 = chembl)\n                                                                                                                           :  :              +- Filter datasourceId#717 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                                                           :  :                 +- Relation [datasourceId#717,targetId#718,alleleOrigins#719,allelicRequirements#720,ancestry#721,ancestryId#722,assays#723,assessments#724,beta#725,betaConfidenceIntervalLower#726,betaConfidenceIntervalUpper#727,biologicalModelAllelicComposition#728,biologicalModelGeneticBackground#729,biologicalModelId#730,biomarkerList#731,biomarkerName#732,biomarkers#733,biosamplesFromSource#734,cellLineBackground#735,cellType#736,clinicalPhase#737,clinicalSignificances#738,clinicalStatus#739,cohortDescription#740,... 83 more fields] parquet\n                                                                                                                           :  +- Project [id#5928 AS target_id#6049, approvedSymbol#5929, description#6034, description_splited#6038, TSorOncogene#6043]\n                                                                                                                           :     +- Project [id#5928, approvedSymbol#5929, description#6034, description_splited#6038, CASE WHEN (RLIKE(description_splited#6038, ncogene) AND RLIKE(description_splited#6038, TSG)) THEN bivalent WHEN RLIKE(description_splited#6038, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#6038, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#6043]\n                                                                                                                           :        +- Project [id#5928, approvedSymbol#5929, description#6034, concat_ws(,, description#6034) AS description_splited#6038]\n                                                                                                                           :           +- Aggregate [id#5928, approvedSymbol#5929], [id#5928, approvedSymbol#5929, collect_set(description#6026, 0, 0) AS description#6034]\n                                                                                                                           :              +- Filter description#6026 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                                                                                                           :                 +- Project [id#5928, approvedSymbol#5929, col#6021.description AS description#6026]\n                                                                                                                           :                    +- Project [id#5928, approvedSymbol#5929, col#6021]\n                                                                                                                           :                       +- Generate explode(hallmarks#5938.attributes), true, [col#6021]\n                                                                                                                           :                          +- Relation [id#5928,approvedSymbol#5929,biotype#5930,transcriptIds#5931,canonicalTranscript#5932,canonicalExons#5933,genomicLocation#5934,alternativeGenes#5935,approvedName#5936,go#5937,hallmarks#5938,synonyms#5939,symbolSynonyms#5940,nameSynonyms#5941,functionDescriptions#5942,subcellularLocations#5943,targetClass#5944,obsoleteSymbols#5945,obsoleteNames#5946,constraint#5947,tep#5948,proteinIds#5949,dbXrefs#5950,chemicalProbes#5951,... 4 more fields] parquet\n                                                                                                                           +- Aggregate [targetId2#6006, drugId2#5999], [targetId2#6006, drugId2#5999, collect_set(actionType#5984, 0, 0) AS actionType#6016]\n                                                                                                                              +- Project [targetId2#6006, drugId2#5999, actionType#5984, mechanismOfAction#5985]\n                                                                                                                                 +- Generate explode(targets#5989), true, [targetId2#6006]\n                                                                                                                                    +- Project [drugId2#5999, actionType#5984, mechanismOfAction#5985, targets#5989]\n                                                                                                                                       +- Generate explode(chemblIds#5986), true, [drugId2#5999]\n                                                                                                                                          +- Relation [actionType#5984,mechanismOfAction#5985,chemblIds#5986,targetName#5987,targetType#5988,targets#5989,references#5990] parquet\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAnalysisException\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_501043/2337046131.py\u001b[0m in \u001b[0;36m<cell line: 811>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    814\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0maggSetups_original\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    815\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 816\u001b[0;31m         \u001b[0maggregations_original\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_analysis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlistado\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtoday_date\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    817\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    818\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"finished analysis\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_501043/2337046131.py\u001b[0m in \u001b[0;36maggregations_original\u001b[0;34m(df, data, listado, comparisonColumn, comparisonType, predictionColumn, predictionType, today_date)\u001b[0m\n\u001b[1;32m    660\u001b[0m         \u001b[0;34m.\u001b[0m\u001b[0mwithColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"predictionType\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpredictionType\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    661\u001b[0m         \u001b[0;34m.\u001b[0m\u001b[0mwithColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"total\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniqIds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 662\u001b[0;31m         \u001b[0;34m.\u001b[0m\u001b[0mwithColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"a\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mF\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"targetId\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mover\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwPredictionComparison\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    663\u001b[0m         .withColumn(\n\u001b[1;32m    664\u001b[0m             \u001b[0;34m\"predictionTotal\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/dataframe.py\u001b[0m in \u001b[0;36mwithColumn\u001b[0;34m(self, colName, col)\u001b[0m\n\u001b[1;32m   3034\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mColumn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3035\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"col should be Column\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3036\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwithColumn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolName\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparkSession\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3037\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3038\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwithColumnRenamed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexisting\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m\"DataFrame\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.5-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1319\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1321\u001b[0;31m         return_value = get_return_value(\n\u001b[0m\u001b[1;32m   1322\u001b[0m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[1;32m   1323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/spark/python/pyspark/sql/utils.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    194\u001b[0m                 \u001b[0;31m# Hide where the exception came from that shows a non-Pythonic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m                 \u001b[0;31m# JVM exception message.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mconverted\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAnalysisException\u001b[0m: Column 'clinicalStatus' does not exist. Did you mean one of the following? [diagonalYes, noEvaluable, oneCellYes, diseaseId, predictionType, targetId, total, GoF_protect, L2GAndColoc, LoF_protect, Phase0, Phase4, Phase>=1, Phase>=2, Phase>=3, comparisonType, maxClinPhase, max_L2GScore, GoF_protect_ch, GoF_risk, LoF_protect_ch, LoF_risk, coherencyOneCell, coherencyDiagonal, hasGeneticEvidence, coherencyDiagonal_ch, max_L2GScore>=0_1, max_L2GScore>=0_2, max_L2GScore>=0_3, max_L2GScore>=0_4, max_L2GScore>=0_5, max_L2GScore>=0_6, max_L2GScore>=0_7, max_L2GScore>=0_8, max_L2GScore>=0_9, coherencyOneCell_ch, diagonalAgreeWithDrugs, max_L2GScore>=0_15, max_L2GScore>=0_25, max_L2GScore>=0_35, max_L2GScore>=0_45, max_L2GScore>=0_55, max_L2GScore>=0_65, max_L2GScore>=0_75, max_L2GScore>=0_85, max_L2GScore>=0_95, oneCellAgreeWithDrugs, max_L2GScore>=0_1&diagonalYes_combined, max_L2GScore>=0_1&oneCellYes_combined, max_L2GScore>=0_2&diagonalYes_combined, max_L2GScore>=0_2&oneCellYes_combined, max_L2GScore>=0_3&diagonalYes_combined, max_L2GScore>=0_3&oneCellYes_combined, max_L2GScore>=0_4&diagonalYes_combined, max_L2GScore>=0_4&oneCellYes_combined, max_L2GScore>=0_5&diagonalYes_combined, max_L2GScore>=0_5&oneCellYes_combined, max_L2GScore>=0_6&diagonalYes_combined, max_L2GScore>=0_6&oneCellYes_combined, max_L2GScore>=0_7&diagonalYes_combined, max_L2GScore>=0_7&oneCellYes_combined, max_L2GScore>=0_8&diagonalYes_combined, max_L2GScore>=0_8&oneCellYes_combined, max_L2GScore>=0_9&diagonalYes_combined, max_L2GScore>=0_9&oneCellYes_combined, max_L2GScore>=0_1&L2GAndColoc_combined, max_L2GScore>=0_15&diagonalYes_combined, max_L2GScore>=0_15&oneCellYes_combined, max_L2GScore>=0_2&L2GAndColoc_combined, max_L2GScore>=0_25&diagonalYes_combined, max_L2GScore>=0_25&oneCellYes_combined, max_L2GScore>=0_3&L2GAndColoc_combined, max_L2GScore>=0_35&diagonalYes_combined, max_L2GScore>=0_35&oneCellYes_combined, max_L2GScore>=0_4&L2GAndColoc_combined, max_L2GScore>=0_45&diagonalYes_combined, max_L2GScore>=0_45&oneCellYes_combined, max_L2GScore>=0_5&L2GAndColoc_combined, max_L2GScore>=0_55&diagonalYes_combined, max_L2GScore>=0_55&oneCellYes_combined, max_L2GScore>=0_6&L2GAndColoc_combined, max_L2GScore>=0_65&diagonalYes_combined, max_L2GScore>=0_65&oneCellYes_combined, max_L2GScore>=0_7&L2GAndColoc_combined, max_L2GScore>=0_75&diagonalYes_combined, max_L2GScore>=0_75&oneCellYes_combined, max_L2GScore>=0_8&L2GAndColoc_combined, max_L2GScore>=0_85&diagonalYes_combined, max_L2GScore>=0_85&oneCellYes_combined, max_L2GScore>=0_9&L2GAndColoc_combined, max_L2GScore>=0_95&diagonalYes_combined, max_L2GScore>=0_95&oneCellYes_combined, max_L2GScore>=0_15&L2GAndColoc_combined, max_L2GScore>=0_25&L2GAndColoc_combined, max_L2GScore>=0_35&L2GAndColoc_combined, max_L2GScore>=0_45&L2GAndColoc_combined, max_L2GScore>=0_55&L2GAndColoc_combined, max_L2GScore>=0_65&L2GAndColoc_combined, max_L2GScore>=0_75&L2GAndColoc_combined, max_L2GScore>=0_85&L2GAndColoc_combined, max_L2GScore>=0_95&L2GAndColoc_combined, max_L2GScore>=0_1&hasGeneticEvidence_combined, max_L2GScore>=0_2&hasGeneticEvidence_combined, max_L2GScore>=0_3&hasGeneticEvidence_combined, max_L2GScore>=0_4&hasGeneticEvidence_combined, max_L2GScore>=0_5&hasGeneticEvidence_combined, max_L2GScore>=0_6&hasGeneticEvidence_combined, max_L2GScore>=0_7&hasGeneticEvidence_combined, max_L2GScore>=0_8&hasGeneticEvidence_combined, max_L2GScore>=0_9&hasGeneticEvidence_combined, max_L2GScore>=0_15&hasGeneticEvidence_combined, max_L2GScore>=0_25&hasGeneticEvidence_combined, max_L2GScore>=0_35&hasGeneticEvidence_combined, max_L2GScore>=0_45&hasGeneticEvidence_combined, max_L2GScore>=0_55&hasGeneticEvidence_combined, max_L2GScore>=0_65&hasGeneticEvidence_combined, max_L2GScore>=0_75&hasGeneticEvidence_combined, max_L2GScore>=0_85&hasGeneticEvidence_combined, max_L2GScore>=0_95&hasGeneticEvidence_combined];\n'Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 96 more fields]\n+- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 95 more fields]\n   +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 94 more fields]\n      +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 93 more fields]\n         +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 92 more fields]\n            +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, ... 2 more fields]\n               +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, oneCellYes#15708, CASE WHEN isnotnull(max_L2GScore#13786) THEN yes ELSE no END AS hasGeneticEvidence#15733]\n                  +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, diagonalYes#15684, CASE WHEN (oneCellAgreeWithDrugs#15561 = coherent) THEN yes ELSE no END AS oneCellYes#15708]\n                     +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, Phase0#15661, CASE WHEN (diagonalAgreeWithDrugs#15544 = coherent) THEN yes ELSE no END AS diagonalYes#15684]\n                        +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, Phase>=1#15639, CASE WHEN (maxClinPhase#8314 = cast(0 as double)) THEN yes ELSE no END AS Phase0#15661]\n                           +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, Phase>=2#15618, CASE WHEN (maxClinPhase#8314 >= cast(1 as double)) THEN yes ELSE no END AS Phase>=1#15639]\n                              +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, Phase>=3#15598, CASE WHEN (maxClinPhase#8314 >= cast(2 as double)) THEN yes ELSE no END AS Phase>=2#15618]\n                                 +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, Phase4#15579, CASE WHEN (maxClinPhase#8314 >= cast(3 as double)) THEN yes ELSE no END AS Phase>=3#15598]\n                                    +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, oneCellAgreeWithDrugs#15561, CASE WHEN (maxClinPhase#8314 = cast(4 as double)) THEN yes ELSE no END AS Phase4#15579]\n                                       +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, diagonalAgreeWithDrugs#15544, CASE WHEN ((coherencyOneCell_ch#13775 = coherent) AND isnotnull(coherencyDiagonal#15508)) THEN CASE WHEN (isnotnull(LoF_protect_ch#13776L) AND (((isnotnull(LoF_protect#15487L) AND isnull(LoF_risk#15488L)) AND isnull(GoF_protect#15485L)) AND isnull(GoF_risk#15486L))) THEN coherent WHEN (isnotnull(GoF_protect_ch#13777L) AND (((isnotnull(GoF_protect#15485L) AND isnull(LoF_risk#15488L)) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L))) THEN coherent ELSE dispar END END AS oneCellAgreeWithDrugs#15561]\n                                          +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L, CASE WHEN ((coherencyDiagonal_ch#13774 = coherent) AND isnotnull(coherencyDiagonal#15508)) THEN CASE WHEN (isnotnull(LoF_protect_ch#13776L) AND (isnotnull(GoF_risk#15486L) OR isnotnull(LoF_protect#15487L))) THEN coherent WHEN (isnotnull(GoF_protect_ch#13777L) AND (isnotnull(LoF_risk#15488L) OR isnotnull(GoF_protect#15485L))) THEN coherent ELSE dispar END END AS diagonalAgreeWithDrugs#15544]\n                                             +- Project [targetId#718, diseaseId#818, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, coherencyOneCell#15518, maxClinPhase#8314, coherencyDiagonal_ch#13774, coherencyOneCell_ch#13775, LoF_protect_ch#13776L, GoF_protect_ch#13777L]\n                                                +- Join RightOuter, ((targetId#933 = targetId#718) AND (diseaseId#4812 = diseaseId#818))\n                                                   :- Project [targetId#933, diseaseId#4812, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, coherencyDiagonal#15508, CASE WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnull(noEvaluable#15489L)) THEN noEvid WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnotnull(noEvaluable#15489L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#15488L) OR isnotnull(LoF_protect#15487L)) OR isnotnull(GoF_risk#15486L)) OR isnotnull(GoF_protect#15485L)) THEN CASE WHEN (isnotnull(LoF_risk#15488L) AND ((isnull(LoF_protect#15487L) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L))) THEN coherent WHEN (isnotnull(GoF_risk#15486L) AND ((isnull(LoF_protect#15487L) AND isnull(LoF_risk#15488L)) AND isnull(GoF_protect#15485L))) THEN coherent WHEN (isnotnull(LoF_protect#15487L) AND ((isnull(LoF_risk#15488L) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L))) THEN coherent WHEN (isnotnull(GoF_protect#15485L) AND ((isnull(LoF_protect#15487L) AND isnull(GoF_risk#15486L)) AND isnull(LoF_risk#15488L))) THEN coherent ELSE dispar END END AS coherencyOneCell#15518]\n                                                   :  +- Project [targetId#933, diseaseId#4812, max_L2GScore#13786, GoF_protect#15485L, GoF_risk#15486L, LoF_protect#15487L, LoF_risk#15488L, noEvaluable#15489L, CASE WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnull(noEvaluable#15489L)) THEN noEvid WHEN ((((isnull(LoF_risk#15488L) AND isnull(LoF_protect#15487L)) AND isnull(GoF_risk#15486L)) AND isnull(GoF_protect#15485L)) AND isnotnull(noEvaluable#15489L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#15488L) OR isnotnull(LoF_protect#15487L)) OR isnotnull(GoF_risk#15486L)) OR isnotnull(GoF_protect#15485L)) THEN CASE WHEN (isnotnull(GoF_risk#15486L) AND isnotnull(LoF_risk#15488L)) THEN dispar WHEN (isnotnull(LoF_protect#15487L) AND isnotnull(LoF_risk#15488L)) THEN dispar WHEN (isnotnull(GoF_protect#15485L) AND isnotnull(GoF_risk#15486L)) THEN dispar WHEN (isnotnull(GoF_protect#15485L) AND isnotnull(LoF_protect#15487L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#15508]\n                                                   :     +- Project [targetId#933, diseaseId#4812, max_L2GScore#13786, __pivot_count(1) AS count AS `count(1) AS count`#15484[0] AS GoF_protect#15485L, __pivot_count(1) AS count AS `count(1) AS count`#15484[1] AS GoF_risk#15486L, __pivot_count(1) AS count AS `count(1) AS count`#15484[2] AS LoF_protect#15487L, __pivot_count(1) AS count AS `count(1) AS count`#15484[3] AS LoF_risk#15488L, __pivot_count(1) AS count AS `count(1) AS count`#15484[4] AS noEvaluable#15489L]\n                                                   :        +- Aggregate [targetId#933, diseaseId#4812, max_L2GScore#13786], [targetId#933, diseaseId#4812, max_L2GScore#13786, pivotfirst(homogenized#5856, count(1) AS count#15472L, GoF_protect, GoF_risk, LoF_protect, LoF_risk, noEvaluable, 0, 0) AS __pivot_count(1) AS count AS `count(1) AS count`#15484]\n                                                   :           +- Aggregate [targetId#933, diseaseId#4812, max_L2GScore#13786, homogenized#5856], [targetId#933, diseaseId#4812, max_L2GScore#13786, homogenized#5856, count(1) AS count(1) AS count#15472L]\n                                                   :              +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800, min_tssDistance_rank#13815]\n                                                   :                 +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800, min_tssDistance_rank#13815, min_tssDistance_rank#13815]\n                                                   :                    +- Window [min(distanceTssMean#5898) windowspecdefinition(targetId#933, diseaseId#4812, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS min_tssDistance_rank#13815], [targetId#933, diseaseId#4812]\n                                                   :                       +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800]\n                                                   :                          +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800]\n                                                   :                             +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, min_footPrintDistance_rank#13800, min_footPrintDistance_rank#13800]\n                                                   :                                +- Window [min(distanceFootprintMean#5897) windowspecdefinition(targetId#933, diseaseId#4812, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS min_footPrintDistance_rank#13800], [targetId#933, diseaseId#4812]\n                                                   :                                   +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786]\n                                                   :                                      +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786]\n                                                   :                                         +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898, max_L2GScore#13786, max_L2GScore#13786]\n                                                   :                                            +- Window [max(resourceScore#1002) windowspecdefinition(targetId#933, diseaseId#4812, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS max_L2GScore#13786], [targetId#933, diseaseId#4812]\n                                                   :                                               +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                  +- Project [leftStudyLocusId#4687, targetId#933, diseaseId#4812, homogenized#5856, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                     +- Join LeftOuter, ((leftStudyLocusId#4687 = leftStudyLocusId#5907) AND (targetId#933 = targetId#5912))\n                                                   :                                                        :- Project [targetId#933, diseaseId#4812, CASE WHEN isnull(homogenized#5806) THEN noEvaluable ELSE homogenized#5806 END AS homogenized#5856, leftStudyLocusId#4687, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248]\n                                                   :                                                        :  +- Project [targetId#933, diseaseId#4812, homogenized#5806, leftStudyLocusId#4687, h4#1268, datasourceId#932, resourceScore#1002, leftVariantId#3247, credibleLeftStudyType#3248]\n                                                   :                                                        :     +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 16 more fields]\n                                                   :                                                        :        +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 17 more fields]\n                                                   :                                                        :           +- Window [first(colocDoE#4890, true) windowspecdefinition(targetId#933, diseaseId#4812, pValueExponent#4712 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS homogenized#5806], [targetId#933, diseaseId#4812], [pValueExponent#4712 ASC NULLS FIRST]\n                                                   :                                                        :              +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                 +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                    +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 14 more fields]\n                                                   :                                                        :                       +- Project [diseaseId#4812, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                          +- Generate explode(concat(array(diseaseId#1033), parents#694)), true, [diseaseId#4812]\n                                                   :                                                        :                             +- Project [diseaseId#1033, leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, ... 15 more fields]\n                                                   :                                                        :                                +- Join LeftOuter, (diseaseId#1033 = diseaseId#4766)\n                                                   :                                                        :                                   :- Project [leftStudyLocusId#4687, targetId#933, rightStudyId#3271, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228, datasourceId#932, ... 12 more fields]\n                                                   :                                                        :                                   :  +- Join RightOuter, ((leftStudyLocusId#1259 = leftStudyLocusId#4687) AND (targetId#4663 = targetId#933))\n                                                   :                                                        :                                   :     :- Project [rightStudyId#3271, rightStudyLocusId#1260, leftStudyLocusId#1259, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, geneId#1200 AS targetId#4663, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228]\n                                                   :                                                        :                                   :     :  +- Filter NOT (rightStudyType#1262 = gwas)\n                                                   :                                                        :                                   :     :     +- Project [rightStudyId#3271, rightStudyLocusId#1260, leftStudyLocusId#1259, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightVariantId#3272, credibleRightStudyType#3273, geneId#1200, projectId#1201, indexStudyType#3324, condition#1224, biosampleId#1228]\n                                                   :                                                        :                                   :     :        +- Join LeftOuter, (rightStudyId#3271 = rightStudyId#3323)\n                                                   :                                                        :                                   :     :           :- Project [rightStudyLocusId#1260, leftStudyLocusId#1259, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248, rightStudyId#3271, rightVariantId#3272, credibleRightStudyType#3273]\n                                                   :                                                        :                                   :     :           :  +- Join LeftOuter, (rightStudyLocusId#1260 = rightStudyLocusId#3270)\n                                                   :                                                        :                                   :     :           :     :- Project [leftStudyLocusId#1259, rightStudyLocusId#1260, chromosome#1261, rightStudyType#1262, numberColocalisingVariants#1263L, h0#1264, h1#1265, h2#1266, h3#1267, h4#1268, colocalisationMethod#1269, betaRatioSignAverage#1270, leftStudyId#3246, leftVariantId#3247, credibleLeftStudyType#3248]\n                                                   :                                                        :                                   :     :           :     :  +- Join LeftOuter, (leftStudyLocusId#1259 = leftStudyLocusId#3245)\n                                                   :                                                        :                                   :     :           :     :     :- Relation [leftStudyLocusId#1259,rightStudyLocusId#1260,chromosome#1261,rightStudyType#1262,numberColocalisingVariants#1263L,h0#1264,h1#1265,h2#1266,h3#1267,h4#1268,colocalisationMethod#1269,betaRatioSignAverage#1270] parquet\n                                                   :                                                        :                                   :     :           :     :     +- Project [studyLocusId#1147 AS leftStudyLocusId#3245, StudyId#1148 AS leftStudyId#3246, variantId#1149 AS leftVariantId#3247, studyType#1172 AS credibleLeftStudyType#3248]\n                                                   :                                                        :                                   :     :           :     :        +- Relation [studyLocusId#1147,studyId#1148,variantId#1149,chromosome#1150,position#1151,region#1152,beta#1153,zScore#1154,pValueMantissa#1155,pValueExponent#1156,effectAlleleFrequencyFromSource#1157,standardError#1158,subStudyDescription#1159,qualityControls#1160,finemappingMethod#1161,credibleSetIndex#1162,credibleSetlog10BF#1163,purityMeanR2#1164,purityMinR2#1165,locusStart#1166,locusEnd#1167,sampleSize#1168,ldSet#1169,locus#1170,... 2 more fields] parquet\n                                                   :                                                        :                                   :     :           :     +- Project [studyLocusId#3278 AS rightStudyLocusId#3270, studyId#3279 AS rightStudyId#3271, variantId#3280 AS rightVariantId#3272, studyType#3303 AS credibleRightStudyType#3273]\n                                                   :                                                        :                                   :     :           :        +- Relation [studyLocusId#3278,studyId#3279,variantId#3280,chromosome#3281,position#3282,region#3283,beta#3284,zScore#3285,pValueMantissa#3286,pValueExponent#3287,effectAlleleFrequencyFromSource#3288,standardError#3289,subStudyDescription#3290,qualityControls#3291,finemappingMethod#3292,credibleSetIndex#3293,credibleSetlog10BF#3294,purityMeanR2#3295,purityMinR2#3296,locusStart#3297,locusEnd#3298,sampleSize#3299,ldSet#3300,locus#3301,... 2 more fields] parquet\n                                                   :                                                        :                                   :     :           +- Project [studyId#1199 AS rightStudyId#3323, geneId#1200, projectId#1201, studyType#3183 AS indexStudyType#3324, condition#1224, biosampleId#1228]\n                                                   :                                                        :                                   :     :              +- Project [studyId#1199, geneId#1200, projectId#1201, studyType#3183, traitFromSource#1203, traitFromSourceMappedIds#1204, biosampleFromSourceId#1205, pubmedId#1206, publicationTitle#1207, publicationFirstAuthor#1208, publicationDate#1209, publicationJournal#1210, backgroundTraitFromSourceMappedIds#1211, initialSampleSize#1212, nCases#1213, nControls#1214, nSamples#1215, cohorts#1216, ldPopulationStructure#1217, discoverySamples#1218, replicationSamples#1219, qualityControls#1220, analysisFlags#1221, summarystatsLocation#1222, ... 6 more fields]\n                                                   :                                                        :                                   :     :                 +- Project [studyId#1199, geneId#1200, projectId#1201, CASE WHEN isnotnull(newStudyType#1643) THEN newStudyType#1643 ELSE studyType#1202 END AS studyType#3183, traitFromSource#1203, traitFromSourceMappedIds#1204, biosampleFromSourceId#1205, pubmedId#1206, publicationTitle#1207, publicationFirstAuthor#1208, publicationDate#1209, publicationJournal#1210, backgroundTraitFromSourceMappedIds#1211, initialSampleSize#1212, nCases#1213, nControls#1214, nSamples#1215, cohorts#1216, ldPopulationStructure#1217, discoverySamples#1218, replicationSamples#1219, qualityControls#1220, analysisFlags#1221, summarystatsLocation#1222, ... 7 more fields]\n                                                   :                                                        :                                   :     :                    +- Project [studyId#1199, geneId#1200, projectId#1201, studyType#1202, traitFromSource#1203, traitFromSourceMappedIds#1204, biosampleFromSourceId#1205, pubmedId#1206, publicationTitle#1207, publicationFirstAuthor#1208, publicationDate#1209, publicationJournal#1210, backgroundTraitFromSourceMappedIds#1211, initialSampleSize#1212, nCases#1213, nControls#1214, nSamples#1215, cohorts#1216, ldPopulationStructure#1217, discoverySamples#1218, replicationSamples#1219, qualityControls#1220, analysisFlags#1221, summarystatsLocation#1222, ... 7 more fields]\n                                                   :                                                        :                                   :     :                       +- Join LeftOuter, (studyId#1199 = studyId#2346)\n                                                   :                                                        :                                   :     :                          :- Relation [studyId#1199,geneId#1200,projectId#1201,studyType#1202,traitFromSource#1203,traitFromSourceMappedIds#1204,biosampleFromSourceId#1205,pubmedId#1206,publicationTitle#1207,publicationFirstAuthor#1208,publicationDate#1209,publicationJournal#1210,backgroundTraitFromSourceMappedIds#1211,initialSampleSize#1212,nCases#1213,nControls#1214,nSamples#1215,cohorts#1216,ldPopulationStructure#1217,discoverySamples#1218,replicationSamples#1219,qualityControls#1220,analysisFlags#1221,summarystatsLocation#1222,... 6 more fields] parquet\n                                                   :                                                        :                                   :     :                          +- Project [studyId#2346, newStudyType#1643]\n                                                   :                                                        :                                   :     :                             +- Project [studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, analysisFlags#2368, summarystatsLocation#2369, ... 7 more fields]\n                                                   :                                                        :                                   :     :                                +- Project [extracted_column#1385, study_type#1337, studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, ... 10 more fields]\n                                                   :                                                        :                                   :     :                                   +- Project [extracted_column#1385, study_type#1337, studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, ... 9 more fields]\n                                                   :                                                        :                                   :     :                                      +- Project [extracted_column#1385, study_type#1337, studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, ... 8 more fields]\n                                                   :                                                        :                                   :     :                                         +- Join RightOuter, (extracted_column#1349 = extracted_column#1385)\n                                                   :                                                        :                                   :     :                                            :- Project [concat_ws(_, study_label#1329, quant_method#1335, sample_group#1330) AS extracted_column#1349, study_type#1337]\n                                                   :                                                        :                                   :     :                                            :  +- LogicalRDD [study_id#1327, dataset_id#1328, study_label#1329, sample_group#1330, tissue_id#1331, tissue_label#1332, condition_label#1333, sample_size#1334, quant_method#1335, pmid#1336, study_type#1337], false\n                                                   :                                                        :                                   :     :                                            +- Project [studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, analysisFlags#2368, summarystatsLocation#2369, ... 7 more fields]\n                                                   :                                                        :                                   :     :                                               +- Project [studyId#2346, geneId#2347, projectId#2348, studyType#2349, traitFromSource#2350, traitFromSourceMappedIds#2351, biosampleFromSourceId#2352, pubmedId#2353, publicationTitle#2354, publicationFirstAuthor#2355, publicationDate#2356, publicationJournal#2357, backgroundTraitFromSourceMappedIds#2358, initialSampleSize#2359, nCases#2360, nControls#2361, nSamples#2362, cohorts#2363, ldPopulationStructure#2364, discoverySamples#2365, replicationSamples#2366, qualityControls#2367, analysisFlags#2368, summarystatsLocation#2369, ... 7 more fields]\n                                                   :                                                        :                                   :     :                                                  +- Filter NOT StartsWith(studyId#2346, UKB_PPP)\n                                                   :                                                        :                                   :     :                                                     +- Filter NOT (studyType#2349 = gwas)\n                                                   :                                                        :                                   :     :                                                        +- Relation [studyId#2346,geneId#2347,projectId#2348,studyType#2349,traitFromSource#2350,traitFromSourceMappedIds#2351,biosampleFromSourceId#2352,pubmedId#2353,publicationTitle#2354,publicationFirstAuthor#2355,publicationDate#2356,publicationJournal#2357,backgroundTraitFromSourceMappedIds#2358,initialSampleSize#2359,nCases#2360,nControls#2361,nSamples#2362,cohorts#2363,ldPopulationStructure#2364,discoverySamples#2365,replicationSamples#2366,qualityControls#2367,analysisFlags#2368,summarystatsLocation#2369,... 6 more fields] parquet\n                                                   :                                                        :                                   :     +- Project [studyLocusId#1011 AS leftStudyLocusId#4687, datasourceId#932, targetId#933, datatypeId#962, diseaseFromSourceMappedId#966, resourceScore#1002, targetFromSourceId#1018, diseaseId#1033, id#1034, score#1035, sourceId#1038, studyId#4704, variantId#4705, betaGwas#4642, pValueExponent#4712]\n                                                   :                                                        :                                   :        +- Project [studyLocusId#1011, datasourceId#932, targetId#933, datatypeId#962, diseaseFromSourceMappedId#966, resourceScore#1002, targetFromSourceId#1018, diseaseId#1033, id#1034, score#1035, sourceId#1038, studyId#4704, variantId#4705, betaGwas#4642, pValueExponent#4712]\n                                                   :                                                        :                                   :           +- Join LeftOuter, (studyLocusId#1011 = studyLocusId#4703)\n                                                   :                                                        :                                   :              :- Project [datasourceId#932, targetId#933, datatypeId#962, diseaseFromSourceMappedId#966, resourceScore#1002, studyLocusId#1011, targetFromSourceId#1018, diseaseId#1033, id#1034, score#1035, sourceId#1038]\n                                                   :                                                        :                                   :              :  +- Filter (datasourceId#932 = gwas_credible_sets)\n                                                   :                                                        :                                   :              :     +- Filter datasourceId#932 IN (gwas_credible_sets)\n                                                   :                                                        :                                   :              :        +- Relation [datasourceId#932,targetId#933,alleleOrigins#934,allelicRequirements#935,ancestry#936,ancestryId#937,assays#938,assessments#939,beta#940,betaConfidenceIntervalLower#941,betaConfidenceIntervalUpper#942,biologicalModelAllelicComposition#943,biologicalModelGeneticBackground#944,biologicalModelId#945,biomarkerList#946,biomarkerName#947,biomarkers#948,biosamplesFromSource#949,cellLineBackground#950,cellType#951,clinicalPhase#952,clinicalSignificances#953,clinicalStatus#954,cohortDescription#955,... 83 more fields] parquet\n                                                   :                                                        :                                   :              +- Project [studyLocusId#4703, studyId#4704, variantId#4705, beta#4709 AS betaGwas#4642, pValueExponent#4712]\n                                                   :                                                        :                                   :                 +- Relation [studyLocusId#4703,studyId#4704,variantId#4705,chromosome#4706,position#4707,region#4708,beta#4709,zScore#4710,pValueMantissa#4711,pValueExponent#4712,effectAlleleFrequencyFromSource#4713,standardError#4714,subStudyDescription#4715,qualityControls#4716,finemappingMethod#4717,credibleSetIndex#4718,credibleSetlog10BF#4719,purityMeanR2#4720,purityMinR2#4721,locusStart#4722,locusEnd#4723,sampleSize#4724,ldSet#4725,locus#4726,... 2 more fields] parquet\n                                                   :                                                        :                                   +- Project [id#687 AS diseaseId#4766, name#691, parents#694, therapeuticAreas#699]\n                                                   :                                                        :                                      +- Relation [id#687,code#688,dbXRefs#689,description#690,name#691,directLocationIds#692,obsoleteTerms#693,parents#694,synonyms#695,ancestors#696,descendants#697,children#698,therapeuticAreas#699,indirectLocationIds#700,ontology#701] parquet\n                                                   :                                                        +- Project [leftStudyLocusId#5907, geneId#5867 AS targetId#5912, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                           +- Project [studyLocusId#5866 AS leftStudyLocusId#5907, geneId#5867, distanceFootprintMean#5897, distanceTssMean#5898]\n                                                   :                                                              +- Project [studyLocusId#5866, geneId#5867, __pivot_first(value) AS `first(value)`#5896[0] AS distanceFootprintMean#5897, __pivot_first(value) AS `first(value)`#5896[1] AS distanceTssMean#5898]\n                                                   :                                                                 +- Aggregate [studyLocusId#5866, geneId#5867], [studyLocusId#5866, geneId#5867, pivotfirst(key#5874, first(value)#5890, distanceFootprintMean, distanceTssMean, 0, 0) AS __pivot_first(value) AS `first(value)`#5896]\n                                                   :                                                                    +- Aggregate [studyLocusId#5866, geneId#5867, key#5874], [studyLocusId#5866, geneId#5867, key#5874, first(value#5875, false) AS first(value)#5890]\n                                                   :                                                                       +- Filter key#5874 IN (distanceFootprintMean,distanceTssMean)\n                                                   :                                                                          +- Project [studyLocusId#5866, geneId#5867, key#5874, value#5875]\n                                                   :                                                                             +- Generate explode(locusToGeneFeatures#5869), true, [key#5874, value#5875]\n                                                   :                                                                                +- Relation [studyLocusId#5866,geneId#5867,score#5868,locusToGeneFeatures#5869] parquet\n                                                   +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, coherencyDiagonal#13753 AS coherencyDiagonal_ch#13774, coherencyOneCell#13763 AS coherencyOneCell_ch#13775, LoF_protect#11572L AS LoF_protect_ch#13776L, GoF_protect#11571L AS GoF_protect_ch#13777L]\n                                                      +- Filter ((isnotnull(GoF_protect#11571L) OR isnotnull(LoF_protect#11572L)) AND (coherencyDiagonal#13753 = coherent))\n                                                         +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, GoF_risk#13442, LoF_risk#13587, coherencyDiagonal#13753, CASE WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnull(noEvaluable#11573L)) THEN noEvid WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnotnull(noEvaluable#11573L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#13587) OR isnotnull(LoF_protect#11572L)) OR isnotnull(GoF_risk#13442)) OR isnotnull(GoF_protect#11571L)) THEN CASE WHEN (isnotnull(LoF_risk#13587) AND ((isnull(LoF_protect#11572L) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L))) THEN coherent WHEN (isnotnull(GoF_risk#13442) AND ((isnull(LoF_protect#11572L) AND isnull(LoF_risk#13587)) AND isnull(GoF_protect#11571L))) THEN coherent WHEN (isnotnull(LoF_protect#11572L) AND ((isnull(LoF_risk#13587) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L))) THEN coherent WHEN (isnotnull(GoF_protect#11571L) AND ((isnull(LoF_protect#11572L) AND isnull(GoF_risk#13442)) AND isnull(LoF_risk#13587))) THEN coherent ELSE dispar END END AS coherencyOneCell#13763]\n                                                            +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, GoF_risk#13442, LoF_risk#13587, CASE WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnull(noEvaluable#11573L)) THEN noEvid WHEN ((((isnull(LoF_risk#13587) AND isnull(LoF_protect#11572L)) AND isnull(GoF_risk#13442)) AND isnull(GoF_protect#11571L)) AND isnotnull(noEvaluable#11573L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#13587) OR isnotnull(LoF_protect#11572L)) OR isnotnull(GoF_risk#13442)) OR isnotnull(GoF_protect#11571L)) THEN CASE WHEN (isnotnull(GoF_risk#13442) AND isnotnull(LoF_risk#13587)) THEN dispar WHEN (isnotnull(LoF_protect#11572L) AND isnotnull(LoF_risk#13587)) THEN dispar WHEN (isnotnull(GoF_protect#11571L) AND isnotnull(GoF_risk#13442)) THEN dispar WHEN (isnotnull(GoF_protect#11571L) AND isnotnull(LoF_protect#11572L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#13753]\n                                                               +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, GoF_risk#13442, null AS LoF_risk#13587]\n                                                                  +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, GoF_protect#11571L, LoF_protect#11572L, noEvaluable#11573L, null AS GoF_risk#13442]\n                                                                     +- Project [targetId#718, diseaseId#818, maxClinPhase#8314, __pivot_count(targetId) AS `count(targetId)`#11570[0] AS GoF_protect#11571L, __pivot_count(targetId) AS `count(targetId)`#11570[1] AS LoF_protect#11572L, __pivot_count(targetId) AS `count(targetId)`#11570[2] AS noEvaluable#11573L]\n                                                                        +- Aggregate [targetId#718, diseaseId#818, maxClinPhase#8314], [targetId#718, diseaseId#818, maxClinPhase#8314, pivotfirst(homogenized#7553, count(targetId)#11562L, GoF_protect, LoF_protect, noEvaluable, 0, 0) AS __pivot_count(targetId) AS `count(targetId)`#11570]\n                                                                           +- Aggregate [targetId#718, diseaseId#818, maxClinPhase#8314, homogenized#7553], [targetId#718, diseaseId#818, maxClinPhase#8314, homogenized#7553, count(targetId#718) AS count(targetId)#11562L]\n                                                                              +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 97 more fields]\n                                                                                 +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 98 more fields]\n                                                                                    +- Window [max(clinicalPhase#737) windowspecdefinition(targetId#718, diseaseId#818, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS maxClinPhase#8314], [targetId#718, diseaseId#818]\n                                                                                       +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 96 more fields]\n                                                                                          +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 96 more fields]\n                                                                                             +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 95 more fields]\n                                                                                                +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 95 more fields]\n                                                                                                   +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 95 more fields]\n                                                                                                      +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 96 more fields]\n                                                                                                         +- Window [collect_set(intogen_function#7068, 0, 0) windowspecdefinition(targetId#718, diseaseId#818, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#7191], [targetId#718, diseaseId#818]\n                                                                                                            +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 94 more fields]\n                                                                                                               +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 94 more fields]\n                                                                                                                  +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 93 more fields]\n                                                                                                                     +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 92 more fields]\n                                                                                                                        +- Join LeftOuter, ((drugId2#5999 = drugId#755) AND (targetId2#6006 = targetId#718))\n                                                                                                                           :- Join LeftOuter, (target_id#6049 = targetId#718)\n                                                                                                                           :  :- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, concat_ws(,, clinicalSignificances#738) AS clinicalSignificances#6271, clinicalStatus#739, cohortDescription#740, ... 83 more fields]\n                                                                                                                           :  :  +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#738, clinicalStatus#739, cohortDescription#740, ... 83 more fields]\n                                                                                                                           :  :     +- Project [datasourceId#717, targetId#718, alleleOrigins#719, allelicRequirements#720, ancestry#721, ancestryId#722, assays#723, assessments#724, cast(beta#725 as double) AS beta#6055, betaConfidenceIntervalLower#726, betaConfidenceIntervalUpper#727, biologicalModelAllelicComposition#728, biologicalModelGeneticBackground#729, biologicalModelId#730, biomarkerList#731, biomarkerName#732, biomarkers#733, biosamplesFromSource#734, cellLineBackground#735, cellType#736, clinicalPhase#737, clinicalSignificances#738, clinicalStatus#739, cohortDescription#740, ... 83 more fields]\n                                                                                                                           :  :        +- Filter datasourceId#717 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                                                           :  :           +- Filter (datasourceId#717 = chembl)\n                                                                                                                           :  :              +- Filter datasourceId#717 IN (ot_genetics_portal,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                                                                                           :  :                 +- Relation [datasourceId#717,targetId#718,alleleOrigins#719,allelicRequirements#720,ancestry#721,ancestryId#722,assays#723,assessments#724,beta#725,betaConfidenceIntervalLower#726,betaConfidenceIntervalUpper#727,biologicalModelAllelicComposition#728,biologicalModelGeneticBackground#729,biologicalModelId#730,biomarkerList#731,biomarkerName#732,biomarkers#733,biosamplesFromSource#734,cellLineBackground#735,cellType#736,clinicalPhase#737,clinicalSignificances#738,clinicalStatus#739,cohortDescription#740,... 83 more fields] parquet\n                                                                                                                           :  +- Project [id#5928 AS target_id#6049, approvedSymbol#5929, description#6034, description_splited#6038, TSorOncogene#6043]\n                                                                                                                           :     +- Project [id#5928, approvedSymbol#5929, description#6034, description_splited#6038, CASE WHEN (RLIKE(description_splited#6038, ncogene) AND RLIKE(description_splited#6038, TSG)) THEN bivalent WHEN RLIKE(description_splited#6038, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#6038, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#6043]\n                                                                                                                           :        +- Project [id#5928, approvedSymbol#5929, description#6034, concat_ws(,, description#6034) AS description_splited#6038]\n                                                                                                                           :           +- Aggregate [id#5928, approvedSymbol#5929], [id#5928, approvedSymbol#5929, collect_set(description#6026, 0, 0) AS description#6034]\n                                                                                                                           :              +- Filter description#6026 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                                                                                                           :                 +- Project [id#5928, approvedSymbol#5929, col#6021.description AS description#6026]\n                                                                                                                           :                    +- Project [id#5928, approvedSymbol#5929, col#6021]\n                                                                                                                           :                       +- Generate explode(hallmarks#5938.attributes), true, [col#6021]\n                                                                                                                           :                          +- Relation [id#5928,approvedSymbol#5929,biotype#5930,transcriptIds#5931,canonicalTranscript#5932,canonicalExons#5933,genomicLocation#5934,alternativeGenes#5935,approvedName#5936,go#5937,hallmarks#5938,synonyms#5939,symbolSynonyms#5940,nameSynonyms#5941,functionDescriptions#5942,subcellularLocations#5943,targetClass#5944,obsoleteSymbols#5945,obsoleteNames#5946,constraint#5947,tep#5948,proteinIds#5949,dbXrefs#5950,chemicalProbes#5951,... 4 more fields] parquet\n                                                                                                                           +- Aggregate [targetId2#6006, drugId2#5999], [targetId2#6006, drugId2#5999, collect_set(actionType#5984, 0, 0) AS actionType#6016]\n                                                                                                                              +- Project [targetId2#6006, drugId2#5999, actionType#5984, mechanismOfAction#5985]\n                                                                                                                                 +- Generate explode(targets#5989), true, [targetId2#6006]\n                                                                                                                                    +- Project [drugId2#5999, actionType#5984, mechanismOfAction#5985, targets#5989]\n                                                                                                                                       +- Generate explode(chemblIds#5986), true, [drugId2#5999]\n                                                                                                                                          +- Relation [actionType#5984,mechanismOfAction#5985,chemblIds#5986,targetName#5987,targetType#5988,targets#5989,references#5990] parquet\n"
     ]
    }
   ],
   "source": [
    "#### BUILDING THE NEW GWAS GENETIC EVIDENCE FROM COLOC\n",
    "from functions import discrepancifier\n",
    "from DoEAssessment import directionOfEffect\n",
    "from functions import relative_success\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date, datetime\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    "    IntegerType,\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\").filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "credibleEvidence = spark.read.parquet(f\"{path}evidence\").filter(\n",
    "    F.col(\"datasourceId\").isin([\"gwas_credible_sets\"])\n",
    ")\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "print(\"read spark files\")\n",
    "\n",
    "print(\"fixing scXQTL and XQTL studies\")\n",
    "#### Fixing scXQTL as XQTLs:\n",
    "## code provided by @ireneisdoomed\n",
    "pd.DataFrame.iteritems = pd.DataFrame.items\n",
    "\n",
    "raw_studies_metadata_schema: StructType = StructType(\n",
    "    [\n",
    "        StructField(\"study_id\", StringType(), True),\n",
    "        StructField(\"dataset_id\", StringType(), True),\n",
    "        StructField(\"study_label\", StringType(), True),\n",
    "        StructField(\"sample_group\", StringType(), True),\n",
    "        StructField(\"tissue_id\", StringType(), True),\n",
    "        StructField(\"tissue_label\", StringType(), True),\n",
    "        StructField(\"condition_label\", StringType(), True),\n",
    "        StructField(\"sample_size\", IntegerType(), True),\n",
    "        StructField(\"quant_method\", StringType(), True),\n",
    "        StructField(\"pmid\", StringType(), True),\n",
    "        StructField(\"study_type\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "    pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "    schema=raw_studies_metadata_schema,\n",
    ")\n",
    "\n",
    "# index = spark.read.parquet(\"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/genetics/parquet/study_index\")\n",
    "\n",
    "study_index_w_correct_type = (\n",
    "    study_table.select(\n",
    "        F.concat_ws(\n",
    "            \"_\",\n",
    "            F.col(\"study_label\"),\n",
    "            F.col(\"quant_method\"),\n",
    "            F.col(\"sample_group\"),\n",
    "        ).alias(\"extracted_column\"),\n",
    "        \"study_type\",\n",
    "    )\n",
    "    .join(\n",
    "        index\n",
    "        # Get eQTL Catalogue studies\n",
    "        .filter(F.col(\"studyType\") != \"gwas\").filter(\n",
    "            ~F.col(\"studyId\").startswith(\"UKB_PPP\")\n",
    "        )\n",
    "        # Remove measured trait\n",
    "        .withColumn(\n",
    "            \"extracted_column\",\n",
    "            F.regexp_replace(F.col(\"studyId\"), r\"(_ENS.*|_ILMN.*|_X.*|_[0-9]+:.*)\", \"\"),\n",
    "        ).withColumn(\n",
    "            \"extracted_column\",\n",
    "            # After the previous cleanup, there are multiple traits from the same publication starting with the gene symbol that need to be removed (e.g. `Sun_2018_aptamer_plasma_ANXA2.4961.17.1..1`)\n",
    "            F.when(\n",
    "                F.col(\"extracted_column\").startswith(\"Sun_2018_aptamer_plasma\"),\n",
    "                F.lit(\"Sun_2018_aptamer_plasma\"),\n",
    "            ).otherwise(F.col(\"extracted_column\")),\n",
    "        ),\n",
    "        on=\"extracted_column\",\n",
    "        how=\"right\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "fixed = (\n",
    "    study_index_w_correct_type.withColumn(\n",
    "        \"toFix\",\n",
    "        F.when(\n",
    "            (F.col(\"study_type\") != \"single-cell\")\n",
    "            & (F.col(\"studyType\").startswith(\"sc\")),\n",
    "            F.lit(True),\n",
    "        ).otherwise(F.lit(False)),\n",
    "    )\n",
    "    # Remove the substring \"sc\" from the studyType column\n",
    "    .withColumn(\n",
    "        \"newStudyType\",\n",
    "        F.when(\n",
    "            F.col(\"toFix\"), F.regexp_replace(F.col(\"studyType\"), r\"sc\", \"\")\n",
    "        ).otherwise(F.col(\"studyType\")),\n",
    "    ).drop(\"toFix\", \"extracted_column\", \"study_type\")\n",
    ").persist()\n",
    "all_studies = index.join(\n",
    "    fixed.selectExpr(\"studyId\", \"newStudyType\"), on=\"studyId\", how=\"left\"\n",
    ").persist()\n",
    "fixedIndex = all_studies.withColumn(\n",
    "    \"studyType\",\n",
    "    F.when(F.col(\"newStudyType\").isNotNull(), F.col(\"newStudyType\")).otherwise(\n",
    "        F.col(\"studyType\")\n",
    "    ),\n",
    ").drop(\"newStudyType\")\n",
    "\n",
    "print(\"fixed scXQTL and XQTL studies\")\n",
    "\n",
    "print(\"creating new coloc\")\n",
    "\n",
    "#### fixed\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        fixedIndex.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "# remove columns without content (only null values on them)\n",
    "df = credibleEvidence.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns).persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")\n",
    "print(\"creating new gwasResolvedColoc\")\n",
    "\n",
    "### bring directionality from QTL\n",
    "\n",
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "\n",
    "print(\"creating new gwasCredibleAssoc\")\n",
    "\n",
    "### modify to include more information\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )\n",
    "    .select(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"homogenized\",\n",
    "        \"leftStudyLocusId\",\n",
    "        \"h4\",\n",
    "        \"datasourceId\",\n",
    "        \"resourceScore\",\n",
    "        \"leftVariantId\",\n",
    "        \"credibleLeftStudyType\",\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")  ### there will be duplicates TargetId-DiseaseId because we are taking the most significant DoE\n",
    "\n",
    "#### LOAD STUDYLOCUSID AND VARIANT DISTANCES\n",
    "l2gPred = spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/locusToGenePredictions\"\n",
    ")\n",
    "l2gTable = (\n",
    "    l2gPred.select(\"studyLocusId\", \"geneId\", F.explode_outer(\"locusToGeneFeatures\"))\n",
    "    .filter(F.col(\"key\").isin([\"distanceFootprintMean\", \"distanceTssMean\"]))\n",
    "    .groupBy(\"studyLocusId\", \"geneId\")\n",
    "    .pivot(\"key\")\n",
    "    .agg(F.first(\"value\"))\n",
    ")\n",
    "print(\"creating gwasCredibleAssocDistances\")\n",
    "gwasCredibleAssocDistances = gwasCredibleAssoc.join(\n",
    "    l2gTable.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\").withColumnRenamed(\n",
    "        \"geneId\", \"targetId\"\n",
    "    ),\n",
    "    on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "print(\"creating analysis_chembl\")\n",
    "analysis_chembl = (\n",
    "    discrepancifier(\n",
    "        directionOfEffect(\n",
    "            evidences.filter((F.col(\"datasourceId\") == \"chembl\")), \"24.09\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "    .filter(  ### ensure drug has annotated MoA and is coherent per Target-Disease\n",
    "        ((F.col(\"GoF_protect\").isNotNull()) | F.col(\"LoF_protect\").isNotNull())\n",
    "        & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    )\n",
    "    .selectExpr(\n",
    "        \"targetId\",\n",
    "        \"diseaseId\",\n",
    "        \"maxClinPhase\",\n",
    "        \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "        \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "        \"LoF_protect as LoF_protect_ch\",\n",
    "        \"GoF_protect as GoF_protect_ch\",\n",
    "    )\n",
    ")\n",
    "\n",
    "### pivot colocdoE grouping by T-D-studyLocusId-distances\n",
    "\n",
    "values = [\"max_L2GScore\", \"min_footPrintDistance_rank\", \"min_tssDistance_rank\"]\n",
    "value = \"max_L2GScore\"\n",
    "\n",
    "list_l2g = [\n",
    "    0.10,\n",
    "    0.15,\n",
    "    0.20,\n",
    "    0.25,\n",
    "    0.30,\n",
    "    0.35,\n",
    "    0.40,\n",
    "    0.45,\n",
    "    0.50,\n",
    "    0.55,\n",
    "    0.60,\n",
    "    0.65,\n",
    "    0.70,\n",
    "    0.75,\n",
    "    0.80,\n",
    "    0.85,\n",
    "    0.90,\n",
    "    0.95,\n",
    "]\n",
    "print(\"creating benchmarkOT function\")\n",
    "\n",
    "dict_comb = {}\n",
    "\n",
    "\n",
    "def benchmarkOT(\n",
    "    dict_comb, value, gwasCredibleAssocDistances, analysis_chembl, list_l2g\n",
    "):\n",
    "\n",
    "    dict_comb = {\n",
    "        \"hasGeneticEvidence\": f\"{value}\",\n",
    "        \"diagonalYes\": f\"{value}\",\n",
    "        \"oneCellYes\": f\"{value}\",\n",
    "        \"L2GAndColoc\": f\"{value}\",\n",
    "    }\n",
    "    return (\n",
    "        discrepancifier(\n",
    "            gwasCredibleAssocDistances\n",
    "            # .filter(F.col(\"h4\").isNotNull()) #### not filter by this because we want to include the L2G AND Coloc question\n",
    "            .withColumn(  ### take maximum L2G score per T-D\n",
    "                \"max_L2GScore\",\n",
    "                F.max(\"resourceScore\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(\n",
    "                \"min_footPrintDistance_rank\",\n",
    "                F.min(\"distanceFootprintMean\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .withColumn(\n",
    "                \"min_tssDistance_rank\",\n",
    "                F.min(\"distanceTssMean\").over(\n",
    "                    Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "                ),\n",
    "            )\n",
    "            .groupBy(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                f\"{value}\",\n",
    "                # \"leftStudyLocusId\",\n",
    "            )\n",
    "            .pivot(\"homogenized\")\n",
    "            .count()\n",
    "        )\n",
    "        .join(analysis_chembl, on=[\"targetId\", \"diseaseId\"], how=\"right\")\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (\n",
    "                    (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                    & (F.col(\"coherencyDiagonal\").isNotNull())\n",
    "                )\n",
    "                # & (F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "                ,\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (\n",
    "                    (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                    & (F.col(\"coherencyDiagonal\").isNotNull())\n",
    "                )\n",
    "                # & (F.col(\"coherencyOneCell\") == \"coherent\")\n",
    "                ,\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "            # ).filter(\n",
    "            #    F.col(\"diagonalAgreeWithDrugs\").isNotNull()\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(F.col(f\"{value}\").isNotNull(), F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"L2GAndColoc\",\n",
    "            F.when(\n",
    "                (F.col(f\"{value}\").isNotNull())\n",
    "                & (F.col(\"coherencyDiagonal\").isin([\"coherent\", \"dispar\"])),\n",
    "                F.lit(\"yes\"),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .select(\n",
    "            [\"*\"]\n",
    "            + (\n",
    "                [  ### single columns\n",
    "                    F.when(F.col(f\"{value}\") >= n, F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{value}>={str(n).replace('.', '_')}\")\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "            )\n",
    "            + (\n",
    "                [  ### column combinations for Yes/No colums Plus has DoE (any agreement)\n",
    "                    F.when((F.col(a) == \"yes\") & (F.col(x) >= n), F.lit(\"yes\"))\n",
    "                    .otherwise(F.lit(\"no\"))\n",
    "                    .alias(f\"{x}>={str(n).replace('.', '_')}&{a}_combined\")\n",
    "                    for a, x in dict_comb.items()\n",
    "                    for n in list_l2g\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "### HERE\n",
    "\n",
    "### Make all the datasets\n",
    "values = [\"max_L2GScore\", \"min_footPrintDistance_rank\", \"min_tssDistance_rank\"]\n",
    "\n",
    "print(\"creating dataframes in loop\")\n",
    "\n",
    "datasetDict = {}\n",
    "for value in values:\n",
    "    if value == \"max_L2GScore\":\n",
    "        datasetDict[f\"df_l2g_original\"] = benchmarkOT(\n",
    "            dict_comb, value, gwasCredibleAssocDistances, analysis_chembl, list_l2g\n",
    "        )\n",
    "    else:\n",
    "        datasetDict[f\"{value}\"] = benchmarkOT(\n",
    "            dict_comb, value, gwasCredibleAssocDistances, analysis_chembl, list_l2g\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df(dataset) -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    toAnalysis = dataset.columns[22:]\n",
    "    dataType = [\"byDatatype\"] * len(toAnalysis)\n",
    "    l_studies = []\n",
    "    l_studies.extend([list(a) for a in zip(toAnalysis, dataType)])\n",
    "\n",
    "    schema = StructType(\n",
    "        [\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "            StructField(\"comparisonType\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    comparisons = spark.createDataFrame(l_studies, schema=schema)\n",
    "    ### include all the columns as predictor\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            # (\"Phase>=3\", \"clinical\"),\n",
    "            # (\"Phase>=2\", \"clinical\"),\n",
    "            # (\"Phase>=1\", \"clinical\"),\n",
    "            # (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "\n",
    "    wComparison = Window.partitionBy(F.col(comparisonColumn))\n",
    "    wPrediction = Window.partitionBy(F.col(predictionColumn))\n",
    "    wPredictionComparison = Window.partitionBy(\n",
    "        F.col(comparisonColumn), F.col(predictionColumn)\n",
    "    )\n",
    "\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "    \"\"\"\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    filePath = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            comparisonType,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            filePath,\n",
    "        ]\n",
    "    )\n",
    "    return results\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio\n",
    "\n",
    "# Initialize an empty list to store the results\n",
    "result_st = []\n",
    "result_ci = []\n",
    "results = []\n",
    "\n",
    "\n",
    "def convertTuple(tup):\n",
    "    st = \",\".join(map(str, tup))\n",
    "    return st\n",
    "\n",
    "\n",
    "print(\"launched function to run analysis\")\n",
    "\n",
    "\n",
    "listado = []\n",
    "today_date = str(date.today())\n",
    "for key, df_analysis in datasetDict.items():\n",
    "    aggSetups_original = comparisons_df(df_analysis)\n",
    "    print(\"corresponding dataframe key: \", key)\n",
    "    df_analysis.persist()\n",
    "    for row in aggSetups_original:\n",
    "        print(key, value)\n",
    "        aggregations_original(df_analysis, key, listado, *row, today_date)\n",
    "\n",
    "print(\"finished analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "aggSetups_original = comparisons_df(df_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Row(comparison='diagonalYes', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='oneCellYes', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='hasGeneticEvidence', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='L2GAndColoc', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_1', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_15', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_2', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_25', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_3', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_35', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_4', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_45', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_5', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_55', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_6', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_65', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_7', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_75', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_8', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_85', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_9', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_95', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_1&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_15&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_2&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_25&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_3&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_35&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_4&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_45&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_5&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_55&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_6&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_65&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_7&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_75&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_8&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_85&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_9&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_95&hasGeneticEvidence_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_1&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_15&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_2&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_25&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_3&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_35&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_4&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_45&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_5&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_55&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_6&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_65&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_7&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_75&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_8&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_85&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_9&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_95&diagonalYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_1&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_15&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_2&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_25&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_3&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_35&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_4&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_45&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_5&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_55&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_6&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_65&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_7&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_75&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_8&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_85&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_9&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_95&oneCellYes_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_1&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_15&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_2&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_25&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_3&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_35&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_4&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_45&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_5&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_55&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_6&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_65&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_7&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_75&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_8&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_85&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_9&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical'),\n",
       " Row(comparison='min_tssDistance_rank>=0_95&L2GAndColoc_combined', comparisonType='byDatatype', _1='Phase4', _2='clinical')]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggSetups_original"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_l2g_original DataFrame[targetId: string, diseaseId: string, max_L2GScore: double, GoF_protect: bigint, GoF_risk: bigint, LoF_protect: bigint, LoF_risk: bigint, noEvaluable: bigint, coherencyDiagonal: string, coherencyOneCell: string, maxClinPhase: double, coherencyDiagonal_ch: string, coherencyOneCell_ch: string, LoF_protect_ch: bigint, GoF_protect_ch: bigint, diagonalAgreeWithDrugs: string, oneCellAgreeWithDrugs: string, Phase4: string, Phase>=3: string, Phase>=2: string, Phase>=1: string, Phase0: string, diagonalYes: string, oneCellYes: string, hasGeneticEvidence: string, L2GAndColoc: string, max_L2GScore>=0_1: string, max_L2GScore>=0_15: string, max_L2GScore>=0_2: string, max_L2GScore>=0_25: string, max_L2GScore>=0_3: string, max_L2GScore>=0_35: string, max_L2GScore>=0_4: string, max_L2GScore>=0_45: string, max_L2GScore>=0_5: string, max_L2GScore>=0_55: string, max_L2GScore>=0_6: string, max_L2GScore>=0_65: string, max_L2GScore>=0_7: string, max_L2GScore>=0_75: string, max_L2GScore>=0_8: string, max_L2GScore>=0_85: string, max_L2GScore>=0_9: string, max_L2GScore>=0_95: string, max_L2GScore>=0_1&hasGeneticEvidence_combined: string, max_L2GScore>=0_15&hasGeneticEvidence_combined: string, max_L2GScore>=0_2&hasGeneticEvidence_combined: string, max_L2GScore>=0_25&hasGeneticEvidence_combined: string, max_L2GScore>=0_3&hasGeneticEvidence_combined: string, max_L2GScore>=0_35&hasGeneticEvidence_combined: string, max_L2GScore>=0_4&hasGeneticEvidence_combined: string, max_L2GScore>=0_45&hasGeneticEvidence_combined: string, max_L2GScore>=0_5&hasGeneticEvidence_combined: string, max_L2GScore>=0_55&hasGeneticEvidence_combined: string, max_L2GScore>=0_6&hasGeneticEvidence_combined: string, max_L2GScore>=0_65&hasGeneticEvidence_combined: string, max_L2GScore>=0_7&hasGeneticEvidence_combined: string, max_L2GScore>=0_75&hasGeneticEvidence_combined: string, max_L2GScore>=0_8&hasGeneticEvidence_combined: string, max_L2GScore>=0_85&hasGeneticEvidence_combined: string, max_L2GScore>=0_9&hasGeneticEvidence_combined: string, max_L2GScore>=0_95&hasGeneticEvidence_combined: string, max_L2GScore>=0_1&diagonalYes_combined: string, max_L2GScore>=0_15&diagonalYes_combined: string, max_L2GScore>=0_2&diagonalYes_combined: string, max_L2GScore>=0_25&diagonalYes_combined: string, max_L2GScore>=0_3&diagonalYes_combined: string, max_L2GScore>=0_35&diagonalYes_combined: string, max_L2GScore>=0_4&diagonalYes_combined: string, max_L2GScore>=0_45&diagonalYes_combined: string, max_L2GScore>=0_5&diagonalYes_combined: string, max_L2GScore>=0_55&diagonalYes_combined: string, max_L2GScore>=0_6&diagonalYes_combined: string, max_L2GScore>=0_65&diagonalYes_combined: string, max_L2GScore>=0_7&diagonalYes_combined: string, max_L2GScore>=0_75&diagonalYes_combined: string, max_L2GScore>=0_8&diagonalYes_combined: string, max_L2GScore>=0_85&diagonalYes_combined: string, max_L2GScore>=0_9&diagonalYes_combined: string, max_L2GScore>=0_95&diagonalYes_combined: string, max_L2GScore>=0_1&oneCellYes_combined: string, max_L2GScore>=0_15&oneCellYes_combined: string, max_L2GScore>=0_2&oneCellYes_combined: string, max_L2GScore>=0_25&oneCellYes_combined: string, max_L2GScore>=0_3&oneCellYes_combined: string, max_L2GScore>=0_35&oneCellYes_combined: string, max_L2GScore>=0_4&oneCellYes_combined: string, max_L2GScore>=0_45&oneCellYes_combined: string, max_L2GScore>=0_5&oneCellYes_combined: string, max_L2GScore>=0_55&oneCellYes_combined: string, max_L2GScore>=0_6&oneCellYes_combined: string, max_L2GScore>=0_65&oneCellYes_combined: string, max_L2GScore>=0_7&oneCellYes_combined: string, max_L2GScore>=0_75&oneCellYes_combined: string, max_L2GScore>=0_8&oneCellYes_combined: string, max_L2GScore>=0_85&oneCellYes_combined: string, max_L2GScore>=0_9&oneCellYes_combined: string, max_L2GScore>=0_95&oneCellYes_combined: string, max_L2GScore>=0_1&L2GAndColoc_combined: string, max_L2GScore>=0_15&L2GAndColoc_combined: string, max_L2GScore>=0_2&L2GAndColoc_combined: string, max_L2GScore>=0_25&L2GAndColoc_combined: string, max_L2GScore>=0_3&L2GAndColoc_combined: string, max_L2GScore>=0_35&L2GAndColoc_combined: string, max_L2GScore>=0_4&L2GAndColoc_combined: string, max_L2GScore>=0_45&L2GAndColoc_combined: string, max_L2GScore>=0_5&L2GAndColoc_combined: string, max_L2GScore>=0_55&L2GAndColoc_combined: string, max_L2GScore>=0_6&L2GAndColoc_combined: string, max_L2GScore>=0_65&L2GAndColoc_combined: string, max_L2GScore>=0_7&L2GAndColoc_combined: string, max_L2GScore>=0_75&L2GAndColoc_combined: string, max_L2GScore>=0_8&L2GAndColoc_combined: string, max_L2GScore>=0_85&L2GAndColoc_combined: string, max_L2GScore>=0_9&L2GAndColoc_combined: string, max_L2GScore>=0_95&L2GAndColoc_combined: string]\n",
      "min_footPrintDistance_rank DataFrame[targetId: string, diseaseId: string, min_footPrintDistance_rank: float, GoF_protect: bigint, GoF_risk: bigint, LoF_protect: bigint, LoF_risk: bigint, noEvaluable: bigint, coherencyDiagonal: string, coherencyOneCell: string, maxClinPhase: double, coherencyDiagonal_ch: string, coherencyOneCell_ch: string, LoF_protect_ch: bigint, GoF_protect_ch: bigint, diagonalAgreeWithDrugs: string, oneCellAgreeWithDrugs: string, Phase4: string, Phase>=3: string, Phase>=2: string, Phase>=1: string, Phase0: string, diagonalYes: string, oneCellYes: string, hasGeneticEvidence: string, L2GAndColoc: string, min_footPrintDistance_rank>=0_1: string, min_footPrintDistance_rank>=0_15: string, min_footPrintDistance_rank>=0_2: string, min_footPrintDistance_rank>=0_25: string, min_footPrintDistance_rank>=0_3: string, min_footPrintDistance_rank>=0_35: string, min_footPrintDistance_rank>=0_4: string, min_footPrintDistance_rank>=0_45: string, min_footPrintDistance_rank>=0_5: string, min_footPrintDistance_rank>=0_55: string, min_footPrintDistance_rank>=0_6: string, min_footPrintDistance_rank>=0_65: string, min_footPrintDistance_rank>=0_7: string, min_footPrintDistance_rank>=0_75: string, min_footPrintDistance_rank>=0_8: string, min_footPrintDistance_rank>=0_85: string, min_footPrintDistance_rank>=0_9: string, min_footPrintDistance_rank>=0_95: string, min_footPrintDistance_rank>=0_1&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_15&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_2&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_25&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_3&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_35&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_4&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_45&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_5&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_55&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_6&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_65&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_7&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_75&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_8&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_85&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_9&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_95&hasGeneticEvidence_combined: string, min_footPrintDistance_rank>=0_1&diagonalYes_combined: string, min_footPrintDistance_rank>=0_15&diagonalYes_combined: string, min_footPrintDistance_rank>=0_2&diagonalYes_combined: string, min_footPrintDistance_rank>=0_25&diagonalYes_combined: string, min_footPrintDistance_rank>=0_3&diagonalYes_combined: string, min_footPrintDistance_rank>=0_35&diagonalYes_combined: string, min_footPrintDistance_rank>=0_4&diagonalYes_combined: string, min_footPrintDistance_rank>=0_45&diagonalYes_combined: string, min_footPrintDistance_rank>=0_5&diagonalYes_combined: string, min_footPrintDistance_rank>=0_55&diagonalYes_combined: string, min_footPrintDistance_rank>=0_6&diagonalYes_combined: string, min_footPrintDistance_rank>=0_65&diagonalYes_combined: string, min_footPrintDistance_rank>=0_7&diagonalYes_combined: string, min_footPrintDistance_rank>=0_75&diagonalYes_combined: string, min_footPrintDistance_rank>=0_8&diagonalYes_combined: string, min_footPrintDistance_rank>=0_85&diagonalYes_combined: string, min_footPrintDistance_rank>=0_9&diagonalYes_combined: string, min_footPrintDistance_rank>=0_95&diagonalYes_combined: string, min_footPrintDistance_rank>=0_1&oneCellYes_combined: string, min_footPrintDistance_rank>=0_15&oneCellYes_combined: string, min_footPrintDistance_rank>=0_2&oneCellYes_combined: string, min_footPrintDistance_rank>=0_25&oneCellYes_combined: string, min_footPrintDistance_rank>=0_3&oneCellYes_combined: string, min_footPrintDistance_rank>=0_35&oneCellYes_combined: string, min_footPrintDistance_rank>=0_4&oneCellYes_combined: string, min_footPrintDistance_rank>=0_45&oneCellYes_combined: string, min_footPrintDistance_rank>=0_5&oneCellYes_combined: string, min_footPrintDistance_rank>=0_55&oneCellYes_combined: string, min_footPrintDistance_rank>=0_6&oneCellYes_combined: string, min_footPrintDistance_rank>=0_65&oneCellYes_combined: string, min_footPrintDistance_rank>=0_7&oneCellYes_combined: string, min_footPrintDistance_rank>=0_75&oneCellYes_combined: string, min_footPrintDistance_rank>=0_8&oneCellYes_combined: string, min_footPrintDistance_rank>=0_85&oneCellYes_combined: string, min_footPrintDistance_rank>=0_9&oneCellYes_combined: string, min_footPrintDistance_rank>=0_95&oneCellYes_combined: string, min_footPrintDistance_rank>=0_1&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_15&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_2&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_25&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_3&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_35&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_4&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_45&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_5&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_55&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_6&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_65&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_7&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_75&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_8&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_85&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_9&L2GAndColoc_combined: string, min_footPrintDistance_rank>=0_95&L2GAndColoc_combined: string]\n",
      "min_tssDistance_rank DataFrame[targetId: string, diseaseId: string, min_tssDistance_rank: float, GoF_protect: bigint, GoF_risk: bigint, LoF_protect: bigint, LoF_risk: bigint, noEvaluable: bigint, coherencyDiagonal: string, coherencyOneCell: string, maxClinPhase: double, coherencyDiagonal_ch: string, coherencyOneCell_ch: string, LoF_protect_ch: bigint, GoF_protect_ch: bigint, diagonalAgreeWithDrugs: string, oneCellAgreeWithDrugs: string, Phase4: string, Phase>=3: string, Phase>=2: string, Phase>=1: string, Phase0: string, diagonalYes: string, oneCellYes: string, hasGeneticEvidence: string, L2GAndColoc: string, min_tssDistance_rank>=0_1: string, min_tssDistance_rank>=0_15: string, min_tssDistance_rank>=0_2: string, min_tssDistance_rank>=0_25: string, min_tssDistance_rank>=0_3: string, min_tssDistance_rank>=0_35: string, min_tssDistance_rank>=0_4: string, min_tssDistance_rank>=0_45: string, min_tssDistance_rank>=0_5: string, min_tssDistance_rank>=0_55: string, min_tssDistance_rank>=0_6: string, min_tssDistance_rank>=0_65: string, min_tssDistance_rank>=0_7: string, min_tssDistance_rank>=0_75: string, min_tssDistance_rank>=0_8: string, min_tssDistance_rank>=0_85: string, min_tssDistance_rank>=0_9: string, min_tssDistance_rank>=0_95: string, min_tssDistance_rank>=0_1&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_15&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_2&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_25&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_3&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_35&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_4&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_45&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_5&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_55&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_6&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_65&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_7&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_75&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_8&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_85&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_9&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_95&hasGeneticEvidence_combined: string, min_tssDistance_rank>=0_1&diagonalYes_combined: string, min_tssDistance_rank>=0_15&diagonalYes_combined: string, min_tssDistance_rank>=0_2&diagonalYes_combined: string, min_tssDistance_rank>=0_25&diagonalYes_combined: string, min_tssDistance_rank>=0_3&diagonalYes_combined: string, min_tssDistance_rank>=0_35&diagonalYes_combined: string, min_tssDistance_rank>=0_4&diagonalYes_combined: string, min_tssDistance_rank>=0_45&diagonalYes_combined: string, min_tssDistance_rank>=0_5&diagonalYes_combined: string, min_tssDistance_rank>=0_55&diagonalYes_combined: string, min_tssDistance_rank>=0_6&diagonalYes_combined: string, min_tssDistance_rank>=0_65&diagonalYes_combined: string, min_tssDistance_rank>=0_7&diagonalYes_combined: string, min_tssDistance_rank>=0_75&diagonalYes_combined: string, min_tssDistance_rank>=0_8&diagonalYes_combined: string, min_tssDistance_rank>=0_85&diagonalYes_combined: string, min_tssDistance_rank>=0_9&diagonalYes_combined: string, min_tssDistance_rank>=0_95&diagonalYes_combined: string, min_tssDistance_rank>=0_1&oneCellYes_combined: string, min_tssDistance_rank>=0_15&oneCellYes_combined: string, min_tssDistance_rank>=0_2&oneCellYes_combined: string, min_tssDistance_rank>=0_25&oneCellYes_combined: string, min_tssDistance_rank>=0_3&oneCellYes_combined: string, min_tssDistance_rank>=0_35&oneCellYes_combined: string, min_tssDistance_rank>=0_4&oneCellYes_combined: string, min_tssDistance_rank>=0_45&oneCellYes_combined: string, min_tssDistance_rank>=0_5&oneCellYes_combined: string, min_tssDistance_rank>=0_55&oneCellYes_combined: string, min_tssDistance_rank>=0_6&oneCellYes_combined: string, min_tssDistance_rank>=0_65&oneCellYes_combined: string, min_tssDistance_rank>=0_7&oneCellYes_combined: string, min_tssDistance_rank>=0_75&oneCellYes_combined: string, min_tssDistance_rank>=0_8&oneCellYes_combined: string, min_tssDistance_rank>=0_85&oneCellYes_combined: string, min_tssDistance_rank>=0_9&oneCellYes_combined: string, min_tssDistance_rank>=0_95&oneCellYes_combined: string, min_tssDistance_rank>=0_1&L2GAndColoc_combined: string, min_tssDistance_rank>=0_15&L2GAndColoc_combined: string, min_tssDistance_rank>=0_2&L2GAndColoc_combined: string, min_tssDistance_rank>=0_25&L2GAndColoc_combined: string, min_tssDistance_rank>=0_3&L2GAndColoc_combined: string, min_tssDistance_rank>=0_35&L2GAndColoc_combined: string, min_tssDistance_rank>=0_4&L2GAndColoc_combined: string, min_tssDistance_rank>=0_45&L2GAndColoc_combined: string, min_tssDistance_rank>=0_5&L2GAndColoc_combined: string, min_tssDistance_rank>=0_55&L2GAndColoc_combined: string, min_tssDistance_rank>=0_6&L2GAndColoc_combined: string, min_tssDistance_rank>=0_65&L2GAndColoc_combined: string, min_tssDistance_rank>=0_7&L2GAndColoc_combined: string, min_tssDistance_rank>=0_75&L2GAndColoc_combined: string, min_tssDistance_rank>=0_8&L2GAndColoc_combined: string, min_tssDistance_rank>=0_85&L2GAndColoc_combined: string, min_tssDistance_rank>=0_9&L2GAndColoc_combined: string, min_tssDistance_rank>=0_95&L2GAndColoc_combined: string]\n"
     ]
    }
   ],
   "source": [
    "for key, df_analysis in datasetDict.items():\n",
    "    print(key, df_analysis)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### debugging ignorenulls=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### 10.12.2024\n",
    "from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "### index with new fix\" \"gs://ot-team/irene/gentropy/study_index_2412_fixed\"\n",
    "index = spark.read.parquet(f\"gs://ot-team/irene/gentropy/study_index_2412_fixed\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "\n",
    "#### Fixing scXQTL as XQTLs:\n",
    "## code provided by @ireneisdoomed\n",
    "pd.DataFrame.iteritems = pd.DataFrame.items\n",
    "\n",
    "raw_studies_metadata_schema: StructType = StructType(\n",
    "    [\n",
    "        StructField(\"study_id\", StringType(), True),\n",
    "        StructField(\"dataset_id\", StringType(), True),\n",
    "        StructField(\"study_label\", StringType(), True),\n",
    "        StructField(\"sample_group\", StringType(), True),\n",
    "        StructField(\"tissue_id\", StringType(), True),\n",
    "        StructField(\"tissue_label\", StringType(), True),\n",
    "        StructField(\"condition_label\", StringType(), True),\n",
    "        StructField(\"sample_size\", IntegerType(), True),\n",
    "        StructField(\"quant_method\", StringType(), True),\n",
    "        StructField(\"pmid\", StringType(), True),\n",
    "        StructField(\"study_type\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "    pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "    schema=raw_studies_metadata_schema,\n",
    ")\n",
    "\n",
    "# index = spark.read.parquet(\"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/genetics/parquet/study_index\")\n",
    "\n",
    "study_index_w_correct_type = (\n",
    "    study_table.select(\n",
    "        F.concat_ws(\n",
    "            \"_\",\n",
    "            F.col(\"study_label\"),\n",
    "            F.col(\"quant_method\"),\n",
    "            F.col(\"sample_group\"),\n",
    "        ).alias(\"extracted_column\"),\n",
    "        \"study_type\",\n",
    "    )\n",
    "    .join(\n",
    "        index\n",
    "        # Get eQTL Catalogue studies\n",
    "        .filter(F.col(\"studyType\") != \"gwas\").filter(\n",
    "            ~F.col(\"studyId\").startswith(\"UKB_PPP\")\n",
    "        )\n",
    "        # Remove measured trait\n",
    "        .withColumn(\n",
    "            \"extracted_column\",\n",
    "            F.regexp_replace(F.col(\"studyId\"), r\"(_ENS.*|_ILMN.*|_X.*|_[0-9]+:.*)\", \"\"),\n",
    "        ).withColumn(\n",
    "            \"extracted_column\",\n",
    "            # After the previous cleanup, there are multiple traits from the same publication starting with the gene symbol that need to be removed (e.g. `Sun_2018_aptamer_plasma_ANXA2.4961.17.1..1`)\n",
    "            F.when(\n",
    "                F.col(\"extracted_column\").startswith(\"Sun_2018_aptamer_plasma\"),\n",
    "                F.lit(\"Sun_2018_aptamer_plasma\"),\n",
    "            ).otherwise(F.col(\"extracted_column\")),\n",
    "        ),\n",
    "        on=\"extracted_column\",\n",
    "        how=\"right\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "fixed = (\n",
    "    study_index_w_correct_type.withColumn(\n",
    "        \"toFix\",\n",
    "        F.when(\n",
    "            (F.col(\"study_type\") != \"single-cell\")\n",
    "            & (F.col(\"studyType\").startswith(\"sc\")),\n",
    "            F.lit(True),\n",
    "        ).otherwise(F.lit(False)),\n",
    "    )\n",
    "    # Remove the substring \"sc\" from the studyType column\n",
    "    .withColumn(\n",
    "        \"newStudyType\",\n",
    "        F.when(\n",
    "            F.col(\"toFix\"), F.regexp_replace(F.col(\"studyType\"), r\"sc\", \"\")\n",
    "        ).otherwise(F.col(\"studyType\")),\n",
    "    ).drop(\"toFix\", \"extracted_column\", \"study_type\")\n",
    ").persist()\n",
    "all_studies = index.join(\n",
    "    fixed.selectExpr(\"studyId\", \"newStudyType\"), on=\"studyId\", how=\"left\"\n",
    ").persist()\n",
    "fixedIndex = all_studies.withColumn(\n",
    "    \"studyType\",\n",
    "    F.when(F.col(\"newStudyType\").isNotNull(), F.col(\"newStudyType\")).otherwise(\n",
    "        F.col(\"studyType\")\n",
    "    ),\n",
    ").drop(\"newStudyType\")\n",
    "#### fixed\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        fixedIndex.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns).persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ").persist()\n",
    "\n",
    "resolvedColoc = (\n",
    "    (\n",
    "        newColoc.withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"inner\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "datasource_filter = [\n",
    "    \"ot_genetics_portal\",\n",
    "    \"gwas_credible_sets\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]\n",
    "\n",
    "assessment, evidences, actionType, oncolabel = temporary_directionOfEffect(\n",
    "    path, datasource_filter\n",
    ")\n",
    "\n",
    "drugApproved = (\n",
    "    spark.read.parquet(\"gs://ot-team/irene/l2g/validation/chembl_w_flags\")\n",
    "    .drop(\"clinicalTrialId\", \"isComplex\")\n",
    "    .withColumn(\n",
    "        \"isApproved\",\n",
    "        F.when(F.col(\"isApproved\") == \"true\", F.lit(1)).otherwise(F.lit(0)),\n",
    "    )\n",
    "    .distinct()\n",
    ")\n",
    "\n",
    "analysis_chembl_indication = (\n",
    "    discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\") == \"chembl\"))\n",
    "        .join(\n",
    "            drugApproved.filter(F.col(\"isApproved\") == 1),\n",
    "            on=[\"targetId\", \"diseaseId\", \"drugId\"],\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approvedDrug\",\n",
    "            F.max(F.col(\"isApproved\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\", \"approvedDrug\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "    )\n",
    "    .filter(F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    .drop(\n",
    "        \"coherencyDiagonal\", \"coherencyOneCell\", \"noEvaluable\", \"GoF_risk\", \"LoF_risk\"\n",
    "    )\n",
    "    .withColumnRenamed(\"GoF_protect\", \"drugGoF_protect\")\n",
    "    .withColumnRenamed(\"LoF_protect\", \"drugLoF_protect\")\n",
    "    .withColumn(\n",
    "        \"approved\",\n",
    "        F.when(F.col(\"approvedDrug\") == 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"newPhases\",\n",
    "        F.when(F.col(\"approvedDrug\") == 1, F.lit(4)).when(\n",
    "            F.col(\"approvedDrug\").isNull(),\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(3)).otherwise(\n",
    "                F.col(\"maxClinPhase\")\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "chemblAssoc = (\n",
    "    discrepancifier(\n",
    "        assessment.filter(\n",
    "            (F.col(\"datasourceId\") == \"chembl\")\n",
    "            & (F.col(\"homogenized\") != \"noEvaluable\")\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(\"clinicalPhase\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .count()\n",
    "    )\n",
    "    .filter(F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    .drop(\n",
    "        \"coherencyDiagonal\", \"coherencyOneCell\", \"noEvaluable\", \"GoF_risk\", \"LoF_risk\"\n",
    "    )\n",
    "    .withColumnRenamed(\"GoF_protect\", \"drugGoF_protect\")\n",
    "    .withColumnRenamed(\"LoF_protect\", \"drugLoF_protect\")\n",
    ")\n",
    "\n",
    "benchmark = (\n",
    "    (\n",
    "        resolvedColoc.filter(F.col(\"betaGwas\") < 0)\n",
    "        .join(  ### select just GWAS giving protection\n",
    "            analysis_chembl_indication, on=[\"targetId\", \"diseaseId\"], how=\"inner\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"AgreeDrug\",\n",
    "            F.when(\n",
    "                (F.col(\"drugGoF_protect\").isNotNull())\n",
    "                & (F.col(\"colocDoE\") == \"GoF_protect\"),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"drugLoF_protect\").isNotNull())\n",
    "                & (F.col(\"colocDoE\") == \"LoF_protect\"),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "    ).filter(\n",
    "        F.col(\"name\") != \"COVID-19\"\n",
    "    )  #### remove COVID-19 associations\n",
    ").join(biosample.select(\"biosampleId\", \"biosampleName\"), on=\"biosampleId\", how=\"left\")\n",
    "\n",
    "#### Analysis\n",
    "\n",
    "#### 1 Build a dictionary with the distinct values as key and column names as value\n",
    "variables_study = [\"projectId\", \"biosampleName\", \"rightStudyType\", \"colocDoE\"]\n",
    "\n",
    "# Initialize an empty dictionary\n",
    "disdic = {}\n",
    "\n",
    "# Iterate over the list of column names\n",
    "for col_name in variables_study:\n",
    "    # Extract distinct values for the column\n",
    "    distinct_values = benchmark.select(col_name).distinct().collect()\n",
    "\n",
    "    # Populate the dictionary\n",
    "    for row in distinct_values:\n",
    "        distinct_value = row[col_name]\n",
    "        if distinct_value is not None:  # Exclude None (null) values\n",
    "            disdic[distinct_value] = col_name\n",
    "\n",
    "####2 Define agregation function\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio\n",
    "from pyspark.sql.types import *\n",
    "\n",
    "\n",
    "def convertTuple(tup):\n",
    "    st = \",\".join(map(str, tup))\n",
    "    return st\n",
    "\n",
    "\n",
    "#####3 run in a function\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "    wComparison = Window.partitionBy(comparisonColumn)\n",
    "    wPrediction = Window.partitionBy(predictionColumn)\n",
    "    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\n",
    "    results = []\n",
    "    # uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"dataset\", F.lit(data))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        # .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\"comparisonColumn\", F.lit(comparisonColumn))\n",
    "        .withColumn(\"predictionColumnValue\", F.lit(predictionColumn))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"dataset\",\n",
    "            \"comparisonColumn\",\n",
    "            \"predictionColumnValue\",\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "    \"\"\"\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + comparisonType\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + comparisonType\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    path = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + comparisonType\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    print(path)\n",
    "    ### making analysis\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "    results.extend(\n",
    "        [\n",
    "            comparisonType,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            # studies,\n",
    "            # tissues,\n",
    "            path,\n",
    "        ]\n",
    "    )\n",
    "    return results\n",
    "\n",
    "\n",
    "#### 3 Loop over different datasets (as they will have different rows and columns)\n",
    "\n",
    "\n",
    "def comparisons_df_iterative(disdic, projectId):\n",
    "    toAnalysis = [(key, value) for key, value in disdic.items() if value == projectId]\n",
    "    schema = StructType(\n",
    "        [\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "            StructField(\"comparisonType\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    comparisons = spark.createDataFrame(toAnalysis, schema=schema)\n",
    "    ### include all the columns as predictor\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"nPhase4\", \"clinical\"),\n",
    "            (\"nPhase>=3\", \"clinical\"),\n",
    "            (\"nPhase>=2\", \"clinical\"),\n",
    "            (\"nPhase>=1\", \"clinical\"),\n",
    "            (\"approved\", \"clinical\"),\n",
    "            # (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "print(\"created full_data and lists\")\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "listado = []\n",
    "result_all = []\n",
    "today_date = str(date.today())\n",
    "variables_study = [\"projectId\", \"biosampleName\", \"rightStudyType\", \"colocDoE\"]\n",
    "\n",
    "print(\"looping for variables_study\")\n",
    "\n",
    "for variable in variables_study:\n",
    "    print(\"analysing\", variable)\n",
    "    #### build list of comparison and prediction columns\n",
    "    rows = comparisons_df_iterative(disdic, variable)\n",
    "    #### prepare aggregation depending on the variable problem\n",
    "    window_spec = Window.partitionBy(\"targetId\", \"diseaseId\", variable).orderBy(\n",
    "        F.col(\"pValueExponent\").asc()\n",
    "    )  ### ignore nulls aded 29.01.2025\n",
    "    #### take directionality from lowest p value\n",
    "    bench2 = (\n",
    "        benchmark.withColumn(\n",
    "            \"agree_lowestPval\", F.first(\"AgreeDrug\", ignorenulls=True).over(window_spec)\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\", \"approved\", \"newPhases\")\n",
    "        .pivot(variable)\n",
    "        .agg(F.collect_set(\"agree_lowestPVal\"))\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(  ###  new phases extracted from aproved label\n",
    "            \"nPhase4\",\n",
    "            F.when(F.col(\"newPhases\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"nPhase>=3\",\n",
    "            F.when(F.col(\"newPhases\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"nPhase>=2\",\n",
    "            F.when(F.col(\"newPhases\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"nPhase>=1\",\n",
    "            F.when(F.col(\"newPhases\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approved\",\n",
    "            F.when(F.col(\"approved\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### debugging doe fro "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-01-30 11:50:25.301146\n",
      "Analysis started on 2025-01-30 at  2025-01-30 11:50:25.301146\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/30 11:50:32 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read spark files\n",
      "fixing scXQTL and XQTL studies\n",
      "fixed scXQTL and XQTL studies\n",
      "creating new coloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating new gwasResolvedColoc\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'tuple' object has no attribute 'persist'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_905774/1881504754.py\u001b[0m in \u001b[0;36m<cell line: 129>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    127\u001b[0m ]\n\u001b[1;32m    128\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 129\u001b[0;31m \u001b[0mprueba_assessment\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemporary_directionOfEffect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatasource_filter\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpersist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    130\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    131\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Moving to step 2\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'tuple' object has no attribute 'persist'"
     ]
    }
   ],
   "source": [
    "#### 11.12.2024\n",
    "#######\n",
    "##########     ATENTION\n",
    "#### change code to work with generated dataframe instead of reading the parquet\n",
    "\n",
    "\"\"\"\n",
    "This scripts run Odds ratio analysis for DoE and \n",
    "genetic information on drug clinical success\n",
    "\n",
    "\"\"\"\n",
    "from functions import (\n",
    "    discrepancifier,\n",
    "    build_gwasResolvedColoc,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "today_date = str(date.today())\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "print(\"Analysis started on \" + today_date + \" at \", c)\n",
    "\"\"\"\n",
    "#coloc = spark.read.parquet(\n",
    "#    \"gs://genetics-portal-dev-data/22.09.1/outputs/v2d_coloc\"\n",
    "#).filter(F.col(\"right_type\") != \"gwas\")\n",
    "\"\"\"\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "evidences = (\n",
    "    spark.read.parquet(f\"{path}evidence\")\n",
    "    .filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "                \"gene_burden\",\n",
    "                \"eva\",\n",
    "                \"eva_somatic\",\n",
    "                \"gene2phenotype\",\n",
    "                \"orphanet\",\n",
    "                \"cancer_gene_census\",\n",
    "                \"intogen\",\n",
    "                \"impc\",\n",
    "                \"chembl\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "gwasResolvedColoc = build_gwasResolvedColoc(path)\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )  ## added 30.01.2025\n",
    "    .select(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "### datasource_filter for temporaryDoE\n",
    "datasource_filter = [\n",
    "    \"ot_genetics_portal\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/01/30 11:53:58 WARN CacheManager: Asked to cache already cached data.\n",
      "25/01/30 11:53:59 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "un, dos, tre, cuatro = temporary_directionOfEffect(path, datasource_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 29:>                                                         (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+---------------+-------------+-------------------+--------+----------+------+-----------+----------+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+-------------+----------+--------------------+------------------+--------+-------------+---------------------+--------------+-----------------+--------+----------------+---------------+----------+--------+-------------------+-------------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-------------------+-----------------+------------------------+-----------------------+-----------------------------+---------------------+--------------------+----------------------------+-------------------+--------------------+------------+--------------------------------+--------------------------------+--------------+--------------+--------+------------------------+----------------------------------+---------------------------+------+-----------------+----------------+------------------+---------+----------------------+---------------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+----------------+------------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----------+-------------------+----+----------------------------+-------------------+-------------------------------------+------------------------------+-------------+---------------+-----------+-------------+--------------------+-------------------+-------------+----------------+------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+\n",
      "|      datasourceId|       targetId|alleleOrigins|allelicRequirements|ancestry|ancestryId|assays|assessments|      beta|betaConfidenceIntervalLower|betaConfidenceIntervalUpper|biologicalModelAllelicComposition|biologicalModelGeneticBackground|biologicalModelId|biomarkerList|biomarkerName|biomarkers|biosamplesFromSource|cellLineBackground|cellType|clinicalPhase|clinicalSignificances|clinicalStatus|cohortDescription|cohortId|cohortPhenotypes|cohortShortName|confidence|contrast|crisprScreenLibrary|         datatypeId|diseaseCellLines|   diseaseFromSource|diseaseFromSourceId|diseaseFromSourceMappedId|diseaseModelAssociatedHumanPhenotypes|diseaseModelAssociatedModelPhenotypes|drugFromSource|drugId|drugResponse|geneInteractionType|geneticBackground|geneticInteractionPValue|geneticInteractionScore|interactingTargetFromSourceId|interactingTargetRole|          literature|log2FoldChangePercentileRank|log2FoldChangeValue|      mutatedSamples|   OddsRatio|oddsRatioConfidenceIntervalLower|oddsRatioConfidenceIntervalUpper|pValueExponent|pValueMantissa|pathways|phenotypicConsequenceFDR|phenotypicConsequenceLogFoldChange|phenotypicConsequencePValue|pmcIds|primaryProjectHit|primaryProjectId|projectDescription|projectId|publicationFirstAuthor|publicationYear|reactionId|reactionName|releaseDate|releaseVersion|resourceScore| sex|significantDriverMethods|statisticalMethod|statisticalMethodOverview|statisticalTestTail|studyCases|studyCasesWithQualifyingVariants|         studyId|studyLocusId|studyOverview|studySampleSize|studyStartDate|studyStopReason|studyStopReasonCategories|targetFromSource|targetFromSourceId|targetInModel|targetInModelEnsemblId|targetInModelMgiId|targetModulation|targetRole|textMiningSentences|urls|variantAminoacidDescriptions|variantFromSourceId|variantFunctionalConsequenceFromQtlId|variantFunctionalConsequenceId|variantHgvsId|      variantId|variantRsId|    diseaseId|                  id|              score|variantEffect|directionOnTrait|          sourceId|target_id|approvedSymbol|description|description_splited|TSorOncogene|targetId2|drugId2|actionType|     inhibitors_list|     activators_list|intogen_function|intogenAnnot|homogenized|\n",
      "+------------------+---------------+-------------+-------------------+--------+----------+------+-----------+----------+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+-------------+----------+--------------------+------------------+--------+-------------+---------------------+--------------+-----------------+--------+----------------+---------------+----------+--------+-------------------+-------------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-------------------+-----------------+------------------------+-----------------------+-----------------------------+---------------------+--------------------+----------------------------+-------------------+--------------------+------------+--------------------------------+--------------------------------+--------------+--------------+--------+------------------------+----------------------------------+---------------------------+------+-----------------+----------------+------------------+---------+----------------------+---------------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+----------------+------------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----------+-------------------+----+----------------------------+-------------------+-------------------------------------+------------------------------+-------------+---------------+-----------+-------------+--------------------+-------------------+-------------+----------------+------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+\n",
      "|ot_genetics_portal|ENSG00000000460|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|genetic_association|            null|Atrial fibrillati...|               null|              EFO_0003911|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|          [36653562]|                        null|               null|                null|1.0986805941|                    1.0667732734|                    1.1315422667|           -10|          3.88|    null|                    null|                              null|                       null|  null|             null|            null|              null|  FINNGEN|            FINNGEN_R6|           2022|      null|        null|       null|          null| 0.1147234887|null|                    null|             null|                     null|               null|      null|                            null|FINNGEN_R6_I9_AF|        null|         null|         164491|          null|           null|                     null|            null|   ENSG00000000460|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                                 null|                    SO_0001628|         null|1_170329868_A_G|  rs4491109|  EFO_0003911|9238fd0c3202953c6...|       0.1147234887|  noEvaluable|            risk|ot_genetics_portal|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|noEvaluable|\n",
      "|              impc|ENSG00000000938|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Fgr<Mhdaali18>/Fg...|            C3HeB/FeJ-Fgr<Mhd...|      MGI:3688148|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Angioosteohypotro...|        ORPHA:75508|            MONDO_0019156|                 [{HP:0002753, Thi...|                 [{MP:0002111, abn...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|          [16964445]|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        46.65|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000000938|          Fgr|    ENSMUSG00000028874|         MGI:95527|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0019156|16870dc14984b399d...|0.46649999999999997|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000000971|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cfh<tm1Mbo>/Cfh<t...|            involves: 129/Sv ...|      MGI:2662550|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Lipodystrophy, Pa...|        OMIM:608709|            MONDO_0012104|                 [{HP:0000093, Pro...|                 [{MP:0002083, pre...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[17517971, 120919...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        53.16|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000000971|          Cfh|    ENSMUSG00000026365|         MGI:88385|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0012104|60554f1565698b146...|             0.5316|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001617|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Sema3f<tm1b(EUCOM...|                     C57BL/6NTac|      MGI:5636922|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Atypical Pantothe...|       ORPHA:216873|            MONDO_0016305|                 [{HP:0100543, Cog...|                 [{MP:0001488, inc...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        49.66|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001617|       Sema3f|    ENSMUSG00000034684|       MGI:1096347|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0016305|4168a3b1a3fa42a3f...|             0.4966|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm1Unc>/Cftr...|            B6.129P2-Cftr<tm1...|      MGI:3689408|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Pulmonary Fibrosi...|        OMIM:616371|              EFO_0000768|                 [{HP:0002206, Pul...|                 [{MP:0014028, dec...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[7560099, 1355249...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        41.34|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|  EFO_0000768|2784b9972ca3a638e...|0.41340000000000005|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm1Unc>/Cftr...|            B6.129P2-Cftr<tm1...|      MGI:3689408|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Idiopathic Pulmon...|         ORPHA:2032|              EFO_0000768|                 [{HP:0002206, Pul...|                 [{MP:0014028, dec...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[7560099, 1355249...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        43.42|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|  EFO_0000768|4e1df7a687e71e817...|0.43420000000000003|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm1Kth>/Cftr...|            B6.129S7-Cftr<tm1...|      MGI:3689378|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|46,Xy Gonadal Dys...|       ORPHA:168563|            MONDO_0011766|                 [{HP:0000786, Pri...|                 [{MP:0014028, dec...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[29924856, 756009...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        54.53|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0011766|0cc68e2a918c55a77...|             0.5453|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm1Unc>/Cftr...|            B6.129P2-Cftr<tm1...|      MGI:2177529|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|46,Xy Gonadal Dys...|       ORPHA:168563|            MONDO_0011766|                 [{HP:0000786, Pri...|                 [{MP:0001923, red...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[29924856, 159215...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        46.97|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0011766|01ffce73002308923...|             0.4697|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm1Kth>/Cftr...|            B6.129S7-Cftr<tm1...|      MGI:3689378|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|46,Xy Gonadal Dys...|        OMIM:607080|            MONDO_0011766|                 [{HP:0000786, Pri...|                 [{MP:0014028, dec...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[29924856, 756009...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        51.51|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0011766|7c5755855498187e7...|             0.5151|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm2Mrc>/Cftr...|            B6.129S6-Cftr<tm2...|      MGI:3689380|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Diamond-Blackfan ...|        OMIM:612527|             Orphanet_124|                 [{HP:0001875, Neu...|                 [{MP:0014028, dec...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[18802965, 105701...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        42.51|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null| Orphanet_124|4858d33c68e59d7d0...|             0.4251|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001626|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cftr<tm2Mrc>/Cftr...|            B6.129S6-Cftr<tm2...|      MGI:3689380|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Diamond-Blackfan ...|        OMIM:618310|             Orphanet_124|                 [{HP:0001875, Neu...|                 [{MP:0014028, dec...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[18802965, 105701...|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        42.28|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null| Orphanet_124|cf0648f0223c44a15...|             0.4228|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000001630|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Cyp51<tm1.1Bfro>/...|            involves: 129P2/O...|      MGI:5014092|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Isolated Tibial H...|        ORPHA:93322|            MONDO_0010144|                 [{HP:0009556, Abs...|                 [{MP:0000431, abs...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|          [21705796]|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        47.95|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000001630|        Cyp51|    ENSMUSG00000001467|        MGI:106040|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0010144|039f580fea3238814...|0.47950000000000004|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|              impc|ENSG00000002746|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Hecw1<tm2b(EUCOMM...|                     C57BL/6NTac|      MGI:6120795|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|         Dystonia 27|        OMIM:616411|            MONDO_0015494|                 [{HP:0002174, Pos...|                 [{MP:0000745, tre...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        65.42|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000002746|        Hecw1|    ENSMUSG00000021301|       MGI:2444115|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|MONDO_0015494|e3ab37882c8a05dac...|             0.6542|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "|ot_genetics_portal|ENSG00000002822|         null|               null|    null|      null|  null|       null| -0.012497|              -0.0159520684|              -0.0090419316|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|genetic_association|            null|Arm fat mass (right)|               null|              EFO_0004995|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|           -12|          1.35|    null|                    null|                              null|                       null|  null|             null|            null|              null|    NEALE|          UKB Neale v2|           2018|      null|        null|       null|          null| 0.4622868001|null|                    null|             null|                     null|               null|      null|                            null|NEALE2_23120_raw|        null|         null|         354736|          null|           null|                     null|            null|   ENSG00000002822|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                           SO_0002316|                    SO_0001627|         null|  7_2068401_G_C| rs17132130|  EFO_0004995|70ec605bd7815bf28...|       0.4622868001|          LoF|         protect|ot_genetics_portal|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|LoF_protect|\n",
      "|ot_genetics_portal|ENSG00000002822|         null|               null|    null|      null|  null|       null|-0.0144155|              -0.0182764452|              -0.0105545548|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|genetic_association|            null| Arm fat mass (left)|               null|              EFO_0004995|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|           -13|          2.52|    null|                    null|                              null|                       null|  null|             null|            null|              null|    NEALE|          UKB Neale v2|           2018|      null|        null|       null|          null| 0.4790070951|null|                    null|             null|                     null|               null|      null|                            null|NEALE2_23124_raw|        null|         null|         354673|          null|           null|                     null|            null|   ENSG00000002822|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                           SO_0002316|                    SO_0001627|         null|  7_2068401_G_C| rs17132130|  EFO_0004995|5fc2eaadbcf7a62bf...|       0.4790070951|          LoF|         protect|ot_genetics_portal|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|LoF_protect|\n",
      "|ot_genetics_portal|ENSG00000002822|         null|               null|    null|      null|  null|       null|-0.0273158|              -0.0358954648|              -0.0187361352|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|genetic_association|            null| Leg fat mass (left)|               null|              EFO_0004995|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|           -10|        4.3744|    null|                    null|                              null|                       null|  null|             null|            null|              null|    NEALE|          UKB Neale v2|           2018|      null|        null|       null|          null| 0.4714113772|null|                    null|             null|                     null|               null|      null|                            null|NEALE2_23116_raw|        null|         null|         354788|          null|           null|                     null|            null|   ENSG00000002822|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                           SO_0002316|                    SO_0001627|         null|  7_2068401_G_C| rs17132130|  EFO_0004995|544f0ce61fbb38db0...|       0.4714113772|          LoF|         protect|ot_genetics_portal|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|LoF_protect|\n",
      "|ot_genetics_portal|ENSG00000002822|         null|               null|    null|      null|  null|       null|-0.0184386|              -0.0248090116|              -0.0120681884|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|genetic_association|            null|Leg predicted mas...|               null|              EFO_0004995|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|            -8|       1.40374|    null|                    null|                              null|                       null|  null|             null|            null|              null|    NEALE|          UKB Neale v2|           2018|      null|        null|       null|          null| 0.3254676163|null|                    null|             null|                     null|               null|      null|                            null|NEALE2_23118_raw|        null|         null|         354766|          null|           null|                     null|            null|   ENSG00000002822|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                           SO_0002316|                    SO_0001627|         null|  7_2068401_G_C| rs17132130|  EFO_0004995|3f3061bf23aa899db...|       0.3254676163|          LoF|         protect|ot_genetics_portal|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|LoF_protect|\n",
      "|ot_genetics_portal|ENSG00000002822|         null|               null|    null|      null|  null|       null|-0.0195723|               -0.026366346|               -0.012778254|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|genetic_association|            null|Leg fat-free mass...|               null|              EFO_0004995|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|                null|                        null|               null|                null|        null|                            null|                            null|            -8|       1.63991|    null|                    null|                              null|                       null|  null|             null|            null|              null|    NEALE|          UKB Neale v2|           2018|      null|        null|       null|          null| 0.2011062503|null|                    null|             null|                     null|               null|      null|                            null|NEALE2_23117_raw|        null|         null|         354771|          null|           null|                     null|            null|   ENSG00000002822|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                           SO_0002316|                    SO_0001627|         null|  7_2068401_G_C| rs17132130|  EFO_0004995|0719b5e16bec43f21...|       0.2011062503|          LoF|         protect|ot_genetics_portal|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|LoF_protect|\n",
      "|cancer_gene_census|ENSG00000002834|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|                             null|                            null|             null|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|   somatic_mutation|            null|                null|               null|              EFO_0006545|                                 null|                                 null|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|          [26496030]|                        null|               null|[{SO_0001059, 1.0...|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|         0.25|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000002834|         null|                  null|              null|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|  EFO_0006545|75f6e7d9ee37285cf...|               0.25|  noEvaluable|            risk|cancer_gene_census|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|noEvaluable|\n",
      "|              impc|ENSG00000002834|         null|               null|    null|      null|  null|       null|      null|                       null|                       null|             Lasp1<Tg(Col1a1-l...|               involves: C57BL/6|      MGI:4366487|         null|         null|      null|                null|              null|    null|         null|                     |          null|             null|    null|            null|           null|      null|    null|               null|       animal_model|            null|Isolated Klippel-...|         ORPHA:2345|              EFO_0700054|                 [{HP:0002650, Sco...|                 [{MP:0004606, abs...|          null|  null|        null|               null|             null|                    null|                   null|                         null|                 null|[12438746, 19378260]|                        null|               null|                null|        null|                            null|                            null|          null|          null|    null|                    null|                              null|                       null|  null|             null|            null|              null|     null|                  null|           null|      null|        null|       null|          null|        45.55|null|                    null|             null|                     null|               null|      null|                            null|            null|        null|         null|           null|          null|           null|                     null|            null|   ENSG00000002834|        Lasp1|    ENSMUSG00000038366|        MGI:109656|            null|      null|               null|null|                        null|               null|                                 null|                          null|         null|           null|       null|  EFO_0700054|d15addcf8dee78bbc...|0.45549999999999996|          LoF|            risk|              impc|     null|          null|       null|               null|        null|     null|   null|      null|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            null|           0|   LoF_risk|\n",
      "+------------------+---------------+-------------+-------------------+--------+----------+------+-----------+----------+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+-------------+----------+--------------------+------------------+--------+-------------+---------------------+--------------+-----------------+--------+----------------+---------------+----------+--------+-------------------+-------------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-------------------+-----------------+------------------------+-----------------------+-----------------------------+---------------------+--------------------+----------------------------+-------------------+--------------------+------------+--------------------------------+--------------------------------+--------------+--------------+--------+------------------------+----------------------------------+---------------------------+------+-----------------+----------------+------------------+---------+----------------------+---------------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+----------------+------------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----------+-------------------+----+----------------------------+-------------------+-------------------------------------+------------------------------+-------------+---------------+-----------+-------------+--------------------+-------------------+-------------+----------------+------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "un.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "assessment = prueba_assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "print(\"defining non propagated,propagated and analysis_drugs functions\")\n",
    "\n",
    "\n",
    "def analysis_nonPropagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "\n",
    "\n",
    "def analysis_drugs(assessment, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "        .join(\n",
    "            drugApproved.filter(F.col(\"isApproved\") == 1),\n",
    "            on=[\"targetId\", \"diseaseId\", \"drugId\"],\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approvedDrug\",\n",
    "            F.max(F.col(\"isApproved\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approved\",\n",
    "            F.when(F.col(\"approvedDrug\") == 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\", \"approved\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "##### approved dataset added 30.01.2025\n",
    "drugApproved = (\n",
    "    spark.read.parquet(\"gs://ot-team/irene/l2g/validation/chembl_w_flags\")\n",
    "    .drop(\"clinicalTrialId\", \"isComplex\")\n",
    "    .withColumn(\n",
    "        \"isApproved\",\n",
    "        F.when(F.col(\"isApproved\") == \"true\", F.lit(1)).otherwise(F.lit(0)),\n",
    "    )\n",
    "    .distinct()\n",
    ")\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment, chembl_ds)\n",
    "\n",
    "#######\n",
    "## include here the analysis\n",
    "#######\n",
    "\n",
    "analysisDatasources = []\n",
    "\n",
    "print(\"defining full_analysis_propagation\")\n",
    "\n",
    "\n",
    "def full_analysis_propagation(\n",
    "    assessment, analysisDatasources, analysis_chembl, terminated_array, diseaseTA\n",
    "):\n",
    "    return (\n",
    "        analysis_propagated(assessment, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"approved\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        #### Should remove the coherencyDiagonal.isNotNull()\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approved\",\n",
    "            F.when(F.col(\"approved\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "#####\n",
    "## no propag\n",
    "#####\n",
    "print(\"defining full analysis no propagation\")\n",
    "\n",
    "\n",
    "def full_analysis_noPropagation(\n",
    "    assessment, analysisDatasources, analysis_chembl, terminated_array, diseaseTA\n",
    "):\n",
    "    return (\n",
    "        analysis_nonPropagated(assessment, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approved\",\n",
    "            F.when(F.col(\"approved\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "print(\"moving to Step 3\")\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "c = datetime.now()\n",
    "print(\"starting dictionaries at\", c)\n",
    "\n",
    "#### continue here on 10.07.2024\n",
    "\n",
    "## 1nd dictionary\n",
    "dfs_dict = {}  ### checked and changed on 01.06.2023\n",
    "dfs_dict_propag = {}\n",
    "\n",
    "\n",
    "wocgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    # \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "]\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "\n",
    "germline_list = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    # \"ot_genetics_portal\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic_list = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "# assessment = prueba_assessment.filter(F.col(\"datasourceId\").isin(datasources_analysis))\n",
    "\n",
    "\n",
    "def dataset_builder(assessment, value, analysis_chembl, terminated_array, diseaseTA):\n",
    "    nonPropagated = full_analysis_noPropagation(\n",
    "        assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "    )\n",
    "    propagated = full_analysis_propagation(\n",
    "        assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "    )\n",
    "    return (\n",
    "        # Non propagation\n",
    "        ## All\n",
    "        nonPropagated,\n",
    "        ## Other\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        nonPropagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "        # Propagation\n",
    "        ## All\n",
    "        propagated,\n",
    "        ## Other\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        propagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "    )\n",
    "\n",
    "\n",
    "for value in datasource_list:\n",
    "    print(value)\n",
    "    if value == \"WOcgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment, wocgc_list, analysis_chembl, terminated_array, diseaseTA\n",
    "        )\n",
    "    elif value == \"germline\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment,\n",
    "            germline_list,\n",
    "            analysis_chembl,\n",
    "            terminated_array,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    elif value == \"somatic\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment,\n",
    "            somatic_list,\n",
    "            analysis_chembl,\n",
    "            terminated_array,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df() -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    comparisons = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"hasGeneticEvidence\", \"byDatatype\"),\n",
    "            (\"diagonalYes\", \"byDatatype\"),\n",
    "            (\"oneCellYes\", \"byDatatype\"),\n",
    "        ],\n",
    "        schema=StructType(\n",
    "            [\n",
    "                StructField(\"comparison\", StringType(), True),\n",
    "                StructField(\"comparisonType\", StringType(), True),\n",
    "            ]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "            (\"approved\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "results = []\n",
    "\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "\n",
    "    wComparison = Window.partitionBy(comparisonColumn)\n",
    "    wPrediction = Window.partitionBy(predictionColumn)\n",
    "    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\n",
    "\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    filePath = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            comparisonType,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            filePath,\n",
    "        ]\n",
    "    )\n",
    "    return results\n",
    "\n",
    "\n",
    "c = datetime.now()\n",
    "\n",
    "print(\"start doing aggregations and writing\")\n",
    "today_date = str(date.today())\n",
    "aggSetups_original = comparisons_df()\n",
    "listado = []\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "\n",
    "for key, df in dfs_dict.items():\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"non propagated files wroten succesfully at\", c)\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "for key, df in dfs_dict_propag.items():\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### make the DoE predicting drug success for T-D with multiple MoA in drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-02-03 16:22:40.625638\n",
      "Analysis started on 2025-02-03 at  2025-02-03 16:22:40.625638\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 16:22:42 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read spark files\n",
      "fixing scXQTL and XQTL studies\n",
      "fixed scXQTL and XQTL studies\n",
      "creating new coloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 16:23:28 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating new gwasResolvedColoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 16:23:29 WARN CacheManager: Asked to cache already cached data.\n",
      "25/02/03 16:23:30 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moving to step 2\n",
      "defining non propagated,propagated and analysis_drugs functions\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there are the same numbers of T-D: 1571\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2032:========================================>          (157 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there are same numbers of T-D: 1571\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "#### 11.12.2024\n",
    "#######\n",
    "##########     ATENTION\n",
    "#### change code to work with generated dataframe instead of reading the parquet\n",
    "\n",
    "\"\"\"\n",
    "This scripts run Odds ratio analysis for DoE and \n",
    "genetic information on drug clinical success\n",
    "\n",
    "\"\"\"\n",
    "from functions import (\n",
    "    discrepancifier,\n",
    "    build_gwasResolvedColoc,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "today_date = str(date.today())\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "print(\"Analysis started on \" + today_date + \" at \", c)\n",
    "\"\"\"\n",
    "#coloc = spark.read.parquet(\n",
    "#    \"gs://genetics-portal-dev-data/22.09.1/outputs/v2d_coloc\"\n",
    "#).filter(F.col(\"right_type\") != \"gwas\")\n",
    "\"\"\"\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    ")\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "evidences = (\n",
    "    spark.read.parquet(f\"{path}evidence\")\n",
    "    .filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "                \"gene_burden\",\n",
    "                \"eva\",\n",
    "                \"eva_somatic\",\n",
    "                \"gene2phenotype\",\n",
    "                \"orphanet\",\n",
    "                \"cancer_gene_census\",\n",
    "                \"intogen\",\n",
    "                \"impc\",\n",
    "                \"chembl\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "gwasResolvedColoc = build_gwasResolvedColoc(path)\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )  ## added 30.01.2025\n",
    "    .select(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "### datasource_filter for temporaryDoE\n",
    "datasource_filter = [\n",
    "    \"ot_genetics_portal\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]\n",
    "\n",
    "prueba_assessment, evidences, actionType, oncolabel = temporary_directionOfEffect(\n",
    "    path, datasource_filter\n",
    ")\n",
    "\n",
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "assessment = prueba_assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "print(\"defining non propagated,propagated and analysis_drugs functions\")\n",
    "\n",
    "\n",
    "def analysis_nonPropagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "drugApproved = (  ### added on 30.01.2025\n",
    "    spark.read.parquet(\"gs://ot-team/irene/l2g/validation/chembl_w_flags\")\n",
    "    .drop(\"clinicalTrialId\", \"isComplex\")\n",
    "    .withColumn(\n",
    "        \"isApproved\",\n",
    "        F.when(F.col(\"isApproved\") == \"true\", F.lit(1)).otherwise(F.lit(0)),\n",
    "    )\n",
    "    .distinct()\n",
    ")\n",
    "\n",
    "\n",
    "def analysis_drugs(assessment, chembl_ds, drugApproved):\n",
    "    return discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "        .join(\n",
    "            drugApproved.filter(F.col(\"isApproved\") == 1),\n",
    "            on=[\"targetId\", \"diseaseId\", \"drugId\"],\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approvedDrug\",\n",
    "            F.max(F.col(\"isApproved\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approved\",\n",
    "            F.when(F.col(\"approvedDrug\") == 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\", \"approved\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment, chembl_ds, drugApproved)\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "\n",
    "def analysis_drugs_dispar(assessment, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "        .withColumn(\n",
    "            \"maxClinPhase_moa\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.first(F.col(\"maxClinPhase_moa\")).alias(\"doe\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl_dispar = analysis_drugs_dispar(assessment, chembl_ds)\n",
    "\n",
    "\n",
    "def analysis_drugs_dispar_approved(assessment, chembl_ds, drugApproved):\n",
    "    return discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "        .join(\n",
    "            drugApproved.filter(F.col(\"isApproved\") == 1),\n",
    "            on=[\"targetId\", \"diseaseId\", \"drugId\"],\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approvedDrug\",\n",
    "            F.max(F.col(\"isApproved\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"approved\",\n",
    "            F.when(F.col(\"approvedDrug\") == 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.first(F.col(\"approved\")).alias(\"doe\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "drugApp = (\n",
    "    analysis_drugs_dispar_approved(assessment, chembl_ds, drugApproved)\n",
    "    .filter(F.col(\"coherencyDiagonal\") == \"dispar\")\n",
    "    .withColumnRenamed(\"GoF_protect\", \"GoF_protectApp\")\n",
    "    .withColumnRenamed(\"LoF_protect\", \"LoF_protectApp\")\n",
    ").select(\"targetId\", \"diseaseId\", \"maxClinPhase\", \"GoF_protectApp\", \"LoF_protectApp\")\n",
    "\n",
    "\n",
    "print(\"there are the same numbers of T-D:\", drugApp.count())\n",
    "\n",
    "disparDrug = analysis_chembl_dispar.filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ").selectExpr(\n",
    "    \"targetId\",\n",
    "    \"diseaseId\",\n",
    "    \"GoF_protect as GoF_protect_ch\",\n",
    "    \"LoF_protect as LoF_protect_ch\",\n",
    ")\n",
    "print(\"there are same numbers of T-D:\", disparDrug.count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "drugsDisparities = discrepancifier(\n",
    "    assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "    .join(\n",
    "        drugApproved.filter(F.col(\"isApproved\") == 1),\n",
    "        on=[\"targetId\", \"diseaseId\", \"drugId\"],\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"approvedDrug\",\n",
    "        F.max(F.col(\"isApproved\")).over(\n",
    "            Window.partitionBy(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"approved\",\n",
    "        F.when(F.col(\"approvedDrug\") == 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"maxClinPhase\",\n",
    "        F.max(F.col(\"clinicalPhase\")).over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "    .pivot(\"approved\")\n",
    "    .agg(F.first(F.col(\"homogenized\")).alias(\"doe\"))\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 17:28:32 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------+-----+\n",
      "|maxClinPhase|count|\n",
      "+------------+-----+\n",
      "|         1.0|   47|\n",
      "|         4.0|   95|\n",
      "|         2.0|   93|\n",
      "|         3.0|   32|\n",
      "+------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "genEvidForDrug = analysis_propagated(assessment, datasource_list)\n",
    "drugDispar.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ").groupBy(\"maxClinPhase\").count().show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### adding warnings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "### load black box warnings\n",
    "warnings = spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/drugWarnings/\"\n",
    ").select(F.explode_outer(\"chemblIds\").alias(\"chemblId\"), \"warningType\", \"toxicityClass\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugWarning = (\n",
    "    warnings.groupBy(\"chemblId\")\n",
    "    .pivot(\"warningType\")\n",
    "    .count()\n",
    "    .withColumnRenamed(\"chemblId\", \"drugId\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------+-----------------+---------+\n",
      "|     chemblId|Black Box Warning|Withdrawn|\n",
      "+-------------+-----------------+---------+\n",
      "|CHEMBL2107840|                1|     null|\n",
      "|CHEMBL1200632|                2|     null|\n",
      "|    CHEMBL852|                4|     null|\n",
      "| CHEMBL188921|             null|       10|\n",
      "|CHEMBL3989522|                1|     null|\n",
      "|CHEMBL1200879|                2|     null|\n",
      "|    CHEMBL468|                2|     null|\n",
      "|CHEMBL2023898|                1|     null|\n",
      "|   CHEMBL1544|                1|     null|\n",
      "|CHEMBL3137358|                1|     null|\n",
      "|CHEMBL2303621|                1|     null|\n",
      "|CHEMBL1201827|                1|     null|\n",
      "|    CHEMBL979|             null|        4|\n",
      "|  CHEMBL87493|                1|        1|\n",
      "|CHEMBL1200585|                1|     null|\n",
      "|  CHEMBL24944|             null|        2|\n",
      "|CHEMBL4594242|             null|        1|\n",
      "|   CHEMBL1233|             null|        5|\n",
      "|CHEMBL1201867|                1|     null|\n",
      "|CHEMBL1201486|                1|     null|\n",
      "+-------------+-----------------+---------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "drugWarning.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugsDisparities = discrepancifier(\n",
    "    assessment.filter(F.col(\"datasourceId\").isin(chembl_ds))\n",
    "    .join(drugWarning, on=\"drugId\", how=\"left\")\n",
    "    .withColumn(\n",
    "        \"BBW\",\n",
    "        F.size(\n",
    "            F.collect_set(F.col(\"`Black Box Warning`\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\", \"drugId\")\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .pivot(\"homogenized\")\n",
    "    .agg(F.first(\"BBW\").alias(\"doe\"))\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------+-----------------+---------+\n",
      "|       drugId|Black Box Warning|Withdrawn|\n",
      "+-------------+-----------------+---------+\n",
      "|CHEMBL2107840|                1|     null|\n",
      "|CHEMBL1200632|                2|     null|\n",
      "|    CHEMBL852|                4|     null|\n",
      "| CHEMBL188921|             null|       10|\n",
      "|CHEMBL3989522|                1|     null|\n",
      "|CHEMBL1200879|                2|     null|\n",
      "|    CHEMBL468|                2|     null|\n",
      "|CHEMBL2023898|                1|     null|\n",
      "|   CHEMBL1544|                1|     null|\n",
      "|CHEMBL3137358|                1|     null|\n",
      "|CHEMBL2303621|                1|     null|\n",
      "|CHEMBL1201827|                1|     null|\n",
      "|    CHEMBL979|             null|        4|\n",
      "|  CHEMBL87493|                1|        1|\n",
      "|CHEMBL1200585|                1|     null|\n",
      "|  CHEMBL24944|             null|        2|\n",
      "|CHEMBL4594242|             null|        1|\n",
      "|   CHEMBL1233|             null|        5|\n",
      "|CHEMBL1201867|                1|     null|\n",
      "|CHEMBL1201486|                1|     null|\n",
      "+-------------+-----------------+---------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "drugWarning.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "drugsDisparitiesWD = discrepancifier(\n",
    "    assessment.filter(F.col(\"datasourceId\").isin(chembl_ds))\n",
    "    .join(drugWarning, on=\"drugId\", how=\"left\")\n",
    "    .withColumn(\n",
    "        \"WD\",\n",
    "        F.size(\n",
    "            F.collect_set(F.col(\"Withdrawn\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\", \"drugId\")\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .pivot(\"homogenized\")\n",
    "    .agg(F.first(\"WD\").alias(\"doe\"))\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 3983:=================================================>  (191 + 8) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+----+----+\n",
      "|coherencyDiagonal|  no| yes|\n",
      "+-----------------+----+----+\n",
      "|           dispar| 194|  73|\n",
      "|         coherent|3284|1243|\n",
      "+-----------------+----+----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "### black bos warnings\n",
    "drugsDisparities.withColumnRenamed(\"GoF_protect\", \"GoF_protect_ch\").withColumnRenamed(\n",
    "    \"LoF_protect\", \"LoF_protect_ch\"\n",
    ").withColumnRenamed(\"coherencyDiagonal\", \"coherencyDiagonal_ch\").drop(\n",
    "    \"GoF_risk\", \"LoF_risk\"\n",
    ").join(\n",
    "    genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\"\n",
    ").filter(\n",
    "    F.col(\"coherencyDiagonal\").isin([\"dispar\", \"coherent\"])\n",
    ").withColumn(\n",
    "    \"bbw\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protect_ch\") > 0) | (F.col(\"LoF_protect_ch\") > 0), F.lit(\"yes\")\n",
    "    ).otherwise(F.lit(\"no\")),\n",
    ").groupBy(\n",
    "    \"coherencyDiagonal\"\n",
    ").pivot(\n",
    "    \"bbw\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 4230:===================================>               (139 + 17) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+----+---+\n",
      "|coherencyDiagonal|  no|yes|\n",
      "+-----------------+----+---+\n",
      "|           dispar| 266|  1|\n",
      "|         coherent|4506| 21|\n",
      "+-----------------+----+---+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "drugsDisparitiesWD.withColumnRenamed(\"GoF_protect\", \"GoF_protect_ch\").withColumnRenamed(\n",
    "    \"LoF_protect\", \"LoF_protect_ch\"\n",
    ").withColumnRenamed(\"coherencyDiagonal\", \"coherencyDiagonal_ch\").drop(\n",
    "    \"GoF_risk\", \"LoF_risk\"\n",
    ").join(\n",
    "    genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\"\n",
    ").filter(\n",
    "    F.col(\"coherencyDiagonal\").isin([\"dispar\", \"coherent\"])\n",
    ").withColumn(\n",
    "    \"wd\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protect_ch\") > 0) | (F.col(\"LoF_protect_ch\") > 0), F.lit(\"yes\")\n",
    "    ).otherwise(F.lit(\"no\")),\n",
    ").groupBy(\n",
    "    \"coherencyDiagonal\"\n",
    ").pivot(\n",
    "    \"wd\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugsDisparities.withColumnRenamed(\"GoF_protect\", \"GoF_protect_ch\").withColumnRenamed(\n",
    "    \"LoF_protect\", \"LoF_protect_ch\"\n",
    ").withColumnRenamed(\"coherencyDiagonal\", \"coherencyDiagonal_ch\").drop(\n",
    "    \"GoF_risk\", \"LoF_risk\"\n",
    ").join(\n",
    "    genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\"\n",
    ").filter(\n",
    "    F.col(\"coherencyDiagonal\").isin([\"dispar\", \"coherent\"])\n",
    ").withColumn(\n",
    "    \"bbw\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protect_ch\") > 0) | (F.col(\"LoF_protect_ch\") > 0), F.lit(\"yes\")\n",
    "    ).otherwise(F.lit(\"no\")),\n",
    ").groupBy(\n",
    "    \"coherencyDiagonal\"\n",
    ").pivot(\n",
    "    \"bbw\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|maxClinPhase|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------------+----------------+\n",
      "|ENSG00000140443|MONDO_0015892|         2.0|           2.0|          null|       null|    null|          1|      13|           dispar|          dispar|\n",
      "|ENSG00000141510|  EFO_0004251|         1.0|          null|           1.0|       null|     186|       null|      14|           dispar|          dispar|\n",
      "|ENSG00000157388|MONDO_0001422|         2.0|          null|           2.0|       null|       1|       null|       1|           dispar|          dispar|\n",
      "|ENSG00000165731|  EFO_0002892|         4.0|          null|           4.0|       null|       1|       null|      26|           dispar|          dispar|\n",
      "|ENSG00000128052|  EFO_0000096|         2.0|          null|           2.0|       null|       2|       null|       1|           dispar|          dispar|\n",
      "|ENSG00000105639|  EFO_0000729|         4.0|          null|           4.0|       null|       1|       null|       1|           dispar|          dispar|\n",
      "|ENSG00000165731|  EFO_1001901|         4.0|          null|           4.0|       null|       1|       null|      11|           dispar|          dispar|\n",
      "|ENSG00000105397|  EFO_0000685|         4.0|          null|           4.0|        291|    null|        113|    null|           dispar|          dispar|\n",
      "|ENSG00000135679|  EFO_1001968|         1.0|          null|           1.0|       null|       1|       null|       3|           dispar|          dispar|\n",
      "|ENSG00000037280|  EFO_0000228|         2.0|          null|           2.0|       null|      15|       null|       2|           dispar|          dispar|\n",
      "|ENSG00000162733|  EFO_0000182|         3.0|          null|           3.0|       null|       1|       null|       1|           dispar|          dispar|\n",
      "|ENSG00000115232|  EFO_0003767|         1.0|          null|           1.0|         69|    null|        165|       1|           dispar|          dispar|\n",
      "|ENSG00000121966|  EFO_0003086|         1.0|           1.0|          null|       null|       1|       null|       1|           dispar|          dispar|\n",
      "|ENSG00000138735|  EFO_0003777|         3.0|          null|           3.0|       null|    null|         64|     667|           dispar|          dispar|\n",
      "|ENSG00000105639|  EFO_0005952|         1.0|          null|           1.0|       null|       3|       null|       2|           dispar|          dispar|\n",
      "|ENSG00000160867|  EFO_0000519|         2.0|          null|           2.0|       null|       1|       null|       2|           dispar|          dispar|\n",
      "|ENSG00000198900|  EFO_0004193|         1.0|          null|           1.0|       null|       1|       null|       1|           dispar|          dispar|\n",
      "|ENSG00000120659|  EFO_0000685|         4.0|          null|           4.0|       null|    null|          6|       1|           dispar|          dispar|\n",
      "|ENSG00000096968|  EFO_0002430|         4.0|          null|           4.0|       null|       1|       null|       4|           dispar|          dispar|\n",
      "|ENSG00000077238|MONDO_0005271|         4.0|          null|           4.0|       null|    null|         43|     104|           dispar|          dispar|\n",
      "+---------------+-------------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "drugDispar.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ").drop(\"noEvaluable\").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 3133:=========================================>         (161 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------+---------------+--------------+--------+-----------+--------+-----------+--------+-----------+-----------------+----------------+--------------------+-------------+------------+------+\n",
      "|    diseaseId|       targetId|       doeDrug|maxPhase|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|             taLabel|taLabelSimple|agreeGenetic|phase4|\n",
      "+-------------+---------------+--------------+--------+-----------+--------+-----------+--------+-----------+-----------------+----------------+--------------------+-------------+------------+------+\n",
      "|MONDO_0005148|ENSG00000112164|LoF_protect_ch|     3.0|          9|    null|       null|       4|          2|         coherent|          dispar|    pancreas disease|        Other|    notAgree|    no|\n",
      "|MONDO_0005148|ENSG00000112164|GoF_protect_ch|     4.0|          9|    null|       null|       4|          2|         coherent|          dispar|    pancreas disease|        Other|       agree|   yes|\n",
      "|MONDO_0007254|ENSG00000198793|LoF_protect_ch|     3.0|       null|       1|       null|    null|       null|         coherent|        coherent|cell proliferatio...|     Oncology|       agree|    no|\n",
      "|MONDO_0007254|ENSG00000198793|GoF_protect_ch|     2.0|       null|       1|       null|    null|       null|         coherent|        coherent|cell proliferatio...|     Oncology|    notAgree|    no|\n",
      "|  EFO_0002618|ENSG00000132170|LoF_protect_ch|     1.0|       null|    null|       null|       1|       null|         coherent|        coherent|cell proliferatio...|     Oncology|    notAgree|    no|\n",
      "|  EFO_0002618|ENSG00000132170|GoF_protect_ch|     2.0|       null|    null|       null|       1|       null|         coherent|        coherent|cell proliferatio...|     Oncology|       agree|    no|\n",
      "|  EFO_0000198|ENSG00000134460|LoF_protect_ch|     2.0|       null|    null|       null|       2|       null|         coherent|        coherent|cell proliferatio...|     Oncology|    notAgree|    no|\n",
      "|  EFO_0000198|ENSG00000134460|GoF_protect_ch|     3.0|       null|    null|       null|       2|       null|         coherent|        coherent|cell proliferatio...|     Oncology|       agree|    no|\n",
      "|MONDO_0011962|ENSG00000091831|LoF_protect_ch|     2.0|       null|       1|       null|    null|       null|         coherent|        coherent|cell proliferatio...|     Oncology|       agree|    no|\n",
      "|MONDO_0011962|ENSG00000091831|GoF_protect_ch|     3.0|       null|       1|       null|    null|       null|         coherent|        coherent|cell proliferatio...|     Oncology|    notAgree|    no|\n",
      "|MONDO_0002009|ENSG00000116329|LoF_protect_ch|     4.0|       null|    null|       null|       1|       null|         coherent|        coherent|nervous system di...|        Other|    notAgree|   yes|\n",
      "|MONDO_0002009|ENSG00000116329|GoF_protect_ch|     2.0|       null|    null|       null|       1|       null|         coherent|        coherent|nervous system di...|        Other|       agree|    no|\n",
      "|  EFO_0000660|ENSG00000169083|LoF_protect_ch|     4.0|       null|    null|       null|       1|       null|         coherent|        coherent|reproductive syst...|        Other|    notAgree|   yes|\n",
      "|  EFO_0000660|ENSG00000169083|GoF_protect_ch|     0.5|       null|    null|       null|       1|       null|         coherent|        coherent|reproductive syst...|        Other|       agree|    no|\n",
      "|MONDO_0005090|ENSG00000112038|LoF_protect_ch|     4.0|       null|    null|       null|       1|       null|         coherent|        coherent|nervous system di...|        Other|    notAgree|   yes|\n",
      "|MONDO_0005090|ENSG00000112038|GoF_protect_ch|     2.0|       null|    null|       null|       1|       null|         coherent|        coherent|nervous system di...|        Other|       agree|    no|\n",
      "|MONDO_0008315|ENSG00000167553|LoF_protect_ch|     2.0|       null|      82|       null|    null|         10|         coherent|        coherent|cell proliferatio...|     Oncology|       agree|    no|\n",
      "|MONDO_0008315|ENSG00000167553|GoF_protect_ch|     2.0|       null|      82|       null|    null|         10|         coherent|        coherent|cell proliferatio...|     Oncology|    notAgree|    no|\n",
      "|  EFO_0001421|ENSG00000132170|LoF_protect_ch|     1.0|       null|    null|       null|       5|          1|         coherent|        coherent|gastrointestinal ...|        Other|    notAgree|    no|\n",
      "|  EFO_0001421|ENSG00000132170|GoF_protect_ch|     4.0|       null|    null|       null|       5|          1|         coherent|        coherent|gastrointestinal ...|        Other|       agree|   yes|\n",
      "+-------------+---------------+--------------+--------+-----------+--------+-----------+--------+-----------+-----------------+----------------+--------------------+-------------+------------+------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "disparDrug.selectExpr(\n",
    "    \"targetId\",\n",
    "    \"diseaseId\",\n",
    "    \"stack(2, 'GoF_protect_ch', GoF_protect_ch, 'LoF_protect_ch', LoF_protect_ch) as (var, val)\",\n",
    ").selectExpr(\"targetId\", \"diseaseId\", \"var as doeDrug\", \"val as maxPhase\").join(\n",
    "    genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\"\n",
    ").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(\n",
    "    diseaseTA.select(\"diseaseId\", \"taLabel\", \"taLabelSimple\"),\n",
    "    on=\"diseaseId\",\n",
    "    how=\"left\",\n",
    ").withColumn(\n",
    "    \"agreeGenetic\",\n",
    "    F.when(\n",
    "        (F.col(\"doeDrug\") == \"LoF_protect_ch\")\n",
    "        & ((F.col(\"LoF_protect\").isNotNull()) | (F.col(\"GoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"doeDrug\") == \"GoF_protect_ch\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notAgree\")),\n",
    ").withColumn(\n",
    "    \"phase4\", F.when(F.col(\"maxPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\"))\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 16:33:12 WARN CacheManager: Asked to cache already cached data.        \n",
      "25/02/03 16:33:13 WARN CacheManager: Asked to cache already cached data.\n",
      "25/02/03 16:33:13 WARN CacheManager: Asked to cache already cached data.\n",
      "25/02/03 16:33:13 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "genEvidForDrug = analysis_propagated(assessment, datasource_list).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ")\n",
    "\n",
    "drugDispar = analysis_drugs_dispar(assessment, chembl_ds).selectExpr(\n",
    "    \"targetId\",\n",
    "    \"diseaseId\",\n",
    "    \"maxClinPhase\",\n",
    "    \"GoF_protect as GoF_protect_ch\",\n",
    "    \"LoF_protect as LoF_protect_ch\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### rows per dif doe "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 16:29:22 WARN CacheManager: Asked to cache already cached data.        \n",
      "25/02/03 16:29:23 WARN CacheManager: Asked to cache already cached data.        \n",
      "25/02/03 16:29:23 WARN CacheManager: Asked to cache already cached data.\n",
      "25/02/03 16:29:23 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "genEvidForDrug = analysis_propagated(assessment, datasource_list).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ")\n",
    "\n",
    "drugDispar = analysis_drugs_dispar_approved(\n",
    "    assessment, chembl_ds, drugApproved\n",
    ").selectExpr(\n",
    "    \"targetId\",\n",
    "    \"diseaseId\",\n",
    "    \"maxClinPhase\",\n",
    "    \"GoF_protect as GoF_protect_ch\",\n",
    "    \"LoF_protect as LoF_protect_ch\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81435"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugDispar.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------+-----------+-----------+\n",
      "|       targetId|    diseaseId|GoF_protect|LoF_protect|noEvaluable|\n",
      "+---------------+-------------+-----------+-----------+-----------+\n",
      "|ENSG00000007314|  EFO_0000555|       null|          1|       null|\n",
      "|ENSG00000007314|  EFO_0003102|       null|          1|       null|\n",
      "|ENSG00000007314|  EFO_0003894|       null|          1|       null|\n",
      "|ENSG00000007314|  EFO_0004699|       null|          1|       null|\n",
      "|ENSG00000007314|  EFO_0801084|       null|          1|       null|\n",
      "|ENSG00000007314|  EFO_1000249|       null|          1|       null|\n",
      "|ENSG00000008018|  EFO_1000453|       null|          1|       null|\n",
      "|ENSG00000010310|  EFO_0003884|          1|       null|       null|\n",
      "|ENSG00000012504|  EFO_0004210|          1|       null|       null|\n",
      "|ENSG00000012504|MONDO_0019052|          1|       null|       null|\n",
      "|ENSG00000012779|MONDO_0004235|       null|          1|       null|\n",
      "|ENSG00000014138|  EFO_1001945|       null|          1|       null|\n",
      "|ENSG00000023228|  EFO_1000657|       null|          1|       null|\n",
      "|ENSG00000023445|  EFO_0000616|       null|          1|       null|\n",
      "|ENSG00000025708|  EFO_0000702|       null|          1|       null|\n",
      "|ENSG00000037280|  EFO_0003897|       null|          1|       null|\n",
      "|ENSG00000058091|  EFO_0000574|       null|          1|       null|\n",
      "|ENSG00000062822|  EFO_0004991|       null|          1|       null|\n",
      "|ENSG00000062822|  EFO_0005537|       null|          1|       null|\n",
      "|ENSG00000062822|  EFO_0005952|       null|          1|       null|\n",
      "+---------------+-------------+-----------+-----------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "drugDispar.groupBy(\"targetId\", \"diseaseId\").pivot(\"homogenized\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|homogenized|maxClinPhase|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+-----------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000145675|  EFO_0004142|LoF_protect|         1.0|          null|          null|       null|    null|       null|       1|          1|         coherent|        coherent|\n",
      "|ENSG00000186868|MONDO_0019037|LoF_protect|         2.0|          null|          null|       null|    null|       null|       1|         31|         coherent|        coherent|\n",
      "|ENSG00000037280|  EFO_0003897|LoF_protect|         2.0|          null|          null|       null|       2|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000144285|  EFO_0005306|LoF_protect|         4.0|          null|          null|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000182866|MONDO_0002691|LoF_protect|         2.0|          null|          null|       null|       1|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000163932|  EFO_0001421|LoF_protect|         1.0|          null|          null|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000132170|  EFO_0000519|GoF_protect|         1.0|          null|          null|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000186297|  EFO_0001063|GoF_protect|         2.0|          null|          null|       null|    null|       null|      17|       null|         coherent|        coherent|\n",
      "|ENSG00000105976|  EFO_0000365|LoF_protect|         2.0|          null|          null|       null|       3|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000146648|  EFO_0001061|LoF_protect|         2.0|          null|          null|       null|       2|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000150594|  EFO_1001109|GoF_protect|         3.0|          null|          null|       null|    null|       null|       2|       null|         coherent|        coherent|\n",
      "|ENSG00000159640|MONDO_0010520|LoF_protect|         2.0|          null|          null|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000134460|  EFO_0000218|GoF_protect|         3.0|          null|          null|       null|    null|       null|       2|       null|         coherent|        coherent|\n",
      "|ENSG00000112964| Orphanet_665|GoF_protect|         2.0|          null|          null|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000165731|MONDO_0002367|LoF_protect|         3.0|          null|          null|       null|    null|       null|       6|          1|         coherent|        coherent|\n",
      "|ENSG00000105639|MONDO_0018305|LoF_protect|         1.0|          null|          null|       null|    null|       null|       7|       null|         coherent|        coherent|\n",
      "|ENSG00000121879|  EFO_0001378|LoF_protect|         2.0|          null|          null|       null|       1|       null|    null|          3|         coherent|        coherent|\n",
      "|ENSG00000012504|MONDO_0019052|GoF_protect|         4.0|          null|          null|       null|    null|       null|       2|       null|         coherent|        coherent|\n",
      "|ENSG00000096968|MONDO_0008903|LoF_protect|         1.0|          null|          null|       null|       1|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000140675|  EFO_0004593|LoF_protect|         3.0|          null|          null|       null|    null|       null|       6|       null|         coherent|        coherent|\n",
      "+---------------+-------------+-----------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "drugDispar.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2569:==================================================> (195 + 6) / 201]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------+------------+------------+----------------+-----+----+\n",
      "|       targetId|    diseaseId|homogenized|minStartDate|maxStartDate|maxClinPhase_moa|count|rest|\n",
      "+---------------+-------------+-----------+------------+------------+----------------+-----+----+\n",
      "|ENSG00000000938|  EFO_0000183|LoF_protect|  2008-10-01|  2012-05-01|             1.0|    2|null|\n",
      "|ENSG00000000938|  EFO_0000519|LoF_protect|  2009-05-01|  2009-10-01|             2.0|    2|null|\n",
      "|ENSG00000000938|  EFO_0000519|LoF_protect|  2009-05-01|        null|             2.0|    2|null|\n",
      "|ENSG00000000938|  EFO_0000565|LoF_protect|  2003-11-01|        null|             3.0|   16|null|\n",
      "|ENSG00000000938|  EFO_0000565|LoF_protect|  2003-11-01|  2022-08-25|             3.0|    2|null|\n",
      "|ENSG00000000938|  EFO_0000702|LoF_protect|  2007-04-01|  2007-04-01|             2.0|    1|null|\n",
      "|ENSG00000000938|  EFO_0002918|LoF_protect|  2017-07-07|  2017-07-07|             1.0|    1|null|\n",
      "|ENSG00000000938|  EFO_1000453|LoF_protect|  2017-12-06|  2017-12-06|             2.0|    1|null|\n",
      "|ENSG00000000938|MONDO_0000873|LoF_protect|  2008-10-01|        null|             2.0|    1|null|\n",
      "|ENSG00000000938|MONDO_0000873|LoF_protect|  2008-10-01|  2017-02-23|             2.0|    2|null|\n",
      "|ENSG00000000938|MONDO_0019472|LoF_protect|  2008-10-01|  2008-10-01|             1.0|    1|null|\n",
      "|ENSG00000000938|MONDO_0019474|LoF_protect|  2008-10-01|  2008-10-01|             1.0|    1|null|\n",
      "|ENSG00000000938|MONDO_0100342|LoF_protect|  2017-12-06|  2017-12-06|             2.0|    1|null|\n",
      "|ENSG00000002549|  EFO_0000222|LoF_protect|  2009-10-01|  2010-06-01|             2.0|    2|null|\n",
      "|ENSG00000004468|  EFO_0000095|LoF_protect|  2018-08-16|  2020-09-16|             2.0|    2|null|\n",
      "|ENSG00000004468|  EFO_0000095|LoF_protect|  2018-08-16|        null|             2.0|    1|null|\n",
      "|ENSG00000004468|  EFO_0000209|LoF_protect|  2023-05-22|  2023-05-25|             2.0|    1|null|\n",
      "|ENSG00000004468|  EFO_0000209|LoF_protect|  2023-05-22|        null|             2.0|    1|null|\n",
      "|ENSG00000004468|  EFO_0000540|LoF_protect|  2020-06-18|  2023-05-26|             1.0|    2|null|\n",
      "|ENSG00000004468|  EFO_0000574|LoF_protect|  2017-02-14|        null|             2.0|    2|null|\n",
      "+---------------+-------------+-----------+------------+------------+----------------+-----+----+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "assessment.filter((F.col(\"datasourceId\").isin(chembl_ds))).withColumn(\n",
    "    \"minStartDate\",\n",
    "    F.min(F.col(\"studyStartDate\")).over(\n",
    "        Window.partitionBy(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "    ),\n",
    ").withColumn(\n",
    "    \"maxClinPhase_moa\",\n",
    "    F.max(F.col(\"clinicalPhase\")).over(\n",
    "        Window.partitionBy(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "    ),\n",
    ").withColumn(\n",
    "    \"maxStartDate\",\n",
    "    F.when(\n",
    "        F.col(\"clinicalPhase\") == F.col(\"maxClinPhase_moa\"),\n",
    "        F.max(F.col(\"studyStartDate\")).over(\n",
    "            Window.partitionBy(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "        ),\n",
    "    ).otherwise(F.lit(None)),\n",
    ").groupBy(\n",
    "    \"targetId\",\n",
    "    \"diseaseId\",\n",
    "    \"homogenized\",\n",
    "    \"minStartDate\",\n",
    "    \"maxStartDate\",\n",
    "    \"maxClinPhase_moa\",\n",
    ").count().withColumn(\n",
    "    \"rest\", F.col(\"maxStartDate\") - F.col(\"minStartDate\")\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+-----+\n",
      "|doeAgreeApp|count|\n",
      "+-----------+-----+\n",
      "|   notAgree|   11|\n",
      "|         no|   80|\n",
      "|      agree|   26|\n",
      "+-----------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "drugApp.join(\n",
    "    diseaseTA.select(\"diseaseId\", \"taLabelSimple\", \"taLabel\"),\n",
    "    on=\"diseaseId\",\n",
    "    how=\"left\",\n",
    ").join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeAgreeApp\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & ((F.col(\"LoF_protect\").isNotNull()) | (F.col(\"GoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protectApp\") == \"no\") & (F.col(\"LoF_protectApp\") == \"no\"),\n",
    "        F.lit(\"no\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notAgree\")),\n",
    ").groupBy(\n",
    "    \"doeAgreeApp\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 208:====================================================>(199 + 1) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+-------------+--------------------+---------------+----------------+--------------+--------------+------+\n",
      "|       targetId|    diseaseId|maxClinPhase|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|taLabelSimple|             taLabel|doeMaxClinPhase|doeAgreeMaxPhase|GoF_protectApp|LoF_protectApp|Phase4|\n",
      "+---------------+-------------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+-------------+--------------------+---------------+----------------+--------------+--------------+------+\n",
      "|ENSG00000134460|  EFO_0000198|         3.0|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000171873|  EFO_0005251|         4.0|           4.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|        Other|cardiovascular di...|    GoF_protect|           agree|           yes|            no|   yes|\n",
      "|ENSG00000155511|MONDO_0002009|         2.0|           2.0|           1.0|       null|    null|       null|       3|          3|         coherent|        coherent|        Other|nervous system di...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000091831| Orphanet_145|         2.0|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000169083|  EFO_0000616|         4.0|           4.0|           4.0|       null|       2|       null|    null|          1|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|        notAgree|           yes|           yes|   yes|\n",
      "|ENSG00000141510|  EFO_0000198|         3.0|           3.0|           1.0|       null|    null|       null|      14|          6|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000134460|  EFO_0000574|         3.0|           3.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000149295|  EFO_0003757|         4.0|           2.0|           4.0|       null|    null|       null|       2|       null|         coherent|        coherent|        Other|nervous system di...|    LoF_protect|        notAgree|            no|            no|   yes|\n",
      "|ENSG00000134460|  EFO_0001378|         2.0|           2.0|           2.0|       null|    null|       null|       1|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000178394|MONDO_0002050|         4.0|           4.0|           4.0|       null|    null|       null|       5|       null|         coherent|        coherent|        Other|nervous system di...|    GoF_protect|           agree|           yes|            no|   yes|\n",
      "|ENSG00000102468|  EFO_0004242|         4.0|           1.0|           4.0|       null|    null|       null|       3|       null|         coherent|        coherent|        Other|nervous system di...|    LoF_protect|        notAgree|            no|            no|   yes|\n",
      "|ENSG00000150594|  EFO_0005207|         3.0|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|        Other|cardiovascular di...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000118432|  EFO_0003756|         2.0|           2.0|           2.0|       null|    null|       null|       4|       null|         coherent|        coherent|        Other|nervous system di...|    GoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000066468|  EFO_0000222|         2.0|           2.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    GoF_protect|        notAgree|            no|            no|    no|\n",
      "|ENSG00000102468|  EFO_0003843|         4.0|           1.0|           4.0|       null|    null|          8|    null|       null|         coherent|        coherent|        Other|           phenotype|    LoF_protect|           agree|            no|            no|   yes|\n",
      "|ENSG00000144891|  EFO_0000537|         4.0|           3.0|           4.0|       null|    null|       null|       7|       null|         coherent|        coherent|        Other|cardiovascular di...|    LoF_protect|        notAgree|            no|           yes|   yes|\n",
      "|ENSG00000066468|  EFO_0000616|         4.0|           3.0|           4.0|       null|       1|       null|    null|         25|         coherent|        coherent|     Oncology|cell proliferatio...|    LoF_protect|           agree|            no|           yes|   yes|\n",
      "|ENSG00000198793|  EFO_0003060|         2.0|           1.0|           2.0|       null|       3|       null|    null|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    LoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000121966|  EFO_0000222|         2.0|           1.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|     Oncology|cell proliferatio...|    LoF_protect|           agree|            no|            no|    no|\n",
      "|ENSG00000134460|  EFO_0000729|         2.0|           1.0|           2.0|       null|    null|       null|       1|          2|         coherent|        coherent|        Other|gastrointestinal ...|    LoF_protect|        notAgree|            no|            no|    no|\n",
      "+---------------+-------------+------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+-------------+--------------------+---------------+----------------+--------------+--------------+------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n.groupBy(\\n    \"taLabelSimple\", \"doeAgreeMaxPhase\"\\n).pivot(\\n    \"Phase4\"\\n).count().select(\\n    \"taLabelSimple\", \"doeAgreeMaxPhase\", \"yes\", \"no\"\\n).sort(\\n    F.col(\"taLabelSimple\").desc(), F.col(\"doeAgreeMaxPhase\").desc()\\n).show()\\n'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugDispar.filter(F.col(\"coherencyDiagonal\") == \"dispar\").drop(\n",
    "    \"coherencyDiagonal\", \"coherencyOneCell\"\n",
    ").join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").join(\n",
    "    diseaseTA.select(\"diseaseId\", \"taLabelSimple\", \"taLabel\"),\n",
    "    on=\"diseaseId\",\n",
    "    how=\"left\",\n",
    ").withColumn(\n",
    "    \"doeMaxClinPhase\",\n",
    "    F.when(F.col(\"GoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"GoF_protect\"))\n",
    "    .when(F.col(\"LoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"LoF_protect\"))\n",
    "    .when(F.col(\"GoF_protect_ch\") == F.col(\"LoF_protect_ch\"), F.lit(\"same\"))\n",
    "    .otherwise(F.lit(\"notConsidered\")),\n",
    ").withColumn(\n",
    "    \"doeAgreeMaxPhase\",\n",
    "    F.when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"GoF_protect\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"LoF_protect\")\n",
    "        & ((F.col(\"LoF_protect\").isNotNull()) | (F.col(\"GoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notAgree\")),\n",
    ").join(\n",
    "    drugApp.drop(\"maxClinPhase\"), on=[\"targetId\", \"diseaseId\"], how=\"left\"\n",
    ").withColumn(\n",
    "    \"Phase4\", F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\"))\n",
    ").show()\n",
    "\n",
    "\"\"\"\n",
    ".groupBy(\n",
    "    \"taLabelSimple\", \"doeAgreeMaxPhase\"\n",
    ").pivot(\n",
    "    \"Phase4\"\n",
    ").count().select(\n",
    "    \"taLabelSimple\", \"doeAgreeMaxPhase\", \"yes\", \"no\"\n",
    ").sort(\n",
    "    F.col(\"taLabelSimple\").desc(), F.col(\"doeAgreeMaxPhase\").desc()\n",
    ").show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000112164|MONDO_0005148|           4.0|           3.0|          9|    null|       null|       4|          2|         coherent|          dispar|\n",
      "|ENSG00000132170|  EFO_0002618|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000198793|MONDO_0007254|           2.0|           3.0|       null|       1|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000102468|MONDO_0002009|           3.0|           4.0|       null|    null|       null|       3|       null|         coherent|        coherent|\n",
      "|ENSG00000134460|  EFO_0000198|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|\n",
      "|ENSG00000171873|  EFO_0005251|           4.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000116329|MONDO_0002009|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000091831|MONDO_0011962|           3.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000169083|  EFO_0000660|           0.5|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000155511|MONDO_0002009|           2.0|           1.0|       null|    null|       null|       3|          3|         coherent|        coherent|\n",
      "|ENSG00000126895|  EFO_0003086|           4.0|           2.0|       null|    null|       null|       7|       null|         coherent|        coherent|\n",
      "|ENSG00000109163|  EFO_0000660|           4.0|           4.0|       null|    null|       null|       5|       null|         coherent|        coherent|\n",
      "|ENSG00000091831| Orphanet_145|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000167553|MONDO_0008315|           2.0|           2.0|       null|      82|       null|    null|         10|         coherent|        coherent|\n",
      "|ENSG00000112038|MONDO_0005090|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000132170|  EFO_0001421|           4.0|           1.0|       null|    null|       null|       5|          1|         coherent|        coherent|\n",
      "|ENSG00000169083|  EFO_0000616|           4.0|           4.0|       null|       2|       null|    null|          1|         coherent|        coherent|\n",
      "|ENSG00000141510|  EFO_0000198|           3.0|           1.0|       null|    null|       null|      14|          6|         coherent|        coherent|\n",
      "|ENSG00000134460|  EFO_0000574|           3.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000149295|  EFO_0003757|           2.0|           4.0|       null|    null|       null|       2|       null|         coherent|        coherent|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+\n",
      "|       targetId|    diseaseId|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|maxClinPhase|GoF_protectApp|LoF_protectApp|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+\n",
      "|ENSG00000112164|MONDO_0005148|           4.0|           3.0|          9|    null|       null|       4|          2|         coherent|          dispar|         4.0|           yes|            no|\n",
      "|ENSG00000132170|  EFO_0002618|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000198793|MONDO_0007254|           2.0|           3.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000102468|MONDO_0002009|           3.0|           4.0|       null|    null|       null|       3|       null|         coherent|        coherent|         4.0|            no|           yes|\n",
      "|ENSG00000134460|  EFO_0000198|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000171873|  EFO_0005251|           4.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|           yes|            no|\n",
      "|ENSG00000116329|MONDO_0002009|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000091831|MONDO_0011962|           3.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000169083|  EFO_0000660|           0.5|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000155511|MONDO_0002009|           2.0|           1.0|       null|    null|       null|       3|          3|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000126895|  EFO_0003086|           4.0|           2.0|       null|    null|       null|       7|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000109163|  EFO_0000660|           4.0|           4.0|       null|    null|       null|       5|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000091831| Orphanet_145|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000167553|MONDO_0008315|           2.0|           2.0|       null|      82|       null|    null|         10|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000112038|MONDO_0005090|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|           yes|\n",
      "|ENSG00000132170|  EFO_0001421|           4.0|           1.0|       null|    null|       null|       5|          1|         coherent|        coherent|         4.0|           yes|            no|\n",
      "|ENSG00000169083|  EFO_0000616|           4.0|           4.0|       null|       2|       null|    null|          1|         coherent|        coherent|         4.0|           yes|           yes|\n",
      "|ENSG00000141510|  EFO_0000198|           3.0|           1.0|       null|    null|       null|      14|          6|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000134460|  EFO_0000574|           3.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000149295|  EFO_0003757|           2.0|           4.0|       null|    null|       null|       2|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+\n",
      "|       targetId|    diseaseId|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|maxClinPhase|GoF_protectApp|LoF_protectApp|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+\n",
      "|ENSG00000112164|MONDO_0005148|           4.0|           3.0|          9|    null|       null|       4|          2|         coherent|          dispar|         4.0|           yes|            no|\n",
      "|ENSG00000132170|  EFO_0002618|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000198793|MONDO_0007254|           2.0|           3.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000102468|MONDO_0002009|           3.0|           4.0|       null|    null|       null|       3|       null|         coherent|        coherent|         4.0|            no|           yes|\n",
      "|ENSG00000134460|  EFO_0000198|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000171873|  EFO_0005251|           4.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|           yes|            no|\n",
      "|ENSG00000116329|MONDO_0002009|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000091831|MONDO_0011962|           3.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000169083|  EFO_0000660|           0.5|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000155511|MONDO_0002009|           2.0|           1.0|       null|    null|       null|       3|          3|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000126895|  EFO_0003086|           4.0|           2.0|       null|    null|       null|       7|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000109163|  EFO_0000660|           4.0|           4.0|       null|    null|       null|       5|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "|ENSG00000091831| Orphanet_145|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000167553|MONDO_0008315|           2.0|           2.0|       null|      82|       null|    null|         10|         coherent|        coherent|         2.0|            no|            no|\n",
      "|ENSG00000112038|MONDO_0005090|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|           yes|\n",
      "|ENSG00000132170|  EFO_0001421|           4.0|           1.0|       null|    null|       null|       5|          1|         coherent|        coherent|         4.0|           yes|            no|\n",
      "|ENSG00000169083|  EFO_0000616|           4.0|           4.0|       null|       2|       null|    null|          1|         coherent|        coherent|         4.0|           yes|           yes|\n",
      "|ENSG00000141510|  EFO_0000198|           3.0|           1.0|       null|    null|       null|      14|          6|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000134460|  EFO_0000574|           3.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         3.0|            no|            no|\n",
      "|ENSG00000149295|  EFO_0003757|           2.0|           4.0|       null|    null|       null|       2|       null|         coherent|        coherent|         4.0|            no|            no|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 806:============================================>       (170 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------------+---+---+\n",
      "|doeAgreeMaxPhase| no|yes|\n",
      "+----------------+---+---+\n",
      "|        notAgree| 21| 22|\n",
      "|           agree| 40| 34|\n",
      "+----------------+---+---+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeMaxClinPhase\",\n",
    "    F.when(F.col(\"GoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"GoF_protect\"))\n",
    "    .when(F.col(\"LoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"LoF_protect\"))\n",
    "    .when(F.col(\"GoF_protect_ch\") == F.col(\"LoF_protect_ch\"), F.lit(\"same\"))\n",
    "    .otherwise(F.lit(\"notConsidered\")),\n",
    ").withColumn(\n",
    "    \"doeAgreeMaxPhase\",\n",
    "    F.when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"GoF_protect\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"LoF_protect\")\n",
    "        & ((F.col(\"LoF_protect\").isNotNull()) | (F.col(\"GoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notAgree\")),\n",
    ").withColumn(\n",
    "    \"Phase4\", F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\"))\n",
    ").groupBy(\n",
    "    \"doeAgreeMaxPhase\"\n",
    ").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+---------------+----------------+\n",
      "|       targetId|    diseaseId|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|maxClinPhase|GoF_protectApp|LoF_protectApp|doeMaxClinPhase|doeAgreeMaxPhase|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+---------------+----------------+\n",
      "|ENSG00000112164|MONDO_0005148|           4.0|           3.0|          9|    null|       null|       4|          2|         coherent|          dispar|         4.0|           yes|            no|    GoF_protect|           agree|\n",
      "|ENSG00000132170|  EFO_0002618|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000198793|MONDO_0007254|           2.0|           3.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|    LoF_protect|           agree|\n",
      "|ENSG00000102468|MONDO_0002009|           3.0|           4.0|       null|    null|       null|       3|       null|         coherent|        coherent|         4.0|            no|           yes|    LoF_protect|        notAgree|\n",
      "|ENSG00000134460|  EFO_0000198|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000171873|  EFO_0005251|           4.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|           yes|            no|    GoF_protect|           agree|\n",
      "|ENSG00000116329|MONDO_0002009|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|    LoF_protect|        notAgree|\n",
      "|ENSG00000091831|MONDO_0011962|           3.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|        notAgree|\n",
      "|ENSG00000169083|  EFO_0000660|           0.5|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|    LoF_protect|        notAgree|\n",
      "|ENSG00000155511|MONDO_0002009|           2.0|           1.0|       null|    null|       null|       3|          3|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000126895|  EFO_0003086|           4.0|           2.0|       null|    null|       null|       7|       null|         coherent|        coherent|         4.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000109163|  EFO_0000660|           4.0|           4.0|       null|    null|       null|       5|       null|         coherent|        coherent|         4.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000091831| Orphanet_145|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000167553|MONDO_0008315|           2.0|           2.0|       null|      82|       null|    null|         10|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|        notAgree|\n",
      "|ENSG00000112038|MONDO_0005090|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|           yes|    LoF_protect|        notAgree|\n",
      "|ENSG00000132170|  EFO_0001421|           4.0|           1.0|       null|    null|       null|       5|          1|         coherent|        coherent|         4.0|           yes|            no|    GoF_protect|           agree|\n",
      "|ENSG00000169083|  EFO_0000616|           4.0|           4.0|       null|       2|       null|    null|          1|         coherent|        coherent|         4.0|           yes|           yes|    GoF_protect|        notAgree|\n",
      "|ENSG00000141510|  EFO_0000198|           3.0|           1.0|       null|    null|       null|      14|          6|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000134460|  EFO_0000574|           3.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|           agree|\n",
      "|ENSG00000149295|  EFO_0003757|           2.0|           4.0|       null|    null|       null|       2|       null|         coherent|        coherent|         4.0|            no|            no|    LoF_protect|        notAgree|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+---------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeMaxClinPhase\",\n",
    "    F.when(F.col(\"GoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"GoF_protect\"))\n",
    "    .when(F.col(\"LoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"LoF_protect\"))\n",
    "    .when(F.col(\"GoF_protect_ch\") == F.col(\"LoF_protect_ch\"), F.lit(\"same\"))\n",
    "    .otherwise(F.lit(\"notConsidered\")),\n",
    ").withColumn(\n",
    "    \"doeAgreeMaxPhase\",\n",
    "    F.when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"GoF_protect\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"LoF_protect\")\n",
    "        & ((F.col(\"LoF_protect\").isNotNull()) | (F.col(\"GoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notAgree\")),\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+---------------+----------------+------+-------------------+\n",
      "|       targetId|    diseaseId|GoF_protect_ch|LoF_protect_ch|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|maxClinPhase|GoF_protectApp|LoF_protectApp|doeMaxClinPhase|doeAgreeMaxPhase|Phase4|     doeAdvancedApp|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+---------------+----------------+------+-------------------+\n",
      "|ENSG00000112164|MONDO_0005148|           4.0|           3.0|          9|    null|       null|       4|          2|         coherent|          dispar|         4.0|           yes|            no|    GoF_protect|           agree|   yes|   agreeAndApproved|\n",
      "|ENSG00000132170|  EFO_0002618|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000198793|MONDO_0007254|           2.0|           3.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|    LoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000102468|MONDO_0002009|           3.0|           4.0|       null|    null|       null|       3|       null|         coherent|        coherent|         4.0|            no|           yes|    LoF_protect|        notAgree|   yes|   agreeAndApproved|\n",
      "|ENSG00000134460|  EFO_0000198|           3.0|           2.0|       null|    null|       null|       2|       null|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000171873|  EFO_0005251|           4.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|           yes|            no|    GoF_protect|           agree|   yes|   agreeAndApproved|\n",
      "|ENSG00000116329|MONDO_0002009|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|    LoF_protect|        notAgree|   yes|NotAgreeNotApproved|\n",
      "|ENSG00000091831|MONDO_0011962|           3.0|           2.0|       null|       1|       null|    null|       null|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|        notAgree|    no|NotAgreeNotApproved|\n",
      "|ENSG00000169083|  EFO_0000660|           0.5|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|            no|    LoF_protect|        notAgree|   yes|NotAgreeNotApproved|\n",
      "|ENSG00000155511|MONDO_0002009|           2.0|           1.0|       null|    null|       null|       3|          3|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000126895|  EFO_0003086|           4.0|           2.0|       null|    null|       null|       7|       null|         coherent|        coherent|         4.0|            no|            no|    GoF_protect|           agree|   yes|    AgreeNoApproved|\n",
      "|ENSG00000109163|  EFO_0000660|           4.0|           4.0|       null|    null|       null|       5|       null|         coherent|        coherent|         4.0|            no|            no|    GoF_protect|           agree|   yes|    AgreeNoApproved|\n",
      "|ENSG00000091831| Orphanet_145|           2.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000167553|MONDO_0008315|           2.0|           2.0|       null|      82|       null|    null|         10|         coherent|        coherent|         2.0|            no|            no|    GoF_protect|        notAgree|    no|NotAgreeNotApproved|\n",
      "|ENSG00000112038|MONDO_0005090|           2.0|           4.0|       null|    null|       null|       1|       null|         coherent|        coherent|         4.0|            no|           yes|    LoF_protect|        notAgree|   yes|   agreeAndApproved|\n",
      "|ENSG00000132170|  EFO_0001421|           4.0|           1.0|       null|    null|       null|       5|          1|         coherent|        coherent|         4.0|           yes|            no|    GoF_protect|           agree|   yes|   agreeAndApproved|\n",
      "|ENSG00000169083|  EFO_0000616|           4.0|           4.0|       null|       2|       null|    null|          1|         coherent|        coherent|         4.0|           yes|           yes|    GoF_protect|        notAgree|   yes|   agreeAndApproved|\n",
      "|ENSG00000141510|  EFO_0000198|           3.0|           1.0|       null|    null|       null|      14|          6|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000134460|  EFO_0000574|           3.0|           1.0|       null|    null|       null|       1|       null|         coherent|        coherent|         3.0|            no|            no|    GoF_protect|           agree|    no|    AgreeNoApproved|\n",
      "|ENSG00000149295|  EFO_0003757|           2.0|           4.0|       null|    null|       null|       2|       null|         coherent|        coherent|         4.0|            no|            no|    LoF_protect|        notAgree|   yes|NotAgreeNotApproved|\n",
      "+---------------+-------------+--------------+--------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+------------+--------------+--------------+---------------+----------------+------+-------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeMaxClinPhase\",\n",
    "    F.when(F.col(\"GoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"GoF_protect\"))\n",
    "    .when(F.col(\"LoF_protect_ch\") == F.col(\"maxClinPhase\"), F.lit(\"LoF_protect\"))\n",
    "    .when(F.col(\"GoF_protect_ch\") == F.col(\"LoF_protect_ch\"), F.lit(\"same\"))\n",
    "    .otherwise(F.lit(\"notConsidered\")),\n",
    ").withColumn(\n",
    "    \"doeAgreeMaxPhase\",\n",
    "    F.when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"GoF_protect\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"doeMaxClinPhase\") == \"LoF_protect\")\n",
    "        & ((F.col(\"LoF_protect\").isNotNull()) | (F.col(\"GoF_risk\").isNotNull())),\n",
    "        F.lit(\"agree\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notAgree\")),\n",
    ").withColumn(\n",
    "    \"Phase4\", F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\"))\n",
    ").withColumn(\n",
    "    \"doeAdvancedApp\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"agreeAndApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"agreeAndApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"doeAgreeMaxPhase\") == \"agree\"),\n",
    "        F.lit(\"AgreeNoApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"doeAgreeMaxPhase\") == \"notAgree\"),\n",
    "        F.lit(\"NotAgreeNotApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"doeAgreeMaxPhase\") == \"agree\"),\n",
    "        F.lit(\"agreeAndApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"doeAgreeMaxPhase\") == \"agree\"),\n",
    "        F.lit(\"agreeAndApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"doeAgreeMaxPhase\") == \"notAgree\"),\n",
    "        F.lit(\"agreeAndApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"doeAgreeMaxPhase\") == \"notAgree\"),\n",
    "        F.lit(\"NotAgreeNotApproved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\") & (F.col(\"GoF_protectApp\") == \"yes\"),\n",
    "        F.lit(\"agreeAndApproved\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"notConsidered\")),\n",
    ").show()\n",
    "#### here adding the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 205:===========================================>        (168 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----+\n",
      "|    doeAdvanced|count|\n",
      "+---------------+-----+\n",
      "|       advanced|   49|\n",
      "|advancedNoagree|   35|\n",
      "|         behind|   25|\n",
      "|          noDif|    8|\n",
      "+---------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n.select(\\n    \"GoF_protect_ch\",\\n    \"LoF_protect_ch\",\\n    \"GoF_protect\",\\n    \"GoF_risk\",\\n    \"LoF_protect\",\\n    \"LoF_risk\",\\n    \"doeAdvanced\",\\n    \"Phase4\",\\n    \"doeAgree\",\\n).groupBy(\\n    \"doeAgree\"\\n).pivot(\\n    \"Phase4\"\\n).count().show()\\n'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeAdvanced\",\n",
    "    F.when(  ### doe agreement and advanced\n",
    "        (F.col(\"GoF_protect_ch\") > F.col(\"LoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"AgreeAndAdvanced\"),\n",
    "    )\n",
    "    .when(  ### doE agreement and advanced\n",
    "        (F.col(\"LoF_protect_ch\") > F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"AgreeAndAdvanced\"),\n",
    "    )\n",
    "    .when(  ### doE NO agreement and advanced\n",
    "        (F.col(\"LoF_protect_ch\") < F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"NoAgreeAdvanced\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protect_ch\") < F.col(\"LoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"NoAgreeAdvanced\"),\n",
    "    )\n",
    "    .when(  ###\n",
    "        (F.col(\"LoF_protect_ch\") == F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"behind\")),\n",
    ").withColumn(\n",
    "    \"doeAdvancedApp\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"approved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"approved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\") & (F.col(\"GoF_protectApp\") == \"no\"),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\") & (F.col(\"GoF_protectApp\") == \"yes\"),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"behind\")),\n",
    ").withColumn(\n",
    "    \"Phase4\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protect_ch\") == 4) | (F.col(\"LoF_protect\") == 4),\n",
    "        F.when((F.col(\"doeAdvanced\") == \"advanced\"), F.lit(\"yes\"))\n",
    "        .when(F.col(\"doeAdvanced\") == \"noDif\", F.lit(\"yes\"))\n",
    "        .when(F.col(\"doeAdvanced\") == \"behind\", F.lit(\"no\")),\n",
    "    ).otherwise(F.lit(\"no\")),\n",
    ").withColumn(\n",
    "    \"doeAgree\",\n",
    "    F.when(F.col(\"doeAdvanced\").isin([\"advanced\", \"noDif\"]), F.lit(\"yes\")).otherwise(\n",
    "        F.lit(\"no\")\n",
    "    ),\n",
    ").groupBy(\n",
    "    \"doeAdvanced\"\n",
    ").count().show()\n",
    "\n",
    "\n",
    "\"\"\"\n",
    ".select(\n",
    "    \"GoF_protect_ch\",\n",
    "    \"LoF_protect_ch\",\n",
    "    \"GoF_protect\",\n",
    "    \"GoF_risk\",\n",
    "    \"LoF_protect\",\n",
    "    \"LoF_risk\",\n",
    "    \"doeAdvanced\",\n",
    "    \"Phase4\",\n",
    "    \"doeAgree\",\n",
    ").groupBy(\n",
    "    \"doeAgree\"\n",
    ").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### to test "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 393:==========================================>         (162 + 17) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+-----+\n",
      "|doeAdvanced|count|\n",
      "+-----------+-----+\n",
      "|        yes|   49|\n",
      "|         no|   35|\n",
      "|      noDif|   33|\n",
      "+-----------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n.select(\\n    \"GoF_protect_ch\",\\n    \"LoF_protect_ch\",\\n    \"GoF_protect\",\\n    \"GoF_risk\",\\n    \"LoF_protect\",\\n    \"LoF_risk\",\\n    \"doeAdvanced\",\\n    \"Phase4\",\\n    \"doeAgree\",\\n).groupBy(\\n    \"doeAgree\"\\n).pivot(\\n    \"Phase4\"\\n).count().show()\\n'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disparDrug.join(genEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeAdvanced\",\n",
    "    F.when(  ### doe agreement and advanced\n",
    "        (F.col(\"GoF_protect_ch\") > F.col(\"LoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"yes\"),\n",
    "    )\n",
    "    .when(  ### doE agreement and advanced\n",
    "        (F.col(\"LoF_protect_ch\") > F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"yes\"),\n",
    "    )\n",
    "    .when(  ### doE NO agreement and advanced\n",
    "        (F.col(\"LoF_protect_ch\") < F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"no\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protect_ch\") < F.col(\"LoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"no\"),\n",
    "    )\n",
    "    .when(  ###\n",
    "        (F.col(\"LoF_protect_ch\") == F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .when(  ###\n",
    "        (F.col(\"LoF_protect_ch\") == F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"LoF_risk\").isNotNull()) | (F.col(\"GoF_protect\").isNotNull())),\n",
    "        F.lit(\"noDif\"),\n",
    "    ),\n",
    "    #    .otherwise(F.lit(\"behind\")),\n",
    ").withColumn(\n",
    "    \"doeAdvancedApp\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"approved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"approved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\") & (F.col(\"GoF_protectApp\") == \"no\"),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\") & (F.col(\"GoF_protectApp\") == \"yes\"),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"behind\")),\n",
    ").withColumn(\n",
    "    \"Phase4\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protect_ch\") == 4) | (F.col(\"LoF_protect\") == 4),\n",
    "        F.when((F.col(\"doeAdvanced\") == \"advanced\"), F.lit(\"yes\"))\n",
    "        .when(F.col(\"doeAdvanced\") == \"noDif\", F.lit(\"yes\"))\n",
    "        .when(F.col(\"doeAdvanced\") == \"behind\", F.lit(\"no\")),\n",
    "    ).otherwise(F.lit(\"no\")),\n",
    ").withColumn(\n",
    "    \"doeAgree\",\n",
    "    F.when(F.col(\"doeAdvanced\").isin([\"advanced\", \"noDif\"]), F.lit(\"yes\")).otherwise(\n",
    "        F.lit(\"no\")\n",
    "    ),\n",
    ").groupBy(\n",
    "    \"doeAdvanced\"\n",
    ").count().show()\n",
    "\n",
    "\"\"\"\n",
    ".select(\n",
    "    \"GoF_protect_ch\",\n",
    "    \"LoF_protect_ch\",\n",
    "    \"GoF_protect\",\n",
    "    \"GoF_risk\",\n",
    "    \"LoF_protect\",\n",
    "    \"LoF_risk\",\n",
    "    \"doeAdvanced\",\n",
    "    \"Phase4\",\n",
    "    \"doeAgree\",\n",
    ").groupBy(\n",
    "    \"doeAgree\"\n",
    ").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "disparDrug.join(genEvidForDrugDispar, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/02/03 10:31:02 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "genEvidForDrugDispar = analysis_propagated(assessment, datasource_list).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### let's do the same but using just germline variation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    # \"intogen\",\n",
    "    # \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    # \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "germlineEvidForDrug = analysis_propagated(assessment, datasource_list).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 465:==========================================>         (164 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+---------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|      diseaseId|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+---------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000198646|  MONDO_0011060|       null|    null|       null|      33|       null|         coherent|        coherent|\n",
      "|ENSG00000161133|  MONDO_0031520|       null|    null|       null|      13|       null|         coherent|        coherent|\n",
      "|ENSG00000075043|  MONDO_0100172|       null|    null|       null|       4|          1|         coherent|        coherent|\n",
      "|ENSG00000227507|    EFO_1001901|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000186350|  MONDO_0016145|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000102001| Orphanet_98666|       null|    null|       null|      75|          3|         coherent|        coherent|\n",
      "|ENSG00000130876|Orphanet_306768|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000124302|  MONDO_0020301|       null|    null|       null|       3|       null|         coherent|        coherent|\n",
      "|ENSG00000115694|    EFO_0007987|       null|     168|       null|    null|          5|         coherent|        coherent|\n",
      "|ENSG00000124145|    EFO_0004747|       null|    null|         63|    null|         13|         coherent|        coherent|\n",
      "|ENSG00000177674|    EFO_0005090|       null|    null|         39|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000162594|   OTAR_0000006|       null|    null|          1|    null|       null|         coherent|        coherent|\n",
      "|ENSG00000080293|  MONDO_0009724|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000088888|  MONDO_0012426|       null|    null|       null|       2|       null|         coherent|        coherent|\n",
      "|ENSG00000100271|  MONDO_0010647|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000103313|  MONDO_0009413|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000108950|Orphanet_100033|       null|    null|       null|       8|       null|         coherent|        coherent|\n",
      "|ENSG00000111275|  MONDO_0015263|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000111837|  MONDO_0030307|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "|ENSG00000116514|  MONDO_0011599|       null|    null|       null|       1|       null|         coherent|        coherent|\n",
      "+---------------+---------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "germlineEvidForDrug.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2222:=============================================>     (178 + 16) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+--------------+-----+\n",
      "|doeAdvanced|doeAdvancedApp|count|\n",
      "+-----------+--------------+-----+\n",
      "|   advanced|         noDif|   20|\n",
      "|     behind|        behind|    9|\n",
      "|      noDif|         noDif|   20|\n",
      "|   advanced|      approved|    8|\n",
      "|     behind|         noDif|   19|\n",
      "|      noDif|      approved|    7|\n",
      "|      noDif|        behind|    1|\n",
      "+-----------+--------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "disparDrug.join(germlineEvidForDrug, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").join(drugApp, on=[\"targetId\", \"diseaseId\"], how=\"left\").filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"coherent\"\n",
    ").withColumn(\n",
    "    \"doeAdvanced\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protect_ch\") > F.col(\"LoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"advanced\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protect_ch\") > F.col(\"GoF_protect_ch\"))\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"advanced\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protect_ch\") == F.col(\"GoF_protect_ch\")),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"behind\")),\n",
    ").withColumn(\n",
    "    \"doeAdvancedApp\",\n",
    "    F.when(\n",
    "        (F.col(\"GoF_protectApp\") == \"yes\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"no\")\n",
    "        & ((F.col(\"GoF_protect\").isNotNull()) | (F.col(\"LoF_risk\").isNotNull())),\n",
    "        F.lit(\"approved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"GoF_protectApp\") == \"no\")\n",
    "        & (F.col(\"LoF_protectApp\") == \"yes\")\n",
    "        & ((F.col(\"GoF_risk\").isNotNull()) | (F.col(\"LoF_protect\").isNotNull())),\n",
    "        F.lit(\"approved\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"no\") & (F.col(\"GoF_protectApp\") == \"no\"),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .when(\n",
    "        (F.col(\"LoF_protectApp\") == \"yes\") & (F.col(\"GoF_protectApp\") == \"yes\"),\n",
    "        F.lit(\"noDif\"),\n",
    "    )\n",
    "    .otherwise(F.lit(\"behind\")),\n",
    ").groupBy(\n",
    "    \"doeAdvanced\", \"doeAdvancedApp\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2284:===========================================>          (13 + 3) / 16]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------------------+-----+\n",
      "|actionType                   |count|\n",
      "+-----------------------------+-----+\n",
      "|RNAI INHIBITOR               |21   |\n",
      "|NEGATIVE ALLOSTERIC MODULATOR|24   |\n",
      "|CROSS-LINKING AGENT          |27   |\n",
      "|ANTAGONIST                   |967  |\n",
      "|BINDING AGENT                |232  |\n",
      "|EXOGENOUS PROTEIN            |39   |\n",
      "|PARTIAL AGONIST              |64   |\n",
      "|ANTISENSE INHIBITOR          |38   |\n",
      "|PROTEOLYTIC ENZYME           |3    |\n",
      "|ACTIVATOR                    |70   |\n",
      "|BLOCKER                      |179  |\n",
      "|VACCINE ANTIGEN              |19   |\n",
      "|INHIBITOR                    |3303 |\n",
      "|SUBSTRATE                    |8    |\n",
      "|POSITIVE ALLOSTERIC MODULATOR|82   |\n",
      "|POSITIVE MODULATOR           |36   |\n",
      "|OTHER                        |31   |\n",
      "|AGONIST                      |942  |\n",
      "|OPENER                       |39   |\n",
      "|INVERSE AGONIST              |15   |\n",
      "|HYDROLYTIC ENZYME            |9    |\n",
      "|RELEASING AGENT              |26   |\n",
      "|MODULATOR                    |106  |\n",
      "|STABILISER                   |22   |\n",
      "|EXOGENOUS GENE               |10   |\n",
      "|DISRUPTING AGENT             |8    |\n",
      "|DEGRADER                     |6    |\n",
      "|NEGATIVE MODULATOR           |5    |\n",
      "|ALLOSTERIC ANTAGONIST        |1    |\n",
      "+-----------------------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/mechanismOfAction/\"\n",
    ").groupBy(\"actionType\").count().show(200, truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/mechanismOfAction/\"\n",
    ").groupBy(\"actionType\").count().show(200, truncate=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PySpark",
   "language": "python",
   "name": "pyspark"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
