{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-12-16 13:53:05.366726\n",
      "Analysis started on 2024-12-16 at  2024-12-16 13:53:05.366726\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:53:06 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:53:09 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:25 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:54:25 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading sources\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:27 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moving to step 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:28 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:28 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:28 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "moving to Step 3\n",
      "starting dictionaries at 2024-12-16 13:54:28.587980\n",
      "gene_burden\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:29 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:29 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:29 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:30 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:30 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:30 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intogen\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:32 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:32 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:32 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:34 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:54:34 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:34 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cancer_gene_census\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:35 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:35 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:35 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:37 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:37 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:37 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eva\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:38 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:38 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:38 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:40 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:40 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:40 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eva_somatic\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:42 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:42 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:42 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:43 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ot_genetics_portal\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:44 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:45 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas_credible_set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:53 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:54:59 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "impc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orphanet\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:04 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:04 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:04 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:06 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:06 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:06 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene2phenotype\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:07 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:07 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:07 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:08 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:08 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:08 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WOcgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:16 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:55:23 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "somatic\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:24 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:24 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:24 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:26 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:26 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:26 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "germline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:33 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:55:40 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\ndef aggregations_original(\\n    df,\\n    data,\\n    listado,\\n    comparisonColumn,\\n    comparisonType,\\n    predictionColumn,\\n    predictionType,\\n    today_date,\\n):\\n\\n    wComparison = Window.partitionBy(comparisonColumn)\\n    wPrediction = Window.partitionBy(predictionColumn)\\n    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\\n\\n    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\\n\\n    out = (\\n        df.withColumn(\"comparisonType\", F.lit(comparisonType))\\n        .withColumn(\"predictionType\", F.lit(predictionType))\\n        .withColumn(\"total\", F.lit(uniqIds))\\n        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\\n        .withColumn(\\n            \"predictionTotal\",\\n            F.count(\"targetId\").over(wPrediction),\\n        )\\n        .withColumn(\\n            \"comparisonTotal\",\\n            F.count(\"targetId\").over(wComparison),\\n        )\\n        .select(\\n            F.col(predictionColumn).alias(\"prediction\"),\\n            F.col(comparisonColumn).alias(\"comparison\"),\\n            \"comparisonType\",\\n            \"predictionType\",\\n            \"a\",\\n            \"predictionTotal\",\\n            \"comparisonTotal\",\\n            \"total\",\\n        )\\n        .filter(F.col(\"prediction\").isNotNull())\\n        .filter(F.col(\"comparison\").isNotNull())\\n        .distinct()\\n    )\\n\\n    out.write.mode(\"overwrite\").parquet(\\n        \"gs://ot-team/jroldan/\"\\n        + str(\\n            today_date\\n            + \"_\"\\n            + \"analysis/\"\\n            + data\\n            # + \"_propagated\"\\n            + \"/\"\\n            + comparisonColumn\\n            + \"_\"\\n            + predictionColumn\\n            + \".parquet\"\\n        )\\n    )\\n\\n    filePath = \"gs://ot-team/jroldan/\" + str(\\n        today_date\\n        + \"_\"\\n        + \"analysis/\"\\n        + data\\n        # + \"_propagated\"\\n        + \"/\"\\n        + comparisonColumn\\n        + \"_\"\\n        + predictionColumn\\n        + \".parquet\"\\n    )\\n    listado.append(\\n        filePath\\n    )\\n    print(\\n        today_date\\n        + \"_\"\\n        + \"analysis/\"\\n        + data\\n        # + \"_propagated\"\\n        + \"/\"\\n        + comparisonColumn\\n        + \"_\"\\n        + predictionColumn\\n        + \".parquet\"\\n    )\\n    c = datetime.now()\\n    c.strftime(\"%H:%M:%S\")\\n    print(c)\\n\\n    array1 = np.delete(\\n        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\\n        .groupBy(\"comparison\")\\n        .pivot(\"prediction\")\\n        .agg(F.first(\"a\"))\\n        .sort(F.col(\"comparison\").desc())\\n        .select(\"comparison\", \"yes\", \"no\")\\n        .fillna(0)\\n        .toPandas()\\n        .to_numpy(),\\n        [0],\\n        1,\\n    )\\n    total = np.sum(array1)\\n    res_npPhaseX = np.array(array1, dtype=int)\\n    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\\n    resx_CI = convertTuple(\\n        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\\n    )\\n\\n    result_st.append(resX)\\n    result_ci.append(resx_CI)\\n    (rs_result, rs_ci) = relative_success(array1)\\n\\n    results.append(\\n        [\\n            comparisonType,\\n            comparisonColumn,\\n            predictionColumn,\\n            round(float(resX.split(\",\")[0]), 2),\\n            float(resX.split(\",\")[1]),\\n            round(float(resx_CI.split(\",\")[0]), 2),\\n            round(float(resx_CI.split(\",\")[1]), 2),\\n            str(total),\\n            np.array(res_npPhaseX).tolist(),\\n            round(float(rs_result), 2),\\n            round(float(rs_ci[0]), 2),\\n            round(float(rs_ci[1]), 2),\\n            filePath,\\n        ]\\n    )\\n    return results\\n\\n\\nc = datetime.now()\\n\\nprint(\"start doing aggregations and writing\")\\ntoday_date = str(date.today())\\naggSetups_original = comparisons_df()\\nlistado = []\\n\\nprint(\"starting with non-propagated aggregations at\", c)\\n\\nfor key, df in dfs_dict.items():\\n    df = df.persist()\\n    for row in aggSetups_original:\\n        aggregations_original(df, key, listado, *row, today_date)\\n    df.unpersist()\\n    print(key + \" df unpersisted\")\\n\\nprint(\"non propagated files wroten succesfully at\", c)\\n\\nprint(\"starting with non-propagated aggregations at\", c)\\nfor key, df in dfs_dict_propag.items():\\n    df = df.persist()\\n    for row in aggSetups_original:\\n        aggregations_original(df, key, listado, *row, today_date)\\n    df.unpersist()\\n    print(key + \" df unpersisted\")\\n\\nprint(\"propagated files wroten succesfully at\", c)\\n\\n##### read files and make spreadsheet\\n\\nprint(\"preparing dataframe\")\\n\\nschema = StructType(\\n    [\\n        StructField(\"group\", StringType(), True),\\n        StructField(\"comparison\", StringType(), True),\\n        StructField(\"phase\", StringType(), True),\\n        StructField(\"oddsRatio\", DoubleType(), True),\\n        StructField(\"pValue\", DoubleType(), True),\\n        StructField(\"lowerInterval\", DoubleType(), True),\\n        StructField(\"upperInterval\", DoubleType(), True),\\n        StructField(\"total\", StringType(), True),\\n        StructField(\"values\", ArrayType(ArrayType(IntegerType())), True),\\n        StructField(\"relSuccess\", DoubleType(), True),\\n        StructField(\"rsLower\", DoubleType(), True),\\n        StructField(\"rsUpper\", DoubleType(), True),\\n        StructField(\"path\", StringType(), True),\\n    ]\\n)\\n\\ndf = spreadSheetFormatter(spark.createDataFrame(results, schema=schema))\\n\\n'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### 11.12.2024\n",
    "#######\n",
    "#######     ATENTION\n",
    "#### change code to work with generated dataframe instead of reading the parquet\n",
    "\n",
    "\"\"\"\n",
    "This scripts run Odds ratio analysis for DoE and \n",
    "genetic information on drug clinical success\n",
    "\n",
    "\"\"\"\n",
    "from functions import discrepancifier\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "today_date = str(date.today())\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "print(\"Analysis started on \" + today_date + \" at \", c)\n",
    "\"\"\"\n",
    "#coloc = spark.read.parquet(\n",
    "#    \"gs://genetics-portal-dev-data/22.09.1/outputs/v2d_coloc\"\n",
    "#).filter(F.col(\"right_type\") != \"gwas\")\n",
    "\"\"\"\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "evidences = (\n",
    "    spark.read.parquet(f\"{path}evidence\")\n",
    "    .filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "                \"gene_burden\",\n",
    "                \"eva\",\n",
    "                \"eva_somatic\",\n",
    "                \"gene2phenotype\",\n",
    "                \"orphanet\",\n",
    "                \"cancer_gene_census\",\n",
    "                \"intogen\",\n",
    "                \"impc\",\n",
    "                \"chembl\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns).persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "### bring directionality from QTL\n",
    "\n",
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\"homogenized\", F.first(\"colocDoE\").over(window_spec))\n",
    "    .select(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "# 1# Make a list of variant of interest (Sequence ontology terms) to subset data of interest.\n",
    "\n",
    "### Bear in mind that SO works with ontology structure as: SO:XXXXXX, but databases has the SO as: SO_XXXXXX\n",
    "\n",
    "var_filter_lof = [\n",
    "    ### High impact variants https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html\n",
    "    \"SO_0001589\",  ## frameshit_variant\n",
    "    \"SO_0001587\",  ## stop_gained\n",
    "    \"SO_0001574\",  ## splice_acceptor_variant\n",
    "    \"SO_0001575\",  ## splice_donor_variant\n",
    "    \"SO_0002012\",  ## start_lost\n",
    "    \"SO_0001578\",  ## stop_lost\n",
    "    \"SO_0001893\",  ## transcript_ablation\n",
    "    # \"SO:0001889\", ## transcript_amplification ## the Only HIGH impact that increase protein.\n",
    "]\n",
    "\n",
    "gof = [\"SO_0002053\"]\n",
    "lof = [\"SO_0002054\"]\n",
    "\n",
    "print(\"loading sources\")\n",
    "\n",
    "## Building Sequence Ontology\n",
    "so_path = \"gs://ot-team/jroldan/sequenceOntology_20221118.csv\"\n",
    "so_ontology = spark.read.csv(so_path, header=True)\n",
    "building = (\n",
    "    so_ontology.select(F.col(\"Accession\"), F.col(\"Parents\"))\n",
    "    .withColumn(\"Parentalind\", F.split(F.col(\"Parents\"), \",\"))\n",
    "    .withColumn(\"Parentalind\", F.explode_outer(\"Parentalind\"))\n",
    "    .groupBy(\"Parentalind\")\n",
    "    .agg(F.collect_list(F.col(\"Accession\")).alias(\"childrens\"))\n",
    "    .join(so_ontology, F.col(\"Parentalind\") == so_ontology.Accession, \"right\")\n",
    ")\n",
    "\n",
    "## others\n",
    "target_path = f\"{path}targets/\"\n",
    "target = spark.read.parquet(target_path)\n",
    "disease_path = f\"{path}diseases/\"\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "dis_name = diseases.select(\"id\", \"name\")\n",
    "indication_path = f\"{path}indication/\"\n",
    "indication = spark.read.parquet(indication_path)\n",
    "mecact_path = f\"{path}mechanismOfAction/\"\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "\n",
    "## annotate TSG/oncogene/bivalent using 'hallmarks.attributes'\n",
    "oncotsg_list = [\n",
    "    \"TSG\",\n",
    "    \"oncogene\",\n",
    "    \"Oncogene\",\n",
    "    \"oncogene\",\n",
    "    \"oncogene,TSG\",\n",
    "    \"TSG,oncogene\",\n",
    "    \"fusion,oncogene\",\n",
    "    \"oncogene,fusion\",\n",
    "]\n",
    "\n",
    "#### rlike('('+Keywords+')(\\s|$)'\n",
    "### on 03.07.2023 we add the categories:\n",
    "# DISRUPTING AGENT - inhibitor\n",
    "# STABILISER - activator\n",
    "\n",
    "### Hacer el join del actionType con el chembl para sacar los mecanismos de accion.\n",
    "inhibitors = [\n",
    "    \"RNAI INHIBITOR\",\n",
    "    \"NEGATIVE MODULATOR\",\n",
    "    \"NEGATIVE ALLOSTERIC MODULATOR\",\n",
    "    \"ANTAGONIST\",\n",
    "    \"ANTISENSE INHIBITOR\",\n",
    "    \"BLOCKER\",\n",
    "    \"INHIBITOR\",\n",
    "    \"DEGRADER\",\n",
    "    \"INVERSE AGONIST\",\n",
    "    \"ALLOSTERIC ANTAGONIST\",\n",
    "    \"DISRUPTING AGENT\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "activators = [\n",
    "    \"PARTIAL AGONIST\",\n",
    "    \"ACTIVATOR\",\n",
    "    \"POSITIVE ALLOSTERIC MODULATOR\",\n",
    "    \"POSITIVE MODULATOR\",\n",
    "    \"AGONIST\",\n",
    "    \"SEQUESTERING AGENT\",\n",
    "    \"STABILISER\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "columnas = [\"activator\", \"inhibitor\"]\n",
    "both = activators + inhibitors\n",
    "\n",
    "actionType = (\n",
    "    mecact.select(\n",
    "        F.explode_outer(\"chemblIds\").alias(\"drugId2\"),\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "        \"targets\",\n",
    "    )\n",
    "    .select(\n",
    "        F.explode_outer(\"targets\").alias(\"targetId2\"),\n",
    "        \"drugId2\",\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "    )\n",
    "    .groupBy(\"targetId2\", \"drugId2\")\n",
    "    .agg(\n",
    "        F.collect_set(\"actionType\").alias(\"actionType\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "oncolabel = (\n",
    "    target.select(\n",
    "        \"id\", \"approvedSymbol\", F.explode_outer(F.col(\"hallmarks.attributes\"))\n",
    "    )\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.description\")\n",
    "    .filter(F.col(\"description\").isin(oncotsg_list))\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"description\").alias(\"description\"))\n",
    "    .withColumn(\"description_splited\", F.concat_ws(\",\", F.col(\"description\")))\n",
    "    .withColumn(\n",
    "        \"TSorOncogene\",\n",
    "        F.when(\n",
    "            (\n",
    "                F.col(\"description_splited\").rlike(\"ncogene\")\n",
    "                & F.col(\"description_splited\").rlike(\"TSG\")\n",
    "            ),\n",
    "            F.lit(\"bivalent\"),\n",
    "        )\n",
    "        .when(F.col(\"description_splited\").rlike(\"ncogene(\\s|$)\"), F.lit(\"oncogene\"))\n",
    "        .when(F.col(\"description_splited\").rlike(\"TSG(\\s|$)\"), F.lit(\"TSG\"))\n",
    "        .otherwise(F.lit(\"noEvaluable\")),\n",
    "    )\n",
    "    .withColumnRenamed(\"id\", \"target_id\")\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "all = evidences.filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "\n",
    "#### version all gene burden\n",
    "prueba_assessment = (\n",
    "    all.withColumn(\"beta\", F.col(\"beta\").cast(\"double\"))  ## ot genetics & gene burden\n",
    "    .withColumn(\n",
    "        \"OddsRatio\", F.col(\"OddsRatio\").cast(\"double\")\n",
    "    )  ## ot genetics & gene burden\n",
    "    .withColumn(\n",
    "        \"clinicalSignificances\", F.concat_ws(\",\", F.col(\"clinicalSignificances\"))\n",
    "    )  ### eva\n",
    "    .join(oncolabel, oncolabel.target_id == F.col(\"targetId\"), \"left\")  ###  cgc\n",
    "    .join(\n",
    "        actionType,  ## chembl\n",
    "        (actionType.drugId2 == F.col(\"drugId\"))\n",
    "        & (actionType.targetId2 == F.col(\"targetId\")),\n",
    "        \"left\",\n",
    "    )\n",
    "    .withColumn(\"inhibitors_list\", F.array([F.lit(i) for i in inhibitors]))\n",
    "    .withColumn(\"activators_list\", F.array([F.lit(i) for i in activators]))\n",
    "    .withColumn(\n",
    "        \"intogen_function\",\n",
    "        F.when(\n",
    "            F.arrays_overlap(\n",
    "                F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                F.array([F.lit(i) for i in (gof)]),\n",
    "            ),\n",
    "            F.lit(\"GoF\"),\n",
    "        ).when(\n",
    "            F.arrays_overlap(\n",
    "                F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                F.array([F.lit(i) for i in (lof)]),\n",
    "            ),\n",
    "            F.lit(\"LoF\"),\n",
    "        ),\n",
    "        # .otherwise(\"nodata\"),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"intogenAnnot\",\n",
    "        F.size(F.collect_set(F.col(\"intogen_function\")).over(windowSpec)),\n",
    "    )\n",
    "    ### variant Effect Column\n",
    "    .withColumn(\n",
    "        \"variantEffect\",\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isNotNull(),\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\").isNull(),\n",
    "                    F.when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                        F.lit(\"LoF\"),\n",
    "                    )\n",
    "                    .when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(gof),\n",
    "                        F.lit(\"GoF\"),\n",
    "                    )\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                )\n",
    "                ### variantFunctionalConsequenceFromQtlId\n",
    "                .when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\").isNotNull(),\n",
    "                    F.when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                            var_filter_lof\n",
    "                        ),  ## when is a LoF variant\n",
    "                        F.when(\n",
    "                            F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                            == \"SO_0002316\",\n",
    "                            F.lit(\"LoF\"),\n",
    "                        )\n",
    "                        .when(\n",
    "                            F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                            == \"SO_0002315\",\n",
    "                            F.lit(\"conflict/noEvaluable\"),\n",
    "                        )\n",
    "                        .otherwise(F.lit(\"LoF\")),\n",
    "                    ).when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof)\n",
    "                        == False,  ## when is not a LoF, still can be a GoF\n",
    "                        F.when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(gof)\n",
    "                            == False,  ##if not GoF\n",
    "                            F.when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002316\",\n",
    "                                F.lit(\"LoF\"),\n",
    "                            )\n",
    "                            .when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002315\",\n",
    "                                F.lit(\"GoF\"),\n",
    "                            )\n",
    "                            .otherwise(F.lit(\"noEvaluable\")),\n",
    "                        ).when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                                gof\n",
    "                            ),  ##if is GoF\n",
    "                            F.when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002316\",\n",
    "                                F.lit(\"conflict/noEvaluable\"),\n",
    "                            ).when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002315\",\n",
    "                                F.lit(\"GoF\"),\n",
    "                            ),\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            ).when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isNull(),\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\") == \"SO_0002316\",\n",
    "                    F.lit(\"LoF\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\") == \"SO_0002315\",\n",
    "                    F.lit(\"GoF\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"datasourceId\") == \"gene_burden\",\n",
    "            F.when(F.col(\"targetId\").isNotNull(), F.lit(\"LoF\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### son tambien no data las que tiene riesgo pero no se ensayan LoF o PT\n",
    "        )\n",
    "        #### Eva_germline\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva\",\n",
    "            #### .filter(F.col('variantFunctionalConsequenceId').isin(var_filter_lof))\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                F.lit(\"LoF\"),\n",
    "            ).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son LoF\n",
    "        )\n",
    "        #### Eva_somatic\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva_somatic\",\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                F.lit(\"LoF\"),\n",
    "            ).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "        )\n",
    "        #### G2P\n",
    "        .when(\n",
    "            F.col(\"datasourceId\")\n",
    "            == \"gene2phenotype\",  ### 6 types of variants [SO_0002318, SO_0002317, SO_0001622, SO_0002315, SO_0001566, SO_0002220]\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002317\",\n",
    "                F.lit(\"LoF\"),\n",
    "            )  ### absent gene product\n",
    "            .when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002315\",\n",
    "                F.lit(\"GoF\"),\n",
    "            )  ### increased gene product level\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "        #### Orphanet\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"orphanet\",\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002054\",\n",
    "                F.lit(\"LoF\"),\n",
    "            )  ### Loss of Function Variant\n",
    "            .when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002053\",\n",
    "                F.lit(\"GoF\"),\n",
    "            )  ### Gain_of_Function Variant\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "        #### CGC\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"cancer_gene_census\",\n",
    "            F.when(F.col(\"TSorOncogene\") == \"oncogene\", F.lit(\"GoF\"))\n",
    "            .when(F.col(\"TSorOncogene\") == \"TSG\", F.lit(\"LoF\"))\n",
    "            .when(F.col(\"TSorOncogene\") == \"bivalent\", F.lit(\"bivalent\"))\n",
    "            .otherwise(\"noEvaluable\"),\n",
    "        )\n",
    "        #### intogen\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"intogen\",\n",
    "            F.when(\n",
    "                F.col(\"intogenAnnot\")\n",
    "                == 1,  ## oncogene/tummor suppressor for a given trait\n",
    "                F.when(\n",
    "                    F.arrays_overlap(\n",
    "                        F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                        F.array([F.lit(i) for i in (gof)]),\n",
    "                    ),\n",
    "                    F.lit(\"GoF\"),\n",
    "                ).when(\n",
    "                    F.arrays_overlap(\n",
    "                        F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                        F.array([F.lit(i) for i in (lof)]),\n",
    "                    ),\n",
    "                    F.lit(\"LoF\"),\n",
    "                ),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"intogenAnnot\") > 1, F.lit(\"bivalentIntogen\")\n",
    "            )  ##oncogene & tumor suppressor for a given trait\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "        #### impc\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"impc\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"LoF\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        ### chembl\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"chembl\",\n",
    "            F.when(\n",
    "                F.size(F.array_intersect(F.col(\"actionType\"), F.col(\"inhibitors_list\")))\n",
    "                >= 1,\n",
    "                F.lit(\"LoF\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.size(F.array_intersect(F.col(\"actionType\"), F.col(\"activators_list\")))\n",
    "                >= 1,\n",
    "                F.lit(\"GoF\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"directionOnTrait\",\n",
    "        ## ot genetics portal\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(F.col(\"beta\") > 0, F.lit(\"risk\"))\n",
    "                .when(F.col(\"beta\") < 0, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.when(F.col(\"OddsRatio\") > 1, F.lit(\"risk\"))\n",
    "                .when(F.col(\"OddsRatio\") < 1, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.lit(\"noEvaluable\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.lit(\"conflict/noEvaluable\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"datasourceId\") == \"gene_burden\",\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(F.col(\"beta\") > 0, F.lit(\"risk\"))\n",
    "                .when(F.col(\"beta\") < 0, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"oddsRatio\").isNotNull()) & (F.col(\"beta\").isNull()),\n",
    "                F.when(F.col(\"oddsRatio\") > 1, F.lit(\"risk\"))\n",
    "                .when(F.col(\"oddsRatio\") < 1, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"oddsRatio\").isNull()),\n",
    "                F.lit(\"noEvaluable\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"oddsRatio\").isNotNull()),\n",
    "                F.lit(\"conflict\"),\n",
    "            ),\n",
    "        )\n",
    "        ## Eva_germline\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva\",  ### the same for eva_somatic\n",
    "            F.when(F.col(\"clinicalSignificances\").rlike(\"(pathogenic)$\"), F.lit(\"risk\"))\n",
    "            .when(F.col(\"clinicalSignificances\").contains(\"protect\"), F.lit(\"protect\"))\n",
    "            .otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "        )\n",
    "        #### Eva_somatic\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva_somatic\",\n",
    "            F.when(F.col(\"clinicalSignificances\").rlike(\"(pathogenic)$\"), F.lit(\"risk\"))\n",
    "            .when(F.col(\"clinicalSignificances\").contains(\"protect\"), F.lit(\"protect\"))\n",
    "            .otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "        )\n",
    "        #### G2P\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"gene2phenotype\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### Orphanet\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"orphanet\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### CGC\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"cancer_gene_census\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### intogen\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"intogen\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### impc\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"impc\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        ### chembl\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"chembl\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"protect\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(\n",
    "            (F.col(\"variantEffect\") == \"LoF\") & (F.col(\"directionOnTrait\") == \"risk\"),\n",
    "            F.lit(\"LoF_risk\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"variantEffect\") == \"LoF\")\n",
    "            & (F.col(\"directionOnTrait\") == \"protect\"),\n",
    "            F.lit(\"LoF_protect\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"variantEffect\") == \"GoF\") & (F.col(\"directionOnTrait\") == \"risk\"),\n",
    "            F.lit(\"GoF_risk\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"variantEffect\") == \"GoF\")\n",
    "            & (F.col(\"directionOnTrait\") == \"protect\"),\n",
    "            F.lit(\"GoF_protect\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"noEvaluable\")),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "assessment = prueba_assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "\n",
    "def analysis_nonPropagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "\n",
    "\n",
    "def analysis_drugs(assessment, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment, chembl_ds)\n",
    "\n",
    "#######\n",
    "## include here the analysis\n",
    "#######\n",
    "\n",
    "analysisDatasources = []\n",
    "\n",
    "\n",
    "def full_analysis_propagation(\n",
    "    assessment, analysisDatasources, analysis_chembl, terminated_array, diseaseTA\n",
    "):\n",
    "    return (\n",
    "        analysis_propagated(assessment, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        #### Should remove the coherencyDiagonal.isNotNull()\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "#####\n",
    "## no propag\n",
    "#####\n",
    "def full_analysis_noPropagation(\n",
    "    assessment, analysisDatasources, analysis_chembl, terminated_array, diseaseTA\n",
    "):\n",
    "    return (\n",
    "        analysis_nonPropagated(assessment, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "print(\"moving to Step 3\")\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "c = datetime.now()\n",
    "print(\"starting dictionaries at\", c)\n",
    "\n",
    "#### continue here on 10.07.2024\n",
    "\n",
    "## 1nd dictionary\n",
    "dfs_dict = {}  ### checked and changed on 01.06.2023\n",
    "dfs_dict_propag = {}\n",
    "\n",
    "\n",
    "wocgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    # \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "]\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "\n",
    "germline_list = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    # \"ot_genetics_portal\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic_list = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "# assessment = prueba_assessment.filter(F.col(\"datasourceId\").isin(datasources_analysis))\n",
    "\n",
    "\n",
    "def dataset_builder(assessment, value, analysis_chembl, terminated_array, diseaseTA):\n",
    "    nonPropagated = full_analysis_noPropagation(\n",
    "        assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "    )\n",
    "    propagated = full_analysis_propagation(\n",
    "        assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "    )\n",
    "    return (\n",
    "        # Non propagation\n",
    "        ## All\n",
    "        nonPropagated,\n",
    "        ## Other\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        nonPropagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "        # Propagation\n",
    "        ## All\n",
    "        propagated,\n",
    "        ## Other\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        propagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "    )\n",
    "\n",
    "\n",
    "for value in datasource_list:\n",
    "    print(value)\n",
    "    if value == \"WOcgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment, wocgc_list, analysis_chembl, terminated_array, diseaseTA\n",
    "        )\n",
    "    elif value == \"germline\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment,\n",
    "            germline_list,\n",
    "            analysis_chembl,\n",
    "            terminated_array,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    elif value == \"somatic\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment,\n",
    "            somatic_list,\n",
    "            analysis_chembl,\n",
    "            terminated_array,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df() -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    comparisons = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"hasGeneticEvidence\", \"byDatatype\"),\n",
    "            (\"diagonalYes\", \"byDatatype\"),\n",
    "            (\"oneCellYes\", \"byDatatype\"),\n",
    "        ],\n",
    "        schema=StructType(\n",
    "            [\n",
    "                StructField(\"comparison\", StringType(), True),\n",
    "                StructField(\"comparisonType\", StringType(), True),\n",
    "            ]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "results = []\n",
    "\"\"\"\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "\n",
    "    wComparison = Window.partitionBy(comparisonColumn)\n",
    "    wPrediction = Window.partitionBy(predictionColumn)\n",
    "    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\n",
    "\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    filePath = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    listado.append(\n",
    "        filePath\n",
    "    )\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            comparisonType,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            filePath,\n",
    "        ]\n",
    "    )\n",
    "    return results\n",
    "\n",
    "\n",
    "c = datetime.now()\n",
    "\n",
    "print(\"start doing aggregations and writing\")\n",
    "today_date = str(date.today())\n",
    "aggSetups_original = comparisons_df()\n",
    "listado = []\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "\n",
    "for key, df in dfs_dict.items():\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"non propagated files wroten succesfully at\", c)\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "for key, df in dfs_dict_propag.items():\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"propagated files wroten succesfully at\", c)\n",
    "\n",
    "##### read files and make spreadsheet\n",
    "\n",
    "print(\"preparing dataframe\")\n",
    "\n",
    "schema = StructType(\n",
    "    [\n",
    "        StructField(\"group\", StringType(), True),\n",
    "        StructField(\"comparison\", StringType(), True),\n",
    "        StructField(\"phase\", StringType(), True),\n",
    "        StructField(\"oddsRatio\", DoubleType(), True),\n",
    "        StructField(\"pValue\", DoubleType(), True),\n",
    "        StructField(\"lowerInterval\", DoubleType(), True),\n",
    "        StructField(\"upperInterval\", DoubleType(), True),\n",
    "        StructField(\"total\", StringType(), True),\n",
    "        StructField(\"values\", ArrayType(ArrayType(IntegerType())), True),\n",
    "        StructField(\"relSuccess\", DoubleType(), True),\n",
    "        StructField(\"rsLower\", DoubleType(), True),\n",
    "        StructField(\"rsUpper\", DoubleType(), True),\n",
    "        StructField(\"path\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "\n",
    "df = spreadSheetFormatter(spark.createDataFrame(results, schema=schema))\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 3979:===========================================>        (169 + 8) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------+-----+-----+\n",
      "|oneCellYes|   no|  yes|\n",
      "+----------+-----+-----+\n",
      "|        no|59560|21813|\n",
      "|       yes|   26|   36|\n",
      "+----------+-----+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "dfs_dict_propag[\"df_gwas_credible_set_All_propag\"].groupBy(\"oneCellYes\").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------+-----+-----+\n",
      "|oneCellYes|   no|  yes|\n",
      "+----------+-----+-----+\n",
      "|        no|59575|21820|\n",
      "|       yes|   11|   29|\n",
      "+----------+-----+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict[\"df_gwas_credible_set_All_original\"].groupBy(\"oneCellYes\").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### extract numbers from coloc to compare with the ones for Gwas "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## potential fail: we are propagating gwasResolvedColoc and in the following steps is propagated again (check that)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\", ### left becacause is to bring data from \n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check numbers of datasets in germline and gwas credible set VS the ones provided by coloc. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sanity checks qtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'index' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_131695/786458137.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'index' is not defined"
     ]
    }
   ],
   "source": [
    "index.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 15:17:48 WARN YarnScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources\n",
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "\n",
    "#### 11.12.2024\n",
    "#######\n",
    "#######     ATENTION\n",
    "#### change code to work with generated dataframe instead of reading the parquet\n",
    "\n",
    "\"\"\"\n",
    "This scripts run Odds ratio analysis for DoE and \n",
    "genetic information on drug clinical success\n",
    "\n",
    "\"\"\"\n",
    "from functions import discrepancifier\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 8:===================================================>       (7 + 1) / 8]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------+------+\n",
      "|studyType| count|\n",
      "+---------+------+\n",
      "|     gwas| 90839|\n",
      "|     sqtl|184609|\n",
      "|     pqtl|  3757|\n",
      "|    tuqtl|301140|\n",
      "|     eqtl|948158|\n",
      "|  sctuqtl| 63143|\n",
      "|   sceqtl|335808|\n",
      "|   scsqtl| 29350|\n",
      "+---------+------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "index.groupBy(\"studyType\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------------------------------------------------------------------------------+--------------------------------+--------------------------------+----------+--------------+--------------------------+----------------------+----------------------+----------------------+---------------------+--------------------+--------------------+--------------------+-------------------------+-----------------+---------------------+------------------+----------------------+---------------+----------+--------------+-----------+--------------+\n",
      "|rightStudyId                                                                      |rightStudyLocusId               |leftStudyLocusId                |chromosome|rightStudyType|numberColocalisingVariants|h0                    |h1                    |h2                    |h3                   |h4                  |colocalisationMethod|betaRatioSignAverage|leftStudyId              |leftVariantId    |credibleLeftStudyType|rightVariantId    |credibleRightStudyType|geneId         |projectId |indexStudyType|condition  |biosampleId   |\n",
      "+----------------------------------------------------------------------------------+--------------------------------+--------------------------------+----------+--------------+--------------------------+----------------------+----------------------+----------------------+---------------------+--------------------+--------------------+--------------------+-------------------------+-----------------+---------------------+------------------+----------------------+---------------+----------+--------------+-----------+--------------+\n",
      "|GCST90014010                                                                      |0005218bc3a62e3875f2ba0a0b63c349|0af529199eecb9b6a0779f85fc9d237f|19        |gwas          |1                         |3.1079843775273334E-23|9.978101816041437E-5  |3.11449443689344E-22  |3.107984377838132E-31|0.9999002189818393  |COLOC               |1.0                 |GCST90018951             |19_38320104_C_G  |gwas                 |19_38320104_C_G   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90014010                                                                      |0005218bc3a62e3875f2ba0a0b63c349|e9cf6c4f6215de207532fb58572346be|19        |gwas          |1                         |1.109663241911948E-20 |1.1058085511196052E-4 |1.111998666051756E-19 |1.083547982371386E-4 |0.9997810643466477  |COLOC               |1.0                 |GCST90019500             |19_38320104_C_G  |gwas                 |19_38320104_C_G   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90014010                                                                      |0005218bc3a62e3875f2ba0a0b63c349|bbf92fabf4ac4ccc8fa09075029a8d07|19        |gwas          |1                         |6.303952273206082E-31 |9.978101816041437E-5  |6.317156684346715E-30 |6.303952273836433E-39|0.9999002189818464  |COLOC               |1.0                 |GCST90013999             |19_38320104_C_G  |gwas                 |19_38320104_C_G   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GTEx_tx_minor_salivary_gland_ENST00000650739                                      |0008e0a6a8515514f42c759a5ab0d592|0db850508874cf448fc1b680e101afae|3         |eqtl          |2                         |1.6692899840976926E-59|2.749129180388412E-56 |5.9629969759684814E-6 |0.008829207391733962 |0.9911648296112845  |COLOC               |1.0                 |GCST90264805             |3_53181859_A_C   |gwas                 |3_53173500_C_G    |eqtl                  |ENSG00000163932|GTEx      |eqtl          |naive      |UBERON_0006330|\n",
      "|Aygun_2021_ge_Progenitor_ENSG00000074071                                          |000c0ca56b1f4c71a9e27c0c6cd04eb4|669f43f12522c92e9fd07912b2075590|16        |sceqtl        |1                         |1.8728911667528647E-12|5.0981005772277637E-11|0.03390557889349678   |0.9228830878380463   |0.043211333215604684|COLOC               |-1.0                |FINNGEN_R11_I9_AF_REIMB  |16_1197322_G_A   |gwas                 |16_1772796_C_CACCT|sceqtl                |ENSG00000074071|Aygun_2021|sceqtl        |naive      |CL_0011020    |\n",
      "|Aygun_2021_ge_Progenitor_ENSG00000074071                                          |000c0ca56b1f4c71a9e27c0c6cd04eb4|640a77f297cdc96f42c645609779b92b|16        |sceqtl        |1                         |1.0018935228511351E-12|4.3044595269698714E-11|0.018137615513845854  |0.7790479753646176   |0.20281440907749101 |COLOC               |-1.0                |FINNGEN_R11_I9_HYPTENSESS|16_1928335_A_C   |gwas                 |16_1772796_C_CACCT|sceqtl                |ENSG00000074071|Aygun_2021|sceqtl        |naive      |CL_0011020    |\n",
      "|Braineac2_ge_putamen_ENSG00000147813                                              |0017a37fdd49794eb924301ebe645257|60034b5445cca196ec80c4e92df85a6f|8         |eqtl          |7                         |2.0286256066408496E-27|1.0454912920951048E-19|1.3487858992721654E-10|0.0059571851890201305|0.9940428146761024  |COLOC               |1.0                 |GCST90199815             |8_143594692_C_T  |gwas                 |8_143599607_G_T   |eqtl                  |ENSG00000147813|Braineac2 |eqtl          |naive      |UBERON_0001874|\n",
      "|Braineac2_ge_putamen_ENSG00000147813                                              |0017a37fdd49794eb924301ebe645257|0b51f477621f47ce312001af30028086|8         |eqtl          |5                         |1.699948265442422E-22 |2.0610315664776821E-19|1.1302559932276585E-5 |0.012716046017728619 |0.9872726514223361  |COLOC               |-1.0                |GCST90278652             |8_143602045_A_G  |gwas                 |8_143599607_G_T   |eqtl                  |ENSG00000147813|Braineac2 |eqtl          |naive      |UBERON_0001874|\n",
      "|GTEx_exon_fibroblast_ENSG00000109667.12_4_10040211_10040248                       |0022d49130165f188cf0ceebaa4ed712|78541a0252645ecc26daa450eeb45fd0|4         |sceqtl        |1                         |0.0                   |5.866267082610005E-7  |0.0                   |0.02623851213462055  |0.9737609012386275  |COLOC               |1.0                 |GCST90399704             |4_10079577_G_A   |gwas                 |4_10102479_C_G    |sceqtl                |ENSG00000109667|GTEx      |sceqtl        |naive      |CL_0000057    |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|382ec123cd4bcde15d03be6f3bff1da3|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |-1.0                |GCST90085882             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|17fb120abc3f297bd2d0728b5324e880|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST90018964             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|3de3540041c44075b6cdb608befa68ba|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |-1.0                |GCST90002402             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|9c70b27b12cd9bf810b335f0a8dc4bf1|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST90018966             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|cb2781338d759a3e4e5b06769e565274|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |-1.0                |GCST004607               |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|b07208c748f744d3084a00e5a1fb5402|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST004630               |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|a37dcb278793694be7011b2ce1f4b734|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST90002334             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|4ddb2d5f048fbc3922dcd0df622aaeea|17        |sctuqtl       |7                         |2.5407265689237554E-42|8.412145756033199E-35 |1.968303847994525E-9  |0.06423322490039893  |0.935766773131297   |COLOC               |-1.0                |GCST90092806             |17_81674504_C_T  |gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|674a46d781eb51b85ab915d278acf5a5|17        |sctuqtl       |8                         |3.4697676160520917E-41|5.492578537938273E-35 |2.6880330350601647E-8 |0.04159266293826143  |0.9584073101814033  |COLOC               |-1.0                |GCST90269557             |17_81677012_A_G  |gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|ac62ddf72269de27d846904cffe3afb7|17        |sctuqtl       |1                         |1.7018073881445055E-43|2.0155117691098697E-35|1.3183921763172077E-10|0.014628822098044246 |0.985371177770121   |COLOC               |-1.0                |GCST90014001             |17_81667032_CTG_C|gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|b8975dcb3b2695d8c53fce3fe4c08a0c|17        |sctuqtl       |10                        |2.981602221346924E-41 |1.6489190383377602E-34|2.3098507321675852E-8 |0.12686881982218193  |0.8731311570793094  |COLOC               |1.0                 |GCST90268498             |17_81688016_C_T  |gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "+----------------------------------------------------------------------------------+--------------------------------+--------------------------------+----------+--------------+--------------------------+----------------------+----------------------+----------------------+---------------------+--------------------+--------------------+--------------------+-------------------------+-----------------+---------------------+------------------+----------------------+---------------+----------+--------------+-----------+--------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "newColoc.show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.types import IntegerType, StringType, StructField, StructType\n",
    "import pandas as pd\n",
    "\n",
    "pd.DataFrame.iteritems = pd.DataFrame.items\n",
    "\n",
    "raw_studies_metadata_schema: StructType = StructType(\n",
    "        [\n",
    "            StructField(\"study_id\", StringType(), True),\n",
    "            StructField(\"dataset_id\", StringType(), True),\n",
    "            StructField(\"study_label\", StringType(), True),\n",
    "            StructField(\"sample_group\", StringType(), True),\n",
    "            StructField(\"tissue_id\", StringType(), True),\n",
    "            StructField(\"tissue_label\", StringType(), True),\n",
    "            StructField(\"condition_label\", StringType(), True),\n",
    "            StructField(\"sample_size\", IntegerType(), True),\n",
    "            StructField(\"quant_method\", StringType(), True),\n",
    "            StructField(\"pmid\", StringType(), True),\n",
    "            StructField(\"study_type\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "            pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "            schema=raw_studies_metadata_schema,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------------------------+\n",
      "|extracted_column                                        |\n",
      "+--------------------------------------------------------+\n",
      "|Alasoo_2018_ge_macrophage_naive                         |\n",
      "|Alasoo_2018_exon_macrophage_naive                       |\n",
      "|Alasoo_2018_tx_macrophage_naive                         |\n",
      "|Alasoo_2018_txrev_macrophage_naive                      |\n",
      "|Alasoo_2018_leafcutter_macrophage_naive                 |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h                      |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h                    |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h                      |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h                   |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h              |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_5h                 |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_5h               |\n",
      "|Alasoo_2018_tx_macrophage_Salmonella_5h                 |\n",
      "|Alasoo_2018_txrev_macrophage_Salmonella_5h              |\n",
      "|Alasoo_2018_leafcutter_macrophage_Salmonella_5h         |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h+Salmonella_5h        |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h+Salmonella_5h      |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h+Salmonella_5h        |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h+Salmonella_5h     |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h+Salmonella_5h|\n",
      "+--------------------------------------------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_table.withColumn(\"extracted_column\", \n",
    "    F.concat_ws(\"_\", \n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"tissue_label\"),\n",
    "        F.col(\"condition_label\"))\n",
    "        ).select(\"extracted_column\").show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------------------------------------------------------------------------+\n",
      "|studyId                                                                             |\n",
      "+------------------------------------------------------------------------------------+\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000015532.10_17_50354854_50355056 |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000106771.13_9_109108453_109108488|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135218.19_7_80546191_80546267  |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135899.19_2_230177366_230177537|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000138600.10_15_50748113_50748202 |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000141524.17_17_78119062_78119296 |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168385.18_2_241324663_241325244|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168994.14_6_3739058_3739161    |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000171109.19_3_179348050_179348121|\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000067334.14_1_93877473_93877489             |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000092108.22_14_30638107_30638124            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000123360.12_12_54561501_54561637            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000128951.14_15_48331011_48331094            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000132199.20_18_691204_691276                |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000135124.16_12_121234071_121234104          |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000145191.15_3_184135389_184135391           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157106.18_16_18860667_18860776            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157379.14_14_24290636_24290995            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000158186.13_3_138405534_138405534           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000177951.18_11_205074_205102                |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000026508.21_11_35190066_35190429      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000092094.11_14_20449558_20450879      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000130177.16_13_114250549_114250674    |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000198912.12_1_3891058_3891534         |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000214113.11_6_5236284_5236461         |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000224272.2_2_241809110_241809368      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000256377.6_12_27696979_27697298       |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000035115.22_2_218136_218143                |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000108826.16_17_50368314_50368837           |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000138398.17_2_169636413_169636706          |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000147677.12_8_116734277_116734324          |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000148396.19_9_136474487_136475197          |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000159063.14_11_78104454_78104549           |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000233927.5_19_8321953_8322091              |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella_ENSG00000197013                           |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella_ENSG00000204103                           |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000001561                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000081791                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000100300                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000114446                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000136250                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000144566                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000165934                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000196862                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000225434                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000227354                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000234171                                      |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_ENSG00000006282                                |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_ENSG00000130414                                |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_ENSG00000142794                                |\n",
      "+------------------------------------------------------------------------------------+\n",
      "only showing top 50 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "index.select(\"studyId\").show(50,truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 55:===========================================>              (3 + 1) / 4]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------------------------------------+\n",
      "|extracted_column                           |\n",
      "+-------------------------------------------+\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella  |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella  |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella       |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella       |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella       |\n",
      "+-------------------------------------------+\n",
      "only showing top 50 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "index.withColumn(\"extracted_column\", F.split(F.col(\"studyId\"), \"_ENS\")[0]).select(\"extracted_column\").show(50,truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------+----------+-----------+--------------------+----------+------------+--------------------+-----------+------------+--------+----------+\n",
      "| study_id|dataset_id|study_label|        sample_group| tissue_id|tissue_label|     condition_label|sample_size|quant_method|    pmid|study_type|\n",
      "+---------+----------+-----------+--------------------+----------+------------+--------------------+-----------+------------+--------+----------+\n",
      "|QTS000001| QTD000001|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000002|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000003|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000004|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000005|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|  leafcutter|29379200|      bulk|\n",
      "|QTS000001| QTD000006|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000007|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000008|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000009|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000010|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|  leafcutter|29379200|      bulk|\n",
      "|QTS000001| QTD000011|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000012|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000013|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000014|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000015|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|  leafcutter|29379200|      bulk|\n",
      "|QTS000001| QTD000016|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000017|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000018|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000019|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000020|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|  leafcutter|29379200|      bulk|\n",
      "+---------+----------+-----------+--------------------+----------+------------+--------------------+-----------+------------+--------+----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_table.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------------------------+----------+\n",
      "|extracted_column                                        |study_type|\n",
      "+--------------------------------------------------------+----------+\n",
      "|Alasoo_2018_ge_macrophage_naive                         |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_naive                       |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_naive                         |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_naive                      |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_naive                 |bulk      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h                      |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h                    |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h                      |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h                   |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h              |bulk      |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_5h                 |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_5h               |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_Salmonella_5h                 |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_Salmonella_5h              |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_Salmonella_5h         |bulk      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h+Salmonella_5h        |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h+Salmonella_5h      |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h+Salmonella_5h        |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h+Salmonella_5h     |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h+Salmonella_5h|bulk      |\n",
      "|BLUEPRINT_ge_monocyte_naive                             |bulk      |\n",
      "|BLUEPRINT_exon_monocyte_naive                           |bulk      |\n",
      "|BLUEPRINT_tx_monocyte_naive                             |bulk      |\n",
      "|BLUEPRINT_txrev_monocyte_naive                          |bulk      |\n",
      "|BLUEPRINT_leafcutter_monocyte_naive                     |bulk      |\n",
      "|BLUEPRINT_ge_neutrophil_naive                           |bulk      |\n",
      "|BLUEPRINT_exon_neutrophil_naive                         |bulk      |\n",
      "|BLUEPRINT_tx_neutrophil_naive                           |bulk      |\n",
      "|BLUEPRINT_txrev_neutrophil_naive                        |bulk      |\n",
      "|BLUEPRINT_leafcutter_neutrophil_naive                   |bulk      |\n",
      "|BLUEPRINT_ge_CD4+ T cell_naive                          |bulk      |\n",
      "|BLUEPRINT_exon_CD4+ T cell_naive                        |bulk      |\n",
      "|BLUEPRINT_tx_CD4+ T cell_naive                          |bulk      |\n",
      "|BLUEPRINT_txrev_CD4+ T cell_naive                       |bulk      |\n",
      "|BLUEPRINT_leafcutter_CD4+ T cell_naive                  |bulk      |\n",
      "|Bossini-Castillo_2019_ge_Treg memory_naive              |bulk      |\n",
      "|Bossini-Castillo_2019_exon_Treg memory_naive            |bulk      |\n",
      "|Bossini-Castillo_2019_tx_Treg memory_naive              |bulk      |\n",
      "|Bossini-Castillo_2019_txrev_Treg memory_naive           |bulk      |\n",
      "|Bossini-Castillo_2019_leafcutter_Treg memory_naive      |bulk      |\n",
      "|Braineac2_ge_brain (putamen)_naive                      |bulk      |\n",
      "|Braineac2_exon_brain (putamen)_naive                    |bulk      |\n",
      "|Braineac2_tx_brain (putamen)_naive                      |bulk      |\n",
      "|Braineac2_txrev_brain (putamen)_naive                   |bulk      |\n",
      "|Braineac2_leafcutter_brain (putamen)_naive              |bulk      |\n",
      "|Braineac2_ge_brain (substantia nigra)_naive             |bulk      |\n",
      "|Braineac2_exon_brain (substantia nigra)_naive           |bulk      |\n",
      "|Braineac2_tx_brain (substantia nigra)_naive             |bulk      |\n",
      "|Braineac2_txrev_brain (substantia nigra)_naive          |bulk      |\n",
      "|Braineac2_leafcutter_brain (substantia nigra)_naive     |bulk      |\n",
      "|BrainSeq_ge_brain (DLPFC)_naive                         |bulk      |\n",
      "|BrainSeq_exon_brain (DLPFC)_naive                       |bulk      |\n",
      "|BrainSeq_tx_brain (DLPFC)_naive                         |bulk      |\n",
      "|BrainSeq_txrev_brain (DLPFC)_naive                      |bulk      |\n",
      "|BrainSeq_leafcutter_brain (DLPFC)_naive                 |bulk      |\n",
      "|CAP_ge_LCL_naive                                        |bulk      |\n",
      "|CAP_exon_LCL_naive                                      |bulk      |\n",
      "|CAP_tx_LCL_naive                                        |bulk      |\n",
      "|CAP_txrev_LCL_naive                                     |bulk      |\n",
      "|CAP_leafcutter_LCL_naive                                |bulk      |\n",
      "|CAP_ge_LCL_statin                                       |bulk      |\n",
      "|CAP_exon_LCL_statin                                     |bulk      |\n",
      "|CAP_tx_LCL_statin                                       |bulk      |\n",
      "|CAP_txrev_LCL_statin                                    |bulk      |\n",
      "|CAP_leafcutter_LCL_statin                               |bulk      |\n",
      "|CEDAR_microarray_CD8+ T cell_naive                      |bulk      |\n",
      "|CEDAR_microarray_CD4+ T cell_naive                      |bulk      |\n",
      "|CEDAR_microarray_transverse colon_naive                 |bulk      |\n",
      "|CEDAR_microarray_monocyte_naive                         |bulk      |\n",
      "|CEDAR_microarray_neutrophil_naive                       |bulk      |\n",
      "|CEDAR_microarray_platelet_naive                         |bulk      |\n",
      "|CEDAR_microarray_rectum_naive                           |bulk      |\n",
      "|CEDAR_microarray_B cell_naive                           |bulk      |\n",
      "|CEDAR_microarray_ileum_naive                            |bulk      |\n",
      "|CommonMind_ge_brain (DLPFC)_naive                       |bulk      |\n",
      "|CommonMind_exon_brain (DLPFC)_naive                     |bulk      |\n",
      "|CommonMind_tx_brain (DLPFC)_naive                       |bulk      |\n",
      "|CommonMind_txrev_brain (DLPFC)_naive                    |bulk      |\n",
      "|CommonMind_leafcutter_brain (DLPFC)_naive               |bulk      |\n",
      "|Fairfax_2012_microarray_B cell_naive                    |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_naive                  |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_IFNg_24h               |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_LPS_2h                 |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_LPS_24h                |bulk      |\n",
      "|FUSION_ge_muscle_naive                                  |bulk      |\n",
      "|FUSION_exon_muscle_naive                                |bulk      |\n",
      "|FUSION_tx_muscle_naive                                  |bulk      |\n",
      "|FUSION_txrev_muscle_naive                               |bulk      |\n",
      "|FUSION_leafcutter_muscle_naive                          |bulk      |\n",
      "|FUSION_ge_adipose_naive                                 |bulk      |\n",
      "|FUSION_exon_adipose_naive                               |bulk      |\n",
      "|FUSION_tx_adipose_naive                                 |bulk      |\n",
      "|FUSION_txrev_adipose_naive                              |bulk      |\n",
      "|FUSION_leafcutter_adipose_naive                         |bulk      |\n",
      "|GENCORD_ge_LCL_naive                                    |bulk      |\n",
      "|GENCORD_exon_LCL_naive                                  |bulk      |\n",
      "|GENCORD_tx_LCL_naive                                    |bulk      |\n",
      "|GENCORD_txrev_LCL_naive                                 |bulk      |\n",
      "|GENCORD_leafcutter_LCL_naive                            |bulk      |\n",
      "|GENCORD_ge_fibroblast_naive                             |bulk      |\n",
      "|GENCORD_exon_fibroblast_naive                           |bulk      |\n",
      "|GENCORD_tx_fibroblast_naive                             |bulk      |\n",
      "|GENCORD_txrev_fibroblast_naive                          |bulk      |\n",
      "|GENCORD_leafcutter_fibroblast_naive                     |bulk      |\n",
      "|GENCORD_ge_T cell_naive                                 |bulk      |\n",
      "|GENCORD_exon_T cell_naive                               |bulk      |\n",
      "|GENCORD_tx_T cell_naive                                 |bulk      |\n",
      "|GENCORD_txrev_T cell_naive                              |bulk      |\n",
      "|GENCORD_leafcutter_T cell_naive                         |bulk      |\n",
      "|GEUVADIS_ge_LCL_naive                                   |bulk      |\n",
      "|GEUVADIS_exon_LCL_naive                                 |bulk      |\n",
      "|GEUVADIS_tx_LCL_naive                                   |bulk      |\n",
      "|GEUVADIS_txrev_LCL_naive                                |bulk      |\n",
      "|GEUVADIS_leafcutter_LCL_naive                           |bulk      |\n",
      "|Gilchrist_2021_microarray_NK cell_naive                 |bulk      |\n",
      "|GTEx_ge_adipose_naive                                   |bulk      |\n",
      "|GTEx_exon_adipose_naive                                 |bulk      |\n",
      "|GTEx_tx_adipose_naive                                   |bulk      |\n",
      "|GTEx_txrev_adipose_naive                                |bulk      |\n",
      "|GTEx_leafcutter_adipose_naive                           |bulk      |\n",
      "|GTEx_ge_adipose (visceral)_naive                        |bulk      |\n",
      "|GTEx_exon_adipose (visceral)_naive                      |bulk      |\n",
      "|GTEx_tx_adipose (visceral)_naive                        |bulk      |\n",
      "|GTEx_txrev_adipose (visceral)_naive                     |bulk      |\n",
      "|GTEx_leafcutter_adipose (visceral)_naive                |bulk      |\n",
      "|GTEx_ge_adrenal gland_naive                             |bulk      |\n",
      "|GTEx_exon_adrenal gland_naive                           |bulk      |\n",
      "|GTEx_tx_adrenal gland_naive                             |bulk      |\n",
      "|GTEx_txrev_adrenal gland_naive                          |bulk      |\n",
      "|GTEx_leafcutter_adrenal gland_naive                     |bulk      |\n",
      "|GTEx_ge_artery (aorta)_naive                            |bulk      |\n",
      "|GTEx_exon_artery (aorta)_naive                          |bulk      |\n",
      "|GTEx_tx_artery (aorta)_naive                            |bulk      |\n",
      "|GTEx_txrev_artery (aorta)_naive                         |bulk      |\n",
      "|GTEx_leafcutter_artery (aorta)_naive                    |bulk      |\n",
      "|GTEx_ge_artery (coronary)_naive                         |bulk      |\n",
      "|GTEx_exon_artery (coronary)_naive                       |bulk      |\n",
      "|GTEx_tx_artery (coronary)_naive                         |bulk      |\n",
      "|GTEx_txrev_artery (coronary)_naive                      |bulk      |\n",
      "|GTEx_leafcutter_artery (coronary)_naive                 |bulk      |\n",
      "|GTEx_ge_artery (tibial)_naive                           |bulk      |\n",
      "|GTEx_exon_artery (tibial)_naive                         |bulk      |\n",
      "|GTEx_tx_artery (tibial)_naive                           |bulk      |\n",
      "|GTEx_txrev_artery (tibial)_naive                        |bulk      |\n",
      "|GTEx_leafcutter_artery (tibial)_naive                   |bulk      |\n",
      "|GTEx_ge_brain (amygdala)_naive                          |bulk      |\n",
      "|GTEx_exon_brain (amygdala)_naive                        |bulk      |\n",
      "|GTEx_tx_brain (amygdala)_naive                          |bulk      |\n",
      "|GTEx_txrev_brain (amygdala)_naive                       |bulk      |\n",
      "|GTEx_leafcutter_brain (amygdala)_naive                  |bulk      |\n",
      "|GTEx_ge_brain (anterior cingulate cortex)_naive         |bulk      |\n",
      "|GTEx_exon_brain (anterior cingulate cortex)_naive       |bulk      |\n",
      "|GTEx_tx_brain (anterior cingulate cortex)_naive         |bulk      |\n",
      "|GTEx_txrev_brain (anterior cingulate cortex)_naive      |bulk      |\n",
      "|GTEx_leafcutter_brain (anterior cingulate cortex)_naive |bulk      |\n",
      "|GTEx_ge_brain (caudate)_naive                           |bulk      |\n",
      "|GTEx_exon_brain (caudate)_naive                         |bulk      |\n",
      "|GTEx_tx_brain (caudate)_naive                           |bulk      |\n",
      "|GTEx_txrev_brain (caudate)_naive                        |bulk      |\n",
      "|GTEx_leafcutter_brain (caudate)_naive                   |bulk      |\n",
      "|GTEx_ge_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_exon_brain (cerebellum)_naive                      |bulk      |\n",
      "|GTEx_tx_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_txrev_brain (cerebellum)_naive                     |bulk      |\n",
      "|GTEx_leafcutter_brain (cerebellum)_naive                |bulk      |\n",
      "|GTEx_ge_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_exon_brain (cerebellum)_naive                      |bulk      |\n",
      "|GTEx_tx_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_txrev_brain (cerebellum)_naive                     |bulk      |\n",
      "|GTEx_leafcutter_brain (cerebellum)_naive                |bulk      |\n",
      "|GTEx_ge_brain (cortex)_naive                            |bulk      |\n",
      "|GTEx_exon_brain (cortex)_naive                          |bulk      |\n",
      "|GTEx_tx_brain (cortex)_naive                            |bulk      |\n",
      "|GTEx_txrev_brain (cortex)_naive                         |bulk      |\n",
      "|GTEx_leafcutter_brain (cortex)_naive                    |bulk      |\n",
      "|GTEx_ge_brain (DLPFC)_naive                             |bulk      |\n",
      "|GTEx_exon_brain (DLPFC)_naive                           |bulk      |\n",
      "|GTEx_tx_brain (DLPFC)_naive                             |bulk      |\n",
      "|GTEx_txrev_brain (DLPFC)_naive                          |bulk      |\n",
      "|GTEx_leafcutter_brain (DLPFC)_naive                     |bulk      |\n",
      "|GTEx_ge_brain (hippocampus)_naive                       |bulk      |\n",
      "|GTEx_exon_brain (hippocampus)_naive                     |bulk      |\n",
      "|GTEx_tx_brain (hippocampus)_naive                       |bulk      |\n",
      "|GTEx_txrev_brain (hippocampus)_naive                    |bulk      |\n",
      "|GTEx_leafcutter_brain (hippocampus)_naive               |bulk      |\n",
      "|GTEx_ge_brain (hypothalamus)_naive                      |bulk      |\n",
      "|GTEx_exon_brain (hypothalamus)_naive                    |bulk      |\n",
      "|GTEx_tx_brain (hypothalamus)_naive                      |bulk      |\n",
      "|GTEx_txrev_brain (hypothalamus)_naive                   |bulk      |\n",
      "|GTEx_leafcutter_brain (hypothalamus)_naive              |bulk      |\n",
      "|GTEx_ge_brain (nucleus accumbens)_naive                 |bulk      |\n",
      "|GTEx_exon_brain (nucleus accumbens)_naive               |bulk      |\n",
      "|GTEx_tx_brain (nucleus accumbens)_naive                 |bulk      |\n",
      "|GTEx_txrev_brain (nucleus accumbens)_naive              |bulk      |\n",
      "|GTEx_leafcutter_brain (nucleus accumbens)_naive         |bulk      |\n",
      "|GTEx_ge_brain (putamen)_naive                           |bulk      |\n",
      "|GTEx_exon_brain (putamen)_naive                         |bulk      |\n",
      "|GTEx_tx_brain (putamen)_naive                           |bulk      |\n",
      "|GTEx_txrev_brain (putamen)_naive                        |bulk      |\n",
      "|GTEx_leafcutter_brain (putamen)_naive                   |bulk      |\n",
      "+--------------------------------------------------------+----------+\n",
      "only showing top 200 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_table.select(\n",
    "    F.concat_ws(\n",
    "        \"_\",\n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"sample_group\"),\n",
    "    ).alias(\"extracted_column\"),\n",
    "    \"study_type\",\n",
    ").show(200,truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "saved=study_table.select(\n",
    "    F.concat_ws(\n",
    "        \"_\",\n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"sample_group\"),\n",
    "    ).alias(\"extracted_column\"),\n",
    "    \"study_type\",\n",
    ").join(\n",
    "    index.withColumn(\"extracted_column\", F.split(F.col(\"studyId\"), \"_ENS\")[0]),\n",
    "    on=\"extracted_column\",\n",
    "    how=\"right\",\n",
    ").persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|    extracted_column|study_type|             studyId|         geneId|  projectId|studyType|     traitFromSource|traitFromSourceMappedIds|biosampleFromSourceId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|initialSampleSize|nCases|nControls|nSamples|cohorts|ldPopulationStructure|discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|           condition|sumstatQCValues|diseaseIds|backgroundDiseaseIds|biosampleId|\n",
      "+--------------------+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000140319|Alasoo_2018|   scsqtl|15:40037018:40038...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000065320|Alasoo_2018|   scsqtl|17:9182969:922116...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000169136|Alasoo_2018|   scsqtl|19:49929170:49929...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000167543|Alasoo_2018|   scsqtl|17:29568508:29569...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000196735|Alasoo_2018|   scsqtl|6:32642784:326429...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000179630|Alasoo_2018|   scsqtl|13:43879743:43880...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|       Salmonella_5h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000120910|Alasoo_2018|   scsqtl|8:22475624:224980...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|       Salmonella_5h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000184110|Alasoo_2018|   scsqtl|16:28715434:28723...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000102858|Alasoo_2018|   scsqtl|16:4686303:468879...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000167863|Alasoo_2018|   scsqtl|17:75042277:75046...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000173875|Alasoo_2018|   scsqtl|19:12611082:12623...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000116473|Alasoo_2018|   scsqtl|1:111699607:11170...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000119782|Alasoo_2018|   scsqtl|2:24049886:240539...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000135605|Alasoo_2018|   scsqtl|4:48141419:481454...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000289047|Alasoo_2018|   scsqtl|6:32894927:329009...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000127952|Alasoo_2018|   scsqtl|7:76001002:760037...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Aygun_2021_leafcu...|      null|Aygun_2021_leafcu...|ENSG00000184009| Aygun_2021|   scsqtl|17:81512360:81512...|                    null|           CL_0011020|34416157|            null|                  null|           null|              null|                              null|             null|  null|     null|      85|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0011020|\n",
      "|Aygun_2021_leafcu...|      null|Aygun_2021_leafcu...|ENSG00000165119| Aygun_2021|   scsqtl|9:83978452:839801...|                    null|           CL_0011020|34416157|            null|                  null|           null|              null|                              null|             null|  null|     null|      85|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0011020|\n",
      "|BLUEPRINT_leafcut...|      null|BLUEPRINT_leafcut...|ENSG00000237489|  BLUEPRINT|   scsqtl|10:130079901:1300...|                    null|           CL_0000624|27863251|            null|                  null|           null|              null|                              null|             null|  null|     null|     167|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000624|\n",
      "|BLUEPRINT_leafcut...|      null|BLUEPRINT_leafcut...|ENSG00000119943|  BLUEPRINT|   scsqtl|10:98387307:98388...|                    null|           CL_0000624|27863251|            null|                  null|           null|              null|                              null|             null|  null|     null|     167|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000624|\n",
      "+--------------------+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "saved.filter(F.col(\"study_type\").isNull()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "index.withColumn(\"extracted_column\", \n",
    "            F.split(F.col(\"studyId\"), \"_ENS\")[0]).select(\"extra\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 31:==================================================>       (7 + 1) / 8]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+--------+----------+-----------+------------+---------+------------+---------------+-----------+------------+----+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|    extracted_column|study_id|dataset_id|study_label|sample_group|tissue_id|tissue_label|condition_label|sample_size|quant_method|pmid|study_type|             studyId|         geneId|  projectId|studyType|     traitFromSource|traitFromSourceMappedIds|biosampleFromSourceId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|initialSampleSize|nCases|nControls|nSamples|cohorts|ldPopulationStructure|discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|           condition|sumstatQCValues|diseaseIds|backgroundDiseaseIds|biosampleId|\n",
      "+--------------------+--------+----------+-----------+------------+---------+------------+---------------+-----------+------------+----+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000026508|Alasoo_2018|   sceqtl|ENSG00000026508.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|       Salmonella_5h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000015532|Alasoo_2018|   sceqtl|ENSG00000015532.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000106771|Alasoo_2018|   sceqtl|ENSG00000106771.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000135218|Alasoo_2018|   sceqtl|ENSG00000135218.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000135899|Alasoo_2018|   sceqtl|ENSG00000135899.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000138600|Alasoo_2018|   sceqtl|ENSG00000138600.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000141524|Alasoo_2018|   sceqtl|ENSG00000141524.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000168385|Alasoo_2018|   sceqtl|ENSG00000168385.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000168994|Alasoo_2018|   sceqtl|ENSG00000168994.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000171109|Alasoo_2018|   sceqtl|ENSG00000171109.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000067334|Alasoo_2018|   sceqtl|ENSG00000067334.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000092108|Alasoo_2018|   sceqtl|ENSG00000092108.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000123360|Alasoo_2018|   sceqtl|ENSG00000123360.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000128951|Alasoo_2018|   sceqtl|ENSG00000128951.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000132199|Alasoo_2018|   sceqtl|ENSG00000132199.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000135124|Alasoo_2018|   sceqtl|ENSG00000135124.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000145191|Alasoo_2018|   sceqtl|ENSG00000145191.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000157106|Alasoo_2018|   sceqtl|ENSG00000157106.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000157379|Alasoo_2018|   sceqtl|ENSG00000157379.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000158186|Alasoo_2018|   sceqtl|ENSG00000158186.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "+--------------------+--------+----------+-----------+------------+---------+------------+---------------+-----------+------------+----+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "study_table.withColumn(\"extracted_column\", \n",
    "    F.concat_ws(\"_\", \n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"tissue_label\"),\n",
    "        F.col(\"condition_label\"))\n",
    "        ).join(index.withColumn(\"extracted_column\", \n",
    "            F.split(F.col(\"studyId\"), \"_ENS\")[0]\n",
    "), on=\"extracted_column\", how=\"right\").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_136690/275740773.py\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m study_table = spark.createDataFrame(\n\u001b[0;32m----> 5\u001b[0;31m             \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mraw_studies_metadata_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m             \u001b[0mschema\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraw_studies_metadata_schema\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "#### read metadata table\n",
    "\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "            pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "            schema=raw_studies_metadata_schema,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------------------------------------------------------------------------+-------------------------------------------+\n",
      "|studyId                                                                             |extracted_column                           |\n",
      "+------------------------------------------------------------------------------------+-------------------------------------------+\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000015532.10_17_50354854_50355056 |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000106771.13_9_109108453_109108488|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135218.19_7_80546191_80546267  |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135899.19_2_230177366_230177537|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000138600.10_15_50748113_50748202 |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000141524.17_17_78119062_78119296 |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168385.18_2_241324663_241325244|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168994.14_6_3739058_3739161    |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000171109.19_3_179348050_179348121|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000067334.14_1_93877473_93877489             |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000092108.22_14_30638107_30638124            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000123360.12_12_54561501_54561637            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000128951.14_15_48331011_48331094            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000132199.20_18_691204_691276                |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000135124.16_12_121234071_121234104          |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000145191.15_3_184135389_184135391           |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157106.18_16_18860667_18860776            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157379.14_14_24290636_24290995            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000158186.13_3_138405534_138405534           |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000177951.18_11_205074_205102                |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "+------------------------------------------------------------------------------------+-------------------------------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "index.withColumn(\"extracted_column\", F.split(F.col(\"studyId\"), \"_ENS\")[0]\n",
    ").select(\"studyId\",\"extracted_column\").show(truncate=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_label + sample_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 15:18:07 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n",
      "[Stage 7:============================================>              (3 + 1) / 4]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|             studyId|         geneId|  projectId|studyType|     traitFromSource|traitFromSourceMappedIds|biosampleFromSourceId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|initialSampleSize|nCases|nControls|nSamples|cohorts|ldPopulationStructure|discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|           condition|sumstatQCValues|diseaseIds|backgroundDiseaseIds|biosampleId|\n",
      "+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|Alasoo_2018_exon_...|ENSG00000015532|Alasoo_2018|   sceqtl|ENSG00000015532.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000106771|Alasoo_2018|   sceqtl|ENSG00000106771.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000135218|Alasoo_2018|   sceqtl|ENSG00000135218.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000135899|Alasoo_2018|   sceqtl|ENSG00000135899.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000138600|Alasoo_2018|   sceqtl|ENSG00000138600.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000141524|Alasoo_2018|   sceqtl|ENSG00000141524.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000168385|Alasoo_2018|   sceqtl|ENSG00000168385.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000168994|Alasoo_2018|   sceqtl|ENSG00000168994.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000171109|Alasoo_2018|   sceqtl|ENSG00000171109.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000067334|Alasoo_2018|   sceqtl|ENSG00000067334.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000092108|Alasoo_2018|   sceqtl|ENSG00000092108.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000123360|Alasoo_2018|   sceqtl|ENSG00000123360.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000128951|Alasoo_2018|   sceqtl|ENSG00000128951.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000132199|Alasoo_2018|   sceqtl|ENSG00000132199.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000135124|Alasoo_2018|   sceqtl|ENSG00000135124.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000145191|Alasoo_2018|   sceqtl|ENSG00000145191.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000157106|Alasoo_2018|   sceqtl|ENSG00000157106.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000157379|Alasoo_2018|   sceqtl|ENSG00000157379.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000158186|Alasoo_2018|   sceqtl|ENSG00000158186.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000177951|Alasoo_2018|   sceqtl|ENSG00000177951.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "index.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PySpark",
   "language": "python",
   "name": "pyspark"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
