{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2024-12-16 13:53:05.366726\n",
      "Analysis started on 2024-12-16 at  2024-12-16 13:53:05.366726\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:53:06 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:53:09 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:25 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:54:25 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading sources\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:27 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moving to step 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:28 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:28 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:28 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "moving to Step 3\n",
      "starting dictionaries at 2024-12-16 13:54:28.587980\n",
      "gene_burden\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:29 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:29 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:29 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:30 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:30 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:30 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intogen\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:32 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:32 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:32 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:34 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:54:34 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:34 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cancer_gene_census\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:35 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:35 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:35 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:37 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:37 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:37 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eva\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:38 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:38 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:38 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:40 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:40 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:40 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eva_somatic\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:41 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:42 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:42 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:42 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:43 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ot_genetics_portal\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:44 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:54:45 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas_credible_set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:54:53 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:54:59 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "impc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:01 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:03 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orphanet\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:04 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:04 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:04 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:06 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:06 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:06 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene2phenotype\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:07 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:07 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:07 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:08 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:08 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:08 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WOcgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:16 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:55:23 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "somatic\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:24 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:24 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:24 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:26 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:26 WARN CacheManager: Asked to cache already cached data.\n",
      "24/12/16 13:55:26 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "germline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 13:55:33 WARN CacheManager: Asked to cache already cached data.        \n",
      "24/12/16 13:55:40 WARN CacheManager: Asked to cache already cached data.        \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\ndef aggregations_original(\\n    df,\\n    data,\\n    listado,\\n    comparisonColumn,\\n    comparisonType,\\n    predictionColumn,\\n    predictionType,\\n    today_date,\\n):\\n\\n    wComparison = Window.partitionBy(comparisonColumn)\\n    wPrediction = Window.partitionBy(predictionColumn)\\n    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\\n\\n    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\\n\\n    out = (\\n        df.withColumn(\"comparisonType\", F.lit(comparisonType))\\n        .withColumn(\"predictionType\", F.lit(predictionType))\\n        .withColumn(\"total\", F.lit(uniqIds))\\n        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\\n        .withColumn(\\n            \"predictionTotal\",\\n            F.count(\"targetId\").over(wPrediction),\\n        )\\n        .withColumn(\\n            \"comparisonTotal\",\\n            F.count(\"targetId\").over(wComparison),\\n        )\\n        .select(\\n            F.col(predictionColumn).alias(\"prediction\"),\\n            F.col(comparisonColumn).alias(\"comparison\"),\\n            \"comparisonType\",\\n            \"predictionType\",\\n            \"a\",\\n            \"predictionTotal\",\\n            \"comparisonTotal\",\\n            \"total\",\\n        )\\n        .filter(F.col(\"prediction\").isNotNull())\\n        .filter(F.col(\"comparison\").isNotNull())\\n        .distinct()\\n    )\\n\\n    out.write.mode(\"overwrite\").parquet(\\n        \"gs://ot-team/jroldan/\"\\n        + str(\\n            today_date\\n            + \"_\"\\n            + \"analysis/\"\\n            + data\\n            # + \"_propagated\"\\n            + \"/\"\\n            + comparisonColumn\\n            + \"_\"\\n            + predictionColumn\\n            + \".parquet\"\\n        )\\n    )\\n\\n    filePath = \"gs://ot-team/jroldan/\" + str(\\n        today_date\\n        + \"_\"\\n        + \"analysis/\"\\n        + data\\n        # + \"_propagated\"\\n        + \"/\"\\n        + comparisonColumn\\n        + \"_\"\\n        + predictionColumn\\n        + \".parquet\"\\n    )\\n    listado.append(\\n        filePath\\n    )\\n    print(\\n        today_date\\n        + \"_\"\\n        + \"analysis/\"\\n        + data\\n        # + \"_propagated\"\\n        + \"/\"\\n        + comparisonColumn\\n        + \"_\"\\n        + predictionColumn\\n        + \".parquet\"\\n    )\\n    c = datetime.now()\\n    c.strftime(\"%H:%M:%S\")\\n    print(c)\\n\\n    array1 = np.delete(\\n        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\\n        .groupBy(\"comparison\")\\n        .pivot(\"prediction\")\\n        .agg(F.first(\"a\"))\\n        .sort(F.col(\"comparison\").desc())\\n        .select(\"comparison\", \"yes\", \"no\")\\n        .fillna(0)\\n        .toPandas()\\n        .to_numpy(),\\n        [0],\\n        1,\\n    )\\n    total = np.sum(array1)\\n    res_npPhaseX = np.array(array1, dtype=int)\\n    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\\n    resx_CI = convertTuple(\\n        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\\n    )\\n\\n    result_st.append(resX)\\n    result_ci.append(resx_CI)\\n    (rs_result, rs_ci) = relative_success(array1)\\n\\n    results.append(\\n        [\\n            comparisonType,\\n            comparisonColumn,\\n            predictionColumn,\\n            round(float(resX.split(\",\")[0]), 2),\\n            float(resX.split(\",\")[1]),\\n            round(float(resx_CI.split(\",\")[0]), 2),\\n            round(float(resx_CI.split(\",\")[1]), 2),\\n            str(total),\\n            np.array(res_npPhaseX).tolist(),\\n            round(float(rs_result), 2),\\n            round(float(rs_ci[0]), 2),\\n            round(float(rs_ci[1]), 2),\\n            filePath,\\n        ]\\n    )\\n    return results\\n\\n\\nc = datetime.now()\\n\\nprint(\"start doing aggregations and writing\")\\ntoday_date = str(date.today())\\naggSetups_original = comparisons_df()\\nlistado = []\\n\\nprint(\"starting with non-propagated aggregations at\", c)\\n\\nfor key, df in dfs_dict.items():\\n    df = df.persist()\\n    for row in aggSetups_original:\\n        aggregations_original(df, key, listado, *row, today_date)\\n    df.unpersist()\\n    print(key + \" df unpersisted\")\\n\\nprint(\"non propagated files wroten succesfully at\", c)\\n\\nprint(\"starting with non-propagated aggregations at\", c)\\nfor key, df in dfs_dict_propag.items():\\n    df = df.persist()\\n    for row in aggSetups_original:\\n        aggregations_original(df, key, listado, *row, today_date)\\n    df.unpersist()\\n    print(key + \" df unpersisted\")\\n\\nprint(\"propagated files wroten succesfully at\", c)\\n\\n##### read files and make spreadsheet\\n\\nprint(\"preparing dataframe\")\\n\\nschema = StructType(\\n    [\\n        StructField(\"group\", StringType(), True),\\n        StructField(\"comparison\", StringType(), True),\\n        StructField(\"phase\", StringType(), True),\\n        StructField(\"oddsRatio\", DoubleType(), True),\\n        StructField(\"pValue\", DoubleType(), True),\\n        StructField(\"lowerInterval\", DoubleType(), True),\\n        StructField(\"upperInterval\", DoubleType(), True),\\n        StructField(\"total\", StringType(), True),\\n        StructField(\"values\", ArrayType(ArrayType(IntegerType())), True),\\n        StructField(\"relSuccess\", DoubleType(), True),\\n        StructField(\"rsLower\", DoubleType(), True),\\n        StructField(\"rsUpper\", DoubleType(), True),\\n        StructField(\"path\", StringType(), True),\\n    ]\\n)\\n\\ndf = spreadSheetFormatter(spark.createDataFrame(results, schema=schema))\\n\\n'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### 11.12.2024\n",
    "#######\n",
    "#######     ATENTION\n",
    "#### change code to work with generated dataframe instead of reading the parquet\n",
    "\n",
    "\"\"\"\n",
    "This scripts run Odds ratio analysis for DoE and \n",
    "genetic information on drug clinical success\n",
    "\n",
    "\"\"\"\n",
    "from functions import discrepancifier\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "c = datetime.now()\n",
    "today_date = str(date.today())\n",
    "print(\"spark session created at\", c)\n",
    "\n",
    "print(\"Analysis started on \" + today_date + \" at \", c)\n",
    "\"\"\"\n",
    "#coloc = spark.read.parquet(\n",
    "#    \"gs://genetics-portal-dev-data/22.09.1/outputs/v2d_coloc\"\n",
    "#).filter(F.col(\"right_type\") != \"gwas\")\n",
    "\"\"\"\n",
    "\n",
    "#### make the dataset from stopped clin trials\n",
    "### read supplementary table 9\n",
    "\"\"\" ### just showing how i did the dataset\n",
    "st9 = spark.read.csv(\"/Users/juanr/Downloads/ST9.csv\", sep=\",\", header=True)\n",
    "st9.filter(\n",
    "    (F.col(\"clinicalStatus\").isin([\"Terminated\", \"Withdrawn\", \"Suspended\"]))\n",
    "    & (F.col(\"prediction\") == \"Negative\")\n",
    ").groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"clinicalStatus\", \"prediction\"\n",
    ").count().toPandas().to_csv(\n",
    "    \"targetDiseaseStoppedNegative.csv\"\n",
    "\"\"\"\n",
    "### target-diseases terminated&withdrawal in clin trials\n",
    "terminated = spark.read.csv(\n",
    "    \"gs://ot-team/jroldan/analysis/targetDiseaseStoppedNegative.csv\",\n",
    "    sep=\",\",\n",
    "    header=True,\n",
    ").drop(\"_c0\", \"Withdrawn\")\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "evidences = (\n",
    "    spark.read.parquet(f\"{path}evidence\")\n",
    "    .filter(\n",
    "        F.col(\"datasourceId\").isin(\n",
    "            [\n",
    "                \"ot_genetics_portal\",\n",
    "                \"gene_burden\",\n",
    "                \"eva\",\n",
    "                \"eva_somatic\",\n",
    "                \"gene2phenotype\",\n",
    "                \"orphanet\",\n",
    "                \"cancer_gene_census\",\n",
    "                \"intogen\",\n",
    "                \"impc\",\n",
    "                \"chembl\",\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "ot_genetics = evidences.filter(F.col(\"datasourceId\") == \"ot_genetics_portal\")\n",
    "\n",
    "#### Now load sources of data to generate credible_set_OT_genetics evidences and associations.\n",
    "\n",
    "target = spark.read.parquet(f\"{path}targets/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path}diseases/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns).persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "### bring directionality from QTL\n",
    "\n",
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")\n",
    "\n",
    "#### take the direction from the lowest p value\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\").orderBy(\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    gwasResolvedColoc.withColumn(\"homogenized\", F.first(\"colocDoE\").over(window_spec))\n",
    "    .select(\"targetId\", \"diseaseId\", \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "# 1# Make a list of variant of interest (Sequence ontology terms) to subset data of interest.\n",
    "\n",
    "### Bear in mind that SO works with ontology structure as: SO:XXXXXX, but databases has the SO as: SO_XXXXXX\n",
    "\n",
    "var_filter_lof = [\n",
    "    ### High impact variants https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html\n",
    "    \"SO_0001589\",  ## frameshit_variant\n",
    "    \"SO_0001587\",  ## stop_gained\n",
    "    \"SO_0001574\",  ## splice_acceptor_variant\n",
    "    \"SO_0001575\",  ## splice_donor_variant\n",
    "    \"SO_0002012\",  ## start_lost\n",
    "    \"SO_0001578\",  ## stop_lost\n",
    "    \"SO_0001893\",  ## transcript_ablation\n",
    "    # \"SO:0001889\", ## transcript_amplification ## the Only HIGH impact that increase protein.\n",
    "]\n",
    "\n",
    "gof = [\"SO_0002053\"]\n",
    "lof = [\"SO_0002054\"]\n",
    "\n",
    "print(\"loading sources\")\n",
    "\n",
    "## Building Sequence Ontology\n",
    "so_path = \"gs://ot-team/jroldan/sequenceOntology_20221118.csv\"\n",
    "so_ontology = spark.read.csv(so_path, header=True)\n",
    "building = (\n",
    "    so_ontology.select(F.col(\"Accession\"), F.col(\"Parents\"))\n",
    "    .withColumn(\"Parentalind\", F.split(F.col(\"Parents\"), \",\"))\n",
    "    .withColumn(\"Parentalind\", F.explode_outer(\"Parentalind\"))\n",
    "    .groupBy(\"Parentalind\")\n",
    "    .agg(F.collect_list(F.col(\"Accession\")).alias(\"childrens\"))\n",
    "    .join(so_ontology, F.col(\"Parentalind\") == so_ontology.Accession, \"right\")\n",
    ")\n",
    "\n",
    "## others\n",
    "target_path = f\"{path}targets/\"\n",
    "target = spark.read.parquet(target_path)\n",
    "disease_path = f\"{path}diseases/\"\n",
    "diseases = spark.read.parquet(disease_path)\n",
    "dis_name = diseases.select(\"id\", \"name\")\n",
    "indication_path = f\"{path}indication/\"\n",
    "indication = spark.read.parquet(indication_path)\n",
    "mecact_path = f\"{path}mechanismOfAction/\"\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "\n",
    "## annotate TSG/oncogene/bivalent using 'hallmarks.attributes'\n",
    "oncotsg_list = [\n",
    "    \"TSG\",\n",
    "    \"oncogene\",\n",
    "    \"Oncogene\",\n",
    "    \"oncogene\",\n",
    "    \"oncogene,TSG\",\n",
    "    \"TSG,oncogene\",\n",
    "    \"fusion,oncogene\",\n",
    "    \"oncogene,fusion\",\n",
    "]\n",
    "\n",
    "#### rlike('('+Keywords+')(\\s|$)'\n",
    "### on 03.07.2023 we add the categories:\n",
    "# DISRUPTING AGENT - inhibitor\n",
    "# STABILISER - activator\n",
    "\n",
    "### Hacer el join del actionType con el chembl para sacar los mecanismos de accion.\n",
    "inhibitors = [\n",
    "    \"RNAI INHIBITOR\",\n",
    "    \"NEGATIVE MODULATOR\",\n",
    "    \"NEGATIVE ALLOSTERIC MODULATOR\",\n",
    "    \"ANTAGONIST\",\n",
    "    \"ANTISENSE INHIBITOR\",\n",
    "    \"BLOCKER\",\n",
    "    \"INHIBITOR\",\n",
    "    \"DEGRADER\",\n",
    "    \"INVERSE AGONIST\",\n",
    "    \"ALLOSTERIC ANTAGONIST\",\n",
    "    \"DISRUPTING AGENT\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "activators = [\n",
    "    \"PARTIAL AGONIST\",\n",
    "    \"ACTIVATOR\",\n",
    "    \"POSITIVE ALLOSTERIC MODULATOR\",\n",
    "    \"POSITIVE MODULATOR\",\n",
    "    \"AGONIST\",\n",
    "    \"SEQUESTERING AGENT\",\n",
    "    \"STABILISER\",  ## added new on 03.07.2023\n",
    "]\n",
    "\n",
    "columnas = [\"activator\", \"inhibitor\"]\n",
    "both = activators + inhibitors\n",
    "\n",
    "actionType = (\n",
    "    mecact.select(\n",
    "        F.explode_outer(\"chemblIds\").alias(\"drugId2\"),\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "        \"targets\",\n",
    "    )\n",
    "    .select(\n",
    "        F.explode_outer(\"targets\").alias(\"targetId2\"),\n",
    "        \"drugId2\",\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "    )\n",
    "    .groupBy(\"targetId2\", \"drugId2\")\n",
    "    .agg(\n",
    "        F.collect_set(\"actionType\").alias(\"actionType\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "oncolabel = (\n",
    "    target.select(\n",
    "        \"id\", \"approvedSymbol\", F.explode_outer(F.col(\"hallmarks.attributes\"))\n",
    "    )\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.description\")\n",
    "    .filter(F.col(\"description\").isin(oncotsg_list))\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"description\").alias(\"description\"))\n",
    "    .withColumn(\"description_splited\", F.concat_ws(\",\", F.col(\"description\")))\n",
    "    .withColumn(\n",
    "        \"TSorOncogene\",\n",
    "        F.when(\n",
    "            (\n",
    "                F.col(\"description_splited\").rlike(\"ncogene\")\n",
    "                & F.col(\"description_splited\").rlike(\"TSG\")\n",
    "            ),\n",
    "            F.lit(\"bivalent\"),\n",
    "        )\n",
    "        .when(F.col(\"description_splited\").rlike(\"ncogene(\\s|$)\"), F.lit(\"oncogene\"))\n",
    "        .when(F.col(\"description_splited\").rlike(\"TSG(\\s|$)\"), F.lit(\"TSG\"))\n",
    "        .otherwise(F.lit(\"noEvaluable\")),\n",
    "    )\n",
    "    .withColumnRenamed(\"id\", \"target_id\")\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "all = evidences.filter(\n",
    "    F.col(\"datasourceId\").isin(\n",
    "        [\n",
    "            \"ot_genetics_portal\",\n",
    "            \"gene_burden\",\n",
    "            \"eva\",\n",
    "            \"eva_somatic\",\n",
    "            \"gene2phenotype\",\n",
    "            \"orphanet\",\n",
    "            \"cancer_gene_census\",\n",
    "            \"intogen\",\n",
    "            \"impc\",\n",
    "            \"chembl\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "\n",
    "#### version all gene burden\n",
    "prueba_assessment = (\n",
    "    all.withColumn(\"beta\", F.col(\"beta\").cast(\"double\"))  ## ot genetics & gene burden\n",
    "    .withColumn(\n",
    "        \"OddsRatio\", F.col(\"OddsRatio\").cast(\"double\")\n",
    "    )  ## ot genetics & gene burden\n",
    "    .withColumn(\n",
    "        \"clinicalSignificances\", F.concat_ws(\",\", F.col(\"clinicalSignificances\"))\n",
    "    )  ### eva\n",
    "    .join(oncolabel, oncolabel.target_id == F.col(\"targetId\"), \"left\")  ###  cgc\n",
    "    .join(\n",
    "        actionType,  ## chembl\n",
    "        (actionType.drugId2 == F.col(\"drugId\"))\n",
    "        & (actionType.targetId2 == F.col(\"targetId\")),\n",
    "        \"left\",\n",
    "    )\n",
    "    .withColumn(\"inhibitors_list\", F.array([F.lit(i) for i in inhibitors]))\n",
    "    .withColumn(\"activators_list\", F.array([F.lit(i) for i in activators]))\n",
    "    .withColumn(\n",
    "        \"intogen_function\",\n",
    "        F.when(\n",
    "            F.arrays_overlap(\n",
    "                F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                F.array([F.lit(i) for i in (gof)]),\n",
    "            ),\n",
    "            F.lit(\"GoF\"),\n",
    "        ).when(\n",
    "            F.arrays_overlap(\n",
    "                F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                F.array([F.lit(i) for i in (lof)]),\n",
    "            ),\n",
    "            F.lit(\"LoF\"),\n",
    "        ),\n",
    "        # .otherwise(\"nodata\"),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"intogenAnnot\",\n",
    "        F.size(F.collect_set(F.col(\"intogen_function\")).over(windowSpec)),\n",
    "    )\n",
    "    ### variant Effect Column\n",
    "    .withColumn(\n",
    "        \"variantEffect\",\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isNotNull(),\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\").isNull(),\n",
    "                    F.when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                        F.lit(\"LoF\"),\n",
    "                    )\n",
    "                    .when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(gof),\n",
    "                        F.lit(\"GoF\"),\n",
    "                    )\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                )\n",
    "                ### variantFunctionalConsequenceFromQtlId\n",
    "                .when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\").isNotNull(),\n",
    "                    F.when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                            var_filter_lof\n",
    "                        ),  ## when is a LoF variant\n",
    "                        F.when(\n",
    "                            F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                            == \"SO_0002316\",\n",
    "                            F.lit(\"LoF\"),\n",
    "                        )\n",
    "                        .when(\n",
    "                            F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                            == \"SO_0002315\",\n",
    "                            F.lit(\"conflict/noEvaluable\"),\n",
    "                        )\n",
    "                        .otherwise(F.lit(\"LoF\")),\n",
    "                    ).when(\n",
    "                        F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof)\n",
    "                        == False,  ## when is not a LoF, still can be a GoF\n",
    "                        F.when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(gof)\n",
    "                            == False,  ##if not GoF\n",
    "                            F.when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002316\",\n",
    "                                F.lit(\"LoF\"),\n",
    "                            )\n",
    "                            .when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002315\",\n",
    "                                F.lit(\"GoF\"),\n",
    "                            )\n",
    "                            .otherwise(F.lit(\"noEvaluable\")),\n",
    "                        ).when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                                gof\n",
    "                            ),  ##if is GoF\n",
    "                            F.when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002316\",\n",
    "                                F.lit(\"conflict/noEvaluable\"),\n",
    "                            ).when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002315\",\n",
    "                                F.lit(\"GoF\"),\n",
    "                            ),\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            ).when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isNull(),\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\") == \"SO_0002316\",\n",
    "                    F.lit(\"LoF\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"variantFunctionalConsequenceFromQtlId\") == \"SO_0002315\",\n",
    "                    F.lit(\"GoF\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"datasourceId\") == \"gene_burden\",\n",
    "            F.when(F.col(\"targetId\").isNotNull(), F.lit(\"LoF\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### son tambien no data las que tiene riesgo pero no se ensayan LoF o PT\n",
    "        )\n",
    "        #### Eva_germline\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva\",\n",
    "            #### .filter(F.col('variantFunctionalConsequenceId').isin(var_filter_lof))\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                F.lit(\"LoF\"),\n",
    "            ).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son LoF\n",
    "        )\n",
    "        #### Eva_somatic\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva_somatic\",\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                F.lit(\"LoF\"),\n",
    "            ).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "        )\n",
    "        #### G2P\n",
    "        .when(\n",
    "            F.col(\"datasourceId\")\n",
    "            == \"gene2phenotype\",  ### 6 types of variants [SO_0002318, SO_0002317, SO_0001622, SO_0002315, SO_0001566, SO_0002220]\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002317\",\n",
    "                F.lit(\"LoF\"),\n",
    "            )  ### absent gene product\n",
    "            .when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002315\",\n",
    "                F.lit(\"GoF\"),\n",
    "            )  ### increased gene product level\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "        #### Orphanet\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"orphanet\",\n",
    "            F.when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002054\",\n",
    "                F.lit(\"LoF\"),\n",
    "            )  ### Loss of Function Variant\n",
    "            .when(\n",
    "                F.col(\"variantFunctionalConsequenceId\") == \"SO_0002053\",\n",
    "                F.lit(\"GoF\"),\n",
    "            )  ### Gain_of_Function Variant\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "        #### CGC\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"cancer_gene_census\",\n",
    "            F.when(F.col(\"TSorOncogene\") == \"oncogene\", F.lit(\"GoF\"))\n",
    "            .when(F.col(\"TSorOncogene\") == \"TSG\", F.lit(\"LoF\"))\n",
    "            .when(F.col(\"TSorOncogene\") == \"bivalent\", F.lit(\"bivalent\"))\n",
    "            .otherwise(\"noEvaluable\"),\n",
    "        )\n",
    "        #### intogen\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"intogen\",\n",
    "            F.when(\n",
    "                F.col(\"intogenAnnot\")\n",
    "                == 1,  ## oncogene/tummor suppressor for a given trait\n",
    "                F.when(\n",
    "                    F.arrays_overlap(\n",
    "                        F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                        F.array([F.lit(i) for i in (gof)]),\n",
    "                    ),\n",
    "                    F.lit(\"GoF\"),\n",
    "                ).when(\n",
    "                    F.arrays_overlap(\n",
    "                        F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                        F.array([F.lit(i) for i in (lof)]),\n",
    "                    ),\n",
    "                    F.lit(\"LoF\"),\n",
    "                ),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"intogenAnnot\") > 1, F.lit(\"bivalentIntogen\")\n",
    "            )  ##oncogene & tumor suppressor for a given trait\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "        #### impc\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"impc\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"LoF\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        ### chembl\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"chembl\",\n",
    "            F.when(\n",
    "                F.size(F.array_intersect(F.col(\"actionType\"), F.col(\"inhibitors_list\")))\n",
    "                >= 1,\n",
    "                F.lit(\"LoF\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.size(F.array_intersect(F.col(\"actionType\"), F.col(\"activators_list\")))\n",
    "                >= 1,\n",
    "                F.lit(\"GoF\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"directionOnTrait\",\n",
    "        ## ot genetics portal\n",
    "        F.when(\n",
    "            F.col(\"datasourceId\") == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(F.col(\"beta\") > 0, F.lit(\"risk\"))\n",
    "                .when(F.col(\"beta\") < 0, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.when(F.col(\"OddsRatio\") > 1, F.lit(\"risk\"))\n",
    "                .when(F.col(\"OddsRatio\") < 1, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.lit(\"noEvaluable\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                F.lit(\"conflict/noEvaluable\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"datasourceId\") == \"gene_burden\",\n",
    "            F.when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                F.when(F.col(\"beta\") > 0, F.lit(\"risk\"))\n",
    "                .when(F.col(\"beta\") < 0, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"oddsRatio\").isNotNull()) & (F.col(\"beta\").isNull()),\n",
    "                F.when(F.col(\"oddsRatio\") > 1, F.lit(\"risk\"))\n",
    "                .when(F.col(\"oddsRatio\") < 1, F.lit(\"protect\"))\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNull()) & (F.col(\"oddsRatio\").isNull()),\n",
    "                F.lit(\"noEvaluable\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"beta\").isNotNull()) & (F.col(\"oddsRatio\").isNotNull()),\n",
    "                F.lit(\"conflict\"),\n",
    "            ),\n",
    "        )\n",
    "        ## Eva_germline\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva\",  ### the same for eva_somatic\n",
    "            F.when(F.col(\"clinicalSignificances\").rlike(\"(pathogenic)$\"), F.lit(\"risk\"))\n",
    "            .when(F.col(\"clinicalSignificances\").contains(\"protect\"), F.lit(\"protect\"))\n",
    "            .otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "        )\n",
    "        #### Eva_somatic\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"eva_somatic\",\n",
    "            F.when(F.col(\"clinicalSignificances\").rlike(\"(pathogenic)$\"), F.lit(\"risk\"))\n",
    "            .when(F.col(\"clinicalSignificances\").contains(\"protect\"), F.lit(\"protect\"))\n",
    "            .otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "        )\n",
    "        #### G2P\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"gene2phenotype\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### Orphanet\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"orphanet\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### CGC\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"cancer_gene_census\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### intogen\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"intogen\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        #### impc\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"impc\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        )\n",
    "        ### chembl\n",
    "        .when(\n",
    "            F.col(\"datasourceId\") == \"chembl\",\n",
    "            F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"protect\")).otherwise(\n",
    "                F.lit(\"noEvaluable\")\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(\n",
    "            (F.col(\"variantEffect\") == \"LoF\") & (F.col(\"directionOnTrait\") == \"risk\"),\n",
    "            F.lit(\"LoF_risk\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"variantEffect\") == \"LoF\")\n",
    "            & (F.col(\"directionOnTrait\") == \"protect\"),\n",
    "            F.lit(\"LoF_protect\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"variantEffect\") == \"GoF\") & (F.col(\"directionOnTrait\") == \"risk\"),\n",
    "            F.lit(\"GoF_risk\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"variantEffect\") == \"GoF\")\n",
    "            & (F.col(\"directionOnTrait\") == \"protect\"),\n",
    "            F.lit(\"GoF_protect\"),\n",
    "        )\n",
    "        .otherwise(F.lit(\"noEvaluable\")),\n",
    "    )\n",
    ").persist()\n",
    "\n",
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    prueba_assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "terminated_array = (\n",
    "    terminated.groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.collect_set(\"clinicalStatus\").alias(\"clinicalStatus\"))\n",
    "    .withColumn(\"prediction\", F.when(F.col(\"clinicalStatus\").isNotNull(), F.lit(\"yes\")))\n",
    ")\n",
    "\n",
    "assessment = prueba_assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "\n",
    "def analysis_nonPropagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "\n",
    "\n",
    "def analysis_drugs(assessment, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\").isin(chembl_ds)))\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment, chembl_ds)\n",
    "\n",
    "#######\n",
    "## include here the analysis\n",
    "#######\n",
    "\n",
    "analysisDatasources = []\n",
    "\n",
    "\n",
    "def full_analysis_propagation(\n",
    "    assessment, analysisDatasources, analysis_chembl, terminated_array, diseaseTA\n",
    "):\n",
    "    return (\n",
    "        analysis_propagated(assessment, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        #### Should remove the coherencyDiagonal.isNotNull()\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "#####\n",
    "## no propag\n",
    "#####\n",
    "def full_analysis_noPropagation(\n",
    "    assessment, analysisDatasources, analysis_chembl, terminated_array, diseaseTA\n",
    "):\n",
    "    return (\n",
    "        analysis_nonPropagated(assessment, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase0\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 0, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(terminated_array, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"prediction\") == \"yes\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "print(\"moving to Step 3\")\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "c = datetime.now()\n",
    "print(\"starting dictionaries at\", c)\n",
    "\n",
    "#### continue here on 10.07.2024\n",
    "\n",
    "## 1nd dictionary\n",
    "dfs_dict = {}  ### checked and changed on 01.06.2023\n",
    "dfs_dict_propag = {}\n",
    "\n",
    "\n",
    "wocgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    # \"ot_genetics_portal\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "]\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"ot_genetics_portal\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"WOcgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "\n",
    "germline_list = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    # \"ot_genetics_portal\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic_list = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "# assessment = prueba_assessment.filter(F.col(\"datasourceId\").isin(datasources_analysis))\n",
    "\n",
    "\n",
    "def dataset_builder(assessment, value, analysis_chembl, terminated_array, diseaseTA):\n",
    "    nonPropagated = full_analysis_noPropagation(\n",
    "        assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "    )\n",
    "    propagated = full_analysis_propagation(\n",
    "        assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "    )\n",
    "    return (\n",
    "        # Non propagation\n",
    "        ## All\n",
    "        nonPropagated,\n",
    "        ## Other\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        nonPropagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "        # Propagation\n",
    "        ## All\n",
    "        propagated,\n",
    "        ## Other\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        propagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "    )\n",
    "\n",
    "\n",
    "for value in datasource_list:\n",
    "    print(value)\n",
    "    if value == \"WOcgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment, wocgc_list, analysis_chembl, terminated_array, diseaseTA\n",
    "        )\n",
    "    elif value == \"germline\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment,\n",
    "            germline_list,\n",
    "            analysis_chembl,\n",
    "            terminated_array,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    elif value == \"somatic\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment,\n",
    "            somatic_list,\n",
    "            analysis_chembl,\n",
    "            terminated_array,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment, value, analysis_chembl, terminated_array, diseaseTA\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df() -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    comparisons = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"hasGeneticEvidence\", \"byDatatype\"),\n",
    "            (\"diagonalYes\", \"byDatatype\"),\n",
    "            (\"oneCellYes\", \"byDatatype\"),\n",
    "        ],\n",
    "        schema=StructType(\n",
    "            [\n",
    "                StructField(\"comparison\", StringType(), True),\n",
    "                StructField(\"comparisonType\", StringType(), True),\n",
    "            ]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "results = []\n",
    "\"\"\"\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "\n",
    "    wComparison = Window.partitionBy(comparisonColumn)\n",
    "    wPrediction = Window.partitionBy(predictionColumn)\n",
    "    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\n",
    "\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    filePath = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    listado.append(\n",
    "        filePath\n",
    "    )\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            comparisonType,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            filePath,\n",
    "        ]\n",
    "    )\n",
    "    return results\n",
    "\n",
    "\n",
    "c = datetime.now()\n",
    "\n",
    "print(\"start doing aggregations and writing\")\n",
    "today_date = str(date.today())\n",
    "aggSetups_original = comparisons_df()\n",
    "listado = []\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "\n",
    "for key, df in dfs_dict.items():\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"non propagated files wroten succesfully at\", c)\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "for key, df in dfs_dict_propag.items():\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"propagated files wroten succesfully at\", c)\n",
    "\n",
    "##### read files and make spreadsheet\n",
    "\n",
    "print(\"preparing dataframe\")\n",
    "\n",
    "schema = StructType(\n",
    "    [\n",
    "        StructField(\"group\", StringType(), True),\n",
    "        StructField(\"comparison\", StringType(), True),\n",
    "        StructField(\"phase\", StringType(), True),\n",
    "        StructField(\"oddsRatio\", DoubleType(), True),\n",
    "        StructField(\"pValue\", DoubleType(), True),\n",
    "        StructField(\"lowerInterval\", DoubleType(), True),\n",
    "        StructField(\"upperInterval\", DoubleType(), True),\n",
    "        StructField(\"total\", StringType(), True),\n",
    "        StructField(\"values\", ArrayType(ArrayType(IntegerType())), True),\n",
    "        StructField(\"relSuccess\", DoubleType(), True),\n",
    "        StructField(\"rsLower\", DoubleType(), True),\n",
    "        StructField(\"rsUpper\", DoubleType(), True),\n",
    "        StructField(\"path\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "\n",
    "df = spreadSheetFormatter(spark.createDataFrame(results, schema=schema))\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 3979:===========================================>        (169 + 8) / 200]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------+-----+-----+\n",
      "|oneCellYes|   no|  yes|\n",
      "+----------+-----+-----+\n",
      "|        no|59560|21813|\n",
      "|       yes|   26|   36|\n",
      "+----------+-----+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "dfs_dict_propag[\"df_gwas_credible_set_All_propag\"].groupBy(\"oneCellYes\").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------+-----+-----+\n",
      "|oneCellYes|   no|  yes|\n",
      "+----------+-----+-----+\n",
      "|        no|59575|21820|\n",
      "|       yes|   11|   29|\n",
      "+----------+-----+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict[\"df_gwas_credible_set_All_original\"].groupBy(\"oneCellYes\").pivot(\n",
    "    \"Phase4\"\n",
    ").count().show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### extract numbers from coloc to compare with the ones for Gwas "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## potential fail: we are propagating gwasResolvedColoc and in the following steps is propagated again (check that)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwasResolvedColoc = (\n",
    "    (\n",
    "        newColoc.filter(F.col(\"rightStudyType\") != \"gwas\")\n",
    "        .withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"right\", ### left becacause is to bring data from \n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    )\n",
    "    .withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check numbers of datasets in germline and gwas credible set VS the ones provided by coloc. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sanity checks qtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'index' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_131695/786458137.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'index' is not defined"
     ]
    }
   ],
   "source": [
    "index.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 15:17:48 WARN YarnScheduler: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources\n",
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "\n",
    "#### 11.12.2024\n",
    "#######\n",
    "#######     ATENTION\n",
    "#### change code to work with generated dataframe instead of reading the parquet\n",
    "\n",
    "\"\"\"\n",
    "This scripts run Odds ratio analysis for DoE and \n",
    "genetic information on drug clinical success\n",
    "\n",
    "\"\"\"\n",
    "from functions import discrepancifier\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    ArrayType,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "path = \"gs://open-targets-pre-data-releases/24.12-uo_test-3/output/etl/parquet/\"\n",
    "\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path}credibleSet\")\n",
    "\n",
    "index = spark.read.parquet(f\"{path}gwasIndex\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path}colocalisation/coloc\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path}variantIndex\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path}biosample\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .persist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 8:===================================================>       (7 + 1) / 8]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------+------+\n",
      "|studyType| count|\n",
      "+---------+------+\n",
      "|     gwas| 90839|\n",
      "|     sqtl|184609|\n",
      "|     pqtl|  3757|\n",
      "|    tuqtl|301140|\n",
      "|     eqtl|948158|\n",
      "|  sctuqtl| 63143|\n",
      "|   sceqtl|335808|\n",
      "|   scsqtl| 29350|\n",
      "+---------+------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "index.groupBy(\"studyType\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------------------------------------------------------------------------------+--------------------------------+--------------------------------+----------+--------------+--------------------------+----------------------+----------------------+----------------------+---------------------+--------------------+--------------------+--------------------+-------------------------+-----------------+---------------------+------------------+----------------------+---------------+----------+--------------+-----------+--------------+\n",
      "|rightStudyId                                                                      |rightStudyLocusId               |leftStudyLocusId                |chromosome|rightStudyType|numberColocalisingVariants|h0                    |h1                    |h2                    |h3                   |h4                  |colocalisationMethod|betaRatioSignAverage|leftStudyId              |leftVariantId    |credibleLeftStudyType|rightVariantId    |credibleRightStudyType|geneId         |projectId |indexStudyType|condition  |biosampleId   |\n",
      "+----------------------------------------------------------------------------------+--------------------------------+--------------------------------+----------+--------------+--------------------------+----------------------+----------------------+----------------------+---------------------+--------------------+--------------------+--------------------+-------------------------+-----------------+---------------------+------------------+----------------------+---------------+----------+--------------+-----------+--------------+\n",
      "|GCST90014010                                                                      |0005218bc3a62e3875f2ba0a0b63c349|0af529199eecb9b6a0779f85fc9d237f|19        |gwas          |1                         |3.1079843775273334E-23|9.978101816041437E-5  |3.11449443689344E-22  |3.107984377838132E-31|0.9999002189818393  |COLOC               |1.0                 |GCST90018951             |19_38320104_C_G  |gwas                 |19_38320104_C_G   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90014010                                                                      |0005218bc3a62e3875f2ba0a0b63c349|e9cf6c4f6215de207532fb58572346be|19        |gwas          |1                         |1.109663241911948E-20 |1.1058085511196052E-4 |1.111998666051756E-19 |1.083547982371386E-4 |0.9997810643466477  |COLOC               |1.0                 |GCST90019500             |19_38320104_C_G  |gwas                 |19_38320104_C_G   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90014010                                                                      |0005218bc3a62e3875f2ba0a0b63c349|bbf92fabf4ac4ccc8fa09075029a8d07|19        |gwas          |1                         |6.303952273206082E-31 |9.978101816041437E-5  |6.317156684346715E-30 |6.303952273836433E-39|0.9999002189818464  |COLOC               |1.0                 |GCST90013999             |19_38320104_C_G  |gwas                 |19_38320104_C_G   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GTEx_tx_minor_salivary_gland_ENST00000650739                                      |0008e0a6a8515514f42c759a5ab0d592|0db850508874cf448fc1b680e101afae|3         |eqtl          |2                         |1.6692899840976926E-59|2.749129180388412E-56 |5.9629969759684814E-6 |0.008829207391733962 |0.9911648296112845  |COLOC               |1.0                 |GCST90264805             |3_53181859_A_C   |gwas                 |3_53173500_C_G    |eqtl                  |ENSG00000163932|GTEx      |eqtl          |naive      |UBERON_0006330|\n",
      "|Aygun_2021_ge_Progenitor_ENSG00000074071                                          |000c0ca56b1f4c71a9e27c0c6cd04eb4|669f43f12522c92e9fd07912b2075590|16        |sceqtl        |1                         |1.8728911667528647E-12|5.0981005772277637E-11|0.03390557889349678   |0.9228830878380463   |0.043211333215604684|COLOC               |-1.0                |FINNGEN_R11_I9_AF_REIMB  |16_1197322_G_A   |gwas                 |16_1772796_C_CACCT|sceqtl                |ENSG00000074071|Aygun_2021|sceqtl        |naive      |CL_0011020    |\n",
      "|Aygun_2021_ge_Progenitor_ENSG00000074071                                          |000c0ca56b1f4c71a9e27c0c6cd04eb4|640a77f297cdc96f42c645609779b92b|16        |sceqtl        |1                         |1.0018935228511351E-12|4.3044595269698714E-11|0.018137615513845854  |0.7790479753646176   |0.20281440907749101 |COLOC               |-1.0                |FINNGEN_R11_I9_HYPTENSESS|16_1928335_A_C   |gwas                 |16_1772796_C_CACCT|sceqtl                |ENSG00000074071|Aygun_2021|sceqtl        |naive      |CL_0011020    |\n",
      "|Braineac2_ge_putamen_ENSG00000147813                                              |0017a37fdd49794eb924301ebe645257|60034b5445cca196ec80c4e92df85a6f|8         |eqtl          |7                         |2.0286256066408496E-27|1.0454912920951048E-19|1.3487858992721654E-10|0.0059571851890201305|0.9940428146761024  |COLOC               |1.0                 |GCST90199815             |8_143594692_C_T  |gwas                 |8_143599607_G_T   |eqtl                  |ENSG00000147813|Braineac2 |eqtl          |naive      |UBERON_0001874|\n",
      "|Braineac2_ge_putamen_ENSG00000147813                                              |0017a37fdd49794eb924301ebe645257|0b51f477621f47ce312001af30028086|8         |eqtl          |5                         |1.699948265442422E-22 |2.0610315664776821E-19|1.1302559932276585E-5 |0.012716046017728619 |0.9872726514223361  |COLOC               |-1.0                |GCST90278652             |8_143602045_A_G  |gwas                 |8_143599607_G_T   |eqtl                  |ENSG00000147813|Braineac2 |eqtl          |naive      |UBERON_0001874|\n",
      "|GTEx_exon_fibroblast_ENSG00000109667.12_4_10040211_10040248                       |0022d49130165f188cf0ceebaa4ed712|78541a0252645ecc26daa450eeb45fd0|4         |sceqtl        |1                         |0.0                   |5.866267082610005E-7  |0.0                   |0.02623851213462055  |0.9737609012386275  |COLOC               |1.0                 |GCST90399704             |4_10079577_G_A   |gwas                 |4_10102479_C_G    |sceqtl                |ENSG00000109667|GTEx      |sceqtl        |naive      |CL_0000057    |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|382ec123cd4bcde15d03be6f3bff1da3|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |-1.0                |GCST90085882             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|17fb120abc3f297bd2d0728b5324e880|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST90018964             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|3de3540041c44075b6cdb608befa68ba|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |-1.0                |GCST90002402             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|9c70b27b12cd9bf810b335f0a8dc4bf1|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST90018966             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|cb2781338d759a3e4e5b06769e565274|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |-1.0                |GCST004607               |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|b07208c748f744d3084a00e5a1fb5402|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST004630               |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|GCST90002322                                                                      |002462a2da2f7c27935710b4d1b08c4b|a37dcb278793694be7011b2ce1f4b734|20        |gwas          |1                         |0.0                   |0.0                   |0.0                   |0.0                  |1.0                 |COLOC               |1.0                 |GCST90002334             |20_31216378_G_A  |gwas                 |20_31216378_G_A   |gwas                  |null           |GCST      |gwas          |null       |null          |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|4ddb2d5f048fbc3922dcd0df622aaeea|17        |sctuqtl       |7                         |2.5407265689237554E-42|8.412145756033199E-35 |1.968303847994525E-9  |0.06423322490039893  |0.935766773131297   |COLOC               |-1.0                |GCST90092806             |17_81674504_C_T  |gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|674a46d781eb51b85ab915d278acf5a5|17        |sctuqtl       |8                         |3.4697676160520917E-41|5.492578537938273E-35 |2.6880330350601647E-8 |0.04159266293826143  |0.9584073101814033  |COLOC               |-1.0                |GCST90269557             |17_81677012_A_G  |gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|ac62ddf72269de27d846904cffe3afb7|17        |sctuqtl       |1                         |1.7018073881445055E-43|2.0155117691098697E-35|1.3183921763172077E-10|0.014628822098044246 |0.985371177770121   |COLOC               |-1.0                |GCST90014001             |17_81667032_CTG_C|gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "|Quach_2016_txrev_monocyte_Pam3CSK4_ENSG00000214087.grp_1.contained.ENST00000573715|002ee135d661029059fe791e40de3731|b8975dcb3b2695d8c53fce3fe4c08a0c|17        |sctuqtl       |10                        |2.981602221346924E-41 |1.6489190383377602E-34|2.3098507321675852E-8 |0.12686881982218193  |0.8731311570793094  |COLOC               |1.0                 |GCST90268498             |17_81688016_C_T  |gwas                 |17_81682485_G_A   |sctuqtl               |ENSG00000214087|Quach_2016|sctuqtl       |Pam3CSK4_6h|CL_0002057    |\n",
      "+----------------------------------------------------------------------------------+--------------------------------+--------------------------------+----------+--------------+--------------------------+----------------------+----------------------+----------------------+---------------------+--------------------+--------------------+--------------------+-------------------------+-----------------+---------------------+------------------+----------------------+---------------+----------+--------------+-----------+--------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "newColoc.show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.types import IntegerType, StringType, StructField, StructType\n",
    "import pandas as pd\n",
    "\n",
    "pd.DataFrame.iteritems = pd.DataFrame.items\n",
    "\n",
    "raw_studies_metadata_schema: StructType = StructType(\n",
    "        [\n",
    "            StructField(\"study_id\", StringType(), True),\n",
    "            StructField(\"dataset_id\", StringType(), True),\n",
    "            StructField(\"study_label\", StringType(), True),\n",
    "            StructField(\"sample_group\", StringType(), True),\n",
    "            StructField(\"tissue_id\", StringType(), True),\n",
    "            StructField(\"tissue_label\", StringType(), True),\n",
    "            StructField(\"condition_label\", StringType(), True),\n",
    "            StructField(\"sample_size\", IntegerType(), True),\n",
    "            StructField(\"quant_method\", StringType(), True),\n",
    "            StructField(\"pmid\", StringType(), True),\n",
    "            StructField(\"study_type\", StringType(), True),\n",
    "        ]\n",
    "    )\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "            pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "            schema=raw_studies_metadata_schema,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------------------------+\n",
      "|extracted_column                                        |\n",
      "+--------------------------------------------------------+\n",
      "|Alasoo_2018_ge_macrophage_naive                         |\n",
      "|Alasoo_2018_exon_macrophage_naive                       |\n",
      "|Alasoo_2018_tx_macrophage_naive                         |\n",
      "|Alasoo_2018_txrev_macrophage_naive                      |\n",
      "|Alasoo_2018_leafcutter_macrophage_naive                 |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h                      |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h                    |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h                      |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h                   |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h              |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_5h                 |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_5h               |\n",
      "|Alasoo_2018_tx_macrophage_Salmonella_5h                 |\n",
      "|Alasoo_2018_txrev_macrophage_Salmonella_5h              |\n",
      "|Alasoo_2018_leafcutter_macrophage_Salmonella_5h         |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h+Salmonella_5h        |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h+Salmonella_5h      |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h+Salmonella_5h        |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h+Salmonella_5h     |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h+Salmonella_5h|\n",
      "+--------------------------------------------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_table.withColumn(\"extracted_column\", \n",
    "    F.concat_ws(\"_\", \n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"tissue_label\"),\n",
    "        F.col(\"condition_label\"))\n",
    "        ).select(\"extracted_column\").show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------------------------------------------------------------------------+\n",
      "|studyId                                                                             |\n",
      "+------------------------------------------------------------------------------------+\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000015532.10_17_50354854_50355056 |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000106771.13_9_109108453_109108488|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135218.19_7_80546191_80546267  |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135899.19_2_230177366_230177537|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000138600.10_15_50748113_50748202 |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000141524.17_17_78119062_78119296 |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168385.18_2_241324663_241325244|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168994.14_6_3739058_3739161    |\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000171109.19_3_179348050_179348121|\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000067334.14_1_93877473_93877489             |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000092108.22_14_30638107_30638124            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000123360.12_12_54561501_54561637            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000128951.14_15_48331011_48331094            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000132199.20_18_691204_691276                |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000135124.16_12_121234071_121234104          |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000145191.15_3_184135389_184135391           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157106.18_16_18860667_18860776            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157379.14_14_24290636_24290995            |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000158186.13_3_138405534_138405534           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000177951.18_11_205074_205102                |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000026508.21_11_35190066_35190429      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000092094.11_14_20449558_20450879      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000130177.16_13_114250549_114250674    |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000198912.12_1_3891058_3891534         |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000214113.11_6_5236284_5236461         |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000224272.2_2_241809110_241809368      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_ENSG00000256377.6_12_27696979_27697298       |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000035115.22_2_218136_218143                |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000108826.16_17_50368314_50368837           |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000138398.17_2_169636413_169636706          |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000147677.12_8_116734277_116734324          |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000148396.19_9_136474487_136475197          |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000159063.14_11_78104454_78104549           |\n",
      "|Alasoo_2018_exon_macrophage_naive_ENSG00000233927.5_19_8321953_8322091              |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella_ENSG00000197013                           |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella_ENSG00000204103                           |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000001561                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000081791                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000100300                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000114446                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000136250                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000144566                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000165934                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000196862                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000225434                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000227354                                      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_ENSG00000234171                                      |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_ENSG00000006282                                |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_ENSG00000130414                                |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_ENSG00000142794                                |\n",
      "+------------------------------------------------------------------------------------+\n",
      "only showing top 50 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "index.select(\"studyId\").show(50,truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 55:===========================================>              (3 + 1) / 4]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------------------------------------+\n",
      "|extracted_column                           |\n",
      "+-------------------------------------------+\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella     |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_exon_macrophage_naive          |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella  |\n",
      "|Alasoo_2018_ge_macrophage_IFNg+Salmonella  |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_IFNg             |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella       |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella       |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella       |\n",
      "+-------------------------------------------+\n",
      "only showing top 50 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "index.withColumn(\"extracted_column\", F.split(F.col(\"studyId\"), \"_ENS\")[0]).select(\"extracted_column\").show(50,truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------+----------+-----------+--------------------+----------+------------+--------------------+-----------+------------+--------+----------+\n",
      "| study_id|dataset_id|study_label|        sample_group| tissue_id|tissue_label|     condition_label|sample_size|quant_method|    pmid|study_type|\n",
      "+---------+----------+-----------+--------------------+----------+------------+--------------------+-----------+------------+--------+----------+\n",
      "|QTS000001| QTD000001|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000002|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000003|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000004|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000005|Alasoo_2018|    macrophage_naive|CL_0000235|  macrophage|               naive|         84|  leafcutter|29379200|      bulk|\n",
      "|QTS000001| QTD000006|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000007|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000008|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000009|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000010|Alasoo_2018|     macrophage_IFNg|CL_0000235|  macrophage|            IFNg_18h|         84|  leafcutter|29379200|      bulk|\n",
      "|QTS000001| QTD000011|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000012|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000013|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000014|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000015|Alasoo_2018|macrophage_Salmon...|CL_0000235|  macrophage|       Salmonella_5h|         84|  leafcutter|29379200|      bulk|\n",
      "|QTS000001| QTD000016|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|          ge|29379200|      bulk|\n",
      "|QTS000001| QTD000017|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|        exon|29379200|      bulk|\n",
      "|QTS000001| QTD000018|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|          tx|29379200|      bulk|\n",
      "|QTS000001| QTD000019|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|       txrev|29379200|      bulk|\n",
      "|QTS000001| QTD000020|Alasoo_2018|macrophage_IFNg+S...|CL_0000235|  macrophage|IFNg_18h+Salmonel...|         84|  leafcutter|29379200|      bulk|\n",
      "+---------+----------+-----------+--------------------+----------+------------+--------------------+-----------+------------+--------+----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_table.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------------------------+----------+\n",
      "|extracted_column                                        |study_type|\n",
      "+--------------------------------------------------------+----------+\n",
      "|Alasoo_2018_ge_macrophage_naive                         |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_naive                       |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_naive                         |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_naive                      |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_naive                 |bulk      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h                      |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h                    |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h                      |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h                   |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h              |bulk      |\n",
      "|Alasoo_2018_ge_macrophage_Salmonella_5h                 |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_Salmonella_5h               |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_Salmonella_5h                 |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_Salmonella_5h              |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_Salmonella_5h         |bulk      |\n",
      "|Alasoo_2018_ge_macrophage_IFNg_18h+Salmonella_5h        |bulk      |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_18h+Salmonella_5h      |bulk      |\n",
      "|Alasoo_2018_tx_macrophage_IFNg_18h+Salmonella_5h        |bulk      |\n",
      "|Alasoo_2018_txrev_macrophage_IFNg_18h+Salmonella_5h     |bulk      |\n",
      "|Alasoo_2018_leafcutter_macrophage_IFNg_18h+Salmonella_5h|bulk      |\n",
      "|BLUEPRINT_ge_monocyte_naive                             |bulk      |\n",
      "|BLUEPRINT_exon_monocyte_naive                           |bulk      |\n",
      "|BLUEPRINT_tx_monocyte_naive                             |bulk      |\n",
      "|BLUEPRINT_txrev_monocyte_naive                          |bulk      |\n",
      "|BLUEPRINT_leafcutter_monocyte_naive                     |bulk      |\n",
      "|BLUEPRINT_ge_neutrophil_naive                           |bulk      |\n",
      "|BLUEPRINT_exon_neutrophil_naive                         |bulk      |\n",
      "|BLUEPRINT_tx_neutrophil_naive                           |bulk      |\n",
      "|BLUEPRINT_txrev_neutrophil_naive                        |bulk      |\n",
      "|BLUEPRINT_leafcutter_neutrophil_naive                   |bulk      |\n",
      "|BLUEPRINT_ge_CD4+ T cell_naive                          |bulk      |\n",
      "|BLUEPRINT_exon_CD4+ T cell_naive                        |bulk      |\n",
      "|BLUEPRINT_tx_CD4+ T cell_naive                          |bulk      |\n",
      "|BLUEPRINT_txrev_CD4+ T cell_naive                       |bulk      |\n",
      "|BLUEPRINT_leafcutter_CD4+ T cell_naive                  |bulk      |\n",
      "|Bossini-Castillo_2019_ge_Treg memory_naive              |bulk      |\n",
      "|Bossini-Castillo_2019_exon_Treg memory_naive            |bulk      |\n",
      "|Bossini-Castillo_2019_tx_Treg memory_naive              |bulk      |\n",
      "|Bossini-Castillo_2019_txrev_Treg memory_naive           |bulk      |\n",
      "|Bossini-Castillo_2019_leafcutter_Treg memory_naive      |bulk      |\n",
      "|Braineac2_ge_brain (putamen)_naive                      |bulk      |\n",
      "|Braineac2_exon_brain (putamen)_naive                    |bulk      |\n",
      "|Braineac2_tx_brain (putamen)_naive                      |bulk      |\n",
      "|Braineac2_txrev_brain (putamen)_naive                   |bulk      |\n",
      "|Braineac2_leafcutter_brain (putamen)_naive              |bulk      |\n",
      "|Braineac2_ge_brain (substantia nigra)_naive             |bulk      |\n",
      "|Braineac2_exon_brain (substantia nigra)_naive           |bulk      |\n",
      "|Braineac2_tx_brain (substantia nigra)_naive             |bulk      |\n",
      "|Braineac2_txrev_brain (substantia nigra)_naive          |bulk      |\n",
      "|Braineac2_leafcutter_brain (substantia nigra)_naive     |bulk      |\n",
      "|BrainSeq_ge_brain (DLPFC)_naive                         |bulk      |\n",
      "|BrainSeq_exon_brain (DLPFC)_naive                       |bulk      |\n",
      "|BrainSeq_tx_brain (DLPFC)_naive                         |bulk      |\n",
      "|BrainSeq_txrev_brain (DLPFC)_naive                      |bulk      |\n",
      "|BrainSeq_leafcutter_brain (DLPFC)_naive                 |bulk      |\n",
      "|CAP_ge_LCL_naive                                        |bulk      |\n",
      "|CAP_exon_LCL_naive                                      |bulk      |\n",
      "|CAP_tx_LCL_naive                                        |bulk      |\n",
      "|CAP_txrev_LCL_naive                                     |bulk      |\n",
      "|CAP_leafcutter_LCL_naive                                |bulk      |\n",
      "|CAP_ge_LCL_statin                                       |bulk      |\n",
      "|CAP_exon_LCL_statin                                     |bulk      |\n",
      "|CAP_tx_LCL_statin                                       |bulk      |\n",
      "|CAP_txrev_LCL_statin                                    |bulk      |\n",
      "|CAP_leafcutter_LCL_statin                               |bulk      |\n",
      "|CEDAR_microarray_CD8+ T cell_naive                      |bulk      |\n",
      "|CEDAR_microarray_CD4+ T cell_naive                      |bulk      |\n",
      "|CEDAR_microarray_transverse colon_naive                 |bulk      |\n",
      "|CEDAR_microarray_monocyte_naive                         |bulk      |\n",
      "|CEDAR_microarray_neutrophil_naive                       |bulk      |\n",
      "|CEDAR_microarray_platelet_naive                         |bulk      |\n",
      "|CEDAR_microarray_rectum_naive                           |bulk      |\n",
      "|CEDAR_microarray_B cell_naive                           |bulk      |\n",
      "|CEDAR_microarray_ileum_naive                            |bulk      |\n",
      "|CommonMind_ge_brain (DLPFC)_naive                       |bulk      |\n",
      "|CommonMind_exon_brain (DLPFC)_naive                     |bulk      |\n",
      "|CommonMind_tx_brain (DLPFC)_naive                       |bulk      |\n",
      "|CommonMind_txrev_brain (DLPFC)_naive                    |bulk      |\n",
      "|CommonMind_leafcutter_brain (DLPFC)_naive               |bulk      |\n",
      "|Fairfax_2012_microarray_B cell_naive                    |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_naive                  |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_IFNg_24h               |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_LPS_2h                 |bulk      |\n",
      "|Fairfax_2014_microarray_monocyte_LPS_24h                |bulk      |\n",
      "|FUSION_ge_muscle_naive                                  |bulk      |\n",
      "|FUSION_exon_muscle_naive                                |bulk      |\n",
      "|FUSION_tx_muscle_naive                                  |bulk      |\n",
      "|FUSION_txrev_muscle_naive                               |bulk      |\n",
      "|FUSION_leafcutter_muscle_naive                          |bulk      |\n",
      "|FUSION_ge_adipose_naive                                 |bulk      |\n",
      "|FUSION_exon_adipose_naive                               |bulk      |\n",
      "|FUSION_tx_adipose_naive                                 |bulk      |\n",
      "|FUSION_txrev_adipose_naive                              |bulk      |\n",
      "|FUSION_leafcutter_adipose_naive                         |bulk      |\n",
      "|GENCORD_ge_LCL_naive                                    |bulk      |\n",
      "|GENCORD_exon_LCL_naive                                  |bulk      |\n",
      "|GENCORD_tx_LCL_naive                                    |bulk      |\n",
      "|GENCORD_txrev_LCL_naive                                 |bulk      |\n",
      "|GENCORD_leafcutter_LCL_naive                            |bulk      |\n",
      "|GENCORD_ge_fibroblast_naive                             |bulk      |\n",
      "|GENCORD_exon_fibroblast_naive                           |bulk      |\n",
      "|GENCORD_tx_fibroblast_naive                             |bulk      |\n",
      "|GENCORD_txrev_fibroblast_naive                          |bulk      |\n",
      "|GENCORD_leafcutter_fibroblast_naive                     |bulk      |\n",
      "|GENCORD_ge_T cell_naive                                 |bulk      |\n",
      "|GENCORD_exon_T cell_naive                               |bulk      |\n",
      "|GENCORD_tx_T cell_naive                                 |bulk      |\n",
      "|GENCORD_txrev_T cell_naive                              |bulk      |\n",
      "|GENCORD_leafcutter_T cell_naive                         |bulk      |\n",
      "|GEUVADIS_ge_LCL_naive                                   |bulk      |\n",
      "|GEUVADIS_exon_LCL_naive                                 |bulk      |\n",
      "|GEUVADIS_tx_LCL_naive                                   |bulk      |\n",
      "|GEUVADIS_txrev_LCL_naive                                |bulk      |\n",
      "|GEUVADIS_leafcutter_LCL_naive                           |bulk      |\n",
      "|Gilchrist_2021_microarray_NK cell_naive                 |bulk      |\n",
      "|GTEx_ge_adipose_naive                                   |bulk      |\n",
      "|GTEx_exon_adipose_naive                                 |bulk      |\n",
      "|GTEx_tx_adipose_naive                                   |bulk      |\n",
      "|GTEx_txrev_adipose_naive                                |bulk      |\n",
      "|GTEx_leafcutter_adipose_naive                           |bulk      |\n",
      "|GTEx_ge_adipose (visceral)_naive                        |bulk      |\n",
      "|GTEx_exon_adipose (visceral)_naive                      |bulk      |\n",
      "|GTEx_tx_adipose (visceral)_naive                        |bulk      |\n",
      "|GTEx_txrev_adipose (visceral)_naive                     |bulk      |\n",
      "|GTEx_leafcutter_adipose (visceral)_naive                |bulk      |\n",
      "|GTEx_ge_adrenal gland_naive                             |bulk      |\n",
      "|GTEx_exon_adrenal gland_naive                           |bulk      |\n",
      "|GTEx_tx_adrenal gland_naive                             |bulk      |\n",
      "|GTEx_txrev_adrenal gland_naive                          |bulk      |\n",
      "|GTEx_leafcutter_adrenal gland_naive                     |bulk      |\n",
      "|GTEx_ge_artery (aorta)_naive                            |bulk      |\n",
      "|GTEx_exon_artery (aorta)_naive                          |bulk      |\n",
      "|GTEx_tx_artery (aorta)_naive                            |bulk      |\n",
      "|GTEx_txrev_artery (aorta)_naive                         |bulk      |\n",
      "|GTEx_leafcutter_artery (aorta)_naive                    |bulk      |\n",
      "|GTEx_ge_artery (coronary)_naive                         |bulk      |\n",
      "|GTEx_exon_artery (coronary)_naive                       |bulk      |\n",
      "|GTEx_tx_artery (coronary)_naive                         |bulk      |\n",
      "|GTEx_txrev_artery (coronary)_naive                      |bulk      |\n",
      "|GTEx_leafcutter_artery (coronary)_naive                 |bulk      |\n",
      "|GTEx_ge_artery (tibial)_naive                           |bulk      |\n",
      "|GTEx_exon_artery (tibial)_naive                         |bulk      |\n",
      "|GTEx_tx_artery (tibial)_naive                           |bulk      |\n",
      "|GTEx_txrev_artery (tibial)_naive                        |bulk      |\n",
      "|GTEx_leafcutter_artery (tibial)_naive                   |bulk      |\n",
      "|GTEx_ge_brain (amygdala)_naive                          |bulk      |\n",
      "|GTEx_exon_brain (amygdala)_naive                        |bulk      |\n",
      "|GTEx_tx_brain (amygdala)_naive                          |bulk      |\n",
      "|GTEx_txrev_brain (amygdala)_naive                       |bulk      |\n",
      "|GTEx_leafcutter_brain (amygdala)_naive                  |bulk      |\n",
      "|GTEx_ge_brain (anterior cingulate cortex)_naive         |bulk      |\n",
      "|GTEx_exon_brain (anterior cingulate cortex)_naive       |bulk      |\n",
      "|GTEx_tx_brain (anterior cingulate cortex)_naive         |bulk      |\n",
      "|GTEx_txrev_brain (anterior cingulate cortex)_naive      |bulk      |\n",
      "|GTEx_leafcutter_brain (anterior cingulate cortex)_naive |bulk      |\n",
      "|GTEx_ge_brain (caudate)_naive                           |bulk      |\n",
      "|GTEx_exon_brain (caudate)_naive                         |bulk      |\n",
      "|GTEx_tx_brain (caudate)_naive                           |bulk      |\n",
      "|GTEx_txrev_brain (caudate)_naive                        |bulk      |\n",
      "|GTEx_leafcutter_brain (caudate)_naive                   |bulk      |\n",
      "|GTEx_ge_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_exon_brain (cerebellum)_naive                      |bulk      |\n",
      "|GTEx_tx_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_txrev_brain (cerebellum)_naive                     |bulk      |\n",
      "|GTEx_leafcutter_brain (cerebellum)_naive                |bulk      |\n",
      "|GTEx_ge_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_exon_brain (cerebellum)_naive                      |bulk      |\n",
      "|GTEx_tx_brain (cerebellum)_naive                        |bulk      |\n",
      "|GTEx_txrev_brain (cerebellum)_naive                     |bulk      |\n",
      "|GTEx_leafcutter_brain (cerebellum)_naive                |bulk      |\n",
      "|GTEx_ge_brain (cortex)_naive                            |bulk      |\n",
      "|GTEx_exon_brain (cortex)_naive                          |bulk      |\n",
      "|GTEx_tx_brain (cortex)_naive                            |bulk      |\n",
      "|GTEx_txrev_brain (cortex)_naive                         |bulk      |\n",
      "|GTEx_leafcutter_brain (cortex)_naive                    |bulk      |\n",
      "|GTEx_ge_brain (DLPFC)_naive                             |bulk      |\n",
      "|GTEx_exon_brain (DLPFC)_naive                           |bulk      |\n",
      "|GTEx_tx_brain (DLPFC)_naive                             |bulk      |\n",
      "|GTEx_txrev_brain (DLPFC)_naive                          |bulk      |\n",
      "|GTEx_leafcutter_brain (DLPFC)_naive                     |bulk      |\n",
      "|GTEx_ge_brain (hippocampus)_naive                       |bulk      |\n",
      "|GTEx_exon_brain (hippocampus)_naive                     |bulk      |\n",
      "|GTEx_tx_brain (hippocampus)_naive                       |bulk      |\n",
      "|GTEx_txrev_brain (hippocampus)_naive                    |bulk      |\n",
      "|GTEx_leafcutter_brain (hippocampus)_naive               |bulk      |\n",
      "|GTEx_ge_brain (hypothalamus)_naive                      |bulk      |\n",
      "|GTEx_exon_brain (hypothalamus)_naive                    |bulk      |\n",
      "|GTEx_tx_brain (hypothalamus)_naive                      |bulk      |\n",
      "|GTEx_txrev_brain (hypothalamus)_naive                   |bulk      |\n",
      "|GTEx_leafcutter_brain (hypothalamus)_naive              |bulk      |\n",
      "|GTEx_ge_brain (nucleus accumbens)_naive                 |bulk      |\n",
      "|GTEx_exon_brain (nucleus accumbens)_naive               |bulk      |\n",
      "|GTEx_tx_brain (nucleus accumbens)_naive                 |bulk      |\n",
      "|GTEx_txrev_brain (nucleus accumbens)_naive              |bulk      |\n",
      "|GTEx_leafcutter_brain (nucleus accumbens)_naive         |bulk      |\n",
      "|GTEx_ge_brain (putamen)_naive                           |bulk      |\n",
      "|GTEx_exon_brain (putamen)_naive                         |bulk      |\n",
      "|GTEx_tx_brain (putamen)_naive                           |bulk      |\n",
      "|GTEx_txrev_brain (putamen)_naive                        |bulk      |\n",
      "|GTEx_leafcutter_brain (putamen)_naive                   |bulk      |\n",
      "+--------------------------------------------------------+----------+\n",
      "only showing top 200 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_table.select(\n",
    "    F.concat_ws(\n",
    "        \"_\",\n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"sample_group\"),\n",
    "    ).alias(\"extracted_column\"),\n",
    "    \"study_type\",\n",
    ").show(200,truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "saved=study_table.select(\n",
    "    F.concat_ws(\n",
    "        \"_\",\n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"sample_group\"),\n",
    "    ).alias(\"extracted_column\"),\n",
    "    \"study_type\",\n",
    ").join(\n",
    "    index.withColumn(\"extracted_column\", F.split(F.col(\"studyId\"), \"_ENS\")[0]),\n",
    "    on=\"extracted_column\",\n",
    "    how=\"right\",\n",
    ").persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|    extracted_column|study_type|             studyId|         geneId|  projectId|studyType|     traitFromSource|traitFromSourceMappedIds|biosampleFromSourceId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|initialSampleSize|nCases|nControls|nSamples|cohorts|ldPopulationStructure|discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|           condition|sumstatQCValues|diseaseIds|backgroundDiseaseIds|biosampleId|\n",
      "+--------------------+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000140319|Alasoo_2018|   scsqtl|15:40037018:40038...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000065320|Alasoo_2018|   scsqtl|17:9182969:922116...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000169136|Alasoo_2018|   scsqtl|19:49929170:49929...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000167543|Alasoo_2018|   scsqtl|17:29568508:29569...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000196735|Alasoo_2018|   scsqtl|6:32642784:326429...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000179630|Alasoo_2018|   scsqtl|13:43879743:43880...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|       Salmonella_5h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000120910|Alasoo_2018|   scsqtl|8:22475624:224980...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|       Salmonella_5h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000184110|Alasoo_2018|   scsqtl|16:28715434:28723...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000102858|Alasoo_2018|   scsqtl|16:4686303:468879...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000167863|Alasoo_2018|   scsqtl|17:75042277:75046...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000173875|Alasoo_2018|   scsqtl|19:12611082:12623...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000116473|Alasoo_2018|   scsqtl|1:111699607:11170...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000119782|Alasoo_2018|   scsqtl|2:24049886:240539...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000135605|Alasoo_2018|   scsqtl|4:48141419:481454...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000289047|Alasoo_2018|   scsqtl|6:32894927:329009...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_leafc...|      null|Alasoo_2018_leafc...|ENSG00000127952|Alasoo_2018|   scsqtl|7:76001002:760037...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000235|\n",
      "|Aygun_2021_leafcu...|      null|Aygun_2021_leafcu...|ENSG00000184009| Aygun_2021|   scsqtl|17:81512360:81512...|                    null|           CL_0011020|34416157|            null|                  null|           null|              null|                              null|             null|  null|     null|      85|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0011020|\n",
      "|Aygun_2021_leafcu...|      null|Aygun_2021_leafcu...|ENSG00000165119| Aygun_2021|   scsqtl|9:83978452:839801...|                    null|           CL_0011020|34416157|            null|                  null|           null|              null|                              null|             null|  null|     null|      85|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0011020|\n",
      "|BLUEPRINT_leafcut...|      null|BLUEPRINT_leafcut...|ENSG00000237489|  BLUEPRINT|   scsqtl|10:130079901:1300...|                    null|           CL_0000624|27863251|            null|                  null|           null|              null|                              null|             null|  null|     null|     167|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000624|\n",
      "|BLUEPRINT_leafcut...|      null|BLUEPRINT_leafcut...|ENSG00000119943|  BLUEPRINT|   scsqtl|10:98387307:98388...|                    null|           CL_0000624|27863251|            null|                  null|           null|              null|                              null|             null|  null|     null|     167|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|               naive|           null|        []|                  []| CL_0000624|\n",
      "+--------------------+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "saved.filter(F.col(\"study_type\").isNull()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "index.withColumn(\"extracted_column\", \n",
    "            F.split(F.col(\"studyId\"), \"_ENS\")[0]).select(\"extra\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 31:==================================================>       (7 + 1) / 8]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+--------+----------+-----------+------------+---------+------------+---------------+-----------+------------+----+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|    extracted_column|study_id|dataset_id|study_label|sample_group|tissue_id|tissue_label|condition_label|sample_size|quant_method|pmid|study_type|             studyId|         geneId|  projectId|studyType|     traitFromSource|traitFromSourceMappedIds|biosampleFromSourceId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|initialSampleSize|nCases|nControls|nSamples|cohorts|ldPopulationStructure|discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|           condition|sumstatQCValues|diseaseIds|backgroundDiseaseIds|biosampleId|\n",
      "+--------------------+--------+----------+-----------+------------+---------+------------+---------------+-----------+------------+----+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000026508|Alasoo_2018|   sceqtl|ENSG00000026508.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|       Salmonella_5h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000015532|Alasoo_2018|   sceqtl|ENSG00000015532.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000106771|Alasoo_2018|   sceqtl|ENSG00000106771.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000135218|Alasoo_2018|   sceqtl|ENSG00000135218.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000135899|Alasoo_2018|   sceqtl|ENSG00000135899.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000138600|Alasoo_2018|   sceqtl|ENSG00000138600.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000141524|Alasoo_2018|   sceqtl|ENSG00000141524.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000168385|Alasoo_2018|   sceqtl|ENSG00000168385.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000168994|Alasoo_2018|   sceqtl|ENSG00000168994.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000171109|Alasoo_2018|   sceqtl|ENSG00000171109.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000067334|Alasoo_2018|   sceqtl|ENSG00000067334.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000092108|Alasoo_2018|   sceqtl|ENSG00000092108.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000123360|Alasoo_2018|   sceqtl|ENSG00000123360.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000128951|Alasoo_2018|   sceqtl|ENSG00000128951.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000132199|Alasoo_2018|   sceqtl|ENSG00000132199.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000135124|Alasoo_2018|   sceqtl|ENSG00000135124.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000145191|Alasoo_2018|   sceqtl|ENSG00000145191.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000157106|Alasoo_2018|   sceqtl|ENSG00000157106.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000157379|Alasoo_2018|   sceqtl|ENSG00000157379.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|    null|      null|       null|        null|     null|        null|           null|       null|        null|null|      null|Alasoo_2018_exon_...|ENSG00000158186|Alasoo_2018|   sceqtl|ENSG00000158186.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "+--------------------+--------+----------+-----------+------------+---------+------------+---------------+-----------+------------+----+----------+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "study_table.withColumn(\"extracted_column\", \n",
    "    F.concat_ws(\"_\", \n",
    "        F.col(\"study_label\"),\n",
    "        F.col(\"quant_method\"),\n",
    "        F.col(\"tissue_label\"),\n",
    "        F.col(\"condition_label\"))\n",
    "        ).join(index.withColumn(\"extracted_column\", \n",
    "            F.split(F.col(\"studyId\"), \"_ENS\")[0]\n",
    "), on=\"extracted_column\", how=\"right\").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_136690/275740773.py\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m study_table = spark.createDataFrame(\n\u001b[0;32m----> 5\u001b[0;31m             \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mraw_studies_metadata_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m             \u001b[0mschema\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraw_studies_metadata_schema\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "#### read metadata table\n",
    "\n",
    "raw_studies_metadata_path = \"https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/fe3c4b4ed911b3a184271a6aadcd8c8769a66aba/data_tables/dataset_metadata.tsv\"\n",
    "\n",
    "study_table = spark.createDataFrame(\n",
    "            pd.read_csv(raw_studies_metadata_path, sep=\"\\t\"),\n",
    "            schema=raw_studies_metadata_schema,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------------------------------------------------------------------------+-------------------------------------------+\n",
      "|studyId                                                                             |extracted_column                           |\n",
      "+------------------------------------------------------------------------------------+-------------------------------------------+\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000015532.10_17_50354854_50355056 |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000106771.13_9_109108453_109108488|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135218.19_7_80546191_80546267  |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000135899.19_2_230177366_230177537|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000138600.10_15_50748113_50748202 |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000141524.17_17_78119062_78119296 |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168385.18_2_241324663_241325244|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000168994.14_6_3739058_3739161    |Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg+Salmonella_ENSG00000171109.19_3_179348050_179348121|Alasoo_2018_exon_macrophage_IFNg+Salmonella|\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000067334.14_1_93877473_93877489             |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000092108.22_14_30638107_30638124            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000123360.12_12_54561501_54561637            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000128951.14_15_48331011_48331094            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000132199.20_18_691204_691276                |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000135124.16_12_121234071_121234104          |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000145191.15_3_184135389_184135391           |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157106.18_16_18860667_18860776            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000157379.14_14_24290636_24290995            |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000158186.13_3_138405534_138405534           |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "|Alasoo_2018_exon_macrophage_IFNg_ENSG00000177951.18_11_205074_205102                |Alasoo_2018_exon_macrophage_IFNg           |\n",
      "+------------------------------------------------------------------------------------+-------------------------------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "index.withColumn(\"extracted_column\", F.split(F.col(\"studyId\"), \"_ENS\")[0]\n",
    ").select(\"studyId\",\"extracted_column\").show(truncate=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_label + sample_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/12/16 15:18:07 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n",
      "[Stage 7:============================================>              (3 + 1) / 4]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|             studyId|         geneId|  projectId|studyType|     traitFromSource|traitFromSourceMappedIds|biosampleFromSourceId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|initialSampleSize|nCases|nControls|nSamples|cohorts|ldPopulationStructure|discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|           condition|sumstatQCValues|diseaseIds|backgroundDiseaseIds|biosampleId|\n",
      "+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "|Alasoo_2018_exon_...|ENSG00000015532|Alasoo_2018|   sceqtl|ENSG00000015532.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000106771|Alasoo_2018|   sceqtl|ENSG00000106771.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000135218|Alasoo_2018|   sceqtl|ENSG00000135218.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000135899|Alasoo_2018|   sceqtl|ENSG00000135899.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000138600|Alasoo_2018|   sceqtl|ENSG00000138600.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000141524|Alasoo_2018|   sceqtl|ENSG00000141524.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000168385|Alasoo_2018|   sceqtl|ENSG00000168385.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000168994|Alasoo_2018|   sceqtl|ENSG00000168994.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000171109|Alasoo_2018|   sceqtl|ENSG00000171109.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|IFNg_18h+Salmonel...|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000067334|Alasoo_2018|   sceqtl|ENSG00000067334.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000092108|Alasoo_2018|   sceqtl|ENSG00000092108.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000123360|Alasoo_2018|   sceqtl|ENSG00000123360.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000128951|Alasoo_2018|   sceqtl|ENSG00000128951.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000132199|Alasoo_2018|   sceqtl|ENSG00000132199.2...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000135124|Alasoo_2018|   sceqtl|ENSG00000135124.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000145191|Alasoo_2018|   sceqtl|ENSG00000145191.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000157106|Alasoo_2018|   sceqtl|ENSG00000157106.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000157379|Alasoo_2018|   sceqtl|ENSG00000157379.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000158186|Alasoo_2018|   sceqtl|ENSG00000158186.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "|Alasoo_2018_exon_...|ENSG00000177951|Alasoo_2018|   sceqtl|ENSG00000177951.1...|                    null|           CL_0000235|29379200|            null|                  null|           null|              null|                              null|             null|  null|     null|      84|   null|                 null|            null|              null|             []|           []|https://ftp.ebi.a...|       true|            IFNg_18h|           null|        []|                  []| CL_0000235|\n",
      "+--------------------+---------------+-----------+---------+--------------------+------------------------+---------------------+--------+----------------+----------------------+---------------+------------------+----------------------------------+-----------------+------+---------+--------+-------+---------------------+----------------+------------------+---------------+-------------+--------------------+-----------+--------------------+---------------+----------+--------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "index.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### make the contradictions and improve them to run the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded files\n",
      "loaded newColoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded gwasComplete\n",
      "loaded resolvedColloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:45 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:46 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run temporary direction of effect\n",
      "Moving to step 2\n",
      "defining non propagated,propagated and analysis_drugs functions\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:47 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:47 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:47 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "defining full_analysis_propagation\n",
      "defining full analysis no propagation\n",
      "moving to Step 3\n",
      "starting dictionaries at 2025-04-24 00:54:47.098964\n",
      "gene_burden\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:47 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:47 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:48 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:48 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intogen\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:49 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:49 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:50 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:50 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cancer_gene_census\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:51 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:51 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:52 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:52 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eva\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:53 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:53 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:54 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:54 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eva_somatic\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:54:55 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:55 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:55 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:56 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:56 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:54:56 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas_credible_set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:55:14 WARN CacheManager: Asked to cache already cached data.        ]\n",
      "25/04/24 00:55:30 WARN CacheManager: Asked to cache already cached data.        ]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "impc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:55:31 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:31 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:31 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:31 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:32 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:32 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:32 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:32 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orphanet\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:55:33 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:33 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:34 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:34 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene2phenotype\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:55:35 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:35 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:36 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:55:36 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WOcgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wCgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "somatic\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 00:56:43 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:56:43 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:56:44 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 00:56:44 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "germline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    }
   ],
   "source": [
    "import time\n",
    "#from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "# from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "# from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "#from itertools import islice\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    StringType,\n",
    "    IntegerType,\n",
    "    ArrayType\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "spark.conf.set(\n",
    "    \"spark.sql.shuffle.partitions\", \"400\"\n",
    ")  # Default is 200, increase if needed\n",
    "\n",
    "\n",
    "path_n='gs://open-targets-data-releases/25.03/output/'\n",
    "\n",
    "target = spark.read.parquet(f\"{path_n}target/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path_n}disease/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path_n}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path_n}credible_set\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path_n}colocalisation_coloc\") \n",
    "\n",
    "index=spark.read.parquet(f\"{path_n}study/\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path_n}variant\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path_n}biosample\")\n",
    "\n",
    "print(\"loaded files\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "            'isTransQtl'\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    ")\n",
    "    # .persist()\n",
    ")\n",
    "\n",
    "print(\"loaded newColoc\")\n",
    "\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns)  # .persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")  # .persist()\n",
    "\n",
    "print(\"loaded gwasComplete\")\n",
    "\n",
    "resolvedColoc = (\n",
    "    (\n",
    "        newColoc.withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"inner\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    ).withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    # .persist()\n",
    ")\n",
    "print(\"loaded resolvedColloc\")\n",
    "\n",
    "datasource_filter = [\n",
    "    \"gwas_credible_set\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]\n",
    "\n",
    "assessment, evidences, actionType, oncolabel = temporary_directionOfEffect(\n",
    "    path_n, datasource_filter\n",
    ")\n",
    "\n",
    "print(\"run temporary direction of effect\")\n",
    "\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\",'leftStudyId').orderBy( ### include gwas study\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    resolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )  ## added 30.01.2025\n",
    "    .select(\"targetId\", \"diseaseId\",'leftStudyId', \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "negativeTD = (\n",
    "    evidences.filter(F.col(\"datasourceId\") == \"chembl\")\n",
    "    .select(\"targetId\", \"diseaseId\", \"studyStopReason\", \"studyStopReasonCategories\")\n",
    "    .filter(F.array_contains(F.col(\"studyStopReasonCategories\"), \"Negative\"))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .count()\n",
    "    .withColumn(\"stopReason\", F.lit(\"Negative\"))\n",
    "    .drop(\"count\")\n",
    ")\n",
    "\n",
    "assessment_all = assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "print(\"defining non propagated,propagated and analysis_drugs functions\")\n",
    "\n",
    "def analysis_nonPropagated(assessment_all, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment_all, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "\n",
    "def analysis_drugs(assessment_all, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter((F.col(\"datasourceId\").isin(chembl_ds))\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment_all, chembl_ds)\n",
    "\n",
    "#######\n",
    "## include here the analysis\n",
    "#######\n",
    "\n",
    "analysisDatasources = []\n",
    "\n",
    "print(\"defining full_analysis_propagation\")\n",
    "\n",
    "doe_columns=[\"LoF_protect\", \"GoF_risk\", \"LoF_risk\", \"GoF_protect\"]\n",
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "\n",
    "def full_analysis_propagation(\n",
    "    doe_columns,assessment_all, analysisDatasources, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "):\n",
    "    conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in doe_columns\n",
    "    ]\n",
    "    \n",
    "    return (\n",
    "        analysis_propagated(assessment_all, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        #### Should remove the coherencyDiagonal.isNotNull()\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        ).withColumn(\n",
    "            \"arrayN\", F.array(*[F.col(c) for c in doe_columns])\n",
    "        ).withColumn(\n",
    "            \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    "        ).withColumn(\"maxDoE_names\", F.array(*conditions)\n",
    "        ).withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\")\n",
    "        ).withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(negativeTD, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"stopReason\") == \"Negative\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        )\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "#####\n",
    "## no propag\n",
    "#####\n",
    "print(\"defining full analysis no propagation\")\n",
    "\n",
    "\n",
    "def full_analysis_noPropagation(\n",
    "    doe_columns,assessment_all, analysisDatasources, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "):\n",
    "    return (\n",
    "        analysis_nonPropagated(assessment_all, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(negativeTD, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"stopReason\") == \"Negative\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        ).withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        )\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "print(\"moving to Step 3\")\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "c = datetime.now()\n",
    "print(\"starting dictionaries at\", c)\n",
    "\n",
    "#### continue here on 10.07.2024\n",
    "\n",
    "## 1nd dictionary\n",
    "dfs_dict = {}  ### checked and changed on 01.06.2023\n",
    "dfs_dict_propag = {}\n",
    "\n",
    "\n",
    "wocgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "]\n",
    "wCgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"cancer_gene_census\",\n",
    "]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"WOcgc\",\n",
    "    \"wCgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "\n",
    "germline_list = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic_list = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "\n",
    "# assessment = prueba_assessment.filter(F.col(\"datasourceId\").isin(datasources_analysis))\n",
    "def dataset_builder(assessment_all, value, analysis_chembl, negativeTD, diseaseTA):\n",
    "    nonPropagated = full_analysis_noPropagation(\n",
    "        doe_columns,assessment_all, value, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "    )\n",
    "    propagated = full_analysis_propagation(\n",
    "        doe_columns,assessment_all, value, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "    )\n",
    "    return (\n",
    "        # Non propagation\n",
    "        ## All\n",
    "        nonPropagated,\n",
    "        ## Other\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        nonPropagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "        # Propagation\n",
    "        ## All\n",
    "        propagated,\n",
    "        ## Other\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        propagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "    )\n",
    "\n",
    "\n",
    "for value in datasource_list:\n",
    "    print(value)\n",
    "    if value == \"WOcgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, wocgc_list, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "    elif value == \"wCgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, wCgc_list, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "    elif value == \"germline\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all,\n",
    "            germline_list,\n",
    "            analysis_chembl,\n",
    "            negativeTD,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    elif value == \"somatic\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all,\n",
    "            somatic_list,\n",
    "            analysis_chembl,\n",
    "            negativeTD,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"]\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, value, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df() -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    comparisons = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"hasGeneticEvidence\", \"byDatatype\"),\n",
    "            (\"diagonalYes\", \"byDatatype\"),\n",
    "            (\"oneCellYes\", \"byDatatype\"),\n",
    "            (\"NdiagonalYes\", \"byDatatype\"),\n",
    "            (\"NoneCellYes\", \"byDatatype\"),\n",
    "        ],\n",
    "        schema=StructType(\n",
    "            [\n",
    "                StructField(\"comparison\", StringType(), True),\n",
    "                StructField(\"comparisonType\", StringType(), True),\n",
    "            ]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+------------------+-------------------+-----+\n",
      "|diagonalYes|N_oneCellAgreeDrug|N_diagonalAgreeDrug|count|\n",
      "+-----------+------------------+-------------------+-----+\n",
      "|         no|                no|                 no|   42|\n",
      "|         no|                no|                yes|   28|\n",
      "|         no|               yes|                yes|   35|\n",
      "+-----------+------------------+-------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "\n",
    "dfs_dict_propag[f\"df_gwas_credible_set_All_propag\"].filter(F.col('coherencyDiagonal')=='dispar').withColumn(\n",
    "    \"N_oneCellAgreeDrug\",\n",
    "    F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "    .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "          ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    ").withColumn(\n",
    "    \"N_diagonalAgreeDrug\",\n",
    "    F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "        (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "        F.lit(\"yes\")\n",
    "    ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "        (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "        F.lit(\"yes\")\n",
    "    ).otherwise(F.lit('no'))\n",
    ").groupBy('diagonalYes','N_oneCellAgreeDrug','N_diagonalAgreeDrug').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+-------------------+----------+------------------+-----+\n",
      "|diagonalYes|N_diagonalAgreeDrug|oneCellYes|N_oneCellAgreeDrug|count|\n",
      "+-----------+-------------------+----------+------------------+-----+\n",
      "|        yes|                yes|        no|                no|  111|\n",
      "|         no|                 no|        no|                no|73934|\n",
      "|        yes|                yes|        no|               yes|    5|\n",
      "|         no|                yes|        no|                no|   37|\n",
      "|        yes|                yes|       yes|               yes|   72|\n",
      "|         no|                yes|        no|               yes|   38|\n",
      "+-----------+-------------------+----------+------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "dfs_dict_propag[f\"df_gwas_credible_set_All_propag\"].withColumn(\n",
    "    \"maxDoEArrayN\",\n",
    "    F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    ").withColumn(\n",
    "    \"N_oneCellAgreeDrug\",\n",
    "    F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "    .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "          ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    ").withColumn(\n",
    "    \"N_diagonalAgreeDrug\",\n",
    "    F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "        (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "        F.lit(\"yes\")\n",
    "    ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "        (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "        F.lit(\"yes\")\n",
    "    ).otherwise(F.lit('no'))\n",
    ").groupBy('diagonalYes','N_diagonalAgreeDrug','oneCellYes','N_oneCellAgreeDrug').count().show()\n",
    "\n",
    "#).filter(F.col('maxDoEArrayN')==2).count()\n",
    "\n",
    "#.show(200,truncate=False)\n",
    "#).groupBy('maxDoEArrayN','N_diagonalAgreeDrug').count().show()\n",
    "#).filter((F.col('diagonalYes')=='no') & (F.col('N_diagonalAgreeDrug')=='yes')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Solo necesito meter las columnas y correr "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "F.size(F.array_intersect(F.col(\"your_array_column\"), F.array([F.lit(x) for x in my_list]))) > 0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+-----------+-----+\n",
      "|diagonalYes_number|diagonalYes|count|\n",
      "+------------------+-----------+-----+\n",
      "|                no|         no|74009|\n",
      "|                no|        yes|  188|\n",
      "+------------------+-----------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict_propag[f\"df_gwas_credible_set_All_propag\"].groupBy('diagonalYes_number','diagonalYes').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_dict_propag[f\"df_gwas_credible_set_All_propag\"]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- diseaseId: string (nullable = true)\n",
      " |-- targetId: string (nullable = true)\n",
      " |-- GoF_protect: long (nullable = true)\n",
      " |-- GoF_risk: long (nullable = true)\n",
      " |-- LoF_protect: long (nullable = true)\n",
      " |-- LoF_risk: long (nullable = true)\n",
      " |-- noEvaluable: void (nullable = true)\n",
      " |-- coherencyDiagonal: string (nullable = true)\n",
      " |-- coherencyOneCell: string (nullable = true)\n",
      " |-- maxClinPhase: double (nullable = true)\n",
      " |-- coherencyDiagonal_ch: string (nullable = true)\n",
      " |-- coherencyOneCell_ch: string (nullable = true)\n",
      " |-- LoF_protect_ch: long (nullable = true)\n",
      " |-- GoF_protect_ch: long (nullable = true)\n",
      " |-- geneticEvidence: string (nullable = false)\n",
      " |-- diagonalAgreeWithDrugs: string (nullable = true)\n",
      " |-- oneCellAgreeWithDrugs: string (nullable = true)\n",
      " |-- Phase4: string (nullable = false)\n",
      " |-- Phase>=3: string (nullable = false)\n",
      " |-- Phase>=2: string (nullable = false)\n",
      " |-- Phase>=1: string (nullable = false)\n",
      " |-- stopReason: string (nullable = true)\n",
      " |-- PhaseT: string (nullable = false)\n",
      " |-- taLabelSimple: string (nullable = true)\n",
      " |-- hasGeneticEvidence: string (nullable = false)\n",
      " |-- diagonalYes: string (nullable = false)\n",
      " |-- oneCellYes: string (nullable = false)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict[f\"df_gwas_credible_set_All_original\"].printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+----------------+------------------+-----+\n",
      "|coherencyDiagonal|coherencyOneCell|hasGeneticEvidence|count|\n",
      "+-----------------+----------------+------------------+-----+\n",
      "|         coherent|        coherent|               yes|  274|\n",
      "|         coherent|          dispar|               yes|   15|\n",
      "|           dispar|          dispar|               yes|   53|\n",
      "+-----------------+----------------+------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict[f\"df_gwas_credible_set_All_original\"].filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").groupBy(\"coherencyDiagonal\", \"coherencyOneCell\", \"hasGeneticEvidence\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------+-----------------+-----+\n",
      "|maxClinPhase|coherencyDiagonal|count|\n",
      "+------------+-----------------+-----+\n",
      "|         0.5|         coherent|    1|\n",
      "|         1.0|           dispar|    1|\n",
      "|         1.0|         coherent|   25|\n",
      "|         2.0|           dispar|   21|\n",
      "|         2.0|         coherent|   96|\n",
      "|         3.0|         coherent|   75|\n",
      "|         3.0|           dispar|   11|\n",
      "|         4.0|           dispar|   20|\n",
      "|         4.0|         coherent|   92|\n",
      "+------------+-----------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict[f\"df_gwas_credible_set_All_original\"].filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()).groupBy('maxClinPhase','coherencyDiagonal').count().sort(F.col('maxClinPhase')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------+-----------------+-----+\n",
      "|Phase4|coherencyDiagonal|count|\n",
      "+------+-----------------+-----+\n",
      "|    no|             NULL|67713|\n",
      "|    no|         coherent|  197|\n",
      "|    no|           dispar|   33|\n",
      "|   yes|             NULL| 6142|\n",
      "|   yes|           dispar|   20|\n",
      "|   yes|         coherent|   92|\n",
      "+------+-----------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict[f\"df_gwas_credible_set_All_original\"].groupBy('Phase4','coherencyDiagonal').count().sort(F.col('Phase4')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_dict[f\"df_gwas_credible_set_All_original\"].filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()).groupBy('maxClinPhase','coherencyDiagonal').count().sort(F.col('maxClinPhase')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+----------------+------------------+-----+\n",
      "|coherencyDiagonal|coherencyOneCell|hasGeneticEvidence|count|\n",
      "+-----------------+----------------+------------------+-----+\n",
      "|         coherent|        coherent|               yes|  303|\n",
      "|         coherent|          dispar|               yes|   15|\n",
      "|           dispar|          dispar|               yes|  105|\n",
      "+-----------------+----------------+------------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dfs_dict_propag[f\"df_gwas_credible_set_All_propag\"].filter(\n",
    "    F.col(\"coherencyDiagonal\").isNotNull()\n",
    ").groupBy(\"coherencyDiagonal\", \"coherencyOneCell\", \"hasGeneticEvidence\").count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_dict[f\"df_gwas_credible_set_All_original\"].withColumn(\n",
    "        \"numbers_coherencyDiagonal\",\n",
    "        F.when(\n",
    "            (F.col(\"LoF_risk\").isNull())\n",
    "            & (F.col(\"LoF_protect\").isNull())\n",
    "            & (F.col(\"GoF_risk\").isNull())\n",
    "            & (F.col(\"GoF_protect\").isNull())\n",
    "            & (F.col(\"noEvaluable\").isNull()),\n",
    "            F.lit(\"noEvid\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"LoF_risk\").isNull())\n",
    "            & (F.col(\"LoF_protect\").isNull())\n",
    "            & (F.col(\"GoF_risk\").isNull())\n",
    "            & (F.col(\"GoF_protect\").isNull())\n",
    "            & (F.col(\"noEvaluable\").isNotNull()),\n",
    "            F.lit(\"EvidNotDoE\"),\n",
    "        )\n",
    "        .when(\n",
    "            ((F.col(\"LoF_risk\")) | (F.col(\"GoF_protect\").isNotNull())) > \n",
    "            (F.col(\"LoF_protect\").isNotNull()) (F.col(\"GoF_risk\").isNotNull())\n",
    "            ,\n",
    "            F.when(\n",
    "                ((F.col(\"GoF_risk\").isNotNull()) & (F.col(\"LoF_risk\").isNotNull())),\n",
    "                F.lit(\"dispar\"),\n",
    "            )\n",
    "            .when(\n",
    "                ((F.col(\"LoF_protect\").isNotNull()) & (F.col(\"LoF_risk\").isNotNull())),\n",
    "                F.lit(\"dispar\"),\n",
    "            )\n",
    "            .when(\n",
    "                ((F.col(\"GoF_protect\").isNotNull()) & (F.col(\"GoF_risk\").isNotNull())),\n",
    "                F.lit(\"dispar\"),\n",
    "            )\n",
    "            .when(\n",
    "                (\n",
    "                    (F.col(\"GoF_protect\").isNotNull())\n",
    "                    & (F.col(\"LoF_protect\").isNotNull())\n",
    "                ),\n",
    "                F.lit(\"dispar\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"coherent\")),\n",
    "        ),\n",
    "    ).withColumn(\n",
    "        \"coherencyOneCell\",\n",
    "        F.when(\n",
    "            (F.col(\"LoF_risk\").isNull())\n",
    "            & (F.col(\"LoF_protect\").isNull())\n",
    "            & (F.col(\"GoF_risk\").isNull())\n",
    "            & (F.col(\"GoF_protect\").isNull())\n",
    "            & (F.col(\"noEvaluable\").isNull()),\n",
    "            F.lit(\"noEvid\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"LoF_risk\").isNull())\n",
    "            & (F.col(\"LoF_protect\").isNull())\n",
    "            & (F.col(\"GoF_risk\").isNull())\n",
    "            & (F.col(\"GoF_protect\").isNull())\n",
    "            & (F.col(\"noEvaluable\").isNotNull()),\n",
    "            F.lit(\"EvidNotDoE\"),\n",
    "        )\n",
    "        .when(\n",
    "            (F.col(\"LoF_risk\").isNotNull())\n",
    "            | (F.col(\"LoF_protect\").isNotNull())\n",
    "            | (F.col(\"GoF_risk\").isNotNull())\n",
    "            | (F.col(\"GoF_protect\").isNotNull()),\n",
    "            F.when(\n",
    "                F.col(\"LoF_risk\").isNotNull()\n",
    "                & (\n",
    "                    (F.col(\"LoF_protect\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                    & (F.col(\"GoF_protect\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"GoF_risk\").isNotNull()\n",
    "                & (\n",
    "                    (F.col(\"LoF_protect\").isNull())\n",
    "                    & (F.col(\"LoF_risk\").isNull())\n",
    "                    & (F.col(\"GoF_protect\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"LoF_protect\").isNotNull()\n",
    "                & (\n",
    "                    (F.col(\"LoF_risk\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                    & (F.col(\"GoF_protect\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .when(\n",
    "                F.col(\"GoF_protect\").isNotNull()\n",
    "                & (\n",
    "                    (F.col(\"LoF_protect\").isNull())\n",
    "                    & (F.col(\"GoF_risk\").isNull())\n",
    "                    & (F.col(\"LoF_risk\").isNull())\n",
    "                ),\n",
    "                F.lit(\"coherent\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"dispar\")),\n",
    "        ),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+---------------+-------------+-------------------+--------+----------+----+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+----------+--------------------+--------+-------------+---------------------+--------------+-----------------+--------+----------------+---------------+----------+--------+-------------------+----------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-----------------+--------------------+----------------------------+-------------------+--------------------+---------+--------------------------------+--------------------------------+--------------+--------------+--------+---------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+-------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----+----------------------------+-------------------+------------------------------+-------------+---------+-----------+------+---------------+------------+-------------------+-------------+--------------------+-------------------+-------------+----------------+------------------+-------------------------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+-----------+\n",
      "|      datasourceId|       targetId|alleleOrigins|allelicRequirements|ancestry|ancestryId|beta|betaConfidenceIntervalLower|betaConfidenceIntervalUpper|biologicalModelAllelicComposition|biologicalModelGeneticBackground|biologicalModelId|biomarkerName|biomarkers|biosamplesFromSource|cellType|clinicalPhase|clinicalSignificances|clinicalStatus|cohortDescription|cohortId|cohortPhenotypes|cohortShortName|confidence|contrast|crisprScreenLibrary|      datatypeId|diseaseCellLines|   diseaseFromSource|diseaseFromSourceId|diseaseFromSourceMappedId|diseaseModelAssociatedHumanPhenotypes|diseaseModelAssociatedModelPhenotypes|drugFromSource|drugId|drugResponse|geneticBackground|          literature|log2FoldChangePercentileRank|log2FoldChangeValue|      mutatedSamples|OddsRatio|oddsRatioConfidenceIntervalLower|oddsRatioConfidenceIntervalUpper|pValueExponent|pValueMantissa|pathways|projectId|reactionId|reactionName|releaseDate|releaseVersion|resourceScore| sex|significantDriverMethods|statisticalMethod|statisticalMethodOverview|statisticalTestTail|studyCases|studyCasesWithQualifyingVariants|studyId|studyOverview|studySampleSize|studyStartDate|studyStopReason|studyStopReasonCategories|targetFromSource|targetFromSourceId|targetInModel|targetInModelEnsemblId|targetInModelMgiId|targetModulation|urls|variantAminoacidDescriptions|variantFromSourceId|variantFunctionalConsequenceId|variantHgvsId|variantId|variantRsId|pmcIds|publicationYear|studyLocusId|textMiningSentences|    diseaseId|                  id|              score|variantEffect|directionOnTrait|          sourceId|variantFunctionalConsequenceFromQtlId|target_id|approvedSymbol|description|description_splited|TSorOncogene|targetId2|drugId2|actionType|     inhibitors_list|     activators_list|intogen_function|intogenAnnot|homogenized|leftStudyId|\n",
      "+------------------+---------------+-------------+-------------------+--------+----------+----+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+----------+--------------------+--------+-------------+---------------------+--------------+-----------------+--------+----------------+---------------+----------+--------+-------------------+----------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-----------------+--------------------+----------------------------+-------------------+--------------------+---------+--------------------------------+--------------------------------+--------------+--------------+--------+---------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+-------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----+----------------------------+-------------------+------------------------------+-------------+---------+-----------+------+---------------+------------+-------------------+-------------+--------------------+-------------------+-------------+----------------+------------------+-------------------------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+-----------+\n",
      "|              impc|ENSG00000000938|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Fgr<Mhdaali18>/Fg...|            C3HeB/FeJ-Fgr<Mhd...|      MGI:3688148|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Angioosteohypotro...|        ORPHA:75508|            MONDO_0019156|                 [{HP:0002753, Thi...|                 [{MP:0002111, abn...|          NULL|  NULL|        NULL|             NULL|          [16964445]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        46.65|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000938|          Fgr|    ENSMUSG00000028874|         MGI:95527|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0019156|16870dc14984b399d...|0.46649999999999997|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000001626|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cftr<tm2Mrc>/Cftr...|            B6.129S6-Cftr<tm2...|      MGI:3689380|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Diamond-Blackfan ...|        OMIM:612527|             Orphanet_124|                 [{HP:0001875, Neu...|                 [{MP:0014028, dec...|          NULL|  NULL|        NULL|             NULL|[18802965, 105701...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        42.51|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL| Orphanet_124|4858d33c68e59d7d0...|             0.4251|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000001626|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cftr<tm2Mrc>/Cftr...|            B6.129S6-Cftr<tm2...|      MGI:3689380|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Diamond-Blackfan ...|        OMIM:618310|             Orphanet_124|                 [{HP:0001875, Neu...|                 [{MP:0014028, dec...|          NULL|  NULL|        NULL|             NULL|[18802965, 105701...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        42.28|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000001626|         Cftr|    ENSMUSG00000041301|         MGI:88388|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL| Orphanet_124|cf0648f0223c44a15...|             0.4228|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000001630|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cyp51<tm1.1Bfro>/...|            involves: 129P2/O...|      MGI:5014092|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Isolated Tibial H...|        ORPHA:93322|            MONDO_0010144|                 [{HP:0009556, Abs...|                 [{MP:0000431, abs...|          NULL|  NULL|        NULL|             NULL|          [21705796]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        47.95|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000001630|        Cyp51|    ENSMUSG00000001467|        MGI:106040|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0010144|039f580fea3238814...|0.47950000000000004|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000002834|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Lasp1<Tg(Col1a1-l...|               involves: C57BL/6|      MGI:4366487|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Isolated Klippel-...|         ORPHA:2345|              EFO_0700054|                 [{HP:0002650, Sco...|                 [{MP:0004606, abs...|          NULL|  NULL|        NULL|             NULL|[12438746, 19378260]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        45.55|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000002834|        Lasp1|    ENSMUSG00000038366|        MGI:109656|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0700054|d15addcf8dee78bbc...|0.45549999999999996|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000002834|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Lasp1<Tg(Col1a1-l...|               involves: C57BL/6|      MGI:4366487|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Spondyloepimetaph...|        OMIM:602557|            MONDO_0011252|                 [{HP:0000926, Pla...|                 [{MP:0004606, abs...|          NULL|  NULL|        NULL|             NULL|[12438746, 19378260]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        48.17|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000002834|        Lasp1|    ENSMUSG00000038366|        MGI:109656|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0011252|182490edd19fd7664...|             0.4817|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000002834|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Lasp1<Tg(Col1a1-l...|               involves: C57BL/6|      MGI:4366487|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Mucopolysaccharid...|          ORPHA:582|            MONDO_0018938|                 [{HP:0000926, Pla...|                 [{MP:0004606, abs...|          NULL|  NULL|        NULL|             NULL|[12438746, 19378260]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        44.89|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000002834|        Lasp1|    ENSMUSG00000038366|        MGI:109656|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0018938|02161a73d40ebc3d1...|             0.4489|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|cancer_gene_census|ENSG00000002834|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|somatic_mutation|            NULL|                NULL|               NULL|            MONDO_0019472|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|[{SO_0001059, 3.0...|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|          0.5|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|    673|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000002834|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0019472|6da76243ec10fd146...|                0.5|  noEvaluable|            risk|cancer_gene_census|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|noEvaluable|       NULL|\n",
      "|              impc|ENSG00000004399|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Plxnd1<b2b3150Clo...|            C57BL/6J-Plxnd1<b...|      MGI:5648021|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|   Scimitar Syndrome|          ORPHA:185|              EFO_1001167|                 [{HP:0010772, Ano...|                 [{MP:0001823, thy...|          NULL|  NULL|        NULL|             NULL|[18992737, 152399...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        42.24|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004399|       Plxnd1|    ENSMUSG00000030123|       MGI:2154244|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_1001167|77f2e49b5523c7b8a...|             0.4224|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004399|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Plxnd1<em1Zho>/Pl...|            C57BL/6-Plxnd1<em...|      MGI:7544908|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|   Scimitar Syndrome|          ORPHA:185|              EFO_1001167|                 [{HP:0010772, Ano...|                 [{MP:0002724, enh...|          NULL|  NULL|        NULL|             NULL|[18992737, 152399...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        59.52|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004399|       Plxnd1|    ENSMUSG00000030123|       MGI:2154244|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_1001167|02c73980a006ba179...| 0.5952000000000001|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004399|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Plxnd1<tm1Joe>/Pl...|            involves: 129S1/S...|      MGI:3051132|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|   Scimitar Syndrome|          ORPHA:185|              EFO_1001167|                 [{HP:0010772, Ano...|                 [{MP:0006355, abn...|          NULL|  NULL|        NULL|             NULL|[15550623, 189927...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        45.34|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004399|       Plxnd1|    ENSMUSG00000030123|       MGI:2154244|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_1001167|12658e09a3df0a356...|             0.4534|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004399|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Plxnd1<b2b1863Clo...|            C57BL/6J-Plxnd1<b...|      MGI:5476432|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|   Scimitar Syndrome|          ORPHA:185|              EFO_1001167|                 [{HP:0010772, Ano...|                 [{MP:0001823, thy...|          NULL|  NULL|        NULL|             NULL|[18992737, 15239958]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        44.21|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004399|       Plxnd1|    ENSMUSG00000030123|       MGI:2154244|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_1001167|502f18650bcb08321...|             0.4421|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004399|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Plxnd1<b2b553Clo>...|            C57BL/6J-Plxnd1<b...|      MGI:5316794|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|   Scimitar Syndrome|          ORPHA:185|              EFO_1001167|                 [{HP:0010772, Ano...|                 [{MP:0010413, com...|          NULL|  NULL|        NULL|             NULL|[18992737, 152399...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        49.57|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004399|       Plxnd1|    ENSMUSG00000030123|       MGI:2154244|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_1001167|db713167f1a06f624...|0.49570000000000003|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004399|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Plxnd1<em1Zho>/Pl...|            C57BL/6-Plxnd1<em...|      MGI:7544908|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Congenital Total ...|        ORPHA:99125|              EFO_1001167|                 [{HP:0011721, Inf...|                 [{MP:0002724, enh...|          NULL|  NULL|        NULL|             NULL|[18992737, 152399...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        43.81|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004399|       Plxnd1|    ENSMUSG00000030123|       MGI:2154244|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_1001167|7e0a53510a2c98bac...|0.43810000000000004|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004468|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cd38<em1(IMPC)Ccp...|                     C57BL/6NCrl|      MGI:6385272|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Acute Panmyelosis...|        ORPHA:86843|            MONDO_0019455|                 [{HP:0001744, Spl...|                 [{MP:0000689, abn...|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        46.25|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004468|         Cd38|    ENSMUSG00000029084|        MGI:107474|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0019455|b156a4dce0fddee37...|             0.4625|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004478|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Fkbp4<tm1Dvds>/Fk...|            either: (involves...|      MGI:3818988|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Spinocerebellar A...|        OMIM:613909|            MONDO_0013486|                 [{HP:0000789, Inf...|                 [{MP:0001925, mal...|          NULL|  NULL|        NULL|             NULL|[15831525, 173079...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         54.2|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004478|        Fkbp4|    ENSMUSG00000030357|         MGI:95543|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0013486|b7dd664c896d3988e...|              0.542|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004478|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Fkbp4<tm1Dvds>/Fk...|            involves: 129X1/S...|      MGI:3583221|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Spinocerebellar A...|       ORPHA:276183|            MONDO_0013486|                 [{HP:0000029, Tes...|                 [{MP:0002212, abn...|          NULL|  NULL|        NULL|             NULL|[16873445, 158315...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        51.89|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004478|        Fkbp4|    ENSMUSG00000030357|         MGI:95543|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0013486|1f35a4163b929d5a9...|             0.5189|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004478|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Fkbp4<tm1Dvds>/Fk...|            either: (involves...|      MGI:3818988|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Spinocerebellar A...|       ORPHA:276183|            MONDO_0013486|                 [{HP:0003251, Mal...|                 [{MP:0001925, mal...|          NULL|  NULL|        NULL|             NULL|[15831525, 173079...|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        56.15|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004478|        Fkbp4|    ENSMUSG00000030357|         MGI:95543|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0013486|188046a8b9de525b8...|             0.5615|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004487|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Kdm1a<tm1Ebg>/Kdm...|            B6.129-Kdm1a<tm1Ebg>|      MGI:5883628|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Developmental And...|        OMIM:300088|            MONDO_0010246|                 [{HP:0025190, Bil...|                 [{MP:0009767, dec...|          NULL|  NULL|        NULL|             NULL|          [24735673]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         66.3|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004487|        Kdm1a|    ENSMUSG00000036940|       MGI:1196256|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0010246|076934417d9b661f3...| 0.6629999999999999|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "|              impc|ENSG00000004487|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Kdm1a<tm1Ebg>/Kdm...|            B6.129-Kdm1a<tm1Ebg>|      MGI:5883628|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|            NULL|           NULL|      NULL|    NULL|               NULL|    animal_model|            NULL|Female Restricted...|       ORPHA:101039|            MONDO_0010246|                 [{HP:0007359, Foc...|                 [{MP:0009767, dec...|          NULL|  NULL|        NULL|             NULL|          [24735673]|                        NULL|               NULL|                NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        56.24|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|   NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000004487|        Kdm1a|    ENSMUSG00000036940|       MGI:1196256|            NULL|NULL|                        NULL|               NULL|                          NULL|         NULL|     NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0010246|e6b449a2ddbfa8fd4...|             0.5624|          LoF|            risk|              impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|       NULL|\n",
      "+------------------+---------------+-------------+-------------------+--------+----------+----+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+----------+--------------------+--------+-------------+---------------------+--------------+-----------------+--------+----------------+---------------+----------+--------+-------------------+----------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-----------------+--------------------+----------------------------+-------------------+--------------------+---------+--------------------------------+--------------------------------+--------------+--------------+--------+---------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+-------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----+----------------------------+-------------------+------------------------------+-------------+---------+-----------+------+---------------+------------+-------------------+-------------+--------------------+-------------------+-------------+----------------+------------------+-------------------------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "assessment_all.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysisDatasources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1259:===================================================>(465 + 2) / 467]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+-------+\n",
      "|homogenized|  count|\n",
      "+-----------+-------+\n",
      "|noEvaluable|2891236|\n",
      "|   GoF_risk|3993490|\n",
      "|LoF_protect|3535651|\n",
      "|   LoF_risk|5214499|\n",
      "|GoF_protect|3956367|\n",
      "+-----------+-------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "assessment_all.filter(F.col(\"datasourceId\").isin(wCgc_list)).groupBy('homogenized').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = Window.partitionBy(\"targetId\", \"diseaseId\",\"homogenized\")\n",
    "\n",
    "df_with_count = assessment_all.filter(F.col(\"datasourceId\").isin(wCgc_list)).withColumn(\n",
    "    \"homogenized_count\", F.count(\"homogenized\").over(w)\n",
    ")\n",
    "\n",
    "# Then, compute the max of those counts over the same window\n",
    "df_with_max = df_with_count.withColumn(\n",
    "    \"maxDoE\", F.max(\"homogenized_count\").over(w)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 3508:>                                                       (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|  diseaseId|homogenized|maxDoE|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-----------+-----------+------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000198911|EFO_0006843|GoF_protect|    14|         14|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|\n",
      "+---------------+-----------+-----------+------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_with_max.filter((F.col('targetId')=='ENSG00000198911') & (F.col('diseaseId')=='EFO_0006843')).groupBy(\n",
    "    \"targetId\", \"diseaseId\", \"homogenized\", \"maxDoE\"\n",
    ").pivot('homogenized').count()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=assessment_all.filter((F.col('targetId')=='ENSG00000198911') & (F.col('diseaseId')=='EFO_0006843')).filter(F.col(\"datasourceId\").isin(wCgc_list)).withColumn(\n",
    "    \"homogenized_count\", F.count(\"homogenized\").over(w)\n",
    ").persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|  diseaseId|homogenized_count|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-----------+-----------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000198911|EFO_0006843|               14|         14|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|\n",
      "+---------------+-----------+-----------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "w = Window.partitionBy(\"targetId\", \"diseaseId\",\"homogenized\")\n",
    "discrepancifier(test.withColumn(\n",
    "    \"homogenized_count\", F.count(\"homogenized\").over(w)\n",
    ").groupBy('targetId','diseaseId','homogenized_count').pivot('homogenized').count()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|  diseaseId|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-----------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000198911|EFO_0006843|         14|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|\n",
      "+---------------+-----------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "discrepancifier(test.groupBy('targetId','diseaseId').pivot('homogenized').count()).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "test2=assessment_all.filter((F.col('targetId')=='ENSG00000085721') & (F.col('diseaseId')=='EFO_0005110')).filter(F.col(\"datasourceId\").isin(wCgc_list)).withColumn(\n",
    "    \"homogenized_count\", F.count(\"homogenized\").over(w)\n",
    ").persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "columnas=['GoF_risk','GoF_protect','LoF_protect','LoF_risk']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/23 23:39:55 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/23 23:39:55 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/23 23:39:55 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------+--------+-----------+-----------+-----------------+----------------+-------------------+------+---------------+\n",
      "|       targetId|  diseaseId|LoF_protect|LoF_risk|GoF_risk|GoF_protect|noEvaluable|coherencyDiagonal|coherencyOneCell|             arrayN|maxDoE|neutralDiagonal|\n",
      "+---------------+-----------+-----------+--------+--------+-----------+-----------+-----------------+----------------+-------------------+------+---------------+\n",
      "|ENSG00000085721|EFO_0005110|         12|       6|    NULL|       NULL|       NULL|           dispar|          dispar|[NULL, NULL, 12, 6]|    12|           this|\n",
      "+---------------+-----------+-----------+--------+--------+-----------+-----------+-----------------+----------------+-------------------+------+---------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "discrepancifier(\n",
    "    test2.groupBy(\"targetId\", \"diseaseId\").pivot(\"homogenized\").count()\n",
    ").withColumn(\"arrayN\", F.array(*columnas)).withColumn(\n",
    "    \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    ").withColumn(\n",
    "    \"neutralDiagonal\",\n",
    "    F.when(\n",
    "        (F.col(\"coherencyDiagonal\") == \"dispar\")\n",
    "        & ((F.col(\"LoF_protect\") == F.col('maxDoE')) | (F.col(\"GoF_risk\") == F.col('maxDoE')))\n",
    "        & ((F.col(\"LoF_risk\") == F.col('maxDoE')) | (F.col(\"GoF_protect\") == F.col('maxDoE'))),\n",
    "        F.lit(\"noResol\"),\n",
    "    ).otherwise(F.lit('this'))\n",
    ").show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/23 23:44:12 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/23 23:44:12 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/23 23:44:12 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "from pyspark.sql import functions as F\n",
    "\n",
    "columnas = [\"LoF_protect\", \"GoF_risk\", \"LoF_risk\", \"GoF_protect\"]\n",
    "\n",
    "# Add array of values\n",
    "df = discrepancifier(\n",
    "    test2.groupBy(\"targetId\", \"diseaseId\").pivot(\"homogenized\").count()\n",
    ").withColumn(\n",
    "    \"arrayN\", F.array(*[F.col(c) for c in columnas])\n",
    ").withColumn(\n",
    "    \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    ")\n",
    "\n",
    "# Create array of column names where value == maxDoE\n",
    "conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in columnas\n",
    "]\n",
    "\n",
    "df = df.withColumn(\"maxDoE_names\", F.array(*conditions))\n",
    "df = df.withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------+--------+--------+-----------+-----------+-----------------+----------------+-------------------+------+-------------+\n",
      "|       targetId|  diseaseId|LoF_protect|LoF_risk|GoF_risk|GoF_protect|noEvaluable|coherencyDiagonal|coherencyOneCell|             arrayN|maxDoE| maxDoE_names|\n",
      "+---------------+-----------+-----------+--------+--------+-----------+-----------+-----------------+----------------+-------------------+------+-------------+\n",
      "|ENSG00000085721|EFO_0005110|         12|       6|    NULL|       NULL|       NULL|           dispar|          dispar|[12, NULL, 6, NULL]|    12|[LoF_protect]|\n",
      "+---------------+-----------+-----------+--------+--------+-----------+-----------+-----------------+----------------+-------------------+------+-------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.withColumn('diagonalYes', F.when)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 4089:===================================================>(466 + 1) / 467]]]]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000085721|  EFO_0005110|       NULL|    NULL|         12|       6|       NULL|           dispar|          dispar|\n",
      "|ENSG00000135409|  EFO_0004732|       NULL|      12|       NULL|       6|       NULL|           dispar|          dispar|\n",
      "|ENSG00000186787|  EFO_0004324|       NULL|      57|         57|       6|       NULL|           dispar|          dispar|\n",
      "|ENSG00000176014|  EFO_1000357|          1|    NULL|         11|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000183624|  EFO_0004730|          6|    NULL|          6|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000180210|MONDO_0004634|       NULL|       3|       NULL|       3|       NULL|           dispar|          dispar|\n",
      "|ENSG00000188994|  EFO_0004306|         19|    NULL|         31|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000145850|  EFO_0008595|       NULL|       1|          8|      47|       NULL|           dispar|          dispar|\n",
      "|ENSG00000151292|  EFO_0008595|          7|      14|          7|       7|       NULL|           dispar|          dispar|\n",
      "|ENSG00000115677|  EFO_0004298|         12|       1|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000122034|  EFO_0004324|          2|     255|         41|    1299|       NULL|           dispar|          dispar|\n",
      "|ENSG00000198060|  EFO_0004529|         40|      17|       NULL|      16|       NULL|           dispar|          dispar|\n",
      "|ENSG00000104447|  EFO_0004338|         16|    NULL|         40|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000171703|  EFO_0803540|         24|      11|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000131183|  EFO_0010968|          5|    NULL|          1|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000164161|  EFO_0010736|       NULL|      28|       NULL|       2|       NULL|           dispar|          dispar|\n",
      "|ENSG00000150938|  EFO_0004339|          1|      16|       NULL|       2|       NULL|           dispar|          dispar|\n",
      "|ENSG00000108179|  EFO_0004584|         41|    NULL|         19|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000146733|  EFO_0004742|       NULL|    NULL|         25|      41|       NULL|           dispar|          dispar|\n",
      "|ENSG00000115232|  EFO_0003767|         34|    NULL|         40|    NULL|       NULL|           dispar|          dispar|\n",
      "+---------------+-------------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(assessment_all.groupBy('targetId','diseaseId').pivot('homogenized').count()).filter(F.col('coherencyDiagonal')=='dispar').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2569:==================>                                  (20 + 37) / 57]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|homogenized_count|maxDoE|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "|ENSG00000110492|  EFO_0008354|                5|     5|       NULL|    NULL|          5|       5|       NULL|           dispar|          dispar|\n",
      "|ENSG00000133030|  EFO_0004587|               21|    21|         21|    NULL|         21|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000133606|  EFO_0004742|                2|     2|       NULL|    NULL|          2|       2|       NULL|           dispar|          dispar|\n",
      "|ENSG00000178573|  EFO_0000508|                1|     1|       NULL|    NULL|          1|       1|       NULL|           dispar|          dispar|\n",
      "|ENSG00000130204|  EFO_0010351|               40|    40|         40|      40|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000167183|  EFO_0004308|                4|     4|          4|    NULL|          4|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000109667|  EFO_0005090|               10|    10|         10|    NULL|         10|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000179833|  EFO_0006841|                5|     5|       NULL|    NULL|          5|       5|       NULL|           dispar|          dispar|\n",
      "|ENSG00000255974|  EFO_0004518|                2|     2|          2|    NULL|          2|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000167258|  EFO_0000403|                1|     1|       NULL|       1|       NULL|       1|       NULL|           dispar|          dispar|\n",
      "|ENSG00000162607|  EFO_0008595|                3|     3|       NULL|    NULL|          3|       3|       NULL|           dispar|          dispar|\n",
      "|ENSG00000140575|  EFO_0004516|               48|    48|         48|      48|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000144401|  EFO_0004309|                1|     1|          1|    NULL|          1|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000198911|  EFO_0006843|                7|     7|          7|    NULL|          7|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000122691|  EFO_0004339|                4|     4|          4|       4|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000151292|  EFO_0005105|                7|     7|       NULL|       7|          7|       7|       NULL|           dispar|          dispar|\n",
      "|ENSG00000114650|  EFO_0004509|                1|     1|       NULL|       1|       NULL|       1|       NULL|           dispar|          dispar|\n",
      "|ENSG00000180210|MONDO_0000473|                4|     4|          4|       4|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000138085|  EFO_0021902|                1|     1|          1|       1|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "|ENSG00000140386|  EFO_0004747|                2|     2|          2|       2|       NULL|    NULL|       NULL|           dispar|          dispar|\n",
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_with_max2).filter(F.col('coherencyDiagonal')=='dispar').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql import functions as F\n",
    "\n",
    "# Get the list of homogenized categories (the pivoted column names)\n",
    "homogenized_cols = [col for col in df_with_max2.columns if col not in {\"targetId\", \"diseaseId\", \"homogenized_count\", \"maxDoE\"}]\n",
    "\n",
    "# Create a new column to capture the column name where value == maxDoE\n",
    "df_result = df_with_max2.withColumn(\n",
    "    \"maxHomogenizedLabel\",\n",
    "    F.expr(\n",
    "        \"CASE \" +\n",
    "        \" \".join(\n",
    "            [f\"WHEN `{col}` = maxDoE THEN '{col}'\" for col in homogenized_cols]\n",
    "        ) +\n",
    "        \" ELSE NULL END\"\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### evaluate this only in the ones that are uncoherent, and different from 1 vs 1. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1707:===========>                                         (12 + 45) / 57]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|homogenized_count|maxDoE|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|maxHomogenizedLabel|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------------+----------------+\n",
      "|ENSG00000037280|  EFO_0000349|                1|     1|       NULL|       1|       NULL|       1|       NULL|           GoF_risk|           dispar|          dispar|\n",
      "|ENSG00000113083|  EFO_0006841|                2|     2|          2|       2|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000179889|  EFO_0008592|                3|     3|       NULL|       3|       NULL|       3|       NULL|           GoF_risk|           dispar|          dispar|\n",
      "|ENSG00000162607|  EFO_0008595|                3|     3|       NULL|    NULL|          3|       3|       NULL|        LoF_protect|           dispar|          dispar|\n",
      "|ENSG00000198099|  EFO_0004747|               12|    12|       NULL|    NULL|         12|      12|       NULL|        LoF_protect|           dispar|          dispar|\n",
      "|ENSG00000166949|MONDO_0004784|                9|     9|       NULL|       9|       NULL|       9|       NULL|           GoF_risk|           dispar|          dispar|\n",
      "|ENSG00000158874|  EFO_0008589|                4|     4|          4|       4|       NULL|       4|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000130203|  EFO_0022174|                3|     3|          3|    NULL|          3|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000025039|  EFO_0004742|               48|    48|         48|      48|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000082701|  EFO_0004574|               13|    13|         13|    NULL|         13|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000140455|  EFO_0000400|                7|     7|       NULL|    NULL|          7|       7|       NULL|        LoF_protect|           dispar|          dispar|\n",
      "|ENSG00000105376|  EFO_0005140|                1|     1|          1|       1|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000133703|  EFO_1001972|                1|     1|       NULL|       1|       NULL|       1|       NULL|           GoF_risk|           dispar|          dispar|\n",
      "|ENSG00000075568|  EFO_0004536|               22|    22|         22|    NULL|         22|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000124570|  EFO_0004516|                1|     1|          1|       1|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000115556|  EFO_0004696|               24|    24|         24|    NULL|         24|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000123636|  EFO_0007989|               12|    12|         12|    NULL|         12|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000160185|  EFO_0803540|                1|     1|       NULL|    NULL|          1|       1|       NULL|        LoF_protect|           dispar|          dispar|\n",
      "|ENSG00000182010|  EFO_1001055|                8|     8|          8|    NULL|          8|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "|ENSG00000151006|  EFO_0004870|                5|     5|          5|       5|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|\n",
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------------+----------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_result).filter(F.col('coherencyDiagonal')=='dispar').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1748:===================================================>(464 + 3) / 467]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+---------------+-------------+-------------------+--------+----------+----+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+----------+--------------------+--------+-------------+---------------------+--------------+-----------------+--------+--------------------+---------------+--------------------+--------+-------------------+-------------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-----------------+--------------------+----------------------------+-------------------+--------------+---------+--------------------------------+--------------------------------+--------------+--------------+--------+---------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+------------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----+----------------------------+-------------------+------------------------------+--------------------+------------------+-----------+------+---------------+------------+-------------------+-------------+--------------------+-------------------+-------------+----------------+--------+-------------------------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+------------+-----------------+------+\n",
      "|     datasourceId|       targetId|alleleOrigins|allelicRequirements|ancestry|ancestryId|beta|betaConfidenceIntervalLower|betaConfidenceIntervalUpper|biologicalModelAllelicComposition|biologicalModelGeneticBackground|biologicalModelId|biomarkerName|biomarkers|biosamplesFromSource|cellType|clinicalPhase|clinicalSignificances|clinicalStatus|cohortDescription|cohortId|    cohortPhenotypes|cohortShortName|          confidence|contrast|crisprScreenLibrary|         datatypeId|diseaseCellLines|   diseaseFromSource|diseaseFromSourceId|diseaseFromSourceMappedId|diseaseModelAssociatedHumanPhenotypes|diseaseModelAssociatedModelPhenotypes|drugFromSource|drugId|drugResponse|geneticBackground|          literature|log2FoldChangePercentileRank|log2FoldChangeValue|mutatedSamples|OddsRatio|oddsRatioConfidenceIntervalLower|oddsRatioConfidenceIntervalUpper|pValueExponent|pValueMantissa|pathways|projectId|reactionId|reactionName|releaseDate|releaseVersion|resourceScore| sex|significantDriverMethods|statisticalMethod|statisticalMethodOverview|statisticalTestTail|studyCases|studyCasesWithQualifyingVariants|     studyId|studyOverview|studySampleSize|studyStartDate|studyStopReason|studyStopReasonCategories|targetFromSource|targetFromSourceId|targetInModel|targetInModelEnsemblId|targetInModelMgiId|targetModulation|urls|variantAminoacidDescriptions|variantFromSourceId|variantFunctionalConsequenceId|       variantHgvsId|         variantId|variantRsId|pmcIds|publicationYear|studyLocusId|textMiningSentences|    diseaseId|                  id|              score|variantEffect|directionOnTrait|sourceId|variantFunctionalConsequenceFromQtlId|target_id|approvedSymbol|description|description_splited|TSorOncogene|targetId2|drugId2|actionType|     inhibitors_list|     activators_list|intogen_function|intogenAnnot|homogenized| leftStudyId|homogenized_count|maxDoE|\n",
      "+-----------------+---------------+-------------+-------------------+--------+----------+----+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+----------+--------------------+--------+-------------+---------------------+--------------+-----------------+--------+--------------------+---------------+--------------------+--------+-------------------+-------------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-----------------+--------------------+----------------------------+-------------------+--------------+---------+--------------------------------+--------------------------------+--------------+--------------+--------+---------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+------------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----+----------------------------+-------------------+------------------------------+--------------------+------------------+-----------+------+---------------+------------+-------------------+-------------+--------------------+-------------------+-------------+----------------+--------+-------------------------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+------------+-----------------+------+\n",
      "|             impc|ENSG00000000938|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Fgr<Mhdaali18>/Fg...|            C3HeB/FeJ-Fgr<Mhd...|      MGI:3688148|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|Gnathodiaphyseal ...|        ORPHA:53697|            MONDO_0008151|                 [{HP:0006487, Bow...|                 [{MP:0002111, abn...|          NULL|  NULL|        NULL|             NULL|          [16964445]|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        55.04|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000938|          Fgr|    ENSMUSG00000028874|         MGI:95527|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0008151|8eded14d700f072e9...|             0.5504|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                1|     1|\n",
      "|             impc|ENSG00000000938|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Fgr<Mhdaali18>/Fg...|            C3HeB/FeJ-Fgr<Mhd...|      MGI:3688148|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|Brachydactylous D...|         ORPHA:2619|            MONDO_0013232|                 [{HP:0005086, Kne...|                 [{MP:0002111, abn...|          NULL|  NULL|        NULL|             NULL|          [16964445]|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         65.7|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000938|          Fgr|    ENSMUSG00000028874|         MGI:95527|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0013232|1f27cc11639c841b9...|              0.657|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                1|     1|\n",
      "|             impc|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cfh<tm1Mbo>/Cfh<t...|              B6.129-Cfh<tm1Mbo>|      MGI:5546609|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|Macular Degenerat...|        OMIM:615439|              EFO_0001365|                 [{HP:0011510, Dru...|                 [{MP:0012143, dec...|          NULL|  NULL|        NULL|             NULL|[28057640, 237990...|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        59.75|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|          Cfh|    ENSMUSG00000026365|         MGI:88385|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0001365|e836735ada4adf51a...|             0.5975|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                2|     2|\n",
      "|             impc|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cfh<tm1Mbo>/Cfh<t...|              B6.129-Cfh<tm1Mbo>|      MGI:5546609|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|Macular Degenerat...|        OMIM:603075|              EFO_0001365|                 [{HP:0000608, Mac...|                 [{MP:0012143, dec...|          NULL|  NULL|        NULL|             NULL|[28057640, 237990...|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        58.54|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|          Cfh|    ENSMUSG00000026365|         MGI:88385|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0001365|a1ab54f0a58aeecbb...|             0.5854|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                2|     2|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|gwas_credible_set|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|                 NULL|          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|               NULL|            NULL|                NULL|               NULL|                     NULL|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|              NULL|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|  EFO_0004747|                NULL|               NULL|         NULL|            NULL|    NULL|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|                NULL|                NULL|            NULL|        NULL|   LoF_risk|GCST90319599|                8|     8|\n",
      "|             impc|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cfh<tm1Mbo>/Cfh<t...|            involves: 129/Sv ...|      MGI:2662550|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|Focal Segmental G...|        OMIM:613237|            MONDO_0013191|                 [{HP:0002907, Mic...|                 [{MP:0002083, pre...|          NULL|  NULL|        NULL|             NULL|[17517971, 120919...|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        60.53|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|          Cfh|    ENSMUSG00000026365|         MGI:88385|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0013191|f8ab808c26ea2b746...| 0.6053000000000001|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                1|     1|\n",
      "|             impc|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cfh<tm1Mbo>/Cfh<t...|              B6.129-Cfh<tm1Mbo>|      MGI:5546609|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|Congenital Statio...|          ORPHA:215|            MONDO_0016293|                 [{HP:0007703, Abn...|                 [{MP:0012143, dec...|          NULL|  NULL|        NULL|             NULL|[28057640, 237990...|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        49.01|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|          Cfh|    ENSMUSG00000026365|         MGI:88385|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0016293|8bf039bb06448145f...|             0.4901|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                1|     1|\n",
      "|             impc|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|             Cfh<tm1Mbo>/Cfh<t...|            involves: 129/Sv ...|      MGI:2662550|         NULL|      NULL|                NULL|    NULL|         NULL|                     |          NULL|             NULL|    NULL|                NULL|           NULL|                NULL|    NULL|               NULL|       animal_model|            NULL|    Renal Hypoplasia|        ORPHA:93101|            MONDO_0019637|                 [{HP:0000093, Pro...|                 [{MP:0002083, pre...|          NULL|  NULL|        NULL|             NULL|[17517971, 120919...|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL|       NULL|          NULL|        49.06|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|        NULL|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|          Cfh|    ENSMUSG00000026365|         MGI:88385|            NULL|NULL|                        NULL|               NULL|                          NULL|                NULL|              NULL|       NULL|  NULL|           NULL|        NULL|               NULL|MONDO_0019637|644e060a3a281c029...|0.49060000000000004|          LoF|            risk|    impc|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|   LoF_risk|        NULL|                1|     1|\n",
      "|              eva|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL| uncertain signifi...|          NULL|             NULL|    NULL|[AHUS, SUSCEPTIBI...|           NULL|criteria provided...|    NULL|               NULL|genetic_association|            NULL| Factor H deficiency|           C0398777|            MONDO_0019738|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL| 2022-12-31|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|RCV002484478|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|       VCV001413155|                    SO_0001630|NC_000001.11:g.19...|   1_196725118_A_C|rs778174105|  NULL|           NULL|        NULL|               NULL|MONDO_0019738|959a7784de893b744...|               0.32|  noEvaluable|     noEvaluable|     eva|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|noEvaluable|        NULL|              223|   223|\n",
      "|              eva|ENSG00000000971|   [germline]|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|               benign|          NULL|             NULL|    NULL|[CFH DEFICIENCY, ...|           NULL|criteria provided...|    NULL|               NULL|genetic_association|            NULL| Factor H deficiency|           C0398777|            MONDO_0019738|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL| 2023-12-09|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|RCV003446800|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|       VCV001168607|                    SO_0002169|NC_000001.11:g.19...|1_196673839_C_CTTT| rs35507625|  NULL|           NULL|        NULL|               NULL|MONDO_0019738|b2eed98ef456d5207...|               0.02|  noEvaluable|     noEvaluable|     eva|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|noEvaluable|        NULL|              223|   223|\n",
      "|              eva|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL|        likely benign|          NULL|             NULL|    NULL|[AHUS, SUSCEPTIBI...|           NULL|criteria provided...|    NULL|               NULL|genetic_association|            NULL| Factor H deficiency|           C0398777|            MONDO_0019738|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL| 2022-12-31|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|RCV002500213|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|       VCV001630939|                    SO_0002169|NC_000001.11:g.19...|   1_196741860_T_C|rs756514818|  NULL|           NULL|        NULL|               NULL|MONDO_0019738|8f3d7a6570d602738...|               0.02|  noEvaluable|     noEvaluable|     eva|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|noEvaluable|        NULL|              223|   223|\n",
      "|              eva|ENSG00000000971|         NULL|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL| uncertain signifi...|          NULL|             NULL|    NULL|[AHUS, SUSCEPTIBI...|           NULL|criteria provided...|    NULL|               NULL|genetic_association|            NULL| Factor H deficiency|           C0398777|            MONDO_0019738|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL| 2022-12-31|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|RCV002478327|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|       VCV001403152|                    SO_0001583|NC_000001.11:g.19...|   1_196677524_G_A|rs370640334|  NULL|           NULL|        NULL|               NULL|MONDO_0019738|a969640f288401d0e...|               0.32|  noEvaluable|     noEvaluable|     eva|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|noEvaluable|        NULL|              223|   223|\n",
      "|              eva|ENSG00000000971|   [germline]|               NULL|    NULL|      NULL|NULL|                       NULL|                       NULL|                             NULL|                            NULL|             NULL|         NULL|      NULL|                NULL|    NULL|         NULL| uncertain signifi...|          NULL|             NULL|    NULL|[CFH DEFICIENCY, ...|           NULL|criteria provided...|    NULL|               NULL|genetic_association|            NULL| Factor H deficiency|           C0398777|            MONDO_0019738|                                 NULL|                                 NULL|          NULL|  NULL|        NULL|             NULL|                NULL|                        NULL|               NULL|          NULL|     NULL|                            NULL|                            NULL|          NULL|          NULL|    NULL|     NULL|      NULL|        NULL| 2023-12-24|          NULL|         NULL|NULL|                    NULL|             NULL|                     NULL|               NULL|      NULL|                            NULL|RCV003453959|         NULL|           NULL|          NULL|           NULL|                     NULL|            NULL|   ENSG00000000971|         NULL|                  NULL|              NULL|            NULL|NULL|                        NULL|       VCV001507082|                    SO_0001583|NC_000001.11:g.19...|   1_196713909_C_T|rs377008918|  NULL|           NULL|        NULL|               NULL|MONDO_0019738|82bf55c6825779f4e...|               0.32|  noEvaluable|     noEvaluable|     eva|                                 NULL|     NULL|          NULL|       NULL|               NULL|        NULL|     NULL|   NULL|      NULL|[RNAI INHIBITOR, ...|[PARTIAL AGONIST,...|            NULL|           0|noEvaluable|        NULL|              223|   223|\n",
      "+-----------------+---------------+-------------+-------------------+--------+----------+----+---------------------------+---------------------------+---------------------------------+--------------------------------+-----------------+-------------+----------+--------------------+--------+-------------+---------------------+--------------+-----------------+--------+--------------------+---------------+--------------------+--------+-------------------+-------------------+----------------+--------------------+-------------------+-------------------------+-------------------------------------+-------------------------------------+--------------+------+------------+-----------------+--------------------+----------------------------+-------------------+--------------+---------+--------------------------------+--------------------------------+--------------+--------------+--------+---------+----------+------------+-----------+--------------+-------------+----+------------------------+-----------------+-------------------------+-------------------+----------+--------------------------------+------------+-------------+---------------+--------------+---------------+-------------------------+----------------+------------------+-------------+----------------------+------------------+----------------+----+----------------------------+-------------------+------------------------------+--------------------+------------------+-----------+------+---------------+------------+-------------------+-------------+--------------------+-------------------+-------------+----------------+--------+-------------------------------------+---------+--------------+-----------+-------------------+------------+---------+-------+----------+--------------------+--------------------+----------------+------------+-----------+------------+-----------------+------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "df_with_max.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2510:===========>                                         (12 + 45) / 57]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------------+------+\n",
      "|neutralDiagonal|coherencyDiagonal| count|\n",
      "+---------------+-----------------+------+\n",
      "|    GoF_protect|         coherent| 68353|\n",
      "|       LoF_risk|         coherent|760316|\n",
      "|    LoF_protect|         coherent| 68390|\n",
      "|    noEvaluable|       EvidNotDoE| 84646|\n",
      "|        neutral|           dispar|  6510|\n",
      "|       GoF_risk|         coherent| 86618|\n",
      "+---------------+-----------------+------+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_result).withColumn(\n",
    "    \"neutralDiagonal\",\n",
    "    F.when(\n",
    "        (F.col(\"coherencyDiagonal\") == \"dispar\")\n",
    "        & ((F.col(\"LoF_protect\") == F.col('maxDoE')) | (F.col(\"GoF_risk\") == F.col('maxDoE')))\n",
    "        & ((F.col(\"LoF_risk\") == F.col('maxDoE')) | (F.col(\"GoF_protect\") == F.col('maxDoE'))),\n",
    "        F.lit(\"neutral\"),\n",
    "    ).otherwise(F.col(\"maxHomogenizedLabel\")),\n",
    ").groupBy('neutralDiagonal','coherencyDiagonal').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2461:=====================================================>(56 + 1) / 57]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------------+----------------+---------------+\n",
      "|       targetId|  diseaseId|homogenized_count|maxDoE|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|maxHomogenizedLabel|coherencyDiagonal|coherencyOneCell|neutralDiagonal|\n",
      "+---------------+-----------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------------+----------------+---------------+\n",
      "|ENSG00000153815|EFO_0004529|              305|   305|       NULL|    NULL|        305|     305|       NULL|        LoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000198099|EFO_0004530|               28|    28|       NULL|      28|       NULL|      28|       NULL|           GoF_risk|           dispar|          dispar|        neutral|\n",
      "|ENSG00000197653|EFO_0803546|               44|    44|         44|      44|       NULL|      44|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000168769|EFO_0000729|                1|     1|       NULL|    NULL|          1|       1|       NULL|        LoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000081721|EFO_0004503|                3|     3|          3|    NULL|          3|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000122691|EFO_0004339|                4|     4|          4|       4|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000198053|EFO_0004528|                9|     9|       NULL|       9|       NULL|       9|       NULL|           GoF_risk|           dispar|          dispar|        neutral|\n",
      "|ENSG00000182362|EFO_0009431|                3|     3|       NULL|       3|       NULL|       3|       NULL|           GoF_risk|           dispar|          dispar|        neutral|\n",
      "|ENSG00000014123|EFO_0004346|               33|    33|         33|    NULL|         33|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000133030|EFO_0004587|               21|    21|         21|    NULL|         21|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000012983|EFO_0004306|                3|     3|       NULL|    NULL|          3|       3|       NULL|        LoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000124920|EFO_0020049|               12|    12|         12|      12|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000141338|EFO_0022336|              149|   149|       NULL|     149|       NULL|     149|       NULL|           GoF_risk|           dispar|          dispar|        neutral|\n",
      "|ENSG00000186188|EFO_0008591|                1|     1|          1|       1|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000170776|EFO_0004308|               43|    43|         43|    NULL|         43|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000138085|EFO_0022255|                1|     1|          1|       1|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000079313|EFO_0010700|                1|     1|       NULL|    NULL|          1|       1|       NULL|        LoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000133065|EFO_0004612|                2|     2|          2|       2|       NULL|    NULL|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000088543|EFO_0004324|                2|     2|          2|    NULL|          2|       2|       NULL|        GoF_protect|           dispar|          dispar|        neutral|\n",
      "|ENSG00000146540|EFO_0005110|               39|    39|       NULL|      39|       NULL|      39|       NULL|           GoF_risk|           dispar|          dispar|        neutral|\n",
      "+---------------+-----------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------------+----------------+---------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_result).withColumn(\n",
    "    \"neutralDiagonal\",\n",
    "    F.when(\n",
    "        (F.col(\"coherencyDiagonal\") == \"dispar\")\n",
    "        & ((F.col(\"LoF_protect\") == F.col('maxDoE')) | (F.col(\"GoF_risk\") == F.col('maxDoE')))\n",
    "        & ((F.col(\"LoF_risk\") == F.col('maxDoE')) | (F.col(\"GoF_protect\") == F.col('maxDoE'))),\n",
    "        F.lit(\"neutral\"),\n",
    "    ).otherwise(F.col(\"maxHomogenizedLabel\")),\n",
    ").filter(F.col('neutralDiagonal')=='neutral').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'from pyspark.sql.functions import col, array, aggregate, lit\n",
    "\n",
    "# Identify the pivot columns only (exclude non-pivot ones)\n",
    "non_pivot_cols = {\"targetId\", \"diseaseId\", \"homogenized_count\", \"maxDoE\",'maxHomogenizedLabel','noEvaluable'}\n",
    "pivot_cols = [c for c in df_result.columns if c not in non_pivot_cols]\n",
    "\n",
    "# Create array of comparisons (col == maxDoE)\n",
    "comparisons = array(*[(col(c) == col(\"maxDoE\")).cast(\"int\") for c in pivot_cols])\n",
    "\n",
    "# Use `aggregate` to count how many columns equal the max\n",
    "df_with_discrepancy = df_result.withColumn(\n",
    "    \"max_matches\",\n",
    "    aggregate(comparisons, lit(0), lambda acc, x: acc + x)\n",
    ").withColumn(\n",
    "    \"hasDiscrepancy\", F.when(\n",
    "    F.col(\"max_matches\") >= 2, F.lit('dispar')\n",
    ").otherwise(F.lit('coherent')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql.functions import col, array, aggregate, lit\n",
    "\n",
    "# List the columns you want to compare\n",
    "cols_to_check = [\"GoF_protect\", \"GoF_risk\", \"LoF_protect\", \"LoF_risk\"]\n",
    "\n",
    "# Create array of Boolean comparisons, cast to int (1 if match, else 0)\n",
    "comparison_array = array(*[(col(c) == col(\"maxDoE\")).cast(\"int\") for c in cols_to_check])\n",
    "\n",
    "# Count how many columns matched the maxDoE\n",
    "df_with_check = df_result.withColumn(\n",
    "    \"num_max_matches\",\n",
    "    aggregate(comparison_array, lit(0), lambda acc, x: acc + x)\n",
    ").withColumn(\n",
    "    \"has_discrepancy\",\n",
    "    col(\"num_max_matches\") >= 2\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "AnalysisException",
     "evalue": "[UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `hasDiscrepancy` cannot be resolved. Did you mean one of the following? [`has_discrepancy`, `diseaseId`, `GoF_protect`, `GoF_risk`, `LoF_protect`].;\n'Aggregate ['hasDiscrepancy], ['hasDiscrepancy, count(1) AS count#227838L]\n+- Filter (coherencyDiagonal#227794 = dispar)\n   +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, num_max_matches#227766, has_discrepancy#227781, coherencyDiagonal#227794, CASE WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnull(noEvaluable#190395L)) THEN noEvid WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnotnull(noEvaluable#190395L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#190394L) OR isnotnull(LoF_protect#190393L)) OR isnotnull(GoF_risk#190392L)) OR isnotnull(GoF_protect#190391L)) THEN CASE WHEN (isnotnull(LoF_risk#190394L) AND ((isnull(LoF_protect#190393L) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L))) THEN coherent WHEN (isnotnull(GoF_risk#190392L) AND ((isnull(LoF_protect#190393L) AND isnull(LoF_risk#190394L)) AND isnull(GoF_protect#190391L))) THEN coherent WHEN (isnotnull(LoF_protect#190393L) AND ((isnull(LoF_risk#190394L) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L))) THEN coherent WHEN (isnotnull(GoF_protect#190391L) AND ((isnull(LoF_protect#190393L) AND isnull(GoF_risk#190392L)) AND isnull(LoF_risk#190394L))) THEN coherent ELSE dispar END END AS coherencyOneCell#227808]\n      +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, num_max_matches#227766, has_discrepancy#227781, CASE WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnull(noEvaluable#190395L)) THEN noEvid WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnotnull(noEvaluable#190395L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#190394L) OR isnotnull(LoF_protect#190393L)) OR isnotnull(GoF_risk#190392L)) OR isnotnull(GoF_protect#190391L)) THEN CASE WHEN (isnotnull(GoF_risk#190392L) AND isnotnull(LoF_risk#190394L)) THEN dispar WHEN (isnotnull(LoF_protect#190393L) AND isnotnull(LoF_risk#190394L)) THEN dispar WHEN (isnotnull(GoF_protect#190391L) AND isnotnull(GoF_risk#190392L)) THEN dispar WHEN (isnotnull(GoF_protect#190391L) AND isnotnull(LoF_protect#190393L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#227794]\n         +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, num_max_matches#227766, (num_max_matches#227766 >= 2) AS has_discrepancy#227781]\n            +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, aggregate(array(cast((GoF_protect#190391L = maxDoE#176188L) as int), cast((GoF_risk#190392L = maxDoE#176188L) as int), cast((LoF_protect#190393L = maxDoE#176188L) as int), cast((LoF_risk#190394L = maxDoE#176188L) as int)), 0, lambdafunction((lambda x_8#227767 + lambda y_9#227768), lambda x_8#227767, lambda y_9#227768, false), lambdafunction(lambda id#227769, lambda id#227769, false)) AS num_max_matches#227766]\n               +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, CASE WHEN (GoF_protect#190391L = maxDoE#176188L) THEN GoF_protect WHEN (GoF_risk#190392L = maxDoE#176188L) THEN GoF_risk WHEN (LoF_protect#190393L = maxDoE#176188L) THEN LoF_protect WHEN (LoF_risk#190394L = maxDoE#176188L) THEN LoF_risk WHEN (noEvaluable#190395L = maxDoE#176188L) THEN noEvaluable ELSE cast(null as string) END AS maxHomogenizedLabel#190426]\n                  +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, __pivot_count(1) AS count AS `count(1) AS count`#190390[0] AS GoF_protect#190391L, __pivot_count(1) AS count AS `count(1) AS count`#190390[1] AS GoF_risk#190392L, __pivot_count(1) AS count AS `count(1) AS count`#190390[2] AS LoF_protect#190393L, __pivot_count(1) AS count AS `count(1) AS count`#190390[3] AS LoF_risk#190394L, __pivot_count(1) AS count AS `count(1) AS count`#190390[4] AS noEvaluable#190395L]\n                     +- Aggregate [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L], [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, pivotfirst(homogenized#4153, count(1) AS count#190378L, GoF_protect, GoF_risk, LoF_protect, LoF_risk, noEvaluable, 0, 0) AS __pivot_count(1) AS count AS `count(1) AS count`#190390]\n                        +- Aggregate [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, homogenized#4153], [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, homogenized#4153, count(1) AS count(1) AS count#190378L]\n                           +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 82 more fields]\n                              +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 83 more fields]\n                                 +- Window [max(homogenized_count#176081L) windowspecdefinition(targetId#2033, diseaseId#2115, homogenized#4153, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS maxDoE#176188L], [targetId#2033, diseaseId#2115, homogenized#4153]\n                                    +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 81 more fields]\n                                       +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 81 more fields]\n                                          +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 82 more fields]\n                                             +- Window [count(homogenized#4153) windowspecdefinition(targetId#2033, diseaseId#2115, homogenized#4153, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS homogenized_count#176081L], [targetId#2033, diseaseId#2115, homogenized#4153]\n                                                +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 80 more fields]\n                                                   +- Filter datasourceId#2032 IN (gene_burden,intogen,eva,eva_somatic,impc,orphanet,gene2phenotype,gwas_credible_set,cancer_gene_census)\n                                                      +- Union false, false\n                                                         :- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 80 more fields]\n                                                         :  +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 79 more fields]\n                                                         :     +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 78 more fields]\n                                                         :        +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 78 more fields]\n                                                         :           +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 78 more fields]\n                                                         :              +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 79 more fields]\n                                                         :                 +- Window [collect_set(intogen_function#3736, 0, 0) windowspecdefinition(targetId#2033, diseaseId#2115, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#3842], [targetId#2033, diseaseId#2115]\n                                                         :                    +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 77 more fields]\n                                                         :                       +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 77 more fields]\n                                                         :                          +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 76 more fields]\n                                                         :                             +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 75 more fields]\n                                                         :                                +- Join LeftOuter, ((drugId2#2820 = drugId#2066) AND (targetId2#2827 = targetId#2033))\n                                                         :                                   :- Join LeftOuter, (target_id#2870 = targetId#2033)\n                                                         :                                   :  :- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, concat_ws(,, clinicalSignificances#2049) AS clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :  +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#2049, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :     +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, cast(beta#2038 as double) AS beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#2049, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :        +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2038, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#2049, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :           +- Filter datasourceId#2032 IN (gwas_credible_set,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                         :                                   :  :              +- Relation [datasourceId#2032,targetId#2033,alleleOrigins#2034,allelicRequirements#2035,ancestry#2036,ancestryId#2037,beta#2038,betaConfidenceIntervalLower#2039,betaConfidenceIntervalUpper#2040,biologicalModelAllelicComposition#2041,biologicalModelGeneticBackground#2042,biologicalModelId#2043,biomarkerName#2044,biomarkers#2045,biosamplesFromSource#2046,cellType#2047,clinicalPhase#2048,clinicalSignificances#2049,clinicalStatus#2050,cohortDescription#2051,cohortId#2052,cohortPhenotypes#2053,cohortShortName#2054,confidence#2055,... 65 more fields] parquet\n                                                         :                                   :  +- Project [id#2747 AS target_id#2870, approvedSymbol#2748, description#2855, description_splited#2859, TSorOncogene#2864]\n                                                         :                                   :     +- Project [id#2747, approvedSymbol#2748, description#2855, description_splited#2859, CASE WHEN (RLIKE(description_splited#2859, ncogene) AND RLIKE(description_splited#2859, TSG)) THEN bivalent WHEN RLIKE(description_splited#2859, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#2859, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#2864]\n                                                         :                                   :        +- Project [id#2747, approvedSymbol#2748, description#2855, concat_ws(,, description#2855) AS description_splited#2859]\n                                                         :                                   :           +- Aggregate [id#2747, approvedSymbol#2748], [id#2747, approvedSymbol#2748, collect_set(description#2847, 0, 0) AS description#2855]\n                                                         :                                   :              +- Filter description#2847 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                                         :                                   :                 +- Project [id#2747, approvedSymbol#2748, col#2842.description AS description#2847]\n                                                         :                                   :                    +- Project [id#2747, approvedSymbol#2748, col#2842]\n                                                         :                                   :                       +- Generate explode(hallmarks#2757.attributes), true, [col#2842]\n                                                         :                                   :                          +- Relation [id#2747,approvedSymbol#2748,biotype#2749,transcriptIds#2750,canonicalTranscript#2751,canonicalExons#2752,genomicLocation#2753,alternativeGenes#2754,approvedName#2755,go#2756,hallmarks#2757,synonyms#2758,symbolSynonyms#2759,nameSynonyms#2760,functionDescriptions#2761,subcellularLocations#2762,targetClass#2763,obsoleteSymbols#2764,obsoleteNames#2765,constraint#2766,tep#2767,proteinIds#2768,dbXrefs#2769,chemicalProbes#2770,... 5 more fields] parquet\n                                                         :                                   +- Aggregate [targetId2#2827, drugId2#2820], [targetId2#2827, drugId2#2820, collect_set(actionType#2805, 0, 0) AS actionType#2837]\n                                                         :                                      +- Project [targetId2#2827, drugId2#2820, actionType#2805, mechanismOfAction#2806]\n                                                         :                                         +- Generate explode(targets#2810), true, [targetId2#2827]\n                                                         :                                            +- Project [drugId2#2820, actionType#2805, mechanismOfAction#2806, targets#2810]\n                                                         :                                               +- Generate explode(chemblIds#2807), true, [drugId2#2820]\n                                                         :                                                  +- Relation [actionType#2805,mechanismOfAction#2806,chemblIds#2807,targetName#2808,targetType#2809,targets#2810,references#2811] parquet\n                                                         +- Project [datasourceId#6381, targetId#1758, null AS alleleOrigins#6387, null AS allelicRequirements#6388, null AS ancestry#6389, null AS ancestryId#6390, null AS beta#6391, null AS betaConfidenceIntervalLower#6392, null AS betaConfidenceIntervalUpper#6393, null AS biologicalModelAllelicComposition#6394, null AS biologicalModelGeneticBackground#6395, null AS biologicalModelId#6396, null AS biomarkerName#6397, null AS biomarkers#6398, null AS biosamplesFromSource#6399, null AS cellType#6400, null AS clinicalPhase#6401, null AS clinicalSignificances#6402, null AS clinicalStatus#6403, null AS cohortDescription#6404, null AS cohortId#6405, null AS cohortPhenotypes#6406, null AS cohortShortName#6407, null AS confidence#6408, ... 80 more fields]\n                                                            +- Project [targetId#1758, diseaseId#1911, leftStudyId#1078, homogenized#6188, gwas_credible_set AS datasourceId#6381]\n                                                               +- Project [targetId#1758, diseaseId#1911, leftStudyId#1078, CASE WHEN isnull(homogenized#6142) THEN noEvaluable ELSE homogenized#6142 END AS homogenized#6188]\n                                                                  +- Project [targetId#1758, diseaseId#1911, leftStudyId#1078, homogenized#6142]\n                                                                     +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 17 more fields]\n                                                                        +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 18 more fields]\n                                                                           +- Window [first(colocDoE#1991, true) windowspecdefinition(targetId#1758, diseaseId#1911, leftStudyId#1078, pValueExponent#1808 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS homogenized#6142], [targetId#1758, diseaseId#1911, leftStudyId#1078], [pValueExponent#1808 ASC NULLS FIRST]\n                                                                              +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                 +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                    +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 15 more fields]\n                                                                                       +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                          +- Generate explode(concat(array(diseaseId#800), parents#694)), true, [diseaseId#1911]\n                                                                                             +- Project [diseaseId#800, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                                +- Join LeftOuter, (diseaseId#800 = diseaseId#1864)\n                                                                                                   :- Project [leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, biosampleId#1002, ... 13 more fields]\n                                                                                                   :  +- Join Inner, ((leftStudyLocusId#949 = leftStudyLocusId#1783) AND (targetId#1758 = targetId#718))\n                                                                                                   :     :- Project [rightStudyId#1103, rightStudyLocusId#950, leftStudyLocusId#949, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, geneId#974 AS targetId#1758, projectId#975, indexStudyType#1159, condition#998, biosampleId#1002]\n                                                                                                   :     :  +- Project [rightStudyId#1103, rightStudyLocusId#950, leftStudyLocusId#949, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, geneId#974, projectId#975, indexStudyType#1159, condition#998, biosampleId#1002]\n                                                                                                   :     :     +- Join LeftOuter, (rightStudyId#1103 = rightStudyId#1158)\n                                                                                                   :     :        :- Project [rightStudyLocusId#950, leftStudyLocusId#949, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightStudyId#1103, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137]\n                                                                                                   :     :        :  +- Join LeftOuter, (rightStudyLocusId#950 = rightStudyLocusId#1102)\n                                                                                                   :     :        :     :- Project [leftStudyLocusId#949, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080]\n                                                                                                   :     :        :     :  +- Join LeftOuter, (leftStudyLocusId#949 = leftStudyLocusId#1077)\n                                                                                                   :     :        :     :     :- Relation [leftStudyLocusId#949,rightStudyLocusId#950,chromosome#951,rightStudyType#952,numberColocalisingVariants#953L,h0#954,h1#955,h2#956,h3#957,h4#958,colocalisationMethod#959,betaRatioSignAverage#960] parquet\n                                                                                                   :     :        :     :     +- Project [studyLocusId#895 AS leftStudyLocusId#1077, StudyId#896 AS leftStudyId#1078, variantId#897 AS leftVariantId#1079, studyType#920 AS credibleLeftStudyType#1080]\n                                                                                                   :     :        :     :        +- Relation [studyLocusId#895,studyId#896,variantId#897,chromosome#898,position#899,region#900,beta#901,zScore#902,pValueMantissa#903,pValueExponent#904,effectAlleleFrequencyFromSource#905,standardError#906,subStudyDescription#907,qualityControls#908,finemappingMethod#909,credibleSetIndex#910,credibleSetlog10BF#911,purityMeanR2#912,purityMinR2#913,locusStart#914,locusEnd#915,sampleSize#916,ldSet#917,locus#918,... 3 more fields] parquet\n                                                                                                   :     :        :     +- Project [studyLocusId#1111 AS rightStudyLocusId#1102, studyId#1112 AS rightStudyId#1103, variantId#1113 AS rightVariantId#1104, studyType#1136 AS credibleRightStudyType#1105, isTransQtl#1137]\n                                                                                                   :     :        :        +- Relation [studyLocusId#1111,studyId#1112,variantId#1113,chromosome#1114,position#1115,region#1116,beta#1117,zScore#1118,pValueMantissa#1119,pValueExponent#1120,effectAlleleFrequencyFromSource#1121,standardError#1122,subStudyDescription#1123,qualityControls#1124,finemappingMethod#1125,credibleSetIndex#1126,credibleSetlog10BF#1127,purityMeanR2#1128,purityMinR2#1129,locusStart#1130,locusEnd#1131,sampleSize#1132,ldSet#1133,locus#1134,... 3 more fields] parquet\n                                                                                                   :     :        +- Project [studyId#973 AS rightStudyId#1158, geneId#974, projectId#975, studyType#976 AS indexStudyType#1159, condition#998, biosampleId#1002]\n                                                                                                   :     :           +- Relation [studyId#973,geneId#974,projectId#975,studyType#976,traitFromSource#977,traitFromSourceMappedIds#978,biosampleFromSourceId#979,pubmedId#980,publicationTitle#981,publicationFirstAuthor#982,publicationDate#983,publicationJournal#984,backgroundTraitFromSourceMappedIds#985,initialSampleSize#986,nCases#987,nControls#988,nSamples#989,cohorts#990,ldPopulationStructure#991,discoverySamples#992,replicationSamples#993,qualityControls#994,analysisFlags#995,summarystatsLocation#996,... 6 more fields] parquet\n                                                                                                   :     +- Project [studyLocusId#798 AS leftStudyLocusId#1783, datasourceId#717, targetId#718, datatypeId#743, diseaseFromSourceMappedId#747, resourceScore#769, targetFromSourceId#784, diseaseId#800, id#801, score#802, sourceId#805, studyId#1800, variantId#1801, betaGwas#1737, pValueExponent#1808]\n                                                                                                   :        +- Project [studyLocusId#798, datasourceId#717, targetId#718, datatypeId#743, diseaseFromSourceMappedId#747, resourceScore#769, targetFromSourceId#784, diseaseId#800, id#801, score#802, sourceId#805, studyId#1800, variantId#1801, betaGwas#1737, pValueExponent#1808]\n                                                                                                   :           +- Join LeftOuter, (studyLocusId#798 = studyLocusId#1799)\n                                                                                                   :              :- Project [datasourceId#717, targetId#718, datatypeId#743, diseaseFromSourceMappedId#747, resourceScore#769, targetFromSourceId#784, studyLocusId#798, diseaseId#800, id#801, score#802, sourceId#805]\n                                                                                                   :              :  +- Filter (datasourceId#717 = gwas_credible_sets)\n                                                                                                   :              :     +- Relation [datasourceId#717,targetId#718,alleleOrigins#719,allelicRequirements#720,ancestry#721,ancestryId#722,beta#723,betaConfidenceIntervalLower#724,betaConfidenceIntervalUpper#725,biologicalModelAllelicComposition#726,biologicalModelGeneticBackground#727,biologicalModelId#728,biomarkerName#729,biomarkers#730,biosamplesFromSource#731,cellType#732,clinicalPhase#733,clinicalSignificances#734,clinicalStatus#735,cohortDescription#736,cohortId#737,cohortPhenotypes#738,cohortShortName#739,confidence#740,... 65 more fields] parquet\n                                                                                                   :              +- Project [studyLocusId#1799, studyId#1800, variantId#1801, beta#1805 AS betaGwas#1737, pValueExponent#1808]\n                                                                                                   :                 +- Relation [studyLocusId#1799,studyId#1800,variantId#1801,chromosome#1802,position#1803,region#1804,beta#1805,zScore#1806,pValueMantissa#1807,pValueExponent#1808,effectAlleleFrequencyFromSource#1809,standardError#1810,subStudyDescription#1811,qualityControls#1812,finemappingMethod#1813,credibleSetIndex#1814,credibleSetlog10BF#1815,purityMeanR2#1816,purityMinR2#1817,locusStart#1818,locusEnd#1819,sampleSize#1820,ldSet#1821,locus#1822,... 3 more fields] parquet\n                                                                                                   +- Project [id#689 AS diseaseId#1864, name#691, parents#694, therapeuticAreas#700]\n                                                                                                      +- Relation [id#689,code#690,name#691,description#692,dbXRefs#693,parents#694,synonyms#695,obsoleteTerms#696,obsoleteXRefs#697,children#698,ancestors#699,therapeuticAreas#700,descendants#701,ontology#702] parquet\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAnalysisException\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[43mdiscrepancifier\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf_with_check\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilter\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[43mF\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcol\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcoherencyDiagonal\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdispar\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[0;32m----> 3\u001b[0m \u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgroupBy\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhasDiscrepancy\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcount\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mshow()\n",
      "File \u001b[0;32m/usr/lib/spark/python/pyspark/sql/group.py:38\u001b[0m, in \u001b[0;36mdfapi.<locals>._api\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_api\u001b[39m(\u001b[38;5;28mself\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGroupedData\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame:\n\u001b[1;32m     37\u001b[0m     name \u001b[38;5;241m=\u001b[39m f\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\n\u001b[0;32m---> 38\u001b[0m     jdf \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_jgd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     39\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m DataFrame(jdf, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msession)\n",
      "File \u001b[0;32m/usr/lib/spark/python/lib/py4j-0.10.9.7-src.zip/py4j/java_gateway.py:1322\u001b[0m, in \u001b[0;36mJavaMember.__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1316\u001b[0m command \u001b[38;5;241m=\u001b[39m proto\u001b[38;5;241m.\u001b[39mCALL_COMMAND_NAME \u001b[38;5;241m+\u001b[39m\\\n\u001b[1;32m   1317\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcommand_header \u001b[38;5;241m+\u001b[39m\\\n\u001b[1;32m   1318\u001b[0m     args_command \u001b[38;5;241m+\u001b[39m\\\n\u001b[1;32m   1319\u001b[0m     proto\u001b[38;5;241m.\u001b[39mEND_COMMAND_PART\n\u001b[1;32m   1321\u001b[0m answer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgateway_client\u001b[38;5;241m.\u001b[39msend_command(command)\n\u001b[0;32m-> 1322\u001b[0m return_value \u001b[38;5;241m=\u001b[39m \u001b[43mget_return_value\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1323\u001b[0m \u001b[43m    \u001b[49m\u001b[43manswer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgateway_client\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtarget_id\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1325\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m temp_arg \u001b[38;5;129;01min\u001b[39;00m temp_args:\n\u001b[1;32m   1326\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(temp_arg, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_detach\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n",
      "File \u001b[0;32m/usr/lib/spark/python/pyspark/errors/exceptions/captured.py:185\u001b[0m, in \u001b[0;36mcapture_sql_exception.<locals>.deco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    181\u001b[0m converted \u001b[38;5;241m=\u001b[39m convert_exception(e\u001b[38;5;241m.\u001b[39mjava_exception)\n\u001b[1;32m    182\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(converted, UnknownException):\n\u001b[1;32m    183\u001b[0m     \u001b[38;5;66;03m# Hide where the exception came from that shows a non-Pythonic\u001b[39;00m\n\u001b[1;32m    184\u001b[0m     \u001b[38;5;66;03m# JVM exception message.\u001b[39;00m\n\u001b[0;32m--> 185\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m converted \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    186\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    187\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "\u001b[0;31mAnalysisException\u001b[0m: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `hasDiscrepancy` cannot be resolved. Did you mean one of the following? [`has_discrepancy`, `diseaseId`, `GoF_protect`, `GoF_risk`, `LoF_protect`].;\n'Aggregate ['hasDiscrepancy], ['hasDiscrepancy, count(1) AS count#227838L]\n+- Filter (coherencyDiagonal#227794 = dispar)\n   +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, num_max_matches#227766, has_discrepancy#227781, coherencyDiagonal#227794, CASE WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnull(noEvaluable#190395L)) THEN noEvid WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnotnull(noEvaluable#190395L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#190394L) OR isnotnull(LoF_protect#190393L)) OR isnotnull(GoF_risk#190392L)) OR isnotnull(GoF_protect#190391L)) THEN CASE WHEN (isnotnull(LoF_risk#190394L) AND ((isnull(LoF_protect#190393L) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L))) THEN coherent WHEN (isnotnull(GoF_risk#190392L) AND ((isnull(LoF_protect#190393L) AND isnull(LoF_risk#190394L)) AND isnull(GoF_protect#190391L))) THEN coherent WHEN (isnotnull(LoF_protect#190393L) AND ((isnull(LoF_risk#190394L) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L))) THEN coherent WHEN (isnotnull(GoF_protect#190391L) AND ((isnull(LoF_protect#190393L) AND isnull(GoF_risk#190392L)) AND isnull(LoF_risk#190394L))) THEN coherent ELSE dispar END END AS coherencyOneCell#227808]\n      +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, num_max_matches#227766, has_discrepancy#227781, CASE WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnull(noEvaluable#190395L)) THEN noEvid WHEN ((((isnull(LoF_risk#190394L) AND isnull(LoF_protect#190393L)) AND isnull(GoF_risk#190392L)) AND isnull(GoF_protect#190391L)) AND isnotnull(noEvaluable#190395L)) THEN EvidNotDoE WHEN (((isnotnull(LoF_risk#190394L) OR isnotnull(LoF_protect#190393L)) OR isnotnull(GoF_risk#190392L)) OR isnotnull(GoF_protect#190391L)) THEN CASE WHEN (isnotnull(GoF_risk#190392L) AND isnotnull(LoF_risk#190394L)) THEN dispar WHEN (isnotnull(LoF_protect#190393L) AND isnotnull(LoF_risk#190394L)) THEN dispar WHEN (isnotnull(GoF_protect#190391L) AND isnotnull(GoF_risk#190392L)) THEN dispar WHEN (isnotnull(GoF_protect#190391L) AND isnotnull(LoF_protect#190393L)) THEN dispar ELSE coherent END END AS coherencyDiagonal#227794]\n         +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, num_max_matches#227766, (num_max_matches#227766 >= 2) AS has_discrepancy#227781]\n            +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, maxHomogenizedLabel#190426, aggregate(array(cast((GoF_protect#190391L = maxDoE#176188L) as int), cast((GoF_risk#190392L = maxDoE#176188L) as int), cast((LoF_protect#190393L = maxDoE#176188L) as int), cast((LoF_risk#190394L = maxDoE#176188L) as int)), 0, lambdafunction((lambda x_8#227767 + lambda y_9#227768), lambda x_8#227767, lambda y_9#227768, false), lambdafunction(lambda id#227769, lambda id#227769, false)) AS num_max_matches#227766]\n               +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, GoF_protect#190391L, GoF_risk#190392L, LoF_protect#190393L, LoF_risk#190394L, noEvaluable#190395L, CASE WHEN (GoF_protect#190391L = maxDoE#176188L) THEN GoF_protect WHEN (GoF_risk#190392L = maxDoE#176188L) THEN GoF_risk WHEN (LoF_protect#190393L = maxDoE#176188L) THEN LoF_protect WHEN (LoF_risk#190394L = maxDoE#176188L) THEN LoF_risk WHEN (noEvaluable#190395L = maxDoE#176188L) THEN noEvaluable ELSE cast(null as string) END AS maxHomogenizedLabel#190426]\n                  +- Project [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, __pivot_count(1) AS count AS `count(1) AS count`#190390[0] AS GoF_protect#190391L, __pivot_count(1) AS count AS `count(1) AS count`#190390[1] AS GoF_risk#190392L, __pivot_count(1) AS count AS `count(1) AS count`#190390[2] AS LoF_protect#190393L, __pivot_count(1) AS count AS `count(1) AS count`#190390[3] AS LoF_risk#190394L, __pivot_count(1) AS count AS `count(1) AS count`#190390[4] AS noEvaluable#190395L]\n                     +- Aggregate [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L], [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, pivotfirst(homogenized#4153, count(1) AS count#190378L, GoF_protect, GoF_risk, LoF_protect, LoF_risk, noEvaluable, 0, 0) AS __pivot_count(1) AS count AS `count(1) AS count`#190390]\n                        +- Aggregate [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, homogenized#4153], [targetId#2033, diseaseId#2115, homogenized_count#176081L, maxDoE#176188L, homogenized#4153, count(1) AS count(1) AS count#190378L]\n                           +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 82 more fields]\n                              +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 83 more fields]\n                                 +- Window [max(homogenized_count#176081L) windowspecdefinition(targetId#2033, diseaseId#2115, homogenized#4153, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS maxDoE#176188L], [targetId#2033, diseaseId#2115, homogenized#4153]\n                                    +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 81 more fields]\n                                       +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 81 more fields]\n                                          +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 82 more fields]\n                                             +- Window [count(homogenized#4153) windowspecdefinition(targetId#2033, diseaseId#2115, homogenized#4153, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS homogenized_count#176081L], [targetId#2033, diseaseId#2115, homogenized#4153]\n                                                +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 80 more fields]\n                                                   +- Filter datasourceId#2032 IN (gene_burden,intogen,eva,eva_somatic,impc,orphanet,gene2phenotype,gwas_credible_set,cancer_gene_census)\n                                                      +- Union false, false\n                                                         :- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 80 more fields]\n                                                         :  +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 79 more fields]\n                                                         :     +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 78 more fields]\n                                                         :        +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 78 more fields]\n                                                         :           +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 78 more fields]\n                                                         :              +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 79 more fields]\n                                                         :                 +- Window [collect_set(intogen_function#3736, 0, 0) windowspecdefinition(targetId#2033, diseaseId#2115, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS _we0#3842], [targetId#2033, diseaseId#2115]\n                                                         :                    +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 77 more fields]\n                                                         :                       +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 77 more fields]\n                                                         :                          +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 76 more fields]\n                                                         :                             +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 75 more fields]\n                                                         :                                +- Join LeftOuter, ((drugId2#2820 = drugId#2066) AND (targetId2#2827 = targetId#2033))\n                                                         :                                   :- Join LeftOuter, (target_id#2870 = targetId#2033)\n                                                         :                                   :  :- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, concat_ws(,, clinicalSignificances#2049) AS clinicalSignificances#3058, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :  +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#2049, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :     +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, cast(beta#2038 as double) AS beta#2876, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#2049, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :        +- Project [datasourceId#2032, targetId#2033, alleleOrigins#2034, allelicRequirements#2035, ancestry#2036, ancestryId#2037, beta#2038, betaConfidenceIntervalLower#2039, betaConfidenceIntervalUpper#2040, biologicalModelAllelicComposition#2041, biologicalModelGeneticBackground#2042, biologicalModelId#2043, biomarkerName#2044, biomarkers#2045, biosamplesFromSource#2046, cellType#2047, clinicalPhase#2048, clinicalSignificances#2049, clinicalStatus#2050, cohortDescription#2051, cohortId#2052, cohortPhenotypes#2053, cohortShortName#2054, confidence#2055, ... 66 more fields]\n                                                         :                                   :  :           +- Filter datasourceId#2032 IN (gwas_credible_set,gene_burden,eva,eva_somatic,gene2phenotype,orphanet,cancer_gene_census,intogen,impc,chembl)\n                                                         :                                   :  :              +- Relation [datasourceId#2032,targetId#2033,alleleOrigins#2034,allelicRequirements#2035,ancestry#2036,ancestryId#2037,beta#2038,betaConfidenceIntervalLower#2039,betaConfidenceIntervalUpper#2040,biologicalModelAllelicComposition#2041,biologicalModelGeneticBackground#2042,biologicalModelId#2043,biomarkerName#2044,biomarkers#2045,biosamplesFromSource#2046,cellType#2047,clinicalPhase#2048,clinicalSignificances#2049,clinicalStatus#2050,cohortDescription#2051,cohortId#2052,cohortPhenotypes#2053,cohortShortName#2054,confidence#2055,... 65 more fields] parquet\n                                                         :                                   :  +- Project [id#2747 AS target_id#2870, approvedSymbol#2748, description#2855, description_splited#2859, TSorOncogene#2864]\n                                                         :                                   :     +- Project [id#2747, approvedSymbol#2748, description#2855, description_splited#2859, CASE WHEN (RLIKE(description_splited#2859, ncogene) AND RLIKE(description_splited#2859, TSG)) THEN bivalent WHEN RLIKE(description_splited#2859, ncogene(\\s|$)) THEN oncogene WHEN RLIKE(description_splited#2859, TSG(\\s|$)) THEN TSG ELSE noEvaluable END AS TSorOncogene#2864]\n                                                         :                                   :        +- Project [id#2747, approvedSymbol#2748, description#2855, concat_ws(,, description#2855) AS description_splited#2859]\n                                                         :                                   :           +- Aggregate [id#2747, approvedSymbol#2748], [id#2747, approvedSymbol#2748, collect_set(description#2847, 0, 0) AS description#2855]\n                                                         :                                   :              +- Filter description#2847 IN (TSG,oncogene,Oncogene,oncogene,oncogene,TSG,TSG,oncogene,fusion,oncogene,oncogene,fusion)\n                                                         :                                   :                 +- Project [id#2747, approvedSymbol#2748, col#2842.description AS description#2847]\n                                                         :                                   :                    +- Project [id#2747, approvedSymbol#2748, col#2842]\n                                                         :                                   :                       +- Generate explode(hallmarks#2757.attributes), true, [col#2842]\n                                                         :                                   :                          +- Relation [id#2747,approvedSymbol#2748,biotype#2749,transcriptIds#2750,canonicalTranscript#2751,canonicalExons#2752,genomicLocation#2753,alternativeGenes#2754,approvedName#2755,go#2756,hallmarks#2757,synonyms#2758,symbolSynonyms#2759,nameSynonyms#2760,functionDescriptions#2761,subcellularLocations#2762,targetClass#2763,obsoleteSymbols#2764,obsoleteNames#2765,constraint#2766,tep#2767,proteinIds#2768,dbXrefs#2769,chemicalProbes#2770,... 5 more fields] parquet\n                                                         :                                   +- Aggregate [targetId2#2827, drugId2#2820], [targetId2#2827, drugId2#2820, collect_set(actionType#2805, 0, 0) AS actionType#2837]\n                                                         :                                      +- Project [targetId2#2827, drugId2#2820, actionType#2805, mechanismOfAction#2806]\n                                                         :                                         +- Generate explode(targets#2810), true, [targetId2#2827]\n                                                         :                                            +- Project [drugId2#2820, actionType#2805, mechanismOfAction#2806, targets#2810]\n                                                         :                                               +- Generate explode(chemblIds#2807), true, [drugId2#2820]\n                                                         :                                                  +- Relation [actionType#2805,mechanismOfAction#2806,chemblIds#2807,targetName#2808,targetType#2809,targets#2810,references#2811] parquet\n                                                         +- Project [datasourceId#6381, targetId#1758, null AS alleleOrigins#6387, null AS allelicRequirements#6388, null AS ancestry#6389, null AS ancestryId#6390, null AS beta#6391, null AS betaConfidenceIntervalLower#6392, null AS betaConfidenceIntervalUpper#6393, null AS biologicalModelAllelicComposition#6394, null AS biologicalModelGeneticBackground#6395, null AS biologicalModelId#6396, null AS biomarkerName#6397, null AS biomarkers#6398, null AS biosamplesFromSource#6399, null AS cellType#6400, null AS clinicalPhase#6401, null AS clinicalSignificances#6402, null AS clinicalStatus#6403, null AS cohortDescription#6404, null AS cohortId#6405, null AS cohortPhenotypes#6406, null AS cohortShortName#6407, null AS confidence#6408, ... 80 more fields]\n                                                            +- Project [targetId#1758, diseaseId#1911, leftStudyId#1078, homogenized#6188, gwas_credible_set AS datasourceId#6381]\n                                                               +- Project [targetId#1758, diseaseId#1911, leftStudyId#1078, CASE WHEN isnull(homogenized#6142) THEN noEvaluable ELSE homogenized#6142 END AS homogenized#6188]\n                                                                  +- Project [targetId#1758, diseaseId#1911, leftStudyId#1078, homogenized#6142]\n                                                                     +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 17 more fields]\n                                                                        +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 18 more fields]\n                                                                           +- Window [first(colocDoE#1991, true) windowspecdefinition(targetId#1758, diseaseId#1911, leftStudyId#1078, pValueExponent#1808 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS homogenized#6142], [targetId#1758, diseaseId#1911, leftStudyId#1078], [pValueExponent#1808 ASC NULLS FIRST]\n                                                                              +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                 +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                    +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 15 more fields]\n                                                                                       +- Project [diseaseId#1911, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                          +- Generate explode(concat(array(diseaseId#800), parents#694)), true, [diseaseId#1911]\n                                                                                             +- Project [diseaseId#800, leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, ... 16 more fields]\n                                                                                                +- Join LeftOuter, (diseaseId#800 = diseaseId#1864)\n                                                                                                   :- Project [leftStudyLocusId#949, targetId#1758, rightStudyId#1103, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, projectId#975, indexStudyType#1159, condition#998, biosampleId#1002, ... 13 more fields]\n                                                                                                   :  +- Join Inner, ((leftStudyLocusId#949 = leftStudyLocusId#1783) AND (targetId#1758 = targetId#718))\n                                                                                                   :     :- Project [rightStudyId#1103, rightStudyLocusId#950, leftStudyLocusId#949, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, geneId#974 AS targetId#1758, projectId#975, indexStudyType#1159, condition#998, biosampleId#1002]\n                                                                                                   :     :  +- Project [rightStudyId#1103, rightStudyLocusId#950, leftStudyLocusId#949, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137, geneId#974, projectId#975, indexStudyType#1159, condition#998, biosampleId#1002]\n                                                                                                   :     :     +- Join LeftOuter, (rightStudyId#1103 = rightStudyId#1158)\n                                                                                                   :     :        :- Project [rightStudyLocusId#950, leftStudyLocusId#949, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080, rightStudyId#1103, rightVariantId#1104, credibleRightStudyType#1105, isTransQtl#1137]\n                                                                                                   :     :        :  +- Join LeftOuter, (rightStudyLocusId#950 = rightStudyLocusId#1102)\n                                                                                                   :     :        :     :- Project [leftStudyLocusId#949, rightStudyLocusId#950, chromosome#951, rightStudyType#952, numberColocalisingVariants#953L, h0#954, h1#955, h2#956, h3#957, h4#958, colocalisationMethod#959, betaRatioSignAverage#960, leftStudyId#1078, leftVariantId#1079, credibleLeftStudyType#1080]\n                                                                                                   :     :        :     :  +- Join LeftOuter, (leftStudyLocusId#949 = leftStudyLocusId#1077)\n                                                                                                   :     :        :     :     :- Relation [leftStudyLocusId#949,rightStudyLocusId#950,chromosome#951,rightStudyType#952,numberColocalisingVariants#953L,h0#954,h1#955,h2#956,h3#957,h4#958,colocalisationMethod#959,betaRatioSignAverage#960] parquet\n                                                                                                   :     :        :     :     +- Project [studyLocusId#895 AS leftStudyLocusId#1077, StudyId#896 AS leftStudyId#1078, variantId#897 AS leftVariantId#1079, studyType#920 AS credibleLeftStudyType#1080]\n                                                                                                   :     :        :     :        +- Relation [studyLocusId#895,studyId#896,variantId#897,chromosome#898,position#899,region#900,beta#901,zScore#902,pValueMantissa#903,pValueExponent#904,effectAlleleFrequencyFromSource#905,standardError#906,subStudyDescription#907,qualityControls#908,finemappingMethod#909,credibleSetIndex#910,credibleSetlog10BF#911,purityMeanR2#912,purityMinR2#913,locusStart#914,locusEnd#915,sampleSize#916,ldSet#917,locus#918,... 3 more fields] parquet\n                                                                                                   :     :        :     +- Project [studyLocusId#1111 AS rightStudyLocusId#1102, studyId#1112 AS rightStudyId#1103, variantId#1113 AS rightVariantId#1104, studyType#1136 AS credibleRightStudyType#1105, isTransQtl#1137]\n                                                                                                   :     :        :        +- Relation [studyLocusId#1111,studyId#1112,variantId#1113,chromosome#1114,position#1115,region#1116,beta#1117,zScore#1118,pValueMantissa#1119,pValueExponent#1120,effectAlleleFrequencyFromSource#1121,standardError#1122,subStudyDescription#1123,qualityControls#1124,finemappingMethod#1125,credibleSetIndex#1126,credibleSetlog10BF#1127,purityMeanR2#1128,purityMinR2#1129,locusStart#1130,locusEnd#1131,sampleSize#1132,ldSet#1133,locus#1134,... 3 more fields] parquet\n                                                                                                   :     :        +- Project [studyId#973 AS rightStudyId#1158, geneId#974, projectId#975, studyType#976 AS indexStudyType#1159, condition#998, biosampleId#1002]\n                                                                                                   :     :           +- Relation [studyId#973,geneId#974,projectId#975,studyType#976,traitFromSource#977,traitFromSourceMappedIds#978,biosampleFromSourceId#979,pubmedId#980,publicationTitle#981,publicationFirstAuthor#982,publicationDate#983,publicationJournal#984,backgroundTraitFromSourceMappedIds#985,initialSampleSize#986,nCases#987,nControls#988,nSamples#989,cohorts#990,ldPopulationStructure#991,discoverySamples#992,replicationSamples#993,qualityControls#994,analysisFlags#995,summarystatsLocation#996,... 6 more fields] parquet\n                                                                                                   :     +- Project [studyLocusId#798 AS leftStudyLocusId#1783, datasourceId#717, targetId#718, datatypeId#743, diseaseFromSourceMappedId#747, resourceScore#769, targetFromSourceId#784, diseaseId#800, id#801, score#802, sourceId#805, studyId#1800, variantId#1801, betaGwas#1737, pValueExponent#1808]\n                                                                                                   :        +- Project [studyLocusId#798, datasourceId#717, targetId#718, datatypeId#743, diseaseFromSourceMappedId#747, resourceScore#769, targetFromSourceId#784, diseaseId#800, id#801, score#802, sourceId#805, studyId#1800, variantId#1801, betaGwas#1737, pValueExponent#1808]\n                                                                                                   :           +- Join LeftOuter, (studyLocusId#798 = studyLocusId#1799)\n                                                                                                   :              :- Project [datasourceId#717, targetId#718, datatypeId#743, diseaseFromSourceMappedId#747, resourceScore#769, targetFromSourceId#784, studyLocusId#798, diseaseId#800, id#801, score#802, sourceId#805]\n                                                                                                   :              :  +- Filter (datasourceId#717 = gwas_credible_sets)\n                                                                                                   :              :     +- Relation [datasourceId#717,targetId#718,alleleOrigins#719,allelicRequirements#720,ancestry#721,ancestryId#722,beta#723,betaConfidenceIntervalLower#724,betaConfidenceIntervalUpper#725,biologicalModelAllelicComposition#726,biologicalModelGeneticBackground#727,biologicalModelId#728,biomarkerName#729,biomarkers#730,biosamplesFromSource#731,cellType#732,clinicalPhase#733,clinicalSignificances#734,clinicalStatus#735,cohortDescription#736,cohortId#737,cohortPhenotypes#738,cohortShortName#739,confidence#740,... 65 more fields] parquet\n                                                                                                   :              +- Project [studyLocusId#1799, studyId#1800, variantId#1801, beta#1805 AS betaGwas#1737, pValueExponent#1808]\n                                                                                                   :                 +- Relation [studyLocusId#1799,studyId#1800,variantId#1801,chromosome#1802,position#1803,region#1804,beta#1805,zScore#1806,pValueMantissa#1807,pValueExponent#1808,effectAlleleFrequencyFromSource#1809,standardError#1810,subStudyDescription#1811,qualityControls#1812,finemappingMethod#1813,credibleSetIndex#1814,credibleSetlog10BF#1815,purityMeanR2#1816,purityMinR2#1817,locusStart#1818,locusEnd#1819,sampleSize#1820,ldSet#1821,locus#1822,... 3 more fields] parquet\n                                                                                                   +- Project [id#689 AS diseaseId#1864, name#691, parents#694, therapeuticAreas#700]\n                                                                                                      +- Relation [id#689,code#690,name#691,description#692,dbXRefs#693,parents#694,synonyms#695,obsoleteTerms#696,obsoleteXRefs#697,children#698,ancestors#699,therapeuticAreas#700,descendants#701,ontology#702] parquet\n"
     ]
    }
   ],
   "source": [
    "\n",
    "discrepancifier(df_with_check).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ").groupBy('hasDiscrepancy').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['targetId',\n",
       " 'diseaseId',\n",
       " 'homogenized_count',\n",
       " 'maxDoE',\n",
       " 'GoF_protect',\n",
       " 'GoF_risk',\n",
       " 'LoF_protect',\n",
       " 'LoF_risk',\n",
       " 'noEvaluable',\n",
       " 'maxHomogenizedLabel']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2061:===================================================>  (54 + 3) / 57]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----+\n",
      "|hasDiscrepancy|count|\n",
      "+--------------+-----+\n",
      "|      coherent| 6430|\n",
      "|        dispar|   11|\n",
      "+--------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_with_discrepancy).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ").groupBy('hasDiscrepancy').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                ]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------+--------------+-----------------+----------------+\n",
      "|       targetId|    diseaseId|homogenized_count|maxDoE|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|maxHomogenizedLabel|max_matches|hasDiscrepancy|coherencyDiagonal|coherencyOneCell|\n",
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------+--------------+-----------------+----------------+\n",
      "|ENSG00000158987|  EFO_0004742|                1|     1|          1|       1|          1|       1|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000135845|  EFO_0010700|               12|    12|         12|      12|         12|      12|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000109180|  EFO_0004747|               73|    73|         73|      73|         73|      73|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000090975|MONDO_0000771|                6|     6|          6|       6|          6|       6|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000103657|  EFO_0004696|                4|     4|          4|       4|          4|       4|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000197826|  EFO_0004191|                2|     2|          2|       2|          2|       2|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000136280|  EFO_0004309|               35|    35|         35|      35|         35|      35|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000100209|  EFO_0004529|                1|     1|          1|       1|          1|       1|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000126653|  EFO_0004747|               11|    11|         11|      11|         11|      11|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000170035|  EFO_0004509|               20|    20|         20|      20|         20|      20|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000157107|  EFO_0004838|                8|     8|          8|       8|          8|       8|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000169756|  EFO_0004309|               10|    10|         10|      10|         10|      10|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000126653|  EFO_0004464|               11|    11|         11|      11|         11|      11|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "|ENSG00000145029|  EFO_0004747|                8|     8|          8|       8|          8|       8|       NULL|        GoF_protect|          4|        dispar|           dispar|          dispar|\n",
      "+---------------+-------------+-----------------+------+-----------+--------+-----------+--------+-----------+-------------------+-----------+--------------+-----------------+----------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_with_discrepancy).filter(F.col('hasDiscrepancy')=='dispar').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2002:=================================================>    (52 + 5) / 57]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+--------------+-----+\n",
      "|max_matches|hasDiscrepancy|count|\n",
      "+-----------+--------------+-----+\n",
      "|       NULL|      coherent| 6531|\n",
      "+-----------+--------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "discrepancifier(df_with_discrepancy).filter(\n",
    "    F.col(\"coherencyDiagonal\") == \"dispar\"\n",
    ").groupBy('max_matches','hasDiscrepancy').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1210:=================================================> (451 + 16) / 467]]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------+-----------+--------+-----------+--------+-----------+\n",
      "|       targetId|     diseaseId|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|\n",
      "+---------------+--------------+-----------+--------+-----------+--------+-----------+\n",
      "|ENSG00000151617|   EFO_0000612|       NULL|       3|       NULL|    NULL|       NULL|\n",
      "|ENSG00000100583|   EFO_0004586|       NULL|      30|       NULL|    NULL|       NULL|\n",
      "|ENSG00000176909|   EFO_0003779|       NULL|      11|       NULL|    NULL|       NULL|\n",
      "|ENSG00000000938| MONDO_0019156|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000060138| MONDO_0033671|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000077782| MONDO_0007227|       NULL|    NULL|       NULL|       4|       NULL|\n",
      "|ENSG00000077935| MONDO_0054728|       NULL|    NULL|       NULL|       3|       NULL|\n",
      "|ENSG00000148584| MONDO_0007152|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000157087| MONDO_0012116|       NULL|    NULL|       NULL|       4|       NULL|\n",
      "|ENSG00000018408| MONDO_0000640|       NULL|       1|       NULL|    NULL|       NULL|\n",
      "|ENSG00000112531| MONDO_0014733|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000144891| MONDO_0007173|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000156650|   EFO_0000308|       NULL|    NULL|       NULL|    NULL|          1|\n",
      "|ENSG00000162434|   EFO_0006544|       NULL|    NULL|       NULL|    NULL|          1|\n",
      "|ENSG00000169045| MONDO_0030801|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000171456|   EFO_0003763|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000112394|Orphanet_79156|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000112893| MONDO_0019740|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "|ENSG00000117400| MONDO_0011173|       NULL|    NULL|       NULL|      14|         19|\n",
      "|ENSG00000120436| MONDO_0011735|       NULL|    NULL|       NULL|       1|       NULL|\n",
      "+---------------+--------------+-----------+--------+-----------+--------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "(assessment_all.filter(F.col(\"datasourceId\").isin(wCgc_list))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded files\n",
      "loaded newColoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded gwasComplete\n",
      "loaded resolvedColloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 01:18:31 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 01:18:32 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run temporary direction of effect\n",
      "Moving to step 2\n",
      "defining non propagated,propagated and analysis_drugs functions\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 01:18:33 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 01:18:33 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 01:18:33 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "defining full_analysis_propagation\n",
      "defining full analysis no propagation\n",
      "moving to Step 3\n",
      "starting dictionaries at 2025-04-24 01:18:33.412624\n",
      "gene_burden\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 01:18:34 WARN CacheManager: Asked to cache already cached data.\n",
      "25/04/24 01:18:34 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intogen\n",
      "cancer_gene_census\n",
      "eva\n",
      "eva_somatic\n",
      "gwas_credible_set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "impc\n",
      "orphanet\n",
      "gene2phenotype\n",
      "WOcgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wCgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "somatic\n",
      "germline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "import time\n",
    "#from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "# from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "# from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "#from itertools import islice\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    StringType,\n",
    "    IntegerType,\n",
    "    ArrayType\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "spark.conf.set(\n",
    "    \"spark.sql.shuffle.partitions\", \"400\"\n",
    ")  # Default is 200, increase if needed\n",
    "\n",
    "\n",
    "path_n='gs://open-targets-data-releases/25.03/output/'\n",
    "\n",
    "target = spark.read.parquet(f\"{path_n}target/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path_n}disease/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path_n}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path_n}credible_set\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path_n}colocalisation_coloc\") \n",
    "\n",
    "index=spark.read.parquet(f\"{path_n}study/\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path_n}variant\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path_n}biosample\")\n",
    "\n",
    "print(\"loaded files\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "            'isTransQtl'\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    ")\n",
    "    # .persist()\n",
    ")\n",
    "\n",
    "print(\"loaded newColoc\")\n",
    "\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns)  # .persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")  # .persist()\n",
    "\n",
    "print(\"loaded gwasComplete\")\n",
    "\n",
    "resolvedColoc = (\n",
    "    (\n",
    "        newColoc.withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"inner\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    ).withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    # .persist()\n",
    ")\n",
    "print(\"loaded resolvedColloc\")\n",
    "\n",
    "datasource_filter = [\n",
    "    \"gwas_credible_set\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]\n",
    "\n",
    "assessment, evidences, actionType, oncolabel = temporary_directionOfEffect(\n",
    "    path_n, datasource_filter\n",
    ")\n",
    "\n",
    "print(\"run temporary direction of effect\")\n",
    "\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\",'leftStudyId').orderBy( ### include gwas study\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    resolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )  ## added 30.01.2025\n",
    "    .select(\"targetId\", \"diseaseId\",'leftStudyId', \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "negativeTD = (\n",
    "    evidences.filter(F.col(\"datasourceId\") == \"chembl\")\n",
    "    .select(\"targetId\", \"diseaseId\", \"studyStopReason\", \"studyStopReasonCategories\")\n",
    "    .filter(F.array_contains(F.col(\"studyStopReasonCategories\"), \"Negative\"))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .count()\n",
    "    .withColumn(\"stopReason\", F.lit(\"Negative\"))\n",
    "    .drop(\"count\")\n",
    ")\n",
    "\n",
    "assessment_all = assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "print(\"defining non propagated,propagated and analysis_drugs functions\")\n",
    "\n",
    "def analysis_nonPropagated(assessment_all, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment_all, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "\n",
    "def analysis_drugs(assessment_all, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter((F.col(\"datasourceId\").isin(chembl_ds))\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment_all, chembl_ds)\n",
    "\n",
    "#######\n",
    "## include here the analysis\n",
    "#######\n",
    "\n",
    "analysisDatasources = []\n",
    "\n",
    "print(\"defining full_analysis_propagation\")\n",
    "\n",
    "doe_columns=[\"LoF_protect\", \"GoF_risk\", \"LoF_risk\", \"GoF_protect\"]\n",
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "\n",
    "def full_analysis_propagation(\n",
    "    doe_columns,assessment_all, analysisDatasources, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "):\n",
    "    conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in doe_columns\n",
    "    ]\n",
    "    \n",
    "    return (\n",
    "        analysis_propagated(assessment_all, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        #### Should remove the coherencyDiagonal.isNotNull()\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        ).withColumn(\n",
    "            \"arrayN\", F.array(*[F.col(c) for c in doe_columns])\n",
    "        ).withColumn(\n",
    "            \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    "        ).withColumn(\"maxDoE_names\", F.array(*conditions)\n",
    "        ).withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\")\n",
    "        ).withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(negativeTD, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"stopReason\") == \"Negative\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        )\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "#####\n",
    "## no propag\n",
    "#####\n",
    "print(\"defining full analysis no propagation\")\n",
    "\n",
    "\n",
    "def full_analysis_noPropagation(\n",
    "    doe_columns,assessment_all, analysisDatasources, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "):\n",
    "    conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in doe_columns\n",
    "    ]\n",
    "    return (\n",
    "        analysis_nonPropagated(assessment_all, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        ).withColumn(\n",
    "            \"arrayN\", F.array(*[F.col(c) for c in doe_columns])\n",
    "        ).withColumn(\n",
    "            \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    "        ).withColumn(\"maxDoE_names\", F.array(*conditions)\n",
    "        ).withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\")\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(negativeTD, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"stopReason\") == \"Negative\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        ).withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        )\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "print(\"moving to Step 3\")\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "c = datetime.now()\n",
    "print(\"starting dictionaries at\", c)\n",
    "\n",
    "#### continue here on 10.07.2024\n",
    "\n",
    "## 1nd dictionary\n",
    "dfs_dict = {}  ### checked and changed on 01.06.2023\n",
    "dfs_dict_propag = {}\n",
    "\n",
    "\n",
    "wocgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "]\n",
    "wCgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"cancer_gene_census\",\n",
    "]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"WOcgc\",\n",
    "    \"wCgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "\n",
    "germline_list = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic_list = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "\n",
    "# assessment = prueba_assessment.filter(F.col(\"datasourceId\").isin(datasources_analysis))\n",
    "def dataset_builder(assessment_all, value, analysis_chembl, negativeTD, diseaseTA):\n",
    "    nonPropagated = full_analysis_noPropagation(\n",
    "        doe_columns,assessment_all, value, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "    )\n",
    "    propagated = full_analysis_propagation(\n",
    "        doe_columns,assessment_all, value, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "    )\n",
    "    return (\n",
    "        # Non propagation\n",
    "        ## All\n",
    "        nonPropagated,\n",
    "        ## Other\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        nonPropagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "        # Propagation\n",
    "        ## All\n",
    "        propagated,\n",
    "        ## Other\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        propagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "    )\n",
    "\n",
    "\n",
    "for value in datasource_list:\n",
    "    print(value)\n",
    "    if value == \"WOcgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, wocgc_list, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "    elif value == \"wCgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, wCgc_list, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "    elif value == \"germline\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all,\n",
    "            germline_list,\n",
    "            analysis_chembl,\n",
    "            negativeTD,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    elif value == \"somatic\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all,\n",
    "            somatic_list,\n",
    "            analysis_chembl,\n",
    "            negativeTD,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"]\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, value, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df() -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    comparisons = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"hasGeneticEvidence\", \"byDatatype\"),\n",
    "            (\"diagonalYes\", \"byDatatype\"),\n",
    "            (\"oneCellYes\", \"byDatatype\"),\n",
    "            (\"NdiagonalYes\", \"byDatatype\"),\n",
    "            (\"NoneCellYes\", \"byDatatype\"),\n",
    "        ],\n",
    "        schema=StructType(\n",
    "            [\n",
    "                StructField(\"comparison\", StringType(), True),\n",
    "                StructField(\"comparisonType\", StringType(), True),\n",
    "            ]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### see the error: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-04-24 06:17:17.038629\n",
      "Analysis started on 2025-04-24 at  2025-04-24 06:17:17.038629\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 06:17:21 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded files\n",
      "loaded newColoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded gwasComplete\n",
      "loaded resolvedColloc\n",
      "run temporary direction of effect\n",
      "Moving to step 2\n",
      "defining non propagated,propagated and analysis_drugs functions\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "defining full_analysis_propagation\n",
      "defining full analysis no propagation\n",
      "moving to Step 3\n",
      "starting dictionaries at 2025-04-24 06:18:22.479136\n",
      "gene_burden\n",
      "intogen\n",
      "cancer_gene_census\n",
      "eva\n",
      "eva_somatic\n",
      "gwas_credible_set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "impc\n",
      "orphanet\n",
      "gene2phenotype\n",
      "WOcgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wCgc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "somatic\n",
      "germline\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "\n",
    "from itertools import islice\n",
    "import time\n",
    "#from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "# from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "# from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "#from itertools import islice\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    StringType,\n",
    "    IntegerType,\n",
    "    ArrayType\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "spark.conf.set(\n",
    "    \"spark.sql.shuffle.partitions\", \"400\"\n",
    ")  # Default is 200, increase if needed\n",
    "\n",
    "\n",
    "path_n='gs://open-targets-data-releases/25.03/output/'\n",
    "\n",
    "target = spark.read.parquet(f\"{path_n}target/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path_n}disease/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path_n}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path_n}credible_set\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path_n}colocalisation_coloc\") \n",
    "\n",
    "index=spark.read.parquet(f\"{path_n}study/\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path_n}variant\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path_n}biosample\")\n",
    "\n",
    "print(\"loaded files\")\n",
    "\n",
    "newColoc = (\n",
    "    new.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "            'isTransQtl'\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    ")\n",
    "    # .persist()\n",
    ")\n",
    "\n",
    "print(\"loaded newColoc\")\n",
    "\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns)  # .persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")  # .persist()\n",
    "\n",
    "print(\"loaded gwasComplete\")\n",
    "\n",
    "resolvedColoc = (\n",
    "    (\n",
    "        newColoc.withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"inner\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    ).withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    # .persist()\n",
    ")\n",
    "print(\"loaded resolvedColloc\")\n",
    "\n",
    "datasource_filter = [\n",
    "    \"gwas_credible_set\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]\n",
    "\n",
    "assessment, evidences, actionType, oncolabel = temporary_directionOfEffect(\n",
    "    path_n, datasource_filter\n",
    ")\n",
    "\n",
    "print(\"run temporary direction of effect\")\n",
    "\n",
    "window_spec = Window.partitionBy(\"targetId\", \"diseaseId\",'leftStudyId').orderBy( ### include gwas study\n",
    "    F.col(\"pValueExponent\").asc()\n",
    ")\n",
    "gwasCredibleAssoc = (\n",
    "    resolvedColoc.withColumn(\n",
    "        \"homogenized\", F.first(\"colocDoE\", ignorenulls=True).over(window_spec)\n",
    "    )  ## added 30.01.2025\n",
    "    .select(\"targetId\", \"diseaseId\",'leftStudyId', \"homogenized\")\n",
    "    .withColumn(\n",
    "        \"homogenized\",\n",
    "        F.when(F.col(\"homogenized\").isNull(), F.lit(\"noEvaluable\")).otherwise(\n",
    "            F.col(\"homogenized\")\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "print(\"Moving to step 2\")\n",
    "\n",
    "columns_chembl = [\"LoF_protect\", \"GoF_protect\"]\n",
    "columns_dataset = [\"LoF_protect\", \"GoF_protect\", \"LoF_risk\", \"GoF_risk\", \"evidenceDif\"]\n",
    "columns = [\"GoF_risk\", \"LoF_protect\", \"LoF_risk\", \"GoF_protect\"]\n",
    "terms = [\"noEvaluable\", \"bivalent_risk\", \"null\", \"dispar\"]\n",
    "\n",
    "taDf = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"MONDO_0045024\", \"cell proliferation disorder\", \"Oncology\"),\n",
    "        (\"EFO_0005741\", \"infectious disease\", \"Other\"),\n",
    "        (\"OTAR_0000014\", \"pregnancy or perinatal disease\", \"Other\"),\n",
    "        (\"EFO_0005932\", \"animal disease\", \"Other\"),\n",
    "        (\"MONDO_0024458\", \"disease of visual system\", \"Other\"),\n",
    "        (\"EFO_0000319\", \"cardiovascular disease\", \"Other\"),\n",
    "        (\"EFO_0009605\", \"pancreas disease\", \"Other\"),\n",
    "        (\"EFO_0010282\", \"gastrointestinal disease\", \"Other\"),\n",
    "        (\"OTAR_0000017\", \"reproductive system or breast disease\", \"Other\"),\n",
    "        (\"EFO_0010285\", \"integumentary system disease\", \"Other\"),\n",
    "        (\"EFO_0001379\", \"endocrine system disease\", \"Other\"),\n",
    "        (\"OTAR_0000010\", \"respiratory or thoracic disease\", \"Other\"),\n",
    "        (\"EFO_0009690\", \"urinary system disease\", \"Other\"),\n",
    "        (\"OTAR_0000006\", \"musculoskeletal or connective tissue disease\", \"Other\"),\n",
    "        (\"MONDO_0021205\", \"disease of ear\", \"Other\"),\n",
    "        (\"EFO_0000540\", \"immune system disease\", \"Other\"),\n",
    "        (\"EFO_0005803\", \"hematologic disease\", \"Other\"),\n",
    "        (\"EFO_0000618\", \"nervous system disease\", \"Other\"),\n",
    "        (\"MONDO_0002025\", \"psychiatric disorder\", \"Other\"),\n",
    "        (\"MONDO_0024297\", \"nutritional or metabolic disease\", \"Other\"),\n",
    "        (\"OTAR_0000018\", \"genetic, familial or congenital disease\", \"Other\"),\n",
    "        (\"OTAR_0000009\", \"injury, poisoning or other complication\", \"Other\"),\n",
    "        (\"EFO_0000651\", \"phenotype\", \"Other\"),\n",
    "        (\"EFO_0001444\", \"measurement\", \"Other\"),\n",
    "        (\"GO_0008150\", \"biological process\", \"Other\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"taId\", StringType(), True),\n",
    "            StructField(\"taLabel\", StringType(), True),\n",
    "            StructField(\"taLabelSimple\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ").withColumn(\"taRank\", F.monotonically_increasing_id())\n",
    "\n",
    "### give us a classification of Oncology VS non oncology\n",
    "wByDisease = Window.partitionBy(\"diseaseId\")  #### checked 31.05.2023\n",
    "diseaseTA = (\n",
    "    diseases.withColumn(\"taId\", F.explode(\"therapeuticAreas\"))\n",
    "    .select(F.col(\"id\").alias(\"diseaseId\"), \"taId\", \"parents\")\n",
    "    .join(taDf, on=\"taId\", how=\"left\")\n",
    "    .withColumn(\"minRank\", F.min(\"taRank\").over(wByDisease))\n",
    "    .filter(F.col(\"taRank\") == F.col(\"minRank\"))\n",
    "    .drop(\"taRank\", \"minRank\")\n",
    ")\n",
    "\n",
    "#### give us propagation of diseases and list of therapeutic areas associated\n",
    "diseases2 = diseases.select(\"id\", \"parents\").withColumn(\n",
    "    \"diseaseIdPropagated\",\n",
    "    F.explode_outer(F.concat(F.array(F.col(\"id\")), F.col(\"parents\"))),\n",
    ")\n",
    "\n",
    "chembl_trials = (\n",
    "    assessment.filter((F.col(\"datasourceId\").isin([\"chembl\"])))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .agg(F.max(F.col(\"clinicalPhase\")).alias(\"maxClinPhase\"))\n",
    ")\n",
    "\n",
    "negativeTD = (\n",
    "    evidences.filter(F.col(\"datasourceId\") == \"chembl\")\n",
    "    .select(\"targetId\", \"diseaseId\", \"studyStopReason\", \"studyStopReasonCategories\")\n",
    "    .filter(F.array_contains(F.col(\"studyStopReasonCategories\"), \"Negative\"))\n",
    "    .groupBy(\"targetId\", \"diseaseId\")\n",
    "    .count()\n",
    "    .withColumn(\"stopReason\", F.lit(\"Negative\"))\n",
    "    .drop(\"count\")\n",
    ")\n",
    "\n",
    "assessment_all = assessment.unionByName(\n",
    "    gwasCredibleAssoc.withColumn(\"datasourceId\", F.lit(\"gwas_credible_set\")),\n",
    "    allowMissingColumns=True,\n",
    ")\n",
    "\n",
    "print(\"defining non propagated,propagated and analysis_drugs functions\")\n",
    "\n",
    "def analysis_nonPropagated(assessment_all, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "def analysis_propagated(assessment_all, analysisDatasources):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter(F.col(\"datasourceId\").isin(analysisDatasources))\n",
    "        .withColumn(\n",
    "            \"datasources\",\n",
    "            F.collect_set(F.col(\"datasourceId\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .join(\n",
    "            diseases2.selectExpr(\"id as diseaseId\", \"diseaseIdPropagated\"),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumnRenamed(\"diseaseId\", \"oldDiseaseId\")\n",
    "        .withColumnRenamed(\"diseaseIdPropagated\", \"diseaseId\")\n",
    "        .groupBy(\n",
    "            \"targetId\",\n",
    "            \"diseaseId\",\n",
    "        )\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "chembl_ds = [\"chembl\"]\n",
    "\n",
    "def analysis_drugs(assessment_all, chembl_ds):\n",
    "    return discrepancifier(\n",
    "        assessment_all.filter((F.col(\"datasourceId\").isin(chembl_ds))\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "        .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "analysis_chembl = analysis_drugs(assessment_all, chembl_ds)\n",
    "\n",
    "#######\n",
    "## include here the analysis\n",
    "#######\n",
    "\n",
    "analysisDatasources = []\n",
    "\n",
    "print(\"defining full_analysis_propagation\")\n",
    "\n",
    "doe_columns=[\"LoF_protect\", \"GoF_risk\", \"LoF_risk\", \"GoF_protect\"]\n",
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "\n",
    "def full_analysis_propagation(\n",
    "    doe_columns,assessment_all, analysisDatasources, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "):\n",
    "    conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in doe_columns\n",
    "    ]\n",
    "    \n",
    "    return (\n",
    "        analysis_propagated(assessment_all, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        #### Should remove the coherencyDiagonal.isNotNull()\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        ).withColumn(\n",
    "            \"arrayN\", F.array(*[F.col(c) for c in doe_columns])\n",
    "        ).withColumn(\n",
    "            \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    "        ).withColumn(\"maxDoE_names\", F.array(*conditions)\n",
    "        ).withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\")\n",
    "        ).withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(negativeTD, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"stopReason\") == \"Negative\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        )\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "\n",
    "#####\n",
    "## no propag\n",
    "#####\n",
    "print(\"defining full analysis no propagation\")\n",
    "\n",
    "\n",
    "def full_analysis_noPropagation(\n",
    "    doe_columns,assessment_all, analysisDatasources, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "):\n",
    "    conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in doe_columns\n",
    "    ]\n",
    "    return (\n",
    "        analysis_nonPropagated(assessment_all, analysisDatasources)\n",
    "        .join(\n",
    "            analysis_chembl.selectExpr(\n",
    "                \"targetId\",\n",
    "                \"diseaseId\",\n",
    "                \"maxClinPhase\",\n",
    "                \"coherencyDiagonal as coherencyDiagonal_ch\",\n",
    "                \"coherencyOneCell as coherencyOneCell_ch\",\n",
    "                \"LoF_protect as LoF_protect_ch\",\n",
    "                \"GoF_protect as GoF_protect_ch\",\n",
    "            ),\n",
    "            on=[\"targetId\", \"diseaseId\"],\n",
    "            how=\"right\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"geneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"coherencyDiagonal\").isNotNull(), F.lit(\"hasGeneticEvidence\")\n",
    "            ).otherwise(F.lit(\"noGeneticEvidence\")),\n",
    "        )\n",
    "        # .filter(F.col(\"coherencyDiagonal_ch\").isNotNull())\n",
    "        .withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        ).withColumn(\n",
    "            \"arrayN\", F.array(*[F.col(c) for c in doe_columns])\n",
    "        ).withColumn(\n",
    "            \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    "        ).withColumn(\"maxDoE_names\", F.array(*conditions)\n",
    "        ).withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\")\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase4\",\n",
    "            F.when(F.col(\"maxClinPhase\") == 4, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=3\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 3, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=2\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 2, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"Phase>=1\",\n",
    "            F.when(F.col(\"maxClinPhase\") >= 1, F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(negativeTD, on=[\"targetId\", \"diseaseId\"], how=\"left\")\n",
    "        .withColumn(\n",
    "            \"PhaseT\",\n",
    "            F.when(F.col(\"stopReason\") == \"Negative\", F.lit(\"yes\")).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .join(\n",
    "            diseaseTA.select(\"diseaseId\", \"taLabelSimple\"), on=\"diseaseId\", how=\"left\"\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"hasGeneticEvidence\",\n",
    "            F.when(\n",
    "                F.col(\"geneticEvidence\") == \"hasGeneticEvidence\", F.lit(\"yes\")\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diagonalYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"diagonalAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"diagonalAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellYes\",\n",
    "            F.when(\n",
    "                F.col(\"hasGeneticEvidence\") == \"yes\",\n",
    "                F.when(F.col(\"oneCellAgreeWithDrugs\") == \"coherent\", F.lit(\"yes\"))\n",
    "                .when(F.col(\"oneCellAgreeWithDrugs\") == \"dispar\", F.lit(\"no\"))\n",
    "                .otherwise(F.lit(\"no\")),\n",
    "            ).otherwise(F.lit(\"no\")),\n",
    "        ).withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when(F.col(\"GoF_protect_ch\").isNotNull() & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when(F.col(\"LoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when(F.col(\"GoF_protect_ch\").isNotNull() & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        )\n",
    "        # .persist()\n",
    "    )\n",
    "\n",
    "print(\"moving to Step 3\")\n",
    "\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "c = datetime.now()\n",
    "print(\"starting dictionaries at\", c)\n",
    "\n",
    "#### continue here on 10.07.2024\n",
    "\n",
    "## 1nd dictionary\n",
    "dfs_dict = {}  ### checked and changed on 01.06.2023\n",
    "dfs_dict_propag = {}\n",
    "\n",
    "\n",
    "wocgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "]\n",
    "wCgc_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"cancer_gene_census\",\n",
    "]\n",
    "\n",
    "datasource_list = [\n",
    "    \"gene_burden\",\n",
    "    \"intogen\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "    \"WOcgc\",\n",
    "    \"wCgc\",\n",
    "    \"somatic\",\n",
    "    \"germline\",\n",
    "]\n",
    "\n",
    "germline_list = [\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"gwas_credible_set\",\n",
    "    \"impc\",\n",
    "    \"orphanet\",\n",
    "    \"gene2phenotype\",\n",
    "]\n",
    "\n",
    "somatic_list = [\"intogen\", \"cancer_gene_census\", \"eva_somatic\"]\n",
    "\n",
    "\n",
    "# assessment = prueba_assessment.filter(F.col(\"datasourceId\").isin(datasources_analysis))\n",
    "def dataset_builder(assessment_all, value, analysis_chembl, negativeTD, diseaseTA):\n",
    "    nonPropagated = full_analysis_noPropagation(\n",
    "        doe_columns,assessment_all, value, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "    )\n",
    "    propagated = full_analysis_propagation(\n",
    "        doe_columns,assessment_all, value, analysis_chembl, negativeTD, diseaseTA,diagonal_lof,diagonal_gof\n",
    "    )\n",
    "    return (\n",
    "        # Non propagation\n",
    "        ## All\n",
    "        nonPropagated,\n",
    "        ## Other\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        nonPropagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        nonPropagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "        # Propagation\n",
    "        ## All\n",
    "        propagated,\n",
    "        ## Other\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Other\"),\n",
    "        ## Other&Null\n",
    "        propagated.filter(\n",
    "            (F.col(\"taLabelSimple\").isNull()) | (F.col(\"taLabelSimple\") == \"Other\")\n",
    "        ),\n",
    "        ## Oncology\n",
    "        propagated.filter(F.col(\"taLabelSimple\") == \"Oncology\"),\n",
    "    )\n",
    "\n",
    "\n",
    "for value in datasource_list:\n",
    "    print(value)\n",
    "    if value == \"WOcgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, wocgc_list, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "    elif value == \"wCgc\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, wCgc_list, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "    elif value == \"germline\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all,\n",
    "            germline_list,\n",
    "            analysis_chembl,\n",
    "            negativeTD,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    elif value == \"somatic\":\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"],\n",
    "        ) = dataset_builder(\n",
    "            assessment_all,\n",
    "            somatic_list,\n",
    "            analysis_chembl,\n",
    "            negativeTD,\n",
    "            diseaseTA,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        (\n",
    "            dfs_dict[f\"df_{value}_All_original\"],\n",
    "            dfs_dict[f\"df_{value}_Other_original\"],\n",
    "            dfs_dict[f\"df_{value}_OtherNull_original\"],\n",
    "            dfs_dict[f\"df_{value}_Oncology_original\"],\n",
    "            dfs_dict_propag[f\"df_{value}_All_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Other_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_OtherNull_propag\"],\n",
    "            dfs_dict_propag[f\"df_{value}_Oncology_propag\"]\n",
    "        ) = dataset_builder(\n",
    "            assessment_all, value, analysis_chembl, negativeTD, diseaseTA\n",
    "        )\n",
    "\n",
    "\n",
    "def comparisons_df() -> list:\n",
    "    \"\"\"Return list of all comparisons to be used in the analysis\"\"\"\n",
    "    comparisons = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"hasGeneticEvidence\", \"byDatatype\"),\n",
    "            (\"diagonalYes\", \"byDatatype\"),\n",
    "            (\"oneCellYes\", \"byDatatype\"),\n",
    "            (\"NdiagonalYes\", \"byDatatype\"),\n",
    "            (\"NoneCellYes\", \"byDatatype\"),\n",
    "        ],\n",
    "        schema=StructType(\n",
    "            [\n",
    "                StructField(\"comparison\", StringType(), True),\n",
    "                StructField(\"comparisonType\", StringType(), True),\n",
    "            ]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    predictions = spark.createDataFrame(\n",
    "        data=[\n",
    "            (\"Phase4\", \"clinical\"),\n",
    "            (\"Phase>=3\", \"clinical\"),\n",
    "            (\"Phase>=2\", \"clinical\"),\n",
    "            (\"Phase>=1\", \"clinical\"),\n",
    "            (\"PhaseT\", \"clinical\"),\n",
    "        ]\n",
    "    )\n",
    "    return comparisons.join(predictions, how=\"full\").collect()\n",
    "\n",
    "\n",
    "\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "results = []\n",
    "\n",
    "\n",
    "def aggregations_original(\n",
    "    df,\n",
    "    data,\n",
    "    listado,\n",
    "    comparisonColumn,\n",
    "    comparisonType,\n",
    "    predictionColumn,\n",
    "    predictionType,\n",
    "    today_date,\n",
    "):\n",
    "\n",
    "    wComparison = Window.partitionBy(comparisonColumn)\n",
    "    wPrediction = Window.partitionBy(predictionColumn)\n",
    "    wPredictionComparison = Window.partitionBy(comparisonColumn, predictionColumn)\n",
    "\n",
    "    uniqIds = df.select(\"targetId\", \"diseaseId\").distinct().count()\n",
    "\n",
    "    out = (\n",
    "        df.withColumn(\"comparisonType\", F.lit(comparisonType))\n",
    "        .withColumn(\"predictionType\", F.lit(predictionType))\n",
    "        .withColumn(\"total\", F.lit(uniqIds))\n",
    "        .withColumn(\"a\", F.count(\"targetId\").over(wPredictionComparison))\n",
    "        .withColumn(\n",
    "            \"predictionTotal\",\n",
    "            F.count(\"targetId\").over(wPrediction),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"comparisonTotal\",\n",
    "            F.count(\"targetId\").over(wComparison),\n",
    "        )\n",
    "        .select(\n",
    "            F.col(predictionColumn).alias(\"prediction\"),\n",
    "            F.col(comparisonColumn).alias(\"comparison\"),\n",
    "            \"comparisonType\",\n",
    "            \"predictionType\",\n",
    "            \"a\",\n",
    "            \"predictionTotal\",\n",
    "            \"comparisonTotal\",\n",
    "            \"total\",\n",
    "        )\n",
    "        .filter(F.col(\"prediction\").isNotNull())\n",
    "        .filter(F.col(\"comparison\").isNotNull())\n",
    "        .distinct()\n",
    "    )\n",
    "\n",
    "    out.write.mode(\"overwrite\").parquet(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    filePath = \"gs://ot-team/jroldan/\" + str(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    listado.append(\n",
    "        \"gs://ot-team/jroldan/\"\n",
    "        + str(\n",
    "            today_date\n",
    "            + \"_\"\n",
    "            + \"analysis/\"\n",
    "            + data\n",
    "            # + \"_propagated\"\n",
    "            + \"/\"\n",
    "            + comparisonColumn\n",
    "            + \"_\"\n",
    "            + predictionColumn\n",
    "            + \".parquet\"\n",
    "        )\n",
    "    )\n",
    "    print(\n",
    "        today_date\n",
    "        + \"_\"\n",
    "        + \"analysis/\"\n",
    "        + data\n",
    "        # + \"_propagated\"\n",
    "        + \"/\"\n",
    "        + comparisonColumn\n",
    "        + \"_\"\n",
    "        + predictionColumn\n",
    "        + \".parquet\"\n",
    "    )\n",
    "    c = datetime.now()\n",
    "    c.strftime(\"%H:%M:%S\")\n",
    "    print(c)\n",
    "\n",
    "    array1 = np.delete(\n",
    "        out.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            data,\n",
    "            comparisonColumn,\n",
    "            predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            filePath,\n",
    "        ]\n",
    "    )\n",
    "    return results\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start doing aggregations and writing\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting with non-propagated aggregations at 2025-04-24 06:21:33.611622\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-04-24_analysis/df_gene_burden_All_original/hasGeneticEvidence_Phase4.parquet\n",
      "2025-04-24 06:21:42.256155\n",
      "df_gene_burden_All_original df unpersisted\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-04-24_analysis/df_gene_burden_Other_original/hasGeneticEvidence_Phase4.parquet\n",
      "2025-04-24 06:21:53.047763\n",
      "df_gene_burden_Other_original df unpersisted\n",
      "non propagated files wroten succesfully at 2025-04-24 06:21:33.611622\n",
      "starting with propagated aggregations at 2025-04-24 06:21:33.611622\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-04-24_analysis/df_gene_burden_All_propag/hasGeneticEvidence_Phase4.parquet\n",
      "2025-04-24 06:22:04.494690\n",
      "df_gene_burden_All_propag df unpersisted\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-04-24_analysis/df_gene_burden_Other_propag/hasGeneticEvidence_Phase4.parquet\n",
      "2025-04-24 06:22:14.903707\n",
      "df_gene_burden_Other_propag df unpersisted\n",
      "propagated files wroten succesfully at 2025-04-24 06:21:33.611622\n",
      "creating pandas dataframe with resulting rows\n",
      "created pandas dataframe\n",
      "converting to spark dataframe\n",
      "preparing dataframe\n",
      "read pattern variables\n",
      "importing functions\n",
      "imported functions\n",
      "processed spreadsheet\n",
      "writting the dataframe\n",
      "dataframe written \n",
      " Analysis finished\n"
     ]
    }
   ],
   "source": [
    "c = datetime.now()\n",
    "print(\"start doing aggregations and writing\")\n",
    "today_date = str(date.today())\n",
    "aggSetups_original = comparisons_df()\n",
    "listado = []\n",
    "results = []\n",
    "\n",
    "print(\"starting with non-propagated aggregations at\", c)\n",
    "for key, df in islice(dfs_dict.items(),2):\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original[:1]:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"non propagated files wroten succesfully at\", c)\n",
    "\n",
    "\n",
    "print(\"starting with propagated aggregations at\", c)\n",
    "for key, df in islice(dfs_dict_propag.items(),2):\n",
    "    df = df.persist()\n",
    "    for row in aggSetups_original[:1]:\n",
    "        aggregations_original(df, key, listado, *row, today_date)\n",
    "    df.unpersist()\n",
    "    print(key + \" df unpersisted\")\n",
    "\n",
    "print(\"propagated files wroten succesfully at\", c)\n",
    "\n",
    "print(\"creating pandas dataframe with resulting rows\")\n",
    "df_results = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"group\",\n",
    "        \"comparison\",\n",
    "        \"phase\",\n",
    "        \"OR\",\n",
    "        \"pValue\",\n",
    "        \"LowCI\",\n",
    "        \"HighCI\",\n",
    "        \"total\",\n",
    "        \"array\",\n",
    "        \"rs\",\n",
    "        \"lowRs\",\n",
    "        \"HighRs\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")\n",
    "print(\"created pandas dataframe\")\n",
    "print(\"converting to spark dataframe\")\n",
    "print(\"preparing dataframe\")\n",
    "\n",
    "schema = StructType(\n",
    "    [\n",
    "        StructField(\"group\", StringType(), True),\n",
    "        StructField(\"comparison\", StringType(), True),\n",
    "        StructField(\"phase\", StringType(), True),\n",
    "        StructField(\"oddsRatio\", DoubleType(), True),\n",
    "        StructField(\"pValue\", DoubleType(), True),\n",
    "        StructField(\"lowerInterval\", DoubleType(), True),\n",
    "        StructField(\"upperInterval\", DoubleType(), True),\n",
    "        StructField(\"total\", StringType(), True),\n",
    "        StructField(\"values\", ArrayType(ArrayType(IntegerType())), True),\n",
    "        StructField(\"relSuccess\", DoubleType(), True),\n",
    "        StructField(\"rsLower\", DoubleType(), True),\n",
    "        StructField(\"rsUpper\", DoubleType(), True),\n",
    "        StructField(\"path\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "\n",
    "print(\"read pattern variables\")\n",
    "df = spreadSheetFormatter(spark.createDataFrame(df_results, schema=schema))\n",
    "print(\"processed spreadsheet\")\n",
    "print(\"writting the dataframe\")\n",
    "\n",
    "# Convert list of lists to DataFrame\n",
    "# Regular expressions\n",
    "'''\n",
    "value_pattern = r\"df_([^_]+)_\"  # Extracts {value}\n",
    "middle_pattern = r\"df_[^_]+_([^_]+)_\"  # Extracts middle part (All, Other, etc.)\n",
    "suffix_pattern = r\"(original|propag)$\"  # Extracts suffix (original or propag)\n",
    "'''\n",
    "\n",
    "df.withColumn(\n",
    "    \"datasource\",\n",
    "    F.regexp_extract(F.col(\"group\"), r\"df_(.*?)_(All|Other|OtherNull|Oncology)_(propag|original)\", 1)\n",
    ").withColumn(\n",
    "    \"therArea\",\n",
    "    F.regexp_extract(F.col(\"group\"), r\"_(All|Other|OtherNull|Oncology)_\", 1)\n",
    ").withColumn(\n",
    "    \"type\",\n",
    "    F.regexp_extract(F.col(\"group\"), r\"_(propag|original)$\", 1)\n",
    ").toPandas().to_csv(\n",
    "    f\"gs://ot-team/jroldan/analysis/{today_date}_genEvidAnalysis_new.csv\"\n",
    ")\n",
    "\n",
    "print(\"dataframe written \\n Analysis finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spark session created at 2025-04-24 09:20:11.034967\n",
      "Analysis started on 2025-04-24 at  2025-04-24 09:20:11.034967\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/04/24 09:20:16 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from itertools import islice\n",
    "import time\n",
    "#from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "# from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "# from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "#from itertools import islice\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    StringType,\n",
    "    IntegerType,\n",
    "    ArrayType\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "spark.conf.set(\n",
    "    \"spark.sql.shuffle.partitions\", \"400\"\n",
    ") \n",
    "\n",
    "from google.cloud import storage\n",
    "\n",
    "bucket_name = \"ot-team\"\n",
    "prefix = \"jroldan/2025-04-24_analysis/\"\n",
    "\n",
    "client = storage.Client()\n",
    "bucket = client.get_bucket(bucket_name)\n",
    "blobs = bucket.list_blobs(prefix=prefix)\n",
    "\n",
    "# List all .parquet files with full GCS paths\n",
    "parquet_files = [\n",
    "    f\"gs://{bucket_name}/{blob.name}\"\n",
    "    for blob in blobs\n",
    "    if blob.name.endswith(\".parquet\")\n",
    "]\n",
    "\n",
    "print(f\"Found {len(parquet_files)} parquet files\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1175 parquet files\n"
     ]
    }
   ],
   "source": [
    "from google.cloud import storage\n",
    "\n",
    "bucket_name = \"ot-team\"\n",
    "prefix = \"jroldan/2025-04-24_analysis/\"\n",
    "\n",
    "client = storage.Client()\n",
    "bucket = client.get_bucket(bucket_name)\n",
    "blobs = bucket.list_blobs(prefix=prefix)\n",
    "\n",
    "# List all .parquet files with full GCS paths\n",
    "parquet_files = [\n",
    "    f\"gs://{bucket_name}/{blob.name}\"\n",
    "    for blob in blobs\n",
    "    if blob.name.endswith(\".parquet\")\n",
    "]\n",
    "\n",
    "print(f\"Found {len(parquet_files)} parquet files\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NdiagonalYes_Phase4.parquet/part-00000-b3d4c9e7-1032-4b19-ab35-d2d5122cbc7f-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-2a4fb3d9-1b5a-4f12-8c1f-5ff8f1c5b0d6-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-3072ccae-482c-41f2-830e-5bb3e55c8137-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-71d6a4c7-d2e7-4450-99a4-43ec3b22efba-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NdiagonalYes_PhaseT.parquet/part-00000-1dd8ef29-5fe0-4472-8ce8-078fe3b5f14d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NoneCellYes_Phase4.parquet/part-00000-f7a92351-ff0c-454e-a90d-d7f1c6318e41-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NoneCellYes_Phase>=1.parquet/part-00000-c919d8a4-7a40-45aa-983c-274974008893-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NoneCellYes_Phase>=2.parquet/part-00000-7e3859b2-e2c2-4d0f-a9d4-8d24503db474-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NoneCellYes_Phase>=3.parquet/part-00000-05892316-8dce-4ee8-b3f0-c69a9729ea5a-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/NoneCellYes_PhaseT.parquet/part-00000-24969534-278a-486f-9547-d9eb77d5811f-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/diagonalYes_Phase4.parquet/part-00000-28f2f74e-b973-449b-b314-ff539e56f492-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/diagonalYes_Phase>=1.parquet/part-00000-62b9beb6-9416-4cfc-bdfe-13c53e5e6e12-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/diagonalYes_Phase>=2.parquet/part-00000-a9344c38-5a7e-4f51-b1de-6b4b2d49f000-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/diagonalYes_Phase>=3.parquet/part-00000-f84a1be6-36ee-4d49-8573-9445268b90be-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/diagonalYes_PhaseT.parquet/part-00000-a45f1bb8-a6e3-4da9-a89c-d682958884f3-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-b471b29c-cbcb-49e7-866a-30aab488b89c-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-e700dc50-4e98-4bc3-a713-b811a6b03eb3-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-2864245e-60b5-408e-9d51-63a26098317c-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-97e6ca28-1e02-4dc4-ab88-65dc025c3479-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-1887c0f4-00e3-40fd-9411-8b7936f05b87-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/oneCellYes_Phase4.parquet/part-00000-f8fec66c-dba2-45fd-b793-e1a382439340-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/oneCellYes_Phase>=1.parquet/part-00000-1fe5581c-312a-4c9b-8e88-627d8a2feeaf-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/oneCellYes_Phase>=2.parquet/part-00000-51911082-e763-43b0-9664-779861b3c5c8-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/oneCellYes_Phase>=3.parquet/part-00000-bc3c48fc-6bb4-4dc7-ba75-1ccb44126f6e-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_All_original/oneCellYes_PhaseT.parquet/part-00000-005efbfb-5ca9-4910-9a53-82892820511d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-8c081ba5-2b77-4d16-be99-ca31e546574a-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-ed214384-cd75-49a7-bf02-98716daead86-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-13bf6a39-0333-4a6f-809a-d42ab80b22ec-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-1e416603-2e17-46c1-8876-575f618a3676-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-8a667edb-c0f9-4239-9094-6d8c0bac3398-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-d941d318-5d68-45ca-9285-4f3a22d0a774-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-4ad498e4-f06b-4c8c-8932-4646a91c6160-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-7d83ce85-0308-49d2-b6fa-2c29e6783568-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-06003869-c7a1-4531-a497-7a88f60dac84-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-86b11a8b-cef5-4019-90bd-fa760ca128f3-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/diagonalYes_Phase4.parquet/part-00000-26fa490a-fd13-4626-80c9-494c6334b3c9-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-3e0ccc4b-750b-4ef2-ad36-72e283d0f040-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-011f18ed-85b2-4cbe-803d-1b1e3a9c0964-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-b59fad06-103a-4cbc-8e3e-3bbab0e1674e-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-06f6c995-ce57-4fbe-8fe5-bb1ffcd8e254-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-a03e48c6-e96d-4720-acf4-27a276677f0f-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-5900e811-e961-42ba-aec4-95a66cf30518-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-81c3f44a-1a2c-41ac-bb35-421d0392fb8a-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-bfff2e94-f206-49af-8464-97e484da3dcd-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-af5bc255-2a83-4de6-b055-290e341a7a72-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/oneCellYes_Phase4.parquet/part-00000-31b931ed-2585-4286-b860-92e0e211fa44-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-13e13faa-a409-40f2-8d54-11ce1e26ba1d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-c6eb7e9a-db50-4808-83d1-73310afe2b82-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-85e5ea1c-2efa-48fb-b6e3-89d57a932dad-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-854f231d-ba3e-4107-9efc-2d11963abe14-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-36183497-442c-4997-95ff-62c893e5ce22-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-14e13fb3-d7ce-48ff-bd01-2931476e4cbc-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-286e1661-49ec-4d41-81f9-5b03d03fbf0c-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-4619ff6a-4ce5-4c8c-8cd2-89af244f2007-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-f890ca60-2a3e-40f6-8093-2a9d68be6620-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-0028b5b2-ad21-46cf-ae09-947f34fdd2e6-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-c664b7de-c611-4e34-aa53-505b472e2997-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-f36f48d7-6373-4cd8-9661-654a992cb846-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-646fefee-96af-45d7-ad21-b0e09dc24eb1-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-1bd2bf09-3008-42b3-b49a-70004131b2b3-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-0e63c0f6-92af-466e-b101-da90e6f0c14d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-f5a6ecf2-8652-4969-b4cc-eed7b46024ed-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-3acc3b0a-566c-4553-9ea3-27a7629a24c5-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-d061a63f-b5d6-45dd-a48c-e90b232ab4ee-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-bd7039f3-fef2-4622-b717-ad4acb0c2f1f-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-27d74f4e-3935-4399-8ec2-6aa41993e11d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-7fd8a25c-f0a3-416e-b115-9f301e03412d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-4848b4a2-56c4-4e62-a8a4-f8d60b101d19-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-9d1c1766-533a-4fcb-9726-52bdeabca69a-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-a06acb2c-f08a-4b17-9c14-ff1c59c4452c-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-e9801ba4-5437-432b-8911-6aed9f4f1f0f-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-76f614ab-db3c-4c46-b934-16c4d1592ec0-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-7bb886c2-38d5-43f8-ae01-3fccaab58b4f-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-3434b723-473b-4f63-a34c-dbcf0910cfae-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-c9c55c90-c6a1-434b-9015-3a0bb12b6384-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NdiagonalYes_Phase4.parquet/part-00000-e74a1a7a-6a8e-4cc3-9c6b-a378812081d1-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-5d37f6d4-2ec3-4511-a17c-b1e0ed582533-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-464ded76-d790-44de-bef4-926d298bb51d-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-c02385a9-c131-485f-b8a8-1b7f51d02695-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-0a448085-b491-4786-8c3c-44037a3c3cb9-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NoneCellYes_Phase4.parquet/part-00000-029d3520-92a2-4fc6-8b6b-b049fa0a4d14-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-0887ae79-401a-43f7-b3fe-814a5ce4d3b9-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-9b6aeb9f-4a37-4169-ad45-cd69fc81a8f0-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-bd0f0bd9-11cf-4783-9a7d-7e63aef66fcb-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/NoneCellYes_PhaseT.parquet/part-00000-56969dc1-4d8b-4f5f-9c3d-9d2f01f233ab-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/diagonalYes_Phase4.parquet/part-00000-0409fc9a-0809-456b-8437-4261a1d562b5-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/diagonalYes_Phase>=1.parquet/part-00000-a28decd8-d9ad-4585-b3e6-d4e0d5f115c2-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/diagonalYes_Phase>=2.parquet/part-00000-b80ded72-efdd-4bbb-9dde-59bea89c82f3-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/diagonalYes_Phase>=3.parquet/part-00000-cdaafb54-c558-4d1b-a6a0-9c3e5debe139-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/diagonalYes_PhaseT.parquet/part-00000-281c092e-e8a0-452b-b75f-7f97c25934db-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-b046aa96-4674-4604-8fa4-e7671fc88096-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-6ea00007-b76c-40e8-aa7a-c59e430cc7f1-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-4fb8182e-4a52-45cc-906a-96e99ecd9ca8-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-db88d04b-5767-4fc8-ab84-a306c7cf9fcd-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-0585712a-fd23-48af-b3d0-f46d376ee3ce-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/oneCellYes_Phase4.parquet/part-00000-d256d359-532a-4b55-ad36-2acb56ebd611-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/oneCellYes_Phase>=1.parquet/part-00000-1c07cf0a-b539-4955-b4ef-5caccdf6e015-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/oneCellYes_Phase>=2.parquet/part-00000-3404855e-3883-49d3-9fdb-f4d556299a75-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/oneCellYes_Phase>=3.parquet/part-00000-7c9f4620-d544-4afa-bf17-71cf60ebe8ad-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_WOcgc_Other_original/oneCellYes_PhaseT.parquet/part-00000-72314556-13fa-45c7-bf9b-8e54a08b8353-c000.snappy.parquet\n",
      "Part 1: df_WOcgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NdiagonalYes_Phase4.parquet/part-00000-4c846f54-c030-4390-8b81-b8d4bc0d69b5-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-41e4bd28-c267-4f8a-bd44-81ec8d0ac808-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-4abfad4e-b765-49a5-93db-4120ca0fdbdf-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-fece8a44-e522-4b54-bce9-af0320983df7-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NdiagonalYes_PhaseT.parquet/part-00000-829136d5-4857-45fd-ba47-7082c3cd5f7e-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NoneCellYes_Phase4.parquet/part-00000-993b4756-f0a2-413c-b013-6ba9005a5dd6-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NoneCellYes_Phase>=1.parquet/part-00000-eec00d59-217c-467c-b186-0edc182ca3d8-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NoneCellYes_Phase>=2.parquet/part-00000-1e646825-b2f3-44ce-adaf-b369a5861049-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NoneCellYes_Phase>=3.parquet/part-00000-ea9fd787-6bad-4f35-9f77-085a6d654c92-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/NoneCellYes_PhaseT.parquet/part-00000-8242115d-d195-4dc2-b28d-76f77841c3c3-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/diagonalYes_Phase4.parquet/part-00000-d435eefd-42ce-4ab5-8bbe-d784486e8e0d-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/diagonalYes_Phase>=1.parquet/part-00000-21c17b8b-1c8d-4838-867c-134d1ecae3f9-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/diagonalYes_Phase>=2.parquet/part-00000-13bd365b-321f-4150-b6f1-ba58be2dfcb1-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/diagonalYes_Phase>=3.parquet/part-00000-50a2bf77-b671-4fe6-a2c5-247700f45eef-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/diagonalYes_PhaseT.parquet/part-00000-d1839b94-fb27-4079-aab4-0c91d8ed49af-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-73b0af59-71b0-4274-8d7d-74f1d8863419-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-dee03968-cbd9-4b8f-ac01-a6696355a748-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-b10a8053-e402-4747-b550-222db3a2a197-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-6f63d6ec-4973-417b-8d4c-d87669b7827d-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-405fd721-4f5b-4399-ba9a-57847b60db4d-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/oneCellYes_Phase4.parquet/part-00000-0bb8bc33-361d-47b7-b799-4a2e20d6f35a-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/oneCellYes_Phase>=1.parquet/part-00000-a6449420-fe7b-4f29-b220-ffadee0ab7b4-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/oneCellYes_Phase>=2.parquet/part-00000-6451919f-23ca-445d-9d77-fd41a6746ccf-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/oneCellYes_Phase>=3.parquet/part-00000-9d6dba29-e6f1-45a5-98f4-32dc91208c47-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_All_original/oneCellYes_PhaseT.parquet/part-00000-3146c022-da0f-4913-8136-2281f1dca67f-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-026c2b18-1b84-4b2e-9c5b-9405757f3806-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-0e079edd-15ba-40bd-9fc6-a99691698337-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-1fb0bd6d-05c5-4336-9798-8d031444ea38-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-acbf074e-0c88-46c8-b71e-bb78d0329f96-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-7db4606a-49f0-4e30-9fda-f696a3a79ba4-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-08f41f40-c4a3-4db1-9e4c-86d7e24b7ac9-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-71559ab5-cb19-43bd-aad0-fa33d14fea27-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-9fe8faf0-3054-4cc0-a577-9d68df2c8332-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-cdaee51b-fa19-4393-a8c6-2a44d6164736-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-ef40b00d-d89b-4018-a4f9-130d1c105a1a-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/diagonalYes_Phase4.parquet/part-00000-54b76440-3103-469a-b561-1da10ae7f3f3-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-48126f3f-1870-4592-8a98-0807b7bb67e8-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-790c3ce7-75da-466a-bb05-a523fb5a7c46-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-774b4301-86be-4659-a36e-9208668d0c33-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-ddf92d9f-0456-4710-8a88-4cf7b7f2a683-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-146b0464-b076-4c93-98f1-8886054bfe2d-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-88a61611-b841-4d98-8a18-6a12d795c799-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-e67d6477-5a4c-4cb8-92b6-8463e69b74db-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-69460e7c-1082-42e8-a01c-1c414f6212d6-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-a7f30dd2-1b18-4b36-b6ea-5a52eb10c1c6-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/oneCellYes_Phase4.parquet/part-00000-91e4789b-bd42-499e-8539-826ba9aaa494-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-a71498d5-ff5d-4333-a879-4ebf14e6e35e-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-919587ef-43fa-49e0-95c5-43152958cadb-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-1900999c-30d7-4e0c-b887-c3cbe0ab70d7-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-2cfd9e75-5626-43b4-963f-aaafe0a9b8f6-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-06ef4110-3f7c-4ba3-8473-9d45f6a19ece-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-e3a682a5-8d61-4d5c-b78d-504cc072dacc-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-bfd5f0fd-03a2-4f0d-a53f-f37fe3bb7a16-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-b08f2249-6190-4d08-8efa-95e5a90a64b9-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-9e624f6d-8b0d-4abf-a894-9212249a85a6-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-966b1975-e7bc-4e71-9201-a30cf4c4cf88-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-f1e4418e-9ae7-45c1-8ea7-db0604ccd911-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-aecae315-2738-4e78-a89e-323a25653c77-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-84baf34e-91c7-4d57-bf5c-790779644966-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-6c8887d1-1cc8-4b10-9d4b-7e566a9bca50-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-8003e0e9-1fd6-476c-aa3a-f53fc2c231d8-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-80e62853-7cc3-4b01-8986-140ca910625b-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-d96c8679-1268-4985-90e5-3326dd0ed94d-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-b794e0d8-5ad9-46d3-9752-182174866253-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-5a3b4e84-9ede-4239-9ed8-be6a005a0199-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-a1668d19-d9ec-4b89-8f66-6e7db51f340d-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-1c205439-bc6c-4d8b-9b81-40804b0b52c1-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-c0905b65-b03e-4bc0-940a-531b944e1bfb-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-22c7caea-53a3-49f7-8f78-685dc75be05f-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-2e652ab9-fa61-45d9-82d7-cc1ff08e3379-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-d0fbd5dd-e127-49cd-9409-b692f357e353-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-5b32a393-3f61-4552-82fd-d6c3295caaba-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-f349c535-ed50-4ec2-84dd-675d9b3dc81b-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-06585aa5-bbec-425d-a3f2-6541868115ad-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-fa70e6a0-1e16-4bda-971c-db6414a9a2c7-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NdiagonalYes_Phase4.parquet/part-00000-5b2016f5-7acb-4622-819a-d2bcad395aed-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-c33e508a-494a-40d7-a222-abad3d359d1f-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-5c8cf342-442d-4d7c-9f2d-585282fb031f-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-4850be1d-cf3f-4ce6-a940-5d62c01f7eb4-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-6f937263-49ea-4a1a-9c59-c4c3570ce507-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NoneCellYes_Phase4.parquet/part-00000-ba97a420-3b03-4067-b06a-6eb4ec27d82f-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-e1269647-540a-4872-b069-803b3a018e11-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-cd3e006a-9056-46af-9fed-bc8afa1f19fb-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-26ea4c12-d88b-404a-9f44-1c7fd585ef4f-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/NoneCellYes_PhaseT.parquet/part-00000-8860fe99-81c8-4ab6-b55a-cd3ba0c12809-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/diagonalYes_Phase4.parquet/part-00000-75efd9c6-b558-42bf-9aa4-ce6e1056a186-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/diagonalYes_Phase>=1.parquet/part-00000-99ff8610-7726-42a1-a39d-831cc50210a9-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/diagonalYes_Phase>=2.parquet/part-00000-cdc75745-a588-4dfd-aa43-3964a9e39ad9-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/diagonalYes_Phase>=3.parquet/part-00000-8358bc2c-27c4-4d3d-9887-99eaf4bcfa53-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/diagonalYes_PhaseT.parquet/part-00000-15955528-70ad-4cab-8628-6cbc1a86561a-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-5e5f7606-a34f-482a-a53f-69cf447f87ca-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-5f81bdc0-4ad6-4a60-a125-b61e8f36babc-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-04454075-805d-49ac-8f2e-79053f1d9908-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-0abd7dc3-ea7e-4173-85e2-45c5aea0c384-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-347f7e2b-1989-4743-8cf5-2e5024558366-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/oneCellYes_Phase4.parquet/part-00000-f136dce0-ff20-484d-8767-f1ffa2f086da-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/oneCellYes_Phase>=1.parquet/part-00000-62244a69-89fc-4098-8e87-3f1d14007a7a-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/oneCellYes_Phase>=2.parquet/part-00000-fc349cae-0a0c-42eb-b5d8-b0f3a3d453d3-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/oneCellYes_Phase>=3.parquet/part-00000-41cb4d06-ad85-4cd1-8992-fe5ed771d6af-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_cancer_gene_census_Other_original/oneCellYes_PhaseT.parquet/part-00000-098ea0b5-3e15-4987-9d69-8f8284d09ed6-c000.snappy.parquet\n",
      "Part 1: df_cancer_gene_census_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NdiagonalYes_Phase4.parquet/part-00000-ce8e7404-7a42-4ba5-8a9a-71cffa02f456-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-ffb0156c-20ff-4d8d-b5f9-7ab382e9db77-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-393e45c9-d31d-4cdb-bcbb-77891e286b9c-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-923246aa-99ae-4ce1-a222-aea337dbd652-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NdiagonalYes_PhaseT.parquet/part-00000-a880a514-9163-4e01-b243-372fa43ba346-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NoneCellYes_Phase4.parquet/part-00000-27412b88-a35a-4043-a09f-0bb9eb37802c-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NoneCellYes_Phase>=1.parquet/part-00000-02491ba4-26b2-4faa-b4f7-95fe4acd4032-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NoneCellYes_Phase>=2.parquet/part-00000-5a779d10-fdd8-465f-bff9-8f8a883fcdeb-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NoneCellYes_Phase>=3.parquet/part-00000-2fe4de72-b513-4566-9179-5250b268dbbb-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/NoneCellYes_PhaseT.parquet/part-00000-403779a0-d8e4-4b74-b926-4addca91e914-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/diagonalYes_Phase4.parquet/part-00000-8ac5e33f-67e7-4d07-ab65-8df56942c753-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/diagonalYes_Phase>=1.parquet/part-00000-bd0f708e-8dce-477e-b500-0d8e5af5b18a-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/diagonalYes_Phase>=2.parquet/part-00000-c3fbf63d-4bd0-447f-884a-b38316dfd56e-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/diagonalYes_Phase>=3.parquet/part-00000-314d4583-2e05-4ad5-b220-21a664e4d21e-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/diagonalYes_PhaseT.parquet/part-00000-8c52dc9d-7264-4190-bf2a-4e3d0a1236e4-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-bfbd58a9-d3f9-4806-8463-3b281b3eb292-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-6d24a5cf-fbe5-4512-9a9b-e2c252c9050d-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-52526dc4-8c1b-45db-b17f-19c3bcd0c809-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-d6034d00-05b0-4393-b7bc-441de2728d1a-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-44e70ea6-213c-470a-98ab-af5af76229ab-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/oneCellYes_Phase4.parquet/part-00000-476b3521-1a74-4bd2-819b-29a16102e095-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/oneCellYes_Phase>=1.parquet/part-00000-80357351-3749-46fd-a9b8-2b80fd0901a6-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/oneCellYes_Phase>=2.parquet/part-00000-13f51312-469c-48da-91fa-0ad0ca67c48e-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/oneCellYes_Phase>=3.parquet/part-00000-c0642cf6-4ca2-4cdd-bd4b-fab2d0ecd6d2-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_All_original/oneCellYes_PhaseT.parquet/part-00000-308151a3-5733-4c9e-858e-e23af13bd2cd-c000.snappy.parquet\n",
      "Part 1: df_eva_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-8dad7b08-c607-408e-94f2-3a819cd9b512-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-02e59527-3f8a-41ef-bbcb-0e39edf9ff76-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-41b34416-0912-4e02-97ed-178cdc48cde5-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-27174a12-df3a-4935-a4da-66f4c2685ed7-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-f7d174cf-9d27-45b2-9721-de77012b88c7-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-71b3bb32-521e-40ca-8eb8-c27f42d8c8c5-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-fcfd7b6d-e3ce-4293-8046-58f1ba8cad5d-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-043027d4-75f7-47c0-b6b1-083b4d98f407-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-e2e87349-588d-412a-bdb1-b9ffe01aabc6-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-20b83afa-ff4e-4175-8c9f-803ee66d506d-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/diagonalYes_Phase4.parquet/part-00000-6c8392ae-47ba-4141-af51-59c473039a14-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-5acc27c8-8813-465e-9618-b14ca63d69ce-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-d0ca3adf-2559-4da7-b181-560a260f344e-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-5e4b62b8-2247-420b-b44e-7681dd941dd0-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-50765740-9428-40e2-b42f-2bc712a18973-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-88e6b1ec-b6b8-41fc-8b06-6611abf88bb1-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-8013adf7-f8e1-43e2-88e4-1ef7ee53f3b4-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-4be2644e-f6db-4810-8a8d-7d5643584489-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-60bf83b1-430c-40d0-8d1c-c3516b850ec7-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-c8ae385f-426c-412c-b43f-3566a30a83b6-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/oneCellYes_Phase4.parquet/part-00000-4869497f-87bb-4f61-94bb-7cf264e39ee7-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-7c51f419-0dac-475d-8865-c508af1335ea-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-08a010f1-42a1-4306-a9c1-c5aa9ebb1d11-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-61bcf3bd-4475-4407-bd81-6eaecebf5718-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-c4be0a5b-14b9-49fc-8c97-aa0d27e3ac59-c000.snappy.parquet\n",
      "Part 1: df_eva_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-2b1f2120-6ddc-4f97-8ff6-84a54dadbe91-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-d2ee89a7-2180-46b5-8e2d-f3e1c4a3429e-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-43168a3d-7c3f-48e8-afb0-d882fa1f70c4-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-0423cc75-5ff2-44b2-8906-86d09129643c-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-6fcde4a5-61a4-4ed3-adff-e50f77da41d9-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-f00e8bd1-1888-4220-bbbf-9d6d983b3485-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-e8941ea4-ac57-4203-b182-d1c5d7cf295c-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-4d31ebb1-9e7c-4ffa-b1ca-b38e51f873b7-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-adc1fe57-6e31-4c10-ab5e-aca1d7109f4c-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-b4b90456-67a7-44f9-9f10-854a6186cd4b-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-76fb7d2b-062c-41dd-a0e0-2ca450d30df1-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-b4194014-7c6d-4ca6-a12d-312d7f323a9b-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-af4ac177-eb48-4328-b4e5-463e024c45e8-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-5f63099d-a4d7-49f4-9113-4b3bfc3eb575-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-9b681665-23d2-4d87-91b0-1da1f30884b2-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-6820f226-0bab-4c10-a9bd-c7d2504a86e8-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-abe29f0c-090a-40aa-ab6a-584ae27941e6-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-1464d931-3d72-49ca-b58b-91f434cb4ebb-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-0095c114-4b49-46b4-b206-77b0a1c218d2-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-1f9ed16d-ad1d-4490-ad54-788ed6fcdbfc-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-6edc79df-64d7-40c5-9783-cb1b6038e04d-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-1804c618-4695-420c-98a6-42986df8bbe7-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-2624ebc7-9426-4580-a7b1-0c95f25bcae4-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-feb800b1-a5df-47d6-8830-fc5fe980b36a-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-b89aac78-1095-4d58-bb26-a0efbf98b69e-c000.snappy.parquet\n",
      "Part 1: df_eva_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NdiagonalYes_Phase4.parquet/part-00000-8e4e034d-810c-445a-af29-2e2eef8dca35-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-4c006a0e-82ef-4447-bcc2-02e2b661d2f2-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-f8dc3c14-6672-499b-b502-a02aa4e0cb08-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-e9341d3e-1c6b-46e1-92ab-fe8f439e708b-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-1bcb97e9-af33-4b26-b286-dba9dc267c73-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NoneCellYes_Phase4.parquet/part-00000-b6c10029-81cb-4173-b018-65c1e55b8168-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-8f772ae5-da31-42b2-9861-6030d0e361c2-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-fb10cc59-62a8-4f63-b625-bab95deec251-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-9754cb2b-4908-45e2-9bcd-4a77b4da4e65-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/NoneCellYes_PhaseT.parquet/part-00000-069328c9-16e3-4bfb-b164-db3c7d250a6c-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/diagonalYes_Phase4.parquet/part-00000-9b28a841-c571-425f-9b59-7f481c69887e-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/diagonalYes_Phase>=1.parquet/part-00000-28ae783b-a576-41aa-825f-273b6abeda9a-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/diagonalYes_Phase>=2.parquet/part-00000-04368def-856e-443a-9c52-d41b55d2e33b-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/diagonalYes_Phase>=3.parquet/part-00000-31c1a2ea-17a2-496d-beac-0cb054f9dabe-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/diagonalYes_PhaseT.parquet/part-00000-48066da1-c128-460d-a3cd-8eca7d996189-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-93ff87e2-7ebb-4850-b215-f44314bac7da-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-d50ad248-40eb-4ad6-a254-8a497e26ad97-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-aee25a2e-b43a-4962-a5a5-ec4186cfb6e1-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-4ea359b5-4ab0-42f3-bc79-d5a1065e038c-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-fe72a2ae-dddb-451a-8ae8-f3c631341090-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/oneCellYes_Phase4.parquet/part-00000-78cf794d-446d-41ac-ae62-f72fba83bf48-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/oneCellYes_Phase>=1.parquet/part-00000-596a5817-4213-4f84-868b-c0a28758096e-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/oneCellYes_Phase>=2.parquet/part-00000-f8442da3-427e-48ef-8b0f-264bff5f1c4d-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/oneCellYes_Phase>=3.parquet/part-00000-b90a4b9c-cfbf-41bd-90c0-072b02005a47-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_Other_original/oneCellYes_PhaseT.parquet/part-00000-a4a7ef20-b509-4f4c-be01-908ea3685c37-c000.snappy.parquet\n",
      "Part 1: df_eva_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NdiagonalYes_Phase4.parquet/part-00000-e1ed01e5-8f48-4c10-8f23-de2571b25822-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-88bd7f42-8258-413f-a6e1-ede04fb27d11-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-7aabc568-86ab-4064-a2c6-e4978880746c-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-b1f47bdb-44de-45ec-af2e-524c8f81c350-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NdiagonalYes_PhaseT.parquet/part-00000-08e8edcd-4765-40d3-a0d1-e6eed72937d4-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NoneCellYes_Phase4.parquet/part-00000-97cad521-8441-4717-ad86-fac865295b47-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NoneCellYes_Phase>=1.parquet/part-00000-a537f9ad-16d4-429c-b7aa-49d084f33e28-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NoneCellYes_Phase>=2.parquet/part-00000-5ae41de8-7495-4f91-9f63-8ba6c3032ec6-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NoneCellYes_Phase>=3.parquet/part-00000-c8dc19cb-7c33-4332-9f8c-1932873fa91c-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/NoneCellYes_PhaseT.parquet/part-00000-688e7157-9dab-4120-82f8-6c231840d87d-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/diagonalYes_Phase4.parquet/part-00000-df735043-7dbe-4284-ad2d-2dd1834a4ee7-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/diagonalYes_Phase>=1.parquet/part-00000-429828e3-2b61-43ce-982f-dcdefdeac3e6-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/diagonalYes_Phase>=2.parquet/part-00000-062d200c-1d79-48eb-8de5-241c5740221f-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/diagonalYes_Phase>=3.parquet/part-00000-ce36c446-39fe-4d37-9b75-bf4cb48bbe59-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/diagonalYes_PhaseT.parquet/part-00000-c700aa82-a102-42b1-b886-5e714caa1cdc-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-774f5be7-8e9c-4619-a64b-b280bcdec22d-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-9991ded6-f01e-4600-ac1b-c6ee5ad4626b-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-f735100e-31df-48b4-b0f6-a38e0e53c995-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-c524c318-ffd5-4203-9128-b0ea504504a2-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-fe99f896-8b96-4aec-8be7-78f19e223b0c-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/oneCellYes_Phase4.parquet/part-00000-110f8cfd-d3a6-4b32-846b-fed9acc896bb-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/oneCellYes_Phase>=1.parquet/part-00000-bf6743a4-0ae0-494b-881c-40732cc91650-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/oneCellYes_Phase>=2.parquet/part-00000-3f6566b7-25ec-41a3-aeec-d75d97b0d83b-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/oneCellYes_Phase>=3.parquet/part-00000-4256174a-e86e-4652-8818-2775b706088b-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_All_original/oneCellYes_PhaseT.parquet/part-00000-25ad6497-1d88-47b1-a487-c1ee44fe44dd-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-a3101831-6732-4972-9c02-1f1461ba9466-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-cbba9ab8-5a6d-487b-8441-4ce134356f75-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-0a67f1aa-2f34-4c3c-a7c8-06c9e51f5055-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-ca4d8a9f-99d7-4877-9709-9dd6b3868a37-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-489c1d33-b23a-49b2-a830-89eb01d99e90-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-2a599ff8-ac33-4000-bb70-19b6008c7263-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-23777d79-26f3-49de-8c2d-c1d6b09a5f72-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-7106ba6d-9b99-4a7c-b526-8bfc3ff08f78-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-80f870f6-e4c1-4a40-b0d1-4fedd8ea6211-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-06670e4d-91fb-48f5-8680-72dfdfaa998d-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/diagonalYes_Phase4.parquet/part-00000-c8425db9-15cc-4d13-97a8-408932f00f21-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-50445fb2-d982-4efd-baae-c44a653479c7-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-da05da02-2cf5-4346-abd1-775f1838f78b-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-7bd205d3-f976-4a99-a0b9-9a55612ea436-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-3e7027ae-2bac-4744-b3ea-ff98fdf7be9c-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-8b98283a-3650-4b48-9326-81b8dfce6558-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-3b7ed259-b75f-46b4-ac47-6437d59a052c-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-05d0381f-8487-4919-a039-e6a54a5db556-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-3586bb95-7d3d-43f7-af83-1f31f3a4d453-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-225be486-2471-4e7b-93fc-0dafe7a01cdd-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/oneCellYes_Phase4.parquet/part-00000-0c642861-7065-46e5-a8d2-38638f93e099-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-b71512f1-a5a7-488c-8f88-030cfde55c59-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-886b84a5-e630-4f47-892d-ac16667815d9-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-b0fb6162-4689-41e6-b94c-8fa00fd28c98-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-a77fcd84-8ef4-4dff-a35d-c95471555eb4-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-1fd55a7a-f208-4a0f-8ec4-8cc860887fdb-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-5b27de60-b716-4ea3-8545-ad4854cbf5fa-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-2697a033-ac1b-449e-9c11-4c6b3dce810f-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-534719c9-cc44-419f-b17a-f2739dc44673-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-2abede16-cd74-4ffe-8e84-84327944b38c-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-2fc0b982-fd67-4e45-8c8a-0a3b88807028-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-70b804b1-81dc-4980-81c3-36fce7a9f075-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-ba2c72bf-2b55-4545-89c2-7a2e2dbce4f6-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-aa8783f0-1ce9-448f-88d1-c85b75b896d6-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-4bfc0b38-c730-43fb-a027-5cec92c885db-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-96d82dca-3efe-4e1e-a54a-676a1dc05a16-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-7c73c048-3b5c-4066-b668-1326d2f0f859-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-d3baff2a-deae-4ae5-b67e-2509f3ab1504-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-80ae41f7-4aca-4269-b6f2-a84db6437775-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-ac1cce69-5c44-4cdb-8ecf-2693dbfd55f3-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-f27528b4-4b42-4aaf-a572-7fc2a25252c4-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-39c6a528-95c1-4d84-b9f9-38cb97d90da1-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-503b8f09-c7c7-40d7-82ec-9eb292c242a0-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-8ad55b10-a7b8-4303-b894-119655402fef-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-7cfd7497-d0af-4648-a214-27b56b428515-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-d0ad00d4-5f80-4f79-bc5b-06d4aa07fda2-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-ba7d943a-bc48-4d1a-9d30-ab65ee97437d-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-f3d89ea5-2016-496c-8afa-5af1e4961231-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-905d9bb5-72f0-43b2-9776-8c11c2f2168b-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-f0ad20a5-38a4-4d76-a0f5-e5db4c62b328-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NdiagonalYes_Phase4.parquet/part-00000-10cceab4-7562-413f-a454-d5c2c59e5b7f-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-2dcd7b18-ea80-4634-a814-d11120454756-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-fd462270-f92d-4aba-9ee8-fd885ffa20c7-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-d59da50e-963a-4457-8ad2-b6b005ebf48f-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-0cb4c5c5-804c-48b2-bbe6-25a7e60fafa5-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NoneCellYes_Phase4.parquet/part-00000-db80ee23-bed5-4333-9ae9-8e1f34395207-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-71b92625-fc7d-49df-8737-6139410c6600-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-ac2180a2-76e3-4d59-8f1d-e1b121055bef-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-e9ec4526-6479-4142-92d1-7eb6a475678b-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/NoneCellYes_PhaseT.parquet/part-00000-98ac234a-ce48-40e1-9ca3-03de8b99d2c9-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/diagonalYes_Phase4.parquet/part-00000-408e1b01-7da0-484f-9b18-0fa9eee3c6c9-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/diagonalYes_Phase>=1.parquet/part-00000-2108aae1-6c16-4d66-acf7-01bd59fb568e-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/diagonalYes_Phase>=2.parquet/part-00000-5361d3ba-dce1-4f35-8011-35b206ed0afc-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/diagonalYes_Phase>=3.parquet/part-00000-abcb3157-5360-44a7-a06b-d67620aaf5fc-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/diagonalYes_PhaseT.parquet/part-00000-4721b6f6-3f2d-43a5-bc16-8cfd29eb5b2e-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-033d40d2-3196-4889-800c-1a88b97ccdd1-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-45fa09c2-a147-4d18-818d-681ee6f667dd-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-fadecce1-793f-473c-a90f-0f77914b23ec-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-dd4b915b-fa2c-4c15-aa92-6226cfb94715-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-82e40c03-7468-43ae-83bf-7630e71e1b88-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/oneCellYes_Phase4.parquet/part-00000-17e8a3a7-3388-4c5d-acd3-65d91aa4adf5-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/oneCellYes_Phase>=1.parquet/part-00000-063bdf0a-114f-4a3c-8c48-b6bd64f90636-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/oneCellYes_Phase>=2.parquet/part-00000-0540d2af-fe5f-4934-af2b-e4341c5694dc-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/oneCellYes_Phase>=3.parquet/part-00000-96e5febb-e7dd-4933-966c-46b124f4bbcf-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_eva_somatic_Other_original/oneCellYes_PhaseT.parquet/part-00000-2de742a5-8276-412f-9100-3ea0ce78a131-c000.snappy.parquet\n",
      "Part 1: df_eva_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NdiagonalYes_Phase4.parquet/part-00000-44db69d3-5afa-45b9-b0cd-daa237a04c43-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-191b88d7-6bb7-4234-8f77-a56d5fcf8ffc-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-190701ac-a039-4664-aa85-04c1b304b2c8-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-68512845-1ac6-487a-a13c-8606796484fe-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NdiagonalYes_PhaseT.parquet/part-00000-f8ce8eaf-46a6-44cd-95c0-a576c886ba35-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NoneCellYes_Phase4.parquet/part-00000-c355418f-e58e-48ce-b243-5c60872c115a-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NoneCellYes_Phase>=1.parquet/part-00000-233b6fe7-44ac-4633-9985-6ea030481bce-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NoneCellYes_Phase>=2.parquet/part-00000-e1ee925f-7d95-4097-89a8-60b5b4c543a9-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NoneCellYes_Phase>=3.parquet/part-00000-7264e8fe-4fcd-4c5f-8f05-555e2a36a8c6-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/NoneCellYes_PhaseT.parquet/part-00000-dd9c0fca-a773-4a92-a043-f79a8a51f780-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/diagonalYes_Phase4.parquet/part-00000-2f080869-836f-4a88-a937-911273d0f65d-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/diagonalYes_Phase>=1.parquet/part-00000-f68d5576-6d27-413e-87cd-a536e87d8ca6-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/diagonalYes_Phase>=2.parquet/part-00000-02a44ccb-eb1c-493c-a478-a239e4ebb7e9-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/diagonalYes_Phase>=3.parquet/part-00000-19e13834-5d17-463d-a388-9cdf62eb6ff6-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/diagonalYes_PhaseT.parquet/part-00000-317500e7-c355-43ae-aa31-517d2555174b-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-c2e8b985-6ea4-404d-9bf6-ba21247c5913-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-7f5f8922-2744-4618-8af9-0483f8dae0c0-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-c73483a0-8a28-4351-9f00-77f10ff70b77-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-ce5f0b5e-dc48-4622-8528-76af5be8a841-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-9f110bd9-b8c1-4a76-a4b4-5ed561a18c23-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/oneCellYes_Phase4.parquet/part-00000-c22d7fd1-b3d3-4233-a868-4d9f5dd9bdb0-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/oneCellYes_Phase>=1.parquet/part-00000-d283c10b-cca0-4c02-a3d5-469ddceda95b-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/oneCellYes_Phase>=2.parquet/part-00000-e6a03a10-c99d-4947-97d5-025a34ec93ea-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/oneCellYes_Phase>=3.parquet/part-00000-c03e14c3-76aa-4077-bcab-38873c8ceaad-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_All_original/oneCellYes_PhaseT.parquet/part-00000-49243ed6-0109-4f4d-bf56-0d3ee0ec4acb-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-54a77a44-78ee-4c92-bd03-b8eff4006b77-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-c8ccbf94-5f0b-4d95-aa19-266d9855add2-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-7645f3d2-1c2b-4532-a2ea-be598fa41b2c-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-ef075c5d-22af-45a2-af3f-5093e04d40f6-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-76fe381b-46f4-4088-bb69-f672f0fea860-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-77391d1e-97b6-40df-bafc-0974a59888a2-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-dc0377f4-b9bc-4696-99c1-f9d86e9e181b-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-a0e907df-d74e-4247-a47c-32475f8dbca5-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-a15bbf08-5f8b-4ba1-9239-7f2e3f6c6665-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-bd2b9433-b6f9-4d75-9fa0-8bb8dcf25a50-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/diagonalYes_Phase4.parquet/part-00000-61fa86a0-26fc-43fc-a966-246d890ae599-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-8ecee244-5cf7-48ac-b552-dda1dc8805bd-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-8ae2630a-a9a8-4115-a0d9-903468b03319-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-c2806d57-1027-405d-b56b-b1fea59597c2-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-5891a538-e335-41c2-92fa-7bba17b33660-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-f280d782-31b6-4788-9aca-475f736e80d2-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-5ec85f02-5e54-490e-bceb-08bb1a70beac-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-db66aecc-1da1-4afe-8513-1e8efc6c9c9b-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-f42c4aa5-6783-4bf6-8821-2e9784bc96a7-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-24c30710-d801-4513-a1e6-8b89bc1f038b-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/oneCellYes_Phase4.parquet/part-00000-abd2c125-264d-4d63-b377-4ecb5310b2c1-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-a68fefba-2421-46e1-ab46-7aecab671eb6-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-d2e1e46f-55b9-4f79-bef7-459bdbfee198-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-53ed08fc-cafc-48d8-a0a1-81cc6f95c63c-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-a5de7208-721e-4267-b317-ac0086b548d5-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-dbbb4a92-709e-4590-a897-630a83cca0fb-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-984a6996-88ed-4742-a02e-e54c16958d19-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-dee22c87-4d53-4df2-b018-8c28a2945b30-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-1627fb76-5531-4302-af84-4bcf45bc0eb1-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-9375679b-5602-49da-97de-a2fb6bdab418-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-08b3f23a-beae-43c8-ac5a-8e1c291bb153-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-278c0313-634e-4386-aa5c-af983efd8c6d-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-c24f2e22-0552-47d7-8a97-084585d25984-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-e6f603c8-53ca-4169-ba08-e3c3500b9367-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-c89f46b2-4f2f-47b4-94e4-c56eab7497b0-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-cfef1683-870c-4544-a268-3464322b1820-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-fcc9c3fb-39eb-44dc-930c-3ed5b15192ef-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-957136d4-7482-4909-b0a8-5b3a6372ff49-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-0fa02ebe-86a6-49d7-86db-1a50a4cc9706-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-7ebe8358-9404-4e30-ad23-6b339613b3f0-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-450ca0eb-9725-4a18-b60d-4ad1d43bb6c5-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-79751b93-fdb2-47a0-b7af-ef7702863419-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-c7dc55c4-1dab-4b56-9941-ab24e58a4374-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-47d1cc0b-9b26-4808-a27a-88df1bd497fd-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-41579b87-0fc3-4d54-8e35-24e9c13f1587-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-d1ecee5a-b4a4-42ca-8035-8d5c542b6dfd-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-a344e843-dcc4-43e2-8691-147071e6b34a-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-ab05e8d1-b365-42da-8226-5bedff74c290-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-2ebcd548-fb03-44c3-928b-1ebe25daded8-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-7856a043-14c8-452c-9bc9-9482f4381c39-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NdiagonalYes_Phase4.parquet/part-00000-0b957047-60cb-40f7-9f74-369fd54e2fc1-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-66aa1d75-65d0-4660-a45a-410f1d54bb26-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-a4f29d6e-9496-47e9-8faa-c6e2f1d265f6-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-acb8bffc-309f-4d81-8a32-8e4128d729ab-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-6d20ce2e-4789-4ba2-b2ff-fbabe5ef540d-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NoneCellYes_Phase4.parquet/part-00000-0eeaf8bf-a350-4226-978e-7c41f21a7c2d-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-3e952161-4a7b-4083-a5e4-26665c67904f-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-58b29207-1464-4b14-a854-70864ecff9af-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-78babce3-e522-4ad7-abf4-5a13dc3e27b7-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/NoneCellYes_PhaseT.parquet/part-00000-fbcae053-ef5c-4daf-8f75-4c6976e998d4-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/diagonalYes_Phase4.parquet/part-00000-9d650a96-a2a5-4ab9-bca2-360974dfcdef-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/diagonalYes_Phase>=1.parquet/part-00000-43a90728-7602-45bb-9ee4-769d9bd42873-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/diagonalYes_Phase>=2.parquet/part-00000-627c49a5-42ce-47b0-af26-4df008a81848-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/diagonalYes_Phase>=3.parquet/part-00000-83f2c2d7-93c0-4cc2-88f5-64191beae999-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/diagonalYes_PhaseT.parquet/part-00000-ff89b92b-7804-41ac-94eb-be12e4159468-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-58a8fd08-c9ce-4062-a6f0-e089a7c8b8f9-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-be838aa7-ae49-4b9f-a433-86aae5f0edf1-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-080d1dd6-6c67-47d7-b419-310ac01532f5-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-0193e4d0-87a3-491b-a702-ed7d5901779f-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-6673c4ec-45e1-44c1-bf90-c91d43a2289e-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/oneCellYes_Phase4.parquet/part-00000-76263f68-38ec-4c45-a759-9a13a03e99ca-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/oneCellYes_Phase>=1.parquet/part-00000-f24e39a6-1339-401a-8e93-6967812b297c-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/oneCellYes_Phase>=2.parquet/part-00000-1e11003e-59e8-4d52-aa1d-d11ae6092d57-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/oneCellYes_Phase>=3.parquet/part-00000-cbe2a478-8d72-4926-8774-82dc579226da-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene2phenotype_Other_original/oneCellYes_PhaseT.parquet/part-00000-cf85bf53-b9f4-4117-8fc6-631d5788d47d-c000.snappy.parquet\n",
      "Part 1: df_gene2phenotype_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NdiagonalYes_Phase4.parquet/part-00000-84b82513-e405-49a0-b759-c3d5f9308d62-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-525c241b-f724-4aab-b28a-37d3551eab6b-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-cb9c7d4f-3837-42ec-b6fb-e384c64a2a23-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-cdd1e3e2-1cc4-4d6c-ba94-212f1c3eef97-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NdiagonalYes_PhaseT.parquet/part-00000-1487ddb9-30c7-4bb0-aca2-77129917188e-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NoneCellYes_Phase4.parquet/part-00000-06833afd-6ff2-4a81-8133-60e5bfe11ac2-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NoneCellYes_Phase>=1.parquet/part-00000-6c956da6-ac4a-40be-851c-2eb8bd41f02b-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NoneCellYes_Phase>=2.parquet/part-00000-6146a41e-4e55-4926-aff9-be2c4c45b3fa-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NoneCellYes_Phase>=3.parquet/part-00000-da4deeda-a7e3-49ca-9a19-01a054364fb3-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/NoneCellYes_PhaseT.parquet/part-00000-7d3a468c-6627-407f-a562-629b4c6707a2-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/diagonalYes_Phase4.parquet/part-00000-7f03eff1-5b68-4698-a05f-2b88bee11467-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/diagonalYes_Phase>=1.parquet/part-00000-12b57269-4e8d-49f8-924f-161951ebf5ce-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/diagonalYes_Phase>=2.parquet/part-00000-890a7dbb-df20-4e7e-8298-3edfd5ae39ba-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/diagonalYes_Phase>=3.parquet/part-00000-e40adb98-6711-4d5f-adc2-2c50c0556b1a-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/diagonalYes_PhaseT.parquet/part-00000-8f4a17a7-a2da-40e8-868e-ec2885961029-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-c2a0f58f-b71c-4a0e-a399-4eb0222d2b22-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-b23c95ef-47db-43c4-8047-10085924f668-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-729e9d68-3d62-46c1-a6e4-9d8b6e9df259-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-253fefb6-6902-4ef2-be19-fe9d3f811dbd-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-a3bdaf7b-5cbf-4871-a70c-458c18d71458-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/oneCellYes_Phase4.parquet/part-00000-a4ab4b34-f564-4ae7-b8f4-a9f125d1f004-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/oneCellYes_Phase>=1.parquet/part-00000-253e005e-3ac0-407f-addd-209f3ebc5c8b-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/oneCellYes_Phase>=2.parquet/part-00000-a31cee4c-cc81-42ad-9db7-e99735339ac1-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/oneCellYes_Phase>=3.parquet/part-00000-10c3558a-bd80-445e-9bc9-63e541658d64-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_original/oneCellYes_PhaseT.parquet/part-00000-1189435b-f91d-4ad7-84c5-7ead1c6081cf-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_All_propag/hasGeneticEvidence_Phase4.parquet/part-00000-4c1721c0-eb1f-4204-abff-f318ccaaaf72-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_All_propag\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-2807c299-0055-4447-8ec5-36e719d452f9-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-852e4cf0-5bd9-4d07-a1e9-5678becad07c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-87380cee-97f6-460b-a44a-80e2fdb0678d-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-05b13b4f-9bde-417f-885d-e82ffefd3cd9-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-0eb66459-554d-485e-84a3-f129112ea314-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-e1f98005-3060-435b-b075-ca9334eb8e41-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-7b08e01c-61fc-42fd-8f30-31df07901185-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-a3c7ffcb-3ec4-4c03-9839-746717f79323-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-74a82ebf-8ca5-422e-ae35-6a88f29c051a-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-50dcc5c3-65d4-4502-83d2-40bd56a69587-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/diagonalYes_Phase4.parquet/part-00000-f97acabd-e91b-4098-9b99-5fe68aef2cf3-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-787b9b80-3343-4bd7-a599-ee77f71a5d85-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-cb12e107-abd6-4157-a2fa-e423bbffced5-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-31445b52-085d-497c-bff4-e149a008fbab-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-c72d0be5-9235-4be5-b121-6dd4537598ff-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-4e23bca4-189b-4ca2-9d5b-8e7303a36e49-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-b7c4a744-a860-417a-b16c-57c7bcf65e4c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-0ffe9f0e-dd1b-4dee-b0cc-3e833adc3001-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-eaf2456c-7321-4758-9e2f-28fa4b0c11fe-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-9021a678-a547-48f4-8f2c-c7ea2804db3c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/oneCellYes_Phase4.parquet/part-00000-9d83f607-a19e-46f5-a1ef-a4daaefc0af1-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-3531c400-bf7c-4ad0-b866-b82f81acdaa2-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-22f763af-15aa-435b-93cf-6daf62eb0bce-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-7a9fce89-8527-4971-97f4-8a786e6578e1-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-ba94fe91-f22b-4766-9037-3fd989dbccb8-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-0f4944d1-196c-420c-bdf1-05ceb6670da3-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-f41c3b7b-3d1f-43c8-b0c8-03b0be192bf4-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-898785e6-ad91-404b-91a9-03cffcc4c8c6-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-bc994e84-dc72-40c2-8558-c285ea77a924-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-fd0d96b7-7bdb-4fd3-b483-99452593985b-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-69a86c85-0fd2-49b8-8ad3-edca2f2fed0f-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-87fdb466-50c9-4e8f-b971-baaa39d2ff67-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-af097d66-0f04-4598-9213-21ee4735db80-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-89f79041-8c69-40cc-a38e-61149e36d229-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-c30cd3dc-c5f5-4229-bab4-14d907db1ea4-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-547261e4-8068-4dec-9e0b-4ed9628394dc-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-89ff0983-4bb8-4aed-8de8-d83c795529d0-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-b9258553-e0ef-4d24-8f5b-81bcab386765-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-4a1f77f7-3ddf-4524-8e0f-e83df7e7a7ce-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-6d071a96-12a2-4e3f-8ae9-171d0f61ac60-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-0c6bc812-4485-40e7-a233-6a93c6c8f21c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-630e4292-dc47-4419-9530-dfd20464170e-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-2e71dabb-87b3-4834-8434-9518ce6c61a7-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-b8dc8983-b239-4ffe-befb-f5f0a717fb7c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-a0d48e88-5c43-4f9b-9de6-5ab9aeb2418c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-9efb1b75-e21d-4fa0-8fc4-ff3a6afc4997-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-274687c0-1d9b-45f6-983d-cf5ab148ff51-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-7d84db05-cd65-4525-ba45-75bd7e6aab04-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-ff8d7760-247c-48e2-80f6-4400fa08815a-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-2840aebf-d5cd-4174-a403-008a2ba9efce-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NdiagonalYes_Phase4.parquet/part-00000-8c2cd8a2-311e-41de-af2d-67525a8b7a13-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-9dbd9ec4-6670-416f-a09b-b78b8964f50b-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-43bceb9d-2ee6-4880-a576-60523f48c9b7-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-5fdad852-9df1-494d-8d0b-1cd7d5635dae-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-d470d6a0-047b-4d6d-a78b-c718c1a37ad4-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NoneCellYes_Phase4.parquet/part-00000-295ec2c9-cda5-4d83-a833-8bdf483c9e93-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-bcab197d-117a-4e9b-9a9d-42587bccd0c0-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-925f2278-b344-454a-ab62-cb8a45a305df-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-a5db1a61-d7a7-4760-bcc2-338e04d24ba2-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/NoneCellYes_PhaseT.parquet/part-00000-fabd4f0e-d41b-4076-8f72-2eae4d3d72a5-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/diagonalYes_Phase4.parquet/part-00000-9282d36a-c5e2-4298-9d5a-df053c3d2c5f-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/diagonalYes_Phase>=1.parquet/part-00000-f48c18a8-9eb6-4007-8a54-df6070e43278-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/diagonalYes_Phase>=2.parquet/part-00000-5b0e233e-084a-4620-8333-a4ffb8c287d4-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/diagonalYes_Phase>=3.parquet/part-00000-e9b55485-fdd5-47b5-a50c-ec3bfda8c49d-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/diagonalYes_PhaseT.parquet/part-00000-b60eb6f6-9605-4400-8893-3ea892ba641e-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-5e69d227-b1c3-45b6-81db-44b5fd379720-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-abde7530-4fe5-4748-bf27-ff43a5c87008-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-9d6c2d13-f9b9-4f1c-b06b-d7e1e186df68-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-6dfd05ae-dbf6-4182-b480-b168c4145606-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-a725d730-e2c8-4d6b-8d2b-f4cf1c0af34b-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/oneCellYes_Phase4.parquet/part-00000-a4d4eac4-7b1b-4f7d-930b-e15fcf3aba63-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/oneCellYes_Phase>=1.parquet/part-00000-2f9b5967-f558-407b-affe-3ef5b6ee46d8-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/oneCellYes_Phase>=2.parquet/part-00000-9795687f-7aa2-4851-b147-6cd327d47c4c-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/oneCellYes_Phase>=3.parquet/part-00000-16dc99ae-62e3-4fea-ab8d-8e9d6e14ae60-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_original/oneCellYes_PhaseT.parquet/part-00000-572b2593-33b9-4646-abd3-7db8c752fe97-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gene_burden_Other_propag/hasGeneticEvidence_Phase4.parquet/part-00000-00020add-22cc-49d3-97d5-847547451c79-c000.snappy.parquet\n",
      "Part 1: df_gene_burden_Other_propag\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NdiagonalYes_Phase4.parquet/part-00000-16d984c6-8469-4d6d-8ec3-07879eb87104-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-92b534ce-ba73-4baf-89f4-268444019588-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-4af0f4c1-68f9-4044-ad6d-ad5026c7b339-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-4c57cf58-097c-45c0-b3aa-959e71db8c89-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NdiagonalYes_PhaseT.parquet/part-00000-00acc11e-6a1e-43ff-8459-260fb8435826-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NoneCellYes_Phase4.parquet/part-00000-ee3bcf50-6980-48ea-9498-798365855d95-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NoneCellYes_Phase>=1.parquet/part-00000-763d0034-9ae3-434d-89ee-b6c06d89bc54-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NoneCellYes_Phase>=2.parquet/part-00000-0a6db45b-a64a-44a0-88c4-6eaed28cba0f-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NoneCellYes_Phase>=3.parquet/part-00000-68c435c2-e7e5-4136-88de-42c3e949432d-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/NoneCellYes_PhaseT.parquet/part-00000-5bf77f19-5e73-42ee-9fae-3c36977f1f56-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/diagonalYes_Phase4.parquet/part-00000-e44e095f-a6d8-4809-9b46-550e783429c7-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/diagonalYes_Phase>=1.parquet/part-00000-3cc9e8a2-1cb9-4a10-9bc2-45efb9931d19-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/diagonalYes_Phase>=2.parquet/part-00000-6bcf3f43-c345-4a3e-b751-66bd0b35ee1d-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/diagonalYes_Phase>=3.parquet/part-00000-7aa8950e-40da-4432-9be9-14dd654d001a-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/diagonalYes_PhaseT.parquet/part-00000-114a9f8b-b27a-4177-954f-87ba8f421aee-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-380862c4-aede-406e-8ce8-17a78af52786-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-768a63d6-2949-4987-beb4-6fecdf586d89-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-1df89159-16e2-43a3-9fff-b6b855a37eb3-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-05455f38-e409-402e-aeb5-c720726e9efd-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-d13d99fc-708a-4982-bd30-ae353d6d6e80-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/oneCellYes_Phase4.parquet/part-00000-385246cd-4220-4a26-af62-db41a9ec0d4d-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/oneCellYes_Phase>=1.parquet/part-00000-640354b2-87f6-4f17-a7fa-84ab1f9340a1-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/oneCellYes_Phase>=2.parquet/part-00000-3f8415bf-5335-4db5-9ad1-6f3890097ee8-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/oneCellYes_Phase>=3.parquet/part-00000-7b188f7d-a85c-48a6-bc49-b84222c52b49-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_All_original/oneCellYes_PhaseT.parquet/part-00000-882e81e7-79de-43a3-8381-b29d6f1605a0-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-431e1918-c33f-4415-9c11-aca1d416b33a-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-8684cec2-c770-47c3-8358-c79137ba77c3-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-59bca9ae-5354-4ec5-b969-8ef5fb45542f-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-231aeed8-bd26-4d0e-b7ad-19d5066f91d8-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-c3367b30-2173-4194-9ad6-ab5477b54157-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-7cbd9054-e8c0-4141-b0c3-3530be8be495-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-445089e9-a9dc-4450-aef0-2d0d34dd143b-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-b3cc7f52-0b47-4b16-9630-87bb37448e22-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-839db227-6f36-4c47-bbe3-5a84f990d5c0-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-e772bbce-a0f6-42cb-abee-8d91ef51a0e9-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/diagonalYes_Phase4.parquet/part-00000-923dc102-6569-4ac1-9ab0-000e643d1483-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-0093aafc-7ed9-4f86-8ed2-7c31d7c15c71-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-063cf7ce-95e2-4a50-8207-f0f8f81175cc-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-a1f72352-6811-4964-a815-d0448003424f-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-3bffa04c-4c0e-43fd-9c2d-44982f3cf270-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-8bd9f767-8448-4a24-9b5c-ce8a633c7d2e-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-250caa0c-feab-4a89-9189-6d60517aa729-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-6205fac3-09a2-4a95-8212-ef52f3e0820d-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-d67a48b5-8ebd-40fd-937a-746adca96882-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-f6e7f7ac-fda3-428d-abaa-1358a43b61be-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/oneCellYes_Phase4.parquet/part-00000-538a3285-6aa5-42e8-bdd0-79311e1a1b79-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-3bea3f4c-b63f-4035-8fcc-3ff7d5849112-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-732dc6cb-e20a-4d3d-ae74-06d07ee9bfaa-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-10c2545c-6e8c-40fb-aab8-e7801a5bec25-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-45a23505-3406-402d-ab77-0f221af3e8b6-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-68095dc9-f48c-4ad1-bbd4-3903371ad7a1-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-069df5a8-472a-4267-bc65-4e29c37a1fbe-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-75752015-87cc-424c-a745-286ba4911f89-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-bc45d32f-3831-4dc1-85c9-c1a3c595a19f-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-90125a61-fcd3-486f-8e66-6aa714d0baff-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-dafcb650-cf79-4b8d-aefb-6ebe5de7a844-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-1c2b7946-56e6-4ae3-ba54-2883b8d77dbb-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-7f5411ea-74b4-4c2a-a1a1-c412ba9af7f2-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-d8b47476-6bcd-4cc5-a6e0-df24d62fc1e3-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-48076691-8f63-46db-8dc0-a513e937092d-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-a2027b43-c901-4af8-80b7-23cafae89c42-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-41e69af6-9ebb-4b19-b118-54c0bbb4c2bb-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-9d689c17-8ac9-48af-8a0b-90a850da1e8c-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-e0997078-16c1-4531-8ea1-8a69b22e865b-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-216d88c1-afe1-4c49-8b89-94c657814128-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-7316eb9b-24b3-46d5-a323-4936a1e803fd-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-d2e3dc12-3489-4e1d-8956-4eab42e3317b-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-5f123356-d928-4448-83fd-d86f58da0d02-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-8233c75b-e6c2-4b6f-bd35-4a05326d4879-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-531f3448-d9d8-48ec-9460-52348ac7569f-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-834218b4-ac3e-4365-8b78-c350f643e726-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-e83cdfb7-33e7-4b25-b0c8-4e8b1f5c1ebb-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-ef324442-71c2-494a-af67-0b17a324ebf5-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-1f6e0fdc-5dff-48b2-a5ec-692043a58d91-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-56916316-47fb-43d7-97ea-f288225dc989-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NdiagonalYes_Phase4.parquet/part-00000-0e077402-2270-4ad8-baa2-be776b286fc2-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-61447863-d589-4128-b756-6e4e7ac8ebad-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-2285b5f3-e10f-464a-9e35-d86fd60cf6cb-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-027f5f66-4c8a-48c1-8b37-aeb218046849-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-69e95382-443f-49aa-8c0d-5345fdeedd4e-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NoneCellYes_Phase4.parquet/part-00000-f523a709-ea47-403f-9c93-ff4846c67b13-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-a9711f3c-fd48-4ff3-9d15-9202a24478f0-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-054fb862-69bf-4858-9ab8-d7a2520b5430-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-72687238-5793-4245-b027-67c17702700d-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/NoneCellYes_PhaseT.parquet/part-00000-1ead7b63-562c-4f2c-b8e9-33cc6c890f7c-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/diagonalYes_Phase4.parquet/part-00000-ad42535d-4e97-41be-bb8e-95a24426b248-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/diagonalYes_Phase>=1.parquet/part-00000-539eaaba-9e4e-416d-bd18-8cc72566ecd5-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/diagonalYes_Phase>=2.parquet/part-00000-be1980dc-fb44-48ca-9afc-dac9c6b55c0f-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/diagonalYes_Phase>=3.parquet/part-00000-ee0af301-d20c-4fa5-819e-0bbd52c832bb-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/diagonalYes_PhaseT.parquet/part-00000-89bc642b-737a-4d12-bd10-98e332e43720-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-4f403ea3-014a-41a9-9d33-1c3709623d6b-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-c4e2798c-7c5a-450d-b428-d1cdca8458d8-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-63d27807-5488-4a82-8bb4-773fc3ef23fe-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-4623d3e6-5cb2-4d8e-b094-f49cd52392fe-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-d565252e-b415-4ae8-bf95-4bcdaf615038-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/oneCellYes_Phase4.parquet/part-00000-523d0be1-7ab0-46b5-88eb-7f34192e29d0-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/oneCellYes_Phase>=1.parquet/part-00000-a312eee2-a166-4933-98a6-df54f2008f7a-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/oneCellYes_Phase>=2.parquet/part-00000-ca30e9a7-fb50-4d01-af1b-94d17190b415-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/oneCellYes_Phase>=3.parquet/part-00000-38d64047-c198-4f51-9e54-28cd8bbdea89-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_gwas_credible_set_Other_original/oneCellYes_PhaseT.parquet/part-00000-a6e265c4-255d-42eb-bc33-78436332fb47-c000.snappy.parquet\n",
      "Part 1: df_gwas_credible_set_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NdiagonalYes_Phase4.parquet/part-00000-7bc4778e-1449-4650-a0d1-517f9cb033c9-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-d194f2ee-0e78-4ebb-a1d2-2f4984072a33-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-4c816de8-a79e-4996-920e-d11f46f57c39-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-1568d1c9-dd53-49ca-9cff-0966f59c28af-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NdiagonalYes_PhaseT.parquet/part-00000-813dfd32-a9db-4310-b77c-c46552c648ad-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NoneCellYes_Phase4.parquet/part-00000-00da70a4-c5b0-4e49-820f-df27624a0d75-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NoneCellYes_Phase>=1.parquet/part-00000-40e854d4-87d6-4a1b-85c3-875991bb86bb-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NoneCellYes_Phase>=2.parquet/part-00000-e2162914-b76e-4659-a7f3-6649c18b98ff-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NoneCellYes_Phase>=3.parquet/part-00000-788b2f5b-2c1e-4605-9a9d-74c9e550217e-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/NoneCellYes_PhaseT.parquet/part-00000-26e96a89-192d-480c-9ca3-5bb6a4c749db-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/diagonalYes_Phase4.parquet/part-00000-6a6144e8-5161-49a7-869e-294daba3c332-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/diagonalYes_Phase>=1.parquet/part-00000-6588b8aa-365e-49e9-9153-ebb02cee6040-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/diagonalYes_Phase>=2.parquet/part-00000-9f391501-e84c-419b-8aa4-fdeaf015e289-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/diagonalYes_Phase>=3.parquet/part-00000-85871c60-de19-4ac1-9755-3aca149d05ee-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/diagonalYes_PhaseT.parquet/part-00000-647a85b7-21f6-4b40-8f4f-1dc4aeef689a-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-76968291-59b8-4a1f-8cd0-c813e2544f08-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-bc1e8040-7606-4fb7-9b1a-1b83a1613f2d-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-e99be44b-a71c-4e39-b138-3eed3e0dd757-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-430a8de4-009f-4df5-a1fb-a57c6742b411-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-81a41af5-431f-4a49-9d75-8cfda74c34d7-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/oneCellYes_Phase4.parquet/part-00000-fa32210d-01e2-444a-9f5d-c1c7df89d78e-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/oneCellYes_Phase>=1.parquet/part-00000-1f55b4ef-8032-492b-a714-d1290a731bbf-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/oneCellYes_Phase>=2.parquet/part-00000-e3d2311e-4edd-4185-83ab-f7101043b25c-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/oneCellYes_Phase>=3.parquet/part-00000-109969b3-3b81-434f-aeb3-9539ebc27529-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_All_original/oneCellYes_PhaseT.parquet/part-00000-0658ec98-e610-4640-9c4b-92f507ef9fc7-c000.snappy.parquet\n",
      "Part 1: df_impc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-f4238045-a1d4-4cdc-a23d-b27da60357d6-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-3b414a51-70fb-4741-a90d-f517d1914e17-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-def305d0-e6c1-41c8-9fcc-f9c716c3b0db-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-d659a0df-8dab-4ee2-ae7f-f372bee44a9d-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-48a386db-d8b0-4742-b5ab-7b4e0ce89564-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-ddac3509-8f2b-434e-adc1-cd4a73c20079-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-97b6e014-1b61-403a-a960-bbc410ab8288-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-1d6c0a56-dac1-4a29-8e04-52e29e845d4c-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-b1c84677-25f0-4710-b68f-828bbd5caf4e-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-be3d6e86-52ca-4fa6-ab19-c7fdc46117c6-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/diagonalYes_Phase4.parquet/part-00000-73719ec1-82c8-409f-94bb-45f7a94c17f3-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-67ea5ed2-543c-4647-a181-8547f4eea158-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-f8db3693-92e5-41b6-9ae6-5050886b1be1-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-3cdcb720-04b2-4697-91bf-3443f8182b38-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-0d5d1b9c-7495-4bca-89ff-ab289be9c821-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-aa6fbadc-0a5c-4473-a1bb-c1dfa07ebf36-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-f8a527a7-d38d-4bbe-b79b-c33acea11ea6-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-9961b150-45b1-410b-b626-f7d2d091421b-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-1c2441cb-5351-4bbe-895f-bcaa55e93b6b-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-700a37fd-8bb8-4303-b319-ad21f2b79cb6-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/oneCellYes_Phase4.parquet/part-00000-e45699bb-fbf0-4e01-9c49-613181c250fd-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-9c8e41dc-9974-4e4e-b69a-cdcfc48d136f-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-d2b8f09a-6948-4629-b53c-76feeb1c1f71-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-cbbe341a-42dc-442e-9b24-df75ea5a7b12-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-be14d5c1-fc70-4550-a030-657c29a87f75-c000.snappy.parquet\n",
      "Part 1: df_impc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-40f382d8-57a5-4b49-b3bb-5f2894ee30a0-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-4d11da9a-b12e-45f1-85af-d5c530cc501f-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-a4d614d2-7c3f-4efd-9e92-8f937afe457d-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-ee3d5156-57cf-4b02-935a-946531c77542-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-d6d832d3-3105-4742-99f0-63c507552641-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-926f5dfb-2a38-42c8-b2e6-198b256d86e7-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-b3a5e617-54a9-4865-acbf-027f97ace572-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-d0085bcf-6502-44f1-875a-b3dba30ce59f-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-ad9e21d1-a762-42e5-b916-5354a6e26c5d-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-171661df-7acb-4a68-96ae-5b2185c1adfd-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-f2bbf21d-589d-4621-8072-f4a8e6421f17-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-d13f17c5-fe9c-4aac-9c39-cdb890148222-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-6a0084d9-8027-4a0f-a408-1a774277e777-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-31f40399-d9f6-4641-aa2f-fd22b70d7c40-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-d573e97f-268a-41a6-b425-406e3952e453-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-17caa927-354c-4bef-9d97-b19b2a1cf6ec-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-03c82af1-c256-4a54-95f5-6d04317bfc3a-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-fba21429-f971-4f85-b261-e93fe6d54442-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-db478f12-2b62-406a-a2dc-366b52257cbc-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-83bad786-9903-44a0-b711-7f30c0f5b307-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-4e349de5-3a28-4836-85a5-38e1616e532a-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-fa543a01-a3aa-43ec-8371-508ab8e1cb4f-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-23199edb-13f1-4e4c-81da-4fd782e53edf-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-6052948e-8056-4370-90f5-36e7ce9d204b-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-1a14960d-ddfa-478b-9e86-ee8b7866b40c-c000.snappy.parquet\n",
      "Part 1: df_impc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NdiagonalYes_Phase4.parquet/part-00000-8d3dd764-bd5f-422f-b436-7e601aeb4204-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-29b50d9b-1446-4a3e-a495-67838c864858-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-67bd0dcf-f773-4dda-b2eb-c7a40b41751a-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-ca4fdea5-c328-4171-8f7f-ca3b5a6a59b3-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-578f94b1-ba77-409a-867e-d012c31292cb-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NoneCellYes_Phase4.parquet/part-00000-f13c2dcb-a690-4eae-9c84-bde2b4ce9306-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-00c16885-5d32-4600-910b-d0bc8b7e0204-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-dda783df-2a06-47b2-831c-38385b32c27a-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-3096e50d-36bd-48ef-bde3-60b4daf82f29-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/NoneCellYes_PhaseT.parquet/part-00000-47c99ef5-c819-484c-be32-219435e51dea-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/diagonalYes_Phase4.parquet/part-00000-e811e1a7-677e-41c6-977a-2e5289b68875-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/diagonalYes_Phase>=1.parquet/part-00000-6e9fa96d-89cb-4cb5-b3e8-1c0687ac3363-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/diagonalYes_Phase>=2.parquet/part-00000-f3f64e40-9540-4a56-ba9b-61af2d6908c7-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/diagonalYes_Phase>=3.parquet/part-00000-0fcd05e2-215d-401c-83f2-bbb1abe93ff1-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/diagonalYes_PhaseT.parquet/part-00000-e719d7b8-e193-4b22-ba88-d1b101b39b92-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-804eb17d-b63f-4f05-b083-883577e0de4d-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-3410fa43-0d48-4eec-8434-39651f89d1ae-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-37e5a500-0a40-459b-b485-f4d1f54aad92-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-0d011176-cec6-4ee1-be8a-e391dc6f1f3f-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-ebef76b1-777b-4f30-8229-46b226718fe5-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/oneCellYes_Phase4.parquet/part-00000-9fb163da-a0fc-4259-b756-0e799483dab9-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/oneCellYes_Phase>=1.parquet/part-00000-8d27736e-42ee-4f0e-83f5-4b27c452fc3b-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/oneCellYes_Phase>=2.parquet/part-00000-9800e6cd-d26e-40f2-b108-e68a3c8f001d-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/oneCellYes_Phase>=3.parquet/part-00000-d51d48cc-fccb-4c52-9fce-3367672f936f-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_impc_Other_original/oneCellYes_PhaseT.parquet/part-00000-855a9f73-ca70-4b02-9594-bf503f1f94b7-c000.snappy.parquet\n",
      "Part 1: df_impc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NdiagonalYes_Phase4.parquet/part-00000-b32857c7-b156-4ef8-b86d-44dc1101c147-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-01f599a1-76a3-431a-82da-765102ee68e4-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-bb91ce48-9b0d-4680-97ba-e346aed24743-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-0628c072-dc57-42a6-9a9e-fbf7762e9656-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NdiagonalYes_PhaseT.parquet/part-00000-3e0c1fa2-6be6-4c04-ad2a-f7368d7650bb-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NoneCellYes_Phase4.parquet/part-00000-5a69d9d3-4d34-4399-b0fa-1a6b449ab3b3-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NoneCellYes_Phase>=1.parquet/part-00000-7c55936d-d3bd-4e5b-b999-4f98c65a3311-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NoneCellYes_Phase>=2.parquet/part-00000-776e18f3-94a9-4310-a6e2-6b6385c5cc82-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NoneCellYes_Phase>=3.parquet/part-00000-f431edd9-3940-40b8-ba62-f455a93553cf-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/NoneCellYes_PhaseT.parquet/part-00000-e65d4e95-2f3c-4bc0-bcac-02ad8f9eccff-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/diagonalYes_Phase4.parquet/part-00000-ede73dd4-323a-423d-8ec7-a6128f207a65-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/diagonalYes_Phase>=1.parquet/part-00000-6cb1bb6e-8a52-4052-a055-981057f4a581-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/diagonalYes_Phase>=2.parquet/part-00000-9a4d5bb1-bf2b-4dba-a7eb-c27e879efc4f-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/diagonalYes_Phase>=3.parquet/part-00000-78d3e355-481d-4464-a5d2-d802e267e869-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/diagonalYes_PhaseT.parquet/part-00000-cb810173-96fd-4184-bdbb-894e4034765b-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-1d80155c-0cd1-435d-917a-d5d10fff1524-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-c677fc53-c640-4439-ad56-7b8d5612b7e6-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-240e714d-90a5-44da-9b52-6aa4ea608981-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-f34c434f-f115-4393-8ba0-05064c124d81-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-f3d8b993-e306-4cef-b0fe-7766f061122c-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/oneCellYes_Phase4.parquet/part-00000-c75c6154-6f4e-48a7-ac8b-08736cf1e29f-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/oneCellYes_Phase>=1.parquet/part-00000-e8a24c23-ad76-43c7-87bf-d9a3de133f9e-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/oneCellYes_Phase>=2.parquet/part-00000-c3f9b73a-cf16-4967-a99c-4c3dbb31ec66-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/oneCellYes_Phase>=3.parquet/part-00000-21c49e17-297a-4dae-b3f8-414506330c9d-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_All_original/oneCellYes_PhaseT.parquet/part-00000-4a32d061-48a5-40a6-b01f-9f8b52eea9d1-c000.snappy.parquet\n",
      "Part 1: df_intogen_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-b656b5ca-ff20-46c5-879f-a509285ca543-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-233b0f44-1e5d-460d-ad30-f99154333144-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-880b344c-7fcc-4adc-88d0-a69dca701970-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-21785d73-17c3-4137-bcf1-765e9e91866f-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-58b7257e-478b-475e-a39a-642a7c0ef0b1-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-d7f97436-7eea-4d81-8dc1-f70fa06a6e9f-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-eb1f13c2-3efa-43c6-b6f2-bbfaccf04990-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-b3a004d4-ae26-4255-af07-4c4a78c8c8dc-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-ff9dc2a2-32fa-45d9-a198-325e5293302e-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-930165d6-1e21-41cd-af2d-cefb46ba6631-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/diagonalYes_Phase4.parquet/part-00000-706cce9b-fc73-4162-ab00-3afb7b09a1e4-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-c9da3f01-f472-4c47-9914-250a085ea9de-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-ef1d1ffe-1e65-4dac-bebd-f12b92a9eee1-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-d3563c3c-5639-40cb-95d4-bcd1e88320b7-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-3282553e-3aa8-4fc0-a72a-aec850d638aa-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-d5b5d359-fefb-4697-b71c-1aaf47fc13c4-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-6abe0cad-a809-4afe-b003-8aada45bf28b-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-1f784a88-5b8a-4585-a1e0-8c25acde28b3-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-762abacb-d2e5-405c-acd6-8f65d0b27178-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-f624e353-5b8e-487c-9240-efd6b6a57936-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/oneCellYes_Phase4.parquet/part-00000-9ba80ee3-b872-47de-8887-90fdc6351519-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-50c1f2e7-99e1-4bc9-b5be-e455116cd690-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-1b5fec94-cddc-4feb-8b78-e50e2cb13ce5-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-9ab1f814-a834-4508-950c-528e3644f2da-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-e0b11501-326e-49d5-abf6-ca4a8f822116-c000.snappy.parquet\n",
      "Part 1: df_intogen_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-fa2fbbc7-373f-4178-a168-057aa6027607-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-0f86938f-7555-42b3-8dfd-a26f3004ff7c-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-4d28278e-d6ff-446e-b6cf-cb73bc40f597-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-77ea5e9d-bd96-4ac5-94a8-db503b0dc701-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-25935b7c-2a23-4231-b554-7d901c180433-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-0f6ba49a-c261-471e-82d1-afb2d1e40f11-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-b59c94a6-d344-484e-aecf-30191a2f9751-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-c059eb7a-8c26-477c-a1c4-00a9802071eb-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-bfd88cb2-ed53-4f26-971f-64eb6bad4184-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-bc2908de-1ca2-4dfc-91f4-fcb51a95ec4e-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-41ba5b06-faa8-4f57-96a0-d7271ea10c8b-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-68753906-3e2a-408c-be47-e26ff36bbc6d-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-8af1ff72-dcf7-4ae7-ab75-48adf508577c-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-b2dffb74-284f-4492-b7b9-08f62bb1dd3d-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-bdb0b7d7-1135-4e09-b48d-714b1616922d-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-08f651af-4098-4cf5-89e8-c38ecd4bd0f6-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-441e9dd2-3df0-46f7-8440-d26eb926f0da-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-d488e4f1-6fa7-43f0-a962-e84dc37653c0-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-29eb91d4-283a-411c-8305-98f0984a8703-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-bf5d79eb-6c89-4654-9c22-4f10bc10b6f0-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-b325068a-a719-415d-bff3-313a7997c890-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-ec3fdb4b-80c8-4d07-a8d4-64e2e6af8409-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-427402b6-1169-4a8f-b0ed-6cc7aec60b13-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-0e962171-9754-4b19-9a8e-e63aee755e72-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-b945788d-1e02-4d63-9094-8d244b93060e-c000.snappy.parquet\n",
      "Part 1: df_intogen_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NdiagonalYes_Phase4.parquet/part-00000-e5129563-1b29-4c78-861b-8b3f6c3f0f98-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-04a7c83c-046e-4c8b-a709-9f9f58134003-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-12dfc115-2a05-4468-8589-7301d049ac86-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-cbfb3e48-e64b-4bed-9a50-487c6baca1d0-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-be7c5f68-a0f1-4327-b5d1-97d5178245cc-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NoneCellYes_Phase4.parquet/part-00000-8c128171-0f03-4e9c-a04d-04c3c3001f06-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-e778bb5c-c9a7-4d9b-b629-4ee6c1696399-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-35590111-623a-4e1e-b28f-d1f272adda79-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-5d3c95ef-6437-4e35-8bc7-41c0dc5530a1-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/NoneCellYes_PhaseT.parquet/part-00000-3fea0c26-84f6-425e-8e6e-4edd59862793-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/diagonalYes_Phase4.parquet/part-00000-03580749-e967-4418-9caa-b36b260c0c7b-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/diagonalYes_Phase>=1.parquet/part-00000-eebe363e-835d-40be-9770-342cf392c757-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/diagonalYes_Phase>=2.parquet/part-00000-2bd993b9-782f-4d46-8f48-dda233780b30-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/diagonalYes_Phase>=3.parquet/part-00000-3f44c8a0-4949-45c5-a0a4-823c370db07b-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/diagonalYes_PhaseT.parquet/part-00000-baeea2c5-7e21-46ff-a50a-f83dae875c20-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-35a5370c-fd04-4e24-b12e-7247ab39a877-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-871876ff-d958-42e1-9b0e-9dcd64c29ea7-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-12250dff-bafe-4fed-afb5-cd0ee6501d76-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-7aec306e-e2b2-4ba1-8c24-9fb60ad57382-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-19921ab5-e9d9-4052-90b5-3cc9b553567c-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/oneCellYes_Phase4.parquet/part-00000-a615bcc4-4799-4a4f-9288-4dbca2e5daad-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/oneCellYes_Phase>=1.parquet/part-00000-11bfa1f3-1759-4c6e-b991-c6bccd124af8-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/oneCellYes_Phase>=2.parquet/part-00000-9c2fa654-4a9b-4252-9b98-cae30ea72de7-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/oneCellYes_Phase>=3.parquet/part-00000-26d6a0f7-3250-474f-936f-c8f2352cb8b2-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_intogen_Other_original/oneCellYes_PhaseT.parquet/part-00000-fafbac2f-9998-479c-a7ae-d7da624c1b77-c000.snappy.parquet\n",
      "Part 1: df_intogen_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NdiagonalYes_Phase4.parquet/part-00000-f3dd8554-819d-475d-8a1d-da4c8485b35d-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-9e8842a8-472a-49d2-be85-148e41498fcc-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-5d970bee-55a1-44f4-ac0b-ded53e779ca1-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-681e2af7-c405-4ea7-9fff-1815963021ba-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NdiagonalYes_PhaseT.parquet/part-00000-58fd67f3-05ba-4b78-a8e3-0dac20cad8a3-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NoneCellYes_Phase4.parquet/part-00000-9f5b7a59-d4f5-471f-8215-008cd8008b44-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NoneCellYes_Phase>=1.parquet/part-00000-af73c1f4-8504-4d86-8c2b-f4b97f30c9ea-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NoneCellYes_Phase>=2.parquet/part-00000-e6287c6d-0d31-440b-a9ec-47ff37e85869-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NoneCellYes_Phase>=3.parquet/part-00000-fb7a299e-5608-4e5a-a3b5-f079df5b0ee1-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/NoneCellYes_PhaseT.parquet/part-00000-b0c35636-ba8f-48d3-91fb-d3207a53c08b-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/diagonalYes_Phase4.parquet/part-00000-27401d0b-eed3-4f2a-8ddb-cd7fe2d5fd87-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/diagonalYes_Phase>=1.parquet/part-00000-219b6d2a-6e85-452c-beac-a769f9297eb8-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/diagonalYes_Phase>=2.parquet/part-00000-5c581308-3eb8-434c-b177-f98e1d12fe6d-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/diagonalYes_Phase>=3.parquet/part-00000-2241110d-3dde-46a9-89eb-79f22dda6798-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/diagonalYes_PhaseT.parquet/part-00000-a5135991-bf97-4e29-9a33-690a0c82c4e8-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-a935d552-d39d-4125-b8eb-90082487f138-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-1c68fbdd-1bdc-41d1-900f-8f6ea47f7b6b-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-c38a49c1-eefa-4839-a929-8a345fe5d8c3-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-c80e6237-7c40-4743-8db0-22326677a459-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-3ff955af-6158-4682-a1df-6b64e449d012-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/oneCellYes_Phase4.parquet/part-00000-cfe1ee63-2c84-4057-8bb9-5a6143fe0da5-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/oneCellYes_Phase>=1.parquet/part-00000-6806fad7-e053-4bad-b028-2343e3f3cba0-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/oneCellYes_Phase>=2.parquet/part-00000-5015e115-b372-4e75-bce5-613fd453fda9-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/oneCellYes_Phase>=3.parquet/part-00000-9c066611-4691-40e2-bf4f-9b926cb04443-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_All_original/oneCellYes_PhaseT.parquet/part-00000-21e2e1a7-c484-4606-838a-d16645b5edc7-c000.snappy.parquet\n",
      "Part 1: df_orphanet_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-da7f3b55-77a1-4fc0-ba0c-a2d199a396f6-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-3b97ccc7-7087-454d-824f-45f274916325-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-dcf7650e-37a0-4c98-aadc-128bf3bcc072-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-c99ab2f1-fdf5-4941-b34b-d175393329f6-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-45f28301-1ada-4109-b72b-3f628010db67-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-86d9e7d9-ba6f-4926-b594-0817e74cb2ca-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-c7d5a38c-b17e-477a-944f-9eae97da72d6-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-23d1676f-b0b7-482a-8055-a307f76ea0d3-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-cdd76c0b-c589-4549-96d8-bef2280e30eb-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-b3bc2772-7bad-4776-b67e-41d0224bd5ee-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/diagonalYes_Phase4.parquet/part-00000-89e646a7-195f-4e98-8663-777b1c8c779c-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-6f60da74-57d9-4f29-a8ac-8c21a4e90946-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-bd808f6d-c52e-46bc-85c0-6e97cdfd19a8-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-1580b4bb-c97e-4957-813c-57abe50539cc-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-b47c7d0d-80ed-4cf6-a92c-17c3a115b3cd-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-b77bdcad-f7d1-458a-b67c-2b631155f96d-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-cf20c9ae-6c88-4663-b5d4-994c3a54cbc0-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-c97d2fb9-a4d0-4c07-a2e5-addff7d7ae57-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-d8cb717e-1350-41a9-b46f-c2ae5a411d23-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-a1ea70b1-f9ea-49f5-821f-a5ff689e09bf-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/oneCellYes_Phase4.parquet/part-00000-953220f3-70e8-40fd-91c4-84de161f16e8-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-98ccbab4-f855-49dc-b7f6-0e065483e7ec-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-1b542063-1ba1-41ad-857c-0876b2d39f62-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-50491f72-ff21-4b96-ad2d-09d107ccaa9a-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-ad117363-3986-48a5-93e1-5a23aaf72ea6-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-97280dfe-5ec6-4d12-96df-f0620007e023-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-5d0cc662-4ad1-4afa-9b2b-ab4093be323d-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-5baeffa0-40db-41c9-9b28-7968513e727a-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-49ca62b2-be9e-40b8-b59b-c033d88f6c5d-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-1ab7c401-d345-4322-b5cf-0257fb050502-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-2777e92c-5c8f-4c24-89ad-39372fe006ca-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-7f05e4f8-40c7-4800-a647-ce7b6b7c42e5-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-ac74da21-cfbc-4018-9766-7279e7b3770a-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-ab4098eb-9f67-4c0f-9873-f53d3949071b-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-9baf8f53-9bc0-4775-ae1e-8b4ea3f2f272-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-34c64e70-2559-446b-b08b-b642fee4874b-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-621b4781-882e-4631-8ce2-95cd2048edbc-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-2d3f516a-d1f4-4171-ba40-ed4d4aac59c7-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-5418ed78-88e5-4772-9421-4d54f0fe4292-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-282d12a8-a2a0-43e4-a4b1-2cb380d8707b-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-5351fd43-c332-45d9-8e19-2866305ee629-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-14b222fc-643e-4dd8-a625-921645f33721-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-d4544e53-df95-4849-885b-7257f3956089-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-c9af076d-f218-4ce1-931c-247018a946c5-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-97fcc8e3-d76e-432c-bcc9-eeab5e9369f2-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-8251d232-aee6-495e-a4d5-c91f0df52305-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-4bdafa77-8bf6-4f72-9f18-4fcce46ce0fa-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-cc0eecbe-6c29-49d4-9c00-7c5a8090e87e-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-15d64c4f-7c12-420b-831a-683ed43bab80-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-00a35f6f-f01d-4ab9-99fc-990766212c0f-c000.snappy.parquet\n",
      "Part 1: df_orphanet_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NdiagonalYes_Phase4.parquet/part-00000-67de924b-b5b9-453f-976e-d1af6a6473c2-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-b3010ae0-7005-4082-b5e4-57a5d8ce639a-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-b09c32de-89f9-45d3-aa52-634f458cd355-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-ebb9d47b-fd92-4128-bb54-325f580596d6-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-c07e8402-3c66-476a-a455-e225113086c2-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NoneCellYes_Phase4.parquet/part-00000-3cb2913c-8a65-4922-969b-cae46fbf7023-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-70ebd309-74b2-4ecb-8590-02b0a8006159-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-e2c4b0da-c084-4614-95e8-54eb9945c5b7-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-33ef3406-adbb-416b-943f-cd4daa609e10-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/NoneCellYes_PhaseT.parquet/part-00000-c301e7f3-9b04-4af9-b519-54f847ec1cc4-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/diagonalYes_Phase4.parquet/part-00000-60ea7042-e5cf-4421-964c-a850780f7645-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/diagonalYes_Phase>=1.parquet/part-00000-d49edb20-95d3-45a1-9a0a-e4250d82a9dc-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/diagonalYes_Phase>=2.parquet/part-00000-37692814-e282-4983-8b76-8f32600633b0-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/diagonalYes_Phase>=3.parquet/part-00000-ec1c66f9-cf64-4700-94b2-18571d6e0a5c-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/diagonalYes_PhaseT.parquet/part-00000-f56a0f4e-b564-4e24-872c-a5a7d0c391ff-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-f98af9ef-2845-4839-af55-518d4c0b728b-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-01b1e5ae-8eb7-41ad-9fa7-824f9dc3dff6-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-1778360b-e548-476d-b63b-cce6429a3829-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-99fad403-6038-4ca8-8dce-e2d9756cb67f-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-c8099e7a-7f5b-47c1-94c2-8588beafcaa2-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/oneCellYes_Phase4.parquet/part-00000-88daaaae-dfc5-4461-b36e-df4b05d437bc-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/oneCellYes_Phase>=1.parquet/part-00000-158fc214-f245-41a0-8da7-929a33449967-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/oneCellYes_Phase>=2.parquet/part-00000-1de56080-bf63-4d53-918c-56ec4e546e69-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/oneCellYes_Phase>=3.parquet/part-00000-1043f0b0-38d6-427b-af5f-207a6deb3229-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_orphanet_Other_original/oneCellYes_PhaseT.parquet/part-00000-53fc6be0-ba7b-4ee2-8930-1d1e3a04dc75-c000.snappy.parquet\n",
      "Part 1: df_orphanet_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NdiagonalYes_Phase4.parquet/part-00000-36908bf2-8b86-4b6c-b383-0fbc9d908c1c-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-e6374479-95eb-4c94-915d-51c43395b8f9-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-00741dfe-9182-4e3d-96fe-3ed5c8aeefb9-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-7a5d8d2b-2bae-4109-93d6-df21c03c9864-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NdiagonalYes_PhaseT.parquet/part-00000-6fc6aab0-c7e7-4c33-a764-49653641dab6-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NoneCellYes_Phase4.parquet/part-00000-888d6834-f996-48c6-b938-09f15a9c349e-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NoneCellYes_Phase>=1.parquet/part-00000-82b50447-1e41-470c-94a5-751ff1f4048b-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NoneCellYes_Phase>=2.parquet/part-00000-5749f297-53a1-4fbd-95bb-930017f6ebf0-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NoneCellYes_Phase>=3.parquet/part-00000-6357d50c-00fd-45e9-995c-91ba13465c8f-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/NoneCellYes_PhaseT.parquet/part-00000-f0410835-7b85-433a-9e0b-716f60914d34-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/diagonalYes_Phase4.parquet/part-00000-9f652a1a-a31f-40e1-a726-416f6259a9b2-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/diagonalYes_Phase>=1.parquet/part-00000-4f84e1dd-23a1-4772-bc86-c8b2dd4723b8-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/diagonalYes_Phase>=2.parquet/part-00000-db70f9fb-c197-49ac-a695-45c26fa5b605-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/diagonalYes_Phase>=3.parquet/part-00000-13153228-cbe4-42a5-b76b-da3c7408d74f-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/diagonalYes_PhaseT.parquet/part-00000-ba3e761e-4649-4a93-99f1-294d636717e7-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-d83709d2-d1f4-4610-bf67-0566497a77df-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-c782071b-18e9-4fb7-9fc7-59d59ac37aef-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-121f9c7a-a70e-46e2-9614-0a21de17d3d9-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-dca1f9f3-e701-4ab3-9fff-086d0af3c841-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-00b13350-ae75-49d1-9596-93f68a2e21b7-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/oneCellYes_Phase4.parquet/part-00000-b2f49713-5bc3-4861-b57d-aafce0438dfb-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/oneCellYes_Phase>=1.parquet/part-00000-3cb6b9a8-6000-4df9-9994-2aa1d6e675cc-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/oneCellYes_Phase>=2.parquet/part-00000-bc2e80b5-ba22-42ab-acee-a5de4d83bc3b-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/oneCellYes_Phase>=3.parquet/part-00000-88c1dbe2-43b5-4bd7-b5bd-95569430c58c-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_All_original/oneCellYes_PhaseT.parquet/part-00000-0e68a68e-e5c1-4099-907e-08eb51970c08-c000.snappy.parquet\n",
      "Part 1: df_somatic_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-5aa5a040-1e17-4cc3-a10d-2b63b44fe2e8-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-79b92573-d611-4119-8548-a03fbc1263df-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-6e69f33a-e2d6-47dc-aad0-2f649e5367ba-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-a00d21f2-884a-4817-9a61-be2f5b2f0939-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-daaca981-ef09-4c23-ab07-9e66b8782674-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-acaf600e-49d6-4b6f-941b-96c519fda7f1-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-9f891b96-d31a-4255-b69c-18759e3be4c4-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-aad62336-acae-4362-8785-39540ef7bed2-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-b06de710-3f9f-4f8c-9e33-775739f532ea-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-af24041a-4232-4a4d-9d32-bb7411cd18b9-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-71420e9d-cf44-4f60-bb9d-1fb50d088072-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-7ca6c1a8-4699-4cf8-95ac-a45b86bdd069-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-7f688378-0708-4289-9d89-799a09009c0c-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-2bd5909a-dcd3-4695-891c-f400f9e246e3-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-41149dca-9ee7-4626-973e-b4a959a31031-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-f18452be-4faa-466f-848c-2f0a908834af-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-988ae928-4823-481b-b7c7-411c3e9217c1-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-4474e65c-7e5a-4c3e-8281-a87f87a46cde-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-d745ab34-2e94-403e-b136-a29cbea5e232-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-cd157288-de7c-439d-9258-eaa3af7f43c9-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-ec2a5749-d864-46ac-90d5-b762a884444f-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-48cf1790-b3ee-4d0b-b8cc-d8c63c0ba384-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-38005e7d-b703-4112-94d6-b5579f353f8b-c000.snappy.parquet\n",
      "Part 1: df_somatic_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NdiagonalYes_Phase4.parquet/part-00000-21fd0153-fc01-4ba9-be4e-6b396c92607c-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-b8ece1a0-2228-4cba-99c4-72875c0aeee2-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-d6c1ae6a-6432-4998-a4b3-b8ca1a02e396-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-0e611dd0-4076-4206-8f20-922d95303def-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-74892957-67da-4167-9db9-6b0d80971db4-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NoneCellYes_Phase4.parquet/part-00000-78421d7f-ef7f-4886-a7d9-287a388885c4-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-f3b58eb5-53b7-4bda-b944-97cf9a62959c-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-29c08c29-7897-4f68-9989-bcd726953eef-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-2079806a-12f4-45fc-aa0c-77d9ef8a1900-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/NoneCellYes_PhaseT.parquet/part-00000-8bc003fb-48ab-4a18-ba73-d83794d1c709-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/diagonalYes_Phase4.parquet/part-00000-98b45813-739b-4a63-beea-8e1e0cc61f64-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/diagonalYes_Phase>=1.parquet/part-00000-14d54faa-5a65-448e-9376-ebb5156045d3-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/diagonalYes_Phase>=2.parquet/part-00000-0687a6bf-b905-4ca6-ab71-ff230a437729-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/diagonalYes_Phase>=3.parquet/part-00000-45535971-4821-4d63-bafb-d280472a2946-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/diagonalYes_PhaseT.parquet/part-00000-b0c2178f-2f45-42c9-8a87-284d60d851e3-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-174d55b9-aa07-44c3-9981-275dcd14e028-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-9bd9558a-1da7-42ee-9df5-4f22e49fb03c-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-877f7ae8-cae7-4f65-94a9-2b96d8bf9cfc-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-a8b0aea9-088f-4167-ab54-ee2cb0410af8-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-e7109b27-d262-4ded-b3c2-cc6c6c96fb14-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/oneCellYes_Phase4.parquet/part-00000-a31651d5-ef45-4a44-8a29-097bfd4e2ca2-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/oneCellYes_Phase>=1.parquet/part-00000-6f245ca2-8308-44a2-b5a4-e78e9f9289bb-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/oneCellYes_Phase>=2.parquet/part-00000-d7abbc0d-e733-49d1-881d-858bc43e1c98-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/oneCellYes_Phase>=3.parquet/part-00000-6f4b2b63-061f-44cb-a773-7fb1dcae5a44-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_somatic_Other_original/oneCellYes_PhaseT.parquet/part-00000-73e08682-33a9-41d0-b429-f0b7eaf0d878-c000.snappy.parquet\n",
      "Part 1: df_somatic_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NdiagonalYes_Phase4.parquet/part-00000-5ce0011e-3681-4873-8191-8ab2ceee78a6-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NdiagonalYes_Phase>=1.parquet/part-00000-2d28f741-ab12-47d2-8928-6c75b971dcae-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NdiagonalYes_Phase>=2.parquet/part-00000-0e4622fb-0b47-45e9-8774-7c98e98f1e76-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NdiagonalYes_Phase>=3.parquet/part-00000-b64388e4-c646-4b0b-8bda-5b37c0fe4fba-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NdiagonalYes_PhaseT.parquet/part-00000-ebec549f-96a3-4af0-9592-57fec7010a24-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NoneCellYes_Phase4.parquet/part-00000-ac8daa10-00bd-4e51-97cf-3f05ed629ad4-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NoneCellYes_Phase>=1.parquet/part-00000-f5c4b1e6-9396-459a-8f66-7344c673263f-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NoneCellYes_Phase>=2.parquet/part-00000-ff916998-8a7c-4e61-a5d0-140fa4759daf-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NoneCellYes_Phase>=3.parquet/part-00000-ce753cab-1604-4dc9-95d4-e7a441f37542-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/NoneCellYes_PhaseT.parquet/part-00000-093c3a91-68ba-41f4-8487-13559c373f5b-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/diagonalYes_Phase4.parquet/part-00000-e90c8929-1644-4fdc-b1ab-4862b64871c6-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/diagonalYes_Phase>=1.parquet/part-00000-9f3959f0-13a1-4b7b-809c-0892dc2b75f3-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/diagonalYes_Phase>=2.parquet/part-00000-e3afd179-5bf5-4936-b2e8-29870c525ff3-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/diagonalYes_Phase>=3.parquet/part-00000-ba7af501-eda7-49a4-bb6b-0abe4b0d5094-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/diagonalYes_PhaseT.parquet/part-00000-9b5d9342-9275-43a0-9b6c-edac0850479b-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/hasGeneticEvidence_Phase4.parquet/part-00000-bd0d7252-8fbe-42ab-8641-aef02076450f-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-c88febee-ab8c-4f3f-8ef2-15fa9badb273-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-b6b7bc17-8f60-4d01-8923-1973684cdd27-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-cbf0afb5-a1c2-4cf0-b0ec-d3f0c9a067a4-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/hasGeneticEvidence_PhaseT.parquet/part-00000-cab494cb-f03c-4256-bfdd-cfcade88d3a7-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/oneCellYes_Phase4.parquet/part-00000-ccb64677-c102-43da-afbc-9cf202b7d2ba-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/oneCellYes_Phase>=1.parquet/part-00000-0c4015d5-5fda-4f48-90ed-81f5568ed0e0-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/oneCellYes_Phase>=2.parquet/part-00000-b6eb0c39-6e80-415e-8d6c-0e22477d1e55-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/oneCellYes_Phase>=3.parquet/part-00000-00388e76-798d-43a5-b796-75779be48550-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_All_original/oneCellYes_PhaseT.parquet/part-00000-ad65e89a-c416-4ae7-ae40-6327fe5a4a08-c000.snappy.parquet\n",
      "Part 1: df_wCgc_All_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NdiagonalYes_Phase4.parquet/part-00000-71b7d8ef-5513-4868-b58a-8a7c01cca693-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NdiagonalYes_Phase>=1.parquet/part-00000-3cbfaf67-8f3e-4ebf-a4f7-e79e0b2fd66d-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NdiagonalYes_Phase>=2.parquet/part-00000-44b79902-577e-4c45-92c7-60c402a2fcf5-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NdiagonalYes_Phase>=3.parquet/part-00000-931efd6a-d03a-4624-9785-522998c19601-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NdiagonalYes_PhaseT.parquet/part-00000-f3b1a6a8-ed45-4aba-85a9-b9250a4f712e-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NoneCellYes_Phase4.parquet/part-00000-8a78f4ef-d469-4723-955c-d79deeb8ab8b-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NoneCellYes_Phase>=1.parquet/part-00000-d01107ff-0b91-43ce-aa8b-43595f4598e1-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NoneCellYes_Phase>=2.parquet/part-00000-1d39a46c-db38-481a-8868-fe2a90821006-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NoneCellYes_Phase>=3.parquet/part-00000-e9bd072b-f00d-4f3e-83d2-804aa974fc14-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/NoneCellYes_PhaseT.parquet/part-00000-72a22f7f-de19-4eb7-bee6-1a6b58988c97-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/diagonalYes_Phase4.parquet/part-00000-20cb2beb-36e4-450c-88c8-4123768c91dc-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/diagonalYes_Phase>=1.parquet/part-00000-81d8259a-a1b5-4943-bb6a-4a5bbe83612c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/diagonalYes_Phase>=2.parquet/part-00000-0d2146ad-0214-413a-823e-764d91871401-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/diagonalYes_Phase>=3.parquet/part-00000-92207aa1-8c5e-4d47-aceb-0980d0754d81-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/diagonalYes_PhaseT.parquet/part-00000-91c7c973-f739-449a-a456-59b7cfcb6a96-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/hasGeneticEvidence_Phase4.parquet/part-00000-8a73be51-6898-47ce-b964-3546c8dde8ae-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-69ba2a59-afe0-45bb-9276-77113e108153-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-71f08bd0-1f47-4b27-82ab-bb6befa98dfd-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-3e042c36-a487-4782-bf57-a4cfb437dfb2-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/hasGeneticEvidence_PhaseT.parquet/part-00000-04223fcd-d03d-44bf-8f61-161eac67658a-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/oneCellYes_Phase4.parquet/part-00000-633215f3-ef3f-4951-85bd-5e89655355ab-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/oneCellYes_Phase>=1.parquet/part-00000-fa0899de-ec72-4194-857b-de4422e50d8b-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/oneCellYes_Phase>=2.parquet/part-00000-f0c4a236-76da-4ded-94c4-c084e0e5dea6-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/oneCellYes_Phase>=3.parquet/part-00000-b9479872-415e-4c36-a34d-b74b425caaa6-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Oncology_original/oneCellYes_PhaseT.parquet/part-00000-e520b4d1-4c30-4fe6-9053-c75ed6c953e3-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Oncology_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NdiagonalYes_Phase4.parquet/part-00000-666511f0-323c-4165-b7c4-41da003d8982-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NdiagonalYes_Phase>=1.parquet/part-00000-bdde5657-5d6e-4943-9b4d-3f21778976b5-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NdiagonalYes_Phase>=2.parquet/part-00000-14108e5c-f702-4b7b-a763-81a57e20156c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NdiagonalYes_Phase>=3.parquet/part-00000-5d946306-4014-4c16-9866-7bb56466e636-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NdiagonalYes_PhaseT.parquet/part-00000-142fd5aa-0b5a-4cad-8fb3-2ee86a3114c3-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NoneCellYes_Phase4.parquet/part-00000-d30f9786-f8b3-4d93-a61a-31022e5b9593-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NoneCellYes_Phase>=1.parquet/part-00000-706fc846-0bb4-4df3-b348-21cbcbce6703-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NoneCellYes_Phase>=2.parquet/part-00000-0ccfe1a5-e83a-46f5-90c7-4e8cb57c3e19-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NoneCellYes_Phase>=3.parquet/part-00000-58292092-57dc-4425-8e9f-79053d60be1c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/NoneCellYes_PhaseT.parquet/part-00000-12b51a00-3501-4362-9a63-7bc726fd1216-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/diagonalYes_Phase4.parquet/part-00000-abf44ba6-78cf-4b0d-829c-b3d31a70db77-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/diagonalYes_Phase>=1.parquet/part-00000-9ea3316d-f83f-409c-9128-d5238f6376bb-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/diagonalYes_Phase>=2.parquet/part-00000-f049aa91-2a0b-4ed1-960f-b64e4b95940a-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/diagonalYes_Phase>=3.parquet/part-00000-7c0f1c17-df06-41d2-a40e-dc7e3248b591-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/diagonalYes_PhaseT.parquet/part-00000-40644b9b-dc60-4f87-98f9-b3ebb60f7733-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/hasGeneticEvidence_Phase4.parquet/part-00000-1a8a9d8c-564a-41a0-94f1-0a4b803048d2-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-b27f6cbd-d8c1-4f0c-a1fe-52e446bd47ba-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-a28bc0bd-8342-4b03-972a-99bbf4f8a6d1-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-a6759003-a2cf-4b36-a05a-5d7add064d0e-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/hasGeneticEvidence_PhaseT.parquet/part-00000-330ad0ac-5582-497f-8baf-358bf66c98c8-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/oneCellYes_Phase4.parquet/part-00000-d0f8624e-f286-42c8-bf5b-9758a510dbe3-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/oneCellYes_Phase>=1.parquet/part-00000-f2d749c2-5bd5-4ce5-a404-64e553ce096d-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/oneCellYes_Phase>=2.parquet/part-00000-b42b63fd-f702-4c81-9d0b-ec11015abf1c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/oneCellYes_Phase>=3.parquet/part-00000-7e004e89-1747-408a-b352-d86bd7606800-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_OtherNull_original/oneCellYes_PhaseT.parquet/part-00000-77d64c82-0971-4c12-99e0-4a050f3b2328-c000.snappy.parquet\n",
      "Part 1: df_wCgc_OtherNull_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NdiagonalYes_Phase4.parquet/part-00000-6d65cc8d-15f3-4fcb-9026-ba862c28f1c2-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NdiagonalYes_Phase>=1.parquet/part-00000-e0a0a9a0-753d-4212-b4a0-a5e25b85da1b-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NdiagonalYes_Phase>=2.parquet/part-00000-7322e616-ee11-4b53-8a8a-6bb31e992aeb-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NdiagonalYes_Phase>=3.parquet/part-00000-005b677d-e9e6-4b54-9a44-6082250e202d-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NdiagonalYes_PhaseT.parquet/part-00000-6c6bd336-7814-4305-9794-6e3eff5f68c2-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NdiagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NoneCellYes_Phase4.parquet/part-00000-83865e22-e66f-4b41-8afd-d255a70c9bf0-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NoneCellYes_Phase>=1.parquet/part-00000-c02f0cb6-799b-4f5c-940b-c553c0df76fe-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NoneCellYes_Phase>=2.parquet/part-00000-9ee59252-3df2-4d41-af7e-6d109fdfab2c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NoneCellYes_Phase>=3.parquet/part-00000-3163f9f0-5ce9-4e2d-a613-b7028a42752b-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/NoneCellYes_PhaseT.parquet/part-00000-e419ee80-ba9c-40ea-a799-ade8a2859f0c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: NoneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/diagonalYes_Phase4.parquet/part-00000-3bb79168-fc63-4bbb-9b5b-ff4f2fc073c4-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/diagonalYes_Phase>=1.parquet/part-00000-733997d1-f6c5-4dc6-aee2-5f0a7a2b75e6-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/diagonalYes_Phase>=2.parquet/part-00000-5b9283c7-aacb-4f02-96b7-f5c2af4c780d-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/diagonalYes_Phase>=3.parquet/part-00000-4946ff77-0e26-4a6a-b74d-0fbe3d1d0432-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/diagonalYes_PhaseT.parquet/part-00000-05ffc085-995f-4899-ad7a-28acc0765a3c-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: diagonalYes\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/hasGeneticEvidence_Phase4.parquet/part-00000-8b05f405-ff2d-49cf-adb6-b04dccf568b0-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/hasGeneticEvidence_Phase>=1.parquet/part-00000-c350bbc8-0e9c-4350-a054-6e84e6f700ab-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/hasGeneticEvidence_Phase>=2.parquet/part-00000-a33c4a78-3ae3-4ba7-9838-a2bdc8b752be-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/hasGeneticEvidence_Phase>=3.parquet/part-00000-94d78a94-b8e1-4adb-b750-20c4329df79e-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/hasGeneticEvidence_PhaseT.parquet/part-00000-74a9317d-625f-4cdd-9c71-1231a8c4710f-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: hasGeneticEvidence\n",
      "Part 3: PhaseT\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/oneCellYes_Phase4.parquet/part-00000-9e2cc398-ed87-4662-a812-f8a125af1351-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase4\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/oneCellYes_Phase>=1.parquet/part-00000-fba02e48-f54d-4217-8e83-8e08251f5cdb-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=1\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/oneCellYes_Phase>=2.parquet/part-00000-d8e3cc9a-02de-40dc-9c66-47213cd754c8-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=2\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/oneCellYes_Phase>=3.parquet/part-00000-2560f917-b8dd-4843-aabd-c92abee1ca93-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: Phase>=3\n",
      "finished\n",
      "Reading gs://ot-team/jroldan/2025-04-24_analysis/df_wCgc_Other_original/oneCellYes_PhaseT.parquet/part-00000-32209e9d-fbe7-4cb0-a543-627f5008ffcb-c000.snappy.parquet\n",
      "Part 1: df_wCgc_Other_original\n",
      "Part 2: oneCellYes\n",
      "Part 3: PhaseT\n",
      "finished\n"
     ]
    }
   ],
   "source": [
    "from itertools import islice\n",
    "import time\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "#from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "from functions import relative_success, spreadSheetFormatter, convertTuple\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats.contingency import odds_ratio, relative_risk\n",
    "# from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "# from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "#from itertools import islice\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    StringType,\n",
    "    IntegerType,\n",
    "    ArrayType\n",
    ")\n",
    "import pandas as pd\n",
    "\n",
    "from pyspark.sql import SparkSession\n",
    "import numpy as np\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "full_data = spark.createDataFrame(\n",
    "    data=[\n",
    "        (\"yes\", \"yes\"),\n",
    "        (\"yes\", \"no\"),\n",
    "        (\"no\", \"yes\"),\n",
    "        (\"no\", \"no\"),\n",
    "    ],\n",
    "    schema=StructType(\n",
    "        [\n",
    "            StructField(\"prediction\", StringType(), True),\n",
    "            StructField(\"comparison\", StringType(), True),\n",
    "        ]\n",
    "    ),\n",
    ")\n",
    "results=[]\n",
    "result = []\n",
    "result_st = []\n",
    "result_ci = []\n",
    "array2 = []\n",
    "results = []\n",
    "for path in parquet_files:\n",
    "    print(f\"Reading {path}\")\n",
    "    df = spark.read.parquet(path)\n",
    "    array1 = np.delete(\n",
    "        df.join(full_data, on=[\"prediction\", \"comparison\"], how=\"outer\")\n",
    "        .groupBy(\"comparison\")\n",
    "        .pivot(\"prediction\")\n",
    "        .agg(F.first(\"a\"))\n",
    "        .sort(F.col(\"comparison\").desc())\n",
    "        .select(\"comparison\", \"yes\", \"no\")\n",
    "        .fillna(0)\n",
    "        .toPandas()\n",
    "        .to_numpy(),\n",
    "        [0],\n",
    "        1,\n",
    "    )\n",
    "    total = np.sum(array1)\n",
    "    res_npPhaseX = np.array(array1, dtype=int)\n",
    "    resX = convertTuple(fisher_exact(res_npPhaseX, alternative=\"two-sided\"))\n",
    "    resx_CI = convertTuple(\n",
    "        odds_ratio(res_npPhaseX).confidence_interval(confidence_level=0.95)\n",
    "    )\n",
    "\n",
    "    result_st.append(resX)\n",
    "    result_ci.append(resx_CI)\n",
    "    (rs_result, rs_ci) = relative_success(array1)\n",
    "    import re\n",
    "\n",
    "    # 1. Extract the element between 'analysis/' and the next '/'\n",
    "    match1 = re.search(r'analysis/([^/]+)/', path)\n",
    "    part1 = match1.group(1) if match1 else None\n",
    "\n",
    "    # 2. Extract the element after the second '/' until '_Phase'\n",
    "    match2 = re.search(r'[^/]+/([^/_]+)_Phase', path)\n",
    "    part2 = match2.group(1) if match2 else None\n",
    "\n",
    "    # 3. Extract the element from 'Phase' until '.parquet'\n",
    "    match3 = re.search(r'(Phase[^.]+)\\.parquet', path)\n",
    "    part3 = match3.group(1) if match3 else None\n",
    "\n",
    "    print(\"Part 1:\", part1)\n",
    "    print(\"Part 2:\", part2)\n",
    "    print(\"Part 3:\", part3)\n",
    "\n",
    "    results.append(\n",
    "        [\n",
    "            part1, #data,\n",
    "            part2, #comparisonColumn,\n",
    "            part3, #predictionColumn,\n",
    "            round(float(resX.split(\",\")[0]), 2),\n",
    "            float(resX.split(\",\")[1]),\n",
    "            round(float(resx_CI.split(\",\")[0]), 2),\n",
    "            round(float(resx_CI.split(\",\")[1]), 2),\n",
    "            str(total),\n",
    "            np.array(res_npPhaseX).tolist(),\n",
    "            round(float(rs_result), 2),\n",
    "            round(float(rs_ci[0]), 2),\n",
    "            round(float(rs_ci[1]), 2),\n",
    "            path,\n",
    "        ]\n",
    "    )\n",
    "    print('finished')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "created pandas dataframe\n",
      "converting to spark dataframe\n",
      "preparing dataframe\n",
      "read pattern variables\n",
      "importing functions\n",
      "imported functions\n",
      "processed spreadsheet\n",
      "writting the dataframe\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dataframe written \n",
      " Analysis finished\n"
     ]
    }
   ],
   "source": [
    "today_date = str(date.today())\n",
    "\n",
    "df_results = pd.DataFrame(\n",
    "    results,\n",
    "    columns=[\n",
    "        \"group\",\n",
    "        \"comparison\",\n",
    "        \"phase\",\n",
    "        \"OR\",\n",
    "        \"pValue\",\n",
    "        \"LowCI\",\n",
    "        \"HighCI\",\n",
    "        \"total\",\n",
    "        \"array\",\n",
    "        \"rs\",\n",
    "        \"lowRs\",\n",
    "        \"HighRs\",\n",
    "        \"path\",\n",
    "    ],\n",
    ")\n",
    "print(\"created pandas dataframe\")\n",
    "print(\"converting to spark dataframe\")\n",
    "print(\"preparing dataframe\")\n",
    "\n",
    "schema = StructType(\n",
    "    [\n",
    "        StructField(\"group\", StringType(), True),\n",
    "        StructField(\"comparison\", StringType(), True),\n",
    "        StructField(\"phase\", StringType(), True),\n",
    "        StructField(\"oddsRatio\", DoubleType(), True),\n",
    "        StructField(\"pValue\", DoubleType(), True),\n",
    "        StructField(\"lowerInterval\", DoubleType(), True),\n",
    "        StructField(\"upperInterval\", DoubleType(), True),\n",
    "        StructField(\"total\", StringType(), True),\n",
    "        StructField(\"values\", ArrayType(ArrayType(IntegerType())), True),\n",
    "        StructField(\"relSuccess\", DoubleType(), True),\n",
    "        StructField(\"rsLower\", DoubleType(), True),\n",
    "        StructField(\"rsUpper\", DoubleType(), True),\n",
    "        StructField(\"path\", StringType(), True),\n",
    "    ]\n",
    ")\n",
    "\n",
    "print(\"read pattern variables\")\n",
    "df = spreadSheetFormatter(spark.createDataFrame(df_results, schema=schema))\n",
    "print(\"processed spreadsheet\")\n",
    "print(\"writting the dataframe\")\n",
    "\n",
    "# Convert list of lists to DataFrame\n",
    "# Regular expressions\n",
    "'''\n",
    "value_pattern = r\"df_([^_]+)_\"  # Extracts {value}\n",
    "middle_pattern = r\"df_[^_]+_([^_]+)_\"  # Extracts middle part (All, Other, etc.)\n",
    "suffix_pattern = r\"(original|propag)$\"  # Extracts suffix (original or propag)\n",
    "'''\n",
    "\n",
    "df.withColumn(\n",
    "    \"datasource\",\n",
    "    F.regexp_extract(F.col(\"group\"), r\"df_(.*?)_(All|Other|OtherNull|Oncology)_(propag|original)\", 1)\n",
    ").withColumn(\n",
    "    \"therArea\",\n",
    "    F.regexp_extract(F.col(\"group\"), r\"_(All|Other|OtherNull|Oncology)_\", 1)\n",
    ").withColumn(\n",
    "    \"type\",\n",
    "    F.regexp_extract(F.col(\"group\"), r\"_(propag|original)$\", 1)\n",
    ").toPandas().to_csv(\n",
    "    f\"gs://ot-team/jroldan/analysis/{today_date}_genEvidAnalysis_new_resolvedDiscr.csv\"\n",
    ")\n",
    "\n",
    "print(\"dataframe written \\n Analysis finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded files\n",
      "loaded newColoc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded gwasComplete\n",
      "loaded resolvedColloc\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/07/10 21:42:12 WARN CacheManager: Asked to cache already cached data.\n",
      "25/07/10 21:42:13 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run temporary direction of effect\n",
      "built drugApproved dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/07/10 21:42:14 WARN CacheManager: Asked to cache already cached data.\n",
      "25/07/10 21:42:14 WARN CacheManager: Asked to cache already cached data.\n",
      "25/07/10 21:42:15 WARN CacheManager: Asked to cache already cached data.\n",
      "25/07/10 21:42:15 WARN CacheManager: Asked to cache already cached data.\n",
      "25/07/10 21:42:15 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "################# INCLUDE DIAGONAL \n",
    "\n",
    "import time\n",
    "from array import ArrayType\n",
    "from functions import (\n",
    "    relative_success,\n",
    "    spreadSheetFormatter,\n",
    "    discrepancifier,\n",
    "    temporary_directionOfEffect,\n",
    ")\n",
    "# from stoppedTrials import terminated_td\n",
    "from DoEAssessment import directionOfEffect\n",
    "# from membraneTargets import target_membrane\n",
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "from datetime import datetime\n",
    "from datetime import date\n",
    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType\n",
    "from pyspark.sql.types import (\n",
    "    StructType,\n",
    "    StructField,\n",
    "    DoubleType,\n",
    "    DecimalType,\n",
    "    StringType,\n",
    "    FloatType,\n",
    ")\n",
    "import pandas as pd\n",
    "from functools import reduce\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "spark.conf.set(\n",
    "    \"spark.sql.shuffle.partitions\", \"400\"\n",
    ")  # Default is 200, increase if needed\n",
    "\n",
    "\n",
    "path_n='gs://open-targets-data-releases/25.06/output/'\n",
    "\n",
    "target = spark.read.parquet(f\"{path_n}target/\")\n",
    "\n",
    "diseases = spark.read.parquet(f\"{path_n}disease/\")\n",
    "\n",
    "evidences = spark.read.parquet(f\"{path_n}evidence\")\n",
    "\n",
    "credible = spark.read.parquet(f\"{path_n}credible_set\")\n",
    "\n",
    "new = spark.read.parquet(f\"{path_n}colocalisation_coloc\") \n",
    "\n",
    "index=spark.read.parquet(f\"{path_n}study/\")\n",
    "\n",
    "variantIndex = spark.read.parquet(f\"{path_n}variant\")\n",
    "\n",
    "biosample = spark.read.parquet(f\"{path_n}biosample\")\n",
    "\n",
    "ecaviar=spark.read.parquet(f\"{path_n}colocalisation_ecaviar\")\n",
    "\n",
    "all_coloc=ecaviar.unionByName(new, allowMissingColumns=True)\n",
    "\n",
    "print(\"loaded files\")\n",
    "\n",
    "newColoc = (\n",
    "    all_coloc.join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on left side\n",
    "            \"studyLocusId as leftStudyLocusId\",\n",
    "            \"StudyId as leftStudyId\",\n",
    "            \"variantId as leftVariantId\",\n",
    "            \"studyType as credibleLeftStudyType\",\n",
    "        ),\n",
    "        on=\"leftStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        credible.selectExpr(  #### studyLocusId from credible set to uncover the codified variants on right side\n",
    "            \"studyLocusId as rightStudyLocusId\",\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"variantId as rightVariantId\",\n",
    "            \"studyType as credibleRightStudyType\",\n",
    "            \"pValueExponent as qtlPValueExponent\",\n",
    "            'isTransQtl'\n",
    "        ),\n",
    "        on=\"rightStudyLocusId\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .join(\n",
    "        index.selectExpr(  ### bring modulated target on right side (QTL study)\n",
    "            \"studyId as rightStudyId\",\n",
    "            \"geneId\",\n",
    "            \"projectId\",\n",
    "            \"studyType as indexStudyType\",\n",
    "            \"condition\",\n",
    "            \"biosampleId\",\n",
    "        ),\n",
    "        on=\"rightStudyId\",\n",
    "        how=\"left\",\n",
    ")\n",
    "    # .persist()\n",
    ")\n",
    "\n",
    "print(\"loaded newColoc\")\n",
    "\n",
    "# remove columns without content (only null values on them)\n",
    "df = evidences.filter((F.col(\"datasourceId\") == \"gwas_credible_sets\"))\n",
    "\n",
    "# Use an aggregation to determine non-null columns\n",
    "non_null_counts = df.select(\n",
    "    *[F.sum(F.col(col).isNotNull().cast(\"int\")).alias(col) for col in df.columns]\n",
    ")\n",
    "\n",
    "# Collect the counts for each column\n",
    "non_null_columns = [\n",
    "    row[0] for row in non_null_counts.collect()[0].asDict().items() if row[1] > 0\n",
    "]\n",
    "\n",
    "# Select only the non-null columns\n",
    "filtered_df = df.select(*non_null_columns)  # .persist()\n",
    "\n",
    "## bring studyId, variantId, beta from Gwas and pValue\n",
    "gwasComplete = filtered_df.join(\n",
    "    credible.selectExpr(\n",
    "        \"studyLocusId\", \"studyId\", \"variantId\", \"beta as betaGwas\", \"pValueExponent\"\n",
    "    ),\n",
    "    on=\"studyLocusId\",\n",
    "    how=\"left\",\n",
    ")  # .persist()\n",
    "\n",
    "print(\"loaded gwasComplete\")\n",
    "\n",
    "resolvedColoc = (\n",
    "    (\n",
    "        newColoc.withColumnRenamed(\"geneId\", \"targetId\")\n",
    "        .join(\n",
    "            gwasComplete.withColumnRenamed(\"studyLocusId\", \"leftStudyLocusId\"),\n",
    "            on=[\"leftStudyLocusId\", \"targetId\"],\n",
    "            how=\"inner\",\n",
    "        )\n",
    "        .join(  ### propagated using parent terms\n",
    "            diseases.selectExpr(\n",
    "                \"id as diseaseId\", \"name\", \"parents\", \"therapeuticAreas\"\n",
    "            ),\n",
    "            on=\"diseaseId\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"diseaseId\",\n",
    "            F.explode_outer(F.concat(F.array(F.col(\"diseaseId\")), F.col(\"parents\"))),\n",
    "        )\n",
    "        .drop(\"parents\", \"oldDiseaseId\")\n",
    "    ).withColumn(\n",
    "        \"colocDoE\",\n",
    "        F.when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"eqtl\", \"pqtl\", \"tuqtl\", \"sceqtl\", \"sctuqtl\"]\n",
    "            ),\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            ),\n",
    "        ).when(\n",
    "            F.col(\"rightStudyType\").isin(\n",
    "                [\"sqtl\", \"scsqtl\"]\n",
    "            ),  ### opposite directionality than sqtl\n",
    "            F.when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") > 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") > 0),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"betaGwas\") < 0) & (F.col(\"betaRatioSignAverage\") < 0),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            ),\n",
    "        ),\n",
    "    )\n",
    "    # .persist()\n",
    ")\n",
    "print(\"loaded resolvedColloc\")\n",
    "\n",
    "datasource_filter = [\n",
    "#   \"ot_genetics_portal\",\n",
    "    \"gwas_credible_sets\",\n",
    "    \"gene_burden\",\n",
    "    \"eva\",\n",
    "    \"eva_somatic\",\n",
    "    \"gene2phenotype\",\n",
    "    \"orphanet\",\n",
    "    \"cancer_gene_census\",\n",
    "    \"intogen\",\n",
    "    \"impc\",\n",
    "    \"chembl\",\n",
    "]\n",
    "\n",
    "assessment, evidences, actionType, oncolabel = temporary_directionOfEffect(\n",
    "    path_n, datasource_filter\n",
    ")\n",
    "\n",
    "print(\"run temporary direction of effect\")\n",
    "\n",
    "\n",
    "print(\"built drugApproved dataset\")\n",
    "\n",
    "analysis_chembl_indication = (\n",
    "    discrepancifier(\n",
    "        assessment.filter((F.col(\"datasourceId\") == \"chembl\"))\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(F.col(\"clinicalPhase\")).over(\n",
    "                Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "            ),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .agg(F.count(\"targetId\"))\n",
    "    )\n",
    "    .filter(F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    .drop(\n",
    "        \"coherencyDiagonal\", \n",
    "        \"coherencyOneCell\", \n",
    "        \"noEvaluable\", \n",
    "        \"GoF_risk\", \n",
    "        \"LoF_risk\"\n",
    "    )\n",
    "    .withColumnRenamed(\"GoF_protect\", \"drugGoF_protect\")\n",
    "    .withColumnRenamed(\"LoF_protect\", \"drugLoF_protect\")\n",
    "    # .persist()\n",
    ")\n",
    "\n",
    "chemblAssoc = (\n",
    "    discrepancifier(\n",
    "        assessment.filter(\n",
    "            (F.col(\"datasourceId\") == \"chembl\")\n",
    "            & (F.col(\"homogenized\") != \"noEvaluable\")\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"maxClinPhase\",\n",
    "            F.max(\"clinicalPhase\").over(Window.partitionBy(\"targetId\", \"diseaseId\")),\n",
    "        )\n",
    "        .groupBy(\"targetId\", \"diseaseId\", \"maxClinPhase\")\n",
    "        .pivot(\"homogenized\")\n",
    "        .count()\n",
    "    )\n",
    "    #.filter(F.col(\"coherencyDiagonal\") == \"coherent\")\n",
    "    .drop(\n",
    "        \"coherencyDiagonal\", \"coherencyOneCell\", \"noEvaluable\", \"GoF_risk\", \"LoF_risk\"\n",
    "    )\n",
    "    .withColumnRenamed(\"GoF_protect\", \"drugGoF_protect\")\n",
    "    .withColumnRenamed(\"LoF_protect\", \"drugLoF_protect\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "resolvedColocFiltered = resolvedColoc.filter((F.col('clpp')>=0.01) | (F.col('h4')>=0.8))\n",
    "benchmark = (\n",
    "    (\n",
    "        resolvedColocFiltered\n",
    "        #.filter(F.col(\"betaGwas\") < 0)\n",
    "        .filter(\n",
    "        F.col(\"name\") != \"COVID-19\"\n",
    "    )\n",
    "        .join(  ### select just GWAS giving protection\n",
    "            analysis_chembl_indication, on=[\"targetId\", \"diseaseId\"], how=\"right\"  ### RIGHT SIDE\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"AgreeDrug\",\n",
    "            F.when(\n",
    "                (F.col(\"drugGoF_protect\").isNotNull())\n",
    "                & (F.col(\"colocDoE\") == \"GoF_protect\"),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"drugLoF_protect\").isNotNull())\n",
    "                & (F.col(\"colocDoE\") == \"LoF_protect\"),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"no\")),\n",
    "        ).withColumn(\n",
    "            \"AgreeDrugDiagonal\",\n",
    "            F.when(\n",
    "                (F.col(\"drugGoF_protect\").isNotNull())\n",
    "                & ((F.col(\"colocDoE\") == \"GoF_protect\") | (F.col(\"colocDoE\") == \"LoF_risk\")),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"drugLoF_protect\").isNotNull())\n",
    "                & ((F.col(\"colocDoE\") == \"LoF_protect\") | (F.col(\"colocDoE\") == \"GoF_risk\")),\n",
    "                F.lit(\"yes\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"no\")),\n",
    "        )\n",
    "    )  #### remove COVID-19 associations\n",
    ").join(biosample.select(\"biosampleId\", \"biosampleName\"), on=\"biosampleId\", how=\"left\").persist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1002:===================================================>(393 + 7) / 400]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+---------------+-------------+----------------+------------+-----------------+----------+--------------+--------------------------+----+--------------------+--------------------+----+----+----+----+----+-----------+-------------+---------------------+--------------+----------------------+-----------------+----------+---------+--------------+---------+------------+----------+-------------------------+-------------+------------------+----+-----+--------+-------+---------+--------+--------------+----+----------------+--------+------------+---------------+---------------+---------+-----------------+-------------+\n",
      "|biosampleId|       targetId|    diseaseId|leftStudyLocusId|rightStudyId|rightStudyLocusId|chromosome|rightStudyType|numberColocalisingVariants|clpp|colocalisationMethod|betaRatioSignAverage|  h0|  h1|  h2|  h3|  h4|leftStudyId|leftVariantId|credibleLeftStudyType|rightVariantId|credibleRightStudyType|qtlPValueExponent|isTransQtl|projectId|indexStudyType|condition|datasourceId|datatypeId|diseaseFromSourceMappedId|resourceScore|targetFromSourceId|  id|score|sourceId|studyId|variantId|betaGwas|pValueExponent|name|therapeuticAreas|colocDoE|maxClinPhase|drugGoF_protect|drugLoF_protect|AgreeDrug|AgreeDrugDiagonal|biosampleName|\n",
      "+-----------+---------------+-------------+----------------+------------+-----------------+----------+--------------+--------------------------+----+--------------------+--------------------+----+----+----+----+----+-----------+-------------+---------------------+--------------+----------------------+-----------------+----------+---------+--------------+---------+------------+----------+-------------------------+-------------+------------------+----+-----+--------+-------+---------+--------+--------------+----+----------------+--------+------------+---------------+---------------+---------+-----------------+-------------+\n",
      "|       NULL|ENSG00000007314|  EFO_0000555|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              3|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000007314|  EFO_1000249|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         3.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000008018|  EFO_1000453|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000012504|MONDO_0019052|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         4.0|              1|           NULL|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000012779|MONDO_0004235|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         3.0|           NULL|              4|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000014138|  EFO_1001945|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         3.0|           NULL|              4|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000023228|  EFO_1000657|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              3|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000062822|  EFO_0004991|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000065883|  EFO_0001378|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         1.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000067191|  EFO_0000319|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         4.0|           NULL|              2|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000070886|  EFO_1000489|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000077782|MONDO_0008315|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              6|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000080293|  EFO_0001073|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|              1|           NULL|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000081248|MONDO_0004992|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         1.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000082175|  EFO_1000796|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|              1|           NULL|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000082556|  EFO_1000906|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000082556|MONDO_0004992|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         2.0|              1|           NULL|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000087085|MONDO_0005301|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         3.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000087586|MONDO_0002516|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         1.0|           NULL|              1|       no|               no|         NULL|\n",
      "|       NULL|ENSG00000088832|  EFO_0008514|            NULL|        NULL|             NULL|      NULL|          NULL|                      NULL|NULL|                NULL|                NULL|NULL|NULL|NULL|NULL|NULL|       NULL|         NULL|                 NULL|          NULL|                  NULL|             NULL|      NULL|     NULL|          NULL|     NULL|        NULL|      NULL|                     NULL|         NULL|              NULL|NULL| NULL|    NULL|   NULL|     NULL|    NULL|          NULL|NULL|            NULL|    NULL|         3.0|           NULL|              1|       no|               no|         NULL|\n",
      "+-----------+---------------+-------------+----------------+------------+-----------------+----------+--------------+--------------------------+----+--------------------+--------------------+----+----+----+----+----+-----------+-------------+---------------------+--------------+----------------------+-----------------+----------+---------+--------------+---------+------------+----------+-------------------------+-------------+------------------+----+-----+--------+-------+---------+--------+--------------+----+----------------+--------+------------+---------------+---------------+---------+-----------------+-------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "benchmark.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "114325"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "benchmark.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "74187"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "benchmark.groupBy('targetId','diseaseId','drugGoF_protect','drugLoF_protect').count().count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/07/10 21:23:50 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+----------------+-----+\n",
      "|coherencyDiagonal|coherencyOneCell|count|\n",
      "+-----------------+----------------+-----+\n",
      "|           noEvid|          noEvid|73699|\n",
      "|         coherent|        coherent|  240|\n",
      "|           dispar|          dispar|  225|\n",
      "|         coherent|          dispar|   23|\n",
      "+-----------------+----------------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "discrepancifier(benchmark.groupBy('targetId','diseaseId','drugGoF_protect','drugLoF_protect').pivot('colocDoE').count()).groupBy('coherencyDiagonal','coherencyOneCell').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "doe_columns=[\"LoF_protect\", \"GoF_risk\", \"LoF_risk\", \"GoF_protect\"]\n",
    "diagonal_lof=['LoF_protect','GoF_risk']\n",
    "diagonal_gof=['LoF_risk','GoF_protect']\n",
    "conditions = [\n",
    "    F.when(F.col(c) == F.col(\"maxDoE\"), F.lit(c)).otherwise(F.lit(None)) for c in doe_columns\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/07/10 21:30:25 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    }
   ],
   "source": [
    "test=discrepancifier(benchmark.groupBy('targetId','diseaseId','drugGoF_protect','drugLoF_protect').pivot('colocDoE').count()\n",
    "        ).withColumnRenamed('drugGoF_protect','GoF_protect_ch').withColumnRenamed('drugLoF_protect','LoF_protect_ch').withColumn(\n",
    "            \"arrayN\", F.array(*[F.col(c) for c in doe_columns])\n",
    "        ).withColumn(\n",
    "            \"maxDoE\", F.array_max(F.col(\"arrayN\"))\n",
    "        ).withColumn(\"maxDoE_names\", F.array(*conditions)\n",
    "        ).withColumn(\"maxDoE_names\", F.expr(\"filter(maxDoE_names, x -> x is not null)\")\n",
    "        ).withColumn(\n",
    "            \"maxDoEArrayN\",\n",
    "            F.expr(\"aggregate(arrayN, 0, (acc, x) -> acc + IF(x = maxDoE, 1, 0))\")\n",
    "        ).withColumn(\n",
    "            \"NoneCellYes\",\n",
    "            F.when((F.col(\"LoF_protect_ch\").isNotNull() & (F.col('GoF_protect_ch').isNull())) & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"LoF_protect\")))==True, F.lit('yes'))\n",
    "            .when((F.col(\"GoF_protect_ch\").isNotNull() & (F.col('LoF_protect_ch').isNull())) & (F.array_contains(F.col(\"maxDoE_names\"), F.lit(\"GoF_protect\")))==True, F.lit('yes')\n",
    "                ).otherwise(F.lit('no'))  # If the value is null, return null # Otherwise, check if name is in array\n",
    "        ).withColumn(\n",
    "            \"NdiagonalYes\",\n",
    "            F.when((F.col(\"LoF_protect_ch\").isNotNull() & (F.col('GoF_protect_ch').isNull())) & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_lof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).when((F.col(\"GoF_protect_ch\").isNotNull() & (F.col('LoF_protect_ch').isNull())) & \n",
    "                (F.size(F.array_intersect(F.col(\"maxDoE_names\"), F.array([F.lit(x) for x in diagonal_gof]))) > 0),\n",
    "                F.lit(\"yes\")\n",
    "            ).otherwise(F.lit('no'))\n",
    "        ).withColumn(\n",
    "            \"diagonalAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyDiagonal_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyDiagonal\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        F.col(\"GoF_risk\").isNotNull() | F.col(\"LoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"GoF_protect_ch\").isNotNull()\n",
    "                    & (\n",
    "                        F.col(\"LoF_risk\").isNotNull() | F.col(\"GoF_protect\").isNotNull()\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"oneCellAgreeWithDrugs\",\n",
    "            F.when(\n",
    "                (F.col(\"coherencyOneCell_ch\") == \"coherent\")\n",
    "                & (F.col(\"coherencyOneCell\") == \"coherent\"),\n",
    "                F.when(\n",
    "                    (F.col(\"LoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"LoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"GoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"GoF_protect_ch\").isNotNull())\n",
    "                    & (\n",
    "                        (F.col(\"GoF_protect\").isNotNull())\n",
    "                        & (F.col(\"LoF_risk\").isNull())\n",
    "                        & (F.col(\"LoF_protect\").isNull())\n",
    "                        & (F.col(\"GoF_risk\").isNull())\n",
    "                    ),\n",
    "                    F.lit(\"coherent\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"dispar\")),\n",
    "            ),\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 868:========================================>           (311 + 18) / 400]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+------------+-----+\n",
      "|coherencyDiagonal|NdiagonalYes|count|\n",
      "+-----------------+------------+-----+\n",
      "|           noEvid|          no|73699|\n",
      "|         coherent|         yes|  146|\n",
      "|           dispar|         yes|  125|\n",
      "|           dispar|          no|  100|\n",
      "|         coherent|          no|  117|\n",
      "+-----------------+------------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "test.groupBy('coherencyDiagonal','NdiagonalYes').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------+--------------+--------------+----+-----------+--------+-----------+--------+-----------+-----------------+----------------+--------------------+------+-------------+------------+-----------+------------+\n",
      "|       targetId|    diseaseId|GoF_protect_ch|LoF_protect_ch|null|GoF_protect|GoF_risk|LoF_protect|LoF_risk|noEvaluable|coherencyDiagonal|coherencyOneCell|              arrayN|maxDoE| maxDoE_names|maxDoEArrayN|NoneCellYes|NdiagonalYes|\n",
      "+---------------+-------------+--------------+--------------+----+-----------+--------+-----------+--------+-----------+-----------------+----------------+--------------------+------+-------------+------------+-----------+------------+\n",
      "|ENSG00000169083|  EFO_0001663|             4|            44|NULL|       NULL|       2|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, 2, NULL, N...|     2|   [GoF_risk]|           1|         no|          no|\n",
      "|ENSG00000079999|MONDO_0005301|          NULL|            24|NULL|          1|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     1|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000197635|  EFO_0003884|          NULL|             2|NULL|       NULL|    NULL|       NULL|       4|       NULL|         coherent|        coherent|[NULL, NULL, 4, N...|     4|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000022355|MONDO_0002050|            30|          NULL|NULL|       NULL|    NULL|          1|    NULL|       NULL|         coherent|        coherent|[1, NULL, NULL, N...|     1|[LoF_protect]|           1|         no|          no|\n",
      "|ENSG00000171303|  EFO_0000684|          NULL|             1|NULL|          6|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     6|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000171385|  EFO_0000275|          NULL|            18|NULL|         18|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|    18|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000136531|  EFO_0001073|          NULL|            22|NULL|       NULL|    NULL|       NULL|       2|       NULL|         coherent|        coherent|[NULL, NULL, 2, N...|     2|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000168065|  EFO_0001073|          NULL|             2|NULL|          8|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     8|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000164258|  EFO_0001663|          NULL|             5|NULL|       NULL|    NULL|       NULL|       2|       NULL|         coherent|        coherent|[NULL, NULL, 2, N...|     2|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000186642|  EFO_0000685|          NULL|             3|   3|          1|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     1|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000164418|MONDO_0002050|          NULL|             1|   8|          4|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     4|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000171557|   HP_0002140|          NULL|             1|   8|       NULL|    NULL|       NULL|       4|       NULL|         coherent|        coherent|[NULL, NULL, 4, N...|     4|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000146648|  EFO_0000272|          NULL|             1|   2|       NULL|    NULL|       NULL|       3|       NULL|         coherent|        coherent|[NULL, NULL, 3, N...|     3|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000014257|  EFO_0001663|          NULL|          NULL|NULL|          8|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     8|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000151067|MONDO_0005090|          NULL|             1|  45|       NULL|    NULL|       NULL|      18|       NULL|         coherent|        coherent|[NULL, NULL, 18, ...|    18|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000186716|  EFO_0003086|          NULL|             1|NULL|       NULL|    NULL|       NULL|       2|       NULL|         coherent|        coherent|[NULL, NULL, 2, N...|     2|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000171557|  EFO_0003777|          NULL|          NULL|   2|       NULL|    NULL|       NULL|       9|       NULL|         coherent|        coherent|[NULL, NULL, 9, N...|     9|   [LoF_risk]|           1|         no|          no|\n",
      "|ENSG00000148926|  EFO_0003144|             1|          NULL|NULL|       NULL|       2|          1|    NULL|       NULL|         coherent|          dispar|  [1, 2, NULL, NULL]|     2|   [GoF_risk]|           1|         no|          no|\n",
      "|ENSG00000134460|  EFO_0000400|          NULL|             1|NULL|          2|    NULL|       NULL|    NULL|       NULL|         coherent|        coherent|[NULL, NULL, NULL...|     2|[GoF_protect]|           1|         no|          no|\n",
      "|ENSG00000022355|MONDO_0002009|           123|          NULL|NULL|       NULL|    NULL|          1|    NULL|       NULL|         coherent|        coherent|[1, NULL, NULL, N...|     1|[LoF_protect]|           1|         no|          no|\n",
      "+---------------+-------------+--------------+--------------+----+-----------+--------+-----------+--------+-----------+-----------------+----------------+--------------------+------+-------------+------------+-----------+------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "test.filter((F.col('coherencyDiagonal')=='coherent') & (F.col('NdiagonalYes')=='no')).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 76:===================================================>  (384 + 8) / 400]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+---------+-----+\n",
      "|AgreeDrugDiagonal|AgreeDrug|count|\n",
      "+-----------------+---------+-----+\n",
      "|               no|       no|96389|\n",
      "|              yes|      yes| 9563|\n",
      "|              yes|       no| 8373|\n",
      "+-----------------+---------+-----+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "benchmark.groupBy('AgreeDrugDiagonal','AgreeDrug').count().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PySpark",
   "language": "python",
   "name": "pyspark"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
