{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql import SparkSession, Window\n",
    "import pyspark.sql.functions as F\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check at evidence level the performance of direction of effect on back-end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Read written evidence file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "#### load the files from Ricardo to check ETL evidences:\n",
    "\n",
    "### pay attention to: gene burden (all are LoF)\n",
    "### oncogene/tumor suppressor from intogen\n",
    "\n",
    "ricardo3 = spark.read.parquet(\n",
    "    \"gs://open-targets-pre-data-releases/ricardo/localrun/evidence/\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In order to compare the results, we are goind to run our internal pipeline on the same evidence files. For that, we need to rename all columns that are created to avoid clashing.\n",
    "    - Make a column list that we do not want to change the name (_R)\n",
    "    - Run the rename over the evidence files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_to_drop = [\n",
    "    \"id\",\n",
    "    \"score\",\n",
    "    \"target_id\",\n",
    "    \"approvedSymbol\",\n",
    "    \"description\",\n",
    "    \"description_splited\",\n",
    "    \"TSorOncogene\",\n",
    "    \"stMethod\",\n",
    "    \"counts\",\n",
    "    \"whatToDo\",\n",
    "    \"targetId2\",\n",
    "    \"drugId2\",\n",
    "    \"actionType\",\n",
    "    \"mechanismOfAction\",\n",
    "    \"variantEffect\",\n",
    "    \"directionOnTrait\",\n",
    "    \"homogenizedVersion\",\n",
    "    \"sourceId\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### rename evidences to avoid clashing name columns\n",
    "newCols = []\n",
    "for word in ricardo3.columns:\n",
    "    if word in list_to_drop:\n",
    "        newCols.append(word + \"_R\")\n",
    "    else:\n",
    "        newCols.append(word)\n",
    "evidences2 = ricardo3.toDF(*newCols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Direction of effect function "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 22.11.2023\n",
    "\n",
    "# 1# defining datasets and hardcoding some variables\n",
    "### external datasets:\n",
    "target_path = \"gs://open-targets-data-releases/23.09/output/etl/parquet/targets/\"\n",
    "target = spark.read.parquet(target_path)\n",
    "mecact_path = (\n",
    "    \"gs://open-targets-data-releases/23.09/output/etl/parquet/mechanismOfAction/\"\n",
    ")\n",
    "mecact = spark.read.parquet(mecact_path)\n",
    "\n",
    "var_filter_lof = [\n",
    "    ### High impact variants https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html\n",
    "    \"SO_0001589\",  ## frameshit_variant\n",
    "    \"SO_0001587\",  ## stop_gained\n",
    "    \"SO_0001574\",  ## splice_acceptor_variant\n",
    "    \"SO_0001575\",  ## splice_donor_variant\n",
    "    \"SO_0002012\",  ## start_lost\n",
    "    \"SO_0001578\",  ## stop_lost\n",
    "    \"SO_0001893\",  ## transcript_ablation\n",
    "]\n",
    "\n",
    "gof = [\"SO_0002053\"]\n",
    "lof = [\"SO_0002054\"]\n",
    "\n",
    "\n",
    "## annotate TSG/oncogene/bivalent using 'hallmarks.attributes'\n",
    "oncotsg_list = [\n",
    "    \"TSG\",\n",
    "    \"oncogene\",\n",
    "    \"Oncogene\",\n",
    "    \"oncogene\",\n",
    "    \"oncogene,TSG\",\n",
    "    \"TSG,oncogene\",\n",
    "    \"fusion,oncogene\",\n",
    "    \"oncogene,fusion\",\n",
    "]\n",
    "\n",
    "### define inhibitors and activators\n",
    "inhibitors = [\n",
    "    \"RNAI INHIBITOR\",\n",
    "    \"NEGATIVE MODULATOR\",\n",
    "    \"NEGATIVE ALLOSTERIC MODULATOR\",\n",
    "    \"ANTAGONIST\",\n",
    "    \"ANTISENSE INHIBITOR\",\n",
    "    \"BLOCKER\",\n",
    "    \"INHIBITOR\",\n",
    "    \"DEGRADER\",\n",
    "    \"INVERSE AGONIST\",\n",
    "    \"ALLOSTERIC ANTAGONIST\",\n",
    "    \"DISRUPTING AGENT\",\n",
    "]\n",
    "\n",
    "activators = [\n",
    "    \"PARTIAL AGONIST\",\n",
    "    \"ACTIVATOR\",\n",
    "    \"POSITIVE ALLOSTERIC MODULATOR\",\n",
    "    \"POSITIVE MODULATOR\",\n",
    "    \"AGONIST\",\n",
    "    \"SEQUESTERING AGENT\",\n",
    "    \"STABILISER\",\n",
    "]\n",
    "\n",
    "actionType = (\n",
    "    mecact.select(\n",
    "        F.explode_outer(\"chemblIds\").alias(\"drugId2\"),\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "        \"targets\",\n",
    "    )\n",
    "    .select(\n",
    "        F.explode_outer(\"targets\").alias(\"targetId2\"),\n",
    "        \"drugId2\",\n",
    "        \"actionType\",\n",
    "        \"mechanismOfAction\",\n",
    "    )\n",
    "    .groupBy(\"targetId2\", \"drugId2\")\n",
    "    .agg(\n",
    "        F.collect_set(\"actionType\").alias(\"actionType\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "oncolabel = (\n",
    "    target.select(\n",
    "        \"id\", \"approvedSymbol\", F.explode_outer(F.col(\"hallmarks.attributes\"))\n",
    "    )\n",
    "    .select(\"id\", \"approvedSymbol\", \"col.description\")\n",
    "    .filter(F.col(\"description\").isin(oncotsg_list))\n",
    "    .groupBy(\"id\", \"approvedSymbol\")\n",
    "    .agg(F.collect_set(\"description\").alias(\"description\"))\n",
    "    .withColumn(\"description_splited\", F.concat_ws(\",\", F.col(\"description\")))\n",
    "    .withColumn(\n",
    "        \"TSorOncogene\",\n",
    "        F.when(\n",
    "            (\n",
    "                F.col(\"description_splited\").rlike(\"ncogene\")\n",
    "                & F.col(\"description_splited\").rlike(\"TSG\")\n",
    "            ),\n",
    "            F.lit(\"bivalent\"),\n",
    "        )\n",
    "        .when(F.col(\"description_splited\").rlike(\"ncogene(\\s|$)\"), F.lit(\"oncogene\"))\n",
    "        .when(F.col(\"description_splited\").rlike(\"TSG(\\s|$)\"), F.lit(\"TSG\"))\n",
    "        .otherwise(F.lit(\"noEvaluable\")),\n",
    "    )\n",
    "    .withColumnRenamed(\"id\", \"target_id\")\n",
    ")\n",
    "\n",
    "# 2# run the transformation of the evidences datasets used.\n",
    "\n",
    "\n",
    "def new_directionOfEffect(\n",
    "    evidences,\n",
    "    oncolabel,\n",
    "    actionType,\n",
    "    var_filter_lof,\n",
    "    gof,\n",
    "    inhibitors,\n",
    "    activators,\n",
    "):\n",
    "    windowSpec = Window.partitionBy(\"targetId\", \"diseaseId\")\n",
    "    doe = (\n",
    "        evidences.withColumn(\n",
    "            \"beta\", F.col(\"beta\").cast(\"float\")\n",
    "        )  ## ot genetics & gene burden\n",
    "        .withColumn(\n",
    "            \"OddsRatio\", F.col(\"OddsRatio\").cast(\"float\")\n",
    "        )  ## ot genetics & gene burden\n",
    "        .withColumn(\n",
    "            \"clinicalSignificances\", F.concat_ws(\",\", F.col(\"clinicalSignificances\"))\n",
    "        )  ### eva\n",
    "        .join(oncolabel, oncolabel.target_id == F.col(\"targetId\"), \"left\")  ###  cgc\n",
    "        .join(\n",
    "            actionType,  ## chembl\n",
    "            (actionType.drugId2 == F.col(\"drugId\"))\n",
    "            & (actionType.targetId2 == F.col(\"targetId\")),\n",
    "            \"left\",\n",
    "        )\n",
    "        .withColumn(\"inhibitors_list\", F.array([F.lit(i) for i in inhibitors]))\n",
    "        .withColumn(\"activators_list\", F.array([F.lit(i) for i in activators]))\n",
    "        .withColumn(\n",
    "            \"intogen_function\",\n",
    "            F.when(\n",
    "                F.arrays_overlap(\n",
    "                    F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                    F.array([F.lit(i) for i in (gof)]),\n",
    "                ),\n",
    "                F.lit(\"GoF\"),\n",
    "            ).when(\n",
    "                F.arrays_overlap(\n",
    "                    F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                    F.array([F.lit(i) for i in (lof)]),\n",
    "                ),\n",
    "                F.lit(\"LoF\"),\n",
    "            )\n",
    "            # .otherwise(\"nodata\"),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"intogenAnnot\",\n",
    "            F.size(F.collect_set(F.col(\"intogen_function\")).over(windowSpec)),\n",
    "        )\n",
    "        ### variant Effect Column\n",
    "        .withColumn(\n",
    "            \"variantEffect\",\n",
    "            F.when(\n",
    "                F.col(\"datasourceId\") == \"ot_genetics_portal\",\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\").isNotNull(),\n",
    "                    F.when(\n",
    "                        F.col(\"variantFunctionalConsequenceFromQtlId\").isNull(),\n",
    "                        F.when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                                var_filter_lof\n",
    "                            ),\n",
    "                            F.lit(\"LoF\"),\n",
    "                        )\n",
    "                        .when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(gof),\n",
    "                            F.lit(\"GoF\"),\n",
    "                        )\n",
    "                        .otherwise(F.lit(\"noEvaluable\")),\n",
    "                    )\n",
    "                    ### variantFunctionalConsequenceFromQtlId\n",
    "                    .when(\n",
    "                        F.col(\"variantFunctionalConsequenceFromQtlId\").isNotNull(),\n",
    "                        F.when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                                var_filter_lof\n",
    "                            ),  ## when is a LoF variant\n",
    "                            F.when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002316\",\n",
    "                                F.lit(\"LoF\"),\n",
    "                            )\n",
    "                            .when(\n",
    "                                F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                == \"SO_0002315\",\n",
    "                                F.lit(\"conflict/noEvaluable\"),\n",
    "                            )\n",
    "                            .otherwise(F.lit(\"LoF\")),\n",
    "                        ).when(\n",
    "                            F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof)\n",
    "                            == False,  ## when is not a LoF, still can be a GoF\n",
    "                            F.when(\n",
    "                                F.col(\"variantFunctionalConsequenceId\").isin(gof)\n",
    "                                == False,  ##if not GoF\n",
    "                                F.when(\n",
    "                                    F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                    == \"SO_0002316\",\n",
    "                                    F.lit(\"LoF\"),\n",
    "                                )\n",
    "                                .when(\n",
    "                                    F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                    == \"SO_0002315\",\n",
    "                                    F.lit(\"GoF\"),\n",
    "                                )\n",
    "                                .otherwise(F.lit(\"noEvaluable\")),\n",
    "                            ).when(\n",
    "                                F.col(\"variantFunctionalConsequenceId\").isin(\n",
    "                                    gof\n",
    "                                ),  ##if is GoF\n",
    "                                F.when(\n",
    "                                    F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                    == \"SO_0002316\",\n",
    "                                    F.lit(\"conflict/noEvaluable\"),\n",
    "                                ).when(\n",
    "                                    F.col(\"variantFunctionalConsequenceFromQtlId\")\n",
    "                                    == \"SO_0002315\",\n",
    "                                    F.lit(\"GoF\"),\n",
    "                                ),\n",
    "                            ),\n",
    "                        ),\n",
    "                    ),\n",
    "                ).when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\").isNull(),\n",
    "                    F.when(\n",
    "                        F.col(\"variantFunctionalConsequenceFromQtlId\") == \"SO_0002316\",\n",
    "                        F.lit(\"LoF\"),\n",
    "                    )\n",
    "                    .when(\n",
    "                        F.col(\"variantFunctionalConsequenceFromQtlId\") == \"SO_0002315\",\n",
    "                        F.lit(\"GoF\"),\n",
    "                    )\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                ),\n",
    "            ).when(\n",
    "                F.col(\"datasourceId\") == \"gene_burden\",\n",
    "                F.when(F.col(\"targetId\").isNotNull(), F.lit(\"LoF\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            #### Eva_germline\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"eva\",\n",
    "                #### .filter(F.col('variantFunctionalConsequenceId').isin(var_filter_lof))\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                    F.lit(\"LoF\"),\n",
    "                ).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),  ### Son todas aquellas que tenen info pero no son LoF\n",
    "            )\n",
    "            #### Eva_somatic\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"eva_somatic\",\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\").isin(var_filter_lof),\n",
    "                    F.lit(\"LoF\"),\n",
    "                ).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "            )\n",
    "            #### G2P\n",
    "            .when(\n",
    "                F.col(\"datasourceId\")\n",
    "                == \"gene2phenotype\",  ### 6 types of variants [SO_0002318, SO_0002317, SO_0001622, SO_0002315, SO_0001566, SO_0002220]\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\") == \"SO_0002317\",\n",
    "                    F.lit(\"LoF\"),\n",
    "                )  ### absent gene product\n",
    "                .when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\") == \"SO_0002315\",\n",
    "                    F.lit(\"GoF\"),\n",
    "                )  ### increased gene product level\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            #### Orphanet\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"orphanet\",\n",
    "                F.when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\") == \"SO_0002054\",\n",
    "                    F.lit(\"LoF\"),\n",
    "                )  ### Loss of Function Variant\n",
    "                .when(\n",
    "                    F.col(\"variantFunctionalConsequenceId\") == \"SO_0002053\",\n",
    "                    F.lit(\"GoF\"),\n",
    "                )  ### Gain_of_Function Variant\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            #### CGC\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"cancer_gene_census\",\n",
    "                F.when(F.col(\"TSorOncogene\") == \"oncogene\", F.lit(\"GoF\"))\n",
    "                .when(F.col(\"TSorOncogene\") == \"TSG\", F.lit(\"LoF\"))\n",
    "                .when(F.col(\"TSorOncogene\") == \"bivalent\", F.lit(\"bivalent\"))\n",
    "                .otherwise(\"noEvaluable\"),\n",
    "            )\n",
    "            #### intogen\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"intogen\",\n",
    "                F.when(\n",
    "                    F.col(\"intogenAnnot\")\n",
    "                    == 1,  ## oncogene/tummor suppressor for a given trait\n",
    "                    F.when(\n",
    "                        F.arrays_overlap(\n",
    "                            F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                            F.array([F.lit(i) for i in (gof)]),\n",
    "                        ),\n",
    "                        F.lit(\"GoF\"),\n",
    "                    ).when(\n",
    "                        F.arrays_overlap(\n",
    "                            F.col(\"mutatedSamples.functionalConsequenceId\"),\n",
    "                            F.array([F.lit(i) for i in (lof)]),\n",
    "                        ),\n",
    "                        F.lit(\"LoF\"),\n",
    "                    ),\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"intogenAnnot\") > 1, F.lit(\"bivalentIntogen\")\n",
    "                )  ##oncogene & tumor suppressor for a given trait\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            )\n",
    "            #### impc\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"impc\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"LoF\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            ### chembl\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"chembl\",\n",
    "                F.when(\n",
    "                    F.size(\n",
    "                        F.array_intersect(F.col(\"actionType\"), F.col(\"inhibitors_list\"))\n",
    "                    )\n",
    "                    >= 1,\n",
    "                    F.lit(\"LoF\"),\n",
    "                )\n",
    "                .when(\n",
    "                    F.size(\n",
    "                        F.array_intersect(F.col(\"actionType\"), F.col(\"activators_list\"))\n",
    "                    )\n",
    "                    >= 1,\n",
    "                    F.lit(\"GoF\"),\n",
    "                )\n",
    "                .otherwise(F.lit(\"noEvaluable\")),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"directionOnTrait\",\n",
    "            ## ot genetics portal\n",
    "            F.when(\n",
    "                F.col(\"datasourceId\")\n",
    "                == \"ot_genetics_portal\",  ### the same for gene_burden\n",
    "                F.when(\n",
    "                    (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                    F.when(F.col(\"beta\") > 0, F.lit(\"risk\"))\n",
    "                    .when(F.col(\"beta\") < 0, F.lit(\"protect\"))\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                    F.when(F.col(\"OddsRatio\") > 1, F.lit(\"risk\"))\n",
    "                    .when(F.col(\"OddsRatio\") < 1, F.lit(\"protect\"))\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"beta\").isNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                    F.lit(\"noEvaluable\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNotNull()),\n",
    "                    F.lit(\"conflict/noEvaluable\"),\n",
    "                ),\n",
    "            ).when(\n",
    "                F.col(\"datasourceId\") == \"gene_burden\",\n",
    "                F.when(\n",
    "                    (F.col(\"beta\").isNotNull()) & (F.col(\"OddsRatio\").isNull()),\n",
    "                    F.when(F.col(\"beta\") > 0, F.lit(\"risk\"))\n",
    "                    .when(F.col(\"beta\") < 0, F.lit(\"protect\"))\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"oddsRatio\").isNotNull()) & (F.col(\"beta\").isNull()),\n",
    "                    F.when(F.col(\"oddsRatio\") > 1, F.lit(\"risk\"))\n",
    "                    .when(F.col(\"oddsRatio\") < 1, F.lit(\"protect\"))\n",
    "                    .otherwise(F.lit(\"noEvaluable\")),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"beta\").isNull()) & (F.col(\"oddsRatio\").isNull()),\n",
    "                    F.lit(\"noEvaluable\"),\n",
    "                )\n",
    "                .when(\n",
    "                    (F.col(\"beta\").isNotNull()) & (F.col(\"oddsRatio\").isNotNull()),\n",
    "                    F.lit(\"conflict\"),\n",
    "                ),\n",
    "            )\n",
    "            ## Eva_germline\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"eva\",  ### the same for eva_somatic\n",
    "                F.when(\n",
    "                    F.col(\"clinicalSignificances\").rlike(\"(pathogenic)$\"), F.lit(\"risk\")\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"clinicalSignificances\").contains(\"protect\"), F.lit(\"protect\")\n",
    "                )\n",
    "                .otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "            )\n",
    "            #### Eva_somatic\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"eva_somatic\",\n",
    "                F.when(\n",
    "                    F.col(\"clinicalSignificances\").rlike(\"(pathogenic)$\"), F.lit(\"risk\")\n",
    "                )\n",
    "                .when(\n",
    "                    F.col(\"clinicalSignificances\").contains(\"protect\"), F.lit(\"protect\")\n",
    "                )\n",
    "                .otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),  ### Son todas aquellas que tenen info pero no son patogenicas/protective  + LoF\n",
    "            )\n",
    "            #### G2P\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"gene2phenotype\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            #### Orphanet\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"orphanet\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            #### CGC\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"cancer_gene_census\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            #### intogen\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"intogen\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            #### impc\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"impc\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"risk\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            )\n",
    "            ### chembl\n",
    "            .when(\n",
    "                F.col(\"datasourceId\") == \"chembl\",\n",
    "                F.when(F.col(\"diseaseId\").isNotNull(), F.lit(\"protect\")).otherwise(\n",
    "                    F.lit(\"noEvaluable\")\n",
    "                ),\n",
    "            ),\n",
    "        )\n",
    "        .withColumn(\n",
    "            \"homogenizedVersion_J\",\n",
    "            F.when(\n",
    "                (F.col(\"variantEffect\") == \"LoF\")\n",
    "                & (F.col(\"directionOnTrait\") == \"risk\"),\n",
    "                F.lit(\"LoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"variantEffect\") == \"LoF\")\n",
    "                & (F.col(\"directionOnTrait\") == \"protect\"),\n",
    "                F.lit(\"LoF_protect\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"variantEffect\") == \"GoF\")\n",
    "                & (F.col(\"directionOnTrait\") == \"risk\"),\n",
    "                F.lit(\"GoF_risk\"),\n",
    "            )\n",
    "            .when(\n",
    "                (F.col(\"variantEffect\") == \"GoF\")\n",
    "                & (F.col(\"directionOnTrait\") == \"protect\"),\n",
    "                F.lit(\"GoF_protect\"),\n",
    "            )\n",
    "            .otherwise(F.lit(\"noEvaluable\")),\n",
    "        )\n",
    "    )\n",
    "    return doe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- run direction of effect function over the generated files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "evidences_JR = new_directionOfEffect(\n",
    "    evidences2,\n",
    "    oncolabel,\n",
    "    actionType,\n",
    "    var_filter_lof,\n",
    "    gof,\n",
    "    inhibitors,\n",
    "    activators,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Compare internal column assessments with BackEnd columns (_R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "evidences_JR.groupBy(\n",
    "    \"datasourceId\",\n",
    "    \"variantEffect_R\",\n",
    "    \"variantEffect\",\n",
    "    \"directionOnTrait_R\",\n",
    "    \"directionOnTrait\",\n",
    ").count().toPandas().to_csv(\"gs://ot-team/jroldan/checkingETL_20231122.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PySpark",
   "language": "python",
   "name": "pyspark"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
